{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/eichhorn-lab/PTMmap2structure/blob/main/PTMmap2structure_v1_4.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3wdSW0izB329"
      },
      "source": [
        "# **PTM map to structure notebook**\n",
        "## *version 1.4, last updated 08/06/2024*\n",
        "####*created by CD Eichhorn, 2024*\n",
        "---\n",
        "\n",
        "**Instructions:**\n",
        "- To run, click Runtime tab, then run all.\n",
        "- Go to \"Import CSV file\" section and upload csv file when upload button appears.\n",
        "- Accession numbers must be UniProt accessions.\n",
        "- If you run AF or Colabfold to generate additional pdb files, create a folder to output/libraries directory named colabfold_pdbs and upload pdb coordinate files with the format accessionnumber.pdb.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qGh2X5iDec6z"
      },
      "source": [
        "# **Install packages and upload csv file**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "T8WI8jRC2MQV",
        "outputId": "95b842fd-f544-49ec-bb2a-a0fba985958d"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Collecting biopython\n",
            "  Downloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (12 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython) (1.26.4)\n",
            "Downloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m3.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython\n",
            "Successfully installed biopython-1.84\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (4.66.4)\n",
            "PREFIX=/usr/local\n",
            "Unpacking payload ...\n",
            "Collecting package metadata (current_repodata.json): - \b\b\\ \b\bdone\n",
            "Solving environment: / \b\b- \b\b\\ \b\bdone\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs:\n",
            "    - _libgcc_mutex==0.1=main\n",
            "    - ca-certificates==2020.1.1=0\n",
            "    - certifi==2020.4.5.1=py37_0\n",
            "    - cffi==1.14.0=py37he30daa8_1\n",
            "    - chardet==3.0.4=py37_1003\n",
            "    - conda-package-handling==1.6.1=py37h7b6447c_0\n",
            "    - conda==4.8.3=py37_0\n",
            "    - cryptography==2.9.2=py37h1ba5d50_0\n",
            "    - idna==2.9=py_1\n",
            "    - ld_impl_linux-64==2.33.1=h53a641e_7\n",
            "    - libedit==3.1.20181209=hc058e9b_0\n",
            "    - libffi==3.3=he6710b0_1\n",
            "    - libgcc-ng==9.1.0=hdf63c60_0\n",
            "    - libstdcxx-ng==9.1.0=hdf63c60_0\n",
            "    - ncurses==6.2=he6710b0_1\n",
            "    - openssl==1.1.1g=h7b6447c_0\n",
            "    - pip==20.0.2=py37_3\n",
            "    - pycosat==0.6.3=py37h7b6447c_0\n",
            "    - pycparser==2.20=py_0\n",
            "    - pyopenssl==19.1.0=py37_0\n",
            "    - pysocks==1.7.1=py37_0\n",
            "    - python==3.7.7=hcff3b4d_5\n",
            "    - readline==8.0=h7b6447c_0\n",
            "    - requests==2.23.0=py37_0\n",
            "    - ruamel_yaml==0.15.87=py37h7b6447c_0\n",
            "    - setuptools==46.4.0=py37_0\n",
            "    - six==1.14.0=py37_0\n",
            "    - sqlite==3.31.1=h62c20be_1\n",
            "    - tk==8.6.8=hbc83047_0\n",
            "    - tqdm==4.46.0=py_0\n",
            "    - urllib3==1.25.8=py37_0\n",
            "    - wheel==0.34.2=py37_0\n",
            "    - xz==5.2.5=h7b6447c_0\n",
            "    - yaml==0.1.7=had09818_2\n",
            "    - zlib==1.2.11=h7b6447c_3\n",
            "\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main\n",
            "  ca-certificates    pkgs/main/linux-64::ca-certificates-2020.1.1-0\n",
            "  certifi            pkgs/main/linux-64::certifi-2020.4.5.1-py37_0\n",
            "  cffi               pkgs/main/linux-64::cffi-1.14.0-py37he30daa8_1\n",
            "  chardet            pkgs/main/linux-64::chardet-3.0.4-py37_1003\n",
            "  conda              pkgs/main/linux-64::conda-4.8.3-py37_0\n",
            "  conda-package-han~ pkgs/main/linux-64::conda-package-handling-1.6.1-py37h7b6447c_0\n",
            "  cryptography       pkgs/main/linux-64::cryptography-2.9.2-py37h1ba5d50_0\n",
            "  idna               pkgs/main/noarch::idna-2.9-py_1\n",
            "  ld_impl_linux-64   pkgs/main/linux-64::ld_impl_linux-64-2.33.1-h53a641e_7\n",
            "  libedit            pkgs/main/linux-64::libedit-3.1.20181209-hc058e9b_0\n",
            "  libffi             pkgs/main/linux-64::libffi-3.3-he6710b0_1\n",
            "  libgcc-ng          pkgs/main/linux-64::libgcc-ng-9.1.0-hdf63c60_0\n",
            "  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-9.1.0-hdf63c60_0\n",
            "  ncurses            pkgs/main/linux-64::ncurses-6.2-he6710b0_1\n",
            "  openssl            pkgs/main/linux-64::openssl-1.1.1g-h7b6447c_0\n",
            "  pip                pkgs/main/linux-64::pip-20.0.2-py37_3\n",
            "  pycosat            pkgs/main/linux-64::pycosat-0.6.3-py37h7b6447c_0\n",
            "  pycparser          pkgs/main/noarch::pycparser-2.20-py_0\n",
            "  pyopenssl          pkgs/main/linux-64::pyopenssl-19.1.0-py37_0\n",
            "  pysocks            pkgs/main/linux-64::pysocks-1.7.1-py37_0\n",
            "  python             pkgs/main/linux-64::python-3.7.7-hcff3b4d_5\n",
            "  readline           pkgs/main/linux-64::readline-8.0-h7b6447c_0\n",
            "  requests           pkgs/main/linux-64::requests-2.23.0-py37_0\n",
            "  ruamel_yaml        pkgs/main/linux-64::ruamel_yaml-0.15.87-py37h7b6447c_0\n",
            "  setuptools         pkgs/main/linux-64::setuptools-46.4.0-py37_0\n",
            "  six                pkgs/main/linux-64::six-1.14.0-py37_0\n",
            "  sqlite             pkgs/main/linux-64::sqlite-3.31.1-h62c20be_1\n",
            "  tk                 pkgs/main/linux-64::tk-8.6.8-hbc83047_0\n",
            "  tqdm               pkgs/main/noarch::tqdm-4.46.0-py_0\n",
            "  urllib3            pkgs/main/linux-64::urllib3-1.25.8-py37_0\n",
            "  wheel              pkgs/main/linux-64::wheel-0.34.2-py37_0\n",
            "  xz                 pkgs/main/linux-64::xz-5.2.5-h7b6447c_0\n",
            "  yaml               pkgs/main/linux-64::yaml-0.1.7-had09818_2\n",
            "  zlib               pkgs/main/linux-64::zlib-1.2.11-h7b6447c_3\n",
            "\n",
            "\n",
            "Preparing transaction: / \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "Executing transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "installation finished.\n",
            "WARNING:\n",
            "    You currently have a PYTHONPATH environment variable set. This may cause\n",
            "    unexpected behavior when running the Python interpreter in Miniconda3.\n",
            "    For best results, please verify that your PYTHONPATH only points to\n",
            "    directories of packages that are compatible with the Python interpreter\n",
            "    in Miniconda3: /usr/local\n",
            "Collecting package metadata (current_repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Solving environment: \\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.8.3\n",
            "  latest version: 24.7.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs:\n",
            "    - dssp\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    bzip2-1.0.8                |       h7b6447c_0          78 KB\n",
            "    ca-certificates-2024.7.2   |       h06a4308_0         127 KB\n",
            "    certifi-2020.6.20          |     pyhd3eb1b0_3         155 KB\n",
            "    dssp-3.0.0                 |       hf484d3e_3         2.2 MB  salilab\n",
            "    icu-58.2                   |       he6710b0_3        10.5 MB\n",
            "    libboost-1.67.0            |       h46d08c1_4        13.0 MB\n",
            "    openssl-1.1.1w             |       h7f8727e_0         3.7 MB\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:        29.8 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  bzip2              pkgs/main/linux-64::bzip2-1.0.8-h7b6447c_0\n",
            "  dssp               salilab/linux-64::dssp-3.0.0-hf484d3e_3\n",
            "  icu                pkgs/main/linux-64::icu-58.2-he6710b0_3\n",
            "  libboost           pkgs/main/linux-64::libboost-1.67.0-h46d08c1_4\n",
            "\n",
            "The following packages will be UPDATED:\n",
            "\n",
            "  ca-certificates                                2020.1.1-0 --> 2024.7.2-h06a4308_0\n",
            "  certifi            pkgs/main/linux-64::certifi-2020.4.5.~ --> pkgs/main/noarch::certifi-2020.6.20-pyhd3eb1b0_3\n",
            "  openssl                                 1.1.1g-h7b6447c_0 --> 1.1.1w-h7f8727e_0\n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages\n",
            "bzip2-1.0.8          | 78 KB     | : 100% 1.0/1 [00:00<00:00,  9.97it/s]\n",
            "openssl-1.1.1w       | 3.7 MB    | : 100% 1.0/1 [00:00<00:00,  2.96it/s]               \n",
            "icu-58.2             | 10.5 MB   | : 100% 1.0/1 [00:00<00:00,  1.65it/s]               \n",
            "libboost-1.67.0      | 13.0 MB   | : 100% 1.0/1 [00:03<00:00,  3.10s/it]               \n",
            "dssp-3.0.0           | 2.2 MB    | : 100% 1.0/1 [00:04<00:00,  4.19s/it]\n",
            "ca-certificates-2024 | 127 KB    | : 100% 1.0/1 [00:00<00:00,  6.48it/s]\n",
            "certifi-2020.6.20    | 155 KB    | : 100% 1.0/1 [00:00<00:00,  6.61it/s]\n",
            "Preparing transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Verifying transaction: \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Executing transaction: \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n",
            "/bin/bash: line 1: dssp: command not found\n"
          ]
        }
      ],
      "source": [
        "#@title Install packages\n",
        "\n",
        "# install biopython, tdqm for progress bar\n",
        "!pip install biopython\n",
        "!pip install tqdm\n",
        "\n",
        "# Install Miniconda\n",
        "!wget -qnc https://repo.anaconda.com/miniconda/Miniconda3-py37_4.8.3-Linux-x86_64.sh\n",
        "!bash Miniconda3-py37_4.8.3-Linux-x86_64.sh -bfp /usr/local\n",
        "!rm Miniconda3-py37_4.8.3-Linux-x86_64.sh\n",
        "\n",
        "# Add conda executable to PATH\n",
        "import sys\n",
        "sys.path.append('/usr/local/lib/python3.7/site-packages')\n",
        "\n",
        "# Install DSSP using conda\n",
        "!conda install -y -c salilab dssp\n",
        "\n",
        "# Test DSSP installation\n",
        "!dssp -h"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "vKNtuoHgLYZY"
      },
      "outputs": [],
      "source": [
        "#@title Import packages\n",
        "import csv\n",
        "import os\n",
        "import re\n",
        "import matplotlib.font_manager\n",
        "import matplotlib.gridspec as gridspec\n",
        "import matplotlib.pyplot as plt\n",
        "import matplotlib.patches as patches\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import requests\n",
        "import seaborn as sns  # Import seaborn for plotting\n",
        "import shutil\n",
        "import zipfile\n",
        "\n",
        "from Bio.PDB import PDBParser\n",
        "from Bio.PDB.DSSP import DSSP\n",
        "from concurrent.futures import ThreadPoolExecutor, as_completed\n",
        "from google.colab import files as colab_files\n",
        "from google.colab import files\n",
        "from matplotlib.font_manager import FontProperties\n",
        "from matplotlib.patches import Rectangle\n",
        "from matplotlib.lines import Line2D\n",
        "from matplotlib.patches import Patch\n",
        "from statsmodels.stats.weightstats import ttest_ind\n",
        "from tqdm import tqdm\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 336
        },
        "id": "bZF9tnpDzc66",
        "outputId": "6947b304-a304-424a-ec19-dddb5c7ab7b2"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Please upload the data file:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "\n",
              "     <input type=\"file\" id=\"files-438af9b1-837e-4503-bce9-98eac5eba478\" name=\"files[]\" multiple disabled\n",
              "        style=\"border:none\" />\n",
              "     <output id=\"result-438af9b1-837e-4503-bce9-98eac5eba478\">\n",
              "      Upload widget is only available when the cell has been executed in the\n",
              "      current browser session. Please rerun this cell to enable.\n",
              "      </output>\n",
              "      <script>// Copyright 2017 Google LLC\n",
              "//\n",
              "// Licensed under the Apache License, Version 2.0 (the \"License\");\n",
              "// you may not use this file except in compliance with the License.\n",
              "// You may obtain a copy of the License at\n",
              "//\n",
              "//      http://www.apache.org/licenses/LICENSE-2.0\n",
              "//\n",
              "// Unless required by applicable law or agreed to in writing, software\n",
              "// distributed under the License is distributed on an \"AS IS\" BASIS,\n",
              "// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
              "// See the License for the specific language governing permissions and\n",
              "// limitations under the License.\n",
              "\n",
              "/**\n",
              " * @fileoverview Helpers for google.colab Python module.\n",
              " */\n",
              "(function(scope) {\n",
              "function span(text, styleAttributes = {}) {\n",
              "  const element = document.createElement('span');\n",
              "  element.textContent = text;\n",
              "  for (const key of Object.keys(styleAttributes)) {\n",
              "    element.style[key] = styleAttributes[key];\n",
              "  }\n",
              "  return element;\n",
              "}\n",
              "\n",
              "// Max number of bytes which will be uploaded at a time.\n",
              "const MAX_PAYLOAD_SIZE = 100 * 1024;\n",
              "\n",
              "function _uploadFiles(inputId, outputId) {\n",
              "  const steps = uploadFilesStep(inputId, outputId);\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  // Cache steps on the outputElement to make it available for the next call\n",
              "  // to uploadFilesContinue from Python.\n",
              "  outputElement.steps = steps;\n",
              "\n",
              "  return _uploadFilesContinue(outputId);\n",
              "}\n",
              "\n",
              "// This is roughly an async generator (not supported in the browser yet),\n",
              "// where there are multiple asynchronous steps and the Python side is going\n",
              "// to poll for completion of each step.\n",
              "// This uses a Promise to block the python side on completion of each step,\n",
              "// then passes the result of the previous step as the input to the next step.\n",
              "function _uploadFilesContinue(outputId) {\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  const steps = outputElement.steps;\n",
              "\n",
              "  const next = steps.next(outputElement.lastPromiseValue);\n",
              "  return Promise.resolve(next.value.promise).then((value) => {\n",
              "    // Cache the last promise value to make it available to the next\n",
              "    // step of the generator.\n",
              "    outputElement.lastPromiseValue = value;\n",
              "    return next.value.response;\n",
              "  });\n",
              "}\n",
              "\n",
              "/**\n",
              " * Generator function which is called between each async step of the upload\n",
              " * process.\n",
              " * @param {string} inputId Element ID of the input file picker element.\n",
              " * @param {string} outputId Element ID of the output display.\n",
              " * @return {!Iterable<!Object>} Iterable of next steps.\n",
              " */\n",
              "function* uploadFilesStep(inputId, outputId) {\n",
              "  const inputElement = document.getElementById(inputId);\n",
              "  inputElement.disabled = false;\n",
              "\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  outputElement.innerHTML = '';\n",
              "\n",
              "  const pickedPromise = new Promise((resolve) => {\n",
              "    inputElement.addEventListener('change', (e) => {\n",
              "      resolve(e.target.files);\n",
              "    });\n",
              "  });\n",
              "\n",
              "  const cancel = document.createElement('button');\n",
              "  inputElement.parentElement.appendChild(cancel);\n",
              "  cancel.textContent = 'Cancel upload';\n",
              "  const cancelPromise = new Promise((resolve) => {\n",
              "    cancel.onclick = () => {\n",
              "      resolve(null);\n",
              "    };\n",
              "  });\n",
              "\n",
              "  // Wait for the user to pick the files.\n",
              "  const files = yield {\n",
              "    promise: Promise.race([pickedPromise, cancelPromise]),\n",
              "    response: {\n",
              "      action: 'starting',\n",
              "    }\n",
              "  };\n",
              "\n",
              "  cancel.remove();\n",
              "\n",
              "  // Disable the input element since further picks are not allowed.\n",
              "  inputElement.disabled = true;\n",
              "\n",
              "  if (!files) {\n",
              "    return {\n",
              "      response: {\n",
              "        action: 'complete',\n",
              "      }\n",
              "    };\n",
              "  }\n",
              "\n",
              "  for (const file of files) {\n",
              "    const li = document.createElement('li');\n",
              "    li.append(span(file.name, {fontWeight: 'bold'}));\n",
              "    li.append(span(\n",
              "        `(${file.type || 'n/a'}) - ${file.size} bytes, ` +\n",
              "        `last modified: ${\n",
              "            file.lastModifiedDate ? file.lastModifiedDate.toLocaleDateString() :\n",
              "                                    'n/a'} - `));\n",
              "    const percent = span('0% done');\n",
              "    li.appendChild(percent);\n",
              "\n",
              "    outputElement.appendChild(li);\n",
              "\n",
              "    const fileDataPromise = new Promise((resolve) => {\n",
              "      const reader = new FileReader();\n",
              "      reader.onload = (e) => {\n",
              "        resolve(e.target.result);\n",
              "      };\n",
              "      reader.readAsArrayBuffer(file);\n",
              "    });\n",
              "    // Wait for the data to be ready.\n",
              "    let fileData = yield {\n",
              "      promise: fileDataPromise,\n",
              "      response: {\n",
              "        action: 'continue',\n",
              "      }\n",
              "    };\n",
              "\n",
              "    // Use a chunked sending to avoid message size limits. See b/62115660.\n",
              "    let position = 0;\n",
              "    do {\n",
              "      const length = Math.min(fileData.byteLength - position, MAX_PAYLOAD_SIZE);\n",
              "      const chunk = new Uint8Array(fileData, position, length);\n",
              "      position += length;\n",
              "\n",
              "      const base64 = btoa(String.fromCharCode.apply(null, chunk));\n",
              "      yield {\n",
              "        response: {\n",
              "          action: 'append',\n",
              "          file: file.name,\n",
              "          data: base64,\n",
              "        },\n",
              "      };\n",
              "\n",
              "      let percentDone = fileData.byteLength === 0 ?\n",
              "          100 :\n",
              "          Math.round((position / fileData.byteLength) * 100);\n",
              "      percent.textContent = `${percentDone}% done`;\n",
              "\n",
              "    } while (position < fileData.byteLength);\n",
              "  }\n",
              "\n",
              "  // All done.\n",
              "  yield {\n",
              "    response: {\n",
              "      action: 'complete',\n",
              "    }\n",
              "  };\n",
              "}\n",
              "\n",
              "scope.google = scope.google || {};\n",
              "scope.google.colab = scope.google.colab || {};\n",
              "scope.google.colab._files = {\n",
              "  _uploadFiles,\n",
              "  _uploadFilesContinue,\n",
              "};\n",
              "})(self);\n",
              "</script> "
            ],
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Saving 07262024_all_proteins.csv to 07262024_all_proteins.csv\n",
            "Input CSV file: 3311 rows containing data, 0 duplicate accession numbers, 84 accession numbers with no modifications.\n",
            "3227 accession numbers with 17794 modifications: 15292 Serine, 2125 Threonine, 334 Tyrosine, 38 Met-loss [N-term], 5 Met-loss+Acetyl [N-term]\n",
            "Separated timepoints have been written to /content/output/modlist_files/all_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "Separated timepoints have been written to /content/output/modlist_files/all_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "Separated timepoints have been written to /content/output/modlist_files/all_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "1354 Total unique accession numbers with 8956 modifications (excluding 17 listed as 'None') in /content/output/modlist_files/sig_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "1957 Total unique accession numbers with 8838 modifications (excluding 67 listed as 'None') in /content/output/modlist_files/notsig_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "15 Total unique accession numbers with 94 modifications (excluding 0 listed as 'None') in /content/output/modlist_files/sig_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "3296 Total unique accession numbers with 17700 modifications (excluding 84 listed as 'None') in /content/output/modlist_files/notsig_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "1076 Total unique accession numbers with 6929 modifications (excluding 14 listed as 'None') in /content/output/modlist_files/sig_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "2235 Total unique accession numbers with 10865 modifications (excluding 70 listed as 'None') in /content/output/modlist_files/notsig_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "Phosphoprotein isoforms with different accession numbers saved to /content/output/modlist_files/phosphoprotein_isoforms.tsv (89 isoforms)\n",
            "Process complete for all files. Logs written to /content/output/logs.\n"
          ]
        }
      ],
      "source": [
        "#@title Import CSV file, separate into all, significant, and not significant data by p-value (0.05)\n",
        "\n",
        "import os\n",
        "import re\n",
        "import csv\n",
        "import pandas as pd\n",
        "\n",
        "def sanitize_filename(filename):\n",
        "    \"\"\"Remove or replace characters that are invalid for file names and spaces.\"\"\"\n",
        "    invalid_chars = ['/', ':', '*', '?', '\"', '<', '>', '|', '\\\\', ' ']\n",
        "    for char in invalid_chars:\n",
        "        filename = filename.replace(char, '_')\n",
        "    return filename\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "def classify_files(input_directory, output_directory, log_file):\n",
        "    \"\"\"Classifies files based on abundance P-value and writes to new files with 'sig_' or 'notsig_' prefix,\n",
        "    and counts unique accession numbers for each output file.\"\"\"\n",
        "    # Ensure the output directory exists\n",
        "    os.makedirs(output_directory, exist_ok=True)\n",
        "\n",
        "    # Initialize DataFrame to store Gene ID and Accession Number pairs\n",
        "    gene_id_accession_pairs = pd.DataFrame(columns=['Gene ID', 'Accession Number'])\n",
        "\n",
        "    # List all TSV files in the input directory\n",
        "    for filename in os.listdir(input_directory):\n",
        "        if filename.endswith(\".tsv\") and \"modlist_Abundance_Ratio_(log2)\" in filename:\n",
        "            # Avoid reprocessing already processed files\n",
        "            if filename.startswith(\"sig_\") or filename.startswith(\"notsig_\"):\n",
        "                continue\n",
        "\n",
        "            input_file_path = os.path.join(input_directory, filename)\n",
        "            with open(input_file_path, mode='r') as file:\n",
        "                reader = csv.reader(file, delimiter='\\t')\n",
        "                headers = next(reader)  # Assume first row is the header\n",
        "\n",
        "                # Find the index for the 'Adj. P-Value' column and 'Accession' column\n",
        "                try:\n",
        "                    pvalue_index = headers.index('Adj. P-Value')\n",
        "                    accession_index = headers.index('Accession Number')\n",
        "                    gene_id_index = headers.index('Gene ID')\n",
        "                except ValueError:\n",
        "                    write_log(log_file, f\"Required columns not found in {filename}\")\n",
        "                    continue\n",
        "\n",
        "                sanitized_filename = sanitize_filename(filename)\n",
        "                sig_output_path = os.path.join(output_directory, f\"sig_{sanitized_filename[4:]}\")\n",
        "                notsig_output_path = os.path.join(output_directory, f\"notsig_{sanitized_filename[4:]}\")\n",
        "\n",
        "                # Prepare to write to new files\n",
        "                sig_writer = csv.writer(open(sig_output_path, 'w', newline=''), delimiter='\\t')\n",
        "                notsig_writer = csv.writer(open(notsig_output_path, 'w', newline=''), delimiter='\\t')\n",
        "\n",
        "                # Write headers to new files\n",
        "                sig_writer.writerow(headers)\n",
        "                notsig_writer.writerow(headers)\n",
        "\n",
        "                # Initialize sets to track unique accession numbers\n",
        "                unique_accessions_sig = set()\n",
        "                unique_accessions_notsig = set()\n",
        "                unique_accessions_no_modifications = set()\n",
        "\n",
        "                # Counters for the number of rows written and modifications listed as \"None\"\n",
        "                sig_rows_written = 0\n",
        "                notsig_rows_written = 0\n",
        "                sig_none_modifications = 0\n",
        "                notsig_none_modifications = 0\n",
        "                sig_modifications_count = 0\n",
        "                notsig_modifications_count = 0\n",
        "\n",
        "                # Iterate over each row in the original file\n",
        "                for row in reader:\n",
        "                    try:\n",
        "                        abundance_pvalue = float(row[pvalue_index])\n",
        "                        accession_number = row[accession_index]\n",
        "                        gene_id = row[gene_id_index]\n",
        "                        modifications = row[headers.index('Modifications')]\n",
        "\n",
        "                        # Append Gene ID and Accession Number pair to DataFrame\n",
        "                        gene_id_accession_pairs = pd.concat([gene_id_accession_pairs, pd.DataFrame([{'Gene ID': gene_id, 'Accession Number': accession_number}])], ignore_index=True)\n",
        "\n",
        "                        # Track unique accession numbers with no letter or numerical values in modifications\n",
        "                        if pd.isna(modifications) or not any(char.isalnum() for char in modifications):\n",
        "                            unique_accessions_no_modifications.add(accession_number)\n",
        "\n",
        "                        # Write row to the appropriate file based on P-value\n",
        "                        if abundance_pvalue <= 0.05:\n",
        "                            sig_writer.writerow(row)\n",
        "                            unique_accessions_sig.add(accession_number)\n",
        "                            sig_rows_written += 1\n",
        "                            if modifications == 'None':\n",
        "                                sig_none_modifications += 1\n",
        "                            else:\n",
        "                                sig_modifications_count += 1\n",
        "                        else:\n",
        "                            notsig_writer.writerow(row)\n",
        "                            unique_accessions_notsig.add(accession_number)\n",
        "                            notsig_rows_written += 1\n",
        "                            if modifications == 'None':\n",
        "                                notsig_none_modifications += 1\n",
        "                            else:\n",
        "                                notsig_modifications_count += 1\n",
        "                    except ValueError:\n",
        "                        # Handle cases where conversion to float fails\n",
        "                        write_log(log_file, f\"Error processing line in {filename}: {row}\")\n",
        "                        continue\n",
        "\n",
        "                # Print the number of unique accession numbers and modifications for each file\n",
        "                write_log(log_file, f\"{len(unique_accessions_sig)} Total unique accession numbers with {sig_modifications_count} modifications (excluding {sig_none_modifications} listed as 'None') in {sig_output_path}\")\n",
        "                write_log(log_file, f\"{len(unique_accessions_notsig)} Total unique accession numbers with {notsig_modifications_count} modifications (excluding {notsig_none_modifications} listed as 'None') in {notsig_output_path}\")\n",
        "\n",
        "    # Identify Gene IDs with multiple different Accession Numbers\n",
        "    gene_id_counts = gene_id_accession_pairs.groupby('Gene ID')['Accession Number'].nunique()\n",
        "    isoform_gene_ids = gene_id_counts[gene_id_counts > 1].index\n",
        "\n",
        "    # Filter the isoform data for exact matches of gene ids\n",
        "    isoform_data = gene_id_accession_pairs[gene_id_accession_pairs['Gene ID'].isin(isoform_gene_ids)]\n",
        "    isoform_data = isoform_data.drop_duplicates()\n",
        "\n",
        "    # Pivot the table to list each Gene ID with all its accession numbers in a single row\n",
        "    isoform_pivot = isoform_data.groupby('Gene ID')['Accession Number'].apply(lambda x: ' '.join(x.unique())).reset_index()\n",
        "\n",
        "    # Create and save the output file for phosphoprotein isoforms\n",
        "    phosphoprotein_isoforms_path = os.path.join(output_directory, 'phosphoprotein_isoforms.tsv')\n",
        "    isoform_pivot.to_csv(phosphoprotein_isoforms_path, sep='\\t', header=True, index=False)\n",
        "    write_log(log_file, f\"Phosphoprotein isoforms with different accession numbers saved to {phosphoprotein_isoforms_path} ({len(isoform_gene_ids)} isoforms)\")\n",
        "\n",
        "def process_files():\n",
        "    # Specify the input and output directories\n",
        "    input_directory = \"/content/input\"\n",
        "    output_directory = \"/content/output/modlist_files\"\n",
        "    logs_directory = os.path.join(\"/content/output/logs\")\n",
        "\n",
        "    # Ensure the directories exist\n",
        "    os.makedirs(input_directory, exist_ok=True)\n",
        "    os.makedirs(output_directory, exist_ok=True)\n",
        "    os.makedirs(logs_directory, exist_ok=True)\n",
        "\n",
        "    # File upload\n",
        "    print(\"Please upload the data file:\")\n",
        "    uploaded_files = files.upload()\n",
        "\n",
        "    # Process each uploaded CSV file\n",
        "    for uploaded_file_name in uploaded_files.keys():\n",
        "        # Define the input file path in the new input directory\n",
        "        input_file_path = os.path.join(input_directory, uploaded_file_name)\n",
        "        log_file = os.path.join(logs_directory, f\"{os.path.splitext(uploaded_file_name)[0]}.log\")\n",
        "\n",
        "        # Save the uploaded file to the new input directory\n",
        "        with open(input_file_path, 'wb') as f:\n",
        "            f.write(uploaded_files[uploaded_file_name])\n",
        "\n",
        "        # Read the uploaded file into a DataFrame\n",
        "        data_df = pd.read_csv(input_file_path)\n",
        "\n",
        "        # Find all columns that include 'Abundance Ratio (log2)' and 'Abundance Ratio Adj. P-Value'\n",
        "        abundance_log2_columns = [col for col in data_df.columns if \"Abundance Ratio (log2)\" in col]\n",
        "        abundance_pvalue_columns = [col for col in data_df.columns if col.startswith(\"Abundance Ratio Adj. P-Value\")]\n",
        "\n",
        "        # Check if the number of 'Abundance Ratio (log2)' and 'Abundance Ratio Adj. P-Value' columns match\n",
        "        if len(abundance_log2_columns) != len(abundance_pvalue_columns):\n",
        "            write_log(log_file, \"The number of 'Abundance Ratio (log2)' and 'Abundance Ratio Adj. P-Value' columns do not match.\")\n",
        "            continue\n",
        "\n",
        "        # Count rows with valid data (valid accession numbers)\n",
        "        valid_data_df = data_df.dropna(subset=['Accession'])\n",
        "        valid_rows_count = valid_data_df.shape[0]\n",
        "\n",
        "        # Count duplicate accession numbers in the input CSV file\n",
        "        duplicate_accessions_count = valid_data_df['Accession'].duplicated().sum()\n",
        "\n",
        "        # Count the number of accession numbers with no modifications\n",
        "        no_modifications_count = valid_data_df[valid_data_df['Modifications'].apply(lambda x: pd.isna(x) or not any(char.isalnum() for char in str(x)))].shape[0]\n",
        "\n",
        "        # Count the total number of accession numbers with modifications\n",
        "        total_accessionmods_count = valid_data_df[valid_data_df['Modifications'].apply(lambda x: not pd.isna(x) and x.strip() not in ['', 'None'])].shape[0]\n",
        "\n",
        "        # Print the general statistics once\n",
        "        write_log(log_file, f\"Input CSV file: {valid_rows_count} rows containing data, {duplicate_accessions_count} duplicate accession numbers, {no_modifications_count} accession numbers with no modifications.\")\n",
        "\n",
        "        # Track modification types\n",
        "        modification_types_count = {\n",
        "            \"Serine\": 0,\n",
        "            \"Threonine\": 0,\n",
        "            \"Tyrosine\": 0,\n",
        "            \"Met-loss [N-term]\": 0,\n",
        "            \"Met-loss+Acetyl [N-term]\": 0\n",
        "        }\n",
        "        total_modifications_count = 0\n",
        "        total_rows_with_modifications = 0\n",
        "\n",
        "        # Iterate over each row in the input data to count modifications\n",
        "        for index, row in valid_data_df.iterrows():\n",
        "            modifications = 'None' if pd.isna(row['Modifications']) else row['Modifications']\n",
        "\n",
        "            # Skip rows without modifications\n",
        "            if modifications == 'None' or modifications.strip() == '':\n",
        "                continue\n",
        "\n",
        "            total_rows_with_modifications += 1\n",
        "\n",
        "            # Extract all modifications, including those in nested parentheses\n",
        "            modification_matches = re.findall(r'\\b[S,T,Y]\\d+\\b', str(modifications))\n",
        "            met_loss_match = re.findall(r'Met-loss\\b', str(modifications))\n",
        "            met_loss_acetyl_match = re.findall(r'Met-loss\\+Acetyl\\b', str(modifications))\n",
        "\n",
        "            # Count modification types\n",
        "            for match in modification_matches:\n",
        "                aa = match[0]\n",
        "                if aa == \"S\":\n",
        "                    modification_types_count[\"Serine\"] += 1\n",
        "                elif aa == \"T\":\n",
        "                    modification_types_count[\"Threonine\"] += 1\n",
        "                elif aa == \"Y\":\n",
        "                    modification_types_count[\"Tyrosine\"] += 1\n",
        "\n",
        "            # Ensure Met-loss is counted only once per row\n",
        "            if met_loss_match:\n",
        "                modification_types_count[\"Met-loss [N-term]\"] += 1\n",
        "            if met_loss_acetyl_match:\n",
        "                modification_types_count[\"Met-loss+Acetyl [N-term]\"] += 1\n",
        "\n",
        "            # Combine all matches into a single list\n",
        "            all_matches = modification_matches\n",
        "            if met_loss_match:\n",
        "                all_matches.append('Met-loss [N-term]')\n",
        "            if met_loss_acetyl_match:\n",
        "                all_matches.append('Met-loss+Acetyl [N-term]')\n",
        "            all_matches = all_matches if all_matches else ['None']\n",
        "\n",
        "            total_modifications_count += len(all_matches)\n",
        "\n",
        "        # Log detailed modification information\n",
        "        modification_types_info = ', '.join([f\"{count} {mod_type}\" for mod_type, count in modification_types_count.items()])\n",
        "        write_log(log_file, f\"{total_accessionmods_count} accession numbers with {total_modifications_count} modifications: {modification_types_info}\")\n",
        "\n",
        "        # Verify that the total count of modifications matches the sum of individual types\n",
        "        total_individual_modifications = sum(modification_types_count.values())\n",
        "        if total_modifications_count != total_individual_modifications:\n",
        "            write_log(log_file, f\"Discrepancy found! Total modifications counted: {total_modifications_count}, Sum of individual types: {total_individual_modifications}\")\n",
        "            write_log(log_file, f\"Number of rows with modifications: {total_rows_with_modifications}\")\n",
        "\n",
        "        # Process each pair of 'Abundance Ratio (log2)' and 'Abundance Ratio Adj. P-Value' columns\n",
        "        for log2_col, pvalue_col in zip(abundance_log2_columns, abundance_pvalue_columns):\n",
        "            all_rows_list = []  # Initialize the list for this specific pair of columns\n",
        "            significant_modifications_count = 0\n",
        "            unique_accessions_no_modifications_set = set()  # To track unique accession numbers with no modifications\n",
        "            sanitized_column_name = sanitize_filename(log2_col)\n",
        "            output_file_path = os.path.join(output_directory, f\"all_modlist_{sanitized_column_name}.tsv\")\n",
        "\n",
        "            # Iterate over each row in the DataFrame, skip header rows\n",
        "            for index, row in data_df.iterrows():\n",
        "                # Default modifications to 'None' if missing or 'NaN'\n",
        "                modifications = 'None' if pd.isna(row['Modifications']) else row['Modifications']\n",
        "\n",
        "                if pd.isna(row['Accession']):\n",
        "                    continue  # Skip rows with missing accession number\n",
        "\n",
        "                accession_number = row['Accession']\n",
        "                gene_id = row['GS Gene ID'] if not pd.isna(row['GS Gene ID']) else 'None'\n",
        "                abundance_ratio_log2 = row[log2_col]\n",
        "                adj_p_value = row[pvalue_col]\n",
        "\n",
        "                # Increment significant modifications count\n",
        "                if adj_p_value <= 0.05:\n",
        "                    significant_modifications_count += 1\n",
        "\n",
        "                # Track unique accession numbers with no modifications\n",
        "                if modifications == 'None' or modifications.startswith(\"Met-loss\"):\n",
        "                    unique_accessions_no_modifications_set.add(accession_number)\n",
        "\n",
        "                # Extract all modifications, including those in nested parentheses\n",
        "                modification_matches = re.findall(r'\\b[S,T,Y]\\d+\\b', str(modifications))\n",
        "                met_loss_match = re.findall(r'Met-loss\\b', str(modifications))\n",
        "                met_loss_acetyl_match = re.findall(r'Met-loss\\+Acetyl\\b', str(modifications))\n",
        "\n",
        "                # Combine all matches into a single list\n",
        "                all_matches = modification_matches\n",
        "                if met_loss_match:\n",
        "                    all_matches.append('Met-loss [N-term]')\n",
        "                if met_loss_acetyl_match:\n",
        "                    all_matches.append('Met-loss+Acetyl [N-term]')\n",
        "                all_matches = all_matches if all_matches else ['None']\n",
        "\n",
        "                # Add rows to the list\n",
        "                for Modification in all_matches:\n",
        "                    all_rows_list.append([\n",
        "                        accession_number,\n",
        "                        gene_id,\n",
        "                        abundance_ratio_log2,\n",
        "                        f\"{adj_p_value:.10f}\",  # Format p-value to 4 decimal places\n",
        "                        Modification\n",
        "                    ])\n",
        "\n",
        "            # Write the data to the output file in a tab-delimited format\n",
        "            with open(output_file_path, mode='w', newline='') as file:\n",
        "                writer = csv.writer(file, delimiter='\\t')\n",
        "                # Write the header\n",
        "                writer.writerow(['Accession Number', 'Gene ID', 'Abundance Ratio (log2)', 'Adj. P-Value', 'Modifications'])\n",
        "                # Write the data from the list of lists\n",
        "                writer.writerows(all_rows_list)\n",
        "\n",
        "            # Print the number of unique accession numbers and total Modifications\n",
        "            unique_accession_numbers = set(row[0] for row in all_rows_list)\n",
        "            write_log(log_file, f\"Separated timepoints have been written to {output_file_path}\")\n",
        "\n",
        "        # Separate files into significant and not significant\n",
        "        classify_files(output_directory, os.path.join(output_directory), log_file)\n",
        "\n",
        "    print(f\"Process complete for all files. Logs written to {logs_directory}.\")\n",
        "\n",
        "if __name__ == '__main__':\n",
        "    process_files()\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "HoKDBL65R0wa"
      },
      "source": [
        "# **Pull sequence, domain information, and predicted structure from Uniprot and AlphaFold databases**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EuqJyXTDRBho",
        "outputId": "b4a17220-868c-4cc4-be5c-2cea754fcb9b"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Total count of accession numbers read from the input file: 3311\n",
            "Input file used to get the accession numbers: 07262024_all_proteins.csv\n",
            "Domain information and sequences have been written to /content//output/libraries/all_protein_seq\n",
            "Number of .fasta files saved: 3311\n"
          ]
        }
      ],
      "source": [
        "#@title Query Uniprot to extract protein sequence and domain information\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "def get_domain_info(accession_number, log_file):\n",
        "    \"\"\" Fetches protein sequence and domain information from the UniProt API. \"\"\"\n",
        "    url = f\"https://www.ebi.ac.uk/proteins/api/proteins/{accession_number}\"\n",
        "    headers = {\"Accept\": \"application/json\"}\n",
        "    response = requests.get(url, headers=headers)\n",
        "    if response.status_code == 200:\n",
        "        protein_info = response.json()\n",
        "        protein_length = protein_info.get('sequence', {}).get('length', 0)\n",
        "        sequence = protein_info.get('sequence', {}).get('sequence', '')\n",
        "        domain_annotations = []\n",
        "        for feature in protein_info.get(\"features\", []):\n",
        "            if feature.get(\"type\") == \"DOMAIN\":\n",
        "                domain_annotations.append((feature.get(\"description\", \"No description\"), feature.get(\"begin\", \"-\"), feature.get(\"end\", \"-\")))\n",
        "        return accession_number, sequence, domain_annotations, protein_length\n",
        "    else:\n",
        "        write_log(log_file, f\"Accession number {accession_number} not found on UniProt.\")\n",
        "    return accession_number, '', [], 0\n",
        "\n",
        "def extract_accession_numbers_from_csv(input_csv_path):\n",
        "    \"\"\"Extracts unique accession numbers from the given CSV file.\"\"\"\n",
        "    data = pd.read_csv(input_csv_path)\n",
        "    accession_numbers = data['Accession'].dropna().unique().tolist()\n",
        "    return accession_numbers, os.path.basename(input_csv_path)\n",
        "\n",
        "def process_accession_numbers(accession_numbers, output_folder, log_file):\n",
        "    \"\"\" Queries UniProt API for each accession number and writes domain and sequence data to the output folder. \"\"\"\n",
        "    libraries_folder = os.path.join(\"output/libraries\")\n",
        "    os.makedirs(libraries_folder, exist_ok=True)\n",
        "    output_file_path = os.path.join(libraries_folder, \"domainlibrary.tsv\")\n",
        "\n",
        "    fasta_files_written = 0\n",
        "\n",
        "    with open(output_file_path, mode='w', newline='') as file:\n",
        "        writer = csv.writer(file, delimiter='\\t')\n",
        "        writer.writerow(['Accession', 'Domain Type', 'Domain Start', 'Domain End', 'Protein Length'])\n",
        "\n",
        "        failed_filename = os.path.join(\"output/failed\", \"uniprot_error.txt\")\n",
        "        os.makedirs(os.path.dirname(failed_filename), exist_ok=True)\n",
        "\n",
        "        with open(failed_filename, 'w') as failed_file:\n",
        "            with ThreadPoolExecutor(max_workers=20) as executor:\n",
        "                future_to_accession = {executor.submit(get_domain_info, acc, log_file): acc for acc in accession_numbers}\n",
        "                for future in as_completed(future_to_accession):\n",
        "                    accession_number, sequence, domains, protein_length = future.result()\n",
        "                    if sequence:  # Check if the sequence was found\n",
        "                        fasta_filename = os.path.join(output_folder, f\"{accession_number}.fasta\")\n",
        "                        with open(fasta_filename, \"w\") as fasta_file:\n",
        "                            fasta_file.write(f\">{accession_number}\\n{sequence}\\n\")\n",
        "                        fasta_files_written += 1\n",
        "\n",
        "                        if domains:\n",
        "                            for domain in domains:\n",
        "                                domain_name, domain_start, domain_end = domain\n",
        "                                writer.writerow([accession_number, domain_name, domain_start, domain_end, protein_length])\n",
        "                        else:\n",
        "                            writer.writerow([accession_number, \"No Domains\", 0, 0, protein_length])\n",
        "                    else:\n",
        "                        failed_file.write(f\"{accession_number}\\n\")\n",
        "\n",
        "    write_log(log_file, f\"Domain information and sequences have been written to {output_folder}\")\n",
        "    write_log(log_file, f\"Number of .fasta files saved: {fasta_files_written}\")\n",
        "\n",
        "def find_first_csv(directory):\n",
        "    \"\"\"Finds the first CSV file in the given directory and returns its full path.\"\"\"\n",
        "    for file in os.listdir(directory):\n",
        "        if file.endswith('.csv'):\n",
        "            return os.path.join(directory, file)\n",
        "    return None\n",
        "\n",
        "# Main execution\n",
        "input_directory = '/content/input'  # Directory containing the CSV file\n",
        "logs_directory = os.path.join('/content/output/logs')\n",
        "os.makedirs(logs_directory, exist_ok=True)\n",
        "\n",
        "input_csv_path = find_first_csv(input_directory)\n",
        "if input_csv_path:\n",
        "    accession_numbers, input_filename = extract_accession_numbers_from_csv(input_csv_path)\n",
        "    log_file = os.path.join(logs_directory, f\"uniprot_proteinseq-domain_process.log\")\n",
        "    if accession_numbers:\n",
        "        output_folder = \"/content//output/libraries/all_protein_seq\"\n",
        "        os.makedirs(output_folder, exist_ok=True)\n",
        "        write_log(log_file, f\"Total count of accession numbers read from the input file: {len(accession_numbers)}\")\n",
        "        write_log(log_file, f\"Input file used to get the accession numbers: {input_filename}\")\n",
        "        process_accession_numbers(accession_numbers, output_folder, log_file)\n",
        "    else:\n",
        "        write_log(log_file, \"No accession numbers found in the input CSV file.\")\n",
        "else:\n",
        "    log_file = os.path.join(logs_directory, \"process_log.log\")  # Default log file name if no CSV is found\n",
        "    write_log(log_file, \"No CSV file found in the input directory.\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "r3jgNP2lyas4",
        "outputId": "a3226f8c-e2c4-45b8-d574-400d24b018c1"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Appended 'Domain Type' to notsig_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv and saved as /content/output/modlist_files/domain_sig_modlist_notsig_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "Appended 'Domain Type' to notsig_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv and saved as /content/output/modlist_files/domain_sig_modlist_notsig_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "Appended 'Domain Type' to sig_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv and saved as /content/output/modlist_files/domain_sig_modlist_sig_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "Appended 'Domain Type' to all_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv and saved as /content/output/modlist_files/domain_all_modlist_all_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "Appended 'Domain Type' to all_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv and saved as /content/output/modlist_files/domain_all_modlist_all_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "Appended 'Domain Type' to sig_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv and saved as /content/output/modlist_files/domain_sig_modlist_sig_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "Appended 'Domain Type' to all_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv and saved as /content/output/modlist_files/domain_all_modlist_all_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "Appended 'Domain Type' to sig_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv and saved as /content/output/modlist_files/domain_sig_modlist_sig_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "Appended 'Domain Type' to notsig_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv and saved as /content/output/modlist_files/domain_sig_modlist_notsig_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n"
          ]
        }
      ],
      "source": [
        "#@title Read modlist files and generate new files appending domain information\n",
        "\n",
        "import os\n",
        "import csv\n",
        "import pandas as pd\n",
        "\n",
        "def sanitize_filename(filename):\n",
        "    \"\"\"Remove or replace characters that are invalid for file names and spaces.\"\"\"\n",
        "    invalid_chars = ['/', ':', '*', '?', '\"', '<', '>', '|', '\\\\', ' ']\n",
        "    for char in invalid_chars:\n",
        "        filename = filename.replace(char, '_')\n",
        "    return filename\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "def check_columns_exist(df, columns, filename, log_file):\n",
        "    \"\"\"Checks if the specified columns exist in the DataFrame.\"\"\"\n",
        "    missing_columns = [col for col in columns if col not in df.columns]\n",
        "    if missing_columns:\n",
        "        write_log(log_file, f\"Missing columns {missing_columns} in file {filename}\")\n",
        "        return False\n",
        "    return True\n",
        "\n",
        "def extract_residue_number(modification):\n",
        "    \"\"\"Extracts the residue number from the modification string.\"\"\"\n",
        "    if isinstance(modification, str):\n",
        "        digits = ''.join(filter(str.isdigit, modification))\n",
        "        if digits:\n",
        "            return int(digits)\n",
        "    return None\n",
        "\n",
        "def append_domain_type_to_modlist_files(modlist_directory, domain_library_path, log_file):\n",
        "    \"\"\"Appends the 'Domain Type' column to the modlist files using the domain library.\"\"\"\n",
        "    # Read the domain library into a DataFrame\n",
        "    domain_library = pd.read_csv(domain_library_path, delimiter='\\t')\n",
        "    if not check_columns_exist(domain_library, ['Accession', 'Domain Type', 'Domain Start', 'Domain End'], domain_library_path, log_file):\n",
        "        return\n",
        "\n",
        "    # Create a dictionary mapping accession numbers and residue positions to domain types\n",
        "    domain_dict = {}\n",
        "    for _, row in domain_library.iterrows():\n",
        "        accession = row['Accession']\n",
        "        domain_type = row['Domain Type']\n",
        "        domain_start = int(row['Domain Start'])\n",
        "        domain_end = int(row['Domain End'])\n",
        "\n",
        "        if accession not in domain_dict:\n",
        "            domain_dict[accession] = []\n",
        "        domain_dict[accession].append((domain_start, domain_end, domain_type))\n",
        "\n",
        "    # List all modlist files in the directory\n",
        "    for filename in os.listdir(modlist_directory):\n",
        "        if filename.endswith(\".tsv\") and \"modlist\" in filename:\n",
        "            modlist_file_path = os.path.join(modlist_directory, filename)\n",
        "\n",
        "            # Read the modlist file into a DataFrame\n",
        "            modlist_df = pd.read_csv(modlist_file_path, delimiter='\\t')\n",
        "            if not check_columns_exist(modlist_df, ['Accession Number', 'Modifications'], modlist_file_path, log_file):\n",
        "                continue\n",
        "\n",
        "            # Initialize the 'Domain Type' column with 'No Domain'\n",
        "            modlist_df['Domain Type'] = 'No Domain'\n",
        "\n",
        "            # Iterate over the rows to determine the domain type for each modification\n",
        "            for index, row in modlist_df.iterrows():\n",
        "                accession_number = row['Accession Number']\n",
        "                residue_number = extract_residue_number(row['Modifications'])\n",
        "\n",
        "                if accession_number in domain_dict and residue_number is not None:\n",
        "                    for domain_start, domain_end, domain_type in domain_dict[accession_number]:\n",
        "                        if domain_start <= residue_number <= domain_end:\n",
        "                            modlist_df.at[index, 'Domain Type'] = domain_type\n",
        "                            break\n",
        "\n",
        "            # Determine the appropriate output filename prefix\n",
        "            if \"sig\" in filename:\n",
        "                prefix = \"domain_sig_modlist_\"\n",
        "            elif \"notsig\" in filename:\n",
        "                prefix = \"domain_notsig_modlist_\"\n",
        "            else:\n",
        "                prefix = \"domain_all_modlist_\"\n",
        "\n",
        "            # Write the updated DataFrame back to a new TSV file\n",
        "            updated_file_path = os.path.join(modlist_directory, f\"{prefix}{filename}\")\n",
        "            modlist_df.to_csv(updated_file_path, sep='\\t', index=False)\n",
        "\n",
        "            write_log(log_file, f\"Appended 'Domain Type' to {filename} and saved as {updated_file_path}\")\n",
        "\n",
        "# Main execution\n",
        "modlist_directory = '/content/output/modlist_files'\n",
        "logs_directory = os.path.join('/content/output/logs')\n",
        "os.makedirs(logs_directory, exist_ok=True)\n",
        "\n",
        "log_file = os.path.join(logs_directory, f\"uniprot_proteinseq-domain_process.log\")\n",
        "\n",
        "# Append domain type to modlist files\n",
        "domain_library_path = \"output/libraries/domainlibrary.tsv\"\n",
        "append_domain_type_to_modlist_files(modlist_directory, domain_library_path, log_file)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "background_save": true,
          "base_uri": "https://localhost:8080/"
        },
        "id": "GNSm2fW7L114",
        "outputId": "0aa8b3eb-0e6e-4fd9-8de4-f031e2509d5f"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Total count of accession numbers read from the input file: 3311\n",
            "Input file used to get the accession numbers: 07262024_all_proteins.csv\n",
            "Processed 100 accession numbers\n",
            "Processed 200 accession numbers\n",
            "Processed 300 accession numbers\n",
            "Processed 400 accession numbers\n",
            "Processed 500 accession numbers\n",
            "Processed 600 accession numbers\n",
            "Processed 700 accession numbers\n",
            "Processed 800 accession numbers\n",
            "Processed 900 accession numbers\n",
            "Processed 1000 accession numbers\n",
            "Processed 1100 accession numbers\n",
            "Processed 1200 accession numbers\n",
            "Processed 1300 accession numbers\n",
            "Processed 1400 accession numbers\n",
            "Processed 1500 accession numbers\n",
            "Processed 1600 accession numbers\n",
            "Processed 1700 accession numbers\n",
            "Processed 1800 accession numbers\n",
            "Processed 1900 accession numbers\n",
            "Processed 2000 accession numbers\n",
            "Processed 2100 accession numbers\n",
            "Processed 2200 accession numbers\n",
            "Processed 2300 accession numbers\n",
            "Processed 2400 accession numbers\n",
            "Processed 2500 accession numbers\n",
            "Processed 2600 accession numbers\n",
            "Processed 2700 accession numbers\n",
            "Processed 2800 accession numbers\n",
            "Processed 2900 accession numbers\n",
            "Processed 3000 accession numbers\n",
            "Processed 3100 accession numbers\n",
            "Processed 3200 accession numbers\n",
            "Processed 3300 accession numbers\n",
            "Keyword search results have been written to /content/output/libraries/keyword_search_results.tsv\n",
            "\n",
            "Keyword counts:\n",
            "cold: 27\n",
            "shock: 33\n",
            "repair: 109\n",
            "damage: 102\n",
            "hibernation: 1\n",
            "brain: 87\n",
            "neuron: 385\n",
            "ischemia: 7\n",
            "stroke: 0\n",
            "parkinson: 1\n",
            "alzheimer: 3\n",
            "Keyword counts have been printed to console.\n"
          ]
        }
      ],
      "source": [
        "#@title Query Uniprot to perform keyword search and organizes the output file by significant or not signicant modification abundance\n",
        "#Results are saved in libraries directory\n",
        "import os\n",
        "import csv\n",
        "import requests\n",
        "import pandas as pd\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "def get_protein_info(accession_number, log_file):\n",
        "    \"\"\"Fetches protein information from the UniProt API and checks for specific keywords.\"\"\"\n",
        "    url = f\"https://www.ebi.ac.uk/proteins/api/proteins/{accession_number}\"\n",
        "    headers = {\"Accept\": \"application/json\"}\n",
        "    try:\n",
        "        response = requests.get(url, headers=headers, timeout=30)\n",
        "        response.raise_for_status()  # Raise an exception for bad responses\n",
        "        protein_info = response.json()\n",
        "        entry = response.text.lower()\n",
        "\n",
        "        keywords = [\"cold\", \"shock\", \"repair\", \"damage\", \"hibernation\", \"brain\", \"neuron\", \"ischemia\", \"stroke\", \"parkinson\", \"alzheimer\"]\n",
        "        found_keywords = []\n",
        "\n",
        "        for keyword in keywords:\n",
        "            if keyword in entry:\n",
        "                found_keywords.append(keyword)\n",
        "\n",
        "        if found_keywords:\n",
        "            keyword_string = \"; \".join(found_keywords)\n",
        "            keyword_index = entry.index(found_keywords[0])\n",
        "            start_index = max(0, keyword_index - 30)\n",
        "            end_index = min(len(entry), keyword_index + len(found_keywords[-1]) + 30)\n",
        "            surrounding_text = entry[start_index:end_index]\n",
        "\n",
        "            protein_name = protein_info.get('protein', {}).get('recommendedName', {}).get('fullName', {}).get('value', 'Unknown')\n",
        "            gene_name = protein_info.get('gene', [{}])[0].get('name', {}).get('value', 'Unknown')\n",
        "            organism = protein_info.get('organism', {}).get('names', [{}])[0].get('value', 'Unknown')\n",
        "\n",
        "            return (accession_number, protein_name, gene_name, organism, keyword_string, surrounding_text)\n",
        "        else:\n",
        "            return (accession_number, '', '', '', '', '')\n",
        "\n",
        "    except requests.exceptions.RequestException as e:\n",
        "        write_log(log_file, f\"Error fetching data for accession number {accession_number}: {str(e)}\")\n",
        "        return (accession_number, '', '', '', '', '')\n",
        "\n",
        "def extract_accession_numbers_from_csv(input_csv_path):\n",
        "    \"\"\"Extracts unique accession numbers and relevant data from the given CSV file.\"\"\"\n",
        "    data = pd.read_csv(input_csv_path)\n",
        "    if 'Accession' not in data.columns:\n",
        "        raise KeyError(f\"Missing 'Accession' column in input CSV file.\")\n",
        "\n",
        "    abundance_ratio_columns = [col for col in data.columns if col.startswith(\"Abundance Ratio (log2)\")]\n",
        "    p_value_columns = [col for col in data.columns if col.startswith(\"Abundance Ratio Adj. P-Value\")]\n",
        "\n",
        "    if not abundance_ratio_columns or not p_value_columns:\n",
        "        missing_columns = []\n",
        "        if not abundance_ratio_columns:\n",
        "            missing_columns.append(\"Abundance Ratio (log2)\")\n",
        "        if not p_value_columns:\n",
        "            missing_columns.append(\"Abundance Ratio Adj. P-Value\")\n",
        "        raise KeyError(f\"Missing columns in input CSV file: {missing_columns}\")\n",
        "\n",
        "    data = data.dropna(subset=['Accession'])\n",
        "    relevant_data = data[['Accession'] + abundance_ratio_columns + p_value_columns].dropna().to_dict(orient='records')\n",
        "    return relevant_data, os.path.basename(input_csv_path), abundance_ratio_columns, p_value_columns\n",
        "\n",
        "def process_accession_numbers(accession_data, abundance_ratio_columns, p_value_columns, output_folder, log_file):\n",
        "    \"\"\"Queries UniProt API for each accession number and writes domain and sequence data to the output folder.\"\"\"\n",
        "    libraries_folder = os.path.join(output_folder, \"libraries\")\n",
        "    os.makedirs(libraries_folder, exist_ok=True)\n",
        "    output_file_path = os.path.join(libraries_folder, \"keyword_search_results.tsv\")\n",
        "\n",
        "    significant_data = {col: [] for col in p_value_columns}\n",
        "    non_significant_data = {col: [] for col in p_value_columns}\n",
        "    keyword_counts = {keyword: 0 for keyword in [\"cold\", \"shock\", \"repair\", \"damage\", \"hibernation\", \"brain\", \"neuron\", \"ischemia\", \"stroke\", \"parkinson\", \"alzheimer\"]}\n",
        "\n",
        "    with open(output_file_path, mode='w', newline='') as file:\n",
        "        writer = csv.writer(file, delimiter='\\t')\n",
        "        writer.writerow(['Accession', 'Protein Name', 'Gene Name', 'Organism', 'Keywords', 'Surrounding Texts'] + abundance_ratio_columns + p_value_columns)\n",
        "\n",
        "        failed_filename = os.path.join(output_folder, \"failed\", \"uniprot_keyword_error.txt\")\n",
        "        os.makedirs(os.path.dirname(failed_filename), exist_ok=True)\n",
        "\n",
        "        for i, acc_data in enumerate(accession_data, 1):\n",
        "            accession_number = acc_data['Accession']\n",
        "            result = get_protein_info(accession_number, log_file)\n",
        "\n",
        "            if result[1]:  # Check if the protein info was retrieved\n",
        "                row = list(result) + [acc_data[col] for col in abundance_ratio_columns + p_value_columns]\n",
        "                # Check for significance and categorize by each p-value column\n",
        "                for col in p_value_columns:\n",
        "                    if float(acc_data[col]) <= 0.05:\n",
        "                        significant_data[col].append(row)\n",
        "                    else:\n",
        "                        non_significant_data[col].append(row)\n",
        "\n",
        "                for keyword in result[4].split(\"; \"):\n",
        "                    keyword_counts[keyword] += 1\n",
        "            else:\n",
        "                with open(failed_filename, 'a') as failed_file:\n",
        "                    failed_file.write(f\"{result[0]}\\n\")\n",
        "\n",
        "            if i % 100 == 0:\n",
        "                write_log(log_file, f\"Processed {i} accession numbers\")\n",
        "\n",
        "        # Write data for each abundance ratio column\n",
        "        for col in p_value_columns:\n",
        "            writer.writerow([f\"Significant {col}\"])\n",
        "            writer.writerows(significant_data[col])\n",
        "            writer.writerow([f\"Not significant {col}\"])\n",
        "            writer.writerows(non_significant_data[col])\n",
        "\n",
        "    write_log(log_file, f\"Keyword search results have been written to {output_file_path}\")\n",
        "\n",
        "    # Print the keyword counts\n",
        "    print(\"\\nKeyword counts:\")\n",
        "    for keyword, count in keyword_counts.items():\n",
        "        print(f\"{keyword}: {count}\")\n",
        "    write_log(log_file, \"Keyword counts have been printed to console.\")\n",
        "\n",
        "def find_first_csv(directory):\n",
        "    \"\"\"Finds the first CSV file in the given directory and returns its full path.\"\"\"\n",
        "    for file in os.listdir(directory):\n",
        "        if file.endswith('.csv'):\n",
        "            return os.path.join(directory, file)\n",
        "    return None\n",
        "\n",
        "# Main execution\n",
        "input_directory = '/content/input'  # Directory containing the CSV file\n",
        "logs_directory = '/content/output/logs'\n",
        "os.makedirs(logs_directory, exist_ok=True)\n",
        "\n",
        "input_csv_path = find_first_csv(input_directory)\n",
        "if input_csv_path:\n",
        "    try:\n",
        "        accession_data, input_filename, abundance_ratio_columns, p_value_columns = extract_accession_numbers_from_csv(input_csv_path)\n",
        "        log_file = os.path.join(logs_directory, \"uniprot_keyword-search.log\")\n",
        "        if accession_data:\n",
        "            output_folder = '/content/output'\n",
        "            os.makedirs(output_folder, exist_ok=True)\n",
        "            write_log(log_file, f\"Total count of accession numbers read from the input file: {len(accession_data)}\")\n",
        "            write_log(log_file, f\"Input file used to get the accession numbers: {input_filename}\")\n",
        "\n",
        "            process_accession_numbers(accession_data, abundance_ratio_columns, p_value_columns, output_folder, log_file)\n",
        "        else:\n",
        "            write_log(log_file, \"No accession numbers found in the input CSV file.\")\n",
        "    except KeyError as e:\n",
        "        write_log(log_file, f\"Error processing input CSV file: {e}\")\n",
        "    except Exception as e:\n",
        "        write_log(log_file, f\"Unexpected error occurred: {str(e)}\")\n",
        "else:\n",
        "    log_file = os.path.join(logs_directory, \"process_log.log\")  # Default log file name if no CSV is found\n",
        "    write_log(log_file, \"No CSV file found in the input directory.\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "background_save": true
        },
        "id": "IPX8JdkRvKOO",
        "outputId": "6f4a263c-7b91-49f5-d6d9-058c9f2e5c1e"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Download complete.\n",
            "2791 successful PDB files downloaded\n",
            "520 Failed accession numbers saved to output/failed/pdb_error.txt\n",
            "Number of .fasta files copied to af2fail_protein_seq folder: 520\n"
          ]
        }
      ],
      "source": [
        "#@title Download PDB coordinate files from alphafold database API, saves fasta files of proteins not in AF2 database to failed directory\n",
        "import shutil\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "def find_first_csv_basename(directory):\n",
        "    \"\"\"Finds the first CSV file in the specified directory and returns its base name without extension.\"\"\"\n",
        "    for file in os.listdir(directory):\n",
        "        if file.endswith('.csv'):\n",
        "            return os.path.splitext(file)[0]\n",
        "    return None  # Returns None if no CSV file is found\n",
        "\n",
        "def extract_failed_accession_numbers(failed_filename):\n",
        "    \"\"\"Extracts failed accession numbers from the failed file.\"\"\"\n",
        "    if os.path.exists(failed_filename):\n",
        "        with open(failed_filename, \"r\") as f:\n",
        "            failed_accessions = f.read().splitlines()\n",
        "        return failed_accessions\n",
        "    return []\n",
        "\n",
        "def copy_failed_fasta_files(failed_accessions, input_folder, output_folder):\n",
        "    \"\"\"Copies corresponding .fasta files of failed accession numbers to the output folder.\"\"\"\n",
        "    copied_files_count = 0\n",
        "    for accession_number in failed_accessions:\n",
        "        fasta_filename = os.path.join(input_folder, f\"{accession_number}.fasta\")\n",
        "        if os.path.exists(fasta_filename):\n",
        "            shutil.copy(fasta_filename, output_folder)\n",
        "            copied_files_count += 1\n",
        "        else:\n",
        "            write_log(log_file, f\"Could not find corresponding .fasta file for {accession_number}\")\n",
        "    return copied_files_count\n",
        "\n",
        "# Define the directory where your CSV files are stored\n",
        "directory = \"/content/input\"\n",
        "\n",
        "# Get the base name of the first CSV file found\n",
        "csv_base_name = find_first_csv_basename(directory)\n",
        "if not csv_base_name:\n",
        "    print(\"No CSV file found in the directory.\")\n",
        "    exit()\n",
        "\n",
        "# Define the directory to save PDB files\n",
        "output_directory = \"/content//output/libraries/all_pdb_files\"\n",
        "os.makedirs(output_directory, exist_ok=True)\n",
        "\n",
        "# Initialize counters for downloads\n",
        "successful_downloads = 0\n",
        "failed_pdb_count = 0\n",
        "\n",
        "# Read unique accession numbers from the CSV file\n",
        "csv_file_path = os.path.join(directory, f\"{csv_base_name}.csv\")\n",
        "if not os.path.exists(csv_file_path):\n",
        "    print(f\"CSV file does not exist: {csv_file_path}\")\n",
        "    exit()\n",
        "\n",
        "modlist_df = pd.read_csv(csv_file_path)\n",
        "unique_accessions = set(modlist_df['Accession'].dropna())\n",
        "\n",
        "# Define log file\n",
        "logs_directory = os.path.join('/content/output/logs')\n",
        "os.makedirs(logs_directory, exist_ok=True)\n",
        "log_file = os.path.join(logs_directory, \"pdb_download.log\")\n",
        "\n",
        "# Open output file for failed downloads\n",
        "failed_directory = os.path.join(\"output/failed\")\n",
        "os.makedirs(failed_directory, exist_ok=True)\n",
        "failed_filename = os.path.join(failed_directory, \"pdb_error.txt\")\n",
        "\n",
        "with open(failed_filename, \"w\") as failed_file:\n",
        "    def download_pdb(accession_number):\n",
        "        pdb_url = f'https://alphafold.ebi.ac.uk/files/AF-{accession_number}-F1-model_v4.pdb'\n",
        "        try:\n",
        "            response = requests.get(pdb_url)\n",
        "            if response.status_code == 200:\n",
        "                pdb_filename = os.path.join(output_directory, f\"{accession_number}.pdb\")\n",
        "                with open(pdb_filename, \"wb\") as f:\n",
        "                    f.write(response.content)\n",
        "                return True\n",
        "            else:\n",
        "                failed_file.write(str(accession_number) + \"\\n\")\n",
        "                return False\n",
        "        except requests.exceptions.RequestException as e:\n",
        "            write_log(log_file, f\"Failed to download PDB file for {accession_number}. Error: {e}\")\n",
        "            failed_file.write(str(accession_number) + \"\\n\")\n",
        "            return False\n",
        "\n",
        "    with ThreadPoolExecutor(max_workers=10) as executor:\n",
        "        futures = {executor.submit(download_pdb, accession): accession for accession in unique_accessions}\n",
        "        for future in as_completed(futures):\n",
        "            accession = futures[future]\n",
        "            try:\n",
        "                result = future.result()\n",
        "                if result:\n",
        "                    successful_downloads += 1\n",
        "                else:\n",
        "                    failed_pdb_count += 1\n",
        "            except Exception as e:\n",
        "                write_log(log_file, f\"Error occurred for accession {accession}: {e}\")\n",
        "                failed_file.write(str(accession) + \"\\n\")\n",
        "                failed_pdb_count += 1\n",
        "\n",
        "write_log(log_file, \"Download complete.\")\n",
        "write_log(log_file, f\"{successful_downloads} successful PDB files downloaded\")\n",
        "write_log(log_file, f\"{failed_pdb_count} Failed accession numbers saved to {failed_filename}\")\n",
        "\n",
        "# Create output folder for failed PDBs\n",
        "af2fail_protein_seq_folder = \"/content/output/failed/af2fail_protein_seq\"\n",
        "os.makedirs(af2fail_protein_seq_folder, exist_ok=True)\n",
        "\n",
        "# Extract failed accession numbers\n",
        "failed_accessions = extract_failed_accession_numbers(failed_filename)\n",
        "\n",
        "# Copy corresponding .fasta files to the output folder and count the copied files\n",
        "if failed_accessions:\n",
        "    copied_files_count = copy_failed_fasta_files(failed_accessions, \"/content//output/libraries/all_protein_seq\", af2fail_protein_seq_folder)\n",
        "    write_log(log_file, f\"Number of .fasta files copied to af2fail_protein_seq folder: {copied_files_count}\")\n",
        "else:\n",
        "    write_log(log_file, \"No failed accession numbers found.\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "background_save": true
        },
        "id": "23yAknYQOVNM",
        "outputId": "978887fa-7949-46e8-e10d-36473d753af7"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Downloading zip file: /content/af2fail_protein_seq.zip\n"
          ]
        },
        {
          "data": {
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ],
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "data": {
            "application/javascript": [
              "download(\"download_f49c7af4-b31a-4099-aad1-385edb5a85ec\", \"af2fail_protein_seq.zip\", 22)"
            ],
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "#@title Zip folder containing fasta files for proteins that do not have alphafold entries </n> *Run if you want to use to predict models elsewhere*\n",
        "\n",
        "def zip_directory(directory_to_zip, zip_output_path):\n",
        "    \"\"\"Create a ZIP file containing the specified directory.\"\"\"\n",
        "    with zipfile.ZipFile(zip_output_path, 'w', zipfile.ZIP_DEFLATED) as zipf:\n",
        "        # Walk through the directory to zip only the contents of the specified directory\n",
        "        for root, dirs, files in os.walk(directory_to_zip):\n",
        "            for file in files:\n",
        "                # Create the full path to the file\n",
        "                file_path = os.path.join(root, file)\n",
        "                # Arcname strips the leading part of the path to avoid excessive folder structure in zip\n",
        "                arcname = os.path.relpath(file_path, directory_to_zip)\n",
        "                zipf.write(file_path, arcname=arcname)\n",
        "\n",
        "def main():\n",
        "    # Define the path to the directory you want to zip\n",
        "    directory_to_zip = \"/content/output/af2fail_protein_seq\"\n",
        "\n",
        "    # Define the name for the zip file\n",
        "    zip_filename = \"af2fail_protein_seq.zip\"\n",
        "    zip_output_path = os.path.join(\"/content\", zip_filename)\n",
        "\n",
        "    # Create a ZIP file\n",
        "    zip_directory(directory_to_zip, zip_output_path)\n",
        "\n",
        "    # Download the zip file, which triggers the browser's save dialog in Google Colab\n",
        "    print(f\"Downloading zip file: {zip_output_path}\")\n",
        "    colab_files.download(zip_output_path)\n",
        "\n",
        "if __name__ == '__main__':\n",
        "    main()\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "background_save": true
        },
        "id": "UFCAVJqidDzL",
        "outputId": "694bb0d3-6f90-4f65-8542-208db46177e7"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Rows without modifications: 84\n",
            "Total unique accession numbers with modifications in input file output/modlist_files/all_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv: 3227\n",
            "Residue position 520 out of range for sequence I3N092\n",
            "Saved file with adjacent sequences as: output/modlist_files/AdjacentSeq_all__(Torpor)___(Aug_Active).tsv\n",
            "Total rows processed: 17878, Rows with no sequence: 1, Rows with no modifications: 84\n",
            "Total unique accession numbers in output file: 3227\n",
            "Total modifications in output file: 17793\n",
            "Rows without modifications: 84\n",
            "Total unique accession numbers with modifications in input file output/modlist_files/all_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv: 3227\n",
            "Residue position 520 out of range for sequence I3N092\n",
            "Saved file with adjacent sequences as: output/modlist_files/AdjacentSeq_all__(Aug_Active)___(Mar_Active).tsv\n",
            "Total rows processed: 17878, Rows with no sequence: 1, Rows with no modifications: 84\n",
            "Total unique accession numbers in output file: 3227\n",
            "Total modifications in output file: 17793\n",
            "Rows without modifications: 84\n",
            "Total unique accession numbers with modifications in input file output/modlist_files/all_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv: 3227\n",
            "Residue position 520 out of range for sequence I3N092\n",
            "Saved file with adjacent sequences as: output/modlist_files/AdjacentSeq_all__(Mar_Active)___(Torpor).tsv\n",
            "Total rows processed: 17878, Rows with no sequence: 1, Rows with no modifications: 84\n",
            "Total unique accession numbers in output file: 3227\n",
            "Total modifications in output file: 17793\n",
            "Processing complete.\n"
          ]
        }
      ],
      "source": [
        "#@title Print AdjacentSeq files with 7 aa before and after site\n",
        "import os\n",
        "import pandas as pd\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "def extract_sequence(accession_number, residue_position, fasta_folder, log_file, modification_type):\n",
        "    fasta_file = os.path.join(fasta_folder, accession_number + '.fasta')\n",
        "    if not os.path.exists(fasta_file):\n",
        "        write_log(log_file, f\"FASTA file not found for accession number: {accession_number}\")\n",
        "        return None\n",
        "\n",
        "    with open(fasta_file, 'r') as f:\n",
        "        _ = f.readline()  # Skip the header line\n",
        "        sequence = f.readline().strip()  # Read the sequence\n",
        "        if len(sequence) < residue_position:\n",
        "            write_log(log_file, f\"Residue position {residue_position} out of range for sequence {accession_number}\")\n",
        "            return None\n",
        "\n",
        "        position_index = residue_position - 1  # Residue position is 1-indexed\n",
        "\n",
        "        # Handle specific cases for \"Met-loss [N-Term]\" and \"Met-loss+Acetyl [N-Term]\"\n",
        "        if modification_type in [\"Met-loss [N-term]\", \"Met-loss+Acetyl [N-term]\"]:\n",
        "            surrounding_sequence = sequence[position_index:position_index + 8]\n",
        "        else:\n",
        "            surrounding_sequence = sequence[max(0, position_index - 7):position_index] + sequence[position_index:position_index + 8]\n",
        "    return surrounding_sequence\n",
        "\n",
        "def process_modlist_file(modlist_file, fasta_folder, output_folder, log_file, input_accessions_with_mods, input_accessions_without_mods):\n",
        "    modlist_data = pd.read_csv(modlist_file, delimiter='\\t')\n",
        "    if 'Modifications' not in modlist_data.columns:\n",
        "        write_log(log_file, f\"'Modifications' column not found in file {modlist_file}. Available columns: {list(modlist_data.columns)}\")\n",
        "        return\n",
        "\n",
        "    # Standardize column names to avoid KeyError\n",
        "    modlist_data.columns = modlist_data.columns.str.strip()\n",
        "\n",
        "    # Filter rows with and without modifications\n",
        "    modlist_data_with_modifications = modlist_data[modlist_data['Modifications'].notna()]\n",
        "    modlist_data_without_modifications = modlist_data[modlist_data['Modifications'].isna()]\n",
        "\n",
        "    # Log count of rows without modifications\n",
        "    rows_without_modifications_count = len(modlist_data_without_modifications)\n",
        "    write_log(log_file, f\"Rows without modifications: {rows_without_modifications_count}\")\n",
        "\n",
        "    output_data = []\n",
        "    no_sequence_count = 0\n",
        "    unique_accession_numbers = set()\n",
        "\n",
        "    # Count unique accession numbers in the input file with modifications\n",
        "    unique_accessions_with_modifications = modlist_data_with_modifications['Accession Number'].nunique()\n",
        "    write_log(log_file, f\"Total unique accession numbers with modifications in input file {modlist_file}: {unique_accessions_with_modifications}\")\n",
        "\n",
        "    for index, row in modlist_data_with_modifications.iterrows():\n",
        "        accession_number = row['Accession Number']\n",
        "        modification = row['Modifications']\n",
        "        try:\n",
        "            if \"Met-loss\" in modification:\n",
        "                residue_position = 1  # Met-loss is always at position 1\n",
        "            else:\n",
        "                residue_position = int(''.join(filter(str.isdigit, modification)))\n",
        "        except ValueError:\n",
        "            write_log(log_file, f\"Invalid modification format '{modification}' for row {index} in file {modlist_file}\")\n",
        "            continue\n",
        "\n",
        "        adjacent_seq = extract_sequence(accession_number, residue_position, fasta_folder, log_file, modification)\n",
        "        if adjacent_seq:\n",
        "            unique_accession_numbers.add(accession_number)\n",
        "            output_data.append({'Accession Number': row['Accession Number'],\n",
        "                                'Gene ID': row['Gene ID'],\n",
        "                                'Abundance Ratio (log2)': row['Abundance Ratio (log2)'],\n",
        "                                'Adj. P-Value': row['Adj. P-Value'],\n",
        "                                'Modifications': row['Modifications'],\n",
        "                                'AdjacentSeq': adjacent_seq})\n",
        "        else:\n",
        "            no_sequence_count += 1\n",
        "\n",
        "    output_df = pd.DataFrame(output_data)\n",
        "    output_file = os.path.join(output_folder, f'{os.path.basename(modlist_file).replace(\"all_modlist_Abundance_Ratio_(log2)_\", \"AdjacentSeq_all_\")}')\n",
        "    output_df.to_csv(output_file, sep='\\t', index=False)\n",
        "    write_log(log_file, f\"Saved file with adjacent sequences as: {output_file}\")\n",
        "\n",
        "    # Check if any 'Met-loss' is present in the 'Modifications' column of rows with modifications\n",
        "    met_loss_present = any(isinstance(mod, str) and 'Met-loss' in mod for mod in modlist_data_with_modifications['Modifications'].tolist())\n",
        "    corrected_rows_without_modifications_count = rows_without_modifications_count\n",
        "\n",
        "    write_log(log_file, f\"Total rows processed: {len(modlist_data)}, Rows with no sequence: {no_sequence_count}, Rows with no modifications: {corrected_rows_without_modifications_count}\")\n",
        "    write_log(log_file, f\"Total unique accession numbers in output file: {len(unique_accession_numbers)}\")\n",
        "    write_log(log_file, f\"Total modifications in output file: {len(output_data)}\")\n",
        "\n",
        "    # Compare unique accession numbers between input and output\n",
        "    missing_accessions = set(modlist_data_with_modifications['Accession Number'].unique()) - unique_accession_numbers\n",
        "    if missing_accessions:\n",
        "        write_log(log_file, f\"Missing accession numbers: {missing_accessions}\")\n",
        "\n",
        "    # Identify and log inconsistencies between input and output files\n",
        "    output_accessions = set(modlist_data_with_modifications['Accession Number'].unique())\n",
        "    inconsistent_accessions_with_mods = output_accessions.intersection(input_accessions_without_mods)\n",
        "    inconsistent_accessions_without_mods = set(input_accessions_with_mods) - output_accessions\n",
        "\n",
        "    if inconsistent_accessions_with_mods:\n",
        "        write_log(log_file, f\"Inconsistent accession numbers with modifications in output file but not in input file: {inconsistent_accessions_with_mods}\")\n",
        "    if inconsistent_accessions_without_mods:\n",
        "        write_log(log_file, f\"Inconsistent accession numbers without modifications in output file but with modifications in input file: {inconsistent_accessions_without_mods}\")\n",
        "\n",
        "# Define the paths\n",
        "modlist_folder = 'output/modlist_files'\n",
        "fasta_folder = '/content/output/libraries/all_protein_seq'\n",
        "output_folder = 'output/modlist_files'\n",
        "logs_directory = 'output/logs'\n",
        "os.makedirs(logs_directory, exist_ok=True)\n",
        "log_file = os.path.join(logs_directory, \"adjacentseq_process.log\")\n",
        "\n",
        "# Ensure the output directory exists\n",
        "os.makedirs(output_folder, exist_ok=True)\n",
        "\n",
        "# List all files in the directory\n",
        "all_files = os.listdir(modlist_folder)\n",
        "\n",
        "# Filter files that start with 'all_modlist_Abundance_Ratio_(log2)__'\n",
        "modlist_files = [file for file in all_files if file.startswith('all_modlist_Abundance_Ratio_(log2)__')]\n",
        "\n",
        "# Find the CSV file in the input directory\n",
        "input_directory = 'input/'\n",
        "input_files = [f for f in os.listdir(input_directory) if f.endswith('.csv')]\n",
        "if not input_files:\n",
        "    raise FileNotFoundError(\"No CSV file found in the input directory.\")\n",
        "input_csv_path = os.path.join(input_directory, input_files[0])\n",
        "\n",
        "# Read the input CSV file to get accession numbers with and without modifications\n",
        "input_data = pd.read_csv(input_csv_path)\n",
        "input_data.columns = input_data.columns.str.strip()  # Ensure no leading/trailing spaces in column names\n",
        "input_accessions_with_mods = set(input_data[input_data['Modifications'].notna()]['Accession'])\n",
        "input_accessions_without_mods = set(input_data[input_data['Modifications'].isna()]['Accession'])\n",
        "\n",
        "# Process each modlist file\n",
        "for modlist_file in modlist_files:\n",
        "    process_modlist_file(os.path.join(modlist_folder, modlist_file), fasta_folder, output_folder, log_file, input_accessions_with_mods, input_accessions_without_mods)\n",
        "\n",
        "write_log(log_file, \"Processing complete.\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xaT4bsSnfD1Y"
      },
      "source": [
        "# **Process data using DSSP (secondary structure & solvent accessibility) and IUPred2A (disorder) prediction**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "background_save": true
        },
        "id": "ps0j20MWUt_X",
        "outputId": "ce72814c-fa1e-4b40-fc47-86ae193233f2"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Initial number of PDB files: 3201\n",
            "Number of PDB files processed from /content//output/libraries/all_pdb_files: 2791\n",
            "Number of PDB files processed from /content/output/libraries/colabfold_pdbs: 410\n",
            "Number of successful PDB files processed: 3201\n",
            "Number of failed PDB files: 0\n",
            "Number of missing accession numbers: 111\n",
            "Processing complete.\n"
          ]
        }
      ],
      "source": [
        "#@title Process PDB files with DSSP to print new ss files with secondary structure and solvent accessibility\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "# Create directories\n",
        "output_directory = \"/content//output/libraries/all_protein_ss\"\n",
        "os.makedirs(output_directory, exist_ok=True)\n",
        "logs_directory = \"/content/output/logs\"\n",
        "os.makedirs(logs_directory, exist_ok=True)\n",
        "log_file = os.path.join(logs_directory, \"dssp-process.log\")\n",
        "\n",
        "# Get the list of PDB files in the directories and track the count for each directory\n",
        "pdb_files_directories = [\"/content//output/libraries/all_pdb_files\", \"/content/output/libraries/colabfold_pdbs\"]\n",
        "pdb_files = []\n",
        "pdb_file_counts = {directory: 0 for directory in pdb_files_directories}\n",
        "\n",
        "for pdb_files_directory in pdb_files_directories:\n",
        "    if os.path.exists(pdb_files_directory):\n",
        "        pdb_files_in_dir = [os.path.join(pdb_files_directory, file) for file in os.listdir(pdb_files_directory) if file.endswith(\".pdb\")]\n",
        "        pdb_files.extend(pdb_files_in_dir)\n",
        "        pdb_file_counts[pdb_files_directory] = len(pdb_files_in_dir)\n",
        "\n",
        "def classify_secondary_structure(pdb_file_path, chain_id, log_file):\n",
        "    try:\n",
        "        # Create a PDB parser object\n",
        "        parser = PDBParser()\n",
        "\n",
        "        # Parse the PDB file\n",
        "        structure = parser.get_structure('structure', pdb_file_path)\n",
        "\n",
        "        # Run DSSP\n",
        "        model = structure[0]\n",
        "        dssp = DSSP(model, pdb_file_path)\n",
        "\n",
        "        # Extract accession number from the file name\n",
        "        accession_number = os.path.splitext(os.path.basename(pdb_file_path))[0]\n",
        "\n",
        "        # Generate output file path without the .pdb extension\n",
        "        output_file_path = os.path.join(output_directory, f\"{accession_number}_ss.tsv\")\n",
        "\n",
        "        # Write output to file using tab-delimited format\n",
        "        with open(output_file_path, mode='w', newline='') as file:\n",
        "            writer = csv.writer(file, delimiter='\\t')\n",
        "            # Write the header\n",
        "            writer.writerow([\"Amino Acid\", \"Residue Number\", \"Secondary Structure\", \"Confidence\", \"Solvent Accessibility\"])\n",
        "\n",
        "            # Iterate over residues in the specified chain\n",
        "            for residue in model[chain_id]:\n",
        "                if residue.id[0] == ' ' and 'CA' in residue:  # Ensure it's an amino acid residue and has a CA atom\n",
        "                    residue_number = residue.id[1]\n",
        "                    ss = dssp[(chain_id, residue_number)]\n",
        "                    amino_acid = residue.get_resname()\n",
        "                    confidence = residue[\"CA\"].get_bfactor()  # Using B-factor of the CA atom\n",
        "                    solvent_accessibility = ss[3]  # Solvent accessibility\n",
        "\n",
        "                    # Write data to the CSV file\n",
        "                    writer.writerow([amino_acid, residue_number, ss[2], f\"{confidence:.3f}\", f\"{solvent_accessibility:.3f}\"])\n",
        "\n",
        "    except Exception as e:\n",
        "        write_log(log_file, f\"Error processing PDB file {pdb_file_path}: {e}\")\n",
        "\n",
        "# Count initial number of accession numbers\n",
        "initial_accession_count = len(pdb_files)\n",
        "write_log(log_file, f\"Initial number of PDB files: {initial_accession_count}\")\n",
        "\n",
        "# Use ThreadPoolExecutor to speed up processing\n",
        "successful_downloads = 0\n",
        "failed_pdb_count = 0\n",
        "\n",
        "with ThreadPoolExecutor(max_workers=10) as executor:\n",
        "    future_to_pdb = {executor.submit(classify_secondary_structure, pdb_file, \"A\", log_file): pdb_file for pdb_file in pdb_files}\n",
        "\n",
        "    for future in as_completed(future_to_pdb):\n",
        "        try:\n",
        "            future.result()\n",
        "            successful_downloads += 1\n",
        "        except Exception as e:\n",
        "            failed_pdb_count += 1\n",
        "\n",
        "# Log the counts for each directory\n",
        "for directory, count in pdb_file_counts.items():\n",
        "    write_log(log_file, f\"Number of PDB files processed from {directory}: {count}\")\n",
        "\n",
        "# Log final counts\n",
        "write_log(log_file, f\"Number of successful PDB files processed: {successful_downloads}\")\n",
        "write_log(log_file, f\"Number of failed PDB files: {failed_pdb_count}\")\n",
        "\n",
        "## fixed below to read dynamic filename in input directory\n",
        "\n",
        "# Find the first CSV file in the input directory\n",
        "input_directory = \"/content/input\"\n",
        "input_csv_path = None\n",
        "\n",
        "for file in os.listdir(input_directory):\n",
        "    if file.endswith(\".csv\"):\n",
        "        input_csv_path = os.path.join(input_directory, file)\n",
        "        break\n",
        "\n",
        "# Compare the accession numbers in the CSV file to the accession numbers of the processed PDBs\n",
        "missing_accessions_count = 0\n",
        "\n",
        "if input_csv_path:\n",
        "    try:\n",
        "        csv_accessions = set(pd.read_csv(input_csv_path)['Accession'])\n",
        "        pdb_accessions = set(os.path.splitext(os.path.basename(pdb_file))[0] for pdb_file in pdb_files)\n",
        "        missing_accessions = csv_accessions - pdb_accessions\n",
        "        missing_accessions_count = len(missing_accessions)\n",
        "        write_log(log_file, f\"Number of missing accession numbers: {missing_accessions_count}\")\n",
        "    except KeyError as e:\n",
        "        write_log(log_file, f\"Error processing input CSV file: {e}\")\n",
        "else:\n",
        "    write_log(log_file, f\"Input CSV file not found in {input_directory}.\")\n",
        "\n",
        "print(\"Processing complete.\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "background_save": true
        },
        "id": "mcbn29XN5uqw",
        "outputId": "f28b8e7f-8261-447d-d59f-48c9050975d8"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Processing file: /content/output/modlist_files/all_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "Overlap found in accession numbers saved and failed: {'I3N092'}\n",
            "Accession number I3N092 is found in both saved and failed outputs. Corresponding rows from failed file:\n",
            "['I3N092', ' MAP6 ', '0.46', '7.707e-07', 'S520', 'SS file not found', '-', '-', '-', 'No SS file']\n",
            "3117 Unique accession numbers saved to /content/output/modlist_files/all_mods_ss_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "17794 Modifications saved to /content/output/modlist_files/all_mods_ss_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "Failed due to no modification: 84 accessions with 84 modifications\n",
            "Failed due to no SS file: 111 accessions with 1768 modifications\n",
            "Total rows in all_mods_ss output: 16026\n",
            "Total rows in sig_mods_ss output: 7103\n",
            "Total rows in notsig_mods_ss output: 8923\n",
            "Processing file: /content/output/modlist_files/all_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "Overlap found in accession numbers saved and failed: {'I3N092'}\n",
            "Accession number I3N092 is found in both saved and failed outputs. Corresponding rows from failed file:\n",
            "['I3N092', ' MAP6 ', '0.01', '1.0', 'S520', 'SS file not found', '-', '-', '-', 'No SS file']\n",
            "3117 Unique accession numbers saved to /content/output/modlist_files/all_mods_ss_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "17794 Modifications saved to /content/output/modlist_files/all_mods_ss_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "Failed due to no modification: 84 accessions with 84 modifications\n",
            "Failed due to no SS file: 111 accessions with 1768 modifications\n",
            "Total rows in all_mods_ss output: 16026\n",
            "Total rows in sig_mods_ss output: 94\n",
            "Total rows in notsig_mods_ss output: 15932\n",
            "Processing file: /content/output/modlist_files/all_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "Overlap found in accession numbers saved and failed: {'I3N092'}\n",
            "Accession number I3N092 is found in both saved and failed outputs. Corresponding rows from failed file:\n",
            "['I3N092', ' MAP6 ', '-0.47', '1.6188e-06', 'S520', 'SS file not found', '-', '-', '-', 'No SS file']\n",
            "3117 Unique accession numbers saved to /content/output/modlist_files/all_mods_ss_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "17794 Modifications saved to /content/output/modlist_files/all_mods_ss_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "Failed due to no modification: 84 accessions with 84 modifications\n",
            "Failed due to no SS file: 111 accessions with 1768 modifications\n",
            "Total rows in all_mods_ss output: 16026\n",
            "Total rows in sig_mods_ss output: 5757\n",
            "Total rows in notsig_mods_ss output: 10269\n",
            "Processing complete.\n"
          ]
        }
      ],
      "source": [
        "#@title Combine all data to print all_mods_ss, sig_mods_ss, and notsig_mods_ss into processed directory\n",
        "\n",
        "import os\n",
        "import pandas as pd\n",
        "import csv\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "def process_modlist_file(modlist_file, ss_directory, domain_library, failed_directory, log_file, output_directory):\n",
        "    # Read the modlist file\n",
        "    modlist_df = pd.read_csv(modlist_file, delimiter='\\t')\n",
        "    base_filename = os.path.basename(modlist_file).replace(\"all_modlist_\", \"\").replace(\".tsv\", \"\")\n",
        "    all_mods_output_file_path = os.path.join(output_directory, f\"all_mods_ss_{base_filename}.tsv\")\n",
        "    sig_mods_output_file_path = os.path.join(output_directory, f\"sig_mods_ss_{base_filename}.tsv\")\n",
        "    notsig_mods_output_file_path = os.path.join(output_directory, f\"notsig_mods_ss_{base_filename}.tsv\")\n",
        "    failed_output_file_path = os.path.join(failed_directory, f\"all_mods_ss_failed_{base_filename}.tsv\")\n",
        "\n",
        "    # Open output files for writing using the CSV writer with tab delimiters\n",
        "    with open(all_mods_output_file_path, \"w\", newline='') as all_mods_file, \\\n",
        "         open(sig_mods_output_file_path, \"w\", newline='') as sig_mods_file, \\\n",
        "         open(notsig_mods_output_file_path, \"w\", newline='') as notsig_mods_file, \\\n",
        "         open(failed_output_file_path, \"w\", newline='') as failed_output_file:\n",
        "\n",
        "        all_mods_writer = csv.writer(all_mods_file, delimiter='\\t')\n",
        "        sig_mods_writer = csv.writer(sig_mods_file, delimiter='\\t')\n",
        "        notsig_mods_writer = csv.writer(notsig_mods_file, delimiter='\\t')\n",
        "        failed_writer = csv.writer(failed_output_file, delimiter='\\t')\n",
        "\n",
        "        headers = [\"Accession Number\", \"Gene ID\", \"Abundance Ratio (log2)\", \"Adj. P-Value\", \"Modification\", \"Secondary Structure\", \"Confidence\", \"Solvent Accessibility\", \"Domain Type\"]\n",
        "        all_mods_writer.writerow(headers)\n",
        "        sig_mods_writer.writerow(headers)\n",
        "        notsig_mods_writer.writerow(headers)\n",
        "        failed_writer.writerow(headers + [\"Failure Reason\"])\n",
        "\n",
        "        unique_accessions_saved = set()  # Set to store unique accession numbers saved to the output file\n",
        "        failed_accessions = {\"No modification\": set(), \"No SS file\": set()}  # Dictionary to store unique accession numbers for each failure reason\n",
        "        failed_modifications = {\"No modification\": 0, \"No SS file\": 0}  # Dictionary to store counts of modifications for each failure reason\n",
        "\n",
        "        row_count_all_mods = 0\n",
        "        row_count_sig_mods = 0\n",
        "        row_count_notsig_mods = 0\n",
        "\n",
        "        for index, row in modlist_df.iterrows():\n",
        "            accession_number = row['Accession Number']\n",
        "            gene_id = row['Gene ID']\n",
        "            Modification = row[\"Modifications\"]\n",
        "\n",
        "            if pd.isna(Modification) or (Modification.strip() == '' and not Modification.startswith('Met-loss')):\n",
        "                failed_writer.writerow([accession_number, gene_id, row[\"Abundance Ratio (log2)\"], row[\"Adj. P-Value\"], Modification, \"-\", \"-\", \"-\", \"-\", \"No modification\"])\n",
        "                failed_accessions[\"No modification\"].add(accession_number)\n",
        "                failed_modifications[\"No modification\"] += 1\n",
        "                continue\n",
        "\n",
        "            Modification_numbers = ''.join(filter(str.isdigit, str(Modification)))\n",
        "            if not Modification_numbers and not Modification.startswith('Met-loss'):\n",
        "                failed_writer.writerow([accession_number, gene_id, row[\"Abundance Ratio (log2)\"], row[\"Adj. P-Value\"], Modification, \"-\", \"-\", \"-\", \"-\", \"Invalid Modification\"])\n",
        "                failed_accessions[\"No modification\"].add(accession_number)\n",
        "                failed_modifications[\"No modification\"] += 1\n",
        "                continue\n",
        "\n",
        "            Modification_number = int(Modification_numbers) if Modification_numbers else 0\n",
        "            abundance_value = row.get(\"Abundance Ratio (log2)\")\n",
        "            abundance_p_value = row.get(\"Adj. P-Value\")\n",
        "\n",
        "            found_ss = False\n",
        "            for ss_filename in [f\"{accession_number}_ss.tsv\"]:\n",
        "                ss_file_path = os.path.join(ss_directory, ss_filename)\n",
        "\n",
        "                if os.path.exists(ss_file_path):\n",
        "                    with open(ss_file_path, \"r\") as ss_file:\n",
        "                        next(ss_file)  # Assuming the first line is a header\n",
        "                        for line in ss_file:\n",
        "                            parts = line.strip().split(\"\\t\")\n",
        "                            if len(parts) > 4:\n",
        "                                residue_number = int(parts[1])\n",
        "                                if residue_number == Modification_number or Modification.startswith('Met-loss'):\n",
        "                                    domain_type = get_domain_type(accession_number, residue_number, domain_library)\n",
        "                                    solvent_accessibility = f\"{float(parts[4]):.3f}\" if parts[4] != \"-\" else \"None\"\n",
        "                                    row_data = [accession_number, gene_id, abundance_value, abundance_p_value, Modification, parts[2], parts[3], solvent_accessibility, domain_type]\n",
        "                                    all_mods_writer.writerow(row_data)\n",
        "                                    unique_accessions_saved.add(accession_number)\n",
        "                                    found_ss = True\n",
        "                                    row_count_all_mods += 1\n",
        "\n",
        "                                    # Write to sig_mods or notsig_mods based on abundance_p_value\n",
        "                                    if abundance_p_value and abundance_p_value < 0.05:\n",
        "                                        sig_mods_writer.writerow(row_data)\n",
        "                                        row_count_sig_mods += 1\n",
        "                                    else:\n",
        "                                        notsig_mods_writer.writerow(row_data)\n",
        "                                        row_count_notsig_mods += 1\n",
        "\n",
        "                                    break\n",
        "                        if found_ss:\n",
        "                            break\n",
        "\n",
        "            if not found_ss:\n",
        "                failed_writer.writerow([accession_number, gene_id, abundance_value, abundance_p_value, Modification, \"SS file not found\", \"-\", \"-\", \"-\", \"No SS file\"])\n",
        "                failed_accessions[\"No SS file\"].add(accession_number)\n",
        "                failed_modifications[\"No SS file\"] += 1\n",
        "\n",
        "        # Check for overlaps between saved and failed\n",
        "        saved_accessions = unique_accessions_saved\n",
        "        failed_accessions_set = failed_accessions[\"No modification\"].union(failed_accessions[\"No SS file\"])\n",
        "        overlap_accessions = saved_accessions.intersection(failed_accessions_set)\n",
        "\n",
        "        if overlap_accessions:\n",
        "            write_log(log_file, f\"Overlap found in accession numbers saved and failed: {overlap_accessions}\")\n",
        "            for accession in overlap_accessions:\n",
        "                write_log(log_file, f\"Accession number {accession} is found in both saved and failed outputs. Corresponding rows from failed file:\")\n",
        "                with open(failed_output_file_path, \"r\") as failed_output_file:\n",
        "                    failed_reader = csv.reader(failed_output_file, delimiter='\\t')\n",
        "                    next(failed_reader)  # Skip header\n",
        "                    for row in failed_reader:\n",
        "                        if row[0] == accession:\n",
        "                            write_log(log_file, f\"{row}\")\n",
        "\n",
        "        # Count the number of unique accession numbers saved to the output file\n",
        "        num_unique_accessions_saved = len(unique_accessions_saved)\n",
        "        write_log(log_file, f\"{num_unique_accessions_saved} Unique accession numbers saved to {all_mods_output_file_path}\")\n",
        "\n",
        "        # Additional logs for modifications\n",
        "        saved_modifications = modlist_df[~modlist_df['Modifications'].isna()].shape[0]\n",
        "        write_log(log_file, f\"{saved_modifications} Modifications saved to {all_mods_output_file_path}\")\n",
        "        write_log(log_file, f\"Failed due to no modification: {len(failed_accessions['No modification'])} accessions with {failed_modifications['No modification']} modifications\")\n",
        "        write_log(log_file, f\"Failed due to no SS file: {len(failed_accessions['No SS file'])} accessions with {failed_modifications['No SS file']} modifications\")\n",
        "\n",
        "        # Log the total rows in each output file\n",
        "        write_log(log_file, f\"Total rows in all_mods_ss output: {row_count_all_mods}\")\n",
        "        write_log(log_file, f\"Total rows in sig_mods_ss output: {row_count_sig_mods}\")\n",
        "        write_log(log_file, f\"Total rows in notsig_mods_ss output: {row_count_notsig_mods}\")\n",
        "\n",
        "        # Ensure the sum of sig and notsig rows equals all_mods rows\n",
        "        if row_count_all_mods != row_count_sig_mods + row_count_notsig_mods:\n",
        "            write_log(log_file, \"Error: The sum of sig and notsig rows does not equal the total rows in all_mods.\")\n",
        "\n",
        "def get_domain_type(accession_number, residue_number, domain_library):\n",
        "    if accession_number in domain_library:\n",
        "        for domain_info in domain_library[accession_number]:\n",
        "            domain_type, start, end = domain_info\n",
        "            if start <= residue_number <= end:\n",
        "                return domain_type\n",
        "    return \"None\"\n",
        "\n",
        "def read_domain_library(domain_library_file):\n",
        "    domain_library = {}\n",
        "    if os.path.exists(domain_library_file):\n",
        "        with open(domain_library_file, \"r\") as file:\n",
        "            next(file)  # Skip the header\n",
        "            for line in file:\n",
        "                parts = line.strip().split(\"\\t\")\n",
        "                if len(parts) >= 4:\n",
        "                    accession_number = parts[0]\n",
        "                    domain_type = parts[1]\n",
        "                    start = int(parts[2])\n",
        "                    end = int(parts[3])\n",
        "                    domain_library.setdefault(accession_number, []).append((domain_type, start, end))\n",
        "    return domain_library\n",
        "\n",
        "def main():\n",
        "    modlist_directory = \"/content/output/modlist_files\"  # Directory containing modlist files\n",
        "    ss_directory = \"/content//output/libraries/all_protein_ss\"\n",
        "    domain_library_file = \"/content/output/libraries/domainlibrary.tsv\"\n",
        "    failed_directory = \"/content/output/failed\"\n",
        "    os.makedirs(failed_directory, exist_ok=True)\n",
        "\n",
        "    log_file = os.path.join(\"/content/output/logs\", \"combine_modinfo_process.log\")\n",
        "    output_directory = \"/content/output/modlist_files\"\n",
        "    os.makedirs(output_directory, exist_ok=True)\n",
        "\n",
        "    domain_library = read_domain_library(domain_library_file)\n",
        "\n",
        "    # Process each modlist file\n",
        "    for filename in os.listdir(modlist_directory):\n",
        "        if filename.startswith(\"all_modlist_Abundance\") and filename.endswith(\".tsv\"):\n",
        "            modlist_file = os.path.join(modlist_directory, filename)\n",
        "            write_log(log_file, f\"Processing file: {modlist_file}\")\n",
        "            process_modlist_file(modlist_file, ss_directory, domain_library, failed_directory, log_file, output_directory)\n",
        "\n",
        "    write_log(log_file, \"Processing complete.\")\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    main()\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0yPV_lCVsLr7",
        "outputId": "629ed773-2cbe-45ce-83f6-d48022729ed5"
      },
      "outputs": [
        {
          "metadata": {
            "tags": null
          },
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Found input CSV file: /content/input/07262024_all_proteins.csv\n",
            "Reading domain library...\n",
            "Domain library contains 3311 entries.\n",
            "Input CSV contains 3311 unique accession numbers.\n",
            "Found modlist file: /content/output/modlist_files/all_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "Total number of modifications read from input all_modlist file: 17794\n",
            "Read modifications for 3227 accessions.\n"
          ]
        },
        {
          "metadata": {
            "tags": null
          },
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Fetching disorder info: 100%|██████████| 3311/3311 [12:44<00:00,  4.33it/s]\n"
          ]
        },
        {
          "metadata": {
            "tags": null
          },
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Number of .txt files saved: 3311\n",
            "Number of accessions that failed: 0\n",
            "Found 3311 disorder files to process.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Processing disorder files: 100%|██████████| 67/67 [19:16<00:00, 17.26s/it]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Processing complete.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n"
          ]
        }
      ],
      "source": [
        "#@title Use IUPred2A to predict disorder for all proteins and save .txt and .pdf files\n",
        "\n",
        "import os\n",
        "import requests\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import gc\n",
        "from concurrent.futures import ThreadPoolExecutor, as_completed\n",
        "from tqdm import tqdm\n",
        "\n",
        "# Define directories\n",
        "disorder_files_dir = \"/content/output/libraries/IUPred2A_disorder\"\n",
        "output_figures_path = \"/content/output/plots/IUPred2A\"\n",
        "logs_directory = \"/content/output/logs\"\n",
        "log_file = os.path.join(logs_directory, \"iupred2a_histogram_process.log\")\n",
        "\n",
        "# Ensure the output directory and logs directory exist\n",
        "os.makedirs(output_figures_path, exist_ok=True)\n",
        "os.makedirs(logs_directory, exist_ok=True)\n",
        "os.makedirs(disorder_files_dir, exist_ok=True)\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "def fetch_disorder_info(accession_number):\n",
        "    \"\"\"Fetches disorder information from the IUPred2A website.\"\"\"\n",
        "    url = f\"https://iupred2a.elte.hu/iupred2a/{accession_number}\"\n",
        "    response = requests.get(url)\n",
        "    if response.status_code == 200:\n",
        "        return response.text\n",
        "    return None\n",
        "\n",
        "def save_disorder_info(accession_number, disorder_info):\n",
        "    \"\"\"Saves the disorder information to a .txt file.\"\"\"\n",
        "    filepath = os.path.join(disorder_files_dir, f\"{accession_number}_iupred2a.txt\")\n",
        "    with open(filepath, 'w') as file:\n",
        "        file.write(disorder_info)\n",
        "    return filepath\n",
        "\n",
        "def read_disorder_values(filepath):\n",
        "    \"\"\"Reads disorder values from an IUPred2A prediction file.\"\"\"\n",
        "    residues = []\n",
        "    disorder_values = []\n",
        "    with open(filepath, 'r') as file:\n",
        "        for line in file:\n",
        "            if line.startswith('#') or not line.strip():\n",
        "                continue\n",
        "            parts = line.split()\n",
        "            if len(parts) == 3:\n",
        "                try:\n",
        "                    residue = int(parts[0])\n",
        "                    disorder_score = float(parts[2])\n",
        "                    residues.append(residue)\n",
        "                    disorder_values.append(disorder_score)\n",
        "                except ValueError:\n",
        "                    continue  # Ignore conversion errors and continue\n",
        "    return residues, disorder_values\n",
        "\n",
        "def read_domain_library(domain_library_file):\n",
        "    \"\"\"Reads domain information from a domain library file.\"\"\"\n",
        "    domain_library = {}\n",
        "    if os.path.exists(domain_library_file):\n",
        "        with open(domain_library_file, 'r') as file:\n",
        "            next(file)  # Skip the header\n",
        "            for line in file:\n",
        "                parts = line.strip().split(\"\\t\")\n",
        "                if len(parts) >= 4:\n",
        "                    accession_number = parts[0]\n",
        "                    domain_type = parts[1]\n",
        "                    start = int(parts[2])\n",
        "                    end = int(parts[3])\n",
        "                    domain_library.setdefault(accession_number, []).append((domain_type, start, end))\n",
        "    return domain_library\n",
        "\n",
        "def read_modifications(modlist_file):\n",
        "    \"\"\"Reads modification information from the modlist file.\"\"\"\n",
        "    modifications = {}\n",
        "    total_modifications = 0\n",
        "    df = pd.read_csv(modlist_file, delimiter='\\t')\n",
        "\n",
        "    if 'Accession Number' in df.columns and 'Modifications' in df.columns:\n",
        "        for _, row in df.iterrows():\n",
        "            accession = row['Accession Number']\n",
        "            mod = row['Modifications']\n",
        "            domain_type = row.get('Domain Type', 'None')\n",
        "            if pd.notna(mod):\n",
        "                total_modifications += 1\n",
        "                try:\n",
        "                    mod_residue = 1 if 'N-term' in mod else int(''.join(filter(str.isdigit, str(mod))))\n",
        "                    modifications.setdefault(accession, []).append(f\"{mod_residue} ({domain_type})\")\n",
        "                except ValueError:\n",
        "                    continue  # Ignore conversion errors and continue\n",
        "    print(f\"Total number of modifications read from input all_modlist file: {total_modifications}\")\n",
        "    return modifications\n",
        "\n",
        "def plot_disorder_prediction(residues, scores, accession, domains, modifications):\n",
        "    \"\"\"Plots the disorder prediction results.\"\"\"\n",
        "    plot_filename = os.path.join(output_figures_path, f'{accession}_disorder_prediction.pdf')\n",
        "    if os.path.exists(plot_filename):\n",
        "        return  # Skip if the plot already exists\n",
        "\n",
        "    plt.figure(figsize=(12, 6))\n",
        "    plt.plot(residues, scores, linestyle='-', color='black', label='Disorder Score')\n",
        "\n",
        "    # Add domain boundaries and labels\n",
        "    if accession in domains:\n",
        "        for domain_type, start, end in domains[accession]:\n",
        "            plt.axvspan(start, end, color='lightgray', alpha=0.5)\n",
        "            plt.text((start + end) / 2, 0.8, domain_type, horizontalalignment='center', verticalalignment='center',\n",
        "                     fontsize=12, fontweight='bold', color='black', fontname='Liberation Sans', bbox=dict(facecolor='white', edgecolor='none', alpha=0.7))\n",
        "\n",
        "    # Mark modification residues\n",
        "    if accession in modifications:\n",
        "        for mod_residue in modifications[accession]:\n",
        "            mod_residue_pos = int(mod_residue.split(' ')[0])\n",
        "            if mod_residue_pos - 1 < len(scores):\n",
        "                plt.plot(mod_residue_pos, scores[mod_residue_pos - 1], 'ro')\n",
        "\n",
        "    plt.xlabel('Residue', fontname='Liberation Sans', fontsize=10)\n",
        "    plt.ylabel('Disorder Score', fontname='Liberation Sans', fontsize=10)\n",
        "    plt.title(f'IUPred2A Disorder Prediction for {accession}', fontname='Liberation Sans', fontsize=10)\n",
        "    plt.xticks(fontname='Liberation Sans', fontsize=8)\n",
        "    plt.yticks(fontname='Liberation Sans', fontsize=8)\n",
        "    plt.ylim(0, 1)\n",
        "    plt.grid(False)  # Disable grid lines\n",
        "    plt.tight_layout()\n",
        "    plt.savefig(plot_filename, bbox_inches='tight')\n",
        "    plt.close()\n",
        "\n",
        "def process_disorder_file(filepath, domains, modifications):\n",
        "    \"\"\"Processes a single IUPred2A disorder file to generate a plot.\"\"\"\n",
        "    accession = os.path.basename(filepath).split('_')[0]\n",
        "    residues, scores = read_disorder_values(filepath)\n",
        "    plot_disorder_prediction(residues, scores, accession, domains, modifications)\n",
        "    gc.collect()  # Collect garbage to free up memory after processing each file\n",
        "\n",
        "def read_accession_numbers(input_csv_file):\n",
        "    \"\"\"Reads accession numbers from the input CSV file.\"\"\"\n",
        "    df = pd.read_csv(input_csv_file)\n",
        "    accession_numbers = df['Accession'].dropna().unique().tolist()\n",
        "    return accession_numbers\n",
        "\n",
        "def find_first_csv(directory):\n",
        "    \"\"\"Finds the first CSV file in the given directory and returns its full path.\"\"\"\n",
        "    for file in os.listdir(directory):\n",
        "        if file.endswith('.csv'):\n",
        "            return os.path.join(directory, file)\n",
        "    return None\n",
        "\n",
        "# Main workflow\n",
        "def main():\n",
        "    domain_library_file = '/content/output/libraries/domainlibrary.tsv'\n",
        "    input_directory = '/content/input'\n",
        "    input_csv_file = find_first_csv(input_directory)\n",
        "\n",
        "    if not input_csv_file:\n",
        "        write_log(log_file, \"No input CSV file found.\")\n",
        "        return\n",
        "\n",
        "    write_log(log_file, f\"Found input CSV file: {input_csv_file}\")\n",
        "\n",
        "    # Read domain information\n",
        "    write_log(log_file, \"Reading domain library...\")\n",
        "    domains = read_domain_library(domain_library_file)\n",
        "    write_log(log_file, f\"Domain library contains {len(domains)} entries.\")\n",
        "\n",
        "    # Read accession numbers from the input CSV file\n",
        "    accession_numbers = read_accession_numbers(input_csv_file)\n",
        "    write_log(log_file, f\"Input CSV contains {len(accession_numbers)} unique accession numbers.\")\n",
        "\n",
        "    # Read modification information from the modlist file\n",
        "    modlist_file = None\n",
        "    for filename in os.listdir('/content/output/modlist_files'):\n",
        "        if filename.startswith('all_modlist_Abundance') and filename.endswith('.tsv'):\n",
        "            modlist_file = os.path.join('/content/output/modlist_files', filename)\n",
        "            break\n",
        "\n",
        "    if not modlist_file:\n",
        "        write_log(log_file, \"No modlist file found.\")\n",
        "        return\n",
        "\n",
        "    write_log(log_file, f\"Found modlist file: {modlist_file}\")\n",
        "\n",
        "    # Read modification information from the modlist file\n",
        "    modifications = read_modifications(modlist_file)\n",
        "    write_log(log_file, f\"Read modifications for {len(modifications)} accessions.\")\n",
        "\n",
        "    # Fetch disorder info and save to files\n",
        "    saved_files = 0\n",
        "    failed_accessions = []\n",
        "\n",
        "    with ThreadPoolExecutor(max_workers=10) as executor:\n",
        "        futures = {executor.submit(fetch_disorder_info, accession): accession for accession in accession_numbers}\n",
        "\n",
        "        for future in tqdm(as_completed(futures), total=len(accession_numbers), desc=\"Fetching disorder info\"):\n",
        "            accession = futures[future]\n",
        "            try:\n",
        "                disorder_info = future.result()\n",
        "                if disorder_info:\n",
        "                    save_disorder_info(accession, disorder_info)\n",
        "                    saved_files += 1\n",
        "                else:\n",
        "                    failed_accessions.append(accession)\n",
        "            except Exception as e:\n",
        "                write_log(log_file, f\"Error fetching disorder info for {accession}: {e}\")\n",
        "                failed_accessions.append(accession)\n",
        "\n",
        "    # Print the number of .txt files saved and any that failed\n",
        "    write_log(log_file, f\"Number of .txt files saved: {saved_files}\")\n",
        "    write_log(log_file, f\"Number of accessions that failed: {len(failed_accessions)}\")\n",
        "    if failed_accessions:\n",
        "        write_log(log_file, f\"Failed accessions: {', '.join(failed_accessions)}\")\n",
        "\n",
        "    # Process disorder files in batches\n",
        "    disorder_files = [os.path.join(disorder_files_dir, f) for f in os.listdir(disorder_files_dir) if f.endswith('_iupred2a.txt')]\n",
        "    write_log(log_file, f\"Found {len(disorder_files)} disorder files to process.\")\n",
        "\n",
        "    batch_size = 50  # Adjust batch size according to available memory\n",
        "    for i in tqdm(range(0, len(disorder_files), batch_size), desc=\"Processing disorder files\"):\n",
        "        batch_files = disorder_files[i:i + batch_size]\n",
        "        for filepath in batch_files:\n",
        "            try:\n",
        "                process_disorder_file(filepath, domains, modifications)\n",
        "            except Exception as e:\n",
        "                write_log(log_file, f\"Error processing {filepath}: {e}\")\n",
        "        gc.collect()  # Collect garbage to free up memory after processing each batch\n",
        "\n",
        "    write_log(log_file, \"Processing complete.\")\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    main()\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XufXocRvaa4s"
      },
      "source": [
        "# **Plots and statistical analysis**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "cellView": "form",
        "id": "d-RW3bA6tEyj",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "3fa59cd3-6d41-40ec-fdb3-88491e54f40a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Skipping I3MEV1: Plot already exists.\n",
            "Skipping I3MSK5: Plot already exists.\n",
            "Skipping A0A287D6Y4: Plot already exists.\n",
            "Skipping A0A287D1I2: Plot already exists.\n",
            "Skipping I3N082: Plot already exists.\n",
            "Skipping I3M8C8: Plot already exists.\n",
            "Skipping A0A287D699: Plot already exists.\n",
            "Skipping I3NBE9: Plot already exists.\n",
            "Skipping I3N240: Plot already exists.\n",
            "Skipping A0A287DEP0: Plot already exists.\n",
            "Skipping I3NB21: Plot already exists.\n",
            "Skipping I3MVD2: Plot already exists.\n",
            "Skipping A0A287D768: Plot already exists.\n",
            "Skipping I3M1C8: Plot already exists.\n",
            "Skipping A0A287D036: Plot already exists.\n",
            "Skipping I3LVX4: Plot already exists.\n",
            "Skipping I3LXZ5: Plot already exists.\n",
            "Skipping I3N4D0: Plot already exists.\n",
            "Skipping I3M801: Plot already exists.\n",
            "Skipping I3MKN0: Plot already exists.\n",
            "Skipping I3N8D8: Plot already exists.\n",
            "Skipping I3NHP2: Plot already exists.\n",
            "Skipping I3MJG5: Plot already exists.\n",
            "Skipping I3MA25: Plot already exists.\n",
            "Skipping I3M058: Plot already exists.\n",
            "Skipping I3MZS8: Plot already exists.\n",
            "Skipping I3MP58: Plot already exists.\n",
            "Skipping I3MD18: Plot already exists.\n",
            "Skipping A0A287CUN7: Plot already exists.\n",
            "Skipping A0A287CW45: Plot already exists.\n",
            "Skipping I3MF67: Plot already exists.\n",
            "Skipping I3M3M5: Plot already exists.\n",
            "Skipping I3M8J0: Plot already exists.\n",
            "Skipping I3MHJ7: Plot already exists.\n",
            "Skipping A0A287D3E9: Plot already exists.\n",
            "Skipping I3MFA4: Plot already exists.\n",
            "Skipping I3MMF6: Plot already exists.\n",
            "Skipping I3MXL6: Plot already exists.\n",
            "Skipping A0A287D166: Plot already exists.\n",
            "Skipping I3NGJ8: Plot already exists.\n",
            "Skipping I3MSI1: Plot already exists.\n",
            "Skipping A0A287D4M5: Plot already exists.\n",
            "Skipping I3MXY3: Plot already exists.\n",
            "Skipping I3MRJ9: Plot already exists.\n",
            "Skipping I3N3L7: Plot already exists.\n",
            "Skipping I3MC52: Plot already exists.\n",
            "Skipping I3LY75: Plot already exists.\n",
            "Skipping I3MNA2: Plot already exists.\n",
            "Skipping I3MR22: Plot already exists.\n",
            "Skipping I3M7C6: Plot already exists.\n",
            "1% complete\n",
            "Skipping I3NHX0: Plot already exists.\n",
            "Skipping I3NCU1: Plot already exists.\n",
            "Skipping I3NAE1: Plot already exists.\n",
            "Skipping I3ND87: Plot already exists.\n",
            "Skipping A0A287CWI8: Plot already exists.\n",
            "Skipping A0A287CTR1: Plot already exists.\n",
            "Skipping I3N1E9: Plot already exists.\n",
            "Skipping I3MEU1: Plot already exists.\n",
            "Skipping I3MH39: Plot already exists.\n",
            "Skipping I3M158: Plot already exists.\n",
            "Skipping I3M3Z5: Plot already exists.\n",
            "Skipping I3MWA4: Plot already exists.\n",
            "Skipping I3N8D6: Plot already exists.\n",
            "Skipping I3NFK0: Plot already exists.\n",
            "Skipping I3M3Q4: Plot already exists.\n",
            "Skipping I3N658: Plot already exists.\n",
            "Skipping I3MN68: Plot already exists.\n",
            "Skipping I3M427: Plot already exists.\n",
            "Skipping I3MCQ4: Plot already exists.\n",
            "Skipping A0A287CTL6: Plot already exists.\n",
            "Skipping I3MW52: Plot already exists.\n",
            "Skipping A0A287DE54: Plot already exists.\n",
            "Skipping I3M257: Plot already exists.\n",
            "Skipping I3LWY7: Plot already exists.\n",
            "Skipping A0A287CXB1: Plot already exists.\n",
            "Skipping A0A287CT11: Plot already exists.\n",
            "Skipping I3NEE0: Plot already exists.\n",
            "Skipping I3NGR3: Plot already exists.\n",
            "Skipping I3MV51: Plot already exists.\n",
            "Skipping I3NFM3: Plot already exists.\n",
            "Skipping I3M7W2: Plot already exists.\n",
            "Skipping I3N3J3: Plot already exists.\n",
            "Skipping I3MLA3: Plot already exists.\n",
            "Skipping I3MUH3: Plot already exists.\n",
            "Skipping A0A287D1J1: Plot already exists.\n",
            "Skipping A0A287D1I1: Plot already exists.\n",
            "Skipping I3MJW9: Plot already exists.\n",
            "Skipping I3M0V1: Plot already exists.\n",
            "Skipping I3MRQ1: Plot already exists.\n",
            "Skipping A0A287CRY7: Plot already exists.\n",
            "Skipping I3M8E8: Plot already exists.\n",
            "Skipping I3LXQ2: Plot already exists.\n",
            "Skipping I3NDF1: Plot already exists.\n",
            "Skipping A0A287CVF2: Plot already exists.\n",
            "Skipping I3N4U5: Plot already exists.\n",
            "Skipping A0A287CZ34: Plot already exists.\n",
            "Skipping A0A287DDA1: Plot already exists.\n",
            "Skipping I3M2Y3: Plot already exists.\n",
            "Skipping I3MXL4: Plot already exists.\n",
            "Skipping A0A287CXD0: Plot already exists.\n",
            "3% complete\n",
            "Skipping I3N5D7: Plot already exists.\n",
            "Skipping I3M826: Plot already exists.\n",
            "Skipping I3M6M0: Plot already exists.\n",
            "Skipping A0A287D5P2: Plot already exists.\n",
            "Skipping A0A287DB79: Plot already exists.\n",
            "Skipping I3MGY9: Plot already exists.\n",
            "Skipping A0A287DFK1: Plot already exists.\n",
            "Skipping I3MLW3: Plot already exists.\n",
            "Skipping A0A287CT90: Plot already exists.\n",
            "Skipping A0A287D5M7: Plot already exists.\n",
            "Skipping I3M511: Plot already exists.\n",
            "Skipping A0A287CYS9: Plot already exists.\n",
            "Skipping I3MBI5: Plot already exists.\n",
            "Skipping A0A287DCT9: Plot already exists.\n",
            "Skipping I3N486: Plot already exists.\n",
            "Skipping I3N6K8: Plot already exists.\n",
            "Skipping I3MXA8: Plot already exists.\n",
            "Skipping I3M3R9: Plot already exists.\n",
            "Skipping I3MTR1: Plot already exists.\n",
            "Skipping I3MWK2: Plot already exists.\n",
            "Skipping I3N688: Plot already exists.\n",
            "Skipping I3NDP7: Plot already exists.\n",
            "Skipping I3MDN9: Plot already exists.\n",
            "Skipping I3M6J1: Plot already exists.\n",
            "Skipping I3MVP5: Plot already exists.\n",
            "Skipping I3MLX9: Plot already exists.\n",
            "Skipping I3MDU3: Plot already exists.\n",
            "Skipping I3MQV7: Plot already exists.\n",
            "Skipping A0A287DCJ7: Plot already exists.\n",
            "Skipping A0A287D9I9: Plot already exists.\n",
            "Skipping I3MCH5: Plot already exists.\n",
            "Skipping I3MX94: Plot already exists.\n",
            "Skipping A0A287CWP4: Plot already exists.\n",
            "Skipping I3M8B3: Plot already exists.\n",
            "Skipping I3MJT1: Plot already exists.\n",
            "Skipping A0A287DDX0: Plot already exists.\n",
            "Skipping I3MDX0: Plot already exists.\n",
            "Skipping I3M0K3: Plot already exists.\n",
            "Skipping A0A287D2T2: Plot already exists.\n",
            "Skipping I3MFX4: Plot already exists.\n",
            "Skipping A0A287DFW0: Plot already exists.\n",
            "Skipping I3M001: Plot already exists.\n",
            "Skipping I3MVX0: Plot already exists.\n",
            "Skipping I3NB63: Plot already exists.\n",
            "Skipping I3MT50: Plot already exists.\n",
            "Skipping I3N5I4: Plot already exists.\n",
            "Skipping I3MCI5: Plot already exists.\n",
            "Skipping I3N466: Plot already exists.\n",
            "Skipping I3MPM4: Plot already exists.\n",
            "Skipping A0A287CSZ8: Plot already exists.\n",
            "4% complete\n",
            "Skipping I3M0V4: Plot already exists.\n",
            "Skipping I3MC28: Plot already exists.\n",
            "Skipping I3LYH3: Plot already exists.\n",
            "Skipping I3MHZ1: Plot already exists.\n",
            "Skipping I3LZE6: Plot already exists.\n",
            "Skipping I3MZM8: Plot already exists.\n",
            "Skipping I3M8G1: Plot already exists.\n",
            "Skipping I3MPL5: Plot already exists.\n",
            "Skipping I3MGR1: Plot already exists.\n",
            "Skipping A0A287D262: Plot already exists.\n",
            "Skipping I3MVN4: Plot already exists.\n",
            "Skipping A0A287D8F6: Plot already exists.\n",
            "Skipping I3M696: Plot already exists.\n",
            "Skipping I3LZ66: Plot already exists.\n",
            "Skipping A0A287CXF2: Plot already exists.\n",
            "Skipping I3MJT9: Plot already exists.\n",
            "Skipping I3MEB1: Plot already exists.\n",
            "Skipping I3MB40: Plot already exists.\n",
            "Skipping A0A287D797: Plot already exists.\n",
            "Skipping I3MD75: Plot already exists.\n",
            "Skipping I3M290: Plot already exists.\n",
            "Skipping I3MR60: Plot already exists.\n",
            "Skipping I3M657: Plot already exists.\n",
            "Skipping I3M2A4: Plot already exists.\n",
            "Skipping I3M1T1: Plot already exists.\n",
            "Skipping I3MM18: Plot already exists.\n",
            "Skipping I3MPC9: Plot already exists.\n",
            "Skipping I3N8U2: Plot already exists.\n",
            "Skipping I3M610: Plot already exists.\n",
            "Skipping I3LYZ6: Plot already exists.\n",
            "Skipping I3N6X3: Plot already exists.\n",
            "Skipping A0A287CVA0: Plot already exists.\n",
            "Skipping I3MHM3: Plot already exists.\n",
            "Skipping I3NG19: Plot already exists.\n",
            "Skipping I3MJI3: Plot already exists.\n",
            "Skipping I3M5V9: Plot already exists.\n",
            "Skipping I3M786: Plot already exists.\n",
            "Skipping I3LXL9: Plot already exists.\n",
            "Skipping A0A287D723: Plot already exists.\n",
            "Skipping I3MS39: Plot already exists.\n",
            "Skipping I3MR54: Plot already exists.\n",
            "Skipping I3M1V3: Plot already exists.\n",
            "Skipping I3M1B5: Plot already exists.\n",
            "Skipping I3LYR4: Plot already exists.\n",
            "Skipping A0A287CUX5: Plot already exists.\n",
            "Skipping I3M7D4: Plot already exists.\n",
            "Skipping I3M4K3: Plot already exists.\n",
            "Skipping I3M7F9: Plot already exists.\n",
            "Skipping I3M3E9: Plot already exists.\n",
            "Skipping I3M2D7: Plot already exists.\n",
            "6% complete\n",
            "Skipping A0A287D6B9: Plot already exists.\n",
            "Skipping A0A287CS82: Plot already exists.\n",
            "Skipping I3MA84: Plot already exists.\n",
            "Skipping A0A287D826: Plot already exists.\n",
            "Skipping A0A287CTR9: Plot already exists.\n",
            "Skipping I3M8Y3: Plot already exists.\n",
            "Skipping I3ML72: Plot already exists.\n",
            "Skipping I3N6G6: Plot already exists.\n",
            "Skipping I3MXC7: Plot already exists.\n",
            "Skipping I3MQG8: Plot already exists.\n",
            "Skipping I3N3N1: Plot already exists.\n",
            "Skipping A0A287CZG7: Plot already exists.\n",
            "Skipping I3MC48: Plot already exists.\n",
            "Skipping I3MIP2: Plot already exists.\n",
            "Skipping A0A287CUQ8: Plot already exists.\n",
            "Skipping I3LZI7: Plot already exists.\n",
            "Skipping I3NHD2: Plot already exists.\n",
            "Skipping I3M4Y1: Plot already exists.\n",
            "Skipping A0A287DG17: Plot already exists.\n",
            "Skipping A0A287DGD0: Plot already exists.\n",
            "Skipping A0A287D1A8: Plot already exists.\n",
            "Skipping I3NBQ6: Plot already exists.\n",
            "Skipping I3MNN3: Plot already exists.\n",
            "Skipping I3M9F0: Plot already exists.\n",
            "Skipping A0A287DCK6: Plot already exists.\n",
            "Skipping I3N6G5: Plot already exists.\n",
            "Skipping I3MC61: Plot already exists.\n",
            "Skipping I3M479: Plot already exists.\n",
            "Skipping I3LY63: Plot already exists.\n",
            "Skipping I3LXP6: Plot already exists.\n",
            "Skipping I3MZ88: Plot already exists.\n",
            "Skipping I3NG99: Plot already exists.\n",
            "Skipping I3MEC9: Plot already exists.\n",
            "Skipping I3M0H9: Plot already exists.\n",
            "Skipping A0A287CZB6: Plot already exists.\n",
            "Skipping I3LX83: Plot already exists.\n",
            "Skipping I3MCX4: Plot already exists.\n",
            "Skipping I3M2W7: Plot already exists.\n",
            "Skipping I3NEG9: Plot already exists.\n",
            "Skipping I3M013: Plot already exists.\n",
            "Skipping A0A287D3H3: Plot already exists.\n",
            "Skipping I3LWA7: Plot already exists.\n",
            "Skipping I3M4N4: Plot already exists.\n",
            "Skipping I3MP42: Plot already exists.\n",
            "Skipping A0A287DAD3: Plot already exists.\n",
            "Skipping I3N2V9: Plot already exists.\n",
            "Skipping I3MDX6: Plot already exists.\n",
            "Skipping I3N654: Plot already exists.\n",
            "Skipping I3M629: Plot already exists.\n",
            "Skipping A0A287D401: Plot already exists.\n",
            "7% complete\n",
            "Skipping I3LY94: Plot already exists.\n",
            "Skipping I3MBP6: Plot already exists.\n",
            "Skipping I3NCX8: Plot already exists.\n",
            "Skipping I3MDY1: Plot already exists.\n",
            "Skipping I3M883: Plot already exists.\n",
            "Skipping I3M480: Plot already exists.\n",
            "Skipping I3MA61: Plot already exists.\n",
            "Skipping I3MJX2: Plot already exists.\n",
            "Skipping I3LW47: Plot already exists.\n",
            "Skipping A0A287CS86: Plot already exists.\n",
            "Skipping I3MSY3: Plot already exists.\n",
            "Skipping I3LZH6: Plot already exists.\n",
            "Skipping I3M8P8: Plot already exists.\n",
            "Skipping I3NGT1: Plot already exists.\n",
            "Skipping I3MN53: Plot already exists.\n",
            "Skipping I3MR64: Plot already exists.\n",
            "Skipping I3N9X5: Plot already exists.\n",
            "Skipping I3LWD8: Plot already exists.\n",
            "Skipping A0A287D107: Plot already exists.\n",
            "Skipping I3M8L2: Plot already exists.\n",
            "Skipping I3M8K3: Plot already exists.\n",
            "Skipping I3MJL5: Plot already exists.\n",
            "Skipping A0A287D008: Plot already exists.\n",
            "Skipping I3NG64: Plot already exists.\n",
            "Skipping I3LYU6: Plot already exists.\n",
            "Skipping I3NDB1: Plot already exists.\n",
            "Skipping I3MP66: Plot already exists.\n",
            "Skipping I3M4U0: Plot already exists.\n",
            "Skipping I3MKV4: Plot already exists.\n",
            "Skipping I3N402: Plot already exists.\n",
            "Skipping I3MD97: Plot already exists.\n",
            "Skipping I3ND54: Plot already exists.\n",
            "Skipping I3MCJ1: Plot already exists.\n",
            "Skipping I3MM53: Plot already exists.\n",
            "Skipping I3MGF9: Plot already exists.\n",
            "Skipping I3LYF3: Plot already exists.\n",
            "Skipping I3LXN2: Plot already exists.\n",
            "Skipping I3MGG0: Plot already exists.\n",
            "Skipping A0A287D258: Plot already exists.\n",
            "Skipping I3MMJ4: Plot already exists.\n",
            "Skipping A0A287D9J7: Plot already exists.\n",
            "Skipping A0A287DC75: Plot already exists.\n",
            "Skipping I3M730: Plot already exists.\n",
            "Skipping I3M2F2: Plot already exists.\n",
            "Skipping A0A287CS30: Plot already exists.\n",
            "Skipping A0A287CU44: Plot already exists.\n",
            "Skipping A0A287CSY2: Plot already exists.\n",
            "Skipping I3LX93: Plot already exists.\n",
            "Skipping I3MKU2: Plot already exists.\n",
            "Skipping I3MJU7: Plot already exists.\n",
            "9% complete\n",
            "Skipping I3M5Q0: Plot already exists.\n",
            "Skipping I3LWX7: Plot already exists.\n",
            "Skipping I3M1H0: Plot already exists.\n",
            "Skipping I3M4R0: Plot already exists.\n",
            "Skipping I3MVD6: Plot already exists.\n",
            "Skipping I3MQJ4: Plot already exists.\n",
            "Skipping A0A287DB06: Plot already exists.\n",
            "Skipping I3N4K9: Plot already exists.\n",
            "Skipping I3N206: Plot already exists.\n",
            "Skipping I3NB22: Plot already exists.\n",
            "Skipping I3MAB2: Plot already exists.\n",
            "Skipping I3NAV6: Plot already exists.\n",
            "Skipping A0A287CWM5: Plot already exists.\n",
            "Skipping I3LYE9: Plot already exists.\n",
            "Skipping A0A287D3A8: Plot already exists.\n",
            "Skipping I3MI89: Plot already exists.\n",
            "Skipping I3MC86: Plot already exists.\n",
            "Skipping I3MMK2: Plot already exists.\n",
            "Skipping I3MXL5: Plot already exists.\n",
            "Skipping I3ME32: Plot already exists.\n",
            "Skipping I3MH01: Plot already exists.\n",
            "Skipping I3M6Y2: Plot already exists.\n",
            "Skipping I3NEZ3: Plot already exists.\n",
            "Skipping I3MJA5: Plot already exists.\n",
            "Skipping I3ML61: Plot already exists.\n",
            "Skipping A0A287D839: Plot already exists.\n",
            "Skipping I3LWF9: Plot already exists.\n",
            "Skipping I3MD89: Plot already exists.\n",
            "Skipping I3MKP5: Plot already exists.\n",
            "Skipping I3MXU7: Plot already exists.\n",
            "Skipping I3N5I1: Plot already exists.\n",
            "Skipping I3NDU3: Plot already exists.\n",
            "Skipping I3LXB5: Plot already exists.\n",
            "Skipping I3MTC1: Plot already exists.\n",
            "Skipping I3MK23: Plot already exists.\n",
            "Skipping I3M8K7: Plot already exists.\n",
            "Skipping A0A287D6Z8: Plot already exists.\n",
            "Skipping A0A287CZI6: Plot already exists.\n",
            "Skipping A0A287DA20: Plot already exists.\n",
            "Skipping A0A287CZ94: Plot already exists.\n",
            "Skipping I3MN80: Plot already exists.\n",
            "Skipping I3MH95: Plot already exists.\n",
            "Skipping A0A287CZR6: Plot already exists.\n",
            "Skipping A0A287DG61: Plot already exists.\n",
            "Skipping A0A287CYW5: Plot already exists.\n",
            "Skipping I3MFS8: Plot already exists.\n",
            "Skipping I3LXD2: Plot already exists.\n",
            "Skipping I3M0A4: Plot already exists.\n",
            "Skipping I3MLH5: Plot already exists.\n",
            "Skipping A0A287CZY9: Plot already exists.\n",
            "10% complete\n",
            "Skipping I3NFA8: Plot already exists.\n",
            "Skipping A0A287CYL2: Plot already exists.\n",
            "Skipping I3M6J7: Plot already exists.\n",
            "Skipping I3M1K0: Plot already exists.\n",
            "Skipping A0A287D3P7: Plot already exists.\n",
            "Skipping I3MFY1: Plot already exists.\n",
            "Skipping I3MDR6: Plot already exists.\n",
            "Skipping I3NE16: Plot already exists.\n",
            "Skipping I3NBS6: Plot already exists.\n",
            "Skipping I3N5G1: Plot already exists.\n",
            "Skipping I3M6E5: Plot already exists.\n",
            "Skipping A0A287D5N6: Plot already exists.\n",
            "Skipping I3N6E7: Plot already exists.\n",
            "Skipping I3MCD2: Plot already exists.\n",
            "Skipping I3MCW1: Plot already exists.\n",
            "Skipping I3NAU5: Plot already exists.\n",
            "Skipping I3MSA6: Plot already exists.\n",
            "Skipping I3M7Y0: Plot already exists.\n",
            "Skipping I3LVV7: Plot already exists.\n",
            "Skipping A0A287D6T9: Plot already exists.\n",
            "Skipping I3MQI2: Plot already exists.\n",
            "Skipping I3MN56: Plot already exists.\n",
            "Skipping I3M830: Plot already exists.\n",
            "Skipping I3NEF3: Plot already exists.\n",
            "Skipping I3ME53: Plot already exists.\n",
            "Skipping I3MM63: Plot already exists.\n",
            "Skipping A0A287CS89: Plot already exists.\n",
            "Skipping I3MLN4: Plot already exists.\n",
            "Skipping A0A287CVU1: Plot already exists.\n",
            "Skipping A0A287D452: Plot already exists.\n",
            "Skipping I3MM92: Plot already exists.\n",
            "Skipping I3M3W6: Plot already exists.\n",
            "Skipping A0A287CYP8: Plot already exists.\n",
            "Skipping I3M658: Plot already exists.\n",
            "Skipping I3NEJ7: Plot already exists.\n",
            "Skipping A0A287D0I4: Plot already exists.\n",
            "Skipping I3MZE5: Plot already exists.\n",
            "Skipping I3M3W7: Plot already exists.\n",
            "Skipping I3MV56: Plot already exists.\n",
            "Skipping I3MRY3: Plot already exists.\n",
            "Skipping I3NF79: Plot already exists.\n",
            "Skipping I3MYB1: Plot already exists.\n",
            "Skipping I3MH59: Plot already exists.\n",
            "Skipping I3MJN6: Plot already exists.\n",
            "Skipping I3MYX1: Plot already exists.\n",
            "Skipping I3M4X7: Plot already exists.\n",
            "Skipping I3LZ12: Plot already exists.\n",
            "Skipping I3MC24: Plot already exists.\n",
            "Skipping I3MRY7: Plot already exists.\n",
            "Skipping I3M6N7: Plot already exists.\n",
            "12% complete\n",
            "Skipping A0A287CWW4: Plot already exists.\n",
            "Skipping A0A287CY93: Plot already exists.\n",
            "Skipping I3MV48: Plot already exists.\n",
            "Skipping I3NEF1: Plot already exists.\n",
            "Skipping A0A287DE13: Plot already exists.\n",
            "Skipping A0A287D8U9: Plot already exists.\n",
            "Skipping I3LY76: Plot already exists.\n",
            "Skipping I3NE97: Plot already exists.\n",
            "Skipping I3MT36: Plot already exists.\n",
            "Skipping I3M181: Plot already exists.\n",
            "Skipping I3LYL7: Plot already exists.\n",
            "Skipping I3MCU1: Plot already exists.\n",
            "Skipping I3MDG2: Plot already exists.\n",
            "Skipping I3MKM4: Plot already exists.\n",
            "Skipping I3MD49: Plot already exists.\n",
            "Skipping I3LZJ8: Plot already exists.\n",
            "Skipping A0A287CVM9: Plot already exists.\n",
            "Skipping I3MV41: Plot already exists.\n",
            "Skipping I3M5E7: Plot already exists.\n",
            "Skipping A0A287DFM2: Plot already exists.\n",
            "Skipping I3MZA2: Plot already exists.\n",
            "Skipping I3MXG1: Plot already exists.\n",
            "Skipping I3M5D6: Plot already exists.\n",
            "Skipping I3MKT3: Plot already exists.\n",
            "Skipping A0A287DCD3: Plot already exists.\n",
            "Skipping I3MEG9: Plot already exists.\n",
            "Skipping I3M584: Plot already exists.\n",
            "Skipping I3NGS3: Plot already exists.\n",
            "Skipping I3MMN5: Plot already exists.\n",
            "Skipping I3MCW4: Plot already exists.\n",
            "Skipping A0A287D0U3: Plot already exists.\n",
            "Skipping I3N0W5: Plot already exists.\n",
            "Skipping A0A287DG79: Plot already exists.\n",
            "Skipping I3MIG8: Plot already exists.\n",
            "Skipping I3M6F5: Plot already exists.\n",
            "Skipping A0A287DDY7: Plot already exists.\n",
            "Skipping I3M0H1: Plot already exists.\n",
            "Skipping A0A287D8Y1: Plot already exists.\n",
            "Skipping I3M698: Plot already exists.\n",
            "Skipping I3MYA6: Plot already exists.\n",
            "Skipping I3MYV9: Plot already exists.\n",
            "Skipping I3N294: Plot already exists.\n",
            "Skipping A0A287D260: Plot already exists.\n",
            "Skipping A0A287DC56: Plot already exists.\n",
            "Skipping I3N3U3: Plot already exists.\n",
            "Skipping I3MNZ8: Plot already exists.\n",
            "Skipping A0A287D651: Plot already exists.\n",
            "Skipping I3NBV6: Plot already exists.\n",
            "Skipping I3LZW0: Plot already exists.\n",
            "Skipping I3NGW2: Plot already exists.\n",
            "14% complete\n",
            "Skipping I3N7W8: Plot already exists.\n",
            "Skipping I3LXJ2: Plot already exists.\n",
            "Skipping I3MWX2: Plot already exists.\n",
            "Skipping I3MCG3: Plot already exists.\n",
            "Skipping I3NF97: Plot already exists.\n",
            "Skipping I3MHK2: Plot already exists.\n",
            "Skipping I3M1B9: Plot already exists.\n",
            "Skipping I3N2M7: Plot already exists.\n",
            "Skipping A0A287DGC7: Plot already exists.\n",
            "Skipping I3MQN9: Plot already exists.\n",
            "Skipping A0A287D8S1: Plot already exists.\n",
            "Skipping A0A287D3E5: Plot already exists.\n",
            "Skipping I3MGR4: Plot already exists.\n",
            "Skipping I3N7S7: Plot already exists.\n",
            "Skipping I3MMF3: Plot already exists.\n",
            "Skipping I3MDF0: Plot already exists.\n",
            "Skipping I3MKI3: Plot already exists.\n",
            "Skipping A0A287CVG5: Plot already exists.\n",
            "Skipping I3NAN5: Plot already exists.\n",
            "Skipping I3MDM4: Plot already exists.\n",
            "Skipping I3N8E6: Plot already exists.\n",
            "Skipping I3M8P9: Plot already exists.\n",
            "Skipping A0A287D265: Plot already exists.\n",
            "Skipping I3M5M4: Plot already exists.\n",
            "Skipping I3N2J3: Plot already exists.\n",
            "Skipping I3M514: Plot already exists.\n",
            "Skipping I3ND75: Plot already exists.\n",
            "Skipping A0A287DDG5: Plot already exists.\n",
            "Skipping I3M227: Plot already exists.\n",
            "Skipping I3MIW3: Plot already exists.\n",
            "Skipping I3LYQ1: Plot already exists.\n",
            "Skipping A0A287D7J8: Plot already exists.\n",
            "Skipping A0A287CYV6: Plot already exists.\n",
            "Skipping I3ME15: Plot already exists.\n",
            "Skipping I3NB96: Plot already exists.\n",
            "Skipping I3M3N9: Plot already exists.\n",
            "Skipping I3LXP2: Plot already exists.\n",
            "Skipping I3MQA9: Plot already exists.\n",
            "Skipping A0A287D5S2: Plot already exists.\n",
            "Skipping A0A287CTM0: Plot already exists.\n",
            "Skipping I3N0V5: Plot already exists.\n",
            "Skipping A0A287D4L3: Plot already exists.\n",
            "Skipping I3MAA4: Plot already exists.\n",
            "Skipping A0A287D4I6: Plot already exists.\n",
            "Skipping A0A287D070: Plot already exists.\n",
            "Skipping A0A287CYR0: Plot already exists.\n",
            "Skipping I3MII7: Plot already exists.\n",
            "Skipping I3M6B0: Plot already exists.\n",
            "Skipping A0A287CVS2: Plot already exists.\n",
            "Skipping I3M8G9: Plot already exists.\n",
            "15% complete\n",
            "Skipping I3MLB8: Plot already exists.\n",
            "Skipping A0A287CU00: Plot already exists.\n",
            "Skipping I3ND43: Plot already exists.\n",
            "Skipping A0A287CUW2: Plot already exists.\n",
            "Skipping I3NEV3: Plot already exists.\n",
            "Skipping A0A287D069: Plot already exists.\n",
            "Skipping A0A287D7C1: Plot already exists.\n",
            "Skipping A0A287DF63: Plot already exists.\n",
            "Skipping I3MZ82: Plot already exists.\n",
            "Skipping I3LWE8: Plot already exists.\n",
            "Skipping I3N2T6: Plot already exists.\n",
            "Skipping A0A287CVH4: Plot already exists.\n",
            "Skipping I3MT45: Plot already exists.\n",
            "Skipping I3MR53: Plot already exists.\n",
            "Skipping I3M856: Plot already exists.\n",
            "Skipping A0A287CZB0: Plot already exists.\n",
            "Skipping I3M4C6: Plot already exists.\n",
            "Skipping I3MXA5: Plot already exists.\n",
            "Skipping I3M1M2: Plot already exists.\n",
            "Skipping A0A287DGA4: Plot already exists.\n",
            "Skipping I3MD71: Plot already exists.\n",
            "Skipping I3LZE4: Plot already exists.\n",
            "Skipping I3M9F3: Plot already exists.\n",
            "Skipping I3M202: Plot already exists.\n",
            "Skipping A0A287CUZ5: Plot already exists.\n",
            "Skipping I3MNF6: Plot already exists.\n",
            "Skipping I3NGE0: Plot already exists.\n",
            "Skipping I3NA13: Plot already exists.\n",
            "Skipping A0A287CT91: Plot already exists.\n",
            "Skipping I3N834: Plot already exists.\n",
            "Skipping I3NDG8: Plot already exists.\n",
            "Skipping I3MLF0: Plot already exists.\n",
            "Skipping I3LZL7: Plot already exists.\n",
            "Skipping I3N2L1: Plot already exists.\n",
            "Skipping I3LZS1: Plot already exists.\n",
            "Skipping I3N7B3: Plot already exists.\n",
            "Skipping I3MU62: Plot already exists.\n",
            "Skipping A0A287CWB3: Plot already exists.\n",
            "Skipping I3N9X9: Plot already exists.\n",
            "Skipping A0A287CTQ0: Plot already exists.\n",
            "Skipping A0A287D5U4: Plot already exists.\n",
            "Skipping A0A287CU37: Plot already exists.\n",
            "Skipping I3MM30: Plot already exists.\n",
            "Skipping I3MY03: Plot already exists.\n",
            "Skipping I3MZ87: Plot already exists.\n",
            "Skipping I3MC41: Plot already exists.\n",
            "Skipping I3MPB0: Plot already exists.\n",
            "Skipping I3M4Y8: Plot already exists.\n",
            "Skipping I3MQ21: Plot already exists.\n",
            "Skipping A0A287CW68: Plot already exists.\n",
            "17% complete\n",
            "Skipping I3MBK9: Plot already exists.\n",
            "Skipping I3MCW8: Plot already exists.\n",
            "Skipping I3MIT2: Plot already exists.\n",
            "Skipping I3MSS5: Plot already exists.\n",
            "Skipping A0A287DB91: Plot already exists.\n",
            "Skipping I3NA98: Plot already exists.\n",
            "Skipping I3NG13: Plot already exists.\n",
            "Skipping I3MZ56: Plot already exists.\n",
            "Skipping I3NGZ0: Plot already exists.\n",
            "Skipping I3LZJ7: Plot already exists.\n",
            "Skipping A0A287D5I3: Plot already exists.\n",
            "Skipping I3MJF8: Plot already exists.\n",
            "Skipping I3M0P3: Plot already exists.\n",
            "Skipping I3N5Z1: Plot already exists.\n",
            "Skipping I3LWM0: Plot already exists.\n",
            "Skipping A0A287CYW4: Plot already exists.\n",
            "Skipping A0A287CS72: Plot already exists.\n",
            "Skipping I3LZ68: Plot already exists.\n",
            "Skipping I3M6H8: Plot already exists.\n",
            "Skipping I3MQI4: Plot already exists.\n",
            "Skipping I3M3E3: Plot already exists.\n",
            "Skipping I3ML63: Plot already exists.\n",
            "Skipping I3M7J1: Plot already exists.\n",
            "Skipping I3MHS1: Plot already exists.\n",
            "Skipping I3N603: Plot already exists.\n",
            "Skipping I3MC92: Plot already exists.\n",
            "Skipping A0A287D1C5: Plot already exists.\n",
            "Skipping I3MF27: Plot already exists.\n",
            "Skipping I3N3M6: Plot already exists.\n",
            "Skipping I3MLB6: Plot already exists.\n",
            "Skipping I3NH81: Plot already exists.\n",
            "Skipping A0A287DAR6: Plot already exists.\n",
            "Skipping I3MFQ3: Plot already exists.\n",
            "Skipping I3M0K6: Plot already exists.\n",
            "Skipping I3MU89: Plot already exists.\n",
            "Skipping I3M5C1: Plot already exists.\n",
            "Skipping I3MG89: Plot already exists.\n",
            "Skipping I3ME43: Plot already exists.\n",
            "Skipping A0A287D3H6: Plot already exists.\n",
            "Skipping A0A287D5X2: Plot already exists.\n",
            "Skipping I3N3Z5: Plot already exists.\n",
            "Skipping I3MLG0: Plot already exists.\n",
            "Skipping I3NEE1: Plot already exists.\n",
            "Skipping I3LVY3: Plot already exists.\n",
            "Skipping I3MB26: Plot already exists.\n",
            "Skipping A0A287D2R2: Plot already exists.\n",
            "Skipping I3M6D8: Plot already exists.\n",
            "Skipping A0A287D7Y5: Plot already exists.\n",
            "Skipping I3ML17: Plot already exists.\n",
            "Skipping I3M627: Plot already exists.\n",
            "18% complete\n",
            "Skipping I3M5Y9: Plot already exists.\n",
            "Skipping I3NCD9: Plot already exists.\n",
            "Skipping I3LYC5: Plot already exists.\n",
            "Skipping A0A287D5V6: Plot already exists.\n",
            "Skipping I3M207: Plot already exists.\n",
            "Skipping A0A287CYW9: Plot already exists.\n",
            "Skipping I3ND92: Plot already exists.\n",
            "Skipping I3MLU2: Plot already exists.\n",
            "Skipping A0A287DEJ8: Plot already exists.\n",
            "Skipping I3N6Z0: Plot already exists.\n",
            "Skipping I3N4G6: Plot already exists.\n",
            "Skipping I3MCM6: Plot already exists.\n",
            "Skipping I3N4T4: Plot already exists.\n",
            "Skipping A0A287DCV3: Plot already exists.\n",
            "Skipping A0A287D3N7: Plot already exists.\n",
            "Skipping A0A287DC20: Plot already exists.\n",
            "Skipping A0A287CUR5: Plot already exists.\n",
            "Skipping A0A287DCE6: Plot already exists.\n",
            "Skipping I3MBV3: Plot already exists.\n",
            "Skipping I3MMP3: Plot already exists.\n",
            "Skipping I3M109: Plot already exists.\n",
            "Skipping I3MA92: Plot already exists.\n",
            "Skipping A0A287CS44: Plot already exists.\n",
            "Skipping I3MIQ7: Plot already exists.\n",
            "Skipping A0A287DA60: Plot already exists.\n",
            "Skipping I3LZY5: Plot already exists.\n",
            "Skipping A0A287DDZ8: Plot already exists.\n",
            "Skipping I3MYP2: Plot already exists.\n",
            "Skipping I3M0J2: Plot already exists.\n",
            "Skipping I3N7N4: Plot already exists.\n",
            "Skipping I3MIN6: Plot already exists.\n",
            "Skipping I3M3H9: Plot already exists.\n",
            "Skipping I3M7U0: Plot already exists.\n",
            "Skipping I3MF96: Plot already exists.\n",
            "Skipping I3MNQ6: Plot already exists.\n",
            "Skipping I3MHG0: Plot already exists.\n",
            "Skipping I3M5A5: Plot already exists.\n",
            "Skipping I3MDB0: Plot already exists.\n",
            "Skipping I3LVX5: Plot already exists.\n",
            "Skipping I3M8J3: Plot already exists.\n",
            "Skipping A0A287DF10: Plot already exists.\n",
            "Skipping I3MU41: Plot already exists.\n",
            "Skipping I3ND56: Plot already exists.\n",
            "Skipping I3N9E1: Plot already exists.\n",
            "Skipping I3LZE2: Plot already exists.\n",
            "Skipping I3LXC2: Plot already exists.\n",
            "Skipping I3N111: Plot already exists.\n",
            "Skipping I3MT89: Plot already exists.\n",
            "Skipping I3MSY2: Plot already exists.\n",
            "Skipping I3MNE1: Plot already exists.\n",
            "20% complete\n",
            "Skipping I3M613: Plot already exists.\n",
            "Skipping A0A287CU61: Plot already exists.\n",
            "Skipping I3LVZ4: Plot already exists.\n",
            "Skipping I3MJG7: Plot already exists.\n",
            "Skipping A0A287DER7: Plot already exists.\n",
            "Skipping I3MUB2: Plot already exists.\n",
            "Skipping I3MF37: Plot already exists.\n",
            "Skipping A0A287D4K2: Plot already exists.\n",
            "Skipping I3MEG6: Plot already exists.\n",
            "Skipping I3N7A8: Plot already exists.\n",
            "Skipping I3N5J8: Plot already exists.\n",
            "Skipping I3MPH9: Plot already exists.\n",
            "Skipping I3LXL2: Plot already exists.\n",
            "Skipping I3N461: Plot already exists.\n",
            "Skipping A0A287DFW9: Plot already exists.\n",
            "Skipping A0A287CZQ1: Plot already exists.\n",
            "Skipping I3MKL8: Plot already exists.\n",
            "Skipping I3NAU9: Plot already exists.\n",
            "Skipping A0A287D104: Plot already exists.\n",
            "Skipping I3N587: Plot already exists.\n",
            "Skipping A0A287DA98: Plot already exists.\n",
            "Skipping I3NGD4: Plot already exists.\n",
            "Skipping A0A287CXK1: Plot already exists.\n",
            "Skipping A0A287CTV2: Plot already exists.\n",
            "Skipping I3MCS0: Plot already exists.\n",
            "Skipping I3LWX2: Plot already exists.\n",
            "Skipping A0A287CZS8: Plot already exists.\n",
            "Skipping A0A287CUQ2: Plot already exists.\n",
            "Skipping I3MMM1: Plot already exists.\n",
            "Skipping I3MFN1: Plot already exists.\n",
            "Skipping I3MCG1: Plot already exists.\n",
            "Skipping I3LYS8: Plot already exists.\n",
            "Skipping I3M766: Plot already exists.\n",
            "Skipping A0A287CSH3: Plot already exists.\n",
            "Skipping A0A287DEP8: Plot already exists.\n",
            "Skipping I3MAG0: Plot already exists.\n",
            "Skipping I3MBH4: Plot already exists.\n",
            "Skipping I3LYG7: Plot already exists.\n",
            "Skipping I3MPQ9: Plot already exists.\n",
            "Skipping A0A287DB77: Plot already exists.\n",
            "Skipping I3NFU1: Plot already exists.\n",
            "Skipping A0A287CYX5: Plot already exists.\n",
            "Skipping I3ML12: Plot already exists.\n",
            "Skipping I3MMB1: Plot already exists.\n",
            "Skipping I3N7R3: Plot already exists.\n",
            "Skipping I3M694: Plot already exists.\n",
            "Skipping I3MM77: Plot already exists.\n",
            "Skipping I3N7F1: Plot already exists.\n",
            "Skipping I3N9C8: Plot already exists.\n",
            "Skipping I3LWC3: Plot already exists.\n",
            "21% complete\n",
            "Skipping A0A287D854: Plot already exists.\n",
            "Skipping A0A287D044: Plot already exists.\n",
            "Skipping A0A287CT51: Plot already exists.\n",
            "Skipping I3M8A4: Plot already exists.\n",
            "Skipping I3N2S3: Plot already exists.\n",
            "Skipping I3N613: Plot already exists.\n",
            "Skipping A0A287CUD2: Plot already exists.\n",
            "Skipping A0A287DEV9: Plot already exists.\n",
            "Skipping I3MLD0: Plot already exists.\n",
            "Skipping I3MHA6: Plot already exists.\n",
            "Skipping I3NG44: Plot already exists.\n",
            "Skipping A0A287D9V6: Plot already exists.\n",
            "Skipping I3N700: Plot already exists.\n",
            "Skipping I3M335: Plot already exists.\n",
            "Skipping I3M250: Plot already exists.\n",
            "Skipping I3MCF2: Plot already exists.\n",
            "Skipping A0A287DCU1: Plot already exists.\n",
            "Skipping I3M9M7: Plot already exists.\n",
            "Skipping I3MRC6: Plot already exists.\n",
            "Skipping I3LWY9: Plot already exists.\n",
            "Skipping A0A287CT88: Plot already exists.\n",
            "Skipping I3LXZ1: Plot already exists.\n",
            "Skipping I3MFM7: Plot already exists.\n",
            "Skipping A0A287D0F3: Plot already exists.\n",
            "Skipping A0A287D9K7: Plot already exists.\n",
            "Skipping I3NDC6: Plot already exists.\n",
            "Skipping A0A287D1C7: Plot already exists.\n",
            "Skipping I3N2V7: Plot already exists.\n",
            "Skipping I3N591: Plot already exists.\n",
            "Skipping I3N2Y4: Plot already exists.\n",
            "Skipping I3MGR7: Plot already exists.\n",
            "Skipping A0A287DFN3: Plot already exists.\n",
            "Skipping A0A287CVL5: Plot already exists.\n",
            "Skipping I3M7Z4: Plot already exists.\n",
            "Skipping I3M2G5: Plot already exists.\n",
            "Skipping I3N4C4: Plot already exists.\n",
            "Skipping I3MP13: Plot already exists.\n",
            "Skipping I3M4E3: Plot already exists.\n",
            "Skipping A0A287CSS9: Plot already exists.\n",
            "Skipping I3M719: Plot already exists.\n",
            "Skipping I3LYS1: Plot already exists.\n",
            "Skipping I3N3M3: Plot already exists.\n",
            "Skipping I3MHY0: Plot already exists.\n",
            "Skipping A0A287D1M2: Plot already exists.\n",
            "Skipping I3NC02: Plot already exists.\n",
            "Skipping A0A287D867: Plot already exists.\n",
            "Skipping I3MLM8: Plot already exists.\n",
            "Skipping I3LYM8: Plot already exists.\n",
            "Skipping I3ME20: Plot already exists.\n",
            "Skipping A0A287CX66: Plot already exists.\n",
            "23% complete\n",
            "Skipping A0A287DFL9: Plot already exists.\n",
            "Skipping I3NDY5: Plot already exists.\n",
            "Skipping I3M1L5: Plot already exists.\n",
            "Skipping A0A287DBH4: Plot already exists.\n",
            "Skipping I3LWZ2: Plot already exists.\n",
            "Skipping I3NAT9: Plot already exists.\n",
            "Skipping I3MWQ1: Plot already exists.\n",
            "Skipping A0A287D148: Plot already exists.\n",
            "Skipping I3N7M0: Plot already exists.\n",
            "Skipping I3MAD1: Plot already exists.\n",
            "Skipping A0A287D1S1: Plot already exists.\n",
            "Skipping A0A287CSR5: Plot already exists.\n",
            "Skipping I3NDZ2: Plot already exists.\n",
            "Skipping A0A287DGF4: Plot already exists.\n",
            "Skipping I3N3U6: Plot already exists.\n",
            "Skipping A0A287CXN3: Plot already exists.\n",
            "Skipping I3MJ71: Plot already exists.\n",
            "Skipping I3M7E0: Plot already exists.\n",
            "Skipping A0A287DF89: Plot already exists.\n",
            "Skipping A0A287DFX3: Plot already exists.\n",
            "Skipping I3M2Z5: Plot already exists.\n",
            "Skipping I3MZP1: Plot already exists.\n",
            "Skipping I3MTH3: Plot already exists.\n",
            "Skipping I3LXA5: Plot already exists.\n",
            "Skipping I3M969: Plot already exists.\n",
            "Skipping I3MM84: Plot already exists.\n",
            "Skipping I3M8E6: Plot already exists.\n",
            "Skipping I3MFG6: Plot already exists.\n",
            "Skipping I3N2Y7: Plot already exists.\n",
            "Skipping I3N8N6: Plot already exists.\n",
            "Skipping I3MNW7: Plot already exists.\n",
            "Skipping A0A287DFG0: Plot already exists.\n",
            "Skipping A0A287D9D3: Plot already exists.\n",
            "Skipping I3MRA0: Plot already exists.\n",
            "Skipping A0A287DC11: Plot already exists.\n",
            "Skipping A0A287DA87: Plot already exists.\n",
            "Skipping I3M3V4: Plot already exists.\n",
            "Skipping I3NH83: Plot already exists.\n",
            "Skipping A0A287CSE2: Plot already exists.\n",
            "Skipping I3MDQ4: Plot already exists.\n",
            "Skipping A0A287CYN5: Plot already exists.\n",
            "Skipping I3MKN2: Plot already exists.\n",
            "Skipping I3MZQ6: Plot already exists.\n",
            "Skipping I3LYT1: Plot already exists.\n",
            "Skipping I3LZX4: Plot already exists.\n",
            "Skipping I3MAJ2: Plot already exists.\n",
            "Skipping I3LY55: Plot already exists.\n",
            "Skipping A0A287CWQ8: Plot already exists.\n",
            "Skipping I3MPY8: Plot already exists.\n",
            "Skipping I3M6R0: Plot already exists.\n",
            "24% complete\n",
            "Skipping I3MFF3: Plot already exists.\n",
            "Skipping I3N3K3: Plot already exists.\n",
            "Skipping I3MTX7: Plot already exists.\n",
            "Skipping A0A287DBB4: Plot already exists.\n",
            "Skipping I3MB89: Plot already exists.\n",
            "Skipping I3MHV2: Plot already exists.\n",
            "Skipping I3NB44: Plot already exists.\n",
            "Skipping I3LWM1: Plot already exists.\n",
            "Skipping A0A287CZ44: Plot already exists.\n",
            "Skipping A0A287DEL1: Plot already exists.\n",
            "Skipping I3M5F9: Plot already exists.\n",
            "Skipping I3MV49: Plot already exists.\n",
            "Skipping A0A287CSI2: Plot already exists.\n",
            "Skipping I3N0V9: Plot already exists.\n",
            "Skipping I3MGN0: Plot already exists.\n",
            "Skipping I3N034: Plot already exists.\n",
            "Skipping A0A287DDM7: Plot already exists.\n",
            "Skipping A0A287D342: Plot already exists.\n",
            "Skipping I3MBK7: Plot already exists.\n",
            "Skipping I3N8X0: Plot already exists.\n",
            "Skipping I3MCM3: Plot already exists.\n",
            "Skipping I3N878: Plot already exists.\n",
            "Skipping A0A287D4M2: Plot already exists.\n",
            "Skipping I3MP75: Plot already exists.\n",
            "Skipping I3M399: Plot already exists.\n",
            "Skipping A0A287CSZ6: Plot already exists.\n",
            "Skipping A0A287D508: Plot already exists.\n",
            "Skipping I3LYK9: Plot already exists.\n",
            "Skipping I3MQJ3: Plot already exists.\n",
            "Skipping I3MHW1: Plot already exists.\n",
            "Skipping A0A287D8Z6: Plot already exists.\n",
            "Skipping I3M1G6: Plot already exists.\n",
            "Skipping I3M5L5: Plot already exists.\n",
            "Skipping A0A287CYN7: Plot already exists.\n",
            "Skipping I3MBY8: Plot already exists.\n",
            "Skipping I3M608: Plot already exists.\n",
            "Skipping A0A287D688: Plot already exists.\n",
            "Skipping I3MEQ9: Plot already exists.\n",
            "Skipping I3NF37: Plot already exists.\n",
            "Skipping A0A287DCP5: Plot already exists.\n",
            "Skipping I3MK80: Plot already exists.\n",
            "Skipping A0A287DFG1: Plot already exists.\n",
            "Skipping I3MBT8: Plot already exists.\n",
            "Skipping I3ME85: Plot already exists.\n",
            "Skipping I3M3X5: Plot already exists.\n",
            "Skipping I3N010: Plot already exists.\n",
            "Skipping I3MK45: Plot already exists.\n",
            "Skipping I3MJZ9: Plot already exists.\n",
            "Skipping I3LXT0: Plot already exists.\n",
            "Skipping I3M9Y5: Plot already exists.\n",
            "26% complete\n",
            "Skipping A0A287DFI0: Plot already exists.\n",
            "Skipping I3MJS7: Plot already exists.\n",
            "Skipping I3NHV7: Plot already exists.\n",
            "Skipping I3N814: Plot already exists.\n",
            "Skipping I3MMP0: Plot already exists.\n",
            "Skipping I3MC57: Plot already exists.\n",
            "Skipping I3M5K7: Plot already exists.\n",
            "Skipping A0A287CSD6: Plot already exists.\n",
            "Skipping A0A287D0G9: Plot already exists.\n",
            "Skipping I3N358: Plot already exists.\n",
            "Skipping I3N1R4: Plot already exists.\n",
            "Skipping I3MT61: Plot already exists.\n",
            "Skipping A0A287CVG7: Plot already exists.\n",
            "Skipping I3M2H1: Plot already exists.\n",
            "Skipping I3MB14: Plot already exists.\n",
            "Skipping I3MRK2: Plot already exists.\n",
            "Skipping I3LWZ6: Plot already exists.\n",
            "Skipping I3MX42: Plot already exists.\n",
            "Skipping I3N5H5: Plot already exists.\n",
            "Skipping I3N068: Plot already exists.\n",
            "Skipping I3MQH3: Plot already exists.\n",
            "Skipping I3MFB2: Plot already exists.\n",
            "Skipping I3LW36: Plot already exists.\n",
            "Skipping I3N319: Plot already exists.\n",
            "Skipping I3MLL1: Plot already exists.\n",
            "Skipping I3M2N2: Plot already exists.\n",
            "Skipping I3MKM7: Plot already exists.\n",
            "Skipping I3MP82: Plot already exists.\n",
            "Skipping I3NHA4: Plot already exists.\n",
            "Skipping I3MIV4: Plot already exists.\n",
            "Skipping A0A287CU32: Plot already exists.\n",
            "Skipping A0A287D818: Plot already exists.\n",
            "Skipping I3MLC8: Plot already exists.\n",
            "Skipping I3MA51: Plot already exists.\n",
            "Skipping A0A287DD02: Plot already exists.\n",
            "Skipping I3M9Z2: Plot already exists.\n",
            "Skipping I3MT92: Plot already exists.\n",
            "Skipping I3N825: Plot already exists.\n",
            "Skipping I3M8D2: Plot already exists.\n",
            "Skipping I3MND5: Plot already exists.\n",
            "Skipping I3MF01: Plot already exists.\n",
            "Skipping I3N6Z7: Plot already exists.\n",
            "Skipping I3MLR0: Plot already exists.\n",
            "Skipping I3M6W8: Plot already exists.\n",
            "Skipping I3MER8: Plot already exists.\n",
            "Skipping I3M2H7: Plot already exists.\n",
            "Skipping I3N1X9: Plot already exists.\n",
            "Skipping I3MI17: Plot already exists.\n",
            "Skipping A0A287D2Y4: Plot already exists.\n",
            "Skipping I3MNH4: Plot already exists.\n",
            "28% complete\n",
            "Skipping I3MD14: Plot already exists.\n",
            "Skipping I3M5A2: Plot already exists.\n",
            "Skipping I3M5M9: Plot already exists.\n",
            "Skipping A0A287DDC5: Plot already exists.\n",
            "Skipping I3M573: Plot already exists.\n",
            "Skipping I3N4I8: Plot already exists.\n",
            "Skipping I3N9P3: Plot already exists.\n",
            "Skipping I3NBV1: Plot already exists.\n",
            "Skipping I3NBE7: Plot already exists.\n",
            "Skipping A0A287DBC8: Plot already exists.\n",
            "Skipping I3M4H9: Plot already exists.\n",
            "Skipping I3M3X4: Plot already exists.\n",
            "Skipping A0A287D3J0: Plot already exists.\n",
            "Skipping A0A287D2Z8: Plot already exists.\n",
            "Skipping A0A287DAX7: Plot already exists.\n",
            "Skipping I3LZW5: Plot already exists.\n",
            "Skipping I3LZP0: Plot already exists.\n",
            "Skipping I3M737: Plot already exists.\n",
            "Skipping I3LXM0: Plot already exists.\n",
            "Skipping I3MLZ0: Plot already exists.\n",
            "Skipping I3LYR1: Plot already exists.\n",
            "Skipping I3M774: Plot already exists.\n",
            "Skipping I3NCN5: Plot already exists.\n",
            "Skipping A0A287CYB5: Plot already exists.\n",
            "Skipping I3NAT5: Plot already exists.\n",
            "Skipping I3MYD5: Plot already exists.\n",
            "Skipping I3LZ83: Plot already exists.\n",
            "Skipping I3NGA9: Plot already exists.\n",
            "Skipping I3M5G7: Plot already exists.\n",
            "Skipping A0A287D6D1: Plot already exists.\n",
            "Skipping I3MCL1: Plot already exists.\n",
            "Skipping I3MQ89: Plot already exists.\n",
            "Skipping A0A287DDM5: Plot already exists.\n",
            "Skipping I3NAW6: Plot already exists.\n",
            "Skipping I3M124: Plot already exists.\n",
            "Skipping I3MIC0: Plot already exists.\n",
            "Skipping A0A287D7P8: Plot already exists.\n",
            "Skipping I3NB31: Plot already exists.\n",
            "Skipping A0A287D9Q9: Plot already exists.\n",
            "Skipping I3M8L9: Plot already exists.\n",
            "Skipping I3MLM5: Plot already exists.\n",
            "Skipping I3M3T6: Plot already exists.\n",
            "Skipping I3MYA2: Plot already exists.\n",
            "Skipping A0A287CY92: Plot already exists.\n",
            "Skipping I3N8X9: Plot already exists.\n",
            "Skipping I3N6S2: Plot already exists.\n",
            "Skipping A0A287DCE4: Plot already exists.\n",
            "Skipping A0A287DD67: Plot already exists.\n",
            "Skipping I3LY86: Plot already exists.\n",
            "Skipping I3MAM0: Plot already exists.\n",
            "29% complete\n",
            "Skipping A0A287D4J6: Plot already exists.\n",
            "Skipping I3MZR9: Plot already exists.\n",
            "Skipping I3MI59: Plot already exists.\n",
            "Skipping I3MFX9: Plot already exists.\n",
            "Skipping I3M0P6: Plot already exists.\n",
            "Skipping I3MMT7: Plot already exists.\n",
            "Skipping I3ND67: Plot already exists.\n",
            "Skipping I3MY15: Plot already exists.\n",
            "Skipping I3MYS1: Plot already exists.\n",
            "Skipping I3MX97: Plot already exists.\n",
            "Skipping I3MQL4: Plot already exists.\n",
            "Skipping A0A287D903: Plot already exists.\n",
            "Skipping A0A287D3M2: Plot already exists.\n",
            "Skipping I3M917: Plot already exists.\n",
            "Skipping I3M5D0: Plot already exists.\n",
            "Skipping I3N7W1: Plot already exists.\n",
            "Skipping A0A287D7N8: Plot already exists.\n",
            "Skipping I3N6S8: Plot already exists.\n",
            "Skipping I3N184: Plot already exists.\n",
            "Skipping A0A287CS90: Plot already exists.\n",
            "Skipping A0A287CRX6: Plot already exists.\n",
            "Skipping I3MS47: Plot already exists.\n",
            "Skipping A0A287DEU3: Plot already exists.\n",
            "Skipping A0A287CXC8: Plot already exists.\n",
            "Skipping I3N0V1: Plot already exists.\n",
            "Skipping I3MD23: Plot already exists.\n",
            "Skipping I3MK60: Plot already exists.\n",
            "Skipping I3MCW3: Plot already exists.\n",
            "Skipping I3NAC9: Plot already exists.\n",
            "Skipping A0A287D6E9: Plot already exists.\n",
            "Skipping A0A287D3E2: Plot already exists.\n",
            "Skipping I3N874: Plot already exists.\n",
            "Skipping I3MQS4: Plot already exists.\n",
            "Skipping I3MM20: Plot already exists.\n",
            "Skipping I3M1X9: Plot already exists.\n",
            "Skipping I3LZS4: Plot already exists.\n",
            "Skipping I3MCM7: Plot already exists.\n",
            "Skipping I3N7U9: Plot already exists.\n",
            "Skipping I3MA01: Plot already exists.\n",
            "Skipping A0A287D4R1: Plot already exists.\n",
            "Skipping A0A287D7T2: Plot already exists.\n",
            "Skipping I3MZJ0: Plot already exists.\n",
            "Skipping A0A287D7E2: Plot already exists.\n",
            "Skipping I3MMD0: Plot already exists.\n",
            "Skipping A0A287DDK4: Plot already exists.\n",
            "Skipping A0A287DBX4: Plot already exists.\n",
            "Skipping I3NFH6: Plot already exists.\n",
            "Skipping I3NF01: Plot already exists.\n",
            "Skipping A0A287D6W1: Plot already exists.\n",
            "Skipping I3MWV0: Plot already exists.\n",
            "31% complete\n",
            "Skipping I3N1C4: Plot already exists.\n",
            "Skipping I3LYT2: Plot already exists.\n",
            "Skipping I3MY55: Plot already exists.\n",
            "Skipping A0A287CYW2: Plot already exists.\n",
            "Skipping I3MR94: Plot already exists.\n",
            "Skipping I3M5W7: Plot already exists.\n",
            "Skipping A0A287D720: Plot already exists.\n",
            "Skipping I3LWB0: Plot already exists.\n",
            "Skipping I3MVX6: Plot already exists.\n",
            "Skipping I3LZW6: Plot already exists.\n",
            "Skipping I3MIY6: Plot already exists.\n",
            "Skipping I3NEN1: Plot already exists.\n",
            "Skipping I3M945: Plot already exists.\n",
            "Skipping I3MW73: Plot already exists.\n",
            "Skipping A0A287CVT8: Plot already exists.\n",
            "Skipping I3NCZ9: Plot already exists.\n",
            "Skipping I3LXJ1: Plot already exists.\n",
            "Skipping A0A287D526: Plot already exists.\n",
            "Skipping I3MDP6: Plot already exists.\n",
            "Skipping A0A287D559: Plot already exists.\n",
            "Skipping I3LWZ8: Plot already exists.\n",
            "Skipping I3N0M3: Plot already exists.\n",
            "Skipping I3N888: Plot already exists.\n",
            "Skipping I3M1N2: Plot already exists.\n",
            "Skipping I3M0A7: Plot already exists.\n",
            "Skipping I3LZE3: Plot already exists.\n",
            "Skipping I3MMC1: Plot already exists.\n",
            "Skipping I3MRW0: Plot already exists.\n",
            "Skipping I3M8K9: Plot already exists.\n",
            "Skipping A0A287D607: Plot already exists.\n",
            "Skipping I3MBS2: Plot already exists.\n",
            "Skipping I3N473: Plot already exists.\n",
            "Skipping A0A287DC35: Plot already exists.\n",
            "Skipping A0A287D0L6: Plot already exists.\n",
            "Skipping A0A287D3J6: Plot already exists.\n",
            "Skipping A0A287DBV6: Plot already exists.\n",
            "Skipping I3MPU3: Plot already exists.\n",
            "Skipping I3N9J3: Plot already exists.\n",
            "Skipping I3MIV7: Plot already exists.\n",
            "Skipping I3LZS2: Plot already exists.\n",
            "Skipping I3NB95: Plot already exists.\n",
            "Skipping A0A287D771: Plot already exists.\n",
            "Skipping A0A287DD97: Plot already exists.\n",
            "Skipping A0A287DB08: Plot already exists.\n",
            "Skipping A0A287D8T6: Plot already exists.\n",
            "Skipping I3MG17: Plot already exists.\n",
            "Skipping I3NHF5: Plot already exists.\n",
            "Skipping I3N3G2: Plot already exists.\n",
            "Skipping I3M9J9: Plot already exists.\n",
            "Skipping I3MGQ0: Plot already exists.\n",
            "32% complete\n",
            "Skipping I3MJR2: Plot already exists.\n",
            "Skipping I3MCA5: Plot already exists.\n",
            "Skipping I3LX77: Plot already exists.\n",
            "Skipping I3MB41: Plot already exists.\n",
            "Skipping A0A287D3X4: Plot already exists.\n",
            "Skipping A0A287CXP7: Plot already exists.\n",
            "Skipping A0A287CZF9: Plot already exists.\n",
            "Skipping I3MPR8: Plot already exists.\n",
            "Skipping I3MCX7: Plot already exists.\n",
            "Skipping I3MM35: Plot already exists.\n",
            "Skipping A0A287D009: Plot already exists.\n",
            "Skipping I3ME39: Plot already exists.\n",
            "Skipping A0A287D4M9: Plot already exists.\n",
            "Skipping A0A287CW89: Plot already exists.\n",
            "Skipping A0A287DE99: Plot already exists.\n",
            "Skipping I3MRK4: Plot already exists.\n",
            "Skipping I3N449: Plot already exists.\n",
            "Skipping I3MZ75: Plot already exists.\n",
            "Skipping I3MSJ6: Plot already exists.\n",
            "Skipping I3MQS7: Plot already exists.\n",
            "Skipping I3N017: Plot already exists.\n",
            "Skipping A0A287D1P0: Plot already exists.\n",
            "Skipping I3M9I5: Plot already exists.\n",
            "Skipping I3MR90: Plot already exists.\n",
            "Skipping I3MQB3: Plot already exists.\n",
            "Skipping I3M606: Plot already exists.\n",
            "Skipping I3MYV7: Plot already exists.\n",
            "Skipping I3MY19: Plot already exists.\n",
            "Skipping A0A287CUD7: Plot already exists.\n",
            "Skipping I3MK62: Plot already exists.\n",
            "Skipping I3MT80: Plot already exists.\n",
            "Skipping I3MC65: Plot already exists.\n",
            "Skipping I3NAW4: Plot already exists.\n",
            "Skipping I3LZU3: Plot already exists.\n",
            "Skipping I3MDH9: Plot already exists.\n",
            "Skipping I3ND88: Plot already exists.\n",
            "Skipping I3NFW1: Plot already exists.\n",
            "Skipping I3MBB1: Plot already exists.\n",
            "Skipping I3MR55: Plot already exists.\n",
            "Skipping I3MV05: Plot already exists.\n",
            "Skipping I3MLC0: Plot already exists.\n",
            "Skipping I3MXI4: Plot already exists.\n",
            "Skipping I3MB96: Plot already exists.\n",
            "Skipping I3LWR0: Plot already exists.\n",
            "Skipping A0A287DD82: Plot already exists.\n",
            "Skipping I3MTT0: Plot already exists.\n",
            "Skipping A0A287DGF5: Plot already exists.\n",
            "Skipping I3MG83: Plot already exists.\n",
            "Skipping I3MXB5: Plot already exists.\n",
            "Skipping A0A287D7A4: Plot already exists.\n",
            "34% complete\n",
            "Skipping I3MJR1: Plot already exists.\n",
            "Skipping I3MHZ8: Plot already exists.\n",
            "Skipping I3M412: Plot already exists.\n",
            "Skipping I3NFL2: Plot already exists.\n",
            "Skipping I3MIX7: Plot already exists.\n",
            "Skipping I3MLT3: Plot already exists.\n",
            "Skipping I3LXD5: Plot already exists.\n",
            "Skipping I3MQH1: Plot already exists.\n",
            "Skipping I3NGI2: Plot already exists.\n",
            "Skipping I3N9H3: Plot already exists.\n",
            "Skipping A0A287D8Q7: Plot already exists.\n",
            "Skipping A0A287CTE0: Plot already exists.\n",
            "Skipping I3MRQ9: Plot already exists.\n",
            "Skipping A0A287DAR9: Plot already exists.\n",
            "Skipping I3MJ55: Plot already exists.\n",
            "Skipping A0A287D0E9: Plot already exists.\n",
            "Skipping I3N7C1: Plot already exists.\n",
            "Skipping A0A287D2K4: Plot already exists.\n",
            "Skipping I3MNT5: Plot already exists.\n",
            "Skipping I3MSG5: Plot already exists.\n",
            "Skipping I3LYX4: Plot already exists.\n",
            "Skipping I3MDD3: Plot already exists.\n",
            "Skipping I3N2B3: Plot already exists.\n",
            "Skipping I3NGM6: Plot already exists.\n",
            "Skipping A0A287DAM7: Plot already exists.\n",
            "Skipping I3ML81: Plot already exists.\n",
            "Skipping I3LXV3: Plot already exists.\n",
            "Skipping A0A287CX35: Plot already exists.\n",
            "Skipping I3LYC4: Plot already exists.\n",
            "Skipping I3N9G1: Plot already exists.\n",
            "Skipping I3MI14: Plot already exists.\n",
            "Skipping A0A287D0H0: Plot already exists.\n",
            "Skipping I3MJT5: Plot already exists.\n",
            "Skipping A0A287D737: Plot already exists.\n",
            "Skipping A0A287DCV0: Plot already exists.\n",
            "Skipping I3N055: Plot already exists.\n",
            "Skipping I3M5W4: Plot already exists.\n",
            "Skipping I3MYW6: Plot already exists.\n",
            "Skipping I3NF07: Plot already exists.\n",
            "Skipping I3MLC5: Plot already exists.\n",
            "Skipping I3MWD7: Plot already exists.\n",
            "Skipping A0A287CZ75: Plot already exists.\n",
            "Skipping I3MGQ7: Plot already exists.\n",
            "Skipping I3MBZ9: Plot already exists.\n",
            "Skipping I3MKC3: Plot already exists.\n",
            "Skipping I3N420: Plot already exists.\n",
            "Skipping I3M8H9: Plot already exists.\n",
            "Skipping I3MR66: Plot already exists.\n",
            "Skipping I3M6Z3: Plot already exists.\n",
            "Skipping A0A287DFQ9: Plot already exists.\n",
            "35% complete\n",
            "Skipping I3N149: Plot already exists.\n",
            "Skipping I3N2I7: Plot already exists.\n",
            "Skipping I3M0M8: Plot already exists.\n",
            "Skipping A0A287D7Q0: Plot already exists.\n",
            "Skipping I3M688: Plot already exists.\n",
            "Skipping A0A287CUP2: Plot already exists.\n",
            "Skipping A0A287DFZ2: Plot already exists.\n",
            "Skipping I3M0T9: Plot already exists.\n",
            "Skipping A0A287DFA4: Plot already exists.\n",
            "Skipping I3N5K9: Plot already exists.\n",
            "Skipping I3M7G6: Plot already exists.\n",
            "Skipping I3N5F2: Plot already exists.\n",
            "Skipping I3MW57: Plot already exists.\n",
            "Skipping A0A287CXH2: Plot already exists.\n",
            "Skipping I3MFR6: Plot already exists.\n",
            "Skipping I3MD55: Plot already exists.\n",
            "Skipping I3MSM8: Plot already exists.\n",
            "Skipping I3MNI2: Plot already exists.\n",
            "Skipping I3MLV7: Plot already exists.\n",
            "Skipping I3MWF3: Plot already exists.\n",
            "Skipping A0A287CXN9: Plot already exists.\n",
            "Skipping I3N4Q1: Plot already exists.\n",
            "Skipping A0A287CVB3: Plot already exists.\n",
            "Skipping I3MGG6: Plot already exists.\n",
            "Skipping I3M512: Plot already exists.\n",
            "Skipping I3NHT5: Plot already exists.\n",
            "Skipping A0A287DE16: Plot already exists.\n",
            "Skipping A0A287DCX2: Plot already exists.\n",
            "Skipping I3MAX1: Plot already exists.\n",
            "Skipping I3MNB0: Plot already exists.\n",
            "Skipping I3M682: Plot already exists.\n",
            "Skipping I3LWU0: Plot already exists.\n",
            "Skipping I3N9K0: Plot already exists.\n",
            "Skipping I3MD33: Plot already exists.\n",
            "Skipping I3MT35: Plot already exists.\n",
            "Skipping I3NAB7: Plot already exists.\n",
            "Skipping A0A287DC24: Plot already exists.\n",
            "Skipping I3N4B3: Plot already exists.\n",
            "Skipping I3N2X9: Plot already exists.\n",
            "Skipping I3M8W0: Plot already exists.\n",
            "Skipping I3M3J2: Plot already exists.\n",
            "Skipping A0A287CTQ6: Plot already exists.\n",
            "Skipping I3MTW9: Plot already exists.\n",
            "Skipping I3MYV0: Plot already exists.\n",
            "Skipping I3M815: Plot already exists.\n",
            "Skipping A0A287CZK2: Plot already exists.\n",
            "Skipping I3MLD8: Plot already exists.\n",
            "Skipping I3MV73: Plot already exists.\n",
            "Skipping I3NBW4: Plot already exists.\n",
            "Skipping I3M3I9: Plot already exists.\n",
            "37% complete\n",
            "Skipping A0A287DB36: Plot already exists.\n",
            "Skipping I3M0Q3: Plot already exists.\n",
            "Skipping I3M5V4: Plot already exists.\n",
            "Skipping I3NA27: Plot already exists.\n",
            "Skipping A0A287D281: Plot already exists.\n",
            "Skipping A0A287D8X7: Plot already exists.\n",
            "Skipping I3MQJ1: Plot already exists.\n",
            "Skipping I3MEN1: Plot already exists.\n",
            "Skipping I3NDC2: Plot already exists.\n",
            "Skipping I3LVU5: Plot already exists.\n",
            "Skipping I3M5D7: Plot already exists.\n",
            "Skipping I3N988: Plot already exists.\n",
            "Skipping I3M2L7: Plot already exists.\n",
            "Skipping I3MHR7: Plot already exists.\n",
            "Skipping I3MPA6: Plot already exists.\n",
            "Skipping I3M1N6: Plot already exists.\n",
            "Skipping I3M1S9: Plot already exists.\n",
            "Skipping I3ME50: Plot already exists.\n",
            "Skipping A0A287D387: Plot already exists.\n",
            "Skipping A0A287DAB5: Plot already exists.\n",
            "Skipping I3ME51: Plot already exists.\n",
            "Skipping I3NBR9: Plot already exists.\n",
            "Skipping I3LWR2: Plot already exists.\n",
            "Skipping I3MDH7: Plot already exists.\n",
            "Skipping I3MQP1: Plot already exists.\n",
            "Skipping I3MSC3: Plot already exists.\n",
            "Skipping I3M650: Plot already exists.\n",
            "Skipping I3MC89: Plot already exists.\n",
            "Skipping I3LWX3: Plot already exists.\n",
            "Skipping I3N7Q6: Plot already exists.\n",
            "Skipping I3MC38: Plot already exists.\n",
            "Skipping I3MK66: Plot already exists.\n",
            "Skipping I3NCG9: Plot already exists.\n",
            "Skipping A0A287DBI9: Plot already exists.\n",
            "Skipping I3M6P1: Plot already exists.\n",
            "Skipping I3MFK1: Plot already exists.\n",
            "Skipping A0A287D210: Plot already exists.\n",
            "Skipping I3N191: Plot already exists.\n",
            "Skipping I3N6E0: Plot already exists.\n",
            "Skipping I3MAX3: Plot already exists.\n",
            "Skipping A0A287CVL4: Plot already exists.\n",
            "Skipping I3M0M0: Plot already exists.\n",
            "Skipping A0A287DAD6: Plot already exists.\n",
            "Skipping I3M2R7: Plot already exists.\n",
            "Skipping I3MQZ2: Plot already exists.\n",
            "Skipping I3M589: Plot already exists.\n",
            "Skipping A0A287DEG3: Plot already exists.\n",
            "Skipping I3N266: Plot already exists.\n",
            "Skipping I3MYJ3: Plot already exists.\n",
            "Skipping I3MMP6: Plot already exists.\n",
            "39% complete\n",
            "Skipping I3N2W9: Plot already exists.\n",
            "Skipping A0A287CZX7: Plot already exists.\n",
            "Skipping I3MBC2: Plot already exists.\n",
            "Skipping I3MQQ6: Plot already exists.\n",
            "Skipping I3N9G0: Plot already exists.\n",
            "Skipping I3N1B7: Plot already exists.\n",
            "Skipping I3N9F5: Plot already exists.\n",
            "Skipping I3MH45: Plot already exists.\n",
            "Skipping A0A287D4C0: Plot already exists.\n",
            "Skipping I3LW28: Plot already exists.\n",
            "Skipping I3N934: Plot already exists.\n",
            "Skipping A0A287CRR5: Plot already exists.\n",
            "Skipping I3N2M4: Plot already exists.\n",
            "Skipping I3MMI2: Plot already exists.\n",
            "Skipping I3N3B4: Plot already exists.\n",
            "Skipping I3M1R0: Plot already exists.\n",
            "Skipping I3LYJ4: Plot already exists.\n",
            "Skipping I3NG60: Plot already exists.\n",
            "Skipping I3MHS4: Plot already exists.\n",
            "Skipping A0A287DGA3: Plot already exists.\n",
            "Skipping A0A287D863: Plot already exists.\n",
            "Skipping A0A287D2I4: Plot already exists.\n",
            "Skipping A0A287D5P1: Plot already exists.\n",
            "Skipping A0A287CTL4: Plot already exists.\n",
            "Skipping I3MRR9: Plot already exists.\n",
            "Skipping A0A287CXP4: Plot already exists.\n",
            "Skipping I3MQU2: Plot already exists.\n",
            "Skipping I3NB47: Plot already exists.\n",
            "Skipping I3NEE2: Plot already exists.\n",
            "Skipping A0A287D0K7: Plot already exists.\n",
            "Skipping A0A287D3A0: Plot already exists.\n",
            "Skipping I3NA64: Plot already exists.\n",
            "Skipping A0A287D8R4: Plot already exists.\n",
            "Skipping I3NFM1: Plot already exists.\n",
            "Skipping A0A287D191: Plot already exists.\n",
            "Skipping I3MNC4: Plot already exists.\n",
            "Skipping I3MPT0: Plot already exists.\n",
            "Skipping I3MSN0: Plot already exists.\n",
            "Skipping I3LZW1: Plot already exists.\n",
            "Skipping I3MPK0: Plot already exists.\n",
            "Skipping I3NCB8: Plot already exists.\n",
            "Skipping I3MTC5: Plot already exists.\n",
            "Skipping A0A287CYJ9: Plot already exists.\n",
            "Skipping I3M472: Plot already exists.\n",
            "Skipping A0A287D239: Plot already exists.\n",
            "Skipping I3MD65: Plot already exists.\n",
            "Skipping I3ME54: Plot already exists.\n",
            "Skipping I3MDR3: Plot already exists.\n",
            "Skipping I3MBJ5: Plot already exists.\n",
            "Skipping I3M1I3: Plot already exists.\n",
            "40% complete\n",
            "Skipping I3N135: Plot already exists.\n",
            "Skipping A0A287D8N3: Plot already exists.\n",
            "Skipping I3N4Y6: Plot already exists.\n",
            "Skipping A0A287CYF9: Plot already exists.\n",
            "Skipping A0A287CSD9: Plot already exists.\n",
            "Skipping I3N4D8: Plot already exists.\n",
            "Skipping I3LWK7: Plot already exists.\n",
            "Skipping I3M6R4: Plot already exists.\n",
            "Skipping A0A287D1H7: Plot already exists.\n",
            "Skipping I3MCV7: Plot already exists.\n",
            "Skipping A0A287D1S3: Plot already exists.\n",
            "Skipping I3N9V0: Plot already exists.\n",
            "Skipping A0A287CVR4: Plot already exists.\n",
            "Skipping I3MQH2: Plot already exists.\n",
            "Skipping I3LX47: Plot already exists.\n",
            "Skipping A0A287DA94: Plot already exists.\n",
            "Skipping I3MN79: Plot already exists.\n",
            "Skipping I3NGZ8: Plot already exists.\n",
            "Skipping I3M6M2: Plot already exists.\n",
            "Skipping I3N585: Plot already exists.\n",
            "Skipping I3MEC5: Plot already exists.\n",
            "Skipping I3MRN5: Plot already exists.\n",
            "Skipping I3N7R9: Plot already exists.\n",
            "Skipping I3MLA5: Plot already exists.\n",
            "Skipping I3M588: Plot already exists.\n",
            "Skipping I3MKS2: Plot already exists.\n",
            "Skipping I3MS80: Plot already exists.\n",
            "Skipping I3MLG6: Plot already exists.\n",
            "Skipping I3MN11: Plot already exists.\n",
            "Skipping A0A287DD92: Plot already exists.\n",
            "Skipping I3MKN5: Plot already exists.\n",
            "Skipping I3MGX5: Plot already exists.\n",
            "Skipping I3M844: Plot already exists.\n",
            "Skipping I3MX54: Plot already exists.\n",
            "Skipping I3MUB7: Plot already exists.\n",
            "Skipping I3MGF5: Plot already exists.\n",
            "Skipping I3MGT2: Plot already exists.\n",
            "Skipping I3MLY4: Plot already exists.\n",
            "Skipping I3N487: Plot already exists.\n",
            "Skipping I3MRP9: Plot already exists.\n",
            "Skipping I3MQU7: Plot already exists.\n",
            "Skipping I3MC16: Plot already exists.\n",
            "Skipping I3LVY7: Plot already exists.\n",
            "Skipping I3MSH1: Plot already exists.\n",
            "Skipping A0A287D3F7: Plot already exists.\n",
            "Skipping I3LY06: Plot already exists.\n",
            "Skipping I3MTF2: Plot already exists.\n",
            "Skipping A0A287D3G6: Plot already exists.\n",
            "Skipping I3M8H5: Plot already exists.\n",
            "Skipping A0A287DBX1: Plot already exists.\n",
            "42% complete\n",
            "Skipping A0A287DC25: Plot already exists.\n",
            "Skipping I3MBH6: Plot already exists.\n",
            "Skipping I3N8B3: Plot already exists.\n",
            "Skipping I3M8G7: Plot already exists.\n",
            "Skipping I3MQ03: Plot already exists.\n",
            "Skipping I3NFR3: Plot already exists.\n",
            "Skipping A0A287D060: Plot already exists.\n",
            "Skipping I3M9X1: Plot already exists.\n",
            "Skipping I3MDX4: Plot already exists.\n",
            "Skipping I3M9L7: Plot already exists.\n",
            "Skipping A0A287D5Z8: Plot already exists.\n",
            "Skipping A0A287CYF1: Plot already exists.\n",
            "Skipping I3ML95: Plot already exists.\n",
            "Skipping A0A287CU31: Plot already exists.\n",
            "Skipping I3MQY3: Plot already exists.\n",
            "Skipping A0A287DCM2: Plot already exists.\n",
            "Skipping I3NB84: Plot already exists.\n",
            "Skipping I3MQU9: Plot already exists.\n",
            "Skipping I3MI30: Plot already exists.\n",
            "Skipping I3N5Y0: Plot already exists.\n",
            "Skipping A0A287D563: Plot already exists.\n",
            "Skipping I3MP99: Plot already exists.\n",
            "Skipping A0A287D6W9: Plot already exists.\n",
            "Skipping I3NF63: Plot already exists.\n",
            "Skipping I3MB80: Plot already exists.\n",
            "Skipping I3LZ54: Plot already exists.\n",
            "Skipping I3MKK1: Plot already exists.\n",
            "Skipping I3M2N5: Plot already exists.\n",
            "Skipping I3LZF5: Plot already exists.\n",
            "Skipping A0A287CZQ2: Plot already exists.\n",
            "Skipping I3ND96: Plot already exists.\n",
            "Skipping I3MAZ4: Plot already exists.\n",
            "Skipping I3M2S4: Plot already exists.\n",
            "Skipping I3N6E9: Plot already exists.\n",
            "Skipping I3M8W7: Plot already exists.\n",
            "Skipping I3MNW5: Plot already exists.\n",
            "Skipping I3MYY4: Plot already exists.\n",
            "Skipping I3N3M4: Plot already exists.\n",
            "Skipping I3N6Q4: Plot already exists.\n",
            "Skipping I3M7T1: Plot already exists.\n",
            "Skipping A0A287D8Q1: Plot already exists.\n",
            "Skipping I3MRF1: Plot already exists.\n",
            "Skipping I3MUG2: Plot already exists.\n",
            "Skipping I3N7P2: Plot already exists.\n",
            "Skipping I3NAI2: Plot already exists.\n",
            "Skipping A0A287CXU4: Plot already exists.\n",
            "Skipping I3MGU6: Plot already exists.\n",
            "Skipping I3MVN1: Plot already exists.\n",
            "Skipping A0A287CTA6: Plot already exists.\n",
            "Skipping I3MIW4: Plot already exists.\n",
            "43% complete\n",
            "Skipping I3MB62: Plot already exists.\n",
            "Skipping I3NCL0: Plot already exists.\n",
            "Skipping A0A287CXS7: Plot already exists.\n",
            "Skipping I3M268: Plot already exists.\n",
            "Skipping I3MQD1: Plot already exists.\n",
            "Skipping I3MJT8: Plot already exists.\n",
            "Skipping I3M245: Plot already exists.\n",
            "Skipping I3MB01: Plot already exists.\n",
            "Skipping I3MIU5: Plot already exists.\n",
            "Skipping I3M5M0: Plot already exists.\n",
            "Skipping I3MX36: Plot already exists.\n",
            "Skipping I3MMF0: Plot already exists.\n",
            "Skipping A0A287D7C6: Plot already exists.\n",
            "Skipping I3MEA3: Plot already exists.\n",
            "Skipping A0A287CZA8: Plot already exists.\n",
            "Skipping I3MTH7: Plot already exists.\n",
            "Skipping I3MAM2: Plot already exists.\n",
            "Skipping I3MNS1: Plot already exists.\n",
            "Skipping I3M126: Plot already exists.\n",
            "Skipping I3NCN6: Plot already exists.\n",
            "Skipping I3MAC4: Plot already exists.\n",
            "Skipping I3MU06: Plot already exists.\n",
            "Skipping I3N4H6: Plot already exists.\n",
            "Skipping A0A287D4M7: Plot already exists.\n",
            "Skipping I3NDV0: Plot already exists.\n",
            "Skipping I3M6G8: Plot already exists.\n",
            "Skipping I3MLP1: Plot already exists.\n",
            "Skipping A0A287CSK2: Plot already exists.\n",
            "Skipping A0A287CY90: Plot already exists.\n",
            "Skipping I3MV12: Plot already exists.\n",
            "Skipping I3N786: Plot already exists.\n",
            "Skipping I3M1U4: Plot already exists.\n",
            "Skipping I3MQF8: Plot already exists.\n",
            "Skipping I3MAM5: Plot already exists.\n",
            "Skipping I3MQE9: Plot already exists.\n",
            "Skipping I3MNJ8: Plot already exists.\n",
            "Skipping A0A287CWS9: Plot already exists.\n",
            "Skipping I3MRA8: Plot already exists.\n",
            "Skipping A0A287DAQ0: Plot already exists.\n",
            "Skipping I3M6L3: Plot already exists.\n",
            "Skipping I3MI04: Plot already exists.\n",
            "Skipping I3MXH8: Plot already exists.\n",
            "Skipping I3MTA4: Plot already exists.\n",
            "Skipping I3MCZ4: Plot already exists.\n",
            "Skipping A0A287DDY4: Plot already exists.\n",
            "Skipping I3LZ25: Plot already exists.\n",
            "Skipping I3LYH0: Plot already exists.\n",
            "Skipping I3LVV6: Plot already exists.\n",
            "Skipping A0A287CRV9: Plot already exists.\n",
            "Skipping A0A287D6P7: Plot already exists.\n",
            "45% complete\n",
            "Skipping I3MVJ7: Plot already exists.\n",
            "Skipping I3MCA4: Plot already exists.\n",
            "Skipping A0A287DE03: Plot already exists.\n",
            "Skipping A0A287DAE5: Plot already exists.\n",
            "Skipping A0A287D6K5: Plot already exists.\n",
            "Skipping I3M9Z7: Plot already exists.\n",
            "Skipping I3M720: Plot already exists.\n",
            "Skipping I3NFM6: Plot already exists.\n",
            "Skipping A0A287D2Z9: Plot already exists.\n",
            "Skipping I3NG95: Plot already exists.\n",
            "Skipping I3MLK8: Plot already exists.\n",
            "Skipping I3M3T0: Plot already exists.\n",
            "Skipping I3MCB8: Plot already exists.\n",
            "Skipping I3M994: Plot already exists.\n",
            "Skipping I3MUG0: Plot already exists.\n",
            "Skipping A0A287CWA8: Plot already exists.\n",
            "Skipping I3NGP1: Plot already exists.\n",
            "Skipping A0A287DF55: Plot already exists.\n",
            "Skipping I3LW37: Plot already exists.\n",
            "Skipping I3NBF0: Plot already exists.\n",
            "Skipping A0A287D1Y0: Plot already exists.\n",
            "Skipping A0A287D9R7: Plot already exists.\n",
            "Skipping I3M0R6: Plot already exists.\n",
            "Skipping A0A287DEF6: Plot already exists.\n",
            "Skipping I3MPK4: Plot already exists.\n",
            "Skipping I3M8E9: Plot already exists.\n",
            "Skipping I3MFM8: Plot already exists.\n",
            "Skipping I3M5P9: Plot already exists.\n",
            "Skipping I3MB17: Plot already exists.\n",
            "Skipping I3N752: Plot already exists.\n",
            "Skipping I3NCW4: Plot already exists.\n",
            "Skipping A0A287CZR4: Plot already exists.\n",
            "Skipping I3M838: Plot already exists.\n",
            "Skipping I3N5Y3: Plot already exists.\n",
            "Skipping A0A287D1L7: Plot already exists.\n",
            "Skipping I3MZY2: Plot already exists.\n",
            "Skipping A0A287D200: Plot already exists.\n",
            "Skipping I3NEI8: Plot already exists.\n",
            "Skipping I3N7A4: Plot already exists.\n",
            "Skipping A0A287D1Q6: Plot already exists.\n",
            "Skipping I3MCX0: Plot already exists.\n",
            "Skipping I3LY72: Plot already exists.\n",
            "Skipping I3MF23: Plot already exists.\n",
            "Skipping I3M5N1: Plot already exists.\n",
            "Skipping I3NG40: Plot already exists.\n",
            "Skipping I3MBI7: Plot already exists.\n",
            "Skipping A0A287DBH5: Plot already exists.\n",
            "Skipping I3NB69: Plot already exists.\n",
            "Skipping I3N0Z2: Plot already exists.\n",
            "Skipping I3MQ25: Plot already exists.\n",
            "46% complete\n",
            "Skipping I3MMV2: Plot already exists.\n",
            "Skipping I3MV43: Plot already exists.\n",
            "Skipping I3MUZ6: Plot already exists.\n",
            "Skipping I3LY74: Plot already exists.\n",
            "Skipping I3M5F7: Plot already exists.\n",
            "Skipping I3MYA1: Plot already exists.\n",
            "Skipping I3MDD9: Plot already exists.\n",
            "Skipping I3MCG6: Plot already exists.\n",
            "Skipping I3NAU7: Plot already exists.\n",
            "Skipping I3MVK0: Plot already exists.\n",
            "Skipping A0A287D2Q5: Plot already exists.\n",
            "Skipping I3NE06: Plot already exists.\n",
            "Skipping I3M537: Plot already exists.\n",
            "Skipping A0A287CTI0: Plot already exists.\n",
            "Skipping I3MDM1: Plot already exists.\n",
            "Skipping I3MLX7: Plot already exists.\n",
            "Skipping I3MU54: Plot already exists.\n",
            "Skipping I3MJE0: Plot already exists.\n",
            "Skipping I3MKC1: Plot already exists.\n",
            "Skipping A0A287D9D2: Plot already exists.\n",
            "Skipping I3M9I0: Plot already exists.\n",
            "Skipping I3MHH4: Plot already exists.\n",
            "Skipping I3ME84: Plot already exists.\n",
            "Skipping I3N200: Plot already exists.\n",
            "Skipping I3MNH2: Plot already exists.\n",
            "Skipping A0A287DA67: Plot already exists.\n",
            "Skipping I3M2I9: Plot already exists.\n",
            "Skipping A0A287DCD0: Plot already exists.\n",
            "Skipping A0A287D3H7: Plot already exists.\n",
            "Skipping I3NDK1: Plot already exists.\n",
            "Skipping I3MZ09: Plot already exists.\n",
            "Skipping I3MJW7: Plot already exists.\n",
            "Skipping I3NA57: Plot already exists.\n",
            "Skipping A0A287CZ22: Plot already exists.\n",
            "Skipping I3MMF9: Plot already exists.\n",
            "Skipping I3MYB5: Plot already exists.\n",
            "Skipping A0A287D8S8: Plot already exists.\n",
            "Skipping A0A287D5A5: Plot already exists.\n",
            "Skipping I3LX68: Plot already exists.\n",
            "Skipping I3N907: Plot already exists.\n",
            "Skipping I3MQN2: Plot already exists.\n",
            "Skipping I3MPD0: Plot already exists.\n",
            "Skipping I3N8K2: Plot already exists.\n",
            "Skipping I3MQJ5: Plot already exists.\n",
            "Skipping A0A287CSP3: Plot already exists.\n",
            "Skipping I3MDV0: Plot already exists.\n",
            "Skipping I3NFD6: Plot already exists.\n",
            "Skipping A0A287DEB6: Plot already exists.\n",
            "Skipping A0A287CVT3: Plot already exists.\n",
            "Skipping I3MLJ0: Plot already exists.\n",
            "48% complete\n",
            "Skipping I3N7J5: Plot already exists.\n",
            "Skipping I3LY67: Plot already exists.\n",
            "Skipping A0A287CT81: Plot already exists.\n",
            "Skipping A0A287CU35: Plot already exists.\n",
            "Skipping I3M255: Plot already exists.\n",
            "Skipping I3MCR1: Plot already exists.\n",
            "Skipping I3N7H8: Plot already exists.\n",
            "Skipping I3NEP6: Plot already exists.\n",
            "Skipping A0A287D382: Plot already exists.\n",
            "Skipping A0A287DDV8: Plot already exists.\n",
            "Skipping A0A287CT27: Plot already exists.\n",
            "Skipping I3LXL5: Plot already exists.\n",
            "Skipping I3MI61: Plot already exists.\n",
            "Skipping I3M618: Plot already exists.\n",
            "Skipping A0A287D6J0: Plot already exists.\n",
            "Skipping I3M1X8: Plot already exists.\n",
            "Skipping I3LWM4: Plot already exists.\n",
            "Skipping A0A287DE58: Plot already exists.\n",
            "Skipping I3MPW8: Plot already exists.\n",
            "Skipping A0A287CTJ4: Plot already exists.\n",
            "Skipping I3MBL6: Plot already exists.\n",
            "Skipping I3M217: Plot already exists.\n",
            "Skipping I3MIR9: Plot already exists.\n",
            "Skipping I3N5Z5: Plot already exists.\n",
            "Skipping I3NEP5: Plot already exists.\n",
            "Skipping I3MI72: Plot already exists.\n",
            "Skipping I3N0K6: Plot already exists.\n",
            "Skipping A0A287D7F6: Plot already exists.\n",
            "Skipping I3M0L6: Plot already exists.\n",
            "Skipping I3NB60: Plot already exists.\n",
            "Skipping I3NA11: Plot already exists.\n",
            "Skipping I3LXE6: Plot already exists.\n",
            "Skipping I3MDB2: Plot already exists.\n",
            "Skipping A0A287D1K4: Plot already exists.\n",
            "Skipping I3ML03: Plot already exists.\n",
            "Skipping I3MF47: Plot already exists.\n",
            "Skipping I3MHF4: Plot already exists.\n",
            "Skipping I3NA77: Plot already exists.\n",
            "Skipping I3MMP5: Plot already exists.\n",
            "Skipping I3N5C3: Plot already exists.\n",
            "Skipping I3MRU8: Plot already exists.\n",
            "Skipping I3MP40: Plot already exists.\n",
            "Skipping I3NFR7: Plot already exists.\n",
            "Skipping I3MJ29: Plot already exists.\n",
            "Skipping I3MYF8: Plot already exists.\n",
            "Skipping I3MEM0: Plot already exists.\n",
            "Skipping I3LX06: Plot already exists.\n",
            "Skipping I3MQ63: Plot already exists.\n",
            "Skipping I3MZQ7: Plot already exists.\n",
            "Skipping I3NHC1: Plot already exists.\n",
            "49% complete\n",
            "Skipping I3NE41: Plot already exists.\n",
            "Skipping I3N4C2: Plot already exists.\n",
            "Skipping I3MUQ0: Plot already exists.\n",
            "Skipping A0A287D763: Plot already exists.\n",
            "Skipping I3MGL8: Plot already exists.\n",
            "Skipping I3MK32: Plot already exists.\n",
            "Skipping I3N3P6: Plot already exists.\n",
            "Skipping I3LXB3: Plot already exists.\n",
            "Skipping I3M9C2: Plot already exists.\n",
            "Skipping I3MPE6: Plot already exists.\n",
            "Skipping A0A287D6F1: Plot already exists.\n",
            "Skipping A0A287DB56: Plot already exists.\n",
            "Skipping I3MF75: Plot already exists.\n",
            "Skipping I3M8A2: Plot already exists.\n",
            "Skipping I3MHA7: Plot already exists.\n",
            "Skipping I3M8V5: Plot already exists.\n",
            "Skipping I3MD48: Plot already exists.\n",
            "Skipping I3M140: Plot already exists.\n",
            "Skipping I3N469: Plot already exists.\n",
            "Skipping I3M5C7: Plot already exists.\n",
            "Skipping I3M3B6: Plot already exists.\n",
            "Skipping A0A287CXR2: Plot already exists.\n",
            "Skipping A0A287DF37: Plot already exists.\n",
            "Skipping I3NFD1: Plot already exists.\n",
            "Skipping A0A287D0G7: Plot already exists.\n",
            "Skipping I3MAE2: Plot already exists.\n",
            "Skipping I3M315: Plot already exists.\n",
            "Skipping I3ME02: Plot already exists.\n",
            "Skipping A0A287CVP5: Plot already exists.\n",
            "Skipping I3N9U4: Plot already exists.\n",
            "Skipping A0A287D954: Plot already exists.\n",
            "Skipping I3N383: Plot already exists.\n",
            "Skipping I3MSS1: Plot already exists.\n",
            "Skipping I3M7C8: Plot already exists.\n",
            "Skipping I3MVT9: Plot already exists.\n",
            "Skipping I3MJW4: Plot already exists.\n",
            "Skipping A0A287DBP1: Plot already exists.\n",
            "Skipping A0A287D6D9: Plot already exists.\n",
            "Skipping I3MJI0: Plot already exists.\n",
            "Skipping I3MHJ0: Plot already exists.\n",
            "Skipping I3MN83: Plot already exists.\n",
            "Skipping I3MQD7: Plot already exists.\n",
            "Skipping I3MDK4: Plot already exists.\n",
            "Skipping I3MRT7: Plot already exists.\n",
            "Skipping I3LW52: Plot already exists.\n",
            "Skipping I3MW18: Plot already exists.\n",
            "Skipping I3M499: Plot already exists.\n",
            "Skipping I3MLK4: Plot already exists.\n",
            "Skipping I3N188: Plot already exists.\n",
            "Skipping I3MPS2: Plot already exists.\n",
            "51% complete\n",
            "Skipping A0A287D5R8: Plot already exists.\n",
            "Skipping A0A287DB82: Plot already exists.\n",
            "Skipping I3MMJ3: Plot already exists.\n",
            "Skipping I3MQ49: Plot already exists.\n",
            "Skipping I3M7T7: Plot already exists.\n",
            "Skipping I3MHG1: Plot already exists.\n",
            "Skipping I3LWQ2: Plot already exists.\n",
            "Skipping I3M9T7: Plot already exists.\n",
            "Skipping I3M4N5: Plot already exists.\n",
            "Skipping I3MZD4: Plot already exists.\n",
            "Skipping I3MMH1: Plot already exists.\n",
            "Skipping I3LXU8: Plot already exists.\n",
            "Skipping I3MNZ2: Plot already exists.\n",
            "Skipping I3NGY3: Plot already exists.\n",
            "Skipping A0A287DEI7: Plot already exists.\n",
            "Skipping I3N336: Plot already exists.\n",
            "Skipping A0A287D2Z4: Plot already exists.\n",
            "Skipping I3MNH9: Plot already exists.\n",
            "Skipping I3MA38: Plot already exists.\n",
            "Skipping I3MVU3: Plot already exists.\n",
            "Skipping A0A287DA49: Plot already exists.\n",
            "Skipping A0A287CSI8: Plot already exists.\n",
            "Skipping I3MMW9: Plot already exists.\n",
            "Skipping A0A287DGC5: Plot already exists.\n",
            "Skipping A0A287CWW5: Plot already exists.\n",
            "Skipping I3MZG6: Plot already exists.\n",
            "Skipping A0A287CWU8: Plot already exists.\n",
            "Skipping I3MNU5: Plot already exists.\n",
            "Skipping I3NE34: Plot already exists.\n",
            "Skipping A0A287D5N3: Plot already exists.\n",
            "Skipping A0A287D7N5: Plot already exists.\n",
            "Skipping A0A287CUU1: Plot already exists.\n",
            "Skipping A0A287D0R6: Plot already exists.\n",
            "Skipping A0A287CRV0: Plot already exists.\n",
            "Skipping A0A287DF09: Plot already exists.\n",
            "Skipping I3M0B2: Plot already exists.\n",
            "Skipping I3M9T6: Plot already exists.\n",
            "Skipping I3M4K2: Plot already exists.\n",
            "Skipping A0A287D2C7: Plot already exists.\n",
            "Skipping I3M452: Plot already exists.\n",
            "Skipping I3MQ22: Plot already exists.\n",
            "Skipping I3MDH0: Plot already exists.\n",
            "Skipping I3MW72: Plot already exists.\n",
            "Skipping I3MCQ8: Plot already exists.\n",
            "Skipping A0A287D6T5: Plot already exists.\n",
            "Skipping I3M6N6: Plot already exists.\n",
            "Skipping I3LXI5: Plot already exists.\n",
            "Skipping I3MMQ1: Plot already exists.\n",
            "Skipping A0A287CS98: Plot already exists.\n",
            "Skipping I3MEF6: Plot already exists.\n",
            "53% complete\n",
            "Skipping I3MDE2: Plot already exists.\n",
            "Skipping I3M5X0: Plot already exists.\n",
            "Skipping I3M4G7: Plot already exists.\n",
            "Skipping I3MCS9: Plot already exists.\n",
            "Skipping I3NB24: Plot already exists.\n",
            "Skipping I3MY01: Plot already exists.\n",
            "Skipping I3LXX7: Plot already exists.\n",
            "Skipping A0A287CUN8: Plot already exists.\n",
            "Skipping I3MA47: Plot already exists.\n",
            "Skipping I3M0Y5: Plot already exists.\n",
            "Skipping I3MI29: Plot already exists.\n",
            "Skipping I3M4U6: Plot already exists.\n",
            "Skipping I3MHN3: Plot already exists.\n",
            "Skipping A0A287CZ62: Plot already exists.\n",
            "Skipping A0A287CVQ6: Plot already exists.\n",
            "Skipping A0A287CS45: Plot already exists.\n",
            "Skipping I3MTZ4: Plot already exists.\n",
            "Skipping I3M7U3: Plot already exists.\n",
            "Skipping A0A287DAA2: Plot already exists.\n",
            "Skipping A0A287CXL5: Plot already exists.\n",
            "Skipping I3MD43: Plot already exists.\n",
            "Skipping A0A287CT46: Plot already exists.\n",
            "Skipping A0A287CT87: Plot already exists.\n",
            "Skipping I3MRA4: Plot already exists.\n",
            "Skipping I3M2W9: Plot already exists.\n",
            "Skipping I3MK05: Plot already exists.\n",
            "Skipping I3NF80: Plot already exists.\n",
            "Skipping I3MV00: Plot already exists.\n",
            "Skipping I3MKB0: Plot already exists.\n",
            "Skipping I3M1C9: Plot already exists.\n",
            "Skipping I3MNG0: Plot already exists.\n",
            "Skipping I3MKL0: Plot already exists.\n",
            "Skipping I3MZY4: Plot already exists.\n",
            "Skipping A0A287DBB5: Plot already exists.\n",
            "Skipping I3N9R4: Plot already exists.\n",
            "Skipping I3M348: Plot already exists.\n",
            "Skipping I3M7B4: Plot already exists.\n",
            "Skipping I3MA74: Plot already exists.\n",
            "Skipping I3MPC7: Plot already exists.\n",
            "Skipping I3MC85: Plot already exists.\n",
            "Skipping I3MJ17: Plot already exists.\n",
            "Skipping I3MCJ8: Plot already exists.\n",
            "Skipping I3LYW6: Plot already exists.\n",
            "Skipping I3MML0: Plot already exists.\n",
            "Skipping A0A287CS43: Plot already exists.\n",
            "Skipping A0A287CSE0: Plot already exists.\n",
            "Skipping I3M649: Plot already exists.\n",
            "Skipping I3N0V0: Plot already exists.\n",
            "Skipping I3MQP7: Plot already exists.\n",
            "Skipping I3MLJ9: Plot already exists.\n",
            "54% complete\n",
            "Skipping I3MJ54: Plot already exists.\n",
            "Skipping I3N3T9: Plot already exists.\n",
            "Skipping I3MCE5: Plot already exists.\n",
            "Skipping A0A287CYG3: Plot already exists.\n",
            "Skipping I3M6R1: Plot already exists.\n",
            "Skipping I3M7U5: Plot already exists.\n",
            "Skipping I3MER0: Plot already exists.\n",
            "Skipping I3MB48: Plot already exists.\n",
            "Skipping A0A287D0J5: Plot already exists.\n",
            "Skipping I3MDH4: Plot already exists.\n",
            "Skipping I3MVZ0: Plot already exists.\n",
            "Skipping I3MG64: Plot already exists.\n",
            "Skipping I3N8S7: Plot already exists.\n",
            "Skipping I3MP39: Plot already exists.\n",
            "Skipping I3M9X4: Plot already exists.\n",
            "Skipping I3MYF1: Plot already exists.\n",
            "Skipping I3N0A9: Plot already exists.\n",
            "Skipping I3LY84: Plot already exists.\n",
            "Skipping I3MUY7: Plot already exists.\n",
            "Skipping I3LYZ3: Plot already exists.\n",
            "Skipping I3NCR2: Plot already exists.\n",
            "Skipping I3MB21: Plot already exists.\n",
            "Skipping I3MGZ7: Plot already exists.\n",
            "Skipping I3MS67: Plot already exists.\n",
            "Skipping I3M4B5: Plot already exists.\n",
            "Skipping I3LXW8: Plot already exists.\n",
            "Skipping I3ND02: Plot already exists.\n",
            "Skipping I3MQZ1: Plot already exists.\n",
            "Skipping I3M3Z4: Plot already exists.\n",
            "Skipping I3NCW2: Plot already exists.\n",
            "56% complete\n",
            "57% complete\n",
            "59% complete\n",
            "60% complete\n",
            "62% complete\n",
            "64% complete\n",
            "65% complete\n",
            "67% complete\n",
            "68% complete\n",
            "70% complete\n",
            "71% complete\n",
            "73% complete\n",
            "74% complete\n",
            "76% complete\n",
            "78% complete\n",
            "79% complete\n",
            "81% complete\n",
            "82% complete\n",
            "84% complete\n",
            "85% complete\n",
            "87% complete\n",
            "89% complete\n",
            "90% complete\n",
            "92% complete\n",
            "93% complete\n",
            "95% complete\n",
            "96% complete\n",
            "98% complete\n",
            "99% complete\n",
            "101% complete\n",
            "Total accession numbers read from input file: 3201\n",
            "Total number of sa.png plots saved: 3201\n",
            "Total number of accession numbers that could not be processed: 0\n",
            "Total accession numbers read from input file: 3201\n",
            "Total number of sa.png plots saved: 3201\n",
            "Total number of accession numbers that could not be processed: 0\n",
            "Processing complete.\n"
          ]
        }
      ],
      "source": [
        "#@title Plots solvent accessibility vs residue for all pdbs\n",
        "#may crash, rerun to start if that occurs\n",
        "\n",
        "import os\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import gc\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "# Ensure the output directory is correctly defined as an absolute path\n",
        "output_dir = \"/content/output/plots/sa_plots\"\n",
        "os.makedirs(output_dir, exist_ok=True)\n",
        "\n",
        "# Define log file\n",
        "log_file = os.path.join(output_dir, \"sa_plot_process.log\")\n",
        "\n",
        "def read_solvent_accessibility(file_path, log_file):\n",
        "    residue_numbers = []\n",
        "    solvent_accessibility = []\n",
        "    try:\n",
        "        with open(file_path, \"r\") as file:\n",
        "            for _ in range(2):  # Skip first two rows\n",
        "                next(file)\n",
        "            for line in file:\n",
        "                columns = line.strip().split('\\t')  # Use tab delimiter for TSV files\n",
        "                residue_number = int(columns[1])\n",
        "                accessibility = float(columns[-1])\n",
        "                residue_numbers.append(residue_number)\n",
        "                solvent_accessibility.append(accessibility)\n",
        "    except Exception as e:\n",
        "        write_log(log_file, f\"Error reading solvent accessibility from {file_path}: {e}\")\n",
        "    return residue_numbers, solvent_accessibility\n",
        "\n",
        "def read_modifications(directory, target_accession, log_file):\n",
        "    modifications = set()\n",
        "    try:\n",
        "        for file in os.listdir(directory):\n",
        "            if file.startswith(\"all_modlist_Abundance\") and file.endswith(\".tsv\"):\n",
        "                file_path = os.path.join(directory, file)\n",
        "                df = pd.read_csv(file_path, delimiter='\\t')\n",
        "                df = df[df['Accession Number'].str.strip() == target_accession]\n",
        "                for _, row in df.iterrows():\n",
        "                    mod = row['Modifications']\n",
        "                    if pd.notna(mod):\n",
        "                        if 'Met-loss' in mod:\n",
        "                            mod_residue = 1\n",
        "                        else:\n",
        "                            mod_residue = int(''.join(filter(str.isdigit, str(mod))))\n",
        "                        modifications.add(mod_residue)\n",
        "    except Exception as e:\n",
        "        write_log(log_file, f\"Error reading modifications from {directory}: {e}\")\n",
        "    return modifications\n",
        "\n",
        "def read_domains(file_path, target_accession, log_file):\n",
        "    domains = []\n",
        "    try:\n",
        "        with open(file_path, \"r\") as file:\n",
        "            next(file)  # Skip header\n",
        "            for line in file:\n",
        "                columns = line.strip().split('\\t')  # Use tab delimiter for TSV files\n",
        "                if len(columns) >= 4:\n",
        "                    accession = columns[0].strip()\n",
        "                    if accession == target_accession:\n",
        "                        domain_type = columns[1].strip()\n",
        "                        start = int(columns[2].strip())\n",
        "                        end = int(columns[3].strip())\n",
        "                        domains.append((start, end, domain_type))\n",
        "    except Exception as e:\n",
        "        write_log(log_file, f\"Error reading domains from {file_path}: {e}\")\n",
        "    return domains\n",
        "\n",
        "def process_accession_file(file, protein_ss_dir, modlist_dir, domain_file_path, output_dir, log_file):\n",
        "    accession_number = file.split('_')[0]\n",
        "    accessibility_file_path = os.path.join(protein_ss_dir, file)\n",
        "    output_file_path = os.path.join(output_dir, f'{accession_number}_sa.pdf')\n",
        "\n",
        "    if os.path.exists(output_file_path):\n",
        "        write_log(log_file, f\"Skipping {accession_number}: Plot already exists.\")\n",
        "        return True\n",
        "\n",
        "    if not os.path.exists(accessibility_file_path):\n",
        "        write_log(log_file, f\"Skipping {accession_number}: Accessibility file not found.\")\n",
        "        with open(os.path.join('/content/output/failed/sa_plot', \"sa_plot_failed.txt\"), \"a\") as failed_file:\n",
        "            failed_file.write(accession_number + \"\\n\")\n",
        "        return False\n",
        "\n",
        "    residue_numbers, solvent_accessibility = read_solvent_accessibility(accessibility_file_path, log_file)\n",
        "    modifications = read_modifications(modlist_dir, accession_number, log_file)\n",
        "    domains = read_domains(domain_file_path, accession_number, log_file)\n",
        "\n",
        "    fig, ax = plt.subplots(figsize=(12, 6))\n",
        "    ax.plot(residue_numbers, solvent_accessibility, marker='', linestyle='-', color='black')\n",
        "\n",
        "    for residue_number in modifications:\n",
        "        if residue_number in residue_numbers:\n",
        "            ax.plot(residue_number, solvent_accessibility[residue_numbers.index(residue_number)], 'ro')\n",
        "\n",
        "    for start, end, domain_type in domains:\n",
        "        ax.axvspan(start, end, color='lightgray', alpha=0.5)\n",
        "        ax.text((start + end) / 2, 0.8, domain_type, horizontalalignment='center', verticalalignment='center', fontsize=14, fontweight='bold', color='black', bbox=dict(facecolor='white', edgecolor='none', alpha=0.7))\n",
        "\n",
        "    ax.set_xlabel('Residue Number')\n",
        "    ax.set_ylabel('Solvent Accessibility')\n",
        "    ax.set_title(f'Solvent Accessibility and Modifications for {accession_number}')\n",
        "    ax.set_ylim(0, 1.0)\n",
        "    ax.grid(False)\n",
        "\n",
        "    plt.savefig(output_file_path)\n",
        "    plt.close(fig)\n",
        "    gc.collect()  # Collect garbage to free up memory\n",
        "\n",
        "    return True\n",
        "\n",
        "def process_files_in_batches(accession_files, protein_ss_dir, modlist_dir, domain_file_path, output_dir, log_file, batch_size=50):\n",
        "    total_files = len(accession_files)\n",
        "    saved_plots = 0\n",
        "    failed_plots = 0\n",
        "\n",
        "    for i in range(0, total_files, batch_size):\n",
        "        batch_files = accession_files[i:i+batch_size]\n",
        "        for file in batch_files:\n",
        "            if process_accession_file(file, protein_ss_dir, modlist_dir, domain_file_path, output_dir, log_file):\n",
        "                saved_plots += 1\n",
        "            else:\n",
        "                failed_plots += 1\n",
        "\n",
        "        gc.collect()  # Collect garbage to free up memory after each batch\n",
        "\n",
        "        # Calculate and log completion percentage\n",
        "        percent_complete = (i + batch_size) / total_files * 100\n",
        "        write_log(log_file, f\"{int(percent_complete)}% complete\")\n",
        "\n",
        "    print(f\"Total accession numbers read from input file: {total_files}\")\n",
        "    print(f\"Total number of sa.png plots saved: {saved_plots}\")\n",
        "    print(f\"Total number of accession numbers that could not be processed: {failed_plots}\")\n",
        "    write_log(log_file, f\"Total accession numbers read from input file: {total_files}\")\n",
        "    write_log(log_file, f\"Total number of sa.png plots saved: {saved_plots}\")\n",
        "    write_log(log_file, f\"Total number of accession numbers that could not be processed: {failed_plots}\")\n",
        "\n",
        "# List all files in the protein_ss/ directory\n",
        "protein_ss_dir = \"/content/output/libraries/all_protein_ss\"\n",
        "modlist_dir = \"/content/output/modlist_files\"\n",
        "domain_file_path = \"/content/output/libraries/domainlibrary.tsv\"\n",
        "accession_files = [f for f in os.listdir(protein_ss_dir) if f.endswith('_ss.tsv')]\n",
        "\n",
        "process_files_in_batches(accession_files, protein_ss_dir, modlist_dir, domain_file_path, output_dir, log_file)\n",
        "\n",
        "write_log(log_file, \"Processing complete.\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "cellView": "form",
        "id": "-FL-LOTBQWso",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "1e7d4728-c484-407a-fa59-59f4c2cd3aee"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Total number of residues: 2697361\n",
            "Number of A residues: 187384\n",
            "Percentage of A residues: 6.946938136941997%\n",
            "Processing file: /content/output/modlist_files/AdjacentSeq_all__(Aug_Active)___(Mar_Active).tsv\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 1200x800 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAABKUAAAMWCAYAAAAgRDUeAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAB88ElEQVR4nOzdd5hU5fk//ntpu9SlCwgiIoqKLSR2KUpAUBBFBY0KaCwRY9SPUYmxYAmWaOy9RQJ2rFGxIUosUYMgGg0gCoqggCxFBWTP7w9/zNdlF9xdljO78Hpd11zXznOeOefemTPtPc95Tk6SJEkAAAAAQIqqZbsAAAAAADY9QikAAAAAUieUAgAAACB1QikAAAAAUieUAgAAACB1QikAAAAAUieUAgAAACB1QikAAAAAUieUAgAAACB1Qilgk3PRRRdFTk5Oqfrm5OTERRddtGELKsHqGufPn1+h6x01alR07NgxatasGQ0bNqzQdVdV9957b+Tk5MSnn36aaevWrVt069atSL958+bFYYcdFk2aNImcnJy49tpr45VXXomcnJx45ZVXUq05ImLLLbeMIUOGpL7dtJV0Hw8ZMiS23HLLrNVEcWs+Zz799NPIycmJe++9t8K3VVhYGJ06dYrLLruswte9sdjQrw8ffvhh1KhRI6ZOnbrBtlEap5xySvz617/Oag2V1YZ+nVy5cmW0adMmbr755g22DWDTIJQCsmZ1GLD6UqNGjdh8881jyJAh8cUXX2S7vI3ORx99FEOGDIn27dvHHXfcEbfffnu2S6pSzjjjjBg3blwMHz48Ro0aFQcccMAG3+brr78eF110USxatGiDb2tDe+aZZyInJydatWoVhYWF2S4nVd9++21cdNFFaw0vJ06cGL17947NN9888vLyYosttoi+ffvGmDFj0i20irj//vtj9uzZceqpp5a4/Oabb46cnJzYfffdU66s9I444ojIycmJc845p9zryObrw/bbbx8HHnhgXHDBBalve7WZM2fGnXfeGX/6058ybavD0JycnLj00ktLvN1vfvObyMnJiXr16qVVaokqYj+dM2dOXHTRRfHee+9VXGGlVLNmzTjzzDPjsssui++///5n+2ezVqCSSwCy5J577kkiIrn44ouTUaNGJXfccUdy/PHHJ9WrV0/at2+ffPfddxtkuytXriz1uiMiufDCCzdIHety4YUXJhGRfP311xW2zltuuSWJiGTatGkVts6Nwer9cObMmZm25cuXJ8uXLy/Sb7PNNkt+85vfFGlbtWpV8t133yWrVq3aILVdddVVxWpb7fvvv09WrFixQba7IRx11FHJlltumURE8sILL5T6duPHj08iIhk/fnymbcWKFcn333+/AarcML7++uu1vpY89NBDSU5OTrLrrrsmV1xxRXL77bcnw4cPT/bee++kW7du6RdbTl27dk26du2auT5z5swkIpJ77rmnwre18847JyeeeOJal++1116Zfa0yvt4VFBQkeXl5yZZbbpm0adMmKSwsLNd6sv368MwzzyQRkUyfPn2Dbmdt/vCHPyTbbLNNkbbV+11eXl6y/fbbF7vN0qVLk7p16yZ5eXlJ3bp10yq1RBWxn7799ttrfZ6l8Tr5zTffJLVq1Uruuuuun+27rlqBTZuRUkDW9e7dO44++uj47W9/G3feeWecddZZMWPGjHjyySc3yPZq1KgReXl5G2TdldlXX30VEVGhh+19++23FbauyqRWrVpRq1atIm1fffVVsfuuWrVqkZeXF9Wqpf92mpubGzVr1kx9u+WxbNmyeOKJJ+LMM8+MXXfdNUaPHr1e66tZs2bk5uZWUHXZddFFF8X2228fb775Zpx99tlxwgknxF/+8peYOHFiPPjgg9kur9KZNGlSTJ48OY444ogSl8+cOTNef/31uOaaa6JZs2brva9tCI8++misWrUq7r777pg9e3a8+uqrFb6NNF4fevToEY0aNYq///3vG3Q7JVm5cmWMHj16rftBnz594sMPP4zJkycXaX/iiSdixYoVFX7IX1nfC9PYT9N4nWzYsGH07NlzgxymC2w6hFJApbPvvvtGRMSMGTOKtH/00Udx2GGHRePGjSMvLy9++ctfFguuVq5cGSNGjIgOHTpEXl5eNGnSJPbZZ5944YUXMn1KmlNq+fLlccYZZ0SzZs2ifv360a9fv/j888+L1ba2ORrWNk/VP/7xj+jcuXPUrl07GjduHIMGDYrZs2eX+r6YP39+HHHEEdGgQYNo0qRJ/OEPfyhxmPzPbWfLLbeMCy+8MCIimjVrVmyurJtvvjl22GGHyM3NjVatWsWwYcOKHRLSrVu36NSpU7z77rvRpUuXqFOnTuawieXLl8eFF14YW2+9deTm5kabNm3i7LPPjuXLl//s/7h6vVOmTImuXbtGnTp1Yuutt45HHnkkIiImTJgQu+++e9SuXTu23XbbePHFF4utY9KkSdG7d+9o0KBB1KtXL/bff/948803i/X74IMPYr/99ovatWtH69at49JLLy3xULKfzo+z+jDTJEnipptuyhwaElHyfEcREW+99Vb06dMnGjVqFHXr1o2ddtoprrvuuszyKVOmxJAhQ2KrrbaKvLy8aNGiRRx33HGxYMGCTJ+LLroo/vjHP0ZERLt27TLbXT33VUlzxnzyySdx+OGHR+PGjaNOnTqxxx57xD//+c8ifVbX/NBDD8Vll10WrVu3jry8vNh///1j+vTpJTxC6++xxx6L7777Lg4//PAYNGhQjB07tsT9+PPPP4/+/ftH3bp1o3nz5nHGGWeUuA+V9Dz861//GnvttVc0adIkateuHZ07d87sQ2v6xz/+EbvttlvUqVMnGjVqFF26dInnn3++SJ9nn3029t1336hbt27Ur18/DjzwwPjggw+K1VGvXr344osvon///lGvXr1o1qxZnHXWWbFq1aqI+PFwombNmkVExIgRIzKP4+rn34wZM+JXv/pVsRA0IqJ58+ZFrhcWFsa1114bO+ywQ+Tl5cVmm20WJ510UnzzzTdF+iVJEpdeemm0bt066tSpE927d48PPvigXPMM3XPPPbHffvtF8+bNIzc3N7bffvu45ZZbyrSOivT4449HrVq1okuXLiUuHz16dDRq1CgOPPDAOOyww0r8sr+25+3a5sF6+OGHY/vtt4+8vLzo1KlTPPbYY+s1X8/o0aPj17/+dXTv3j222267tQYSH330URxxxBHRrFmzzOvfeeedFxFle3145513Iicnp8TwaNy4cZGTkxNPP/10pu2LL76I4447LjbbbLPIzc2NHXbYIe6+++5it61Zs2Z069YtnnjiiXLdD+tj4sSJMX/+/OjRo0eJy/fcc89o165dsUNgR48eHQcccEA0bty42G2eeOKJOPDAA6NVq1aRm5sb7du3j0suuSTzXF5tXe+FpVWa/TQiYtGiRXHGGWfElltuGbm5udG6des49thjY/78+fHKK6/Er371q4iIGDp0aGYfWL3//nQfXblyZTRu3DiGDh1abBuLFy+OvLy8OOusszJtZXlP//Wvfx0TJ06MhQsXrvX//blap02bFgMGDIgWLVpEXl5etG7dOgYNGhQFBQUREdGpU6fo3r17sfUWFhbG5ptvHocddthatw1UfkIpoNJZ/aG6UaNGmbYPPvgg9thjj/jvf/8b5557blx99dVRt27d6N+/fzz22GOZfhdddFGMGDEiunfvHjfeeGOcd955scUWW8R//vOfdW7zt7/9bVx77bXRs2fPuPzyy6NmzZpx4IEHrtf/cdlll8Wxxx4bHTp0iGuuuSZOP/30eOmll6JLly6lngPkiCOOiO+//z5GjhwZffr0ieuvvz5OPPHEMm/n2muvjUMOOSQiIm655ZYYNWpUHHrooRHx4302bNiwaNWqVVx99dUxYMCAuO2226Jnz56xcuXKIttasGBB9O7dO3bZZZe49tpro3v37lFYWBj9+vWLv/71r9G3b9+44YYbon///vG3v/0tBg4cWKr/85tvvomDDjoodt9997jyyisjNzc3Bg0aFA8++GAMGjQo+vTpE5dffnksW7YsDjvssFiyZEnmth988EHsu+++MXny5Dj77LPj/PPPj5kzZ0a3bt3irbfeyvSbO3dudO/ePd57770499xz4/TTT4/77ruvSFhUki5dusSoUaMi4scP36NGjcpcL8kLL7wQXbp0iQ8//DD+8Ic/xNVXXx3du3cv8qXvhRdeiE8++SSGDh0aN9xwQwwaNCgeeOCB6NOnTyRJEhERhx56aBx55JEREfG3v/0ts93VAcea5s2bF3vttVeMGzcuTjnllMw8H/369SvyHFnt8ssvj8ceeyzOOuusGD58eLz55pvxm9/8Zp33RXmNHj06unfvHi1atIhBgwbFkiVL4qmnnirS57vvvov9998/xo0bF6eeemqcd9558dprr8XZZ59dqm1cd911seuuu8bFF18cf/nLX6JGjRpx+OGHFwvlRowYEcccc0zUrFkzLr744hgxYkS0adMmXn755UyfUaNGxYEHHhj16tWLK664Is4///z48MMPY5999ikyIX5ExKpVq6JXr17RpEmT+Otf/xpdu3aNq6++OjNnW7NmzTIhziGHHJJ5HFc//9q2bRsvvfRSiSH4mk466aT44x//GHvvvXdcd911MXTo0Bg9enT06tWryHP1ggsuiPPPPz923nnnuOqqq2KrrbaKnj17xrJly0p1X/7ULbfcEm3bto0//elPcfXVV0ebNm3ilFNOiZtuuqnM66oIr7/+enTq1Gmto4BGjx4dhx56aNSqVSuOPPLImDZtWrz99tvl3t4///nPGDhwYNSsWTNGjhwZhx56aBx//PHx7rvvlmt9c+bMifHjx2ee20ceeWQ88sgjsWLFiiL9pkyZErvvvnu8/PLLccIJJ8R1110X/fv3zzxvyvL68Mtf/jK22mqreOihh4ote/DBB6NRo0bRq1eviPjxdWSPPfaIF198MU499dS47rrrYuutt47jjz8+rr322mK379y5c0ydOjUWL15crvujvF5//fXIycmJXXfdda19jjzyyHjggQcyr6nz58+P559/Po466qgS+997771Rr169OPPMM+O6666Lzp07xwUXXBDnnntusb4lvReWRWn206VLl8a+++4bN9xwQ/Ts2TOuu+66OPnkk+Ojjz6Kzz//PLbbbru4+OKLIyLixBNPzOwDJQW2NWvWjEMOOSQef/zxYvva448/HsuXL49BgwZFRJT5Pb1z586RJEm8/vrra/1/11XrihUrolevXvHmm2/G73//+7jpppvixBNPjE8++STzGWbgwIHx6quvxty5c4usd+LEiTFnzpxM7UAVldWDB4FN2uq5fF588cXk66+/TmbPnp088sgjSbNmzZLc3Nxk9uzZmb77779/suOOOxaZH6GwsDDZa6+9kg4dOmTadt555+TAAw9c53ZXz9e02nvvvZdERHLKKacU6XfUUUcVmwdm8ODBSdu2bX92nZ9++mlSvXr15LLLLivS7/33309q1KhRrH1t6+vXr1+R9lNOOSWJiGTy5Mll3k5J81R99dVXSa1atZKePXsWmRfpxhtvTCIiufvuuzNtXbt2TSIiufXWW4tsa9SoUUm1atWS1157rUj7rbfemkRE8q9//Wud/+vq9Y4ZMybT9tFHHyURkVSrVi158803M+3jxo0rNidF//79k1q1aiUzZszItM2ZMyepX79+0qVLl0zb6aefnkRE8tZbbxX5//Pz84vNy7Lm/DhJ8uP8YsOGDSvStuZ8Rz/88EPSrl27pG3btsk333xTpO9P54359ttvi90P999/fxIRyauvvpppW9ecMW3btk0GDx5c7P/76eOwZMmSpF27dsmWW26ZeXxX17zddtsVmTfruuuuSyIief/994tta33MmzcvqVGjRnLHHXdk2vbaa6/k4IMPLtLv2muvTSIieeihhzJty5YtS7beeutic0qV9Dxc8z5dsWJF0qlTp2S//fbLtE2bNi2pVq1acsghhxSbB2z147NkyZKkYcOGyQknnFBk+dy5c5P8/Pwi7YMHD87Mi/dTu+66a9K5c+fM9XXNKXXXXXclEZHUqlUr6d69e3L++ecnr732WrH6XnvttSQiktGjRxdpf+6554q0r35OH3jggUX2uT/96U9JRBTZZ0qjpH21V69eyVZbbVWkLa05pVq3bp0MGDCgxGXvvPNOkTnLCgsLk9atWyd/+MMfivQraZ6ytdW84447Jq1bt06WLFmSaXvllVeSiCjxveDn/PWvf01q166dLF68OEmSJPnf//6XRETy2GOPFenXpUuXpH79+slnn31WpP2nj2lZXh+GDx+e1KxZM1m4cGGmbfny5UnDhg2T4447LtN2/PHHJy1btkzmz59fZH2DBg1K8vPzi+0PY8aMKfa6moajjz46adKkSbH21Y/hVVddlUydOrXIa+JNN92U1KtXL1m2bFkyePDgYnNKlbSvn3TSSUmdOnWKfPZY23thaZV2P73ggguSiEjGjh1bbB2r94N1zdO05uvk6vfPp556qki/Pn36FHk+l/U9fc6cOUlEJFdcccU6/++11Tpp0qQkIpKHH354rbf9+OOPk4hIbrjhhiLtp5xySlKvXr0SHzug6jBSCsi6Hj16RLNmzaJNmzZx2GGHRd26dePJJ5+M1q1bR0TEwoUL4+WXX44jjjgilixZEvPnz4/58+fHggULolevXjFt2rTM2foaNmwYH3zwQUybNq3U23/mmWciIuK0004r0n766aeX+38aO3ZsFBYWxhFHHJGpd/78+dGiRYvo0KFDjB8/vlTrGTZsWJHrv//974vUvL7befHFF2PFihVx+umnF5kX6YQTTogGDRoUG2WSm5tbbPj/ww8/HNttt1107NixSA377bdfRESp/td69eoV+aVz2223jYYNG8Z2221X5MxEq//+5JNPIuLHUSrPP/989O/fP7baaqtMv5YtW8ZRRx0VEydOzPyC/8wzz8Qee+wRu+22W6Zfs2bNKnR00KRJk2LmzJlx+umnF5t/6qeHd9auXTvz9/fffx/z58+PPfbYIyLiZ0f1rc0zzzwTu+22W+yzzz6Ztnr16sWJJ54Yn376aXz44YdF+g8dOrTIIWOrD5tdfd9WlAceeCCqVasWAwYMyLQdeeSR8eyzzxY57OyZZ56Jli1bFjkMo06dOsVGBq7NT+/Tb775JgoKCmLfffctcn8+/vjjUVhYGBdccEGxecBWPz4vvPBCLFq0KI488sgi+3P16tVj9913L3F/Pvnkk4tc33fffUt9Px533HHx3HPPRbdu3WLixIlxySWXxL777hsdOnQoMvLg4Ycfjvz8/Pj1r39dpK7OnTtHvXr1MnWtfk7//ve/L7LPlff17Kf3a0FBQcyfPz+6du0an3zySebQmjQtWLCgyCjanxo9enRsttlmmVErOTk5MXDgwHjggQeKHYJVGnPmzIn3338/jj322CJnauvatWvsuOOO5ap/9OjRceCBB0b9+vUjIqJDhw7RuXPnIodvff311/Hqq6/GcccdF1tssUWR25d0mHhpDBw4MFauXBljx47NtD3//POxaNGizOiXJEni0Ucfjb59+0aSJEX2s169ekVBQUGx16fVj8X8+fPLVVd5rWs/WG2HHXaInXbaKe6///6IiBgzZkwcfPDBUadOnRL7/3RfX/1ZY999941vv/02PvrooyJ9S3ovLK3S7qePPvpo7LzzzplRzj9Vnv1gv/32i6ZNmxaZq+6bb76JF154ocgIqLK+p6/vPpCfnx8RPx5Kura5ubbZZpvYZZdditS+atWqeOSRR6Jv375FHjug6hFKAVl30003xQsvvBCPPPJI9OnTJ+bPn19kcs7p06dHkiRx/vnnR7NmzYpcVs+TtHoS74svvjgWLVoU22yzTey4447xxz/+MaZMmbLO7X/22WdRrVq1aN++fZH2bbfdttz/07Rp0yJJkujQoUOxmv/73/9m6v05HTp0KHK9ffv2Ua1atcwhROu7nc8++ywiiv+vtWrViq222iqzfLXNN9+82Nw306ZNiw8++KDY9rfZZpuIiFL9r61bty72ITs/Pz/atGlTrC0iMmHG119/Hd9++22Jj9V2220XhYWFmbm1Pvvss2L3Z0n/+/pYPQ9ap06d1tlv4cKF8Yc//CE222yzqF27djRr1izatWsXEVHuL/qfffbZWu+H1ct/as0vu6u/WKw5P9FPrVixIubOnVvk8nNf9lfP37RgwYKYPn16TJ8+PXbddddYsWJFPPzww0Xq33rrrYvtB6V9fJ5++unYY489Ii8vLxo3bpw5bO6n9+eMGTOiWrVqsf322691PasD7f3226/YPv38888X25/z8vKKHTLVqFGjdd6Pa+rVq1eMGzcuFi1aFK+++moMGzYsPvvsszjooIMy25s2bVoUFBRE8+bNi9W1dOnSTL/Vj/Oa+3qzZs1+9kt8Sf71r39Fjx49om7dutGwYcNo1qxZZv6cbIRSEZE5HOunVq1aFQ888EB07949Zs6cmdnXdt9995g3b1689NJLZd7O6vty6623LraspLaf89///jcmTZoUe++9d6a+6dOnR7du3eLpp5/OBOirA82fex0pi5133jk6duxY5Ev9gw8+GE2bNs2EDV9//XUsWrQobr/99mL72OoAZs39f/VjUd6wbH2UtB+s6aijjoqHH344pk+fHq+//vpaD92L+PFQ8EMOOSTy8/OjQYMG0axZszj66KMjovi+XtJ7YWmUZT+dMWNGhe4DNWrUiAEDBsQTTzyRmRtq7NixsXLlyiKhVFnf09d3H2jXrl2ceeaZceedd0bTpk2jV69ecdNNNxW7zwcOHBj/+te/Mj9CvvLKK/HVV1+VepoAoPKqke0CAHbbbbf45S9/GRER/fv3j3322SeOOuqo+Pjjj6NevXqZiajPOuuszLwXa1r9BaFLly4xY8aMeOKJJ+L555+PO++8M/72t7/FrbfeGr/97W/Xu9a1feha84t5YWFh5OTkxLPPPhvVq1cv1v+nv7qvz/Y31HbWpqRfIwsLC2PHHXeMa665psTbrBkslaSk2tfVXpovI5XZEUccEa+//nr88Y9/jF122SWznx9wwAElTry+IZTnvn399deLzZ0yc+bMtU74/NN5UkoKBEePHl3qkVDr8tprr0W/fv2iS5cucfPNN0fLli2jZs2acc899xSb6PjnrL7/R40aFS1atCi2vEaNoh+d1nY/lkedOnVi3333jX333TeaNm0aI0aMiGeffTYGDx4chYWF0bx587VOiLy2ucbWx4wZM2L//fePjh07xjXXXBNt2rSJWrVqxTPPPBN/+9vfUttXf6pJkyYlBn4vv/xyfPnll/HAAw/EAw88UGz56NGjo2fPnhFR+tfxivaPf/wjIiLOOOOMOOOMM4otf/TRR8s9+qY0Bg4cGJdddlnMnz8/6tevH08++WQceeSRmX169eN59NFHx+DBg0tcx0477VTk+urHomnTphus7pKsbT9Y05FHHhnDhw+PE044IZo0aZLZB9a0aNGi6Nq1azRo0CAuvvjiaN++feTl5cV//vOfOOecc4rt6+UdmVOW/XRDGDRoUNx2223x7LPPRv/+/eOhhx6Kjh07xs4775zpU9b39IrYB66++uoYMmRI5rPbaaedFiNHjow333wzM2p+4MCBMXz48Hj44Yfj9NNPj4ceeijy8/PjgAMOKPd2gcpBKAVUKtWrV4+RI0dmJio/99xzM4dl1axZc61n2vmp1WeYGTp0aCxdujS6dOkSF1100VpDqbZt20ZhYWHMmDGjyKiMjz/+uFjfRo0alThJ+ZqjUNq3bx9JkkS7du0yvy6Wx7Rp0zIjaCJ+HDVWWFiYCQHWdztt27aNiB//158e/rZixYqYOXNmqe7v9u3bx+TJk2P//fdP/dfyZs2aRZ06dUp8rD766KOoVq1a5gN027ZtSzyss6Tbltfq0XZTp05d6333zTffxEsvvRQjRoyICy64INNeUm1luT/btm271vth9fL1tfPOOxc5k2VElBjcrDZ69OioWbNmjBo1qlh4M3HixLj++utj1qxZscUWW0Tbtm1j6tSpkSRJkf+7NI/Po48+Gnl5eTFu3LgioyzvueeeIv3at28fhYWF8eGHH8Yuu+xS4rpWP4bNmzcv1f5fGuV5XqwO6r/88stMXS+++GLsvffe6/xCvPpxnjZtWpHn9Ndff12m0VsREU899VQsX748nnzyySIj60p7+PGG0LFjx5g5c2ax9tGjR0fz5s1LnIB97Nix8dhjj8Wtt94atWvXzowYW/O1fM3X8dX3ZUlnpSzrmSqTJIkxY8ZE9+7d45RTTim2/JJLLonRo0fH0KFDM4/b1KlT17nOsu5XAwcOjBEjRsSjjz4am222WSxevLjIYdOrzz67atWqUu/7M2fOjGrVqq3X+1x5dOzYMUaPHh0FBQWZEbQl2WKLLWLvvfeOV155JX73u98VC5VXe+WVV2LBggUxduzYIhOFl7SvrY+y7Kft27ev8H2gS5cu0bJly3jwwQdjn332iZdffjlzRsfVyvqevvo+Wj0qt7y17rjjjrHjjjvGn//853j99ddj7733jltvvTUuvfTSiPhxRNVuu+0WDz74YJx66qkxduzY6N+/f5HXfKBqcvgeUOl069Ytdtttt7j22mvj+++/j+bNm0e3bt3itttuy3xB+6mvv/468/eCBQuKLKtXr15svfXWJZ7GeLXevXtHRMT1119fpL2kMw21b98+CgoKihwS+OWXXxY7u9mhhx4a1atXjxEjRhQbeZIkSbE612bND6433HBDkZrXdzs9evSIWrVqxfXXX1/k9nfddVcUFBSU6gyERxxxRHzxxRdxxx13FFv23XffleuMX6VVvXr16NmzZzzxxBNFzoo2b968GDNmTOyzzz7RoEGDiIjo06dPvPnmm/Hvf/870+/rr79e68iT8vjFL34R7dq1i2uvvbbYF97V9+/qcGbNx6uk/a1u3boRUfzLc0n69OkT//73v+ONN97ItC1btixuv/322HLLLdd5yFppNWrUKHr06FHkkpeXt9b+o0ePjn333TcGDhwYhx12WJHL6tPZr57vpU+fPjFnzpx45JFHMrf/9ttvM2exW5fq1atHTk5OkZEun376aTz++ONF+vXv3z+qVasWF198cbGRD6sfj169ekWDBg3iL3/5S7GzT0YUfb0prdVz2JT0OK7tsLLV88atDsqPOOKIWLVqVVxyySXF+v7www+Zdffo0SNq1qwZN9xwQ5F9rKT96+eUtK8WFBQUC/vStOeee8bUqVOLvKZ/9913MXbs2DjooIOK7WeHHXZYnHrqqbFkyZJ48sknI+LHsKl69erx6quvFln3zTffXOR6q1atolOnTnHffffF0qVLM+0TJkyI999/v0x1/+tf/4pPP/00hg4dWmKNAwcOjPHjx8ecOXOiWbNm0aVLl7j77rtj1qxZRdbz08eiLK8PET+GBjvuuGM8+OCD8eCDD0bLli2LBDDVq1ePAQMGxKOPPlpiGFLSvv/uu+/GDjvssM5gaEPYc889I0mSUp0F8dJLL40LL7wwMydjSUra11esWFFsn1gfZd1PBwwYEJMnTy7x7Kmr6yzrPlCtWrU47LDD4qmnnopRo0bFDz/8UOzwt7K+p7/77ruRk5MTe+65Z6Zt9TxcP51nam21Ll68OH744YcibTvuuGNUq1at2Ge3gQMHxptvvhl33313zJ8/36F7sJEwUgqolP74xz/G4YcfHvfee2+cfPLJcdNNN8U+++wTO+64Y5xwwgmx1VZbxbx58+KNN96Izz//PCZPnhwREdtvv31069YtOnfuHI0bN4533nknHnnkkTj11FPXuq1ddtkljjzyyLj55pujoKAg9tprr3jppZdK/CV80KBBcc4558QhhxwSp512Wnz77bdxyy23xDbbbFNkAtj27dvHpZdeGsOHD49PP/00+vfvH/Xr14+ZM2fGY489FieeeGKcddZZP3s/zJw5M/r16xcHHHBAvPHGG/GPf/wjjjrqqMxQ+/XdTrNmzWL48OExYsSIOOCAA6Jfv37x8ccfx8033xy/+tWvMvNprMsxxxwTDz30UJx88skxfvz42HvvvWPVqlXx0UcfxUMPPRTjxo3LjPrYEC699NJ44YUXYp999olTTjklatSoEbfddlssX748rrzyyky/s88+O0aNGhUHHHBA/OEPf4i6devG7bffHm3btv3ZecdKq1q1anHLLbdE3759Y5dddomhQ4dGy5Yt46OPPooPPvggxo0bFw0aNIguXbrElVdeGStXrozNN988nn/++RJ/ke/cuXNERJx33nkxaNCgqFmzZvTt2zfz4f6nzj333Lj//vujd+/ecdppp0Xjxo3j73//e8ycOTMeffTRYhN7b2hvvfVWTJ8+fa3Pvc033zx+8YtfxOjRo+Occ86JE044IW688cY49thj4913342WLVvGqFGj1jop8U8deOCBcc0118QBBxwQRx11VHz11Vdx0003xdZbb13ksd16663jvPPOy0wmfuihh0Zubm68/fbb0apVqxg5cmQ0aNAgbrnlljjmmGPiF7/4RQwaNCiaNWsWs2bNin/+85+x9957x4033lim+6J27dqx/fbbx4MPPhjbbLNNNG7cODp16hSdOnWKgw8+ONq1axd9+/aN9u3bx7Jly+LFF1+Mp556Kn71q19F3759I+LHybVPOumkGDlyZLz33nvRs2fPqFmzZkybNi0efvjhuO666+Kwww6LZs2axVlnnRUjR46Mgw46KPr06ROTJk2KZ599tsyH1/Ts2TNq1aoVffv2jZNOOimWLl0ad9xxRzRv3rzEHwjScPDBB8cll1wSEyZMyBzm9OSTT8aSJUuiX79+Jd5mjz32iGbNmsXo0aNj4MCBkZ+fH4cffnjccMMNkZOTE+3bt4+nn366xPnv/vKXv8TBBx8ce++9dwwdOjS++eabuPHGG6NTp05FgqqfM3r06Khevfpag/5+/frFeeedFw888ECceeaZcf3118c+++wTv/jFL+LEE0+Mdu3axaeffhr//Oc/47333ouIsr0+rDZw4MC44IILIi8vL44//vhirwuXX355jB8/Pnbfffc44YQTYvvtt4+FCxfGf/7zn3jxxRdj4cKFmb4rV66MCRMmlDjya0PbZ599okmTJvHiiy9m5sRam65du0bXrl3X2WevvfaKRo0axeDBg+O0006LnJycGDVqVIUeKl7W/fSPf/xjPPLII3H44YfHcccdF507d46FCxfGk08+GbfeemvsvPPO0b59+2jYsGHceuutUb9+/ahbt27svvvuRUZYr2ngwIFxww03xIUXXhg77rhjsRFOZX1Pf+GFF2LvvfeOJk2aZNr+/e9/R/fu3ePCCy+Miy66KCJirbVOnjw5Tj311Dj88MNjm222iR9++CEzuvanJ8iI+DEwO+uss+Kss86Kxo0bV9hoViDL0jjFH0BJ7rnnniQikrfffrvYslWrViXt27dP2rdvn/zwww9JkiTJjBkzkmOPPTZp0aJFUrNmzWTzzTdPDjrooOSRRx7J3O7SSy9Ndtttt6Rhw4ZJ7dq1k44dOyaXXXZZsmLFikyfCy+8MFnz5e+7775LTjvttKRJkyZJ3bp1k759+yazZ88u8TTuzz//fNKpU6ekVq1aybbbbpv84x//KHGdSZIkjz76aLLPPvskdevWTerWrZt07NgxGTZsWPLxxx+v875Zvb4PP/wwOeyww5L69esnjRo1Sk499dTku+++K9d2Vq/z66+/Lnb7G2+8MenYsWNSs2bNZLPNNkt+97vfJd98802RPl27dk122GGHEutdsWJFcsUVVyQ77LBDkpubmzRq1Cjp3LlzMmLEiKSgoGCd/+va1tu2bdvkwAMPLNYeEcmwYcOKtP3nP/9JevXqldSrVy+pU6dO0r179+T1118vdtspU6YkXbt2TfLy8pLNN988ueSSS5K77rqr2GnV1zy9/dq2u7ZTy0+cODH59a9/ndSvXz+pW7dustNOOxU5lfXnn3+eHHLIIUnDhg2T/Pz85PDDD8+cVnvN/e2SSy5JNt9886RatWpF6lzzlO9J8uNz5LDDDksaNmyY5OXlJbvttlvy9NNPl1jzmqffXn0q9ZJOLV4ev//975OISGbMmLHWPhdddFESEcnkyZOTJEmSzz77LOnXr19Sp06dpGnTpskf/vCH5Lnnnit2H695qvMkSZK77ror6dChQ5Kbm5t07Ngxueeee9b6vLz77ruTXXfdNbOvdu3aNXN69tXGjx+f9OrVK8nPz0/y8vKS9u3bJ0OGDEneeeedInWseVr5JCn5Neb1119POnfunNSqVavI43z//fcngwYNStq3b5/Url07ycvLS7bffvvkvPPOSxYvXlxs3bfffnvSuXPnpHbt2kn9+vWTHXfcMTn77LOTOXPmZPqsWrUqGTFiRNKyZcukdu3aSbdu3ZKpU6eWuM/8nCeffDLZaaedkry8vGTLLbdMrrjiiuTuu+/+2edMRe9PP7XTTjslxx9/fOZ63759k7y8vGTZsmVrvc2QIUOSmjVrJvPnz0+SJEm+/vrrZMCAAUmdOnWSRo0aJSeddFIyderUEmt+4IEHko4dOya5ublJp06dkieffDIZMGBA0rFjx1LVu2LFiqRJkybJvvvuu85+7dq1S3bdddfM9alTp2ZeJ/Ly8pJtt902Of/884vcpiyvD0mSJNOmTUsiIomIZOLEiSXWMW/evGTYsGFJmzZtkpo1ayYtWrRI9t9//+T2228v0u/ZZ59NIiKZNm1aKe6FinfaaaclW2+9dZG21fvdVVddtc7blvTc/de//pXsscceSe3atZNWrVolZ599djJu3Lhirz/rei9cl/LspwsWLEhOPfXUZPPNN09q1aqVtG7dOhk8eHBmeZIkyRNPPJFsv/32SY0aNYrsvyW9TiZJkhQWFiZt2rRJIiK59NJLS6yjtO/pixYtSmrVqpXceeedRW6/+n1mzfezkmr95JNPkuOOOy5p3759kpeXlzRu3Djp3r178uKLL5ZY2957751ERPLb3/52rfcjULXkJEkVny0WANikHHPMMfHGG2+UeV6fTd2WW24Z3bp1i3vvvTfbpayXUaNGxbBhw2LWrFnRsGHDrNSwyy67RLNmzYrNsbYp6d+/f+Tk5JR4eFkaPvnkk+jYsWM8++yzsf/++2elhk3dtddeG1deeWXMmDGj3JO/A5hTCgCoUr788svUz/ZF5fGb3/wmtthiixIni65oK1euLDbfzSuvvBKTJ0+Obt26bfDtV1b//e9/4+mnny5xjrO0bLXVVnH88cfH5ZdfnrUaNmUrV66Ma665Jv785z8LpID1Yk4pAKBKmDJlSjz++OPx6quvZiZKp3zmzp27zuW1a9eusMmrV6xYUWQuopKsXLkyatasudbltWrVisaNG0fEj3O3/dxZySrKF198ET169Iijjz46WrVqFR999FHceuut0aJFizj55JMjImLhwoWxYsWKta6jevXq0axZs1TqTct2221XLKz77rvvoqCgYJ23W7FiRdSqVWuty2vXrh3ffffdOtfx0zN+3nLLLaWodsNbunTpz84x1qxZs2JnIa3KatasWWwifoDyEEoBAFXC2LFj44YbbohBgwbF8OHDs11OldayZct1Lh88eHCFHeb3+uuvR/fu3dfZp23btvHZZ5+tdXnXrl3jlVdeqZB6yqJRo0bRuXPnuPPOO+Prr7+OunXrxoEHHhiXX355ZmLnQw89NCZMmLDWdbRt27bI2UE3Vg8++GAMHTp0vdYxePDg+Pvf/77OPpVx5pG//vWvMWLEiHX2mTlzZmy55ZbpFARQhWR1Tqlbbrklbrnllswb9Q477BAXXHBB5lTn3bp1K/Ymf9JJJ8Wtt96adqkAABuNF198cZ3LW7VqFdtvv32FbOubb76Jd999d519fm6EzOpwqDJ6991345tvvlnr8tq1a8fee++dYkXZ8eWXX8YHH3ywzj7ffPNNNGrUaK3LW7VqFXPmzFnnOirjGdc++eST+OSTT9bZZ5999om8vLyUKgKoOrIaSj311FNRvXr16NChQyRJEn//+9/jqquuikmTJsUOO+wQ3bp1i2222SYuvvjizG3q1KkTDRo0yFbJAAAAAFSArB6+17dv3yLXL7vssrjlllvizTffjB122CEifgyhfnrsOAAAAABVX6WZU2rVqlXx8MMPx7Jly2LPPffMtI8ePTr+8Y9/RIsWLaJv375x/vnnR506dda6nuXLl8fy5csz1wsLC2PhwoXRpEmTyMnJ2aD/AwAAAMCmLkmSWLJkSbRq1SqqVau21n5ZD6Xef//92HPPPeP777+PevXqxWOPPZaZw+Coo46Ktm3bRqtWrWLKlClxzjnnxMcffxxjx45d6/pGjhz5sxMNAgAAALBhzZ49O1q3br3W5VmdUyrix1PDzpo1KwoKCuKRRx6JO++8MyZMmFDi5Jovv/xy7L///jF9+vRo3759ietbc6RUQUFBbLHFFjF79mxzUQEAAABsYIsXL442bdrEokWLIj8/f639sh5KralHjx7Rvn37uO2224otW7ZsWdSrVy+ee+656NWrV6nWt3jx4sjPz4+CggKhFAAAAMAGVtosZu0H9mVJYWFhkZFOP/Xee+9FRETLli1TrAgAAACAipbVOaWGDx8evXv3ji222CKWLFkSY8aMiVdeeSXGjRsXM2bMiDFjxkSfPn2iSZMmMWXKlDjjjDOiS5cusdNOO2WzbAAAAADWU1ZDqa+++iqOPfbY+PLLLyM/Pz922mmnGDduXPz617+O2bNnx4svvhjXXnttLFu2LNq0aRMDBgyIP//5z9ksGQAAAIAKUOnmlKpo5pQCAAAASE+VnVMKAAAAgI2fUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEhdVkOpW265JXbaaado0KBBNGjQIPbcc8949tlnM8u///77GDZsWDRp0iTq1asXAwYMiHnz5mWxYgAAAAAqQlZDqdatW8fll18e7777brzzzjux3377xcEHHxwffPBBREScccYZ8dRTT8XDDz8cEyZMiDlz5sShhx6azZIBAAAAqAA5SZIk2S7ipxo3bhxXXXVVHHbYYdGsWbMYM2ZMHHbYYRER8dFHH8V2220Xb7zxRuyxxx6lWt/ixYsjPz8/CgoKokGDBhuydAAAAIBNXmmzmEozp9SqVavigQceiGXLlsWee+4Z7777bqxcuTJ69OiR6dOxY8fYYost4o033ljrepYvXx6LFy8ucgEAAACgcsl6KPX+++9HvXr1Ijc3N04++eR47LHHYvvtt4+5c+dGrVq1omHDhkX6b7bZZjF37ty1rm/kyJGRn5+fubRp02YD/wcAAAAAlFXWQ6ltt9023nvvvXjrrbfid7/7XQwePDg+/PDDcq9v+PDhUVBQkLnMnj27AqsFAAAAoCLUyHYBtWrViq233joiIjp37hxvv/12XHfddTFw4MBYsWJFLFq0qMhoqXnz5kWLFi3Wur7c3NzIzc3d0GUDAAAAsB6yPlJqTYWFhbF8+fLo3Llz1KxZM1566aXMso8//jhmzZoVe+65ZxYrBAAAAGB9ZXWk1PDhw6N3796xxRZbxJIlS2LMmDHxyiuvxLhx4yI/Pz+OP/74OPPMM6Nx48bRoEGD+P3vfx977rlnqc+8BwAAAEDllNVQ6quvvopjjz02vvzyy8jPz4+ddtopxo0bF7/+9a8jIuJvf/tbVKtWLQYMGBDLly+PXr16xc0335zNkgEAAACoADlJkiTZLmJDWrx4ceTn50dBQUE0aNAg2+UAAAAAbNRKm8VUujmlAAAAANj4CaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASF1WQ6mRI0fGr371q6hfv340b948+vfvHx9//HGRPt26dYucnJwil5NPPjlLFQMAAABQEbIaSk2YMCGGDRsWb775ZrzwwguxcuXK6NmzZyxbtqxIvxNOOCG+/PLLzOXKK6/MUsUAAAAAVIQa2dz4c889V+T6vffeG82bN4933303unTpkmmvU6dOtGjRIu3yAAAAANhAKtWcUgUFBRER0bhx4yLto0ePjqZNm0anTp1i+PDh8e2332ajPAAAAAAqSFZHSv1UYWFhnH766bH33ntHp06dMu1HHXVUtG3bNlq1ahVTpkyJc845Jz7++OMYO3ZsietZvnx5LF++PHN98eLFG7x2AAAAAMqm0oRSw4YNi6lTp8bEiROLtJ944omZv3fcccdo2bJl7L///jFjxoxo3759sfWMHDkyRowYscHrBQAAAKD8KsXhe6eeemo8/fTTMX78+GjduvU6++6+++4RETF9+vQSlw8fPjwKCgoyl9mzZ1d4vQAAAACsn6yOlEqSJH7/+9/HY489Fq+88kq0a9fuZ2/z3nvvRUREy5YtS1yem5sbubm5FVkmAAAAABUsq6HUsGHDYsyYMfHEE09E/fr1Y+7cuRERkZ+fH7Vr144ZM2bEmDFjok+fPtGkSZOYMmVKnHHGGdGlS5fYaaedslk6AAAAAOshJ0mSJGsbz8kpsf2ee+6JIUOGxOzZs+Poo4+OqVOnxrJly6JNmzZxyCGHxJ///Odo0KBBqbaxePHiyM/Pj4KCglLfBgAAAIDyKW0Wk/XD99alTZs2MWHChJSqAQAAACAtlWKicwAAAAA2LUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdVkNpUaOHBm/+tWvon79+tG8efPo379/fPzxx0X6fP/99zFs2LBo0qRJ1KtXLwYMGBDz5s3LUsUAAAAAVISshlITJkyIYcOGxZtvvhkvvPBCrFy5Mnr27BnLli3L9DnjjDPiqaeeiocffjgmTJgQc+bMiUMPPTSLVQMAAACwvnKSJEmyXcRqX3/9dTRv3jwmTJgQXbp0iYKCgmjWrFmMGTMmDjvssIiI+Oijj2K77baLN954I/bYY4+fXefixYsjPz8/CgoKokGDBhv6XwAAAADYpJU2i6lUc0oVFBRERETjxo0jIuLdd9+NlStXRo8ePTJ9OnbsGFtssUW88cYbWakRAAAAgPVXI9sFrFZYWBinn3567L333tGpU6eIiJg7d27UqlUrGjZsWKTvZpttFnPnzi1xPcuXL4/ly5dnri9evHiD1QwAAABA+VSakVLDhg2LqVOnxgMPPLBe6xk5cmTk5+dnLm3atKmgCgEAAACoKJUilDr11FPj6aefjvHjx0fr1q0z7S1atIgVK1bEokWLivSfN29etGjRosR1DR8+PAoKCjKX2bNnb8jSAQAAACiHrIZSSZLEqaeeGo899li8/PLL0a5duyLLO3fuHDVr1oyXXnop0/bxxx/HrFmzYs899yxxnbm5udGgQYMiFwAAAAAql6zOKTVs2LAYM2ZMPPHEE1G/fv3MPFH5+flRu3btyM/Pj+OPPz7OPPPMaNy4cTRo0CB+//vfx5577lmqM+8BAAAAUDnlJEmSZG3jOTkltt9zzz0xZMiQiIj4/vvv4//+7//i/vvvj+XLl0evXr3i5ptvXuvhe2sq7WkIAQAAAFh/pc1ishpKpUEoBQAAAJCe0mYxlWKicwAAAAA2LUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFJX7lBq0aJFceedd8bw4cNj4cKFERHxn//8J7744osKKw4AAACAjVON8txoypQp0aNHj8jPz49PP/00TjjhhGjcuHGMHTs2Zs2aFffdd19F1wkAAADARqRcI6XOPPPMGDJkSEybNi3y8vIy7X369IlXX321wooDAAAAYONUrlDq7bffjpNOOqlY++abbx5z585d76IAAAAA2LiVK5TKzc2NxYsXF2v/3//+F82aNVvvogAAAADYuJUrlOrXr19cfPHFsXLlyoiIyMnJiVmzZsU555wTAwYMqNACAQAAANj4lCuUuvrqq2Pp0qXRvHnz+O6776Jr166x9dZbR/369eOyyy6r6BoBAAAA2MiU6+x7+fn58cILL8TEiRNjypQpsXTp0vjFL34RPXr0qOj6AAAAANgI5SRJkmS7iA1p8eLFkZ+fHwUFBdGgQYNslwMAAACwUSttFlOukVLXX399ie05OTmRl5cXW2+9dXTp0iWqV69entUDAAAAsJErVyj1t7/9Lb7++uv49ttvo1GjRhER8c0330SdOnWiXr168dVXX8VWW20V48ePjzZt2lRowQAAAABUfeWa6Pwvf/lL/OpXv4pp06bFggULYsGCBfG///0vdt9997juuuti1qxZ0aJFizjjjDMqul4AAAAANgLlmlOqffv28eijj8Yuu+xSpH3SpEkxYMCA+OSTT+L111+PAQMGxJdffllRtZaLOaUAAAAA0lPaLKZcI6W+/PLL+OGHH4q1//DDDzF37tyIiGjVqlUsWbKkPKsHAAAAYCNXrlCqe/fucdJJJ8WkSZMybZMmTYrf/e53sd9++0VExPvvvx/t2rWrmCoBAAAA2KiUK5S66667onHjxtG5c+fIzc2N3Nzc+OUvfxmNGzeOu+66KyIi6tWrF1dffXWFFgsAAADAxqFcc0qt9tFHH8X//ve/iIjYdtttY9ttt62wwiqKOaUAAAAA0lPaLKbG+mykY8eO0bFjx/VZBQAAAACboHKHUp9//nk8+eSTMWvWrFixYkWRZddcc816FwYAAADAxqtcodRLL70U/fr1i6222io++uij6NSpU3z66aeRJEn84he/qOgaAQAAANjIlGui8+HDh8dZZ50V77//fuTl5cWjjz4as2fPjq5du8bhhx9e0TUCAAAAsJEpVyj13//+N4499tiIiKhRo0Z89913Ua9evbj44ovjiiuuqNACAQAAANj4lCuUqlu3bmYeqZYtW8aMGTMyy+bPn18xlQEAAACw0SrXnFJ77LFHTJw4Mbbbbrvo06dP/N///V+8//77MXbs2Nhjjz0qukYAAAAANjLlCqWuueaaWLp0aUREjBgxIpYuXRoPPvhgdOjQwZn3AAAAAPhZOUmSJNkuYkNavHhx5OfnR0FBQTRo0CDb5QAAAABs1EqbxZRrTqmtttoqFixYUKx90aJFsdVWW5VnlQAAAABsQsoVSn366aexatWqYu3Lly+PL774Yr2LAgAAAGDjVqY5pZ588snM3+PGjYv8/PzM9VWrVsVLL70UW265ZYUVBwAAAMDGqUyhVP/+/SMiIicnJwYPHlxkWc2aNWPLLbeMq6++usKKAwAAAGDjVKZQqrCwMCIi2rVrF2+//XY0bdp0gxQFAAAAwMatTKHUajNnzqzoOgAAAADYhJQrlIqIeOmll+Kll16Kr776KjOCarW77757vQsDAAAAYONVrlBqxIgRcfHFF8cvf/nLaNmyZeTk5FR0XQAAAABsxMoVSt16661x7733xjHHHFPR9QAAAACwCahWnhutWLEi9tprr4quBQAAAIBNRLlCqd/+9rcxZsyYiq4FAAAAgE1EuQ7f+/777+P222+PF198MXbaaaeoWbNmkeXXXHNNhRQHAAAAwMapXKHUlClTYpdddomIiKlTpxZZZtJzAAAAAH5OuUKp8ePHV3QdAAAAAGxCyjWn1GrTp0+PcePGxXfffRcREUmSVEhRAAAAAGzcyhVKLViwIPbff//YZpttok+fPvHll19GRMTxxx8f//d//1ehBQIAAACw8SlXKHXGGWdEzZo1Y9asWVGnTp1M+8CBA+O5556rsOIAAAAA2DiVa06p559/PsaNGxetW7cu0t6hQ4f47LPPKqQwAAAAADZe5RoptWzZsiIjpFZbuHBh5ObmrndRAAAAAGzcyhVK7bvvvnHfffdlrufk5ERhYWFceeWV0b179worDgAAAICNU7kO37vyyitj//33j3feeSdWrFgRZ599dnzwwQexcOHC+Ne//lXRNQIAAACwkSnXSKlOnTrF//73v9hnn33i4IMPjmXLlsWhhx4akyZNivbt21d0jQAAAABsZHKSJEmyXcSGtHjx4sjPz4+CgoJo0KBBtssBAAAA2KiVNosp10ipe+65Jx5++OFi7Q8//HD8/e9/L88qAQAAANiElCuUGjlyZDRt2rRYe/PmzeMvf/nLehcFAAAAwMatXKHUrFmzol27dsXa27ZtG7NmzVrvogAAAADYuJUrlGrevHlMmTKlWPvkyZOjSZMm610UAAAAABu3coVSRx55ZJx22mkxfvz4WLVqVaxatSpefvnl+MMf/hCDBg2q6BoBAAAA2MjUKM+NLrnkkvj0009j//33jxo1flxFYWFhHHvsseaUAgAAAOBn5SRJkpTlBkmSxOzZs6NZs2bx+eefx3vvvRe1a9eOHXfcMdq2bbuh6iy30p6GEAAAAID1V9ospswjpZIkia233jo++OCD6NChQ3To0GG9CgUAAABg01PmOaWqVasWHTp0iAULFmyIegAAAADYBJRrovPLL788/vjHP8bUqVMruh4AAAAANgFlnlMqIqJRo0bx7bffxg8//BC1atWK2rVrF1m+cOHCCitwfZlTCgAAACA9G2xOqYiIa6+9trx1AQAAAED5QqnBgwdXdB0AAAAAbELKNadURMSMGTPiz3/+cxx55JHx1VdfRUTEs88+Gx988EGFFQcAAADAxqlcodSECRNixx13jLfeeivGjh0bS5cujYiIyZMnx4UXXlihBQIAAACw8SlXKHXuuefGpZdeGi+88ELUqlUr077ffvvFm2++WWHFAQAAALBxKlco9f7778chhxxSrL158+Yxf/789S4KAAAAgI1buUKphg0bxpdfflmsfdKkSbH55puvd1EAAAAAbNzKFUoNGjQozjnnnJg7d27k5OREYWFh/Otf/4qzzjorjj322IquEQAAAICNTLlCqb/85S+x3XbbxRZbbBFLly6N7bffPrp06RJ77bVX/PnPf67oGgEAAADYyNQoS+fCwsK46qqr4sknn4wVK1bEMcccEwMGDIilS5fGrrvuGh06dNhQdQIAAACwESlTKHXZZZfFRRddFD169IjatWvHmDFjIkmSuPvuuzdUfQAAAABshMp0+N59990XN998c4wbNy4ef/zxeOqpp2L06NFRWFi4oeoDAAAAYCNUplBq1qxZ0adPn8z1Hj16RE5OTsyZM6fCCwMAAABg41WmUOqHH36IvLy8Im01a9aMlStXVmhRAAAAAGzcyjSnVJIkMWTIkMjNzc20ff/993HyySdH3bp1M21jx46tuAoBAAAA2OiUKZQaPHhwsbajjz66wooBAAAAYNNQplDqnnvu2VB1AAAAALAJKdOcUgAAAABQEYRSAAAAAKQuq6HUq6++Gn379o1WrVpFTk5OPP7440WWDxkyJHJycopcDjjggOwUCwAAAECFyWootWzZsth5553jpptuWmufAw44IL788svM5f7770+xQgAAAAA2hDJNdF7RevfuHb17915nn9zc3GjRokVKFQEAAACQhko/p9Qrr7wSzZs3j2233TZ+97vfxYIFC9bZf/ny5bF48eIiFwAAAAAql0odSh1wwAFx3333xUsvvRRXXHFFTJgwIXr37h2rVq1a621GjhwZ+fn5mUubNm1SrBgAAACA0shJkiTJdhERETk5OfHYY49F//7919rnk08+ifbt28eLL74Y+++/f4l9li9fHsuXL89cX7x4cbRp0yYKCgqiQYMGFV02AAAAAD+xePHiyM/P/9ksplKPlFrTVlttFU2bNo3p06evtU9ubm40aNCgyAUAAACAyqVKhVKff/55LFiwIFq2bJntUgAAAABYD1k9+97SpUuLjHqaOXNmvPfee9G4ceNo3LhxjBgxIgYMGBAtWrSIGTNmxNlnnx1bb7119OrVK4tVAwAAALC+shpKvfPOO9G9e/fM9TPPPDMiIgYPHhy33HJLTJkyJf7+97/HokWLolWrVtGzZ8+45JJLIjc3N1slAwAAAFABKs1E5xtKaSfXAgAAAGD9bZQTnQMAAACwcRBKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJC6GtkuAAAAIFsunzS/1H3P3bXpBqwEYNNjpBQAAAAAqTNSCiDbcnLK1j9JNkwdALCa9yYAUmCkFAAAAACpE0oBAAAAkDqhFAAAAACpE0oBAAAAkDqhFAAAAACpE0oBAAAAkDqhFAAAAACpE0oBAAAAkDqhFAAAAACpE0oBAAAAkDqhFAAAAACpE0oBAAAAkDqhFAAAAACpE0oBAAAAkDqhFAAAAACpE0oBAAAAkDqhFAAAAACpq5HtAgAom8snzS9T/3N3bbqBKgEAACg/I6UAAAAASJ1QCgAAAIDUCaUAAICNR05O2S4AZI1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUZTWUevXVV6Nv377RqlWryMnJiccff7zI8iRJ4oILLoiWLVtG7dq1o0ePHjFt2rTsFAsAAABAhclqKLVs2bLYeeed46abbipx+ZVXXhnXX3993HrrrfHWW29F3bp1o1evXvH999+nXCkAAAAAFalGNjfeu3fv6N27d4nLkiSJa6+9Nv785z/HwQcfHBER9913X2y22Wbx+OOPx6BBg9IsFQAAAIAKVGnnlJo5c2bMnTs3evTokWnLz8+P3XffPd5444213m758uWxePHiIhcAAAAAKpdKG0rNnTs3IiI222yzIu2bbbZZZllJRo4cGfn5+ZlLmzZtNmidAAAAAJRdpQ2lymv48OFRUFCQucyePTvbJQEAAACwhkobSrVo0SIiIubNm1ekfd68eZllJcnNzY0GDRoUuQAAAABQuVTaUKpdu3bRokWLeOmllzJtixcvjrfeeiv23HPPLFYGAAAAwPrK6tn3li5dGtOnT89cnzlzZrz33nvRuHHj2GKLLeL000+PSy+9NDp06BDt2rWL888/P1q1ahX9+/fPXtFQwS6fNL9M/c/dtekGqgQAAADSk9VQ6p133onu3btnrp955pkRETF48OC499574+yzz45ly5bFiSeeGIsWLYp99tknnnvuucjLy8tWyQAAAABUgKyGUt26dYskSda6PCcnJy6++OK4+OKLU6wKAAAAgA2t0s4pBQAAAMDGSygFAAAAQOqEUgAAAACkTigFAAAAQOqyOtE5AFC15eTklKn/uk5wAgDApsVIKQAAAABSJ5QCAAAAIHVCKQAAAABSJ5QCAAAAIHVCKQAAAABSJ5QCAAAAIHU1sl0AAADA2uTk5JSpf7KB6gCg4hkpBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApK5GtgsAAAAgInJyytY/STZMHQApMVIKAAAAgNQZKQUAAKyXyyfNL3Xfc3dtugErAaAqMVIKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNTVyHYBwIaVk5NTpv5JkmygSgAAAOD/MVIKAAAAgNQZKQWU2+WT5pep/7m7Nt1AlQAAAFDVGCkFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqcfQ8ASE9OTtn6J8mGqQMAgKwzUgoAAACA1BkpBQDARuPySfNL3ffcXZtuwEoAgJ9jpBQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6E52TUZaJQSNMDgoAAACUn5FSAAAAAKTOSKkqJicnp0z9kyTZQJUAAAAAlJ+RUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOpqZLsAAAA2HTk5OWXqnyTJBqoEAMg2I6UAAAAASJ1QamOXk1P6CwBAZVOWzzI+zwBAlSKUAgAAACB15pQC2ADKMmeK2VIAAIBNkZFSAAAAAKTOSCkAoNK6fNL8MvU/d9emG6gSAAAqmpFSAAAAAKTOSCkAyLKyzEEWEZEkZiIDAKDqM1IKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInYnOAQBgI1fmEypsoDqoOpyEA0iDkVIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApK5GtgsAAMooJ6ds/ZNkw9QBAKuV5b3J+xLw/6vUI6UuuuiiyMnJKXLp2LFjtstiE7TmfvhzFwAA8BkSYN0q/UipHXbYIV588cXM9Ro1Kn3JAAAAAPyMSp/w1KhRI1q0aJHtMgAAAACoQJX68L2IiGnTpkWrVq1iq622it/85jcxa9asbJcEAAAAwHqq1COldt9997j33ntj2223jS+//DJGjBgR++67b0ydOjXq169f4m2WL18ey5cvz1xfvHhxWuUCAAAAUEqVOpTq3bt35u+ddtopdt9992jbtm089NBDcfzxx5d4m5EjR8aIESPSKhEAAACAcqj0h+/9VMOGDWObbbaJ6dOnr7XP8OHDo6CgIHOZPXt2ihUCAAAAUBpVKpRaunRpzJgxI1q2bLnWPrm5udGgQYMiFwAAWF85OTllugAA61apQ6mzzjorJkyYEJ9++mm8/vrrccghh0T16tXjyCOPzHZpAAAAAKyHSj2n1Oeffx5HHnlkLFiwIJo1axb77LNPvPnmm9GsWbNslwYAAADAeqjUodQDDzyQ7RIAAAAA2AAq9eF7AAAAAGycKvVIKQBg/V0+aX6Z+p+7a9MNVAkAAPw/RkoBAAAAkDqhFAAAAACpE0oBAAAAkDqhFAAAAACpE0oBAAAAkDqhFAAAAACpq5HtAgCownJyytY/STZMHcC6ea5mh/sdANZJKAXARiOnDF8AE1/+AAAgqxy+BwAAAEDqhFIAAAAApM7hewAAUAlcPml+qfueu2vTDVgJAKRDKAUAAFAFCTKBqs7hewAAAACkzkgpAACKKMvoiwgjMACA8jFSCgAAAIDUGSkFAABAaozGBFYzUgoAAACA1AmlAAAAAEidUAoAAACA1JlTCgCgisnJySlT/2QD1QEAsD6MlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdc6+B1RaZTm7VJI4txQAAEBVYqQUbAg5OaW/VDZVuXYAAACqDKEUAAAAAKkTSgEAAACQOqEUAAAAAKkTSgEAAACQOmffAyA1l0+aX6b+5+7adANVAgAAZJuRUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkrka2CwCgcsnJySl132QD1gEAsKkpy+ewiIgk8WmMqs1IKQAAAABSJ5QCAAAAIHVCKQAAAABSZ04pACiFyyfNL1P/c3dtuoEqAQCAjYORUgAAAACkzkgpYONQxjOVhDOVwCavzGc4KusGvM4AwNqV5X3Ye+pGy0gpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFLn7HvAJunySfPL1P/cXZtuoEoAAAA2TUZKAQAAAJA6I6VITU5OTpn6J0mygSoBAAAAss1IKQAAAABSJ5QCAAAAIHVCKQAAAABSJ5QCAAAAIHVCKQAAAABSJ5QCAAAAIHVCKQAAAABSJ5QCAAAAIHU1sl0AAMDG6PJJ88vU/9xdm26gSgAAKiehFAAAAGzk/FhCZeTwPQAAAABSJ5QCAACAqignp/QXqIQcvgcAAMBGI6eMAUySJBuoEuDnGCkFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqcfY+NwuWT5pep/7m7Nt1AlQAAAAClYaQUAAAAAKkTSgEAAACQOqEUAAAAAKkTSgEAAACQOqEUAAAAAKkTSgEAAACQOqEUAAAAAKkTSgEAAACQuhrZLqA0brrpprjqqqti7ty5sfPOO8cNN9wQu+22W7bLAgAAYBNy+aT5Zep/7q5NN1Almxb3+8ar0o+UevDBB+PMM8+MCy+8MP7zn//EzjvvHL169Yqvvvoq26UBAAAAUE6VPpS65ppr4oQTToihQ4fG9ttvH7feemvUqVMn7r777myXBgAAAEA5VerD91asWBHvvvtuDB8+PNNWrVq16NGjR7zxxhsl3mb58uWxfPnyzPWCgoKIiFi8ePGGLbaSKst//f3SJWVb9+JaZSumjMrymFW62svQV+0Vo6zPcLVXjCpdexnfFypV7WXsr/aKofZ1rL+K1l6Z3lMj1F5R1L6WdVei2ivTa8yP6/e9o8R1b+Day6Iq176pWv28SpJknf1ykp/rkUVz5syJzTffPF5//fXYc889M+1nn312TJgwId56661it7noootixIgRaZYJAAAAwBpmz54drVu3XuvySj1SqjyGDx8eZ555ZuZ6YWFhLFy4MJo0aRI5OTlZrKzyWLx4cbRp0yZmz54dDRo0yHY5pVZV645Qe7aoPTvUnh1qzw61Z4fas0Pt2aH27KiqtVfVuiPUvrFJkiSWLFkSrVq1Wme/Sh1KNW3aNKpXrx7z5s0r0j5v3rxo0aJFibfJzc2N3NzcIm0NGzbcUCVWaQ0aNKiST5iqWneE2rNF7dmh9uxQe3aoPTvUnh1qzw61Z0dVrb2q1h2h9o1Jfn7+z/ap1BOd16pVKzp37hwvvfRSpq2wsDBeeumlIofzAQAAAFC1VOqRUhERZ555ZgwePDh++ctfxm677RbXXnttLFu2LIYOHZrt0gAAAAAop0ofSg0cODC+/vrruOCCC2Lu3Lmxyy67xHPPPRebbbZZtkursnJzc+PCCy8sdphjZVdV645Qe7aoPTvUnh1qzw61Z4fas0Pt2aH27KiqtVfVuiPUvqmq1GffAwAAAGDjVKnnlAIAAABg4ySUAgAAACB1QikAAAAAUieUAgAAgCpkzpw52S4BKoRQahM3derUbJcAwAawdOnSbJcArGH+/PmxePHibJdBJXXxxRfHt99+m+0yyqUqf6eYOXNmtksolx122CHGjBmT7TJgvQmlNkFLliyJ22+/PXbbbbfYeeeds11OiV5++eXYfvvtS/zgVlBQEDvssEO89tprWaiMyuyTTz6JqnhC0e+++y6efvrpzPXhw4fHmWeembn88Y9/jO+//z6LFVLZ/O1vf1vn8iVLlkSvXr1SqqZiff7553HiiSdmuwyoMIsWLYphw4ZF06ZNY7PNNotGjRpFixYtYvjw4VU2gFjtu+++y3YJJXrjjTeKvK9GRNx3333Rrl27aN68eZx44omxfPnyLFW3diNGjKiyPyjstNNOsfvuu8cdd9wRS5YsyXY5ZdK+ffto165dHHfccTFq1Kj4/PPPs11SqVx22WVx0kknxeGHHx4LFy7MdjlQbjlJVfwGR7m8+uqrcdddd8Wjjz4arVq1ikMPPTQGDBgQv/rVr7JdWjH9+vWL7t27xxlnnFHi8uuvvz7Gjx8fjz32WMqVlV5hYWHce++9MXbs2Pj0008jJycn2rVrF4cddlgcc8wxkZOTk+0SS9SnT5+4//77Iz8/PyIiLr/88jj55JOjYcOGERGxYMGC2HfffePDDz/MYpUlq169enz55ZfRvHnziIgYOHBgXH/99bHZZptlubJ1u/XWW+Of//xnPPXUUxERUb9+/dhhhx2idu3aERHx0Ucfxdlnn73W50O2lfZX/wYNGmzgSsru0EMPLVW/sWPHbuBKyqZ27dpx2223xbHHHlts2bJly6Jnz56xYMGC+Oijj7JQ3fqZPHly/OIXv4hVq1Zlu5RijjvuuFL1u/vuuzdwJZuWwsLCuOqqq+LJJ5+MFStWxP777x8XXnhh5jWyMlu4cGHsueee8cUXX8RvfvOb2G677SIi4sMPP4wxY8ZEx44dY+LEiTFlypR4880347TTTstyxaWzfPnyuPHGG+Oqq66KuXPnZrucYnr37h3dunWLc845JyIi3n///fjFL34RQ4YMie222y6uuuqqOOmkk+Kiiy7KbqFrqFatWsydOzfzOaYqee211+Kee+6JRx55JAoLC2PAgAHx29/+Nvbdd99sl/azXnnllczlrbfeihUrVsRWW20V++23X3Tv3j26d+9eaT9Lzpw5M44//vj48MMP44477oi+fftmu6T1tnTp0igsLCzSVhk/Q642Z86cmDhxYnz11VfF6q4qr+nZJpTayM2dOzfuvffeuOuuu2Lx4sVxxBFHxK233hqTJ0+O7bffPtvlrVXbtm3jueeey3x4W9NHH30UPXv2jFmzZqVcWekkSRJ9+/aNZ555Jnbeeefo2LFjJEkS//3vf+P999+Pfv36xeOPP57tMku0ZrDToEGDeO+992KrrbaKiIh58+ZFq1atKuUXxjU/zNWvXz8mT56cqb2y2nfffePss8/OfJBYs+5//OMfcdNNN8Ubb7yRzTLXqlq1ausMWZMkiZycnEq5zwwdOrRU/e65554NXEnZPPLII3HMMcfEgw8+GP369cu0L1u2LHr16hVfffVVTJgwIVq2bJnFKsunModS1apVi7Zt28auu+66zlGZlfEHk597nkZE5OTkxA8//JBSRaV3ySWXxEUXXRQ9evSI2rVrx7hx4+LII4+sEuHf6aefHi+99FK8+OKLxb7Uzp07N3r27BnbbrttPP/883H99dfH4MGDs1RpccuXL4+LLrooXnjhhahVq1acffbZ0b9//7jnnnvivPPOi+rVq8epp56aCX4qk5YtW8ZTTz0Vv/zlLyMi4rzzzosJEybExIkTIyLi4YcfjgsvvLDS/cBWrVq1mDdvXjRr1izbpZTbsmXL4qGHHop77703Xnvttdh6663j+OOPj8GDB0eLFi2yXd7P+v777+P111/PhFT//ve/Y+XKldGxY8f44IMPsl3eWt14441xxhlnxHbbbRc1atQosuw///lPlqoqvZkzZ8app54ar7zySpGjAyrzZ8iIiHvvvTdOOumkqFWrVjRp0qTI+2xOTk588sknWayu6qjx812oqvr27RuvvvpqHHjggXHttdfGAQccENWrV49bb70126X9rHnz5kXNmjXXurxGjRrx9ddfp1hR2dx7773x6quvxksvvRTdu3cvsuzll1+O/v37x3333VfiKIdsW/OLltx6w5s+fXrsuOOOmet5eXlRrdr/O7p6t912i2HDhmWjtFIZP3585u8kSaJPnz5x5513xuabb57FqkqnsoVNpXXYYYfFokWL4sgjj4x//vOf0a1bt1i2bFkccMABMW/evCobSFV2v/vd7+L++++PmTNnxtChQ+Poo4+Oxo0bZ7usUllXUPbGG2/E9ddfX+wX3srivvvui5tvvjlOOumkiIh48cUX48ADD4w777yzyGtlZfT444/HbbfdVuIoixYtWsSVV14Zffr0iQsvvLBSBVIRERdccEHcdttt0aNHj3j99dfj8MMPj6FDh8abb74Z11xzTRx++OFRvXr1bJdZom+++abIfT5hwoTo3bt35vqvfvWrmD17djZK+1nbbLPNzwbIlflQrbp168bQoUNj6NChMX369LjnnnvipptuivPPPz8OOOCAePLJJ7Nd4jrl5eXFfvvtF/vss0907949nn322bjtttsq9cjjzz77LMaOHRuNGjWKgw8+uFgoVRUcffTRkSRJ3H333bHZZptV2iNK1nT++efHBRdcEMOHD6/070eVWdXbYym1Z599Nk477bT43e9+Fx06dMh2OWWy+eabx9SpU2PrrbcucfmUKVMq9Reu+++/P/70pz8VC6QiIvbbb78499xzY/To0ZUylKrKcnJyir2JVYU3tUWLFhWZ22LNwLWwsLBSzn2xWteuXYtcr169euyxxx6VfoRaVffb3/42Fi5cGAcffHA88cQTccEFF8ScOXNiwoQJ0apVq2yXt1G66aab4pprromxY8fG3XffHcOHD48DDzwwjj/++OjZs2elfr05+OCDi7V9/PHHce6558ZTTz0Vv/nNb+Liiy/OQmU/b9asWdGnT5/M9R49ekROTk7MmTMnWrduncXKft6XX34ZO+yww1qXd+rUKapVqxYXXnhhilWVzsMPPxz33Xdf9OvXL6ZOnRo77bRT/PDDDzF58uRKva9HRGy22WYxc+bMaNOmTaxYsSL+85//xIgRIzLLlyxZss4fP7NpxIgRmSkUqrqtt946/vSnP0Xbtm1j+PDh8c9//jPbJa3VihUr4s0334zx48dnDuNr06ZNdOnSJW688cZin3UqizvuuCP+7//+L3r06BEffPBBlR1lN3ny5Hj33Xdj2223zXYpZfLtt9/GoEGDBFLrSSi1EZs4cWLcdddd0blz59huu+3imGOOiUGDBmW7rFLp06dP5heVvLy8Isu+++67uPDCC+Oggw7KUnU/b8qUKXHllVeudXnv3r3j+uuvT7Gi0quqwU7Ej6N0hgwZErm5uRHx4xDsk08+OerWrVukX2WbH6h169YxderUtb4RT5kypdJ/8SI7zj777Fi4cGHsv//+seWWW8Yrr7xS6feVn5vHa9GiRekUUk65ublx5JFHxpFHHhmfffZZ3HvvvXHKKafEDz/8EB988EHUq1cv2yX+rDlz5sSFF14Yf//736NXr17x3nvvRadOnbJd1lr98MMPxT4L1KxZM1auXJmlikqvadOm8emnn671eTlz5sxKO3/Q559/Hp07d46IH8Oz3NzcOOOMM6rEZ4I+ffrEueeeG1dccUU8/vjjUadOnSJzG02ZMiXat2+fxQrXbtCgQZV2nyiLV199Ne6+++549NFHo1q1anHEEUfE8ccfn+2ySrTffvvFW2+9Fe3atYuuXbvGSSedFGPGjKnUP4BHRBxwwAHx73//O2688cYq/0P36tGLVS2UOv744+Phhx+Oc889N9ulVGlCqY3YHnvsEXvssUdce+218eCDD8bdd98dZ555ZhQWFsYLL7wQbdq0ifr162e7zBL9+c9/jrFjx8Y222wTp556auYF6qOPPoqbbropVq1aFeedd16Wq1y7hQsXrnNCxM022yy++eabFCsqvZ8LdirziJ01D304+uijs1RJ2fTp0ycuuOCCOPDAA0sMYUeMGBEHHnhglqqjMloz2KlZs2Y0bdo0/vCHPxRpr2wBbET87AiA/Pz8KvPhevU8TUmSVNr5Ln6qoKAg/vKXv8QNN9wQu+yyS7z00ktVYhLiNd+XIkr+0aEy7u+9evWK8847LzMv008tX7488wNcZbRq1aoiNdeoUaNKhK4RP85Dduihh0bXrl2jXr168fe//73I/3L33XdHz549s1hhyapC4Lcuc+bMiXvvvTfuvffemD59euy1115x/fXXxxFHHFHsB8LK5LXXXouWLVvGfvvtF926dYuuXbtGkyZNsl3Wz1q1atVG88PlnXfeGSeffHJ88cUX0alTp2IjGXfaaacsVbZuI0eOjIMOOiiee+652HHHHYvVfc0112SpsqrFROebmI8//jjuuuuuGDVqVCxatCh+/etfV9pjuz/77LP43e9+F+PGjcvMa5STkxO9evWKm266Kdq1a5flCteuevXqMXfu3LUOoa3Mk4VX1Ymfq7J58+bFLrvsErVq1YpTTz01ttlmm4j48fl64403xg8//BCTJk2qtGd+WVP9+vVjypQplfo5WtV5nmbP8uXLM4fvTZw4MQ466KAYOnRoHHDAAZV6+P6VV14ZV1xxRbRo0SL+8pe/lHg4X2VVlff3zz//PH75y19Gbm5uDBs2rMiJT26++eZYvnx5vP3227HFFltku9RiqlWrFr17986EgU899VTst99+lX708U8VFBREvXr1is19tXDhwqhXr16xoDDbqvLZ93r37h0vvvhiNG3aNI499tg47rjjqsyol2XLlsVrr70Wr7zySowfPz7ee++92GabbaJr166ZkKqqHhZXVbz55ptx1FFHxaeffpppW/2jT2We6PzSSy+NCy64ILbddttic2Hl5OTEyy+/nMXqqg6h1CZq1apV8dRTT8Xdd99daUOp1b755puYPn16JEkSHTp0iEaNGmW7pJ+15ge5NS1fvjyee+65SvsCS/pmzpwZv/vd7+KFF14oEsL++te/jptvvrlSz8+05qidqvjFBUrjlFNOiQceeCDatGkTxx13XPzmN7+Jpk2bZrusUqlWrVrUrl07evTosc7JqT1PK97MmTPjlFNOieeff77Y6/uNN9641vkzs60qh4Gkr1+/fnH88cfHQQcdVGknwC+tJUuWxMSJEzPzS02ePDk6dOgQU6dOzXZpG63tt98+tttuuzj77LNLnOi8bdu2Waps3Ro1ahR/+9vfYsiQIdkupUoTSsEG4IMc5bVw4cKYPn16RPw4QWhVOLOX/Z1NRbVq1WKLLbaIXXfddZ2H2VTGYGfIkCGlOjTI83TD+eabb2LatGkRUXVe32FTVFhYGG+//XaMHz8+xo8fHxMnTozvv//ej8kbUN26dWPy5MmVNqRfmxYtWsRrr71W5U4qVtkIpQAASkGwA7DxKSwsjHfeeSdz+N6//vWvWLZsWWy++ebRvXv3zKWyjtbZGPTt2zeGDBkSAwYMyHYpZTJy5Mj48ssvK+0JrKoKoRQAAACbpAYNGsSyZcuiRYsWmQCqW7dulfYMjRuj22+/PS699NI47rjjSpwwvF+/flmqbN0OOeSQePnll6NJkyaxww47FKu7Mo6croyEUgAAAGySbrvttujevXvmRDOkb10nCqnME53/3BQWRk6XjlAKAAAAoJR++OGHGDNmTPTs2TNatGiR7XKqtMp77mIAAABgo7Vy5cqoUaNGlTu7YY0aNeLkk0+O5cuXZ7uUKk8oBQAAAKSuZs2ascUWW1TaQ/TWZbfddotJkyZlu4wqr0a2CwAAAAA2Teedd1786U9/ilGjRkXjxo2zXU6pnXLKKfF///d/8fnnn0fnzp2jbt26RZbvtNNOWaqsajGnFAAAAJAVu+66a0yfPj1WrlwZbdu2LRbu/Oc//8lSZetW0gTtOTk5kSRJpZ6gvbIxUgoAAADIiv79+2e7hHKZOXNmtkvYKBgpBQAAAEDqTHQOAAAAZM2iRYvizjvvjOHDh8fChQsj4sfD9r744ossV7Zuo0aNir333jtatWoVn332WUREXHvttfHEE09kubKqQygFAAAAZMWUKVNim222iSuuuCL++te/xqJFiyIiYuzYsTF8+PDsFrcOt9xyS5x55pnRp0+fWLRoUWYOqYYNG8a1116b3eKqEKEUAAAAkBVnnnlmDBkyJKZNmxZ5eXmZ9j59+sSrr76axcrW7YYbbog77rgjzjvvvKhevXqm/Ze//GW8//77WaysahFKAQAAAFnx9ttvx0knnVSsffPNN4+5c+dmoaLSmTlzZuy6667F2nNzc2PZsmVZqKhqEkoBAAAAWZGbmxuLFy8u1v6///0vmjVrloWKSqddu3bx3nvvFWt/7rnnYrvttku/oCqqRrYLAAAAADZN/fr1i4svvjgeeuihiIjIycmJWbNmxTnnnBMDBgzIcnVrd+aZZ8awYcPi+++/jyRJ4t///nfcf//9MXLkyLjzzjuzXV6VkZMkSZLtIgAAAIBNT0FBQRx22GHxzjvvxJIlS6JVq1Yxd+7c2HPPPeOZZ56JunXrZrvEtRo9enRcdNFFMWPGjIiIaNWqVYwYMSKOP/74LFdWdQilAAAAgKz617/+FZMnT46lS5fGL37xi+jRo0e2Syq1b7/9NpYuXRrNmzfPdilVjlAKAAAAyIr77rsvBg4cGLm5uUXaV6xYEQ888EAce+yxWaps3fbbb78YO3ZsNGzYsEj74sWLo3///vHyyy9np7AqRigFAAAAZEX16tXjyy+/LDbKaMGCBdG8efNYtWpVlipbt2rVqsXcuXOL1f3VV1/F5ptv/v+1d+8xVZcPHMc/h4NcBkhuEiARlxBFIcC5mvyheQmxxWLQddLIgJnBH7HEsAVaLUCWxtQjWtqRomZbF01FnHZbdhklknIRAwVqImmhKd4A+f3RPL9OQAm1c6Lzfm1nOzzf5/vl8/2PffY8D+rp6bFTstGFg84BAAAAAIBd9Pf3y2AwDBj/8ccf5e3tbYdEf+7w4cOW7w0NDTp16pTl576+PlVVVSkgIMAe0UYlSikAAAAAAGBTsbGxMhgMMhgMmjt3rpyd/19P9PX16cSJE0pISLBjwsHFxMRYcs+ZM2fAdXd3d61bt84OyUYnSikAAAAAAGBTSUlJkqTa2lrNnz9fnp6elmsuLi4KDg5WSkqKndIN7cSJE+rv71doaKiqq6vl4+Njuebi4qKbb75ZRqPRjglHF86UAgAAAAAAdlFeXq6HHnpIbm5u9o4CO3CydwAAAAAAAOCY0tLSLIXUk08+qTNnztg50fCNHTtWx48ft3eMUYlSCgAAAAAA2F1FRYV+/fVXe8cYNjagjRylFAAAAAAAsDvKHcdDKQUAAAAAADBCqampGjt2rL1jjEqUUgAAAAAAwO7Onz+v0NBQe8f4Ux9//LGmTJlitc2wrKxM48eP17lz5zR16lR9/vnndkw4ulBKAQAAAAAAm3JycpLRaPzTj7Ozs71jDlBaWqrMzMxBV0Z5e3tr8eLFWrNmjR2SjU6GfjZtAgAAAAAAG9qxY8eQ17766iutXbtW165d0+XLl22Y6q8FBQWpqqpKERERg14/evSo4uPj1d7ebuNko9O/r3YEAAAAAAD/affdd9+AsaamJuXl5Wnnzp1auHChXnjhBTsk+3OdnZ0aM2bMkNednZ11+vRpGyYa3di+BwAAAAAA7ObkyZPKzMxUVFSUent7VVtbq/LycgUFBdk72gABAQGqq6sb8vrhw4fl7+9vw0SjG6UUAAAAAACwuXPnzumZZ55RWFiY6uvr9dFHH2nnzp2KjIy0d7Qh3XPPPcrPzx90W+GlS5e0YsUK3XvvvXZINjpxphQAAAAAALCpkpISrVq1Sn5+fiosLBx0O9+/UWdnp6ZNmyaj0ajs7GxNmjRJ0m9nSZlMJvX19ammpka+vr52Tjo6UEoBAAAAAACbcnJykru7u+bNmyej0TjkvPfff9+GqW5MW1ublixZor179+p6pWIwGDR//nyZTCaFhITYOeHoQSkFAAAAAABs6rHHHpPBYPjLeWaz2QZpRqarq0vNzc3q7+/XxIkTNW7cOHtHGnUopQAAAAAAAGBzHHQOAAAAAAAAm6OUAgAAAAAAgM1RSgEAAAAAAMDmKKUAAAAAAABgc5RSAAAA/5DW1lYZDAbV1tYOOefTTz+VwWDQ2bNnbZYLAADg34hSCgAAOIzr/37aYDBozJgxCgkJ0bJly3T58uV/5PmBgYHq6OhQZGTkP/K8kbr+jkN9Vq5cadd8AAAAkuRs7wAAAAC2lJCQILPZrJ6eHh08eFBpaWkyGAxatWrV33620WiUn5/fP5Dy7+no6LB8f+edd1RQUKCmpibLmKenpz1iAQAAWGGlFAAAcCiurq7y8/NTYGCgkpKSNG/ePO3bt89y/dq1ayoqKlJISIjc3d0VHR2td99913K9q6tLCxculI+Pj9zd3TVx4kSZzWZJg2/fq6ysVHh4uNzd3TV79my1trZa5Vm5cqViYmKsxkpLSxUcHGw1tnnzZkVERMjNzU2TJ0/Whg0bhnxHPz8/y8fb21sGg0F+fn7y8vJSeHi4qqqqrOZv375dHh4eOn/+vOUdtm3bpri4OLm5uSkyMlKfffaZ1T11dXVasGCBPD095evrq0cffVRnzpwZMhMAAMAfUUoBAACHVVdXpy+//FIuLi6WsaKiIr3xxhvauHGj6uvrlZOTo9TUVEspk5+fr4aGBu3Zs0eNjY0qKyvT+PHjB33+Dz/8oOTkZCUmJqq2tlYZGRnKy8sbds633npLBQUFeumll9TY2KjCwkLl5+ervLx8WM/x8PDQww8/bCnRrjObzbr//vvl5eVlGcvNzdXTTz+tQ4cOacaMGUpMTNTPP/8sSTp79qzmzJmj2NhYffvtt6qqqlJnZ6cefPDBYb8bAABwXGzfAwAADmXXrl3y9PRUb2+vrly5IicnJ61fv16SdOXKFRUWFmr//v2aMWOGJCk0NFQHDhzQpk2bNGvWLLW3tys2NlbTp0+XpAErmn6vrKxMt912m1avXi1JmjRpko4cOTLsrYIrVqzQ6tWrlZycLEkKCQlRQ0ODNm3apLS0tGE9KyMjQ3Fxcero6JC/v79++uknVVZWav/+/VbzsrOzlZKSYnmPqqoqbdmyRcuWLdP69esVGxurwsJCy/zXX39dgYGBOnbsmMLDw4eVCQAAOCZKKQAA4FBmz56tsrIydXd365VXXpGzs7OlfGlubtbFixd19913W91z9epVxcbGSpKWLFmilJQU1dTUKD4+XklJSYqLixv0dzU2NurOO++0Grtedt2o7u5utbS0KD09XZmZmZbx3t5eeXt7D+tZknTHHXdo6tSpKi8vV15enioqKhQUFKSZM2cOmdPZ2VnTp09XY2OjJOm7777TJ598MujZVC0tLZRSAADghlBKAQAAh+Lh4aGwsDBJv63uiY6O1pYtW5Senq4LFy5Iknbv3q2AgACr+1xdXSVJCxYsUFtbmyorK7Vv3z7NnTtXWVlZevnll0eUx8nJSf39/VZjPT09lu/XM7322msDCi6j0Tii35mRkSGTyaS8vDyZzWYtWrRIBoPhhu+/cOGCEhMTB13x5e/vP6JMAADA8XCmFAAAcFhOTk569tln9dxzz+nSpUuaMmWKXF1d1d7errCwMKtPYGCg5T4fHx+lpaWpoqJCpaWlevXVVwd9fkREhKqrq63Gvv76a6uffXx8dOrUKati6vcHpfv6+mrChAk6fvz4gEwhISEjeu/U1FS1tbVp7dq1amhoGHQL4O9z9vb26uDBg4qIiJAkTZs2TfX19QoODh6QycPDY0SZAACA46GUAgAADu2BBx6Q0WiUyWSSl5eXli5dqpycHJWXl6ulpUU1NTVat26d5VDxgoIC7dixQ83Nzaqvr9euXbssZc0fPfHEE/r++++Vm5urpqYmvf3229q6davVnLvuukunT59WSUmJWlpaZDKZtGfPHqs5zz//vIqKirR27VodO3ZMR44ckdls1po1a0b0zuPGjVNycrJyc3MVHx+vW265ZcAck8mkDz74QEePHlVWVpa6urr0+OOPS5KysrL0yy+/6JFHHtE333yjlpYW7d27V4sWLVJfX9+IMgEAAMdDKQUAAByas7OzsrOzVVJSou7ubr344ovKz89XUVGRIiIilJCQoN27d1tWJbm4uGj58uW6/fbbNXPmTBmNRm3btm3QZ99666167733tH37dkVHR2vjxo1Wh4NLv62m2rBhg0wmk6Kjo1VdXa2lS5dazcnIyNDmzZtlNpsVFRWlWbNmaevWrSNeKSVJ6enpunr1qqVo+qPi4mIVFxcrOjpaBw4c0Icffmj5L4MTJkzQF198ob6+PsXHxysqKkpPPfWUbrrpJjk58eclAAC4MYb+Px5iAAAAgP+8N998Uzk5OTp58qRcXFws462trQoJCdGhQ4cUExNjv4AAAOA/j4POAQAAHMjFixfV0dGh4uJiLV682KqQAgAAsCXWVwMAADiQkpISTZ48WX5+flq+fLm94wAAAAfG9j0AAAAAAADYHCulAAAAAAAAYHOUUgAAAAAAALA5SikAAAAAAADYHKUUAAAAAAAAbI5SCgAAAAAAADZHKQUAAAAAAACbo5QCAAAAAACAzVFKAQAAAAAAwOYopQAAAAAAAGBz/wMw74ZbBeI9GgAAAABJRU5ErkJggg==\n"
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of Unique Accession Numbers for Significant Data in AdjacentSeq_all__(Aug_Active)___(Mar_Active).tsv: 15\n",
            "Number of Unique Accession Numbers for Not Significant Data in AdjacentSeq_all__(Aug_Active)___(Mar_Active).tsv: 3212\n",
            "Total number of modifications in AdjacentSeq_all__(Aug_Active)___(Mar_Active).tsv: 17793\n",
            "Significant N-terminal (Met-loss) modifications: 2\n",
            "Non-significant N-terminal (Met-loss) modifications: 41\n",
            "Processing file: /content/output/modlist_files/AdjacentSeq_all__(Torpor)___(Aug_Active).tsv\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 1200x800 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAABKUAAAMWCAYAAAAgRDUeAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAB60ElEQVR4nOzdd5hU5fk/4GdoS1+6FBERsWIL9gqKNBWxo4mCmsSCUTHGSEwUbFgSezcKFrB3o2LHbiwootEAgmAELMhSlAXZ8/vDH/N12QV2l+XMLtz3dc11Me9555xnZs4cZj/znvdkkiRJAgAAAABSVCPXBQAAAACw7hFKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAaymYcOGRSaTKVPfTCYTw4YNW7MFlWJZjd9++22lrveuu+6KzTbbLGrXrh1NmjSp1HVXV6NGjYpMJhPTpk3LtnXr1i26detWrN/s2bPj0EMPjebNm0cmk4mrrroqXn755chkMvHyyy+nWnNExIYbbhiDBg1KfbtpK+01HjRoUGy44YY5q4mSlv/MTJs2LTKZTIwaNarSt1VUVBRdunSJiy66qNLXXVWcffbZsdNOO+W0hgULFkSrVq1i9OjROa2jKlvT3xGeeeaZaNiwYXzzzTdrbBsA5SWUAtYay8KAZbdatWpFu3btYtCgQfG///0v1+WtdT799NMYNGhQdOrUKW699da45ZZbcl1StTJkyJAYO3ZsDB06NO66667o3bv3Gt/mG2+8EcOGDYu5c+eu8W2taU899VRkMplo27ZtFBUV5bqcVP3www8xbNiwFYaXr732WvTp0yfatWsXdevWjQ022CAOOOCAGDNmTLqFVhP33HNPzJgxI0455ZSIiGL/j6zslovwuKJOP/30+PDDD+Pxxx/PWQ1XX311NGrUKAYMGFDq8rPOOisymUwcccQRKVdWdjvuuGNkMpm48cYbK7yOp556Kic/TkVE9O7dOzbeeOMYMWJEmfrnslZg3VEr1wUAVLbzzz8/OnbsGIsWLYq33norRo0aFa+99lpMnDgx6tatW+nb++tf/xpnn312pa+3qnv55ZejqKgorr766th4441zXU6V9uyzz5Zoe/HFF+PAAw+MM888M9u2ySabxI8//hh16tRZI3W88cYbMXz48Bg0aFCJkW2fffZZ1KhRfX6rGj16dGy44YYxbdq0ePHFF6NHjx4VXtett95arYKtH374IYYPHx4RUWIE3gMPPBBHHHFEbLvttnHaaadF06ZNY+rUqfHKK6/ErbfeGkcddVQOKq7aLr/88hgwYEDk5+dHxM8jQH/pzjvvjOeee65E++abb55ajaurdevWceCBB8bf//736NevX+rbX7JkSVx99dUxZMiQqFmzZonlSZLEPffcExtuuGE88cQTMX/+/GjUqFHqda7MpEmT4p133okNN9wwRo8eHSeddFKF1vPUU0/F9ddfX2rY8+OPP0atWmv2z7MTTjghzjzzzBg+fPgqX+OV1QpQWYRSwFqnT58+sf3220dExG9/+9to0aJFXHrppfH444/H4YcfXunbq1Wr1hr/ElkVff311xERlXra3g8//BD169evtPVVFaWFTF9//XWJ165GjRprJDgti7y8vJxstyIWLlwYjz32WIwYMSJGjhwZo0ePXq1Qqnbt2pVYXW4NGzYstthii3jrrbdK7HfLPrP8n/Hjx8eHH34Y//jHP7Jtv/nNb4r1eeutt+K5554r0V4RRUVFsXjx4tQ+5wsXLowGDRpERMThhx8ehx12WHz++eex0UYbpbL9ZZ588sn45ptvVvh/8MsvvxxffvllvPjii9GrV694+OGHY+DAganWuCp33313tGrVKv7xj3/EoYceGtOmTav0037T2C8OOeSQ+MMf/hAPPPBAHHfccWt8ewCrUn1+EgWooD322CMiIqZMmVKs/dNPP41DDz00mjVrFnXr1o3tt9++xKkNS5YsieHDh0fnzp2jbt260bx589h9993jueeey/YpbU6pwsLCGDJkSLRs2TIaNWoU/fr1iy+//LJEbSuay2ZF81Tdfffd0bVr16hXr140a9YsBgwYEDNmzCjza/Htt9/G4YcfHo0bN47mzZvHaaedFosWLSr3djbccMM477zzIiKiZcuWJebBuOGGG2LLLbeMvLy8aNu2bQwePLjEKWPdunWLLl26xHvvvRd77rln1K9fP/7yl79ExM+v33nnnRcbb7xx5OXlRfv27eOss86KwsLCVT7HZeudMGFC7LXXXlG/fv3YeOON48EHH4yIiHHjxsVOO+0U9erVi0033TSef/75EusYP3589OnTJxo3bhwNGzaMffbZJ956660S/T7++OPYe++9o169erH++uvHhRdeWOqIm1/Oj7PsNNMkSeL666/PngoUUfp8RxERb7/9dvTt2zeaNm0aDRo0iK233jquvvrq7PIJEybEoEGDYqONNoq6detG69at47jjjovvvvsu22fYsGHxpz/9KSIiOnbsmN3usrmvSptT6vPPP4/DDjssmjVrFvXr14+dd945/vWvfxXrs6zm+++/Py666KJYf/31o27durHPPvvE5MmTS3mHVt8jjzwSP/74Yxx22GExYMCAePjhh0vdj7/88svo379/NGjQIFq1ahVDhgwpdR8q7XP497//PXbddddo3rx51KtXL7p27Zrdh5Z39913x4477hj169ePpk2bxp577llidNzTTz8de+yxRzRo0CAaNWoU++23X3z88ccl6mjYsGH873//i/79+0fDhg2jZcuWceaZZ8bSpUsj4ue5lVq2bBkREcOHD8++j8s+f1OmTIkddtih1CC0VatWxe4XFRXFVVddFVtuuWXUrVs31ltvvTjhhBPi+++/L9YvSZK48MILY/3114/69etH9+7d4+OPP67QPGQjR46MvffeO1q1ahV5eXmxxRZbrNapUKvr0UcfjTp16sSee+5ZrsctXLgw/vjHP0b79u0jLy8vNt100/j73/8eSZIU65fJZOKUU06J0aNHZ4+JzzzzTHaOrL///e9x5ZVXRocOHaJevXqx1157xcSJE0ts78UXX8zuP02aNIkDDzww/vOf/xTrs+z/jU8++SSOOuqoaNq0aey+++7Z5cuC28cee6xcz7UyPProo7HhhhtGp06dSl0+evTo2GKLLaJ79+7Ro0ePUuedKm2+vogVHzevv/762GijjaJevXqx4447xquvvlrq/H5lNWbMmDj00ENj//33j/z8/BWeDruy4/WgQYPi+uuvj4jip4ku88vP8oMPPhiZTCbGjRtXYhs333xzZDKZYvtKWb7TRPx8HNh6661XuR+sqtZ77703unbtGo0aNYrGjRvHVlttlX2e7777bmQymbjjjjtKrHfs2LGRyWTiySefXOn2gXXHuvfTPrDOWfYFtmnTptm2jz/+OHbbbbdo165dnH322dGgQYO4//77o3///vHQQw/FQQcdFBE/f8kfMWJE/Pa3v40dd9wx5s2bF++++268//77se+++65wm7/97W/j7rvvjqOOOip23XXXePHFF2O//fZbredx0UUXxd/+9rc4/PDD47e//W188803ce2118aee+4Z48ePL9OIpcMPPzw23HDDGDFiRLz11ltxzTXXxPfffx933nlnubZz1VVXxZ133hmPPPJI3HjjjdGwYcPYeuuts6/Z8OHDo0ePHnHSSSfFZ599FjfeeGO888478frrrxcblfLdd99Fnz59YsCAAfGb3/wm1ltvvSgqKop+/frFa6+9Fr///e9j8803j48++iiuvPLK+O9//xuPPvroKp/n999/H/vvv38MGDAgDjvssLjxxhtjwIABMXr06Dj99NPjxBNPjKOOOiouv/zyOPTQQ2PGjBnZ0xg+/vjj2GOPPaJx48Zx1llnRe3atePmm2+Obt26ZQOtiIhZs2ZF9+7d46effsruQ7fcckvUq1dvpbXtueeecdddd8XRRx8d++67bxxzzDEr7f/cc8/F/vvvH23atInTTjstWrduHf/5z3/iySefjNNOOy3b5/PPP49jjz02WrduHR9//HHccsst8fHHH8dbb70VmUwmDj744Pjvf/8b99xzT1x55ZXRokWLiIhswLG82bNnx6677ho//PBDnHrqqdG8efO44447ol+/fvHggw9mPyPLXHLJJVGjRo0488wzo6CgIC677LL49a9/HW+//fYq36/yGj16dHTv3j1at24dAwYMiLPPPjueeOKJOOyww7J9fvzxx9hnn31i+vTpceqpp0bbtm3jrrvuihdffLFM27j66qujX79+8etf/zoWL14c9957bxx22GHx5JNPFvssDx8+PIYNGxa77rprnH/++VGnTp14++2348UXX4yePXtGxM+ngw0cODB69eoVl156afzwww9x4403xu677x7jx48vFogtXbo0evXqFTvttFP8/e9/j+effz7+8Y9/RKdOneKkk06Kli1bxo033hgnnXRSHHTQQXHwwQdHRGQ/fx06dIgXXnghvvzyy1h//fVX+hxPOOGEGDVqVBx77LFx6qmnxtSpU+O6666L8ePHF/usnnvuuXHhhRdG3759o2/fvvH+++9Hz549Y/HixWV6LX/pxhtvjC233DL69esXtWrViieeeCJOPvnkKCoqisGDB5d7favrjTfeiC5dupRrtFySJNGvX7946aWX4vjjj49tt902xo4dG3/605/if//7X1x55ZXF+r/44otx//33xymnnBItWrQo9n7feeedMX/+/Bg8eHAsWrQorr766th7773jo48+ivXWWy8iIp5//vno06dPbLTRRjFs2LD48ccf49prr43ddtst3n///RKB6mGHHRadO3eOiy++uFhIlp+fH506dYrXX389hgwZUv4XazW88cYb8atf/arUZYWFhfHQQw/FH//4x4iIOPLII+PYY4+NWbNmRevWrSu0vRtvvDFOOeWU2GOPPWLIkCExbdq06N+/fzRt2nSVn4vSvP322zF58uQYOXJk1KlTJw4++OAYPXp09oeUZVZ1vD7hhBPiq6++KvV00OXtt99+0bBhw7j//vtjr732Krbsvvvuiy233DK6dOkSEWX/TrNM165dV/l/6cpqfe655+LII4+MffbZJy699NKIiPjPf/4Tr7/+epx22mmx/fbbx0YbbRT3339/iRFv9913XzRt2jR69eq10u0D65AEYC0xcuTIJCKS559/Pvnmm2+SGTNmJA8++GDSsmXLJC8vL5kxY0a27z777JNstdVWyaJFi7JtRUVFya677pp07tw527bNNtsk++2330q3e9555yW/PJx+8MEHSUQkJ598crF+Rx11VBIRyXnnnZdtGzhwYNKhQ4dVrnPatGlJzZo1k4suuqhYv48++iipVatWifYVra9fv37F2k8++eQkIpIPP/yw3NtZts5vvvkm2/b1118nderUSXr27JksXbo0237dddclEZHcfvvt2ba99toriYjkpptuKratu+66K6lRo0by6quvFmu/6aabkohIXn/99ZU+12XrHTNmTLbt008/TSIiqVGjRvLWW29l28eOHZtERDJy5MhsW//+/ZM6deokU6ZMybZ99dVXSaNGjZI999wz23b66acnEZG8/fbbxZ5/fn5+EhHJ1KlTi9W01157FaszIpLBgwcXa3vppZeSiEheeumlJEmS5Keffko6duyYdOjQIfn++++L9S0qKsr++4cffijxOtxzzz1JRCSvvPJKtu3yyy8vUdsyHTp0SAYOHFji+f3yfZg/f37SsWPHZMMNN8y+v8tq3nzzzZPCwsJs36uvvjqJiOSjjz4qsa3VMXv27KRWrVrJrbfemm3bddddkwMPPLBYv6uuuiqJiOT+++/Pti1cuDDZeOONi73GSVL653D513Tx4sVJly5dkr333jvbNmnSpKRGjRrJQQcdVGx/T5L/e3/mz5+fNGnSJPnd735XbPmsWbOS/Pz8Yu0DBw5MIiI5//zzi/Xdbrvtkq5du2bvf/PNNyWOJcvcdtttSUQkderUSbp375787W9/S1599dUS9b366qtJRCSjR48u1v7MM88Ua1/2md5vv/2K7XN/+ctfkogots+URWn7aq9evZKNNtqoWNvyn5mpU6eW+KxWhvXXXz855JBDVtpn8ODBxY7Hjz76aBIRyYUXXlis36GHHppkMplk8uTJ2bZlx52PP/64WN9lz6devXrJl19+mW1/++23k4hIhgwZkm3bdtttk1atWiXfffddtu3DDz9MatSokRxzzDHZtmXH5COPPHKFz6Vnz57J5ptvvtLnW9mWLFmSZDKZ5I9//GOpyx988MEkIpJJkyYlSZIk8+bNS+rWrZtceeWVxfot+z9++ePX8sfNwsLCpHnz5skOO+yQLFmyJNtv1KhRSUSUOBaXxSmnnJK0b98++xl49tlnk4hIxo8fn+1T1uP18vvTLy3/uT7yyCOTVq1aJT/99FO2bebMmUmNGjWKHSfK+p1mmYsvvjiJiGT27Nkrfd4rqvW0005LGjduXKyu5Q0dOjSpXbt2MmfOnGxbYWFh0qRJk+S4445b6XaBdYvT94C1To8ePaJly5bRvn37OPTQQ6NBgwbx+OOPZ38dnTNnTrz44otx+OGHx/z58+Pbb7+Nb7/9Nr777rvo1atXTJo0KXu1viZNmsTHH38ckyZNKvP2n3rqqYiIOPXUU4u1n3766RV+Tg8//HAUFRXF4Ycfnq3322+/jdatW0fnzp3jpZdeKtN6lh+J8Ic//KFYzau7neeffz4WL14cp59+erFJs3/3u99F48aNS5z6lZeXF8cee2yxtgceeCA233zz2GyzzYrVsPfee0dElOm5NmzYsNgVnjbddNNo0qRJbL755sUui77s359//nlE/DxK5dlnn43+/fsXm3OlTZs2cdRRR8Vrr70W8+bNy75mO++8c+y4447Zfi1btoxf//rXq6yvrMaPHx9Tp06N008/vcRIuF+eRvHL0VmLFi2Kb7/9NnbeeeeIiHj//fcrtO2nnnoqdtxxx2Kn/zRs2DB+//vfx7Rp0+KTTz4p1v/YY48tdsrYstNml722leXee++NGjVqxCGHHJJtO/LII+Ppp58udtrZU089FW3atIlDDz0021a/fv34/e9/X6bt/PI1/f7776OgoCD22GOPYq/no48+GkVFRXHuueeWmCR+2fvz3HPPxdy5c+PII48stj/XrFkzdtppp1L35xNPPLHY/T322KPMr+Nxxx0XzzzzTHTr1i1ee+21uOCCC2KPPfaIzp07xxtvvJHt98ADD0R+fn7su+++xerq2rVrNGzYMFvXss/0H/7wh2L7XEWPZ798XQsKCuLbb7+NvfbaKz7//PMoKCio0DpXx3fffVdsFG1ZPPXUU1GzZs0Sx/g//vGPkSRJPP3008Xa99prr9hiiy1KXVf//v2jXbt22fs77rhj7LTTTtlj8syZM+ODDz6IQYMGRbNmzbL9tt5669h3332z/X5p+f3nl5o2bRrffvvtqp9kJZozZ04kSbLC13n06NGx/fbbZy+Ysez01tJO4SuLd999N7777rv43e9+V2y+x1//+tflfq8jIn766ae477774ogjjsh+BpadgvrLGst6vC6PI444Ir7++utipyY++OCDUVRUlL1KYXm+0yyz7HWo6L7QpEmTWLhwYbGpDEqrfcmSJfHwww9n25599tmYO3dulb7CIpA+p+8Ba53rr78+NtlkkygoKIjbb789XnnllWKTOE+ePDmSJIm//e1v8be//a3UdXz99dfRrl27OP/88+PAAw+MTTbZJLp06RK9e/eOo48+OnuqTGm++OKLqFGjRom5MzbddNMKP6dJkyZFkiTRuXPnUpeX9dST5R/fqVOnqFGjRvYUx9XdzhdffBERJZ9rnTp1YqONNsouX6Zdu3Yl5r6ZNGlS/Oc//1nhaWVlmax5/fXXL/FHQH5+frRv375EW0Rkw4xvvvkmfvjhh1Lfq8033zyKiopixowZseWWW8YXX3xRLOBaZnXe5+Utmwdt2SkaKzJnzpwYPnx43HvvvSVen4r+ob+i57fsimNffPFFsbo22GCDYv2W/dGz/PxEv7R48eKYM2dOsbaWLVuWenWuZZbN3/Tdd99l58zabrvtYvHixfHAAw9kQ6cvvvgiNt544xL7QVnfnyeffDIuvPDC+OCDD4rNQ/XL9U2ZMiVq1KixwsAhIrKB9rJQdXmNGzcudr9u3bol9v2mTZuu9HVcXq9evaJXr17xww8/xHvvvRf33Xdf3HTTTbH//vvHp59+Gq1atYpJkyZFQUFBiXmmllm2Hy37zC5/TGjZsmWF/sB//fXX47zzzos333wzfvjhh2LLCgoKsp/JNCXLzQO1Kl988UW0bdu2xJXLfvnZ+KWOHTuucF2lHWs32WSTuP/++4uta0XHpLFjxxabzHxV20uSpMIByeoq7XWeO3duPPXUU3HKKacUm4Nut912i4ceeij++9//xiabbFKu7Sx7zZa/KmytWrUqNDH5s88+G998803suOOOxWrs3r173HPPPXHppZdGjRo1yny8Lo/evXtHfn5+3HfffbHPPvtExM+nv2277bbZ16U832mWWfZeVHRfOPnkk+P++++PPn36RLt27aJnz55x+OGHR+/evbN9ttlmm9hss83ivvvui+OPPz5be4sWLVZ4PATWTUIpYK2z4447Zq++179//9h9993jqKOOis8++ywaNmyYnYj6zDPPXOGcBsu+zO65554xZcqUeOyxx+LZZ5+Nf/7zn3HllVfGTTfdFL/97W9Xu9YVfSFcNqnxMkVFRZHJZOLpp58u9Q/2hg0bVsr219R2VqS0+ZeKiopiq622iiuuuKLUxywfLJVmRaHGitrL+0dpVXP44YfHG2+8EX/6059i2223ze7nvXv3LnXi9TWhIq/tG2+8Ed27dy/WNnXq1BX+4bjskuwRpf8xP3r06DKPhFqZV199Nfr16xd77rln3HDDDdGmTZuoXbt2jBw5coWTG6/Istf/rrvuKnV+nOWv3LmyQK686tevH3vssUfsscce0aJFixg+fHg8/fTTMXDgwCgqKiox0uOXVhQKr44pU6bEPvvsE5tttllcccUV0b59+6hTp0489dRTceWVV6a2r/5S8+bNyxX4VcSq5plLc3vff/99dj65tDRr1iwymUypr/MDDzwQhYWF8Y9//KPYFRCXGT16dAwfPjwiyv7/ZWVb9hlZ0ZUDx40bV+I4Vlny8vKif//+8cgjj8QNN9wQs2fPjtdffz0uvvjibJ/yfKdZZtl7UdF9oVWrVvHBBx/E2LFj4+mnn46nn346Ro4cGcccc0yxyc2POOKIuOiii+Lbb7+NRo0axeOPPx5HHnnkOnnFYmDFHBGAtVrNmjVjxIgR0b1797juuuvi7LPPzp6WVbt27TJdRr5Zs2Zx7LHHxrHHHhsLFiyIPffcM4YNG7bCUKpDhw5RVFQUU6ZMKfbr9meffVaib9OmTUtclS6i5C/tnTp1iiRJomPHjuX+1fiXJk2aVOxX9MmTJ0dRUVE2BFjd7XTo0CEifn6uvzz9bfHixTF16tQyvd6dOnWKDz/8MPbZZ5/Uf9Fv2bJl1K9fv9T36tNPP40aNWpkQ7EOHTqUelpnaY+tqGWj7SZOnLjC1+7777+PF154IYYPHx7nnntutr202srzenbo0GGFr8Oy5atrm222KXH6x8omNh49enTUrl077rrrrhLhzWuvvRbXXHNNTJ8+PTbYYIPo0KFDTJw4scTIkLK8Pw899FDUrVs3xo4dW2yU5ciRI4v169SpUxQVFcUnn3wS2267banrWvYetmrVqkz7f1lU5HOxLKifOXNmtq7nn38+dtttt5WGGMve50mTJhX7TH/zzTflDnOeeOKJKCwsjMcff7zYyLqynn68Jmy22WYxderUcj2mQ4cO8fzzz8f8+fOLjZaqyGejtM/pf//73+wx+ZfH1OV9+umn0aJFi2KjpFZl6tSpsc0225S5f2WoVatWdOrUqdTXefTo0dGlS5fs1Vx/6eabb44xY8ZkQ6llI/OW/z9z+f8vl71mkydPLhYW/fTTTzFt2rSVjnRe3sKFC+Oxxx6LI444otipwMuceuqp2QsvlOV4HVH+z+8RRxwRd9xxR7zwwgvxn//8J5IkKXb6W3m/00T8vB+0aNFileHzymqtU6dOHHDAAXHAAQdEUVFRnHzyyXHzzTfH3/72t2wIdsQRR8Tw4cPjoYceivXWWy/mzZtX7NR6gIgIc0oBa71u3brFjjvuGFdddVUsWrQoWrVqFd26dYubb745+wfaL33zzTfZfy87PWiZhg0bxsYbb1zqZeWX6dOnT0REXHPNNcXar7rqqhJ9O3XqFAUFBTFhwoRs28yZM+ORRx4p1u/ggw+OmjVrxvDhw0uMPEmSpESdK7Ls8s7LXHvttcVqXt3t9OjRI+rUqRPXXHNNscffdtttUVBQUKYrEB5++OHxv//9L2699dYSy3788cdYuHDhKtdRUTVr1oyePXvGY489Vuyy47Nnz44xY8bE7rvvnj3dqm/fvvHWW2/Fv//972y/b775psLzoJTmV7/6VXTs2DGuuuqqEn+ILXt9l4Uzy79fpe1vy/54LS0IXV7fvn3j3//+d7z55pvZtoULF8Ytt9wSG2644UpPWSurpk2bRo8ePYrd6tatu8L+o0ePjj322CP7B+Ivb3/6058iIuKee+7J1v/VV1/Fgw8+mH38Dz/8ELfccssq66pZs2ZkMpliIzCmTZtW4mpV/fv3jxo1asT5559fYpTPsvejV69e0bhx47j44otjyZIlJbb1y+NNWdWvXz8iSn8fX3jhhVIfs2zuoWVB+eGHHx5Lly6NCy64oETfn376KbvuHj16RO3atePaa68tto+Vtn+tSmn7akFBQYmwL0277LJLTJw4caXH9OX17ds3li5dGtddd12x9iuvvDIymUz2eFoWjz76aLH5fv7973/H22+/nV1HmzZtYtttt4077rij2Ps9ceLEePbZZ6Nv375l3lZBQUFMmTIldt111zI/prLssssu8e677xZrmzFjRrzyyitx+OGHl/g8H3rooXHsscfG5MmTs1fwXBb6vPLKK9l1LF26tMRnevvtt4/mzZvHrbfeGj/99FO2ffTo0eUOUh955JFYuHBhDB48uNQa999//3jooYeisLCwTMfriPIdhyN+/gw2a9Ys7rvvvrjvvvtixx13LPbjUnm+0yzz3nvvxS677FKsbebMmfHpp58WO06tqNblvwvUqFEjG/b98rO0+eabx1ZbbZWtvU2bNrHnnnuW6XkD6w4jpYB1wp/+9Kc47LDDYtSoUXHiiSfG9ddfH7vvvntstdVW8bvf/S422mijmD17drz55pvx5ZdfxocffhgREVtssUV069YtunbtGs2aNYt33303HnzwwTjllFNWuK1tt902jjzyyLjhhhuioKAgdt1113jhhReKzUWxzIABA+LPf/5zHHTQQXHqqadmLxW/ySabFJtQuVOnTnHhhRfG0KFDs5e2btSoUUydOjUeeeSR+P3vfx9nnnnmKl+HqVOnRr9+/aJ3797x5ptvxt133x1HHXVU9pfz1d1Oy5YtY+jQoTF8+PDo3bt39OvXLz777LO44YYbYocddojf/OY3q6zx6KOPjvvvvz9OPPHEeOmll2K33XaLpUuXxqeffhr3339/jB07NjvqY0248MIL47nnnovdd989Tj755KhVq1bcfPPNUVhYGJdddlm231lnnRV33XVX9O7dO0477bRo0KBB3HLLLdGhQ4diIePqqFGjRtx4441xwAEHxLbbbhvHHntstGnTJj799NP4+OOPY+zYsdG4cePYc88947LLLoslS5ZEu3bt4tlnny11VELXrl0jIuKcc86JAQMGRO3ateOAAw4odaTF2WefHffcc0/06dMnTj311GjWrFnccccdMXXq1HjooYdKTOy9pi27JPuKPnvt2rWLX/3qVzF69Oj485//HL/73e/iuuuui2OOOSbee++9aNOmTdx1113ZQGdl9ttvv7jiiiuid+/ecdRRR8XXX38d119/fWy88cbF3tuNN944zjnnnOxk4gcffHDk5eXFO++8E23bto0RI0ZE48aN48Ybb4yjjz46fvWrX8WAAQOiZcuWMX369PjXv/4Vu+22W4lwY1Xq1asXW2yxRdx3332xySabRLNmzaJLly7RpUuXOPDAA6Njx45xwAEHRKdOnWLhwoXx/PPPxxNPPBE77LBDHHDAARHx8+TbJ5xwQowYMSI++OCD6NmzZ9SuXTsmTZoUDzzwQFx99dVx6KGHRsuWLePMM8+MESNGxP777x99+/aN8ePHx9NPP13uU3969uyZHV1xwgknxIIFC+LWW2+NVq1alfrHdBoOPPDAuOCCC2LcuHHRs2fPMj3mgAMOiO7du8c555wT06ZNi2222SaeffbZeOyxx+L0008vMZ/gymy88cax++67x0knnRSFhYVx1VVXRfPmzeOss87K9rn88sujT58+scsuu8Txxx8fP/74Y1x77bWRn58fw4YNK/O2nn/++UiSJA488MAyP6ayHHjggXHXXXcVmyNqzJgxkSRJ9OvXr9TH9O3bN2rVqhWjR4+OnXbaKbbccsvYeeedY+jQoTFnzpxo1qxZ3HvvvcWCp4ifR/AMGzYs/vCHP8Tee+8dhx9+eEybNi1GjRoVnTp1KtdIpdGjR0fz5s1XGOT169cvbr311vjXv/4VBx988CqP1xH/dxw+9dRTo1evXlGzZs2Vjh6qXbt2HHzwwXHvvffGwoUL4+9//3uJPmX9ThPx8/xSEyZMKHHhk6FDh2aP8ctG6q2o1t/+9rcxZ86c2HvvvWP99dePL774Iq699trYdttts3OrLXPEEUfEueeeG3Xr1o3jjz8+9f87gGoglWv8AaRg2eWi33nnnRLLli5dmnTq1Cnp1KlT9hLGU6ZMSY455pikdevWSe3atZN27dol+++/f/Lggw9mH3fhhRcmO+64Y9KkSZOkXr16yWabbZZcdNFFyeLFi7N9ll2G+5d+/PHH5NRTT02aN2+eNGjQIDnggAOSGTNmlHoZ92effTbp0qVLUqdOnWTTTTdN7r777lLXmSRJ8tBDDyW777570qBBg6RBgwbJZpttlgwePDj57LPPVvraLFvfJ598khx66KFJo0aNkqZNmyannHJK8uOPP1ZoO8vW+c0335R4/HXXXZdsttlmSe3atZP11lsvOemkk0pcInuvvfZKttxyy1LrXbx4cXLppZcmW265ZZKXl5c0bdo06dq1azJ8+PCkoKBgpc91Revt0KFDst9++5Voj4hk8ODBxdref//9pFevXknDhg2T+vXrJ927d0/eeOONEo+dMGFCstdeeyV169ZN2rVrl1xwwQXJbbfdVuKy5ctf3n5F213+0ubLvPbaa8m+++6bNGrUKGnQoEGy9dZbJ9dee212+ZdffpkcdNBBSZMmTZL8/PzksMMOS7766qtS97cLLrggadeuXVKjRo1idXbo0CEZOHBgsb5TpkxJDj300KRJkyZJ3bp1kx133DF58sknS635gQceKNa+7JL3I0eOLPG6VcQf/vCHJCKSKVOmrLDPsGHDkohIPvzwwyRJkuSLL75I+vXrl9SvXz9p0aJFctpppyXPPPNMidd44MCBSYcOHYqt67bbbks6d+6c5OXlJZtttlkycuTIFX4ub7/99mS77bbL7qt77bVX8txzzxXr89JLLyW9evVK8vPzk7p16yadOnVKBg0alLz77rvF6mjQoEGJ9Ze23TfeeCPp2rVrUqdOnWLv8z333JMMGDAg6dSpU1KvXr2kbt26yRZbbJGcc845ybx580qs+5Zbbkm6du2a1KtXL2nUqFGy1VZbJWeddVby1VdfZfssXbo0GT58eNKmTZukXr16Sbdu3ZKJEyeWus+syuOPP55svfXWSd26dZMNN9wwufTSS5Pbb799lZ+Zyt6ffmnrrbdOjj/++BUuHzx4cInXf/78+cmQIUOStm3bJrVr1046d+6cXH755UlRUVGxfqV9zpPk/57P5ZdfnvzjH/9I2rdvn+Tl5SV77LFHdv/9peeffz7Zbbfdknr16iWNGzdODjjggOSTTz4p1mdlx+QkSZIjjjgi2X333Vf4PNekwsLCpEWLFskFF1yQbdtqq62SDTbYYKWP69atW9KqVatkyZIlSZL8fEzq0aNHkpeXl6y33nrJX/7yl+S5554r9bh5zTXXJB06dEjy8vKSHXfcMXn99deTrl27Jr179y5TzbNnz05q1aqVHH300Svs88MPPyT169dPDjrooGzbqo7XP/30U/KHP/whadmyZZLJZIrtW6Uds5MkyT7HTCaTzJgxo9RayvKdJkmS5MYbb0zq169f4ngwcODAEp/DFdX64IMPJj179kxatWqV1KlTJ9lggw2SE044IZk5c2aJuiZNmpRERBIRyWuvvbbC1xJYd2WSpJrP7goAsBqOPvroePPNN0sdzciKbbjhhtGtW7cYNWpUrktZLXfddVcMHjw4pk+fHk2aNEllm9OmTYuOHTvG5ZdfXqZRrqtr1qxZ0bFjx7j33ntzMlIqIuKCCy6IkSNHxqRJkyp1Qv+yKioqipYtW8bBBx9c6unh64rtttsuunXrFldeeWWuSwGICHNKAQDruJkzZ6Z+RTKqjl//+texwQYblJhzb21y1VVXxVZbbZWzQCoiYsiQIbFgwYK499571/i2Fi1aVGKevTvvvDPmzJkT3bp1W+Pbr6qeeeaZmDRpUgwdOjTXpQBkmVMKAFgnTZgwIR599NF45ZVXshOlUzGzZs1a6fJ69epFfn5+pWxr8eLFMWfOnJX2WbJkSdSuXXuFy+vUqRPNmjWLiJ/nbps4cWKl1FZVXXLJJSXa5syZE4sXL17hY1b1Gkb8/F7UqVNnhct/+b43bNgwvv766zJWvHreeuutGDJkSBx22GHRvHnzeP/99+O2226LLl26xGGHHRYRP08A/suLGSzvl/vI2qJ3796xYMGCXJcBUIxQCgBYJz388MNx7bXXxoABA4wcWE1t2rRZ6fKBAwdW2ml+b7zxRnTv3n2lfTp06BBffPHFCpfvtdde8fLLL1dKPdXVwQcfHOPGjVvh8lW9hmVRme97eWy44YbRvn37uOaaa7KToh9zzDFxySWXZEO0HXbYwT4CUAXkdE6pG2+8MW688cbsZbe33HLLOPfcc7OXwe3WrVuJ/yxPOOGEuOmmm9IuFQCAFXj++edXurxt27axxRZbVMq2vv/++3jvvfdW2qdevXrx448/rnB506ZNs1cWW1e999578f33369w+apew4if34umTZuucHllvu+V7fXXX7ePAFQBOQ2lnnjiiahZs2Z07tw5kiSJO+64Iy6//PIYP358bLnlltGtW7fYZJNN4vzzz88+pn79+tG4ceNclQwAAABAJcjp6XsHHHBAsfsXXXRR3HjjjfHWW2/FlltuGRE/h1CtW7fORXkAAAAArCFVZk6ppUuXxgMPPBALFy6MXXbZJds+evTouPvuu6N169ZxwAEHxN/+9reoX7/+CtdTWFgYhYWF2ftFRUUxZ86caN68eWQymTX6HAAAAADWdUmSxPz586Nt27ZRo0aNFfbLeSj10UcfxS677BKLFi2Khg0bxiOPPJI99/yoo46KDh06RNu2bWPChAnx5z//OT777LN4+OGHV7i+ESNGxPDhw9MqHwAAAIBSzJgxI9Zff/0VLs/pnFIRP19Kdvr06VFQUBAPPvhg/POf/4xx48aVOiniiy++GPvss09Mnjw5OnXqVOr6lh8pVVBQEBtssEHMmDHDXFQAAAAAa9i8efOiffv2MXfu3MjPz19hv5yHUsvr0aNHdOrUKW6++eYSyxYuXBgNGzaMZ555Jnr16lWm9c2bNy/y8/OjoKBAKAUAAACwhpU1i1nxiX05UlRUVGyk0y998MEHERHRpk2bFCsCAAAAoLLldE6poUOHRp8+fWKDDTaI+fPnx5gxY+Lll1+OsWPHxpQpU2LMmDHRt2/faN68eUyYMCGGDBkSe+65Z2y99da5LBsAAACA1ZTTUOrrr7+OY445JmbOnBn5+fmx9dZbx9ixY2PfffeNGTNmxPPPPx9XXXVVLFy4MNq3bx+HHHJI/PWvf81lyQAAAABUgio3p1RlM6cUAAAAQHqq7ZxSAAAAAKz9hFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqchpK3XjjjbH11ltH48aNo3HjxrHLLrvE008/nV2+aNGiGDx4cDRv3jwaNmwYhxxySMyePTuHFQMAAABQGXIaSq2//vpxySWXxHvvvRfvvvtu7L333nHggQfGxx9/HBERQ4YMiSeeeCIeeOCBGDduXHz11Vdx8MEH57JkAAAAACpBJkmSJNdF/FKzZs3i8ssvj0MPPTRatmwZY8aMiUMPPTQiIj799NPYfPPN480334ydd965TOubN29e5OfnR0FBQTRu3HhNlg4AAACwzitrFlNl5pRaunRp3HvvvbFw4cLYZZdd4r333oslS5ZEjx49sn0222yz2GCDDeLNN99c4XoKCwtj3rx5xW4AAAAAVC05D6U++uijaNiwYeTl5cWJJ54YjzzySGyxxRYxa9asqFOnTjRp0qRY//XWWy9mzZq1wvWNGDEi8vPzs7f27duv4WcAAAAAQHnlPJTadNNN44MPPoi33347TjrppBg4cGB88sknFV7f0KFDo6CgIHubMWNGJVYLAAAAQGWolesC6tSpExtvvHFERHTt2jXeeeeduPrqq+OII46IxYsXx9y5c4uNlpo9e3a0bt16hevLy8uLvLy8NV02AAAAAKsh5yOllldUVBSFhYXRtWvXqF27drzwwgvZZZ999llMnz49dtlllxxWCAAAAMDqyulIqaFDh0afPn1igw02iPnz58eYMWPi5ZdfjrFjx0Z+fn4cf/zxccYZZ0SzZs2icePG8Yc//CF22WWXMl95DwAAAICqKaeh1Ndffx3HHHNMzJw5M/Lz82PrrbeOsWPHxr777hsREVdeeWXUqFEjDjnkkCgsLIxevXrFDTfckMuSAQAAAKgEmSRJklwXsSbNmzcv8vPzo6CgIBo3bpzrcgAAAADWamXNYqrcnFIAAAAArP2EUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkTigFAAAAQOqEUgAAAACkLqeh1IgRI2KHHXaIRo0aRatWraJ///7x2WefFevTrVu3yGQyxW4nnnhijioGAAAAoDLkNJQaN25cDB48ON5666147rnnYsmSJdGzZ89YuHBhsX6/+93vYubMmdnbZZddlqOKAQAAAKgMtXK58WeeeabY/VGjRkWrVq3ivffeiz333DPbXr9+/WjdunXa5QEAAACwhlSpOaUKCgoiIqJZs2bF2kePHh0tWrSILl26xNChQ+OHH37IRXkAAAAAVJKcjpT6paKiojj99NNjt912iy5dumTbjzrqqOjQoUO0bds2JkyYEH/+85/js88+i4cffrjU9RQWFkZhYWH2/rx589Z47QAAAACUT5UJpQYPHhwTJ06M1157rVj773//++y/t9pqq2jTpk3ss88+MWXKlOjUqVOJ9YwYMSKGDx++xusFAAAAoOKqxOl7p5xySjz55JPx0ksvxfrrr7/SvjvttFNEREyePLnU5UOHDo2CgoLsbcaMGZVeLwAAAACrJ6cjpZIkiT/84Q/xyCOPxMsvvxwdO3Zc5WM++OCDiIho06ZNqcvz8vIiLy+vMssEAAAAoJLlNJQaPHhwjBkzJh577LFo1KhRzJo1KyIi8vPzo169ejFlypQYM2ZM9O3bN5o3bx4TJkyIIUOGxJ577hlbb711LksHAAAAYDVkkiRJcrbxTKbU9pEjR8agQYNixowZ8Zvf/CYmTpwYCxcujPbt28dBBx0Uf/3rX6Nx48Zl2sa8efMiPz8/CgoKyvwYAAAAACqmrFlMzk/fW5n27dvHuHHjUqoGAAAAgLRUiYnOAQAAAFi3CKUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDU5TSUGjFiROywww7RqFGjaNWqVfTv3z8+++yzYn0WLVoUgwcPjubNm0fDhg3jkEMOidmzZ+eoYgAAAAAqQ05DqXHjxsXgwYPjrbfeiueeey6WLFkSPXv2jIULF2b7DBkyJJ544ol44IEHYty4cfHVV1/FwQcfnMOqAQAAAFhdmSRJklwXscw333wTrVq1inHjxsWee+4ZBQUF0bJlyxgzZkwceuihERHx6aefxuabbx5vvvlm7Lzzzqtc57x58yI/Pz8KCgqicePGa/opAAAAAKzTyprFVKk5pQoKCiIiolmzZhER8d5778WSJUuiR48e2T6bbbZZbLDBBvHmm2/mpEYAAAAAVl+tXBewTFFRUZx++umx2267RZcuXSIiYtasWVGnTp1o0qRJsb7rrbdezJo1q9T1FBYWRmFhYfb+vHnz1ljNAAAAAFRMlRkpNXjw4Jg4cWLce++9q7WeESNGRH5+fvbWvn37SqoQAAAAgMpSJUKpU045JZ588sl46aWXYv3118+2t27dOhYvXhxz584t1n/27NnRunXrUtc1dOjQKCgoyN5mzJixJksHAAAAoAJyGkolSRKnnHJKPPLII/Hiiy9Gx44diy3v2rVr1K5dO1544YVs22effRbTp0+PXXbZpdR15uXlRePGjYvdAAAAAKhacjqn1ODBg2PMmDHx2GOPRaNGjbLzROXn50e9evUiPz8/jj/++DjjjDOiWbNm0bhx4/jDH/4Qu+yyS5muvAcAAABA1ZRJkiTJ2cYzmVLbR44cGYMGDYqIiEWLFsUf//jHuOeee6KwsDB69eoVN9xwwwpP31teWS9DCAAAAMDqK2sWk9NQKg1CKQAAAID0lDWLqRITnQMAAACwbhFKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJC6CodSc+fOjX/+858xdOjQmDNnTkREvP/++/G///2v0ooDAAAAYO1UqyIPmjBhQvTo0SPy8/Nj2rRp8bvf/S6aNWsWDz/8cEyfPj3uvPPOyq4TAAAAgLVIhUZKnXHGGTFo0KCYNGlS1K1bN9vet2/feOWVVyqtOAAAAADWThUKpd5555044YQTSrS3a9cuZs2atdpFAQAAALB2q1AolZeXF/PmzSvR/t///jdatmy52kUBAAAAsHarUCjVr1+/OP/882PJkiUREZHJZGL69Onx5z//OQ455JBKLRAAAACAtU+FQql//OMfsWDBgmjVqlX8+OOPsddee8XGG28cjRo1iosuuqiyawQAAABgLVOhq+/l5+fHc889F6+99lpMmDAhFixYEL/61a+iR48elV0fAAAAAGuhTJIkSa6LWJPmzZsX+fn5UVBQEI0bN851OQAAAABrtbJmMRUaKXXNNdeU2p7JZKJu3bqx8cYbx5577hk1a9asyOoBAAAAWMtVKJS68sor45tvvokffvghmjZtGhER33//fdSvXz8aNmwYX3/9dWy00Ubx0ksvRfv27Su1YAAAAACqvwpNdH7xxRfHDjvsEJMmTYrvvvsuvvvuu/jvf/8bO+20U1x99dUxffr0aN26dQwZMqSy6wUAAABgLVChOaU6deoUDz30UGy77bbF2sePHx+HHHJIfP755/HGG2/EIYccEjNnzqysWivEnFIAAAAA6SlrFlOhkVIzZ86Mn376qUT7Tz/9FLNmzYqIiLZt28b8+fMrsnoAAAAA1nIVCqW6d+8eJ5xwQowfPz7bNn78+DjppJNi7733joiIjz76KDp27Fg5VQIAAACwVqlQKHXbbbdFs2bNomvXrpGXlxd5eXmx/fbbR7NmzeK2226LiIiGDRvGP/7xj0otFgAAAIC1Q4XmlFrm008/jf/+978REbHpppvGpptuWmmFVRZzSgEAAACkp6xZTK3V2chmm20Wm2222eqsAgAAAIB1UIVDqS+//DIef/zxmD59eixevLjYsiuuuGK1CwMAAABg7VWhUOqFF16Ifv36xUYbbRSffvppdOnSJaZNmxZJksSvfvWryq4RAAAAgLVMhSY6Hzp0aJx55pnx0UcfRd26deOhhx6KGTNmxF577RWHHXZYZdcIAAAAwFqmQqHUf/7znzjmmGMiIqJWrVrx448/RsOGDeP888+PSy+9tFILBAAAAGDtU6FQqkGDBtl5pNq0aRNTpkzJLvv2228rpzIAAAAA1loVmlNq5513jtdeey0233zz6Nu3b/zxj3+Mjz76KB5++OHYeeedK7tGAAAAANYyFQqlrrjiiliwYEFERAwfPjwWLFgQ9913X3Tu3NmV9wAAAABYpUySJEmui1iT5s2bF/n5+VFQUBCNGzfOdTkAAAAAa7WyZjEVmlNqo402iu+++65E+9y5c2OjjTaqyCoBAAAAWIdUKJSaNm1aLF26tER7YWFh/O9//1vtogAAAABYu5VrTqnHH388+++xY8dGfn5+9v7SpUvjhRdeiA033LDSigMAAABg7VSuUKp///4REZHJZGLgwIHFltWuXTs23HDD+Mc//lFpxQEAAACwdipXKFVUVBQRER07dox33nknWrRosUaKAgAAAGDtVq5QapmpU6dWdh0AAAAArEMqFEpFRLzwwgvxwgsvxNdff50dQbXM7bffvtqFAQAAALD2qlAoNXz48Dj//PNj++23jzZt2kQmk6nsugAAAABYi1UolLrpppti1KhRcfTRR1d2PQAAAACsA2pU5EGLFy+OXXfdtbJrAQAAAGAdUaFQ6re//W2MGTOmsmsBAAAAYB1RodP3Fi1aFLfccks8//zzsfXWW0ft2rWLLb/iiisqpTgAAAAA1k4VCqUmTJgQ2267bURETJw4sdgyk54DAAAAsCoVCqVeeumlyq4DAAAAgHVIheaUWmby5MkxduzY+PHHHyMiIkmSSikKAAAAgLVbhUKp7777LvbZZ5/YZJNNom/fvjFz5syIiDj++OPjj3/8Y6UWCAAAAMDap0Kh1JAhQ6J27doxffr0qF+/frb9iCOOiGeeeabSigMAAABg7VShOaWeffbZGDt2bKy//vrF2jt37hxffPFFpRQGAAAAwNqrQiOlFi5cWGyE1DJz5syJvLy81S4KAAAAgLVbhUKpPfbYI+68887s/UwmE0VFRXHZZZdF9+7dK604AAAAANZOFTp977LLLot99tkn3n333Vi8eHGcddZZ8fHHH8ecOXPi9ddfr+waAQAAAFjLVGikVJcuXeK///1v7L777nHggQfGwoUL4+CDD47x48dHp06dKrtGAAAAANYymSRJklwXsSbNmzcv8vPzo6CgIBo3bpzrcgAAAADWamXNYio0UmrkyJHxwAMPlGh/4IEH4o477qjIKgEAAABYh1QolBoxYkS0aNGiRHurVq3i4osvXu2iAAAAAFi7VSiUmj59enTs2LFEe4cOHWL69OmrXRQAAAAAa7cKhVKtWrWKCRMmlGj/8MMPo3nz5qtdFAAAAABrtwqFUkceeWSceuqp8dJLL8XSpUtj6dKl8eKLL8Zpp50WAwYMqOwaAQAAAFjL1KrIgy644IKYNm1a7LPPPlGr1s+rKCoqimOOOcacUgAAAACsUiZJkqQ8D0iSJGbMmBEtW7aML7/8Mj744IOoV69ebLXVVtGhQ4c1VWeFlfUyhAAAAACsvrJmMeUeKZUkSWy88cbx8ccfR+fOnaNz586rVSgAAAAA655yzylVo0aN6Ny5c3z33Xdroh4AAAAA1gEVmuj8kksuiT/96U8xceLEyq4HAAAAgHVAueeUioho2rRp/PDDD/HTTz9FnTp1ol69esWWz5kzp9IKXF3mlAIAAABIzxqbUyoi4qqrrqpoXQAAAABQsVBq4MCBlV0HAAAAAOuQCs0pFRExZcqU+Otf/xpHHnlkfP311xER8fTTT8fHH39cacUBAAAAsHaqUCg1bty42GqrreLtt9+Ohx9+OBYsWBARER9++GGcd955lVogAAAAAGufCoVSZ599dlx44YXx3HPPRZ06dbLte++9d7z11luVVhwAAAAAa6cKhVIfffRRHHTQQSXaW7VqFd9+++1qFwUAAADA2q1CoVSTJk1i5syZJdrHjx8f7dq1W+2iAAAAAFi7VSiUGjBgQPz5z3+OWbNmRSaTiaKionj99dfjzDPPjGOOOaayawQAAABgLVOhUOriiy+OzTffPDbYYINYsGBBbLHFFrHnnnvGrrvuGn/9618ru0YAAAAA1jK1ytO5qKgoLr/88nj88cdj8eLFcfTRR8chhxwSCxYsiO222y46d+68puoEAAAAYC1SrlDqoosuimHDhkWPHj2iXr16MWbMmEiSJG6//fY1VR8AAAAAa6Fynb535513xg033BBjx46NRx99NJ544okYPXp0FBUVran6AAAAAFgLlSuUmj59evTt2zd7v0ePHpHJZOKrr76q9MIAAAAAWHuVK5T66aefom7dusXaateuHUuWLKnUogAAAABYu5VrTqkkSWLQoEGRl5eXbVu0aFGceOKJ0aBBg2zbww8/XHkVAgAAALDWKVcoNXDgwBJtv/nNbyqtGAAAAADWDeUKpUaOHLmm6gAAAABgHVKuOaUAAAAAoDIIpQAAAABIXU5DqVdeeSUOOOCAaNu2bWQymXj00UeLLR80aFBkMplit969e+emWAAAAAAqTU5DqYULF8Y222wT119//Qr79O7dO2bOnJm93XPPPSlWCAAAAMCaUK6Jzitbnz59ok+fPivtk5eXF61bt06pIgAAAADSUOXnlHr55ZejVatWsemmm8ZJJ50U33333Ur7FxYWxrx584rdAAAAAKhaqnQo1bt377jzzjvjhRdeiEsvvTTGjRsXffr0iaVLl67wMSNGjIj8/PzsrX379ilWDAAAAEBZZJIkSXJdREREJpOJRx55JPr377/CPp9//nl06tQpnn/++dhnn31K7VNYWBiFhYXZ+/PmzYv27dtHQUFBNG7cuLLLBgAAAOAX5s2bF/n5+avMYqr0SKnlbbTRRtGiRYuYPHnyCvvk5eVF48aNi90AAAAAqFqqVSj15ZdfxnfffRdt2rTJdSkAAAAArIacXn1vwYIFxUY9TZ06NT744INo1qxZNGvWLIYPHx6HHHJItG7dOqZMmRJnnXVWbLzxxtGrV68cVg0AAADA6sppKPXuu+9G9+7ds/fPOOOMiIgYOHBg3HjjjTFhwoS44447Yu7cudG2bdvo2bNnXHDBBZGXl5erkgEAAACoBFVmovM1payTawEAAACw+tbKic4BAAAAWDsIpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABIXa1cFwAAAFBpMpny9U+SNVMHAKtkpBQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJC6WrkuAAAAIFcuGf9tmfuevV2LNVgJwLrHSCkAAAAAUieUAgAAACB1QikAAAAAUmdOKYBqpjxzX0SY/wIAAKiajJQCAAAAIHVCKQAAAABSJ5QCAAAAIHVCKQAAAABSJ5QCAAAAIHVCKQAAAABSJ5QCyLVMpnw3AACAtYBQCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASF1OQ6lXXnklDjjggGjbtm1kMpl49NFHiy1PkiTOPffcaNOmTdSrVy969OgRkyZNyk2xAAAAAFSanIZSCxcujG222Sauv/76Updfdtllcc0118RNN90Ub7/9djRo0CB69eoVixYtSrlSAAAAACpTrVxuvE+fPtGnT59SlyVJEldddVX89a9/jQMPPDAiIu68885Yb7314tFHH40BAwakWSoAAAAAlSinodTKTJ06NWbNmhU9evTItuXn58dOO+0Ub7755gpDqcLCwigsLMzenzdv3hqvFVbHJeO/LVf/s7drsYYqAQAAgPRU2YnOZ82aFRER6623XrH29dZbL7usNCNGjIj8/PzsrX379mu0TgAAAADKr8qGUhU1dOjQKCgoyN5mzJiR65IAAAAAWE6VDaVat24dERGzZ88u1j579uzsstLk5eVF48aNi90AAAAAqFqqbCjVsWPHaN26dbzwwgvZtnnz5sXbb78du+yySw4rAwAAAGB15XSi8wULFsTkyZOz96dOnRoffPBBNGvWLDbYYIM4/fTT48ILL4zOnTtHx44d429/+1u0bds2+vfvn7uiAQAAAFhtOQ2l3n333ejevXv2/hlnnBEREQMHDoxRo0bFWWedFQsXLozf//73MXfu3Nh9993jmWeeibp16+aqZAAAAAAqQU5DqW7dukWSJCtcnslk4vzzz4/zzz8/xaoAAAAAWNOq7JxSAAAAAKy9hFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBWtCJlP2G0A1lslkynUDAIBlhFIAAAAApE4oBQAAAEDqauW6AAAAoIop7+m2SbJm6gBgrWakFAAAAACpE0oBAAAAkDqhFAAAAACpM6cUAABQZWXKOb+V2a0Aqg8jpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABIXa1cFwBUX5eM/7Zc/c/ersUaqgQAAIDqRigFa7lMJlOu/kmSrKFKAAAA4P84fQ8AAACA1AmlAAAAAEidUAoAAACA1AmlAAAAAEidUAoAAACA1Ln6HlBcea7W9/43a64OAKDauGT8t2Xue/Z2LdZgJQBUJ0ZKAQAAAJA6I6UAAKi6yjOCNyIiSdZMHQBApTNSCgAAAIDUCaUAAAAASJ3T98gqzwSVESapBAAAACpOKAUAwFrDVeAAoPpw+h4AAAAAqRNKAQAAAJA6oRQAAAAAqRNKAQAAAJA6E51XM5lMplz9kyRZQ5UAAAAAVJyRUgAAAACkTii1tstkyn4DAAAASIlQCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASF2tXBcAAMC6I5PJlKt/sobqAAByTygFAFRZl4z/tlz9z96uxRqqBACAyub0PQAAAABSJ5QCAAAAIHVO3wNYA8ozZ4r5UgAAgHWRkVIAAAAApE4oBQAAAEDqhFIAAAAApM6cUgCQY+WZgywiIknMRAYAQPVnpBQAkJ5Mpnw3AADWWkIpAAAAAFInlAIAAAAgdUIpAAAAAFJnonMAAFjLlfuCCmuoDqoPF+EA0mCkFAAAAACpE0oBAAAAkDqhFAAAAACpE0oBAAAAkDqhFAAAAACpE0oBAAAAkDqhFAAAAACpE0oBAAAAkDqhFAAAAACpE0oBAAAAkLoqHUoNGzYsMplMsdtmm22W67JYBy2/H67qBgAAvkMCrFytXBewKltuuWU8//zz2fu1alX5kgGgSrlk/Lfl6n/2di3WUCUAAPB/qnzCU6tWrWjdunWuywAAAACgElXp0/ciIiZNmhRt27aNjTbaKH7961/H9OnTc10SAAAAAKupSo+U2mmnnWLUqFGx6aabxsyZM2P48OGxxx57xMSJE6NRo0alPqawsDAKCwuz9+fNm5dWuQAAAACUUZUOpfr06ZP999Zbbx077bRTdOjQIe6///44/vjjS33MiBEjYvjw4WmVCAAAAEAFVPnT936pSZMmsckmm8TkyZNX2Gfo0KFRUFCQvc2YMSPFCgEAAAAoi2oVSi1YsCCmTJkSbdq0WWGfvLy8aNy4cbEbAACsrkwmU64bALByVTqUOvPMM2PcuHExbdq0eOONN+Kggw6KmjVrxpFHHpnr0gAAAABYDVV6Tqkvv/wyjjzyyPjuu++iZcuWsfvuu8dbb70VLVu2zHVpAAAAAKyGKh1K3XvvvbkuAQAAAIA1oEqfvgcAAADA2kkoBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApE4oBQAAAEDqhFIAAAAApK5WrgsAgMqSyWTK3DdJkjVYCUBElOOYFBERjksArGOMlAIAAAAgdUZKAUB1U97RF+9/s2bqYK11yfhvy9X/7O1arKFKAIC1mVAKAACgKnDKJ7COEUoBAABUQ+UZ1bjGRzSWJ1ATpgH/nzmlAAAAAEidUAoAAACA1Dl9DwAAqoAqdSoWAKTASCkAAAAAUieUAgBY22Uy5bsBAKRAKAUAAABA6swpBQAAQGrKM39ahDnUYG0mlAIAqGYy5TzFLllDdQAArA6n7wEAAACQOqEUAAAAAKkTSgEAAACQOqEUAAAAAKkTSgEAAACQOlffA6qs8lxdKklcWwoAAKA6MVIKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInYnOAai4ckxGHxFxyfvflKv/2du1KFd/AACg+jBSCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASJ1QCgAAAIDUCaUAAAAASF2tXBcAQNWSyWTK3DdZg3UAAKxryvM9LCIiSXwbo3ozUgoAAACA1AmlAAAAAEid0/cAWDeVc3h8GB4PAACVykgpAAAAAFJnpBQAlMEl478tV/+zt2uxhioBAIC1g1AKAFgnucIRAEBuOX0PAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInYnOgXWSK6kBAADklpFSAAAAAKROKAUAAABA6oRSpCaTyZTrBuWSyZTvBgAAQE4JpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNQJpQAAAABInVAKAAAAgNTVynUBAADVQiZTru6XvP9NufqfvV2LcvUHgPK4ZPy35erv/yXSYKQUAAAAAKkTSgEAAEB1lMmU/QZVkFAKAAAAgNSZUwoAAIC1Rqaco4KSJFlDlQCrYqQUAAAAAKkTSgEAAACQOqEUAAAAAKkTSgEAAACQOhOds1a4ZPy35ep/9nYt1lAlAAAAQFkYKQUAAABA6oRSAAAAAKROKEXVlcmU/QYAAABUK0IpAAAAAFInlAIAAAAgdUIpAAAAAFInlAIAAAAgdUIpAAAAAFJXK9cFAAAAAKzIJeO/LVf/s7drsYYqobJVi5FS119/fWy44YZRt27d2GmnneLf//53rksCAABgbZDJlP1G5fG6E9VgpNR9990XZ5xxRtx0002x0047xVVXXRW9evWKzz77LFq1apXr8gAAAFhHGLEDlavKj5S64oor4ne/+10ce+yxscUWW8RNN90U9evXj9tvvz3XpQEAAABQQVV6pNTixYvjvffei6FDh2bbatSoET169Ig333yz1McUFhZGYWFh9n5BQUFERMybN2/NFltFledZL1owv3zrnlenfMWUk9pXsG61l77ucvZXe+VQ+0rWr/bS113O/mqvHGpfwbqr0P+pEWqvLGpfwbqrUO1V6RgTofYVrnsN114e1bn2ddWyDCZJkpX2yySr6pFDX331VbRr1y7eeOON2GWXXbLtZ511VowbNy7efvvtEo8ZNmxYDB8+PM0yAQAAAFjOjBkzYv3111/h8io9Uqoihg4dGmeccUb2flFRUcyZMyeaN28eGROkRcTPiWX79u1jxowZ0bhx41yXU2bVte4IteeK2nND7bmh9txQe26oPTfUnhtqz43qWnt1rTtC7WubJEli/vz50bZt25X2q9KhVIsWLaJmzZoxe/bsYu2zZ8+O1q1bl/qYvLy8yMvLK9bWpEmTNVVitda4ceNq+YGprnVHqD1X1J4bas8NteeG2nND7bmh9txQe25U19qra90Ral+b5Ofnr7JPlZ7ovE6dOtG1a9d44YUXsm1FRUXxwgsvFDudDwAAAIDqpUqPlIqIOOOMM2LgwIGx/fbbx4477hhXXXVVLFy4MI499thclwYAAABABVX5UOqII46Ib775Js4999yYNWtWbLvttvHMM8/Eeuutl+vSqq28vLw477zzSpzmWNVV17oj1J4ras8NteeG2nND7bmh9txQe26oPTeqa+3Vte4Ita+rqvTV9wAAAABYO1XpOaUAAAAAWDsJpQAAAABInVAKAAAAgNQJpQAAAKAa+eqrr3JdAlQKodQ6buLEibkuAYA1YMGCBbkuAVjOt99+G/Pmzct1GVRR559/fvzwww+5LqNCqvPfFFOnTs11CRWy5ZZbxpgxY3JdBqw2odQ6aP78+XHLLbfEjjvuGNtss02uyynViy++GFtssUWpX9wKCgpiyy23jFdffTUHlVGVff7551EdLyj6448/xpNPPpm9P3To0DjjjDOytz/96U+xaNGiHFZIVXPllVeudPn8+fOjV69eKVVTub788sv4/e9/n+syoNLMnTs3Bg8eHC1atIj11lsvmjZtGq1bt46hQ4dW2wBimR9//DHXJZTqzTffLPb/akTEnXfeGR07doxWrVrF73//+ygsLMxRdSs2fPjwavuDwtZbbx077bRT3HrrrTF//vxcl1MunTp1io4dO8Zxxx0Xd911V3z55Ze5LqlMLrroojjhhBPisMMOizlz5uS6HKiwTFId/4KjQl555ZW47bbb4qGHHoq2bdvGwQcfHIccckjssMMOuS6thH79+kX37t1jyJAhpS6/5ppr4qWXXopHHnkk5crKrqioKEaNGhUPP/xwTJs2LTKZTHTs2DEOPfTQOProoyOTyeS6xFL17ds37rnnnsjPz4+IiEsuuSROPPHEaNKkSUREfPfdd7HHHnvEJ598ksMqS1ezZs2YOXNmtGrVKiIijjjiiLjmmmtivfXWy3FlK3fTTTfFv/71r3jiiSciIqJRo0ax5ZZbRr169SIi4tNPP42zzjprhZ+HXCvrr/6NGzdew5WU38EHH1ymfg8//PAarqR86tWrFzfffHMcc8wxJZYtXLgwevbsGd999118+umnOahu9Xz44Yfxq1/9KpYuXZrrUko47rjjytTv9ttvX8OVrFuKiori8ssvj8cffzwWL14c++yzT5x33nnZY2RVNmfOnNhll13if//7X/z617+OzTffPCIiPvnkkxgzZkxsttlm8dprr8WECRPirbfeilNPPTXHFZdNYWFhXHfddXH55ZfHrFmzcl1OCX369Ilu3brFn//854iI+Oijj+JXv/pVDBo0KDbffPO4/PLL44QTTohhw4blttDl1KhRI2bNmpX9HlOdvPrqqzFy5Mh48MEHo6ioKA455JD47W9/G3vssUeuS1ull19+OXt7++23Y/HixbHRRhvF3nvvHd27d4/u3btX2e+SU6dOjeOPPz4++eSTuPXWW+OAAw7IdUmrbcGCBVFUVFSsrSp+h1zmq6++itdeey2+/vrrEnVXl2N6rgml1nKzZs2KUaNGxW233Rbz5s2Lww8/PG666ab48MMPY4sttsh1eSvUoUOHeOaZZ7Jf3pb36aefRs+ePWP69OkpV1Y2SZLEAQccEE899VRss802sdlmm0WSJPGf//wnPvroo+jXr188+uijuS6zVMsHO40bN44PPvggNtpoo4iImD17drRt27ZK/sG4/Je5Ro0axYcffpitvaraY4894qyzzsp+kVi+7rvvvjuuv/76ePPNN3NZ5grVqFFjpSFrkiSRyWSq5D5z7LHHlqnfyJEj13Al5fPggw/G0UcfHffdd1/069cv275w4cLo1atXfP311zFu3Lho06ZNDqusmKocStWoUSM6dOgQ22233UpHZVbFH0xW9TmNiMhkMvHTTz+lVFHZXXDBBTFs2LDo0aNH1KtXL8aOHRtHHnlktQj/Tj/99HjhhRfi+eefL/FH7axZs6Jnz56x6aabxrPPPhvXXHNNDBw4MEeVllRYWBjDhg2L5557LurUqRNnnXVW9O/fP0aOHBnnnHNO1KxZM0455ZRs8FOVtGnTJp544onYfvvtIyLinHPOiXHjxsVrr70WEREPPPBAnHfeeVXuB7YaNWrE7Nmzo2XLlrkupcIWLlwY999/f4waNSpeffXV2HjjjeP444+PgQMHRuvWrXNd3iotWrQo3njjjWxI9e9//zuWLFkSm222WXz88ce5Lm+FrrvuuhgyZEhsvvnmUatWrWLL3n///RxVVXZTp06NU045JV5++eViZwdU5e+QERGjRo2KE044IerUqRPNmzcv9v9sJpOJzz//PIfVVR+1Vt2F6uqAAw6IV155Jfbbb7+46qqronfv3lGzZs246aabcl3aKs2ePTtq1669wuW1atWKb775JsWKymfUqFHxyiuvxAsvvBDdu3cvtuzFF1+M/v37x5133lnqKIdcW/4PLbn1mjd58uTYaqutsvfr1q0bNWr839nVO+64YwwePDgXpZXJSy+9lP13kiTRt2/f+Oc//xnt2rXLYVVlU9XCprI69NBDY+7cuXHkkUfGv/71r+jWrVssXLgwevfuHbNnz662gVRVd9JJJ8U999wTU6dOjWOPPTZ+85vfRLNmzXJdVpmsLCh7880345prrinxC29Vceedd8YNN9wQJ5xwQkREPP/887HffvvFP//5z2LHyqro0UcfjZtvvrnUURatW7eOyy67LPr27RvnnXdelQqkIiLOPffcuPnmm6NHjx7xxhtvxGGHHRbHHntsvPXWW3HFFVfEYYcdFjVr1sx1maX6/vvvi73m48aNiz59+mTv77DDDjFjxoxclLZKm2yyySoD5Kp8qlaDBg3i2GOPjWOPPTYmT54cI0eOjOuvvz7+9re/Re/evePxxx/PdYkrVbdu3dh7771j9913j+7du8fTTz8dN998c5UeefzFF1/Eww8/HE2bNo0DDzywRChVHfzmN7+JJEni9ttvj/XWW6/KnlGyvL/97W9x7rnnxtChQ6v8/0dVWfXbYymzp59+Ok499dQ46aSTonPnzrkup1zatWsXEydOjI033rjU5RMmTKjSf3Ddc8898Ze//KVEIBURsffee8fZZ58do0ePrpKhVHWWyWRK/CdWHf5Tmzt3brG5LZYPXIuKiqrk3BfL7LXXXsXu16xZM3beeecqP0Ktuvvtb38bc+bMiQMPPDAee+yxOPfcc+Orr76KcePGRdu2bXNd3lrp+uuvjyuuuCIefvjhuP3222Po0KGx3377xfHHHx89e/as0sebAw88sETbZ599FmeffXY88cQT8etf/zrOP//8HFS2atOnT4++fftm7/fo0SMymUx89dVXsf766+ewslWbOXNmbLnllitc3qVLl6hRo0acd955KVZVNg888EDceeed0a9fv5g4cWJsvfXW8dNPP8WHH35Ypff1iIj11lsvpk6dGu3bt4/FixfH+++/H8OHD88unz9//kp//Myl4cOHZ6dQqO423njj+Mtf/hIdOnSIoUOHxr/+9a9cl7RCixcvjrfeeiteeuml7Gl87du3jz333DOuu+66Et91qopbb701/vjHP0aPHj3i448/rraj7D788MN47733YtNNN811KeXyww8/xIABAwRSq0kotRZ77bXX4rbbbouuXbvG5ptvHkcffXQMGDAg12WVSd++fbO/qNStW7fYsh9//DHOO++82H///XNU3apNmDAhLrvsshUu79OnT1xzzTUpVlR21TXYifh5lM6gQYMiLy8vIn4egn3iiSdGgwYNivWravMDrb/++jFx4sQV/kc8YcKEKv+HF7lx1llnxZw5c2KfffaJDTfcMF5++eUqv6+sah6vuXPnplNIBeXl5cWRRx4ZRx55ZHzxxRcxatSoOPnkk+Onn36Kjz/+OBo2bJjrElfpq6++ivPOOy/uuOOO6NWrV3zwwQfRpUuXXJe1Qj/99FOJ7wK1a9eOJUuW5KiismvRokVMmzZthZ/LqVOnVtn5g7788svo2rVrRPwcnuXl5cWQIUOqxXeCvn37xtlnnx2XXnppPProo1G/fv1icxtNmDAhOnXqlMMKV2zAgAFVdp8oj1deeSVuv/32eOihh6JGjRpx+OGHx/HHH5/rskq19957x9tvvx0dO3aMvfbaK0444YQYM2ZMlf4BPCKid+/e8e9//zuuu+66av9D97LRi9UtlDr++OPjgQceiLPPPjvXpVRrQqm12M477xw777xzXHXVVXHffffF7bffHmeccUYUFRXFc889F+3bt49GjRrlusxS/fWvf42HH344NtlkkzjllFOyB6hPP/00rr/++li6dGmcc845Oa5yxebMmbPSCRHXW2+9+P7771OsqOxWFexU5RE7y5/68Jvf/CZHlZRP375949xzz4399tuv1BB2+PDhsd9+++WoOqqi5YOd2rVrR4sWLeK0004r1l7VAtiIWOUIgPz8/Grz5XrZPE1JklTZ+S5+qaCgIC6++OK49tprY9ttt40XXnihWkxCvPz/SxGl/+hQFff3Xr16xTnnnJOdl+mXCgsLsz/AVUVLly4tVnOtWrWqRega8fM8ZAcffHDstdde0bBhw7jjjjuKPZfbb789evbsmcMKS1cdAr+V+eqrr2LUqFExatSomDx5cuy6665xzTXXxOGHH17iB8Kq5NVXX402bdrE3nvvHd26dYu99tormjdvnuuyVmnp0qVrzQ+X//znP+PEE0+M//3vf9GlS5cSIxm33nrrHFW2ciNGjIj9998/nnnmmdhqq61K1H3FFVfkqLLqxUTn65jPPvssbrvttrjrrrti7ty5se+++1bZc7u/+OKLOOmkk2Ls2LHZeY0ymUz06tUrrr/++ujYsWOOK1yxmjVrxqxZs1Y4hLYqTxZeXSd+rs5mz54d2267bdSpUydOOeWU2GSTTSLi58/rddddFz/99FOMHz++yl75ZXmNGjWKCRMmVOnPaHXnc5o7hYWF2dP3Xnvttdh///3j2GOPjd69e1fp4fuXXXZZXHrppdG6deu4+OKLSz2dr6qqzvv7l19+Gdtvv33k5eXF4MGDi1345IYbbojCwsJ45513YoMNNsh1qSXUqFEj+vTpkw0Dn3jiidh7772r/OjjXyooKIiGDRuWmPtqzpw50bBhwxJBYa5V56vv9enTJ55//vlo0aJFHHPMMXHcccdVm1EvCxcujFdffTVefvnleOmll+KDDz6ITTbZJPbaa69sSFVdT4urLt5666046qijYtq0adm2ZT/6VOWJzi+88MI499xzY9NNNy0xF1Ymk4kXX3wxh9VVH0KpddTSpUvjiSeeiNtvv73KhlLLfP/99zF58uRIkiQ6d+4cTZs2zXVJq7T8F7nlFRYWxjPPPFNlD7Ckb+rUqXHSSSfFc889VyyE3XfffeOGG26o0vMzLT9qpzr+4QJlcfLJJ8e9994b7du3j+OOOy5+/etfR4sWLXJdVpnUqFEj6tWrFz169Fjp5NQ+p5Vv6tSpcfLJJ8ezzz5b4vh+3XXXrXD+zFyrzmEg6evXr18cf/zxsf/++1fZCfDLav78+fHaa69l55f68MMPo3PnzjFx4sRcl7bW2mKLLWLzzTePs846q9SJzjt06JCjylauadOmceWVV8agQYNyXUq1JpSCNcAXOSpqzpw5MXny5Ij4eYLQ6nBlL/s764oaNWrEBhtsENttt91KT7OpisHOoEGDynRqkM/pmvP999/HpEmTIqL6HN9hXVRUVBTvvPNOvPTSS/HSSy/Fa6+9FosWLfJj8hrUoEGD+PDDD6tsSL8irVu3jldffbXaXVSsqhFKAQCUgWAHYO1TVFQU7777bvb0vddffz0WLlwY7dq1i+7du2dvVXW0ztrggAMOiEGDBsUhhxyS61LKZcSIETFz5swqewGr6kIoBQAAwDqpcePGsXDhwmjdunU2gOrWrVuVvULj2uiWW26JCy+8MI477rhSJwzv169fjipbuYMOOihefPHFaN68eWy55ZYl6q6KI6erIqEUAAAA66Sbb745unfvnr3QDOlb2YVCqvJE56uawsLI6bIRSgEAAACU0U8//RRjxoyJnj17RuvWrXNdTrVWda9dDAAAAKy1lixZErVq1ap2VzesVatWnHjiiVFYWJjrUqo9oRQAAACQutq1a8cGG2xQZU/RW5kdd9wxxo8fn+syqr1auS4AAAAAWDedc8458Ze//CXuuuuuaNasWa7LKbOTTz45/vjHP8aXX34ZXbt2jQYNGhRbvvXWW+eosurFnFIAAABATmy33XYxefLkWLJkSXTo0KFEuPP+++/nqLKVK22C9kwmE0mSVOkJ2qsaI6UAAACAnOjfv3+uS6iQqVOn5rqEtYKRUgAAAACkzkTnAAAAQM7MnTs3/vnPf8bQoUNjzpw5EfHzaXv/+9//clzZyt11112x2267Rdu2beOLL76IiIirrroqHnvssRxXVn0IpQAAAICcmDBhQmyyySZx6aWXxt///veYO3duREQ8/PDDMXTo0NwWtxI33nhjnHHGGdG3b9+YO3dudg6pJk2axFVXXZXb4qoRoRQAAACQE2eccUYMGjQoJk2aFHXr1s229+3bN1555ZUcVrZy1157bdx6661xzjnnRM2aNbPt22+/fXz00Uc5rKx6EUoBAAAAOfHOO+/ECSecUKK9Xbt2MWvWrBxUVDZTp06N7bbbrkR7Xl5eLFy4MAcVVU9CKQAAACAn8vLyYt68eSXa//vf/0bLli1zUFHZdOzYMT744IMS7c8880xsvvnm6RdUTdXKdQEAAADAuqlfv35x/vnnx/333x8REZlMJqZPnx5//vOf45BDDslxdSt2xhlnxODBg2PRokWRJEn8+9//jnvuuSdGjBgR//znP3NdXrWRSZIkyXURAAAAwLqnoKAgDj300Hj33Xdj/vz50bZt25g1a1bssssu8dRTT0WDBg1yXeIKjR49OoYNGxZTpkyJiIi2bdvG8OHD4/jjj89xZdWHUAoAAADIqddffz0+/PDDWLBgQfzqV7+KHj165LqkMvvhhx9iwYIF0apVq1yXUu0IpQAAAICcuPPOO+OII46IvLy8Yu2LFy+Oe++9N4455pgcVbZye++9dzz88MPRpEmTYu3z5s2L/v37x4svvpibwqoZoRQAAACQEzVr1oyZM2eWGGX03XffRatWrWLp0qU5qmzlatSoEbNmzSpR99dffx3t2rWLJUuW5Kiy6sVE5wAAAEBOJEkSmUymRPuXX34Z+fn5Oaho5SZMmJD99yeffBKzZs3K3l+6dGk888wz0a5du1yUVi0JpQAAAIBUbbfddpHJZCKTycQ+++wTtWr9XzyxdOnSmDp1avTu3TuHFZZu2223zda99957l1her169uPbaa3NQWfUklAIAAABS1b9//4iI+OCDD6JXr17RsGHD7LI6derEhhtuGIccckiOqluxqVOnRpIksdFGG8W///3vaNmyZXZZnTp1olWrVlGzZs0cVli9mFMKAAAAyIk77rgjjjjiiKhbt26uSyEHauS6AAAAAGDdNHDgwGwgdfLJJ8e3336b44rKr3HjxvH555/nuoxqSSgFAAAA5Nzdd98d8+bNy3UZ5eYEtIoTSgEAAAA5J9xZ9wilAAAAACroN7/5TTRu3DjXZVRLQikAAAAg5+bPnx8bbbRRrstYqRdffDG22GKLYqcZ3njjjf+vvfuPibp+4Dj++tydIANGbjIOjfAK0UvoxLma/KFZRtpiOaxW00YGzBz8EUsMW6DVAmRpjDzR0k7Kmm390FTEaauW/ZglUvJDjGtAm0RW6BJ/gnz/aN46gVK+7nPRPR/bbfD+vD9vXp//2Gvvz/s0duxYnTp1SlOmTNHnn38ewIQjC6UUAAAAAAAwlcVikdVq/duPzWYLdMwBKioqlJOTM+jOqKioKC1ZskRr164NQLKRyejnpU0AAAAAAGCiHTt2DHntq6++UmVlpS5duqRz586ZmOqfxcfHq7a2Vk6nc9DrR48eVVpamjo6OkxONjL9+2pHAAAAAADwn/bAAw8MGGtpaVFhYaF27typhQsX6oUXXghAsr/X1dWlUaNGDXndZrPpxIkTJiYa2Xh9DwAAAAAABMzx48eVk5Oj5ORk9fb2qr6+XtXV1YqPjw90tAHGjx+vhoaGIa9///33io2NNTHRyEYpBQAAAAAATHfq1Ck988wzSkhIUGNjoz7++GPt3LlTSUlJgY42pPvuu09FRUWDvlZ49uxZrVy5Uvfff38Ako1MnCkFAAAAAABMVV5ertWrV8tut6ukpGTQ1/n+jbq6ujRt2jRZrVbl5eVp0qRJkv48S8rtdquvr091dXWKiYkJcNKRgVIKAAAAAACYymKxKCwsTHPmzJHVah1y3gcffGBiqqvT3t6upUuXau/evbpcqRiGoXvvvVdut1sOhyPACUcOSikAAAAAAGCqxx9/XIZh/OM8j8djQprh6e7uVmtrq/r7+zVx4kSNGTMm0JFGHEopAAAAAAAAmI6DzgEAAAAAAGA6SikAAAAAAACYjlIKAAAAAAAApqOUAgAAAAAAgOkopQAAAK6TtrY2GYah+vr6Ied8+umnMgxDJ0+eNC0XAADAvxGlFAAACBqXv37aMAyNGjVKDodDy5cv17lz567L+nFxcers7FRSUtJ1WW+4Lj/jUJ9Vq1YFNB8AAIAk2QIdAAAAwExz586Vx+PRxYsXdejQIWVmZsowDK1evfr/Xttqtcput1+HlP+fzs5O38/vvvuuiouL1dLS4huLiIgIRCwAAAA/7JQCAABBJTQ0VHa7XXFxcZo/f77mzJmjffv2+a5funRJpaWlcjgcCgsLk8vl0nvvvee73t3drYULFyo6OlphYWGaOHGiPB6PpMFf36upqVFiYqLCwsI0e/ZstbW1+eVZtWqVpk6d6jdWUVGhCRMm+I1t2rRJTqdTo0eP1uTJk7V+/fohn9Fut/s+UVFRMgxDdrtdkZGRSkxMVG1trd/87du3Kzw8XH/88YfvGbZt26bU1FSNHj1aSUlJ+uyzz/zuaWho0Lx58xQREaGYmBg99thj+vXXX4fMBAAAcCVKKQAAELQaGhr05ZdfKiQkxDdWWlqqN998Uxs2bFBjY6Py8/O1aNEiXylTVFSkpqYm7dmzR83NzaqqqtLYsWMHXf+nn35SRkaG0tPTVV9fr+zsbBUWFl5zzrffflvFxcV66aWX1NzcrJKSEhUVFam6uvqa1gkPD9cjjzziK9Eu83g8evDBBxUZGekbKygo0NNPP63Dhw9rxowZSk9P12+//SZJOnnypO666y6lpKTo22+/VW1trbq6uvTwww9f87MBAIDgxet7AAAgqOzatUsRERHq7e3V+fPnZbFYtG7dOknS+fPnVVJSov3792vGjBmSpJtvvlkHDhzQxo0bNWvWLHV0dCglJUXTp0+XpAE7mv6qqqpKt9xyi9asWSNJmjRpko4cOXLNrwquXLlSa9asUUZGhiTJ4XCoqalJGzduVGZm5jWtlZ2drdTUVHV2dio2Nla//PKLampqtH//fr95eXl5WrBgge85amtrtXnzZi1fvlzr1q1TSkqKSkpKfPPfeOMNxcXF6dixY0pMTLymTAAAIDhRSgEAgKAye/ZsVVVVqaenR6+88opsNpuvfGltbdWZM2d0zz33+N1z4cIFpaSkSJKWLl2qBQsWqK6uTmlpaZo/f75SU1MH/VvNzc264447/MYul11Xq6enR16vV1lZWcrJyfGN9/b2Kioq6prWkqTbb79dU6ZMUXV1tQoLC7V161bFx8dr5syZQ+a02WyaPn26mpubJUnfffedPvnkk0HPpvJ6vZRSAADgqlBKAQCAoBIeHq6EhARJf+7ucblc2rx5s7KysnT69GlJ0u7duzV+/Hi/+0JDQyVJ8+bNU3t7u2pqarRv3z7dfffdys3N1csvvzysPBaLRf39/X5jFy9e9P18OdPrr78+oOCyWq3D+pvZ2dlyu90qLCyUx+PR4sWLZRjGVd9/+vRppaenD7rjKzY2dliZAABA8OFMKQAAELQsFoueffZZPffcczp79qxuvfVWhYaGqqOjQwkJCX6fuLg4333R0dHKzMzU1q1bVVFRoddee23Q9Z1Opw4ePOg39vXXX/v9Hh0drZ9//tmvmPrrQekxMTEaN26cfvzxxwGZHA7HsJ570aJFam9vV2VlpZqamgZ9BfCvOXt7e3Xo0CE5nU5J0rRp09TY2KgJEyYMyBQeHj6sTAAAIPhQSgEAgKD20EMPyWq1yu12KzIyUsuWLVN+fr6qq6vl9XpVV1enV1991XeoeHFxsXbs2KHW1lY1NjZq165dvrLmSk8++aR++OEHFRQUqKWlRe+88462bNniN+fOO+/UiRMnVF5eLq/XK7fbrT179vjNef7551VaWqrKykodO3ZMR44ckcfj0dq1a4f1zGPGjFFGRoYKCgqUlpamG2+8ccAct9utDz/8UEePHlVubq66u7v1xBNPSJJyc3P1+++/69FHH9U333wjr9ervXv3avHixerr6xtWJgAAEHwopQAAQFCz2WzKy8tTeXm5enp69OKLL6qoqEilpaVyOp2aO3eudu/e7duVFBISohUrVui2227TzJkzZbVatW3btkHXvummm/T+++9r+/btcrlc2rBhg9/h4NKfu6nWr18vt9stl8ulgwcPatmyZX5zsrOztWnTJnk8HiUnJ2vWrFnasmXLsHdKSVJWVpYuXLjgK5quVFZWprKyMrlcLh04cEAfffSR71sGx40bpy+++EJ9fX1KS0tTcnKynnrqKd1www2yWPj3EgAAXB2j/8pDDAAAAPCf99Zbbyk/P1/Hjx9XSEiIb7ytrU0Oh0OHDx/W1KlTAxcQAAD853HQOQAAQBA5c+aMOjs7VVZWpiVLlvgVUgAAAGZifzUAAEAQKS8v1+TJk2W327VixYpAxwEAAEGM1/cAAAAAAABgOnZKAQAAAAAAwHSUUgAAAAAAADAdpRQAAAAAAABMRykFAAAAAAAA01FKAQAAAAAAwHSUUgAAAAAAADAdpRQAAAAAAABMRykFAAAAAAAA01FKAQAAAAAAwHT/A0KsB+f35wciAAAAAElFTkSuQmCC\n"
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of Unique Accession Numbers for Significant Data in AdjacentSeq_all__(Torpor)___(Aug_Active).tsv: 1337\n",
            "Number of Unique Accession Numbers for Not Significant Data in AdjacentSeq_all__(Torpor)___(Aug_Active).tsv: 1890\n",
            "Total number of modifications in AdjacentSeq_all__(Torpor)___(Aug_Active).tsv: 17793\n",
            "Significant N-terminal (Met-loss) modifications: 19\n",
            "Non-significant N-terminal (Met-loss) modifications: 24\n",
            "Processing file: /content/output/modlist_files/AdjacentSeq_all__(Mar_Active)___(Torpor).tsv\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 1200x800 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAABKUAAAMWCAYAAAAgRDUeAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAB31klEQVR4nOzdd3hW9f0//ldYYUemDBFZiooDsY5WGYosBVEcaFVAah1Yq36sSq0DF46quDc4Cu5tVVRElDrqQkSrRUXBMhxIQBRQcn5/+OP+GhIgCeHcCT4e13VfV3LOuc/9zJ1z30meeZ/3yUmSJAkAAAAASFGVbAcAAAAA4NdHKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQWwns4777zIyckp0bY5OTlx3nnnbdhAxViV8euvvy7X/d59993RsWPHqF69emyyySbluu/K6o477oicnJz47LPPMsu6d+8e3bt3L7TdggUL4qCDDopGjRpFTk5OjBkzJl588cXIycmJF198MdXMERFbbLFFDB06NPXHTVtxz/HQoUNjiy22yFomilr9NfPZZ59FTk5O3HHHHeX+WAUFBdGpU6e46KKLyn3fG4MN/d7wwQcfRLVq1WLGjBkb7DFK4oQTToh99tknqxk2pGeeeSbq1q0bX331VbajABSilAI2GqvKgFW3atWqRcuWLWPo0KHxv//9L9vxNjoffvhhDB06NNq1axe33npr3HLLLdmOVKmccsopMXHixBg5cmTcfffd0adPnw3+mK+88kqcd955sWjRog3+WBvaU089FTk5OdGiRYsoKCjIdpxUff/993HeeeetsbycOnVq9O3bN1q2bBk1a9aMzTffPPr37x8TJkxIN2glcc8998ScOXPixBNPzCz75c+TqVOnFrlPkiTRqlWryMnJif322y/NuEUccsghkZOTE2eccUaZ95HN94Ztttkm9t133zjnnHNSf+xVZs2aFbfddlv89a9/jYifS9Ff/j6xpls2/slUVn369In27dvH6NGjS7T9U089Vam+PqDyqpbtAADl7fzzz482bdrEsmXL4rXXXos77rgjpk6dGjNmzIiaNWuW++P97W9/izPPPLPc91vRvfjii1FQUBBXX311tG/fPttxKrRnn322yLIXXngh9t9//zjttNMyy7bccsv44YcfokaNGhskxyuvvBKjRo2KoUOHFhnZ9tFHH0WVKpXnf1Xjx4+PLbbYIj777LN44YUXomfPnmXe16233lqpiq3vv/8+Ro0aFRFRZATeAw88EIceemjsuOOO8ec//zkaNGgQs2bNipdeeiluvfXWOPzww7OQuGK7/PLLY/DgwZGXl1dkXc2aNWPChAmxxx57FFo+ZcqU+OKLLyI3NzetmMVavHhxPPHEE7HFFlvEPffcE5dcckmJR+7+UrbfG4477rjo169ffPLJJ9GuXbsN+ljFufrqq6NNmzbRo0ePiIg466yz4g9/+ENm/RtvvBHXXHNN/PWvf42tt946s3z77bdPPev6OPbYY+O0006LUaNGRb169da67VNPPRXXX3+9YgrY4JRSwEanb9++sfPOO0dExB/+8Ido3LhxXHrppfH444/HIYccUu6PV61atahW7df3dvrll19GRJTraXvff/991K5du9z2V1EUVzJ9+eWXRZ67KlWqbJDitCSy/cd1aSxdujQee+yxGD16dIwbNy7Gjx+/XqVU9erVyzFddp133nmxzTbbxGuvvVbkuFv1muX/eeedd+Ldd9+NK664otj1/fr1iwceeCCuueaaQu/zEyZMiC5dupTrKdEFBQWxYsWKUr0HPPTQQ7Fy5coYO3Zs7LXXXvHSSy9Ft27dyi1TRDrvDT179owGDRrEnXfeGeeff/4Gf7xf+vHHH2P8+PFx3HHHZZatfhpfzZo145prrol99tmnSBFcFmn+rFu2bFnUqFEjqlSpEoMGDYo//elP8cADD8TRRx+dyuMDrEvl+ZcoQBntueeeERHxySefFFr+4YcfxkEHHRQNGzaMmjVrxs477xyPP/54oW1+/PHHGDVqVHTo0CFq1qwZjRo1ij322COee+65zDbFzSm1fPnyOOWUU6JJkyZRr169GDBgQHzxxRdFsq1pLps1zVP1j3/8I7p06RK1atWKhg0bxuDBg2POnDklfi6+/vrrOOSQQ6J+/frRqFGj+POf/xzLli0r9eNsscUWce6550ZERJMmTYqcxnDDDTfEtttuG7m5udGiRYsYMWJEkdNCunfvHp06dYq33norunbtGrVr186cOrF8+fI499xzo3379pGbmxutWrWK008/PZYvX77Or3HVfqdPnx7dunWL2rVrR/v27ePBBx+MiJ9HOOy6665Rq1at2GqrreL5558vso933nkn+vbtG/Xr14+6devG3nvvHa+99lqR7d5///3Ya6+9olatWrHZZpvFhRdeWOyIm1/Oj7PqtKAkSeL666/PnAYSUfx8RxERr7/+evTr1y8aNGgQderUie233z6uvvrqzPrp06fH0KFDo23btlGzZs1o1qxZHH300fHNN99ktjnvvPPiL3/5S0REtGnTJvO4q+a+Km7emE8//TQOPvjgaNiwYdSuXTt22223+Oc//1lom1WZ77///rjoootis802i5o1a8bee+8dH3/8cTHfofX3yCOPxA8//BAHH3xwDB48OB5++OFij+MvvvgiBg4cGHXq1ImmTZvGKaecUuwxVNzr8O9//3v89re/jUaNGkWtWrWiS5cumWNodf/4xz9il112idq1a0eDBg2ia9euRUbHPf3007HnnntGnTp1ol69erHvvvvG+++/XyRH3bp143//+18MHDgw6tatG02aNInTTjstVq5cGRE/z63UpEmTiIgYNWpUkdOIPvnkk/jNb35TbBHatGnTQp8XFBTEmDFjYtttt42aNWvGpptuGscee2x8++23hbZLkiQuvPDC2GyzzaJ27drRo0ePeP/998s019C4ceNir732iqZNm0Zubm5ss802ceONN5ZqH+Xp0UcfjRo1akTXrl2LXX/YYYfFN998U+g9f8WKFfHggw+ucdRZSY+dnJycOPHEE2P8+PGZ98tnnnmmVPnHjx8f++yzT/To0SO23nrrGD9+fLHbffjhh3HIIYdEkyZNMu99Z511VkSU7r3hzTffjJycnLjzzjuLPMbEiRMjJycnnnzyycyy//3vf3H00UfHpptuGrm5ubHtttvG2LFji9y3evXq0b1793jsscdK9fWXh6lTp8bXX39dpmJ7fX/WbbHFFrHffvvFs88+GzvuuGPUrFkzttlmm3j44YeLPFZp3o/vvffe+Nvf/hYtW7aM2rVrx+LFiyPi5/eA7bfffp3P89ChQ+P666+PiCh0uuIq9957b3Tp0iXq1asX9evXj+222y7zM6m0xwiAUgrY6K36xbpBgwaZZe+//37stttu8Z///CfOPPPMuOKKK6JOnToxcODAeOSRRzLbnXfeeTFq1Kjo0aNHXHfddXHWWWfF5ptvHm+//fZaH/MPf/hDjBkzJnr16hWXXHJJVK9ePfbdd9/1+jouuuiiOOqoo6JDhw5x5ZVXxsknnxyTJk2Krl27lngekEMOOSSWLVsWo0ePjn79+sU111wTf/zjH0v9OGPGjIkDDjggIiJuvPHGuPvuu+PAAw+MiJ+fsxEjRkSLFi3iiiuuiEGDBsXNN98cvXr1ih9//LHQY33zzTfRt2/f2HHHHWPMmDHRo0ePKCgoiAEDBsTf//736N+/f1x77bUxcODAuOqqq+LQQw8t0df57bffxn777Re77rprXHbZZZGbmxuDBw+O++67LwYPHhz9+vWLSy65JJYuXRoHHXRQLFmyJHPf999/P/bcc89499134/TTT4+zzz47Zs2aFd27d4/XX389s938+fOjR48eMW3atDjzzDPj5JNPjrvuuqtQWVScrl27xt133x0RP/83/u677858XpznnnsuunbtGh988EH8+c9/jiuuuCJ69OhR6Jf65557Lj799NMYNmxYXHvttTF48OC49957o1+/fpEkSUREHHjggXHYYYdFRMRVV12VedxVBcfqFixYEL/97W9j4sSJccIJJ8RFF10Uy5YtiwEDBhR6jaxyySWXxCOPPBKnnXZajBw5Ml577bX4/e9/v9bnoqzGjx8fPXr0iGbNmsXgwYNjyZIl8cQTTxTa5ocffoi99947Jk6cGCeeeGKcddZZ8fLLL8fpp59eose4+uqro3PnznH++efHxRdfHNWqVYuDDz64yB+Bo0aNiiOPPDKqV68e559/fowaNSpatWoVL7zwQmabu+++O/bdd9+oW7duXHrppXH22WfHBx98EHvssUehCfEjIlauXBm9e/eORo0axd///vfo1q1bXHHFFZk525o0aZIpcQ444IDM93HV669169YxadKkYkvw1R177LHxl7/8JX73u9/F1VdfHcOGDYvx48dH7969C71WzznnnDj77LNjhx12iMsvvzzatm0bvXr1iqVLl5boufylG2+8MVq3bh1//etf44orrohWrVrFCSeckPkDOG2vvPJKdOrUaY2j5bbYYovYfffd45577skse/rppyM/Pz8GDx5c7H1KeuxE/Hwa7ymnnBKHHnpoXH311aWacH/u3LkxefLkzOv6sMMOiwcffDBWrFhRaLvp06fHrrvuGi+88EIcc8wxcfXVV8fAgQMzr5nSvDfsvPPO0bZt27j//vuLrLvvvvuiQYMG0bt374j4+T1kt912i+effz5OPPHEzKnew4cPjzFjxhS5f5cuXWLGjBmZAiUtr7zySuTk5ETnzp1Ldb/1/Vm3ysyZM+PQQw+Nvn37xujRozPHyy+L0NK+H19wwQXxz3/+M0477bS4+OKLC5XUXbp0iVdeeWWtX9uxxx6bGS226nhY9XPqueeei8MOOywaNGgQl156aVxyySXRvXv3+Ne//hURpTtGACIiIgHYSIwbNy6JiOT5559Pvvrqq2TOnDnJgw8+mDRp0iTJzc1N5syZk9l27733Trbbbrtk2bJlmWUFBQXJb3/726RDhw6ZZTvssEOy7777rvVxzz333OSXb6fTpk1LIiI54YQTCm13+OGHJxGRnHvuuZllQ4YMSVq3br3OfX722WdJ1apVk4suuqjQdu+9915SrVq1IsvXtL8BAwYUWn7CCSckEZG8++67pX6cVfv86quvMsu+/PLLpEaNGkmvXr2SlStXZpZfd911SUQkY8eOzSzr1q1bEhHJTTfdVOix7r777qRKlSrJyy+/XGj5TTfdlERE8q9//WutX+uq/U6YMCGz7MMPP0wiIqlSpUry2muvZZZPnDgxiYhk3LhxmWUDBw5MatSokXzyySeZZXPnzk3q1auXdO3aNbPs5JNPTiIief311wt9/Xl5eUlEJLNmzSqUqVu3boVyRkQyYsSIQssmT56cREQyefLkJEmS5KeffkratGmTtG7dOvn2228LbVtQUJD5+Pvvvy/yPNxzzz1JRCQvvfRSZtnll19eJNsqrVu3ToYMGVLk6/vl92HJkiVJmzZtki222CLz/V2Veeutt06WL1+e2fbqq69OIiJ57733ijzW+liwYEFSrVq15NZbb80s++1vf5vsv//+hbYbM2ZMEhHJ/fffn1m2dOnSpH379oWe4yQp/nW4+nO6YsWKpFOnTslee+2VWTZz5sykSpUqyQEHHFDoeE+S//f9WbJkSbLJJpskxxxzTKH18+fPT/Ly8gotHzJkSBIRyfnnn19o286dOyddunTJfP7VV18VeS9Z5fbbb08iIqlRo0bSo0eP5Oyzz05efvnlIvlefvnlJCKS8ePHF1r+zDPPFFq+6jW97777Fjrm/vrXvyYRUeiYKYnijtXevXsnbdu2LbRs9dfMrFmzirxWy8Nmm22WDBo0qMjyVT9P3njjjeS6665L6tWrl8l+8MEHJz169EiS5OfXzeo/I0py7CRJknlPev/998uU/e9//3tSq1atZPHixUmSJMl///vfJCKSRx55pNB2Xbt2TerVq5d8/vnnhZb/8vtZmveGkSNHJtWrV08WLlyYWbZ8+fJkk002SY4++ujMsuHDhyfNmzdPvv7660L7Gzx4cJKXl1fkeZowYUKR99Q0HHHEEUmjRo3Wus0DDzxQ6H2jPH7WJcnPz21EJA899FBmWX5+ftK8efOkc+fOmWWlfT9u27Ztsa+1JEmSiy++OImIZMGCBWv9mkeMGFHo95BV/vznPyf169dPfvrppzXet6THCECSJImRUsBGp2fPntGkSZNo1apVHHTQQVGnTp14/PHHY7PNNouIiIULF8YLL7wQhxxySCxZsiS+/vrr+Prrr+Obb76J3r17x8yZMzNX69tkk03i/fffj5kzZ5b48Z966qmIiDjppJMKLT/55JPL/DU9/PDDUVBQEIccckgm79dffx3NmjWLDh06xOTJk0u0nxEjRhT6/E9/+lOhzOv7OM8//3ysWLEiTj755EIT4x5zzDFRv379IiMFcnNzY9iwYYWWPfDAA7H11ltHx44dC2XYa6+9IiJK9LXWrVu30CiGrbbaKjbZZJPYeuutY9ddd80sX/Xxp59+GhE/j1J59tlnY+DAgdG2bdvMds2bN4/DDz88pk6dmvkv/lNPPRW77bZb7LLLLpntmjRpUq6jg955552YNWtWnHzyyUXmn/rlqRS1atXKfLxs2bL4+uuvY7fddouIWOeovjV56qmnYpdddik0wXPdunXjj3/8Y3z22WfxwQcfFNp+2LBhhf4bv+q02VXPbXm59957M3OjrHLYYYfF008/Xei0s6eeeiqaN28eBx10UGZZ7dq1i4wMXJNfPqfffvtt5Ofnx5577lno+Xz00UejoKAgzjnnnCITQa/6/jz33HOxaNGiOOywwwodz1WrVo1dd9212OP5l3PbRPz8XJb0eTz66KPjmWeeie7du8fUqVPjggsuiD333DM6dOhQaHTEAw88EHl5ebHPPvsUytWlS5eoW7duJteq1/Sf/vSnQsdcWd/Pfvm85ufnx9dffx3dunWLTz/9NPLz88u0z/XxzTffFBpFW5xDDjkkfvjhh3jyySdjyZIl8eSTT651wviSHDurdOvWLbbZZpsyZR8/fnzsu+++mQmrO3ToEF26dCl0Ct9XX30VL730Uhx99NGx+eabF7p/WSZEj4g49NBD48cffyx0itmzzz4bixYtyoxmTZIkHnrooejfv38kSVLoGOvdu3fk5+cXeT5WfR/Kc56ukijJMbC68vhZt0qLFi0yI48jIurXrx9HHXVUvPPOOzF//vyIKP378ZAhQwodh7+0vs/zJptsEkuXLi00kmt1JTlGAFb59c3MC2z0rr/++thyyy0jPz8/xo4dGy+99FKhiVo//vjjSJIkzj777Dj77LOL3ceXX34ZLVu2jPPPPz/233//2HLLLaNTp07Rp0+fOPLII9d6xZ3PP/88qlSpUuQKQltttVWZv6aZM2dGkiTRoUOHYteXdKLm1e/frl27qFKlSuYUovV9nM8//zwiin6tNWrUiLZt22bWr9KyZcsic9/MnDkz/vOf/6zxtLKSTNa82WabFfmDKy8vL1q1alVkWURkyoyvvvoqvv/++2K/V1tvvXUUFBTEnDlzYtttt43PP/+8UMG1yvp8n1e3ah60Tp06rXW7hQsXxqhRo+Lee+8t8vyU9Q/9NX19q6489fnnnxfKtfofvKv+8Fl9fqJfWrFiRSxcuLDQsiZNmkTVqlXXeJ9V8zd98803mTmzOnfuHCtWrIgHHnggUzp9/vnn0b59+yLHQUm/P08++WRceOGFMW3atELzUP1yf5988klUqVJlraXCqkJ7Vam6uvr16xf6vGbNmkWO/QYNGqz1eVxd7969o3fv3vH999/HW2+9Fffdd1/cdNNNsd9++8WHH34YTZs2jZkzZ0Z+fn6ReaZWWXUcrXrNrv6e0KRJk1L/IR8R8a9//SvOPffcePXVV+P7778vtC4/P7/YK+BtaMn/f4rrmjRp0iR69uwZEyZMiO+//z5WrlxZqOxcXUmOnVXatGlTpsz/+c9/4p133omjjjqq0Nxt3bt3j+uvvz4WL14c9evXz5SZ63oPKY0ddtghOnbsGPfdd18MHz48In4+Latx48aZ4/yrr76KRYsWxS233JI59XR1q79Xrfo+lLUsWx/rOgZWVx4/61Yp7n1qyy23jIifpx9o1qxZqd+P13Zcre/zfMIJJ8T9998fffv2jZYtW0avXr3ikEMOiT59+mS2KckxArCKUgrY6Oyyyy6Zq+8NHDgw9thjjzj88MPjo48+irp162Ymoj7ttNPWOK9B+/btI+Ln+X8++eSTeOyxx+LZZ5+N2267La666qq46aabCl0uuqzW9EvhqkmNVykoKIicnJx4+umni/2DvW7duuXy+BvqcdakuP/kFhQUxHbbbRdXXnllsfdZvVgqzppKjTUtL+0fJBXNIYccEq+88kr85S9/iR133DFznPfp06fYidc3hLI8t6+88kqhuVUiImbNmrXGeXVmzpwZb7zxRkQULUkifh45UtKRUGvz8ssvx4ABA6Jr165xww03RPPmzaN69eoxbty4mDBhQqn2ter5v/vuu6NZs2ZF1q9+5c61FXKlVbt27dhzzz1jzz33jMaNG8eoUaPi6aefjiFDhkRBQUE0bdp0jRNjr6kUXh+ffPJJ7L333tGxY8e48soro1WrVlGjRo146qmn4qqrrkrtWP2lRo0alajwO/zww+OYY46J+fPnR9++fdd41dHSHjtrGs2yLv/4xz8iIuKUU06JU045pcj6hx56aI0jc8rDoYceGhdddFF8/fXXUa9evXj88cfjsMMOyxzPq76XRxxxRAwZMqTYfaz+z51V34fGjRtvsNzFKekxsD7K+n3eEI+3vs9z06ZNY9q0aTFx4sR4+umn4+mnn45x48bFUUcdVWhy83UdIwCreFcANmpVq1aN0aNHZyYqP/PMMzOnZVWvXr1EV9tp2LBhDBs2LIYNGxbfffdddO3aNc4777w1llKtW7eOgoKC+OSTTwr9F/Wjjz4qsm2DBg2KnaR89f+ytmvXLpIkiTZt2mT+g1oWM2fOLPQf1I8//jgKCgoyJcD6Pk7r1q0j4uev9Zenv61YsSJmzZpVoue7Xbt28e6778bee++d+n/MmzRpErVr1y72e/Xhhx9GlSpVMqVY69atiz2ts7j7ltWq0XYzZsxY43P37bffxqRJk2LUqFFxzjnnZJYXl600z2fr1q3X+DysWr++dthhhyKngBRX3Kwyfvz4qF69etx9991FypupU6fGNddcE7Nnz47NN988WrduHTNmzIgkSQp93SX5/jz00ENRs2bNmDhxYqFRluPGjSu0Xbt27aKgoCA++OCD2HHHHYvd16rvYdOmTct0da/ilOV1saqonzdvXibX888/H7/73e/W+gfsqu/zzJkzC72mv/rqq1L/If/EE0/E8uXL4/HHHy80sq6kpx9vCB07doxZs2atc7sDDjggjj322HjttdfivvvuW+N2JT121keSJDFhwoTo0aNHnHDCCUXWX3DBBTF+/PgYNmxY5ns2Y8aMte6ztMfUoYceGqNGjYqHHnooNt1001i8eHGhU6ZXXXl25cqVJT7uZ82aFVWqVFmvn3Fl0bFjxxg/fnypRuqVx8+6VVaN3v7l9+C///1vRETmZ3N5vh/PmjUrGjduvM7ieW3HRI0aNaJ///7Rv3//KCgoiBNOOCFuvvnmOPvsszP/1FvXMQKwijmlgI1e9+7dY5dddokxY8bEsmXLomnTptG9e/e4+eabM3+g/dJXX32V+XjV6UGr1K1bN9q3b1/sZeVX6du3b0REXHPNNYWWF3e1oXbt2kV+fn5Mnz49s2zevHlFrqZz4IEHRtWqVWPUqFFFRp4kSVIk55qsfoWra6+9tlDm9X2cnj17Ro0aNeKaa64pdP/bb7898vPzS3QFwkMOOST+97//xa233lpk3Q8//FCmK36VVNWqVaNXr17x2GOPFboq2oIFC2LChAmxxx57ZE636tevX7z22mvx73//O7PdV199tcaRJ2Wx0047RZs2bWLMmDFFystVz++qcmb171dxx1udOnUiIkp0tcZ+/frFv//973j11Vczy5YuXRq33HJLbLHFFmWeB+eXGjRoED179ix0q1mz5hq3Hz9+fOy5555x6KGHxkEHHVTotuqS9quuktavX7+YO3duPPjgg5n7f//992s8leiXqlatGjk5OYVGLH722Wfx6KOPFtpu4MCBUaVKlTj//POLjPJZ9f3o3bt31K9fPy6++OIiV+SKKPx+U1K1a9eOiOK/j5MmTSr2PqvmjVtVlB9yyCGxcuXKuOCCC4ps+9NPP2X23bNnz6hevXpce+21hY6x4o6vdSnuWM3Pzy/Xwqa0dt9995gxY8Za39Mjfn7vv/HGG+O8886L/v37r3G7kh476+Nf//pXfPbZZzFs2LAir4ODDjooDj300Jg8eXLMnTs3mjRpEl27do2xY8fG7NmzC+3nl9+H0rw3RPx82th2220X9913X9x3333RvHnz6Nq1a2Z91apVY9CgQfHQQw8VW4gVd9y/9dZbse2226Z+Cufuu+8eSZLEW2+9VeL7lMfPulXmzp1b6Gf+4sWL46677oodd9wxU9KX5/vxW2+9FbvvvnuhZfPmzYsPP/yw0HvUmo6J1X8PqFKlSmbU2y9fR+s6RgBWMVIK+FX4y1/+EgcffHDccccdcdxxx8X1118fe+yxR2y33XZxzDHHRNu2bWPBggXx6quvxhdffBHvvvtuRERss8020b179+jSpUs0bNgw3nzzzXjwwQfjxBNPXONj7bjjjnHYYYfFDTfcEPn5+fHb3/42Jk2aVGjej1UGDx4cZ5xxRhxwwAFx0kknxffffx833nhjbLnlloUmgW3Xrl1ceOGFMXLkyPjss89i4MCBUa9evZg1a1Y88sgj8cc//jFOO+20dT4Ps2bNigEDBkSfPn3i1VdfjX/84x9x+OGHxw477FAuj9OkSZMYOXJkjBo1Kvr06RMDBgyIjz76KG644Yb4zW9+E0ccccQ6Mx555JFx//33x3HHHReTJ0+O3/3ud7Fy5cr48MMP4/7774+JEydmRn1sCBdeeGE899xzsccee8QJJ5wQ1apVi5tvvjmWL18el112WWa7008/Pe6+++7o06dP/PnPf446derELbfcEq1bty5UMq6PKlWqxI033hj9+/ePHXfcMYYNGxbNmzePDz/8MN5///2YOHFi1K9fP7p27RqXXXZZ/Pjjj9GyZct49tlnix390aVLl4iIOOuss2Lw4MFRvXr16N+/f+aPj18688wz45577om+ffvGSSedFA0bNow777wzZs2aFQ899FCRib03tNdffz0+/vjjNb72WrZsGTvttFOMHz8+zjjjjDjmmGPiuuuui6OOOireeuutaN68edx9992ZQmdt9t1337jyyiujT58+cfjhh8eXX34Z119/fbRv377Q97Z9+/Zx1llnZSYTP/DAAyM3NzfeeOONaNGiRYwePTrq168fN954Yxx55JGx0047xeDBg6NJkyYxe/bs+Oc//xm/+93v4rrrrivVc1GrVq3YZptt4r777ostt9wyGjZsGJ06dYpOnTrF/vvvH23atIn+/ftHu3btYunSpfH888/HE088Eb/5zW8yhUq3bt3i2GOPjdGjR8e0adOiV69eUb169Zg5c2Y88MADcfXVV8dBBx0UTZo0idNOOy1Gjx4d++23X/Tr1y/eeeedePrpp0t9+k+vXr0yIyyOPfbY+O677+LWW2+Npk2bFvsPgjTsv//+ccEFF8SUKVOiV69ea912Taeh/VJJj531MX78+Khateoai48BAwbEWWedFffee2+ceuqpcc0118Qee+wRO+20U/zxj3+MNm3axGeffRb//Oc/Y9q0aRFRuveGVQ499NA455xzombNmjF8+PAi7wmXXHJJTJ48OXbdddc45phjYptttomFCxfG22+/Hc8//3yh+eR+/PHHmDJlSrEjvza0PfbYIxo1ahTPP/98iec7Ko+fdatsueWWMXz48HjjjTdi0003jbFjx8aCBQsKlbXl9X785ZdfxvTp04tc9GTkyJGZ/a0anbXqmDjppJOid+/eUbVq1Rg8eHD84Q9/iIULF8Zee+0Vm222WXz++edx7bXXxo477piZ42qVdR0jABFRzHU+ASqpX17Ce3UrV65M2rVrl7Rr1y5zGeNPPvkkOeqoo5JmzZol1atXT1q2bJnst99+yYMPPpi534UXXpjssssuySabbJLUqlUr6dixY3LRRRclK1asyGxz7rnnFrls8g8//JCcdNJJSaNGjZI6deok/fv3T+bMmVPsZdyfffbZpFOnTkmNGjWSrbbaKvnHP/5R7D6TJEkeeuihZI899kjq1KmT1KlTJ+nYsWMyYsSI5KOPPlrrc7Nqfx988EFy0EEHJfXq1UsaNGiQnHjiickPP/xQpsdZtc+vvvqqyP2vu+66pGPHjkn16tWTTTfdNDn++OOTb7/9ttA23bp1S7bddtti865YsSK59NJLk2233TbJzc1NGjRokHTp0iUZNWpUkp+fv9avdU37Le7S7Uny82XZR4wYUWjZ22+/nfTu3TupW7duUrt27aRHjx7JK6+8UuS+06dPT7p165bUrFkzadmyZXLBBRckt99+e5FLq69+efs1Pe6qy3mvuuz4KlOnTk322WefpF69ekmdOnWS7bffPrn22msz67/44ovkgAMOSDbZZJMkLy8vOfjgg5O5c+cWe7xdcMEFScuWLZMqVaoUyrn6Zd+T5OfXyEEHHZRssskmSc2aNZNddtklefLJJ4vN/MADDxRaPmvWrCQiknHjxhV53sriT3/6UxIRySeffLLGbc4777wkIpJ33303SZIk+fzzz5MBAwYktWvXTho3bpz8+c9/Tp555pkiz/GQIUOS1q1bF9rX7bffnnTo0CHJzc1NOnbsmIwbN26Nr8uxY8cmnTt3zhyr3bp1S5577rlC20yePDnp3bt3kpeXl9SsWTNp165dMnTo0OTNN98slKNOnTpF9l/c477yyitJly5dkho1ahT6Pt9zzz3J4MGDk3bt2iW1atVKatasmWyzzTbJWWedlSxevLjIvm+55ZakS5cuSa1atZJ69eol2223XXL66acnc+fOzWyzcuXKZNSoUUnz5s2TWrVqJd27d09mzJhR7DGzLo8//niy/fbbJzVr1ky22GKL5NJLL03Gjh27ztdMeR9Pv7T99tsnw4cPL7RsbT9Pfqm495WSHjvFvQesy4oVK5JGjRole+6551q3a9OmTdK5c+fM5zNmzMi8R9SsWTPZaqutkrPPPrvQfUrz3pAkSTJz5swkIpKISKZOnVpsjgULFiQjRoxIWrVqlVSvXj1p1qxZsvfeeye33HJLoe2efvrpJCKSmTNnluBZKH8nnXRS0r59+zWuf+CBB4p9b17fn3Wrjp+JEycm22+/feaYWf39NEnW7/14lRtvvDGpXbt2kfeCIUOGFHkN/vTTT8mf/vSnpEmTJklOTk7m+H3wwQeTXr16JU2bNk1q1KiRbL755smxxx6bzJs3r8jjleQYAchJkko+uysAwHo48sgj49VXXy12NCNrtsUWW0T37t3jjjvuyHaU9XL33XfHiBEjYvbs2WucwJwNa+DAgZGTk1Pk1PW0fPrpp9GxY8d4+umnY++9907tcbfYYovo1KlTPPnkk6k8XufOnaN79+5x1VVXpfJ4ACVhDCUA8Ks2b9681K/4RcXx+9//PjbffPMic+6Rjv/85z/x5JNPFju/WVratm0bw4cPj0suuSRrGTa0Z555JmbOnBkjR47MdhSAQswpBQD8Kk2fPj0effTReOmllzITpVM28+fPX+v6WrVqldsE1itWrCg0H1Fxfvzxx6hevfoa19eoUSMaNmwYET/P3bauq9OlZeHChbFixYo1rq9ateo6r5pW2Wy99dbx008/FVr2ww8/RH5+/lrvt2LFiqhRo8Ya19eqVSt++OGHte7jl1f7vPHGG0uQtvLq06dPfPfdd9mOAVCEUgoA+FV6+OGH49prr43BgwcbPbCemjdvvtb1Q4YMKbfT/F555ZXo0aPHWrdp3bp1fP7552tc361bt3jxxRfLJU95OvDAA2PKlClrXN+6detCVwbdWN13330xbNiw9drHkCFD4s4771zrNmYxAci+rM4pdeONN8aNN96Y+eG67bbbxjnnnJO5NHn37t2L/GA+9thj46abbko7KgAAa/D888+vdX2LFi1Kddn6tfn222/jrbfeWus26xol06BBg8zVxSqSt956K7799ts1rq9Vq1b87ne/SzFRdsybNy/ef//9tW7z7bffRoMGDda4vkWLFjF37ty17qNnz55lygdA+clqKfXEE09E1apVo0OHDpEkSdx5551x+eWXxzvvvBPbbrttdO/ePbbccss4//zzM/epXbt21K9fP1uRAQAAACgHWT19r3///oU+v+iii+LGG2+M1157LbbddtuI+LmE+uX53gAAAABUfhVmTqmVK1fGAw88EEuXLo3dd989s3z8+PHxj3/8I5o1axb9+/ePs88+O2rXrr3G/SxfvjyWL1+e+bygoCAWLlwYjRo1ipycnA36NQAAAAD82iVJEkuWLIkWLVpElSpV1rhd1kup9957L3bfffdYtmxZ1K1bNx555JHMnAOHH354tG7dOlq0aBHTp0+PM844Iz766KN4+OGH17i/0aNHx6hRo9KKDwAAAEAx5syZE5ttttka12d1TqmIny/nOnv27MjPz48HH3wwbrvttpgyZUqxk2G+8MILsffee8fHH38c7dq1K3Z/q4+Uys/Pj8033zzmzJljLioAAACADWzx4sXRqlWrWLRoUeTl5a1xu6yXUqvr2bNntGvXLm6++eYi65YuXRp169aNZ555Jnr37l2i/S1evDjy8vIiPz9fKQUAAACwgZW0i1nziX1ZUlBQUGik0y9NmzYtIiKaN2+eYiIAAAAAyltW55QaOXJk9O3bNzbffPNYsmRJTJgwIV588cWYOHFifPLJJzFhwoTo169fNGrUKKZPnx6nnHJKdO3aNbbffvtsxgYAAABgPWW1lPryyy/jqKOOinnz5kVeXl5sv/32MXHixNhnn31izpw58fzzz8eYMWNi6dKl0apVqxg0aFD87W9/y2ZkAAAAAMpBhZtTqryZUwoAAAAgPZV2TikAAAAANn5KKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVZLaVuvPHG2H777aN+/fpRv3792H333ePpp5/OrF+2bFmMGDEiGjVqFHXr1o1BgwbFggULspgYAAAAgPKQ1VJqs802i0suuSTeeuutePPNN2OvvfaK/fffP95///2IiDjllFPiiSeeiAceeCCmTJkSc+fOjQMPPDCbkQEAAAAoBzlJkiTZDvFLDRs2jMsvvzwOOuigaNKkSUyYMCEOOuigiIj48MMPY+utt45XX301dttttxLtb/HixZGXlxf5+flRv379DRkdAAAA4FevpF1MhZlTauXKlXHvvffG0qVLY/fdd4+33norfvzxx+jZs2dmm44dO8bmm28er7766hr3s3z58li8eHGhGwAAAAAVS9ZLqffeey/q1q0bubm5cdxxx8UjjzwS22yzTcyfPz9q1KgRm2yySaHtN91005g/f/4a9zd69OjIy8vL3Fq1arWBvwIAAAAASivrpdRWW20V06ZNi9dffz2OP/74GDJkSHzwwQdl3t/IkSMjPz8/c5szZ045pgUAAACgPFTLdoAaNWpE+/btIyKiS5cu8cYbb8TVV18dhx56aKxYsSIWLVpUaLTUggULolmzZmvcX25ubuTm5m7o2AAAAACsh6yPlFpdQUFBLF++PLp06RLVq1ePSZMmZdZ99NFHMXv27Nh9992zmBAAAACA9ZXVkVIjR46Mvn37xuabbx5LliyJCRMmxIsvvhgTJ06MvLy8GD58eJx66qnRsGHDqF+/fvzpT3+K3XffvcRX3gMAAACgYspqKfXll1/GUUcdFfPmzYu8vLzYfvvtY+LEibHPPvtERMRVV10VVapUiUGDBsXy5cujd+/eccMNN2QzMgAAAADlICdJkiTbITakxYsXR15eXuTn50f9+vWzHQcAAABgo1bSLqbCzSkFAAAAwMZPKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6rJaSo0ePTp+85vfRL169aJp06YxcODA+Oijjwpt071798jJySl0O+6447KUGAAAAIDykNVSasqUKTFixIh47bXX4rnnnosff/wxevXqFUuXLi203THHHBPz5s3L3C677LIsJQYAAACgPFTL5oM/88wzhT6/4447omnTpvHWW29F165dM8tr164dzZo1SzseAAAAABtIhZpTKj8/PyIiGjZsWGj5+PHjo3HjxtGpU6cYOXJkfP/999mIBwAAAEA5yepIqV8qKCiIk08+OX73u99Fp06dMssPP/zwaN26dbRo0SKmT58eZ5xxRnz00Ufx8MMPF7uf5cuXx/LlyzOfL168eINnBwAAAKB0KkwpNWLEiJgxY0ZMnTq10PI//vGPmY+32267aN68eey9997xySefRLt27YrsZ/To0TFq1KgNnhcAAACAsqsQp++deOKJ8eSTT8bkyZNjs802W+u2u+66a0REfPzxx8WuHzlyZOTn52duc+bMKfe8AAAAAKyfrI6USpIk/vSnP8UjjzwSL774YrRp02ad95k2bVpERDRv3rzY9bm5uZGbm1ueMQEAAAAoZ1ktpUaMGBETJkyIxx57LOrVqxfz58+PiIi8vLyoVatWfPLJJzFhwoTo169fNGrUKKZPnx6nnHJKdO3aNbbffvtsRgcAAABgPeQkSZJk7cFzcopdPm7cuBg6dGjMmTMnjjjiiJgxY0YsXbo0WrVqFQcccED87W9/i/r165foMRYvXhx5eXmRn59f4vsAAAAAUDYl7WKyfvre2rRq1SqmTJmSUhoAAAAA0lIhJjoHAAAA4NdFKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKQuq6XU6NGj4ze/+U3Uq1cvmjZtGgMHDoyPPvqo0DbLli2LESNGRKNGjaJu3boxaNCgWLBgQZYSAwAAAFAeslpKTZkyJUaMGBGvvfZaPPfcc/Hjjz9Gr169YunSpZltTjnllHjiiSfigQceiClTpsTcuXPjwAMPzGJqAAAAANZXTpIkSbZDrPLVV19F06ZNY8qUKdG1a9fIz8+PJk2axIQJE+Kggw6KiIgPP/wwtt5663j11Vdjt912W+c+Fy9eHHl5eZGfnx/169ff0F8CAAAAwK9aSbuYCjWnVH5+fkRENGzYMCIi3nrrrfjxxx+jZ8+emW06duwYm2++ebz66qtZyQgAAADA+quW7QCrFBQUxMknnxy/+93volOnThERMX/+/KhRo0ZssskmhbbddNNNY/78+cXuZ/ny5bF8+fLM54sXL95gmQEAAAAomwozUmrEiBExY8aMuPfee9drP6NHj468vLzMrVWrVuWUEAAAAIDyUiFKqRNPPDGefPLJmDx5cmy22WaZ5c2aNYsVK1bEokWLCm2/YMGCaNasWbH7GjlyZOTn52duc+bM2ZDRAQAAACiDrJZSSZLEiSeeGI888ki88MIL0aZNm0Lru3TpEtWrV49JkyZlln300Ucxe/bs2H333YvdZ25ubtSvX7/QDQAAAICKJatzSo0YMSImTJgQjz32WNSrVy8zT1ReXl7UqlUr8vLyYvjw4XHqqadGw4YNo379+vGnP/0pdt999xJdeQ8AAACAiiknSZIkaw+ek1Ps8nHjxsXQoUMjImLZsmXxf//3f3HPPffE8uXLo3fv3nHDDTes8fS91ZX0MoQAAAAArL+SdjFZLaXSoJQCAAAASE9Ju5gKMdE5AAAAAL8uSikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUlfmUmrRokVx2223xciRI2PhwoUREfH222/H//73v3ILBwAAAMDGqVpZ7jR9+vTo2bNn5OXlxWeffRbHHHNMNGzYMB5++OGYPXt23HXXXeWdEwAAAICNSJlGSp166qkxdOjQmDlzZtSsWTOzvF+/fvHSSy+VWzgAAAAANk5lKqXeeOONOPbYY4ssb9myZcyfP3+9QwEAAACwcStTKZWbmxuLFy8usvy///1vNGnSZL1DAQAAALBxK1MpNWDAgDj//PPjxx9/jIiInJycmD17dpxxxhkxaNCgcg0IAAAAwManTKXUFVdcEd999100bdo0fvjhh+jWrVu0b98+6tWrFxdddFF5ZwQAAABgI1Omq+/l5eXFc889F1OnTo3p06fHd999FzvttFP07NmzvPMBAAAAsBHKSZIkyXaIDWnx4sWRl5cX+fn5Ub9+/WzHAQAAANiolbSLKdNIqWuuuabY5Tk5OVGzZs1o3759dO3aNapWrVqW3QMAAACwkStTKXXVVVfFV199Fd9//300aNAgIiK+/fbbqF27dtStWze+/PLLaNu2bUyePDlatWpVroEBAAAAqPzKNNH5xRdfHL/5zW9i5syZ8c0338Q333wT//3vf2PXXXeNq6++OmbPnh3NmjWLU045pbzzAgAAALARKNOcUu3atYuHHnoodtxxx0LL33nnnRg0aFB8+umn8corr8SgQYNi3rx55ZW1TMwpBQAAAJCeknYxZRopNW/evPjpp5+KLP/pp59i/vz5ERHRokWLWLJkSVl2DwAAAMBGrkylVI8ePeLYY4+Nd955J7PsnXfeieOPPz722muviIh47733ok2bNuWTEgAAAICNSplKqdtvvz0aNmwYXbp0idzc3MjNzY2dd945GjZsGLfffntERNStWzeuuOKKcg0LAAAAwMahTHNKrfLhhx/Gf//734iI2GqrrWKrrbYqt2DlxZxSAAAAAOkpaRdTbX0epGPHjtGxY8f12QUAAAAAv0JlLqW++OKLePzxx2P27NmxYsWKQuuuvPLK9Q4GAAAAwMarTKXUpEmTYsCAAdG2bdv48MMPo1OnTvHZZ59FkiSx0047lXdGAAAAADYyZZrofOTIkXHaaafFe++9FzVr1oyHHnoo5syZE926dYuDDz64vDMCAAAAsJEpUyn1n//8J4466qiIiKhWrVr88MMPUbdu3Tj//PPj0ksvLdeAAAAAAGx8ylRK1alTJzOPVPPmzeOTTz7JrPv666/LJxkAAAAAG60yzSm12267xdSpU2PrrbeOfv36xf/93//Fe++9Fw8//HDstttu5Z0RAAAAgI1MmUqpK6+8Mr777ruIiBg1alR89913cd9990WHDh1ceQ8AAACAdcpJkiTJdogNafHixZGXlxf5+flRv379bMcBAAAA2KiVtIsp05xSbdu2jW+++abI8kWLFkXbtm3LsksAAAAAfkXKVEp99tlnsXLlyiLLly9fHv/73//WOxQAAAAAG7dSzSn1+OOPZz6eOHFi5OXlZT5fuXJlTJo0KbbYYotyCwcAAADAxqlUpdTAgQMjIiInJyeGDBlSaF316tVjiy22iCuuuKLcwgEAAACwcSpVKVVQUBAREW3atIk33ngjGjduvEFCAQAAALBxK1UptcqsWbPKOwcAAAAAvyJlKqUiIiZNmhSTJk2KL7/8MjOCapWxY8eudzAAAAAANl5lKqVGjRoV559/fuy8887RvHnzyMnJKe9cAAAAAGzEylRK3XTTTXHHHXfEkUceWd55AAAAAPgVqFKWO61YsSJ++9vflncWAAAAAH4lylRK/eEPf4gJEyaUdxYAAAAAfiXKdPresmXL4pZbbonnn38+tt9++6hevXqh9VdeeWW5hAMAAABg41SmUmr69Omx4447RkTEjBkzCq0z6TkAAAAA61KmUmry5MnlnQMAAACAX5EyzSm1yscffxwTJ06MH374ISIikiQpl1AAAAAAbNzKVEp98803sffee8eWW24Z/fr1i3nz5kVExPDhw+P//u//yjUgAAAAABufMpVSp5xySlSvXj1mz54dtWvXziw/9NBD45lnnim3cAAAAABsnMo0p9Szzz4bEydOjM0226zQ8g4dOsTnn39eLsEAAAAA2HiVaaTU0qVLC42QWmXhwoWRm5u73qEAAAAA2LiVqZTac88946677sp8npOTEwUFBXHZZZdFjx49yi0cAAAAABunMp2+d9lll8Xee+8db775ZqxYsSJOP/30eP/992PhwoXxr3/9q7wzAgAAALCRKdNIqU6dOsV///vf2GOPPWL//fePpUuXxoEHHhjvvPNOtGvXrrwzAgAAALCRyUmSJMl2iA1p8eLFkZeXF/n5+VG/fv1sxwEAAADYqJW0iynTSKlx48bFAw88UGT5Aw88EHfeeWdZdgkAAADAr0iZSqnRo0dH48aNiyxv2rRpXHzxxesdCgAAAICNW5lKqdmzZ0ebNm2KLG/dunXMnj17vUMBAAAAsHErUynVtGnTmD59epHl7777bjRq1Gi9QwEAAACwcStTKXXYYYfFSSedFJMnT46VK1fGypUr44UXXog///nPMXjw4PLOCAAAAMBGplpZ7nTBBRfEZ599FnvvvXdUq/bzLgoKCuKoo44ypxQAAAAA65STJElSmjskSRJz5syJJk2axBdffBHTpk2LWrVqxXbbbRetW7feUDnLrKSXIQQAAABg/ZW0iyn1SKkkSaJ9+/bx/vvvR4cOHaJDhw7rFRQAAACAX59SzylVpUqV6NChQ3zzzTcbIg8AAAAAvwJlmuj8kksuib/85S8xY8aM8s4DAAAAwK9AqeeUioho0KBBfP/99/HTTz9FjRo1olatWoXWL1y4sNwCri9zSgEAAACkZ4PNKRURMWbMmLLmAgAAAICylVJDhgwp7xwAAAAA/IqUaU6piIhPPvkk/va3v8Vhhx0WX375ZUREPP300/H++++XWzgAAAAANk5lKqWmTJkS2223Xbz++uvx8MMPx3fffRcREe+++26ce+655RoQAAAAgI1PmUqpM888My688MJ47rnnokaNGpnle+21V7z22mvlFg4AAACAjVOZSqn33nsvDjjggCLLmzZtGl9//fV6hwIAAABg41amUmqTTTaJefPmFVn+zjvvRMuWLdc7FAAAAAAbtzKVUoMHD44zzjgj5s+fHzk5OVFQUBD/+te/4rTTToujjjqqvDMCAAAAsJEpUyl18cUXx9Zbbx2bb755fPfdd7HNNttE165d47e//W387W9/K++MAAAAAGxkqpVm44KCgrj88svj8ccfjxUrVsSRRx4ZgwYNiu+++y46d+4cHTp02FA5AQAAANiIlKqUuuiii+K8886Lnj17Rq1atWLChAmRJEmMHTt2Q+UDAAAAYCNUqtP37rrrrrjhhhti4sSJ8eijj8YTTzwR48ePj4KCgg2VDwAAAICNUKlKqdmzZ0e/fv0yn/fs2TNycnJi7ty55R4MAAAAgI1XqUqpn376KWrWrFloWfXq1ePHH38s11AAAAAAbNxKNadUkiQxdOjQyM3NzSxbtmxZHHfccVGnTp3Msocffrj8EgIAAACw0SlVKTVkyJAiy4444ohyCwMAAADAr0OpSqlx48ZtqBwAAAAA/IqUak4pAAAAACgPSikAAAAAUpfVUuqll16K/v37R4sWLSInJyceffTRQuuHDh0aOTk5hW59+vTJTlgAAAAAyk1WS6mlS5fGDjvsENdff/0at+nTp0/Mmzcvc7vnnntSTAgAAADAhlCqic7LW9++faNv375r3SY3NzeaNWuWUiIAAAAA0lDh55R68cUXo2nTprHVVlvF8ccfH998881at1++fHksXry40A0AAACAiqVCl1J9+vSJu+66KyZNmhSXXnppTJkyJfr27RsrV65c431Gjx4deXl5mVurVq1STAwAAABASeQkSZJkO0RERE5OTjzyyCMxcODANW7z6aefRrt27eL555+Pvffeu9htli9fHsuXL898vnjx4mjVqlXk5+dH/fr1yzs2AAAAAL+wePHiyMvLW2cXU6FHSq2ubdu20bhx4/j444/XuE1ubm7Ur1+/0A0AAACAiqVSlVJffPFFfPPNN9G8efNsRwEAAABgPWT16nvfffddoVFPs2bNimnTpkXDhg2jYcOGMWrUqBg0aFA0a9YsPvnkkzj99NOjffv20bt37yymBgAAAGB9ZbWUevPNN6NHjx6Zz0899dSIiBgyZEjceOONMX369Ljzzjtj0aJF0aJFi+jVq1dccMEFkZubm63IAAAAAJSDCjPR+YZS0sm1AAAAAFh/G+VE5wAAAABsHJRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6qplOwAAAEC5yckp3fZJsmFyALBORkoBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkLpq2Q4AAACQLZe883WJtz2zc+MNmATg10cpBVDJlOaX5wi/QAMAABWT0/cAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCiDbcnJKdwMAANgIZLWUeumll6J///7RokWLyMnJiUcffbTQ+iRJ4pxzzonmzZtHrVq1omfPnjFz5szshAUAAACg3GS1lFq6dGnssMMOcf311xe7/rLLLotrrrkmbrrppnj99dejTp060bt371i2bFnKSQEAAAAoT9Wy+eB9+/aNvn37FrsuSZIYM2ZM/O1vf4v9998/IiLuuuuu2HTTTePRRx+NwYMHpxkVAAAAgHJUYeeUmjVrVsyfPz969uyZWZaXlxe77rprvPrqq2u83/Lly2Px4sWFbgAAAABULBW2lJo/f35ERGy66aaFlm+66aaZdcUZPXp05OXlZW6tWrXaoDkBAAAAKL0KW0qV1ciRIyM/Pz9zmzNnTrYjAQAAALCaCltKNWvWLCIiFixYUGj5ggULMuuKk5ubG/Xr1y90AwAAAKBiyepE52vTpk2baNasWUyaNCl23HHHiIhYvHhxvP7663H88cdnNxyUo0ve+bpU25/ZufEGSgIAAADpyWop9d1338XHH3+c+XzWrFkxbdq0aNiwYWy++eZx8sknx4UXXhgdOnSINm3axNlnnx0tWrSIgQMHZi80AAAAAOstq6XUm2++GT169Mh8fuqpp0ZExJAhQ+KOO+6I008/PZYuXRp//OMfY9GiRbHHHnvEM888EzVr1sxWZAAAAADKQVZLqe7du0eSJGtcn5OTE+eff36cf/75KaYCAAAAYEOrsBOdAwAAALDxUkoBAAAAkDqlFGwIOTklvwEAAMCvkFIKAAAAgNRldaJzAKByyynliM+1XeAEAIBfFyOlAAAAAEidUgoAAACA1CmlAAAAAEidUgoAAACA1JnoHAAAKKyUFzEIFzEAoAyMlAIAAAAgdUZKAQAAFVZOKUdtGbMFUHkYKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKSuWrYDAAAAldsl73xd4m3P7Nx4AyYBoDIxUgoAAACA1BkpBRu5nJycUm2fJMkGSgIAAAD/j5FSAAAAAKTOSCmgsNKMrDKqCgAAgDJSSgFlVppJTSNMbAoAAMD/4/Q9AAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdSY6BwBgo1Gai3C4AAcAZJeRUgAAAACkTikFAAAAQOqUUgAAAACkzpxSZJRmDoYI8zAAAAAAZWekFAAAFVdOTuluAECloZQCAAAAIHVKKQAAAABSp5QCAAAAIHUmOq9kcko5V0KSJBsoCQAAAEDZGSkFAAAAQOqUUgAAAACkTikFAAAAQOqUUhu7nJyS3wAAAABSopQCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSVy3bAQAA1uSSd74u1fZndm68gZJQXnJKeXGVZAPlAACyz0gpAAAAAFKnlAIAAAAgdUopAAAAAFJnTimADaA0c6aYLwUAAPg1MlIKAAAAgNQppQAAAABInVIKAAAAgNSZUwoAsqw0c5BFRCSJmcgAAKj8jJQCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSZ6JzAADYyJX6ggobKAeVh4twAGkwUgoASE9OTuluAABstJRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKSuQpdS5513XuTk5BS6dezYMdux+BVa/Thc1w0AAPwOCbB21bIdYF223XbbeP755zOfV6tW4SMDAAAAsA4VvuGpVq1aNGvWLNsxAAAAAChHFfr0vYiImTNnRosWLaJt27bx+9//PmbPnp3tSAAAAACspwo9UmrXXXeNO+64I7baaquYN29ejBo1Kvbcc8+YMWNG1KtXr9j7LF++PJYvX575fPHixWnFBQAAAKCEKnQp1bdv38zH22+/fey6667RunXruP/++2P48OHF3mf06NExatSotCICAAAAUAYV/vS9X9pkk01iyy23jI8//niN24wcOTLy8/Mztzlz5qSYEAAAAICSqFSl1HfffReffPJJNG/efI3b5ObmRv369QvdAABgfeXk5JTqBgCsXYUupU477bSYMmVKfPbZZ/HKK6/EAQccEFWrVo3DDjss29EAIHtyckp3AwCACqhCzyn1xRdfxGGHHRbffPNNNGnSJPbYY4947bXXokmTJtmOBgAAAMB6qNCl1L333pvtCAAAAABsABX69D0AAAAANk5KKQAAAABSp5QCAAAAIHUVek4pAGD9XfLO16Xa/szOjTdQEgAA+H+MlAIAAAAgdUopAAAAAFLn9D0AANgQcnJKt32SbJgcAFBBGSkFAAAAQOqUUgAAAACkzul7AGw0ckpxqkziNBkAAMgqI6UAAAAASJ1SCgAAAIDUOX0PAACgErrkna9LvO2ZnRtvwCQAZaOUAgCgkNL8oRvhj10oN6WYGzEiIt7+asPkAEiJ0/cAAAAASJ1SCgAAAIDUOX0PAACA9VOaUw+TZMPlACoVpRQAwMbOPDWVgkmr+bUwbx2witP3AAAAAEidUgoAAACA1CmlAAAAAEidOaUAACqZnFLOEWVKYQCgIjJSCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUufoeUGGV5upSSeLaUgAAAJWJkVIAAAAApE4pBQAAAEDqnL4HQGoueefrUm1/ZufGGygJAACQbUZKAQAAAJA6pRQAAAAAqVNKAQAAAJA6pRQAAAAAqVNKAQAAAJA6pRQAZZeTU7obAADA/08pBQAAAEDqlFIAAAAApE4pBQAAAEDqlFIAAAAApE4pBQAAAEDqlFIAAAAApK5atgMAULHk5OSUeNtkA+YAAPi1Kc3vYRERSeK3MSo3I6UAAAAASJ1SCgAAAIDUOX0PgF+nUg6PD8PjAQCgXBkpBQAAAEDqjJQCgBK45J2vS7X9mZ0bb6AkAACwcVBKAQC/Sq5wBACQXU7fAwAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1rr4HbBxKeRWtS97+qlTbn9m5cam2BwAAYO2MlAIAAAAgdUZKkZqcUo5kSZJkAyUBAAAAss1IKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSVy3bAQAANkaXvPN1qbY/s3PjDZQEAPxcomIyUgoAAACA1CmlAABKIiendDcAANZKKQUAAABA6pRSAAAAAKROKQUAAMBGIycnp1S3Ss1p5VRySikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB11bIdAMrDJe98Xartz+zceAMlAQAAAErCSCkAAAAAUqeUAgAAACB1Sikqrpyckt8AAACASkUpBQAAAEDqlFIAAAAApE4pBQAAAEDqlFIAAAAApE4pBQAAAEDqqmU7AAAAAMCaXPLO16Xa/szOjTdQEspbpSilrr/++rj88stj/vz5scMOO8S1114bu+yyS7ZjAQAA8CuiHIHyVeFP37vvvvvi1FNPjXPPPTfefvvt2GGHHaJ3797x5ZdfZjsaAAAAlV1OTslvlB/PO1EJSqkrr7wyjjnmmBg2bFhss802cdNNN0Xt2rVj7Nix2Y4GAAAAQBlV6NP3VqxYEW+99VaMHDkys6xKlSrRs2fPePXVV4u9z/Lly2P58uWZz/Pz8yMiYvHixRs2bAVVmq962XdLSrfvxTVKF6aUZF/DvmUvft+l3F728iH7WvYve/H7LuX2spcP2dew7wr0MzVC9vIi+xr2XYGyV6T3mAjZ17jvDZy9NCpz9l+rVR1MkiRr3S4nWdcWWTR37txo2bJlvPLKK7H77rtnlp9++ukxZcqUeP3114vc57zzzotRo0alGRMAAACA1cyZMyc222yzNa6v0COlymLkyJFx6qmnZj4vKCiIhQsXRqNGjSLHuagR8XNj2apVq5gzZ07Ur18/23FKrLLmjpA9W2TPDtmzQ/bskD07ZM8O2bND9uyorNkra+4I2Tc2SZLEkiVLokWLFmvdrkKXUo0bN46qVavGggULCi1fsGBBNGvWrNj75ObmRm5ubqFlm2yyyYaKWKnVr1+/Ur5gKmvuCNmzRfbskD07ZM8O2bND9uyQPTtkz47Kmr2y5o6QfWOSl5e3zm0q9ETnNWrUiC5dusSkSZMyywoKCmLSpEmFTucDAAAAoHKp0COlIiJOPfXUGDJkSOy8886xyy67xJgxY2Lp0qUxbNiwbEcDAAAAoIwqfCl16KGHxldffRXnnHNOzJ8/P3bcccd45plnYtNNN812tEorNzc3zj333CKnOVZ0lTV3hOzZInt2yJ4dsmeH7Nkhe3bInh2yZ0dlzV5Zc0fI/mtVoa++BwAAAMDGqULPKQUAAADAxkkpBQAAAEDqlFIAAAAApE4pBQAAAJXI3Llzsx0ByoVS6lduxowZ2Y4AwAbw3XffZTsCsJqvv/46Fi9enO0YVFDnn39+fP/999mOUSaV+W+KWbNmZTtCmWy77bYxYcKEbMeA9aaU+hVasmRJ3HLLLbHLLrvEDjvskO04xXrhhRdim222KfYXt/z8/Nh2223j5ZdfzkIyKrJPP/00KuMFRX/44Yd48sknM5+PHDkyTj311MztL3/5SyxbtiyLCalorrrqqrWuX7JkSfTu3TulNOXriy++iD/+8Y/ZjgHlZtGiRTFixIho3LhxbLrpptGgQYNo1qxZjBw5stIWEKv88MMP2Y5QrFdffbXQz9WIiLvuuivatGkTTZs2jT/+8Y+xfPnyLKVbs1GjRlXafyhsv/32seuuu8att94aS5YsyXacUmnXrl20adMmjj766Lj77rvjiy++yHakErnooovi2GOPjYMPPjgWLlyY7ThQZjlJZfwLjjJ56aWX4vbbb4+HHnooWrRoEQceeGAMGjQofvOb32Q7WhEDBgyIHj16xCmnnFLs+muuuSYmT54cjzzySMrJSq6goCDuuOOOePjhh+Ozzz6LnJycaNOmTRx00EFx5JFHRk5OTrYjFqtfv35xzz33RF5eXkREXHLJJXHcccfFJptsEhER33zzTey5557xwQcfZDFl8apWrRrz5s2Lpk2bRkTEoYceGtdcc01suummWU62djfddFP885//jCeeeCIiIurVqxfbbrtt1KpVKyIiPvzwwzj99NPX+HrItpL+179+/fobOEnpHXjggSXa7uGHH97ASUqnVq1acfPNN8dRRx1VZN3SpUujV69e8c0338SHH36YhXTr5913342ddtopVq5cme0oRRx99NEl2m7s2LEbOMmvS0FBQVx++eXx+OOPx4oVK2LvvfeOc889N/MeWZEtXLgwdt999/jf//4Xv//972PrrbeOiIgPPvggJkyYEB07doypU6fG9OnT47XXXouTTjopy4lLZvny5XHdddfF5ZdfHvPnz892nCL69u0b3bt3jzPOOCMiIt57773YaaedYujQobH11lvH5ZdfHscee2ycd9552Q26mipVqsT8+fMzv8dUJi+//HKMGzcuHnzwwSgoKIhBgwbFH/7wh9hzzz2zHW2dXnzxxczt9ddfjxUrVkTbtm1jr732ih49ekSPHj0q7O+Ss2bNiuHDh8cHH3wQt956a/Tv3z/bkdbbd999FwUFBYWWVcTfIVeZO3duTJ06Nb788ssiuSvLe3q2KaU2cvPnz4877rgjbr/99li8eHEccsghcdNNN8W7774b22yzTbbjrVHr1q3jmWeeyfzytroPP/wwevXqFbNnz045WckkSRL9+/ePp556KnbYYYfo2LFjJEkS//nPf+K9996LAQMGxKOPPprtmMVavdipX79+TJs2Ldq2bRsREQsWLIgWLVpUyD8YV/9lrl69evHuu+9msldUe+65Z5x++umZXyRWz/2Pf/wjrr/++nj11VezGXONqlSpstaSNUmSyMnJqZDHzLBhw0q03bhx4zZwktJ58MEH48gjj4z77rsvBgwYkFm+dOnS6N27d3z55ZcxZcqUaN68eRZTlk1FLqWqVKkSrVu3js6dO691VGZF/IfJul6nERE5OTnx008/pZSo5C644II477zzomfPnlGrVq2YOHFiHHbYYZWi/Dv55JNj0qRJ8fzzzxf5o3b+/PnRq1ev2GqrreLZZ5+Na665JoYMGZKlpEUtX748zjvvvHjuueeiRo0acfrpp8fAgQNj3LhxcdZZZ0XVqlXjxBNPzBQ/FUnz5s3jiSeeiJ133jkiIs4666yYMmVKTJ06NSIiHnjggTj33HMr3D/YqlSpEgsWLIgmTZpkO0qZLV26NO6///6444474uWXX4727dvH8OHDY8iQIdGsWbNsx1unZcuWxSuvvJIpqf7973/Hjz/+GB07doz3338/2/HW6LrrrotTTjkltt5666hWrVqhdW+//XaWUpXcrFmz4sQTT4wXX3yx0NkBFfl3yIiIO+64I4499tioUaNGNGrUqNDP2ZycnPj000+zmK7yqLbuTais+vfvHy+99FLsu+++MWbMmOjTp09UrVo1brrppmxHW6cFCxZE9erV17i+WrVq8dVXX6WYqHTuuOOOeOmll2LSpEnRo0ePQuteeOGFGDhwYNx1113FjnLIttX/0NJbb3gff/xxbLfddpnPa9asGVWq/L+zq3fZZZcYMWJENqKVyOTJkzMfJ0kS/fr1i9tuuy1atmyZxVQlU9HKppI66KCDYtGiRXHYYYfFP//5z+jevXssXbo0+vTpEwsWLKi0hVRFd/zxx8c999wTs2bNimHDhsURRxwRDRs2zHasEllbUfbqq6/GNddcU+Q/vBXFXXfdFTfccEMce+yxERHx/PPPx7777hu33XZboffKiujRRx+Nm2++udhRFs2aNYvLLrss+vXrF+eee26FKqQiIs4555y4+eabo2fPnvHKK6/EwQcfHMOGDYvXXnstrrzyyjj44IOjatWq2Y5ZrG+//bbQcz5lypTo27dv5vPf/OY3MWfOnGxEW6ctt9xynQVyRT5Vq06dOjFs2LAYNmxYfPzxxzFu3Li4/vrr4+yzz44+ffrE448/nu2Ia1WzZs3Ya6+9Yo899ogePXrE008/HTfffHOFHnn8+eefx8MPPxwNGjSI/fffv0gpVRkcccQRkSRJjB07NjbddNMKe0bJ6s4+++w455xzYuTIkRX+51FFVvmOWErs6aefjpNOOimOP/746NChQ7bjlErLli1jxowZ0b59+2LXT58+vUL/wXXPPffEX//61yKFVETEXnvtFWeeeWaMHz++QpZSlVlOTk6RH2KV4YfaokWLCs1tsXrhWlBQUCHnvlilW7duhT6vWrVq7LbbbhV+hFpl94c//CEWLlwY+++/fzz22GNxzjnnxNy5c2PKlCnRokWLbMfbKF1//fVx5ZVXxsMPPxxjx46NkSNHxr777hvDhw+PXr16Vej3m/3337/Iso8++ijOPPPMeOKJJ+L3v/99nH/++VlItm6zZ8+Ofv36ZT7v2bNn5OTkxNy5c2OzzTbLYrJ1mzdvXmy77bZrXN+pU6eoUqVKnHvuuSmmKpkHHngg7rrrrhgwYEDMmDEjtt9++/jpp5/i3XffrdDHekTEpptuGrNmzYpWrVrFihUr4u23345Ro0Zl1i9ZsmSt//zMplGjRmWmUKjs2rdvH3/961+jdevWMXLkyPjnP/+Z7UhrtGLFinjttddi8uTJmdP4WrVqFV27do3rrruuyO86FcWtt94a//d//xc9e/aM999/v9KOsnv33Xfjrbfeiq222irbUUrl+++/j8GDByuk1pNSaiM2derUuP3226NLly6x9dZbx5FHHhmDBw/OdqwS6devX+Y/KjVr1iy07ocffohzzz039ttvvyylW7fp06fHZZddtsb1ffv2jWuuuSbFRCVXWYudiJ9H6QwdOjRyc3Mj4uch2Mcdd1zUqVOn0HYVbX6gzTbbLGbMmLHGH8TTp0+v8H94kR2nn356LFy4MPbee+/YYost4sUXX6zwx8q65vFatGhROkHKKDc3Nw477LA47LDD4vPPP4877rgjTjjhhPjpp5/i/fffj7p162Y74jrNnTs3zj333Ljzzjujd+/eMW3atOjUqVO2Y63RTz/9VOR3gerVq8ePP/6YpUQl17hx4/jss8/W+LqcNWtWhZ0/6IsvvoguXbpExM/lWW5ubpxyyimV4neCfv36xZlnnhmXXnppPProo1G7du1CcxtNnz492rVrl8WEazZ48OAKe0yUxksvvRRjx46Nhx56KKpUqRKHHHJIDB8+PNuxirXXXnvF66+/Hm3atIlu3brFscceGxMmTKjQ/wCPiOjTp0/8+9//juuuu67S/6N71ejFylZKDR8+PB544IE488wzsx2lUlNKbcR222232G233WLMmDFx3333xdixY+PUU0+NgoKCeO6556JVq1ZRr169bMcs1t/+9rd4+OGHY8stt4wTTzwx8wb14YcfxvXXXx8rV66Ms846K8sp12zhwoVrnRBx0003jW+//TbFRCW3rmKnIo/YWf3UhyOOOCJLSUqnX79+cc4558S+++5bbAk7atSo2HfffbOUjopo9WKnevXq0bhx4/jzn/9caHlFK2AjYp0jAPLy8irNL9er5mlKkqTCznfxS/n5+XHxxRfHtddeGzvuuGNMmjSpUkxCvPrPpYji/+lQEY/33r17x1lnnZWZl+mXli9fnvkHXEW0cuXKQpmrVatWKUrXiJ/nITvwwAOjW7duUbdu3bjzzjsLfS1jx46NXr16ZTFh8SpD4bc2c+fOjTvuuCPuuOOO+Pjjj+O3v/1tXHPNNXHIIYcU+QdhRfLyyy9H8+bNY6+99oru3btHt27dolGjRtmOtU4rV67caP5xedttt8Vxxx0X//vf/6JTp05FRjJuv/32WUq2dqNHj4799tsvnnnmmdhuu+2K5L7yyiuzlKxyMdH5r8xHH30Ut99+e9x9992xaNGi2GeffSrsud2ff/55HH/88TFx4sTMvEY5OTnRu3fvuP7666NNmzZZTrhmVatWjfnz569xCG1Fniy8sk78XJktWLAgdtxxx6hRo0aceOKJseWWW0bEz6/X6667Ln766ad45513KuyVX1ZXr169mD59eoV+jVZ2XqfZs3z58szpe1OnTo399tsvhg0bFn369KnQw/cvu+yyuPTSS6NZs2Zx8cUXF3s6X0VVmY/3L774InbeeefIzc2NESNGFLrwyQ033BDLly+PN954IzbffPNsRy2iSpUq0bdv30wZ+MQTT8Ree+1V4Ucf/1J+fn7UrVu3yNxXCxcujLp16xYpCrOtMl99r2/fvvH8889H48aN46ijjoqjjz660ox6Wbp0abz88svx4osvxuTJk2PatGmx5ZZbRrdu3TIlVWU9La6yeO211+Lwww+Pzz77LLNs1T99KvJE5xdeeGGcc845sdVWWxWZCysnJydeeOGFLKarPJRSv1IrV66MJ554IsaOHVthS6lVvv322/j4448jSZLo0KFDNGjQINuR1mn1X+RWt3z58njmmWcq7Bss6Zs1a1Ycf/zx8dxzzxUqYffZZ5+44YYbKvT8TKuP2qmMf7hASZxwwglx7733RqtWreLoo4+O3//+99G4ceNsxyqRKlWqRK1ataJnz55rnZza67T8zZo1K0444YR49tlni7y/X3fddWucPzPbKnMZSPoGDBgQw4cPj/3226/CToBfUkuWLImpU6dm5pd69913o0OHDjFjxoxsR9tobbPNNrH11lvH6aefXuxE561bt85SsrVr0KBBXHXVVTF06NBsR6nUlFKwAfhFjrJauHBhfPzxxxHx8wShleHKXo53fi2qVKkSm2++eXTu3Hmtp9lUxGJn6NChJTo1yOt0w/n2229j5syZEVF53t/h16igoCDeeOONmDx5ckyePDmmTp0ay5Yt88/kDahOnTrx7rvvVtiSfk2aNWsWL7/8cqW7qFhFo5QCACgBxQ7AxqegoCDefPPNzOl7//rXv2Lp0qXRsmXL6NGjR+ZWUUfrbAz69+8fQ4cOjUGDBmU7SqmMHj065s2bV2EvYFVZKKUAAAD4Vapfv34sXbo0mjVrlimgunfvXmGv0LgxuuWWW+LCCy+Mo48+utgJwwcMGJClZGt3wAEHxAsvvBCNGjWKbbfdtkjuijhyuiJSSgEAAPCrdPPNN0ePHj0yF5ohfWu7UEhFnuh8XVNYGDldMkopAAAAgBL66aefYsKECdGrV69o1qxZtuNUahX32sUAAADARuvHH3+MatWqVbqrG1arVi2OO+64WL58ebajVHpKKQAAACB11atXj80337zCnqK3Nrvssku888472Y5R6VXLdgAAAADg1+mss86Kv/71r3H33XdHw4YNsx2nxE444YT4v//7v/jiiy+iS5cuUadOnULrt99++ywlq1zMKQUAAABkRefOnePjjz+OH3/8MVq3bl2k3Hn77bezlGztipugPScnJ5IkqdATtFc0RkoBAAAAWTFw4MBsRyiTWbNmZTvCRsFIKQAAAABSZ6JzAAAAIGsWLVoUt912W4wcOTIWLlwYET+ftve///0vy8nW7u67747f/e530aJFi/j8888jImLMmDHx2GOPZTlZ5aGUAgAAALJi+vTpseWWW8all14af//732PRokUREfHwww/HyJEjsxtuLW688cY49dRTo1+/frFo0aLMHFKbbLJJjBkzJrvhKhGlFAAAAJAVp556agwdOjRmzpwZNWvWzCzv169fvPTSS1lMtnbXXntt3HrrrXHWWWdF1apVM8t33nnneO+997KYrHJRSgEAAABZ8cYbb8Sxxx5bZHnLli1j/vz5WUhUMrNmzYrOnTsXWZ6bmxtLly7NQqLKSSkFAAAAZEVubm4sXry4yPL//ve/0aRJkywkKpk2bdrEtGnTiix/5plnYuutt04/UCVVLdsBAAAAgF+nAQMGxPnnnx/3339/RETk5OTE7Nmz44wzzohBgwZlOd2anXrqqTFixIhYtmxZJEkS//73v+Oee+6J0aNHx2233ZbteJVGTpIkSbZDAAAAAL8++fn5cdBBB8Wbb74ZS5YsiRYtWsT8+fNj9913j6eeeirq1KmT7YhrNH78+DjvvPPik08+iYiIFi1axKhRo2L48OFZTlZ5KKUAAACArPrXv/4V7777bnz33Xex0047Rc+ePbMdqcS+//77+O6776Jp06bZjlLpKKUAAACArLjrrrvi0EMPjdzc3ELLV6xYEffee28cddRRWUq2dnvttVc8/PDDsckmmxRavnjx4hg4cGC88MIL2QlWySilAAAAgKyoWrVqzJs3r8goo2+++SaaNm0aK1euzFKytatSpUrMnz+/SO4vv/wyWrZsGT/++GOWklUuJjoHAAAAsiJJksjJySmy/Isvvoi8vLwsJFq76dOnZz7+4IMPYv78+ZnPV65cGc8880y0bNkyG9EqJaUUAAAAkKrOnTtHTk5O5OTkxN577x3Vqv2/emLlypUxa9as6NOnTxYTFm/HHXfM5N5rr72KrK9Vq1Zce+21WUhWOSmlAAAAgFQNHDgwIiKmTZsWvXv3jrp162bW1ahRI7bYYosYNGhQltKt2axZsyJJkmjbtm38+9//jiZNmmTW1ahRI5o2bRpVq1bNYsLKxZxSAAAAQFbceeedceihh0bNmjWzHYUsqJLtAAAAAMCv05AhQzKF1AknnBBff/11lhOVXv369ePTTz/NdoxKSSkFAAAAZN0//vGPWLx4cbZjlJoT0MpOKQUAAABknXLn10cpBQAAAFBGRxxxRNSvXz/bMSolpRQAAACQdUuWLIm2bdtmO8ZavfDCC7HNNtsUOs3wxhtvjMaNG0d+fn5su+228fLLL2cxYeWilAIAAABSVaVKlahatepab9WqVct2zCLGjBkTxxxzTLEjo/Ly8uLYY4+NK6+8MgvJKqecxEmbAAAAQIoee+yxNa579dVX45prromCgoJYtmxZiqnWrXXr1vHMM8/E1ltvXez6Dz/8MHr16hWzZ89OOVnlVPFqRwAAAGCjtv/++xdZ9tFHH8WZZ54ZTzzxRPz+97+P888/PwvJ1m7BggVRvXr1Na6vVq1afPXVVykmqtycvgcAAABkzdy5c+OYY46J7bbbLn766aeYNm1a3HnnndG6detsRyuiZcuWMWPGjDWunz59ejRv3jzFRJWbUgoAAABIXX5+fpxxxhnRvn37eP/992PSpEnxxBNPRKdOnbIdbY369esXZ599drGnFf7www9x7rnnxn777ZeFZJWTOaUAAACAVF122WVx6aWXRrNmzeLiiy8u9nS+imjBggWx0047RdWqVePEE0+MrbbaKiJ+nkvq+uuvj5UrV8bbb78dm266aZaTVg5KKQAAACBVVapUiVq1akXPnj2jatWqa9zu4YcfTjFVyXz++edx/PHHx8SJE2NVpZKTkxO9e/eO66+/Ptq0aZPlhJWHUgoAAABI1dChQyMnJ2ed240bNy6FNGXz7bffxscffxxJkkSHDh2iQYMG2Y5U6SilAAAAAEidic4BAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAMrJZ599Fjk5OTFt2rQ1bvPiiy9GTk5OLFq0KLVcAAAVkVIKgP+vvXsNabqL4wD+nTMvTLEgcVorV2autLmQQl9odhF9IYhWFBliLir0RZKKRpoROZMsMeeltGlZGBRZeAuFCCzCUgsvS3OhBq27QWp5q+fFg39aGaRPLHj2/cAftrNzzn7n3fhyzhmRxZj++2mRSIR58+ZBLpcjJSUFX758+SPzy2QyGI1GeHt7/5H55mp6jb96MjMz/2p9RERERABg/bcLICIiIjKn0NBQ6HQ6TExMoLW1FTExMRCJRDh58uR/nlssFkMqlf6BKv8bo9EovL569SoyMjLQ09MjtDk4OPyNsoiIiIhMcKcUERERWRRbW1tIpVLIZDJERERg8+bNaGxsFD7/+vUrNBoN5HI57O3toVQqce3aNeHzoaEh7Nq1C87OzrC3t8eKFSug0+kAzHx8r66uDp6enrC3t0dwcDD6+/tN6snMzISvr69JW15eHtzd3U3aSktLoVAoYGdnBy8vLxQWFv5yjVKpVHicnJwgEokglUrh6OgIT09PNDQ0mPSvrq6GRCLBp0+fhDVUVVUhICAAdnZ28Pb2xt27d03GdHZ2IiwsDA4ODnBxccHu3bvx7t27X9ZERERE9COGUkRERGSxOjs7cf/+fdjY2AhtGo0GFy9eRHFxMbq6upCYmIjo6GghlElPT0d3dzfq6+uh1+tRVFSEhQsXzjj/ixcvEBkZifDwcDx+/BhqtRqpqamzrvPy5cvIyMjAiRMnoNfrkZWVhfT0dFRUVMxqHolEgh07dggh2jSdToetW7fC0dFRaEtOTsahQ4fQ3t4Of39/hIeH4/379wCAjx8/YuPGjVCpVHj06BEaGhrw+vVrbN++fdZrIyIiIsvF43tERERkUWpqauDg4IDJyUmMjY3BysoKBQUFAICxsTFkZWWhqakJ/v7+AIBly5ahubkZJSUlCAoKwuDgIFQqFfz8/ADgpx1N3ysqKsLy5cuRm5sLAFi5ciU6OjpmfVTw6NGjyM3NRWRkJABALpeju7sbJSUliImJmdVcarUaAQEBMBqNcHV1xZs3b1BXV4empiaTfgkJCYiKihLW0dDQgLKyMqSkpKCgoAAqlQpZWVlC/wsXLkAmk6G3txeenp6zqomIiIgsE0MpIiIisijBwcEoKirCyMgIzpw5A2trayF86evrw+joKLZs2WIyZnx8HCqVCgBw4MABREVFoa2tDSEhIYiIiEBAQMCM36XX67F+/XqTtumw63eNjIzAYDAgLi4Oe/fuFdonJyfh5OQ0q7kAYN26dVi9ejUqKiqQmpqKyspKLF26FIGBgb+s09raGn5+ftDr9QCAJ0+e4M6dOzPeTWUwGBhKERER0W9hKEVEREQWRSKRwMPDA8C/u3uUSiXKysoQFxeH4eFhAEBtbS0WLVpkMs7W1hYAEBYWhoGBAdTV1aGxsRGbNm1CfHw8Tp06Nad6rKys8O3bN5O2iYkJ4fV0TefPn/8p4BKLxXP6TrVaDa1Wi9TUVOh0OsTGxkIkEv32+OHhYYSHh8+448vV1XVONREREZHl4Z1SREREZLGsrKxw+PBhHDlyBJ8/f8aqVatga2uLwcFBeHh4mDwymUwY5+zsjJiYGFRWViIvLw/nzp2bcX6FQoGWlhaTtgcPHpi8d3Z2xqtXr0yCqe8vSndxcYGbmxueP3/+U01yuXxO646OjsbAwADy8/PR3d094xHA7+ucnJxEa2srFAoFAGDt2rXo6uqCu7v7TzVJJJI51URERESWh6EUERERWbRt27ZBLBZDq9XC0dERSUlJSExMREVFBQwGA9ra2nD27FnhUvGMjAzcvHkTfX196OrqQk1NjRDW/Gj//v149uwZkpOT0dPTgytXrqC8vNykz4YNG/D27Vvk5OTAYDBAq9Wivr7epM+xY8eg0WiQn5+P3t5edHR0QKfT4fTp03Na84IFCxAZGYnk5GSEhIRg8eLFP/XRarW4ceMGnj59ivj4eAwNDWHPnj0AgPj4eHz48AE7d+7Ew4cPYTAYcPv2bcTGxmJqampONREREZHlYShFREREFs3a2hoJCQnIycnByMgIjh8/jvT0dGg0GigUCoSGhqK2tlbYlWRjY4O0tDSsWbMGgYGBEIvFqKqqmnHuJUuW4Pr166iuroZSqURxcbHJ5eDAv7upCgsLodVqoVQq0dLSgqSkJJM+arUapaWl0Ol08PHxQVBQEMrLy+e8UwoA4uLiMD4+LgRNP8rOzkZ2djaUSiWam5tx69Yt4V8G3dzccO/ePUxNTSEkJAQ+Pj44ePAg5s+fDysr/rwkIiKi3yP69uMlBkRERET0v3fp0iUkJibi5cuXsLGxEdr7+/shl8vR3t4OX1/fv1cgERER/e/xonMiIiIiCzI6Ogqj0Yjs7Gzs27fPJJAiIiIiMifuryYiIiKyIDk5OfDy8oJUKkVaWtrfLoeIiIgsGI/vERERERERERGR2XGnFBERERERERERmR1DKSIiIiIiIiIiMjuGUkREREREREREZHYMpYiIiIiIiIiIyOwYShERERERERERkdkxlCIiIiIiIiIiIrNjKEVERERERERERGbHUIqIiIiIiIiIiMyOoRQREREREREREZndP091v4gI+72oAAAAAElFTkSuQmCC\n"
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of Unique Accession Numbers for Significant Data in AdjacentSeq_all__(Mar_Active)___(Torpor).tsv: 1062\n",
            "Number of Unique Accession Numbers for Not Significant Data in AdjacentSeq_all__(Mar_Active)___(Torpor).tsv: 2165\n",
            "Total number of modifications in AdjacentSeq_all__(Mar_Active)___(Torpor).tsv: 17793\n",
            "Significant N-terminal (Met-loss) modifications: 13\n",
            "Non-significant N-terminal (Met-loss) modifications: 30\n",
            "Processing complete.\n"
          ]
        }
      ],
      "source": [
        "#@title Plot of amino acid type BEFORE modification\n",
        "\n",
        "import os\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import numpy as np\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "def calculate_aa_percentages(fasta_dir):\n",
        "    aa_counts = {'A': 0, 'C': 0, 'D': 0, 'E': 0, 'F': 0, 'G': 0, 'H': 0, 'I': 0, 'K': 0, 'L': 0, 'M': 0, 'N': 0, 'P': 0, 'Q': 0, 'R': 0, 'S': 0, 'T': 0, 'V': 0, 'W': 0, 'Y': 0}\n",
        "    n_term_counts = {key: 0 for key in aa_counts}\n",
        "    c_term_counts = {key: 0 for key in aa_counts}\n",
        "    total_residues = 0\n",
        "\n",
        "    for filename in os.listdir(fasta_dir):\n",
        "        if filename.endswith('.fasta'):\n",
        "            with open(os.path.join(fasta_dir, filename), 'r') as f:\n",
        "                _ = f.readline()  # Skip header\n",
        "                sequence = f.readline().strip()\n",
        "                if sequence:\n",
        "                    total_residues += len(sequence)\n",
        "                    for aa in sequence:\n",
        "                        if aa in aa_counts:\n",
        "                            aa_counts[aa] += 1\n",
        "                    if sequence[0] in n_term_counts:\n",
        "                        n_term_counts[sequence[0]] += 1\n",
        "                    if sequence[-1] in c_term_counts:\n",
        "                        c_term_counts[sequence[-1]] += 1\n",
        "\n",
        "    aa_percentages = {key: (count / total_residues) * 100 for key, count in aa_counts.items()}\n",
        "    n_term_percentages = {key: (count / total_residues) * 100 for key, count in n_term_counts.items()}\n",
        "    c_term_percentages = {key: (count / total_residues) * 100 for key, count in c_term_counts.items()}\n",
        "\n",
        "    # Print statements for total residues, A residues, and percentage of A residues\n",
        "    print(f\"Total number of residues: {total_residues}\")\n",
        "    print(f\"Number of A residues: {aa_counts['A']}\")\n",
        "    print(f\"Percentage of A residues: {aa_percentages['A']}%\")\n",
        "\n",
        "    return aa_percentages, n_term_percentages, c_term_percentages\n",
        "\n",
        "def plot_last_column(file_path, log_file, aa_percentages, n_term_percentages, c_term_percentages):\n",
        "    significant_characters_count = {key: 0 for key in aa_percentages}\n",
        "    non_significant_characters_count = {key: 0 for key in aa_percentages}\n",
        "\n",
        "    # Include 'N-term' and 'C-term' in the dictionaries\n",
        "    significant_characters_count['N-term'] = 0\n",
        "    significant_characters_count['C-term'] = 0\n",
        "    non_significant_characters_count['N-term'] = 0\n",
        "    non_significant_characters_count['C-term'] = 0\n",
        "\n",
        "    unique_accession_numbers_significant = set()\n",
        "    unique_accession_numbers_non_significant = set()\n",
        "\n",
        "    total_modifications = 0\n",
        "    significant_modifications = 0\n",
        "    non_significant_modifications = 0\n",
        "\n",
        "    significant_met_loss_count = 0\n",
        "    non_significant_met_loss_count = 0\n",
        "\n",
        "    with open(file_path, 'r') as file:\n",
        "        # Skip header\n",
        "        header = next(file)\n",
        "        write_log(log_file, f\"Processing file: {file_path}\")\n",
        "\n",
        "        for line in file:\n",
        "            line = line.strip()\n",
        "            if line:\n",
        "                columns = line.split('\\t')\n",
        "                if len(columns) != 6:  # Ensure the line has exactly 6 columns\n",
        "                    write_log(log_file, f\"Skipping malformed line: {line}\")\n",
        "                    continue\n",
        "\n",
        "                accession_number = columns[0]\n",
        "                p_value = float(columns[3])\n",
        "                last_column = columns[4]  # Ensure the modifications are in the correct column\n",
        "                sequence_column = columns[5]  # Ensure the sequence is in the correct column\n",
        "\n",
        "                total_modifications += 1\n",
        "\n",
        "                is_n_term = last_column.startswith('Met-loss')\n",
        "                if is_n_term:\n",
        "                    character = 'N-term'\n",
        "                else:\n",
        "                    if len(sequence_column) >= 7:\n",
        "                        character = sequence_column[6]\n",
        "                    else:\n",
        "                        write_log(log_file, f\"Skipping sequence due to insufficient length: {sequence_column}\")\n",
        "                        continue\n",
        "\n",
        "                if p_value < 0.05:\n",
        "                    significant_modifications += 1\n",
        "                    unique_accession_numbers_significant.add(accession_number)\n",
        "                    if character in significant_characters_count:\n",
        "                        significant_characters_count[character] += 1\n",
        "                    if is_n_term:\n",
        "                        significant_met_loss_count += 1\n",
        "                else:\n",
        "                    non_significant_modifications += 1\n",
        "                    unique_accession_numbers_non_significant.add(accession_number)\n",
        "                    if character in non_significant_characters_count:\n",
        "                        non_significant_characters_count[character] += 1\n",
        "                    if is_n_term:\n",
        "                        non_significant_met_loss_count += 1\n",
        "\n",
        "    # Calculate percentages for significant and non-significant modifications\n",
        "    significant_characters_percent = {key: (value / significant_modifications) * 100 if significant_modifications > 0 else 0 for key, value in significant_characters_count.items()}\n",
        "    non_significant_characters_percent = {key: (value / non_significant_modifications) * 100 if non_significant_modifications > 0 else 0 for key, value in non_significant_characters_count.items()}\n",
        "\n",
        "    # Plot data\n",
        "    categories = list(significant_characters_percent.keys())\n",
        "    bar_width = 0.25\n",
        "    x = np.arange(len(categories))\n",
        "    fig, ax1 = plt.subplots(figsize=(12, 8))\n",
        "\n",
        "    ax1.bar(x - bar_width, [aa_percentages.get(cat, 0) for cat in categories], bar_width, color='black', label='All Data')\n",
        "    ax1.bar(x, [significant_characters_percent.get(cat, 0) for cat in categories], bar_width, color='red', label='Significant (p < 0.05)')\n",
        "    ax1.bar(x + bar_width, [non_significant_characters_percent.get(cat, 0) for cat in categories], bar_width, color='skyblue', label='Not Significant (p >= 0.05)')\n",
        "\n",
        "    ax1.set_ylabel('Percentage')\n",
        "    ax1.set_xticks(x)\n",
        "    ax1.set_xticklabels(categories)\n",
        "    ax1.set_xlabel('Residue Type')\n",
        "    ax1.set_title(f'Residue before modification - {os.path.basename(file_path)}')\n",
        "    ax1.set_ylim(0, 35)\n",
        "\n",
        "    plt.xticks(rotation=90)\n",
        "    plt.tight_layout()\n",
        "\n",
        "    # Save plot\n",
        "    output_dir = 'output/plots/adjacent_aa'\n",
        "    if not os.path.exists(output_dir):\n",
        "        os.makedirs(output_dir)\n",
        "    output_filename = os.path.basename(file_path).replace('.tsv', '')\n",
        "    plt.savefig(os.path.join(output_dir, f'AA_beforemod_{output_filename}.eps'))      # Saves as eps format image\n",
        "\n",
        "    plt.show()\n",
        "\n",
        "    # Log number of unique accession numbers and modifications\n",
        "    write_log(log_file, f\"Number of Unique Accession Numbers for Significant Data in {os.path.basename(file_path)}: {len(unique_accession_numbers_significant)}\")\n",
        "    write_log(log_file, f\"Number of Unique Accession Numbers for Not Significant Data in {os.path.basename(file_path)}: {len(unique_accession_numbers_non_significant)}\")\n",
        "    write_log(log_file, f\"Total number of modifications in {os.path.basename(file_path)}: {total_modifications}\")\n",
        "\n",
        "    # Print statements for Met-loss modifications\n",
        "    print(f\"Significant N-terminal (Met-loss) modifications: {significant_met_loss_count}\")\n",
        "    print(f\"Non-significant N-terminal (Met-loss) modifications: {non_significant_met_loss_count}\")\n",
        "\n",
        "# Directory containing the files\n",
        "directory = '/content/output/modlist_files'\n",
        "fasta_dir = '/content/output/libraries/all_protein_seq'\n",
        "logs_directory = os.path.join('output/logs')\n",
        "os.makedirs(logs_directory, exist_ok=True)\n",
        "log_file = os.path.join(logs_directory, \"AA_before_mod.log\")\n",
        "\n",
        "# Calculate amino acid percentages for all protein residues\n",
        "aa_percentages, n_term_percentages, c_term_percentages = calculate_aa_percentages(fasta_dir)\n",
        "\n",
        "# Process each file\n",
        "for filename in os.listdir(directory):\n",
        "    if filename.startswith('AdjacentSeq') and filename.endswith('.tsv'):\n",
        "        file_path = os.path.join(directory, filename)\n",
        "        plot_last_column(file_path, log_file, aa_percentages, n_term_percentages, c_term_percentages)\n",
        "\n",
        "write_log(log_file, \"Processing complete.\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "cellView": "form",
        "id": "-MBF_20SXeFf",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "9cfe74b1-dcf7-415d-8c18-e4eaac5a374d"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Total number of residues: 2697361\n",
            "Number of A residues: 187384\n",
            "Percentage of A residues: 6.946938136941997%\n",
            "Processing file: /content/output/modlist_files/AdjacentSeq_all__(Aug_Active)___(Mar_Active).tsv\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 1200x800 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAABKUAAAMWCAYAAAAgRDUeAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAB6NUlEQVR4nOzdd5wU9f0/8PfSDqRKE1BEwAKiqMHeAEWaDcVewa4Yo8aoxETEqERN1GhssYAFYsWCiSA27MaCYokFA4IKKCJVOMrN7w9/7Nfj7uDuOGbv8Pl8PPahOzO7+2Jvdnb3tTOfySRJkgQAAAAApKhargMAAAAA8MujlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAL4mcsuuywymUypls1kMnHZZZet20Br6fPPP4+ePXtGw4YNI5PJxOOPP57rSDlR3N91s802iwEDBhSaVtzzNWLEiMhkMjF16tT0Av9/VWEdqwjFPcfdunWLbt265SwTRa36mnnxxRcjk8nEiy++WOGPtXDhwmjevHmMHDmywu97fbGutw9jx46NevXqxXfffbfOHqM0+vbtG6eeempOM1RW63o7+f3330fdunXj3//+9zp7DAClFFAprfySuvJSo0aN2HjjjWPAgAHx9ddf5zpelXHiiSfGBx98EFdeeWXcd999seOOO8aoUaPihhtuyHW0Sqm452td+/e//73eFE+33HJLZDKZ2GWXXXIdJXXffPNNXHbZZfHee+8VO3/MmDHRtWvXaN68eWywwQbRrl27OOKII2Ls2LHpBq0i/va3v0X9+vXjqKOOKnb+hRdeGJlMJo488siUk5XezjvvHJlMJm699dZy30cutw+9e/eOzTffPIYNG5aTx4+IePXVV+OZZ56Jiy66KDttZRmayWTi/vvvL/Z2e+yxR2Qymdhmm23SilqsilhPP/7447jsssty8sNIkyZN4pRTTok//vGPpVo+l1mBKiwBqISGDx+eRERy+eWXJ/fdd19yxx13JCeffHJSvXr1pH379snixYvXyeMuW7as1PcdEcmQIUPWSY6K8OOPPyYRkVxyySWFpu+///5JmzZtchMqR4YMGZKs+pa3ZMmSZOnSpdnrJT1fy5cvTxYvXpwUFBSsk2yDBg0qkm2lxYsXJ8uWLVsnj7su7L777slmm22WRETy+eefl/p2K1/vU6ZMyU7Lz89P8vPz10HKdeOtt95KIiIZPnx4kXnXXnttEhFJ165dk+uuuy657bbbkgsuuCDZfvvtkxNPPDH1rOXVpk2bQnlfeOGFJCKSF154oUIfZ+nSpUmzZs2Sq666qtj5BQUFySabbJJsttlmSZ06dZL58+dX6ONXhM8++yyJiGSzzTZL9thjj3LfT663D7fcckuywQYb5Ow5Pvjgg5OePXsWmrZyvatdu3bSp0+fIreZMmVKdn6nTp3SilpERa2nDz/8cImvszS2kx9//HESEclzzz23xmVXlxWgJPaUAiq1Pn36xHHHHRennHJK3HnnnXHBBRfEF198EU8++eQ6ebwaNWpE7dq118l9p23lIReNGjVa549VUFAQS5YsWeePU5Hy8vKiZs2a2eslPV/Vq1eP2rVrl/qwzopUu3btqFGjRuqPWx5TpkyJ1157La677rpo1qzZWh92VatWrahVq1YFpcud5cuXx5/+9KfYb7/94sUXX4zzzjsvTj/99Lj22mtj4sSJcc011+Q6YqXz1FNPxXfffRdHHHFEsfNffPHF+Oqrr+Luu++O5cuXx+jRo1NOuGb3339/NG/ePP7617/Ga6+9tk72HElj+9C/f//Iz8+Phx9+eJ0+TnG+/fbb+Ne//lXietC3b98YP358zJ49u9D0UaNGxUYbbVShe7omSRKLFy8u023SWE/T2E527NgxttlmmxgxYsQ6fRzgl0spBVQpe+21V0REfPHFF4Wmf/LJJ3HYYYdF48aNo3bt2rHjjjsWKa6WLVsWQ4cOjS222CJq164dTZo0iT333DPGjx+fXaa4sYfy8/PjvPPOi2bNmkX9+vXjoIMOiq+++qpItgEDBsRmm21WZHpJ41Tdf//90aVLl6hTp040btw4jjrqqJg+ffoan4Mvv/wyzjrrrNhqq62iTp060aRJkzj88MMLfem57LLLok2bNhER8bvf/S4ymUxsttlm0a1bt/jXv/4VX375Zfbwh59nzs/PjyFDhsTmm28eeXl50bp167jwwgsjPz+/UIZMJhNnn312jBw5Mjp16hR5eXmrPQxps802iwMOOCBefPHF2HHHHaNOnTqx7bbbZseiGT16dGy77bZRu3bt6NKlS0ycOLHIfTz//POx1157Rd26daNRo0Zx8MEHx3//+98iy73yyiux0047Re3ataN9+/Zx++23l5hp5fg4JT1fEcWPdxQR8fTTT0fXrl2jfv360aBBg9hpp51i1KhR2fkvv/xyHH744bHppptmn8vzzjuv0BebAQMGxM0335x9Tldefv48r3rozsSJE6NPnz7RoEGDqFevXuy7777xxhtvFFpmZeZXX301zj///GjWrFnUrVs3DjnkkHU2PszIkSNjww03jP333z8OO+ywEkupjz76KPbZZ5+oU6dObLLJJnHFFVdEQUFBkeVWHStl6dKlcemll0aXLl2iYcOGUbdu3dhrr73ihRdeKHLbgoKC+Nvf/pZdp5o1axa9e/eOt99+u9BypXkNduvWLbbZZpv4+OOPo3v37rHBBhvExhtvXKhMevHFF2OnnXaKiIiBAwdm/44jRoyI2bNnx/z582OPPfYo9vlo3rx5oeulfQ2WtF0qzzhDf/nLX2L33XePJk2aRJ06daJLly7xyCOPlOk+KtLjjz8em222WbRv377Y+SNHjoytt946unfvHj169Ch2XSvpdVvSOFg333xztGvXLurUqRM777xzvPzyy2s1Xs+oUaPisMMOiwMOOCAaNmxYaNvwc2+++Wb07ds3Ntxww6hbt2507tw5/va3v0VE2bYPjzzySGQymZgwYUKRx7j99tsjk8nEhx9+mJ1WmvfMiJ/Wz86dO8cTTzxRrudhbfzrX/+K5cuXR48ePYqdf/DBB0deXl6RwmzUqFFxxBFHRPXq1YvcZvjw4bHPPvtE8+bNIy8vL7beeutiD69c+Z41bty47HtWSe8lJSnNehoR8fXXX8fJJ58crVq1iry8vGjbtm2ceeaZsXTp0hgxYkQcfvjhERHRvXv37Dqwcv39+To6a9asqFGjRgwdOrTIY3z66aeRyWTi73//e3ba3Llz49xzz43WrVtHXl5ebL755nH11VcXuz3eb7/9YsyYMZEkSYn/3jVlffvtt6NXr17RtGnTqFOnTrRt2zZOOumkiPjp81njxo1j4MCBRe53/vz5Ubt27bjgggtKfGygaqsaP78C/H8rv2BsuOGG2WkfffRR7LHHHrHxxhvHxRdfHHXr1o2HHnoo+vXrF48++mgccsghEfFT8TBs2LA45ZRTYuedd4758+fH22+/He+++27st99+JT7mKaecEvfff38cc8wxsfvuu8fzzz8f+++//1r9O6688sr44x//GEcccUSccsop8d1338VNN90Ue++9d0ycOHG1eze99dZb8dprr8VRRx0Vm2yySUydOjVuvfXW6NatW3z88cexwQYbxKGHHhqNGjWK8847L44++ujo27dv1KtXL+rWrRvz5s2Lr776Kq6//vqIiKhXr15E/PRF/qCDDopXXnklTjvttOjYsWN88MEHcf3118dnn31WZJD0559/Ph566KE4++yzo2nTpsUWcj83efLkOOaYY+L000+P4447Lv7yl7/EgQceGLfddlv8/ve/j7POOisiIoYNGxZHHHFEfPrpp1Gt2k+/nTz77LPRp0+faNeuXVx22WWxePHiuOmmm2KPPfaId999N/vYH3zwQfTs2TOaNWsWl112WSxfvjyGDBkSG2200WqzlfR8lWTEiBFx0kknRadOnWLw4MHRqFGjmDhxYowdOzaOOeaYiIh4+OGH48cff4wzzzwzmjRpEv/5z3/ipptuiq+++ir7Jer000+Pb775JsaPHx/33XffajNG/LSu77XXXtGgQYO48MILo2bNmnH77bdHt27dYsKECUXGcvr1r38dG264YQwZMiSmTp0aN9xwQ5x99tnx4IMPrvGxymrkyJFx6KGHRq1ateLoo4+OW2+9Nd56661sWRMRMXPmzOjevXssX748+1r9xz/+EXXq1Fnj/c+fPz/uvPPOOProo+PUU0+NBQsWxF133RW9evWK//znP7H99ttnlz355JNjxIgR0adPnzjllFNi+fLl8fLLL8cbb7yR3XuiLK/BH374IXr37h2HHnpoHHHEEfHII4/ERRddFNtuu2306dMnOnbsGJdffnlceumlcdppp2XL89133z2aN28ederUiTFjxsSvf/3raNy4cYn/xrK8Bityu/S3v/0tDjrooDj22GNj6dKl8cADD8Thhx8eTz311Fpv68rjtddei1/96lfFzsvPz49HH300fvvb30ZExNFHHx0DBw6MmTNnRosWLcr1eLfeemucffbZsddee8V5550XU6dOjX79+sWGG24Ym2yySZnv780334zJkyfH8OHDo1atWnHooYfGyJEj4/e//32h5caPHx8HHHBAtGzZMn7zm99EixYt4r///W889dRT8Zvf/KZM24f9998/6tWrFw899FB07dq10LwHH3wwOnXqlB1fqbTvmSt16dIlJyfJeO2116JJkybZHwxWtcEGG8TBBx8c//znP+PMM8+MiIj3338/Pvroo7jzzjtj0qRJRW5z6623RqdOneKggw6KGjVqxJgxY+Kss86KgoKCGDRoUKFlP/300zj66KPj9NNPj1NPPTW22mqrUmcv7Xr6zTffxM477xxz586N0047LTp06BBff/11PPLII/Hjjz/G3nvvHeecc07ceOON8fvf/z46duwYEZH9789ttNFG0bVr13jooYdiyJAhheY9+OCDUb169Wxp9OOPP0bXrl3j66+/jtNPPz023XTTeO2112Lw4MExY8aMIuNOdunSJa6//vr46KOPShyna3VZv/322+x788UXXxyNGjWKqVOnZvceq1mzZhxyyCExevTouP322wvt/fX4449Hfn5+iePLAeuBXB8/CFCclWPMPPvss8l3332XTJ8+PXnkkUeSZs2aJXl5ecn06dOzy+67777JtttumyxZsiQ7raCgINl9992TLbbYIjttu+22S/bff//VPu6qYw+99957SUQkZ511VqHljjnmmCJjSp144onFjtW06n1OnTo1qV69enLllVcWWu6DDz5IatSoUWT6qn788cci015//fUkIpJ77703O23luBrXXnttoWVLGlPqvvvuS6pVq5a8/PLLhabfdtttSUQkr776anZaRCTVqlVLPvroo9VmXalNmzZJRCSvvfZadtq4ceOSiEjq1KmTfPnll9npt99+e5ExKbbffvukefPmyffff5+d9v777yfVqlVLTjjhhOy0fv36JbVr1y50fx9//HFSvXr1IuOyrDo+TknP16rjHc2dOzepX79+sssuuxQZf+zn404V93caNmxYkslkCuVb3Zgxq65j/fr1S2rVqpV88cUX2WnffPNNUr9+/WTvvfcukrlHjx6FMp133nlJ9erVk7lz5xb7eOX19ttvJxGRjB8/PkmS/xtL5Te/+U2h5c4999wkIpI333wzO+3bb79NGjZsWGRMqa5duyZdu3bNXl++fHmRsVN++OGHZKONNkpOOumk7LTnn38+iYjknHPOKZJz5XNRltdg165di7y28vPzkxYtWiT9+/fPTlvdmFKXXnppEhFJ3bp1kz59+iRXXnll8s477xRZrrSvwbJsl0pj1XV16dKlyTbbbJPss88+haanMabUsmXLkkwmk/z2t78tdv4jjzxSaMyy+fPnJ7Vr106uv/76QssVN05ZcZnz8/OTJk2aJDvttFOh8ZlGjBiRHQesrM4+++ykdevW2fXtmWeeSSIimThxYnaZ5cuXJ23btk3atGmT/PDDD4Vu//PXbFm2D0cffXTSvHnzZPny5dlpM2bMSKpVq5Zcfvnl2Wmlfc9c6aqrrkoiIpk1a1ap/v0VZc8990y6dOlSZPrKv+HDDz+cPPXUU0kmk0mmTZuWJEmS/O53v0vatWuXJMlPr91Vx5Qqbrvcq1ev7G1WWvmeNXbs2HJlL+16esIJJyTVqlVL3nrrrSL3sXI9WN04TatuJ1e+f37wwQeFltt6660LvZ7/9Kc/JXXr1k0+++yzQstdfPHFSfXq1bPP50qvvfZaEhHJgw8+uNp/d0lZH3vssSQiiv13rrTyM8GYMWMKTe/bt2+Rvw+wfnH4HlCp9ejRI5o1axatW7eOww47LOrWrRtPPvlk9tfrOXPmxPPPPx9HHHFELFiwIGbPnh2zZ8+O77//Pnr16hWff/559mx9jRo1io8++ig+//zzUj/+ytMgn3POOYWmn3vuueX+N40ePToKCgriiCOOyOadPXt2tGjRIrbYYotiD0f6uZ/vVbJs2bL4/vvvY/PNN49GjRrFu+++W+5cDz/8cHTs2DE6dOhQKNc+++wTEVEkV9euXWPrrbcu9f1vvfXWsdtuu2Wvr9yrZ5999olNN920yPT//e9/ERExY8aMeO+992LAgAGF9jLp3Llz7Lffftm/0YoVK2LcuHHRr1+/QvfXsWPH6NWrV6lzrsn48eNjwYIFcfHFFxcZf+znh9b8/O+0aNGimD17duy+++6RJEmxhyeuyYoVK+KZZ56Jfv36Rbt27bLTW7ZsGcccc0y88sorMX/+/EK3Oe200wpl2muvvWLFihXx5ZdflvnxV2fkyJGx0UYbRffu3SMismebeuCBB2LFihXZ5f7973/HrrvuGjvvvHN2WrNmzeLYY49d42NUr149++t5QUFBzJkzJ5YvXx477rhjofX+0UcfjUwmU2RPgZW5Isr+GqxXr14cd9xx2eu1atWKnXfeObuOrsnQoUNj1KhRscMOO8S4cePikksuiS5dusSvfvWrQoeglvY1WNHbpZ+vqz/88EPMmzcv9tprr7XanpTXnDlzIkmSQnvD/tzIkSNjxx13jM033zwiIurXrx/7779/uccwe/vtt+P777+PU089tdD4TMcee2yJGVZn+fLl8eCDD8aRRx6ZXd9WHi7284wTJ06MKVOmxLnnnltkz9jyjl935JFHxrffflvo0MRHHnkkCgoKsmd/K8t75korn4dVx25a177//vs1/g169uwZjRs3jgceeCCSJIkHHnggjj766BKX//m6Pm/evJg9e3Z07do1/ve//8W8efMKLdu2bdtyv3eUZj0tKCiIxx9/PA488MBix78qz3pw6KGHRo0aNQrtDfvhhx/Gxx9/XOgMgA8//HDstddeseGGGxba1vTo0SNWrFgRL730UqH7Xdt1YOU6/tRTT8WyZcuKXWafffaJpk2bFsr+ww8/xPjx4yv1WTaBtaeUAiq1m2++OcaPHx+PPPJI9O3bN2bPnh15eXnZ+ZMnT44kSeKPf/xjNGvWrNBl5ZfSb7/9NiIiLr/88pg7d25sueWWse2228bvfve7Ynfv/7kvv/wyqlWrVmRsk7Lsxr+qzz//PJIkiS222KJI5v/+97/ZvCVZvHhxXHrppdlxIJo2bRrNmjWLuXPnFvlQXdZcH330UZFMW265ZUREkVxt27Yt0/3/vCiKiGjYsGFERLRu3brY6T/88ENERLZAKe4579ixY8yePTsWLVoU3333XSxevDi22GKLIsutzd9rVSvHM1vTqcanTZuWLdLq1asXzZo1yx5WU56/03fffRc//vhjic9DQUFBkfGQVn3OV36xWPncFmfx4sUxc+bMQpfVWbFiRTzwwAPRvXv3mDJlSkyePDkmT54cu+yyS8yaNSuee+657LJffvnlWv197rnnnujcuXN2TLhmzZrFv/71r0LP5xdffBGtWrVa7WFyZX0NbrLJJkW+IG644YarfR5XdfTRR8fLL78cP/zwQzzzzDNxzDHHxMSJE+PAAw/MniSgtK/Bit4uPfXUU7HrrrtG7dq1o3HjxtGsWbO49dZb12p7sraSYsaumTt3bvz73/+Orl27ZtezyZMnxx577BFvv/12fPbZZ2V+nJXbl5XlwUo1atRY4yHJxXnmmWfiu+++i5133jmbb8qUKdG9e/f45z//mR2vp7TbkbLo3bt3NGzYsNCX+gcffDC233777DpUlvfMlVb+LXJxsofi1oOfq1mzZhx++OExatSoeOmll2L69OnZQ6iL8+qrr0aPHj2yYxM2a9Yse1hlcaVUeZR2Pf3uu+9i/vz5FboONG3aNPbdd9946KGHstMefPDBqFGjRhx66KHZaZ9//nmMHTu2yDqwcvyuil4HunbtGv3794+hQ4dG06ZN4+CDD47hw4cXGiuvRo0a0b9//3jiiSey00ePHh3Lli1TSsF6zphSQKW28847Z39B7NevX+y5555xzDHHxKeffhr16tXLfsC/4IILSvxFc+WXjb333ju++OKLeOKJJ+KZZ56JO++8M66//vq47bbb4pRTTlnrrCV9WPv5niIRP/06mslk4umnny52INbVjWUU8dM4QcOHD49zzz03dtttt2jYsGFkMpk46qijih2gtLQKCgpi2223jeuuu67Y+auWR6UZB+jnivu3rm76mr6MVGYrVqyI/fbbL+bMmRMXXXRRdOjQIerWrRtff/11DBgwYK3+TmVRnuf2wQcfLDLY7OqWf/7552PGjBnxwAMPxAMPPFBk/siRI6Nnz56lTFyy+++/PwYMGBD9+vWL3/3ud9G8efOoXr16DBs2rMiJD9akrK/BilxHGzRoEPvtt1/st99+UbNmzbjnnnvizTffjK5du5b5NVgRXn755TjooINi7733jltuuSVatmwZNWvWjOHDh5c4OPe61Lhx48hkMsUWfg8//HDk5+fHX//61/jrX/9aZP7IkSOzgzyXdntc0VbuCVPSGeMmTJiQ3aOwouXl5UW/fv3isccei1tuuSVmzZoVr776alx11VXZZcrynrnSyr9F06ZN10nukjRp0qRUxe8xxxwTt912W1x22WWx3XbblbgH7xdffBH77rtvdOjQIa677rpo3bp11KpVK/7973/H9ddfX2S7XNb3uJXKsp6uC0cddVQMHDgw3nvvvdh+++3joYcein333bfQ36+goCD222+/uPDCC4u9j5Ul5kpruw5kMpl45JFH4o033ogxY8bEuHHj4qSTToq//vWv8cYbb2S3uUcddVTcfvvt8fTTT0e/fv3ioYceig4dOsR2221XrscFqgalFFBlrPwC2r179/j73/8eF198cfYwppo1a5Z4hp6fW3l2l4EDB8bChQtj7733jssuu6zEUqpNmzZRUFAQX3zxRaG9ED799NMiy2644YYxd+7cItNXPVSqffv2kSRJtG3btsgHv9J45JFH4sQTTyz0YXfJkiXFPnZxSvqy1r59+3j//fdj3333zckv4iVZOchtcc/5J598Ek2bNo26detG7dq1o06dOsUenlncbctr5d4pH374YZEvbyt98MEH8dlnn8U999wTJ5xwQnb6z8/0uFJpn+tmzZrFBhtsUOLzUK1atQopLXr16lVszpKMHDkymjdvnj1L2M+NHj06HnvssbjtttuiTp060aZNm3L/fR555JFo165djB49utBztupheu3bt49x48bFnDlzStxbam1fg8Upz2tmxx13jHvuuSdmzJiRzVWa12BZtktr8uijj0bt2rVj3LhxhfZCHT58eJnvqyLUqFEj2rdvH1OmTCkyb+TIkbHNNtsUe2jm7bffHqNGjcp+2V+5V+Cq28VVt8crty+TJ08uVBYtX748pk6dGp07dy519kWLFsUTTzwRRx55ZBx22GFF5p9zzjkxcuTI6N69e6HtyOreu8q6Xh155JFxzz33xHPPPRf//e9/I0mSQnuZlPU9MyJiypQp2T1y09ShQ4d49NFH17jcnnvuGZtuumm8+OKLcfXVV5e43JgxYyI/Pz+efPLJQnuRrumQ+bIq7XrarFmzaNCgQaGzIhanrOtAv3794vTTT8/uMffZZ5/F4MGDCy3Tvn37WLhwYZnWgYjiB1gvS9Zdd901dt1117jyyitj1KhRceyxx8YDDzyQ/Qy29957R8uWLePBBx+MPffcM55//vm45JJLSpURqLocvgdUKd26dYudd945brjhhliyZEk0b948unXrFrfffnv2i93Pfffdd9n///777wvNq1evXmy++eZFTrX+c3369ImIiBtvvLHQ9FXPTBPx04e8efPmFTokcMaMGfHYY48VWu7QQw+N6tWrx9ChQ4vsaZEkSZGcq6pevXqR2910002l3gNg5Rn4VnXEEUfE119/HXfccUeReYsXL45FixaV6v4rWsuWLWP77bePe+65p9AXzA8//DCeeeaZ6Nu3b0T89Lz06tUrHn/88Zg2bVp2uf/+978xbty4CsvTs2fPqF+/fgwbNix72NVKK/8uK/es+fnfKUmS7Knef65u3boRUfTL86qqV68ePXv2jCeeeKLQae5nzZoVo0aNij333DMaNGhQnn9SIS1btowePXoUupRk8eLFMXr06DjggAPisMMOK3I5++yzY8GCBdlTzfft2zfeeOON+M9//pO9j++++65U4wEV95y++eab8frrrxdarn///pEkSbF7Iqy87dq+BotT0t/xxx9/LJJxpaeffjoi/u+wu9K+BsuyXVqT6tWrRyaTKbT9mDp1ak7OtrbSbrvtFm+//XahadOnT4+XXnopjjjiiGLXtYEDB8bkyZPjzTffjIj/K49/PjbOihUr4h//+Eeh+91xxx2jSZMmcccdd8Ty5cuz00eOHFmmwzMjIh577LFYtGhRDBo0qNiMBxxwQDz66KORn58fv/rVr6Jt27Zxww03FFlnfr5Olnb7sFKPHj2icePG8eCDD8aDDz4YO++8c6HD0MrynrnSO++8U2g8wLTstttu8cMPP6xx7LZMJhM33nhjDBkyJI4//vgSlytuGzJv3rwKLWDLsp5Wq1Yt+vXrF2PGjCmyvv88Z1nXgUaNGkWvXr3ioYceigceeCBq1aoV/fr1K7TMEUccEa+//nqx741z584t9FqI+GkdaNiwYXTq1Ck7bd68efHJJ58U+jxRUtYffvihyLZ25RlTf/4ZrFq1anHYYYfFmDFj4r777ovly5c7dA9+AewpBVQ5v/vd7+Lwww+PESNGxBlnnBE333xz7LnnnrHtttvGqaeeGu3atYtZs2bF66+/Hl999VW8//77EfHTQNvdunWLLl26ROPGjePtt9+ORx55JM4+++wSH2v77bePo48+Om655ZaYN29e7L777vHcc8/F5MmTiyx71FFHxUUXXRSHHHJInHPOOfHjjz/GrbfeGltuuWWhAYPbt28fV1xxRQwePDh76vH69evHlClT4rHHHovTTjstLrjgghIzHXDAAXHfffdFw4YNY+utt47XX389nn322WjSpEmpnr8uXbrEgw8+GOeff37stNNOUa9evTjwwAPj+OOPj4ceeijOOOOMeOGFF2KPPfaIFStWxCeffBIPPfRQjBs3rtjBWNNw7bXXRp8+fWK33XaLk08+ORYvXhw33XRTNGzYMC677LLsckOHDo2xY8fGXnvtFWeddVYsX748brrppujUqdMaxw8rrQYNGsT1118fp5xySuy0005xzDHHxIYbbhjvv/9+/Pjjj3HPPfdEhw4don379nHBBRfE119/HQ0aNIhHH3202C+5Xbp0iYif9qLo1atXVK9evcRTX19xxRUxfvz42HPPPeOss86KGjVqxO233x75+flxzTXXVMi/ryyefPLJWLBgQRx00EHFzt91112jWbNmMXLkyDjyyCPjwgsvjPvuuy969+4dv/nNb6Ju3brxj3/8I9q0abPGv88BBxwQo0ePjkMOOST233//mDJlStx2222x9dZbx8KFC7PLde/ePY4//vi48cYb4/PPP4/evXtHQUFBvPzyy9G9e/c4++yz1/o1WJz27dtHo0aN4rbbbov69etH3bp1Y5dddon69evH7rvvHrvuumv07t07WrduHXPnzo3HH388Xn755ejXr1/ssMMOERGlfg2WZbu0Jvvvv39cd9110bt37zjmmGPi22+/jZtvvjk233zzCnvNlNXBBx8c9913X3z22WfZPdlGjRoVSZKUuK717ds3atSoESNHjoxddtklOnXqFLvuumsMHjw4u9fcAw88UOTLdq1ateKyyy6LX//617HPPvvEEUccEVOnTo0RI0ZE+/bty7SXysiRI6NJkyax++67Fzv/oIMOijvuuCP+9a9/xaGHHhq33nprHHjggbH99tvHwIEDo2XLlvHJJ5/ERx99lC0LyrJ9iPhpD6hDDz00HnjggVi0aFH85S9/KbJMad8zI34aW2jSpEkxaNCgUj8PFWX//fePGjVqxLPPPhunnXbaapc9+OCD4+CDD17tMj179oxatWrFgQceGKeffnosXLgw7rjjjmjevHmxBV15lHU9veqqq+KZZ56Jrl27xmmnnRYdO3aMGTNmxMMPPxyvvPJKNGrUKLbffvuoXr16XH311TFv3rzIy8vLDp5fkiOPPDKOO+64uOWWW6JXr15FBtP/3e9+F08++WQccMABMWDAgOjSpUssWrQoPvjgg3jkkUdi6tSphQ7VGz9+fBx44IGFXg+PPfZYDBw4MIYPHx4DBgyIiCgx66hRo+KWW26JQw45JNq3bx8LFiyIO+64Ixo0aJD9Yenn2W+66aYYMmRIbLvttmvcOwtYD6zjs/sBlMvK03kXd/rgFStWJO3bt0/at2+fPfX1F198kZxwwglJixYtkpo1ayYbb7xxcsABBySPPPJI9nZXXHFFsvPOOyeNGjVK6tSpk3To0CG58sork6VLl2aXGTJkSJHTby9evDg555xzkiZNmiR169ZNDjzwwGT69OnFnnr9mWeeSbbZZpukVq1ayVZbbZXcf//9xd5nkiTJo48+muy5555J3bp1k7p16yYdOnRIBg0alHz66aerfW5++OGHZODAgUnTpk2TevXqJb169Uo++eSTIqdrnzJlShIRybXXXlvo9gsXLkyOOeaYpFGjRklEJG3atMnOW7p0aXL11VcnnTp1SvLy8pINN9ww6dKlSzJ06NBk3rx52eUiIhk0aNBqc/5cmzZtkv3337/I9OLup6Tczz77bLLHHnskderUSRo0aJAceOCByccff1zkPidMmJB06dIlqVWrVtKuXbvktttuK/ZvUNrnq6RTyz/55JPJ7rvvns2z8847J//85z+z8z/++OOkR48eSb169ZKmTZsmp556avL+++8nEZEMHz48u9zy5cuTX//610mzZs2STCZTKGdx69i7776b9OrVK6lXr16ywQYbJN27d09ee+21YjOv+vpZeSr14k4tXh4HHnhgUrt27WTRokUlLjNgwICkZs2ayezZs5MkSZJJkyYlXbt2TWrXrp1svPHGyZ/+9KfkrrvuKvIcr3qq84KCguSqq65K2rRpk+Tl5SU77LBD8tRTTyUnnnhioXU4SX56Tq+99tqkQ4cOSa1atZJmzZolffr0Sd55551Cy5XmNVjcaeWTJCn2cZ944olk6623TmrUqJH9Oy9btiy54447kn79+mWzb7DBBskOO+yQXHvttUl+fn6h+yjta7As26U1ueuuu5ItttgiycvLSzp06JAMHz68VK+Zil6fVsrPz0+aNm2a/OlPf8pO23bbbZNNN910tbfr1q1b0rx582TZsmVJkvz0vtCjR48kLy8v2WijjZLf//73yfjx44vNfOONN2b/PjvvvHPy6quvJl26dEl69+5dqsyzZs1KatSokRx//PElLvPjjz8mG2ywQXLIIYdkp73yyivJfvvtl9SvXz+pW7du0rlz5+Smm27Kzi/r9iFJkuy/MZPJJNOnTy82S2neM5MkSW699dZkgw02SObPn1+q56GiHXTQQcm+++5baNrK9e7hhx9e7W2Le+0++eSTSefOnZPatWsnm222WXL11Vcnd999d5HtT0nvWWtSnvX0yy+/TE444YSkWbNmSV5eXtKuXbtk0KBBhbYNd9xxR9KuXbukevXqhdbfVbeTK82fPz+pU6dOEhHJ/fffX2yOBQsWJIMHD04233zzpFatWknTpk2T3XffPfnLX/5S6HPRf//73yQikmeffbbQ7Ve+z/z8/aykrO+++25y9NFHJ5tuummSl5eXNG/ePDnggAOSt99+u0iugoKCpHXr1klEJFdcccVqn0tg/ZBJkio8kiwAsF7aa6+9Ii8vL5599tlcR6lSMplMDBkypNAehFXRn/70pxg+fHh8/vnnJQ40vy4VFBREs2bN4tBDDy32cMpfih122CG6desW119/fU4e/+WXX45u3brFJ598UuyZO1n3zj333HjppZfinXfeqVTjTQLrD2NKAQCVzowZM1I/2xeVx3nnnRcLFy4s9oyOFW3JkiVFxru59957Y86cOdGtW7d1/viV1dixY+Pzzz8vMkh2mvbaa6/o2bNnTg5P5qexOO+888644oorFFLAOmNMKQCg0njttddi9OjR8cUXX8RFF12U6zhV1ooVK4odtPrn6tWrlz0V+9pavHhxsSdQ+LmlS5dGrVq1Spxfp06daNiwYTbbt99+WyHZ1uSNN96I8847Lw4//PBo0qRJvPvuu3HXXXfFNttsE4cffnhE/DQA+OpOJlGrVq0Sz/ZYVfXu3bvQeG0REQsXLiwy7edWPker27ttTetB9erVC53pb+UJAXJt3rx5sXjx4tUu06JFi5TSpKNJkyar/XsDVASlFABQadxxxx3x9NNPx7nnnhsDBw7MdZwqa/r06YXOulacijzM78EHH1zrv9eJJ54YI0aMqJA8ZbHZZptF69at48Ybb8wOin7CCSfEn//852x5stNOO8WXX35Z4n107do1XnzxxZQS585f/vKXYs9sWRZdu3aNCRMmlDi/TZs2hc4wWln85je/iXvuuWe1yxgVBaDscjqm1K233hq33npr9o2nU6dOcemll2ZPddytW7cib1qnn3563HbbbWlHBQCoMpYsWRKvvPLKapdp165dtGvXrkIeb8aMGfHRRx+tdpkffvghNtxwwxLnt2rVKrbeeusKyVPRXn311dXuJbPhhhtmz5S3Pvvf//4X//vf/0qcv2TJkoiIqF27donLbLjhhsWeiXSlOnXqxB577FH+kOvIxx9/HN98881ql+nRo0dKaQDWHzktpcaMGRPVq1ePLbbYIpIkiXvuuSeuvfbamDhxYnTq1Cm6desWW265ZVx++eXZ22ywwQbRoEGDXEUGAAAAoALk9PC9Aw88sND1K6+8Mm699dZ44403olOnThHxUwm1vh2fDQAAAPBLV2nGlFqxYkU8/PDDsWjRothtt92y00eOHBn3339/tGjRIg488MD44x//GBtssEGJ95Ofnx/5+fnZ6wUFBTFnzpxo0qSJs0YAAAAArGNJksSCBQuiVatWUa1atRKXy3kp9cEHH8Ruu+0WS5YsiXr16sVjjz2WHU/gmGOOiTZt2kSrVq1i0qRJcdFFF8Wnn34ao0ePLvH+hg0bttYDMAIAAACwdqZPnx6bbLJJifNzOqZUxE+nhZ02bVrMmzcvHnnkkbjzzjtjwoQJxQ50+fzzz8e+++4bkydPjvbt2xd7f6vuKTVv3rzYdNNNY/r06caiAgAAAFjH5s+fH61bt465c+dGw4YNS1wu56XUqnr06BHt27eP22+/vci8RYsWRb169WLs2LHRq1evUt3f/Pnzo2HDhjFv3jylFAAAAMA6VtoupuQD+3KkoKCg0J5OP/fee+9FRETLli1TTAQAAABARcvpmFKDBw+OPn36xKabbhoLFiyIUaNGxYsvvhjjxo2LL774IkaNGhV9+/aNJk2axKRJk+K8886LvffeOzp37pzL2AAAAACspZyWUt9++22ccMIJMWPGjGjYsGF07tw5xo0bF/vtt19Mnz49nn322bjhhhti0aJF0bp16+jfv3/84Q9/yGVkAAAAACpApRtTqqIZUwoAAAAgPVV2TCkAAAAA1n9KKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHU1ch0AAAAqyp8nzi71shfv0HQdJgEA1sSeUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkLqel1K233hqdO3eOBg0aRIMGDWK33XaLp59+Ojt/yZIlMWjQoGjSpEnUq1cv+vfvH7NmzcphYgAAAAAqQk5LqU022ST+/Oc/xzvvvBNvv/127LPPPnHwwQfHRx99FBER5513XowZMyYefvjhmDBhQnzzzTdx6KGH5jIyAAAAABUgkyRJkusQP9e4ceO49tpr47DDDotmzZrFqFGj4rDDDouIiE8++SQ6duwYr7/+euy6666lur/58+dHw4YNY968edGgQYN1GR0AgBz788TZpV724h2arsMkAPDLVdouptKMKbVixYp44IEHYtGiRbHbbrvFO++8E8uWLYsePXpkl+nQoUNsuumm8frrr5d4P/n5+TF//vxCFwAAAAAql5yXUh988EHUq1cv8vLy4owzzojHHnsstt5665g5c2bUqlUrGjVqVGj5jTbaKGbOnFni/Q0bNiwaNmyYvbRu3Xod/wsAAAAAKKucl1JbbbVVvPfee/Hmm2/GmWeeGSeeeGJ8/PHH5b6/wYMHx7x587KX6dOnV2BaAAAAACpCjVwHqFWrVmy++eYREdGlS5d466234m9/+1sceeSRsXTp0pg7d26hvaVmzZoVLVq0KPH+8vLyIi8vb13HBgAAAGAt5HxPqVUVFBREfn5+dOnSJWrWrBnPPfdcdt6nn34a06ZNi9122y2HCQEAAABYWzndU2rw4MHRp0+f2HTTTWPBggUxatSoePHFF2PcuHHRsGHDOPnkk+P888+Pxo0bR4MGDeLXv/517LbbbqU+8x4AAAAAlVNOS6lvv/02TjjhhJgxY0Y0bNgwOnfuHOPGjYv99tsvIiKuv/76qFatWvTv3z/y8/OjV69eccstt+QyMgAAAAAVIJMkSZLrEOvS/Pnzo2HDhjFv3rxo0KBBruMAALAO/Xni7FIve/EOTddhEgD45SptF1PpxpQCAAAAYP2nlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdTktpYYNGxY77bRT1K9fP5o3bx79+vWLTz/9tNAy3bp1i0wmU+hyxhln5CgxAAAAABUhp6XUhAkTYtCgQfHGG2/E+PHjY9myZdGzZ89YtGhRoeVOPfXUmDFjRvZyzTXX5CgxAAAAABWhRi4ffOzYsYWujxgxIpo3bx7vvPNO7L333tnpG2ywQbRo0SLteAAAAACsI5VqTKl58+ZFRETjxo0LTR85cmQ0bdo0ttlmmxg8eHD8+OOPuYgHAAAAQAXJ6Z5SP1dQUBDnnntu7LHHHrHNNttkpx9zzDHRpk2baNWqVUyaNCkuuuii+PTTT2P06NHF3k9+fn7k5+dnr8+fP3+dZwcAAACgbCpNKTVo0KD48MMP45VXXik0/bTTTsv+/7bbbhstW7aMfffdN7744oto3759kfsZNmxYDB06dJ3nBQAAAKD8KsXhe2effXY89dRT8cILL8Qmm2yy2mV32WWXiIiYPHlysfMHDx4c8+bNy16mT59e4XkBAAAAWDs53VMqSZL49a9/HY899li8+OKL0bZt2zXe5r333ouIiJYtWxY7Py8vL/Ly8ioyJgAAAAAVLKel1KBBg2LUqFHxxBNPRP369WPmzJkREdGwYcOoU6dOfPHFFzFq1Kjo27dvNGnSJCZNmhTnnXde7L333tG5c+dcRgcAAABgLeS0lLr11lsjIqJbt26Fpg8fPjwGDBgQtWrVimeffTZuuOGGWLRoUbRu3Tr69+8ff/jDH3KQFgAAAICKkvPD91andevWMWHChJTSAAAAAJCWSjHQOQAAAAC/LEopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAqr0ymbBcAoMpQSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQupyWUsOGDYuddtop6tevH82bN49+/frFp59+WmiZJUuWxKBBg6JJkyZRr1696N+/f8yaNStHiQEAAACoCDktpSZMmBCDBg2KN954I8aPHx/Lli2Lnj17xqJFi7LLnHfeeTFmzJh4+OGHY8KECfHNN9/EoYcemsPUAAAAAKytTJIkSa5DrPTdd99F8+bNY8KECbH33nvHvHnzolmzZjFq1Kg47LDDIiLik08+iY4dO8brr78eu+666xrvc/78+dGwYcOYN29eNGjQYF3/EwAAqEiZTJkW//O735V62Yt3aFrWNABAKZS2i6lUY0rNmzcvIiIaN24cERHvvPNOLFu2LHr06JFdpkOHDrHpppvG66+/npOMAAAAAKy9GrkOsFJBQUGce+65sccee8Q222wTEREzZ86MWrVqRaNGjQotu9FGG8XMmTOLvZ/8/PzIz8/PXp8/f/46ywwAAABA+VSaPaUGDRoUH374YTzwwANrdT/Dhg2Lhg0bZi+tW7euoIQAAAAAVJRKUUqdffbZ8dRTT8ULL7wQm2yySXZ6ixYtYunSpTF37txCy8+aNStatGhR7H0NHjw45s2bl71Mnz59XUYHAAAAoBxyWkolSRJnn312PPbYY/H8889H27ZtC83v0qVL1KxZM5577rnstE8//TSmTZsWu+22W7H3mZeXFw0aNCh0AQAAAKByyemYUoMGDYpRo0bFE088EfXr18+OE9WwYcOoU6dONGzYME4++eQ4//zzo3HjxtGgQYP49a9/HbvttlupzrwHAAAAQOWU01Lq1ltvjYiIbt26FZo+fPjwGDBgQEREXH/99VGtWrXo379/5OfnR69eveKWW25JOSkAAAAAFSmnpVSSJGtcpnbt2nHzzTfHzTffnEIiAAAAANJQKQY6BwAAAOCXRSkFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOrKXUrNnTs37rzzzhg8eHDMmTMnIiLefffd+PrrryssHAAAAADrpxrludGkSZOiR48e0bBhw5g6dWqceuqp0bhx4xg9enRMmzYt7r333orOCQAAAMB6pFx7Sp1//vkxYMCA+Pzzz6N27drZ6X379o2XXnqpwsIBAAAAsH4qVyn11ltvxemnn15k+sYbbxwzZ85c61AAAAAArN/KVUrl5eXF/Pnzi0z/7LPPolmzZmsdCgAAAID1W7lKqYMOOiguv/zyWLZsWUREZDKZmDZtWlx00UXRv3//Cg0IAAAAwPqnXKXUX//611i4cGE0b948Fi9eHF27do3NN9886tevH1deeWVFZwQAAABgPVOus+81bNgwxo8fH6+88kpMmjQpFi5cGL/61a+iR48eFZ0PAAAAgPVQuUqplfbcc8/Yc889KyoLAAAAAL8Q5SqlbrzxxmKnZzKZqF27dmy++eax9957R/Xq1dcqHAAAAADrp3KVUtdff31899138eOPP8aGG24YERE//PBDbLDBBlGvXr349ttvo127dvHCCy9E69atKzQwAAAAAFVfuQY6v+qqq2KnnXaKzz//PL7//vv4/vvv47PPPotddtkl/va3v8W0adOiRYsWcd5551V0XgAAAADWA+XaU+oPf/hDPProo9G+ffvstM033zz+8pe/RP/+/eN///tfXHPNNdG/f/8KCwoAAADA+qNce0rNmDEjli9fXmT68uXLY+bMmRER0apVq1iwYMHapQMAAABgvVSuUqp79+5x+umnx8SJE7PTJk6cGGeeeWbss88+ERHxwQcfRNu2bSsmJQAAAADrlXKVUnfddVc0btw4unTpEnl5eZGXlxc77rhjNG7cOO66666IiKhXr1789a9/rdCwAAAAAKwfyjWmVIsWLWL8+PHxySefxGeffRYREVtttVVstdVW2WW6d+9eMQkBAAAAWO+Uq5RaqUOHDtGhQ4eKygIAAADAL0S5S6mvvvoqnnzyyZg2bVosXbq00LzrrrturYMBAAAAsP4qVyn13HPPxUEHHRTt2rWLTz75JLbZZpuYOnVqJEkSv/rVryo6IwAAAADrmXINdD548OC44IIL4oMPPojatWvHo48+GtOnT4+uXbvG4YcfXtEZAQAAAFjPlKuU+u9//xsnnHBCRETUqFEjFi9eHPXq1YvLL788rr766goNCAAAAMD6p1ylVN26dbPjSLVs2TK++OKL7LzZs2dXTDIAAAAA1lvlGlNq1113jVdeeSU6duwYffv2jd/+9rfxwQcfxOjRo2PXXXet6IwAAAAArGfKVUpdd911sXDhwoiIGDp0aCxcuDAefPDB2GKLLZx5DwAAAIA1Klcp1a5du+z/161bN2677bYKCwQAAADA+q9cY0q1a9cuvv/++yLT586dW6iwAgAAAIDilKuUmjp1aqxYsaLI9Pz8/Pj666/XOhQAAAAA67cyHb735JNPZv9/3Lhx0bBhw+z1FStWxHPPPRebbbZZhYUDAAAAYP1UplKqX79+ERGRyWTixBNPLDSvZs2asdlmm8Vf//rXCgsHAAAAwPqpTKVUQUFBRES0bds23nrrrWjatOk6CQUAAADA+q1cZ9+bMmVKRecAAAAA4BekXKVURMRzzz0Xzz33XHz77bfZPahWuvvuu9c6GAAAAADrr3KVUkOHDo3LL788dtxxx2jZsmVkMpmKzgUAAADAeqxcpdRtt90WI0aMiOOPP76i8wAAAADwC1CtPDdaunRp7L777hWdBQAAAIBfiHKVUqecckqMGjWqorMAAAAA8AtRrsP3lixZEv/4xz/i2Wefjc6dO0fNmjULzb/uuusqJBwAAAAA66dylVKTJk2K7bffPiIiPvzww0LzDHoOAAAAwJqUq5R64YUXKjoHAAAAAL8g5RpTaqXJkyfHuHHjYvHixRERkSRJhYQCAAAAYP1WrlLq+++/j3333Te23HLL6Nu3b8yYMSMiIk4++eT47W9/W6EBAQAAAFj/lKuUOu+886JmzZoxbdq02GCDDbLTjzzyyBg7dmyFhQMAAABg/VSuMaWeeeaZGDduXGyyySaFpm+xxRbx5ZdfVkgwAAAAANZf5dpTatGiRYX2kFppzpw5kZeXt9ahAAAAAFi/lauU2muvveLee+/NXs9kMlFQUBDXXHNNdO/evcLCAQAAALB+Ktfhe9dcc03su+++8fbbb8fSpUvjwgsvjI8++ijmzJkTr776akVnBAAAAGA9U649pbbZZpv47LPPYs8994yDDz44Fi1aFIceemhMnDgx2rdvX9EZAQAAAFjPlGtPqYiIhg0bxiWXXFKRWQAAAAD4hSjXnlLDhw+Phx9+uMj0hx9+OO655561DgUAAADA+q1cpdSwYcOiadOmRaY3b948rrrqqrUOBQAAAMD6rVyl1LRp06Jt27ZFprdp0yamTZu21qEAAAAAWL+Vq5Rq3rx5TJo0qcj0999/P5o0abLWoQAAAABYv5WrlDr66KPjnHPOiRdeeCFWrFgRK1asiOeffz5+85vfxFFHHVXRGQEAAABYz5Tr7Ht/+tOfYurUqbHvvvtGjRo/3UVBQUGccMIJxpQCAAAAYI3KXEolSRIzZ86MESNGxBVXXBHvvfde1KlTJ7bddtto06bNusgIAAAAwHqmXKXU5ptvHh999FFsscUWscUWW6yLXAAAAACsx8o8plS1atViiy22iO+//35d5AEAAADgF6BcA53/+c9/jt/97nfx4YcfVnQeAAAAAH4ByjXQ+QknnBA//vhjbLfddlGrVq2oU6dOoflz5sypkHAAAAAArJ/KVUrdcMMNFRwDAAAAgF+ScpVSJ554YkXnAAAAAOAXpFxjSkVEfPHFF/GHP/whjj766Pj2228jIuLpp5+Ojz76qMLCAQAAALB+KlcpNWHChNh2223jzTffjNGjR8fChQsjIuL999+PIUOGVGhAAAAAANY/5SqlLr744rjiiiti/PjxUatWrez0ffbZJ954440KCwcAAADA+qlcpdQHH3wQhxxySJHpzZs3j9mzZ691KAAAAADWb+UqpRo1ahQzZswoMn3ixImx8cYbr3UoAAAAANZv5SqljjrqqLjoooti5syZkclkoqCgIF599dW44IIL4oQTTqjojAAAAACsZ8pVSl111VXRsWPH2HTTTWPhwoWx9dZbx9577x277757/OEPf6jojAAAAACsZ2qUZeGCgoK49tpr48knn4ylS5fG8ccfH/3794+FCxfGDjvsEFtsscW6ygkAAADAeqRMpdSVV14Zl112WfTo0SPq1KkTo0aNiiRJ4u67715X+QAAAABYD5Xp8L177703brnllhg3blw8/vjjMWbMmBg5cmQUFBSsq3wAAAAArIfKVEpNmzYt+vbtm73eo0ePyGQy8c0331R4MAAAAADWX2UqpZYvXx61a9cuNK1mzZqxbNmyCg0FAAAAwPqtTGNKJUkSAwYMiLy8vOy0JUuWxBlnnBF169bNThs9enTFJQQAAABgvVOmUurEE08sMu24446rsDAAAAAA/DKUqZQaPnz4usoBAAAAwC9ImcaUAgAAAICKoJQCAAAAIHU5LaVeeumlOPDAA6NVq1aRyWTi8ccfLzR/wIABkclkCl169+6dm7AAAAAAVJicllKLFi2K7bbbLm6++eYSl+ndu3fMmDEje/nnP/+ZYkIAAAAA1oUyDXRe0fr06RN9+vRZ7TJ5eXnRokWLlBIBAAAAkIZKP6bUiy++GM2bN4+tttoqzjzzzPj+++9Xu3x+fn7Mnz+/0AUAAACAyqVSl1K9e/eOe++9N5577rm4+uqrY8KECdGnT59YsWJFibcZNmxYNGzYMHtp3bp1iokBAAAAKI2cHr63JkcddVT2/7fddtvo3LlztG/fPl588cXYd999i73N4MGD4/zzz89enz9/vmIKAAAAoJKp1HtKrapdu3bRtGnTmDx5conL5OXlRYMGDQpdAAAAAKhcqlQp9dVXX8X3338fLVu2zHUUAAAAANZCTg/fW7hwYaG9nqZMmRLvvfdeNG7cOBo3bhxDhw6N/v37R4sWLeKLL76ICy+8MDbffPPo1atXDlMDAAAAsLZyWkq9/fbb0b179+z1lWNBnXjiiXHrrbfGpEmT4p577om5c+dGq1atomfPnvGnP/0p8vLychUZAAAAgAqQ01KqW7dukSRJifPHjRuXYhoAAAAA0lKlxpQCAAAAYP2glAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFJXI9cBAIBfkEymbMsnybrJAQBAztlTCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDU5bSUeumll+LAAw+MVq1aRSaTiccff7zQ/CRJ4tJLL42WLVtGnTp1okePHvH555/nJiwAAAAAFSanpdSiRYtiu+22i5tvvrnY+ddcc03ceOONcdttt8Wbb74ZdevWjV69esWSJUtSTgoAAABARaqRywfv06dP9OnTp9h5SZLEDTfcEH/4wx/i4IMPjoiIe++9NzbaaKN4/PHH46ijjkozKgAAAAAVqNKOKTVlypSYOXNm9OjRIzutYcOGscsuu8Trr79e4u3y8/Nj/vz5hS4AAAAAVC6VtpSaOXNmRERstNFGhaZvtNFG2XnFGTZsWDRs2DB7ad269TrNCQAAAEDZVdpSqrwGDx4c8+bNy16mT5+e60gAAAAArKLSllItWrSIiIhZs2YVmj5r1qzsvOLk5eVFgwYNCl0AAAAAqFxyOtD56rRt2zZatGgRzz33XGy//fYRETF//vx4880348wzz8xtOKhAf544u0zLX7xD03WUBAAAANKT01Jq4cKFMXny5Oz1KVOmxHvvvReNGzeOTTfdNM4999y44oorYosttoi2bdvGH//4x2jVqlX069cvd6EBAAAAWGs5LaXefvvt6N69e/b6+eefHxERJ554YowYMSIuvPDCWLRoUZx22mkxd+7c2HPPPWPs2LFRu3btXEUGAAAAoALktJTq1q1bJElS4vxMJhOXX355XH755SmmAgAAAGBdq7QDnQMAAACw/lJKAQAAAJA6pRQAAAAAqVNKAQAAAJC6nA50DgBUbZlMpkzLl3x6EwAAfmnsKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKSuRq4DAAAAlCSTyZRp+SRJ1lESACqaPaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASF2NXAcAAADIlT9PnF3qZS/eoek6TALwy6OUAsi1TKZsyyfJuskBAACQIqUUUG5l+WUxwq+LAAAA/B9jSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKmrkesAwLqVyWTKtHySJOsoCQAAAPwfe0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAArAuZTNkuVAzPO0CVoZQCAAAAIHU1ch0AAKAkf544u0zLX7xD03WUBACAimZPKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHU1ch2AsslkMmVaPkmSdZQEAAAAoPzsKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKSuRq4DAADwy5HJZMq0fLKOcgAAuWdPKQAAAABSZ08pAACgsDLu0RaJfdoAKDt7SgEAAACQOntKAawDZRkzxW/LAADAL5E9pQAAAABInVIKAAAAgNQppQAAAABInTGlACDHyjIGWURE4ixXAACsB+wpBQAAAEDqlFIAAAAApE4pBQAAAEDqlFIAAAAApM5A5wAAsJ4r8wkV1lEOqg4n4QDSYE8pAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlIJ1IZMp/QUAAAB+gSp1KXXZZZdFJpMpdOnQoUOuY/ELtOp6uKYLAAD4DAmwejVyHWBNOnXqFM8++2z2eo0alT4yAAAAAGtQ6RueGjVqRIsWLXIdAwAAAIAKVKkP34uI+Pzzz6NVq1bRrl27OPbYY2PatGm5jgQAAADAWqrUe0rtsssuMWLEiNhqq61ixowZMXTo0Nhrr73iww8/jPr16xd7m/z8/MjPz89enz9/flpxAQAAACilSl1K9enTJ/v/nTt3jl122SXatGkTDz30UJx88snF3mbYsGExdOjQtCICAAAAUA6VupRaVaNGjWLLLbeMyZMnl7jM4MGD4/zzz89enz9/frRu3TqNeACp+PPE2WVa/uIdmq6jJAAAAOVX6ceU+rmFCxfGF198ES1btixxmby8vGjQoEGhCwAArK1MJlOmCwCwepW6lLrgggtiwoQJMXXq1HjttdfikEMOierVq8fRRx+d62gAAAAArIVKffjeV199FUcffXR8//330axZs9hzzz3jjTfeiGbNmuU6GgAAAABroVKXUg888ECuIwAAAACwDlTqw/cAAAAAWD8ppQAAAABInVIKAAAAgNQppQAAAABInVIKAAAAgNQppQAAAABInVIKAAAAgNQppQAAAABIXY1cB2Ady2RKv2ySrLscVB1lWWfe/W7d5SirsuSOsL4Dvyy2kcC65nsHUA5KKQDWG5kyfCBOfCAGAICccvgeAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQuhq5DkDl8eeJs8u0/MU7NF1HSQAAAID1nT2lAAAAAEidUgoAAACA1CmlAAAAAEidUgoAAACA1BnoHACgislkMmVaPllHOQAA1oY9pQAAAABInVIKAAAAgNQppQAAAABInTGlAACAtfLnibNLvezFOzRdh0moCsqyvkRYZ2B9Zk8pAAAAAFJnTymg0irL2aWcWQoAAKBqUUoBAEAl4BA4yJEy/BAaiZ9CoSI5fA8AAACA1CmlAAAAAEidUgoAAACA1CmlAAAAAEidUgoAAACA1CmlAAAAAEidUgoAAACA1CmlAAAAAEidUgoAAACA1CmlAAAAAEidUgoAAACA1CmlAAAAAEhdjVwHAAAAoOz+PHF2qZe9eIem6zAJQPnYUwoAAACA1NlTCoBCMplMqZdNkmQdJgEA+GUpy+ewCJ/FqPqUUgBQCmU5RCLCYRIAALAmDt8DAAAAIHVKKQAAAABSp5QCAAAAIHXGlAIAoBBjqAEAaVBKAUBVU8Yz84Qz8xTLGY4AqPLK8l7mfaxi+BxWoRy+BwAAAEDqlFIAAAAApM7he8AvkvFSAAAAcsueUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOoMdE5qMplMmZZPkmQdJQH4ZTGwPwAAlZE9pQAAAABInVIKAAAAgNQppQAAAABInVIKAAAAgNQppQAAAABInVIKAAAAgNTVyHUAAID10Z8nzi7T8hfv0HQdJQEAqJzsKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQXAL1MmU7YLWGcAACqUs+8BAACw3siU8YeBZB3lANZMKQVAav48cXaZlr94h6brKAkAAJBrSikAAABYz/lxkMrImFIAAAAApE4pBQAAAEDqlFIAAAAApE4pBQAAAEDqlFIAAAAApE4pBQAAAEDqlFIAAAAApK5GrgMAAABAVfDnibPLtPzFOzRdR0lg/aCUYr3gzQEAAACqFqUUAAAAwDpgB4rVM6YUAAAAAKlTSgEAAACQuipRSt18882x2WabRe3atWOXXXaJ//znP7mOBAAAAMBaqPSl1IMPPhjnn39+DBkyJN59993YbrvtolevXvHtt9/mOhoAAAAA5VTpS6nrrrsuTj311Bg4cGBsvfXWcdttt8UGG2wQd999d66jAQAAAFBOlfrse0uXLo133nknBg8enJ1WrVq16NGjR7z++uvF3iY/Pz/y8/Oz1+fNmxcREfPnz1+3YSupsvyrlyxcULb7nl+rbGHKqCx/s0qXvQzLyl4xyvoKl71ilHXbWqmyl3F52SuG7Ku5f9mLv+8yLl+W7JXpPTVC9opSpbOvw/fVyvS8V6ZtTITsJd73Os7+S/FLfd5Xbs+SJFntcplkTUvk0DfffBMbb7xxvPbaa7Hbbrtlp1944YUxYcKEePPNN4vc5rLLLouhQ4emGRMAAACAVUyfPj022WSTEudX6j2lymPw4MFx/vnnZ68XFBTEnDlzokmTJpHJZHKYrPKYP39+tG7dOqZPnx4NGjTIdZxSq6q5I2TPFdlzQ/bckD03ZM8N2XND9tyQPTeqavaqmjtC9vVNkiSxYMGCaNWq1WqXq9SlVNOmTaN69eoxa9asQtNnzZoVLVq0KPY2eXl5kZeXV2hao0aN1lXEKq1BgwZV8gVTVXNHyJ4rsueG7Lkhe27Inhuy54bsuSF7blTV7FU1d4Ts65OGDRuucZlKPdB5rVq1okuXLvHcc89lpxUUFMRzzz1X6HA+AAAAAKqWSr2nVETE+eefHyeeeGLsuOOOsfPOO8cNN9wQixYtioEDB+Y6GgAAAADlVOlLqSOPPDK+++67uPTSS2PmzJmx/fbbx9ixY2OjjTbKdbQqKy8vL4YMGVLkMMfKrqrmjpA9V2TPDdlzQ/bckD03ZM8N2XND9tyoqtmrau4I2X+pKvXZ9wAAAABYP1XqMaUAAAAAWD8ppQAAAABInVIKAAAAgNQppQAAAKAK+eabb3IdASqEUuoX7sMPP8x1BADWgYULF+Y6ArCK2bNnx/z583Mdg0rq8ssvjx9//DHXMcqlKn+nmDJlSq4jlEunTp1i1KhRuY4Ba00p9Qu0YMGC+Mc//hE777xzbLfddrmOU6znn38+tt5662I/uM2bNy86deoUL7/8cg6SUZn973//i6p4QtHFixfHU089lb0+ePDgOP/887OX3/3ud7FkyZIcJqSyuf7661c7f8GCBdGrV6+U0lSsr776Kk477bRcx4AKM3fu3Bg0aFA0bdo0Ntpoo9hwww2jRYsWMXjw4CpbQKy0ePHiXEco1uuvv17ofTUi4t577422bdtG8+bN47TTTov8/PwcpSvZ0KFDq+wPCp07d45ddtkl7rjjjliwYEGu45RJ+/bto23btnHSSSfFfffdF1999VWuI5XKlVdeGaeffnocfvjhMWfOnFzHgXLLJFXxGxzl8tJLL8Vdd90Vjz76aLRq1SoOPfTQ6N+/f+y00065jlbEQQcdFN27d4/zzjuv2Pk33nhjvPDCC/HYY4+lnKz0CgoKYsSIETF69OiYOnVqZDKZaNu2bRx22GFx/PHHRyaTyXXEYvXt2zf++c9/RsOGDSMi4s9//nOcccYZ0ahRo4iI+P7772OvvfaKjz/+OIcpi1e9evWYMWNGNG/ePCIijjzyyLjxxhtjo402ynGy1bvtttviX//6V4wZMyYiIurXrx+dOnWKOnXqRETEJ598EhdeeGGJr4dcK+2v/g0aNFjHScru0EMPLdVyo0ePXsdJyqZOnTpx++23xwknnFBk3qJFi6Jnz57x/fffxyeffJKDdGvn/fffj1/96lexYsWKXEcp4qSTTirVcnffffc6TvLLUlBQENdee208+eSTsXTp0th3331jyJAh2W1kZTZnzpzYbbfd4uuvv45jjz02OnbsGBERH3/8cYwaNSo6dOgQr7zySkyaNCneeOONOOecc3KcuHTy8/Pj73//e1x77bUxc+bMXMcpok+fPtGtW7e46KKLIiLigw8+iF/96lcxYMCA6NixY1x77bVx+umnx2WXXZbboKuoVq1azJw5M/s5pip5+eWXY/jw4fHII49EQUFB9O/fP0455ZTYa6+9ch1tjV588cXs5c0334ylS5dGu3btYp999onu3btH9+7dK+1nySlTpsTJJ58cH3/8cdxxxx1x4IEH5jrSWlu4cGEUFBQUmlYZP0Ou9M0338Qrr7wS3377bZHcVWWbnmtKqfXczJkzY8SIEXHXXXfF/Pnz44gjjojbbrst3n///dh6661zHa9Ebdq0ibFjx2Y/vK3qk08+iZ49e8a0adNSTlY6SZLEgQceGP/+979ju+22iw4dOkSSJPHf//43PvjggzjooIPi8ccfz3XMYq1a7DRo0CDee++9aNeuXUREzJo1K1q1alUpvzCu+mGufv368f7772ezV1Z77bVXXHjhhdkPEqvmvv/+++Pmm2+O119/PZcxS1StWrXVlqxJkkQmk6mU68zAgQNLtdzw4cPXcZKyeeSRR+L444+PBx98MA466KDs9EWLFkWvXr3i22+/jQkTJkTLli1zmLJ8KnMpVa1atWjTpk3ssMMOq90rszL+YLKm12lERCaTieXLl6eUqPT+9Kc/xWWXXRY9evSIOnXqxLhx4+Loo4+uEuXfueeeG88991w8++yzRb7Uzpw5M3r27BlbbbVVPPPMM3HjjTfGiSeemKOkReXn58dll10W48ePj1q1asWFF14Y/fr1i+HDh8cll1wS1atXj7PPPjtb/FQmLVu2jDFjxsSOO+4YERGXXHJJTJgwIV555ZWIiHj44YdjyJAhle4HtmrVqsWsWbOiWbNmuY5SbosWLYqHHnooRowYES+//HJsvvnmcfLJJ8eJJ54YLVq0yHW8NVqyZEm89tpr2ZLqP//5Tyxbtiw6dOgQH330Ua7jlejvf/97nHfeedGxY8eoUaNGoXnvvvtujlKV3pQpU+Lss8+OF198sdDRAZX5M2RExIgRI+L000+PWrVqRZMmTQq9z2Yymfjf//6Xw3RVR401L0JVdeCBB8ZLL70U+++/f9xwww3Ru3fvqF69etx22225jrZGs2bNipo1a5Y4v0aNGvHdd9+lmKhsRowYES+99FI899xz0b1790Lznn/++ejXr1/ce++9xe7lkGurftHSW697kydPjm233TZ7vXbt2lGt2v8dXb3zzjvHoEGDchGtVF544YXs/ydJEn379o0777wzNt544xymKp3KVjaV1mGHHRZz586No48+Ov71r39Ft27dYtGiRdG7d++YNWtWlS2kKrszzzwz/vnPf8aUKVNi4MCBcdxxx0Xjxo1zHatUVleUvf7663HjjTcW+YW3srj33nvjlltuidNPPz0iIp599tnYf//948477yy0rayMHn/88bj99tuL3cuiRYsWcc0110Tfvn1jyJAhlaqQioi49NJL4/bbb48ePXrEa6+9FocffngMHDgw3njjjbjuuuvi8MMPj+rVq+c6ZrF++OGHQs/5hAkTok+fPtnrO+20U0yfPj0X0dZoyy23XGOBXJkP1apbt24MHDgwBg4cGJMnT47hw4fHzTffHH/84x+jd+/e8eSTT+Y64mrVrl079tlnn9hzzz2je/fu8fTTT8ftt99eqfc8/vLLL2P06NGx4YYbxsEHH1yklKoKjjvuuEiSJO6+++7YaKONKu0RJav64x//GJdeemkMHjy40r8fVWZVb42l1J5++uk455xz4swzz4wtttgi13HKZOONN44PP/wwNt9882LnT5o0qVJ/4frnP/8Zv//974sUUhER++yzT1x88cUxcuTISllKVWWZTKbIm1hVeFObO3duobEtVi1cCwoKKuXYFyt17dq10PXq1avHrrvuWun3UKvqTjnllJgzZ04cfPDB8cQTT8Sll14a33zzTUyYMCFatWqV63jrpZtvvjmuu+66GD16dNx9990xePDg2H///ePkk0+Onj17VurtzcEHH1xk2qeffhoXX3xxjBkzJo499ti4/PLLc5BszaZNmxZ9+/bNXu/Ro0dkMpn45ptvYpNNNslhsjWbMWNGdOrUqcT522yzTVSrVi2GDBmSYqrSefjhh+Pee++Ngw46KD788MPo3LlzLF++PN5///1Kva5HRGy00UYxZcqUaN26dSxdujTefffdGDp0aHb+ggULVvvjZy4NHTo0O4RCVbf55pvH73//+2jTpk0MHjw4/vWvf+U6UomWLl0ab7zxRrzwwgvZw/hat24de++9d/z9738v8lmnsrjjjjvit7/9bfTo0SM++uijKruX3fvvvx/vvPNObLXVVrmOUiY//vhjHHXUUQqptaSUWo+98sorcdddd0WXLl2iY8eOcfzxx8dRRx2V61il0rdv3+wvKrVr1y40b/HixTFkyJA44IADcpRuzSZNmhTXXHNNifP79OkTN954Y4qJSq+qFjsRP+2lM2DAgMjLy4uIn3bBPuOMM6Ju3bqFlqts4wNtsskm8eGHH5b4Rjxp0qRK/8WL3Ljwwgtjzpw5se+++8Zmm20WL774YqVfV9Y0jtfcuXPTCVJOeXl5cfTRR8fRRx8dX375ZYwYMSLOOuusWL58eXz00UdRr169XEdco2+++SaGDBkS99xzT/Tq1Svee++92GabbXIdq0TLly8v8lmgZs2asWzZshwlKr2mTZvG1KlTS3xdTpkypdKOH/TVV19Fly5dIuKn8iwvLy/OO++8KvGZoG/fvnHxxRfH1VdfHY8//nhssMEGhcY2mjRpUrRv3z6HCUt21FFHVdp1oixeeumluPvuu+PRRx+NatWqxRFHHBEnn3xyrmMVa5999ok333wz2rZtG127do3TTz89Ro0aVal/AI+I6N27d/znP/+Jv//971X+h+6Vey9WtVLq5JNPjocffjguvvjiXEep0pRS67Fdd901dt1117jhhhviwQcfjLvvvjvOP//8KCgoiPHjx0fr1q2jfv36uY5ZrD/84Q8xevTo2HLLLePss8/ObqA++eSTuPnmm2PFihVxySWX5DhlyebMmbPaARE32mij+OGHH1JMVHprKnYq8x47qx76cNxxx+UoSdn07ds3Lr300th///2LLWGHDh0a+++/f47SURmtWuzUrFkzmjZtGr/5zW8KTa9sBWxErHEPgIYNG1aZD9crx2lKkqTSjnfxc/PmzYurrroqbrrppth+++3jueeeqxKDEK/6vhRR/I8OlXF979WrV1xyySXZcZl+Lj8/P/sDXGW0YsWKQplr1KhRJUrXiJ/GITv00EOja9euUa9evbjnnnsK/Vvuvvvu6NmzZw4TFq8qFH6r880338SIESNixIgRMXny5Nh9993jxhtvjCOOOKLID4SVycsvvxwtW7aMffbZJ7p16xZdu3aNJk2a5DrWGq1YsWK9+eHyzjvvjDPOOCO+/vrr2GabbYrsydi5c+ccJVu9YcOGxQEHHBBjx46Nbbfdtkju6667LkfJqhYDnf/CfPrpp3HXXXfFfffdF3Pnzo399tuv0h7b/eWXX8aZZ54Z48aNy45rlMlkolevXnHzzTdH27Ztc5ywZNWrV4+ZM2eWuAttZR4svKoO/FyVzZo1K7bffvuoVatWnH322bHllltGxE+v17///e+xfPnymDhxYqU988uq6tevH5MmTarUr9Gqzus0d/Lz87OH773yyitxwAEHxMCBA6N3796Vevf9a665Jq6++upo0aJFXHXVVcUezldZVeX1/auvvoodd9wx8vLyYtCgQYVOfHLLLbdEfn5+vPXWW7HpppvmOmoR1apViz59+mTLwDFjxsQ+++xT6fc+/rl58+ZFvXr1iox9NWfOnKhXr16RojDXqvLZ9/r06RPPPvtsNG3aNE444YQ46aSTqsxeL4sWLYqXX345XnzxxXjhhRfivffeiy233DK6du2aLamq6mFxVcUbb7wRxxxzTEydOjU7beWPPpV5oPMrrrgiLr300thqq62KjIWVyWTi+eefz2G6qkMp9Qu1YsWKGDNmTNx9992VtpRa6YcffojJkydHkiSxxRZbxIYbbpjrSGu06ge5VeXn58fYsWMr7QaW9E2ZMiXOPPPMGD9+fKESdr/99otbbrmlUo/PtOpeO1XxiwuUxllnnRUPPPBAtG7dOk466aQ49thjo2nTprmOVSrVqlWLOnXqRI8ePVY7OLXXacWbMmVKnHXWWfHMM88U2b7//e9/L3H8zFyrymUg6TvooIPi5JNPjgMOOKDSDoBfWgsWLIhXXnklO77U+++/H1tssUV8+OGHuY623tp6662jY8eOceGFFxY70HmbNm1ylGz1Ntxww7j++utjwIABuY5SpSmlYB3wQY7ymjNnTkyePDkifhogtCqc2cv6zi9FtWrVYtNNN40ddthhtYfZVMZiZ8CAAaU6NMjrdN354Ycf4vPPP4+IqrN9h1+igoKCeOutt+KFF16IF154IV555ZVYsmSJH5PXobp168b7779faUv6krRo0SJefvnlKndSscpGKQUAUAqKHYD1T0FBQbz99tvZw/deffXVWLRoUWy88cbRvXv37KWy7q2zPjjwwANjwIAB0b9//1xHKZNhw4bFjBkzKu0JrKoKpRQAAAC/SA0aNIhFixZFixYtsgVUt27dKu0ZGtdH//jHP+KKK66Ik046qdgBww866KAcJVu9Qw45JJ5//vlo0qRJdOrUqUjuyrjndGWklAIAAOAX6fbbb4/u3btnTzRD+lZ3opDKPND5moawsOd06SilAAAAAEpp+fLlMWrUqOjZs2e0aNEi13GqtMp77mIAAABgvbVs2bKoUaNGlTu7YY0aNeKMM86I/Pz8XEep8pRSAAAAQOpq1qwZm266aaU9RG91dt5555g4cWKuY1R5NXIdAAAAAPhluuSSS+L3v/993HfffdG4ceNcxym1s846K37729/GV199FV26dIm6desWmt+5c+ccJatajCkFAAAA5MQOO+wQkydPjmXLlkWbNm2KlDvvvvtujpKtXnEDtGcymUiSpFIP0F7Z2FMKAAAAyIl+/frlOkK5TJkyJdcR1gv2lAIAAAAgdQY6BwAAAHJm7ty5ceedd8bgwYNjzpw5EfHTYXtff/11jpOt3n333Rd77LFHtGrVKr788suIiLjhhhviiSeeyHGyqkMpBQAAAOTEpEmTYsstt4yrr746/vKXv8TcuXMjImL06NExePDg3IZbjVtvvTXOP//86Nu3b8ydOzc7hlSjRo3ihhtuyG24KkQpBQAAAOTE+eefHwMGDIjPP/88ateunZ3et2/feOmll3KYbPVuuummuOOOO+KSSy6J6tWrZ6fvuOOO8cEHH+QwWdWilAIAAABy4q233orTTz+9yPSNN944Zs6cmYNEpTNlypTYYYcdikzPy8uLRYsW5SBR1aSUAgAAAHIiLy8v5s+fX2T6Z599Fs2aNctBotJp27ZtvPfee0Wmjx07Njp27Jh+oCqqRq4DAAAAAL9MBx10UFx++eXx0EMPRUREJpOJadOmxUUXXRT9+/fPcbqSnX/++TFo0KBYsmRJJEkS//nPf+Kf//xnDBs2LO68885cx6syMkmSJLkOAQAAAPzyzJs3Lw477LB4++23Y8GCBdGqVauYOXNm7LbbbvHvf/876tatm+uIJRo5cmRcdtll8cUXX0RERKtWrWLo0KFx8skn5zhZ1aGUAgAAAHLq1Vdfjffffz8WLlwYv/rVr6JHjx65jlRqP/74YyxcuDCaN2+e6yhVjlIKAAAAyIl77703jjzyyMjLyys0fenSpfHAAw/ECSeckKNkq7fPPvvE6NGjo1GjRoWmz58/P/r16xfPP/98boJVMUopAAAAICeqV68eM2bMKLKX0ffffx/NmzePFStW5CjZ6lWrVi1mzpxZJPe3334bG2+8cSxbtixHyaoWA50DAAAAOZEkSWQymSLTv/rqq2jYsGEOEq3epEmTsv//8ccfx8yZM7PXV6xYEWPHjo2NN944F9GqJKUUAAAAkKoddtghMplMZDKZ2HfffaNGjf+rJ1asWBFTpkyJ3r175zBh8bbffvts7n322afI/Dp16sRNN92Ug2RVk1IKAAAASFW/fv0iIuK9996LXr16Rb169bLzatWqFZtttln0798/R+lKNmXKlEiSJNq1axf/+c9/olmzZtl5tWrViubNm0f16tVzmLBqMaYUAAAAkBP33HNPHHnkkVG7du1cRyEHquU6AAAAAPDLdOKJJ2YLqbPOOitmz56d40Rl16BBg/jf//6X6xhVklIKAAAAyLn7778/5s+fn+sYZeYAtPJTSgEAAAA5p9z55VFKAQAAAJTTcccdFw0aNMh1jCpJKQUAAADk3IIFC6Jdu3a5jrFazz//fGy99daFDjO89dZbo2nTpjFv3rzo1KlTvPzyyzlMWLUopQAAAIBUVatWLapXr77aS40aNXIds4gbbrghTj311GL3jGrYsGGcfvrpcd111+UgWdWUSRy0CQAAAKToiSeeKHHe66+/HjfeeGMUFBTEkiVLUky1Zm3atImxY8dGx44di53/ySefRM+ePWPatGkpJ6uaKl/tCAAAAKzXDj744CLTPv3007j44otjzJgxceyxx8bll1+eg2SrN2vWrKhZs2aJ82vUqBHfffddiomqNofvAQAAADnzzTffxKmnnhrb/r/27j8k6vuB4/jrc+c0UWnBxLNmdZtZLu0yopF/9GOVq5EQuo2NNaw0KvSPpGwW04pITWqL6rK22nXNRcF+tK3MqLGCiuiHWfmjH15owVxry8ayVml+/4iOrpRvuvi4ds8HHJzve3/e9/r8Jy/en/fFx6ulpUWVlZVyu93q169fd0d7Qp8+fVRVVdXh52fOnFFkZKSJiZ5vlFIAAAAAAMB0f/75pz766CNFR0erurpaP/30k3788UfFxcV1d7QOvfXWW8rLy2v3scLbt29r8eLFmjx5cjckez5xphQAAAAAADBVcXGxVqxYIZvNpoKCgnYf5/s3unr1qoYNGyar1aqsrCwNHDhQ0oOzpJxOp1pbW1VRUaGIiIhuTvp8oJQCAAAAAACmslgsCg4O1vjx42W1Wjuc9+2335qY6uk0NDRozpw52rt3rx5WKoZh6M0335TT6ZTdbu/mhM8PSikAAAAAAGCqadOmyTCM/zvP5XKZkKZrmpqaVFdXp7a2Ng0YMEC9evXq7kjPHUopAAAAAAAAmI6DzgEAAAAAAGA6SikAAAAAAACYjlIKAAAAAAAApqOUAgAAAAAAgOkopQAAAJ6R+vp6GYahysrKDuccOHBAhmHoxo0bpuUCAAD4N6KUAgAAfuPhz08bhqEXXnhBdrtdCxYs0N9///1M1o+KilJjY6Pi4uKeyXpd9fAeO3otWbKkW/MBAABIUkB3BwAAADDTxIkT5XK5dO/ePZ08eVJpaWkyDEMrVqz4x2tbrVbZbLZnkPKfaWxs9L7fsWOH8vPzdf78ee9YaGhod8QCAADwwU4pAADgV4KCgmSz2RQVFaUpU6Zo/Pjx2rdvn/fz+/fvq7CwUHa7XcHBwXI4HPr666+9nzc1NemDDz5QeHi4goODNWDAALlcLkntP75XVlammJgYBQcHa+zYsaqvr/fJs2TJEg0dOtRnbPXq1erfv7/P2KZNmxQbG6sePXpo0KBBWr9+fYf3aLPZvK+ePXvKMAzZbDaFhYUpJiZG5eXlPvN37typkJAQ/fXXX9572L59uxITE9WjRw/FxcXp4MGDPtdUVVVp0qRJCg0NVUREhD788EP9/vvvHWYCAAB4HKUUAADwW1VVVTpy5IgCAwO9Y4WFhdq6das2bNig6upqZWdna+rUqd5SJi8vTzU1NdqzZ49qa2tVUlKil156qd31r1y5opSUFCUnJ6uyslIZGRnKzc3tdM6vvvpK+fn5Wr58uWpra1VQUKC8vDy53e5OrRMSEqL33nvPW6I95HK59PbbbyssLMw7lpOTo3nz5unUqVMaOXKkkpOT9ccff0iSbty4oTfeeEMJCQk6ceKEysvLdfXqVb377rudvjcAAOC/eHwPAAD4lV27dik0NFQtLS26c+eOLBaL1q1bJ0m6c+eOCgoKtH//fo0cOVKS9Morr+jQoUPauHGjRo8ercuXLyshIUHDhw+XpCd2ND2qpKREr776qlatWiVJGjhwoM6ePdvpRwUXL16sVatWKSUlRZJkt9tVU1OjjRs3Ki0trVNrZWRkKDExUY2NjYqMjNRvv/2msrIy7d+/32deVlaWUlNTvfdRXl6uzZs3a8GCBVq3bp0SEhJUUFDgnf/FF18oKipKFy5cUExMTKcyAQAA/0QpBQAA/MrYsWNVUlKi5uZmffrppwoICPCWL3V1dbp165YmTJjgc83du3eVkJAgSZozZ45SU1NVUVGhpKQkTZkyRYmJie1+V21trV5//XWfsYdl19Nqbm6Wx+NRenq6Zs6c6R1vaWlRz549O7WWJI0YMUKDBw+W2+1Wbm6uSktL1a9fP40aNarDnAEBARo+fLhqa2slSadPn9bPP//c7tlUHo+HUgoAADwVSikAAOBXQkJCFB0dLenB7h6Hw6HNmzcrPT1dN2/elCTt3r1bffr08bkuKChIkjRp0iQ1NDSorKxM+/bt07hx45SZmamVK1d2KY/FYlFbW5vP2L1797zvH2b6/PPPnyi4rFZrl74zIyNDTqdTubm5crlcmj59ugzDeOrrb968qeTk5HZ3fEVGRnYpEwAA8D+cKQUAAPyWxWLRokWL9PHHH+v27dt67bXXFBQUpMuXLys6OtrnFRUV5b0uPDxcaWlpKi0t1erVq/XZZ5+1u35sbKyOHTvmM3b06FGfv8PDw/Xrr7/6FFOPHpQeERGh3r1769KlS09kstvtXbrvqVOnqqGhQWvWrFFNTU27jwA+mrOlpUUnT55UbGysJGnYsGGqrq5W//79n8gUEhLSpUwAAMD/UEoBAAC/9s4778hqtcrpdCosLEzz589Xdna23G63PB6PKioqtHbtWu+h4vn5+fr+++9VV1en6upq7dq1y1vWPG727Nm6ePGicnJydP78eW3btk1btmzxmTNmzBhdu3ZNxcXF8ng8cjqd2rNnj8+cpUuXqrCwUGvWrNGFCxd09uxZuVwuffLJJ1265169eiklJUU5OTlKSkrSyy+//MQcp9Op7777TufOnVNmZqaampo0Y8YMSVJmZqauX7+u999/X8ePH5fH49HevXs1ffp0tba2dikTAADwP5RSAADArwUEBCgrK0vFxcVqbm7WsmXLlJeXp8LCQsXGxmrixInavXu3d1dSYGCgFi5cqCFDhmjUqFGyWq3avn17u2v37dtX33zzjXbu3CmHw6ENGzb4HA4uPdhNtX79ejmdTjkcDh07dkzz58/3mZORkaFNmzbJ5XIpPj5eo0eP1pYtW7q8U0qS0tPTdffuXW/R9LiioiIVFRXJ4XDo0KFD+uGHH7y/Mti7d28dPnxYra2tSkpKUnx8vObOnasXX3xRFgv/XgIAgKdjtD1+iAEAAAD+87788ktlZ2frl19+UWBgoHe8vr5edrtdp06d0tChQ7svIAAA+M/joHMAAAA/cuvWLTU2NqqoqEizZs3yKaQAAADMxP5qAAAAP1JcXKxBgwbJZrNp4cKF3R0HAAD4MR7fAwAAAAAAgOnYKQUAAAAAAADTUUoBAAAAAADAdJRSAAAAAAAAMB2lFAAAAAAAAExHKQUAAAAAAADTUUoBAAAAAADAdJRSAAAAAAAAMB2lFAAAAAAAAExHKQUAAAAAAADT/Q9Q7sauvON32QAAAABJRU5ErkJggg==\n"
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of Unique Accession Numbers for Significant Data in AdjacentSeq_all__(Aug_Active)___(Mar_Active).tsv: 15\n",
            "Number of Unique Accession Numbers for Not Significant Data in AdjacentSeq_all__(Aug_Active)___(Mar_Active).tsv: 3212\n",
            "Total number of modifications in AdjacentSeq_all__(Aug_Active)___(Mar_Active).tsv: 17793\n",
            "Significant N-terminal (Met-loss) modifications: 2\n",
            "Non-significant N-terminal (Met-loss) modifications: 41\n",
            "Processing file: /content/output/modlist_files/AdjacentSeq_all__(Torpor)___(Aug_Active).tsv\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 1200x800 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAABKUAAAMWCAYAAAAgRDUeAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAB4RUlEQVR4nOzdd5QUZf4+7E+ThgySBBQREbOuiAkTQQTBxIp5VXDNcRe/a2ANiKuLaRWzrgEMYMa4q6iImLOAWXFRUAkqEiVPvX/40j+HmYGZYaieges6p8+hq6qrb3o6zT1PPZVJkiQJAAAAAEhRlVwHAAAAAGDdo5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCWIMuueSSyGQyJdo2k8nEJZdcsmYDraavvvoqunfvHg0aNIhMJhNPPPFEriPlRFE/14033jj69etXYFlRj9ewYcMik8nEN998k17g/19leI6Vh6Ie486dO0fnzp1zlonCVnzNvPzyy5HJZOLll18u9/uaN29eNGvWLIYPH17u+64ojjjiiDjssMNymmHKlClRs2bNeP3113Oao6L65ptvIpPJxLBhw9bYfdx2222x0UYbxaJFi9bYfQCUJ6UUsE5Y/kvq8ku1atVigw02iH79+sX333+f63iVRt++feOjjz6Kyy+/PO67777YcccdY8SIETFkyJBcR6uQinq81rT//ve/a03xdMstt0Qmk4lddtkl11FS98MPP8Qll1wS48aNK3L9008/HZ06dYpmzZpF7dq1Y5NNNonDDjssnnvuuXSDVhLXX3991KtXL4444ohsMVCSSy7K47I677zz4rHHHovx48fnLMOll14au+yyS+y+++5Frj/ssMMik8nEeeedl3Kyklm2bFm0bNkyMplMPPvss2XeTy4/F/v16xeLFy+O22+/vUTb+wwHcq1argMApOnSSy+NNm3axMKFC+Ott96KYcOGxWuvvRYff/xx1KxZs9zv78ILL4zzzz+/3PebCwsWLIg333wzLrjggjjjjDOyy0eMGBEff/xx/PWvf81duArgiy++iCpV/t/feop7vI455pg44ogjIi8vb43k+O9//xs333xzkcXUggULolq1yvPRP3z48Nh4443jnXfeiYkTJ8amm25a5n09//zz5Zhszfvhhx9i0KBBsfHGG8f2229fYN0111wT55xzTnTq1CkGDBgQtWvXjokTJ8aLL74YDz74YOy77765CV1BLVmyJK6//vro379/VK1aNZo2bRr33XdfgW3+9a9/xXfffRfXXXddgeVNmzZNM+pqad++fey4447xr3/9K+69997U7//HH3+Me+65J+65554i18+ZMyeefvrp2HjjjeOBBx6IK664osQjidPy0ksvxdSpU2PjjTeO4cOHR8+ePcu0n+I+F1u3bh0LFiyI6tWrl0PaotWsWTP69u0b1157bZx55pmrfIx9hgO5Vnm+mQKUg549e2ZHq5xwwgnRpEmTuPLKK+Opp55aI4c9VKtWrVKVACvz448/RkREw4YN1/h95efnx+LFi9dIUbimrFgyFfd4Va1aNapWrZpWrAIq0+M5adKkeOONN2LkyJFx8sknx/Dhw2PgwIFl3l+NGjXKMV3uLF26NP7xj3/EPvvsU2TRNmPGjBykqtieeeaZ+PHHH7Pv8XXq1Imjjz66wDYPPvhg/PLLL4WWl8XSpUsjPz8/tefc/Pnzo06dOhHx20ikgQMHxi233BJ169ZN5f6Xu//++6NatWpxwAEHFLn+sccei2XLlsXdd98dXbt2jVdeeSU6deqUasZVuf/++2OHHXaIvn37xt///vcCj215yGQyqbwPH3bYYXHVVVfFmDFjomvXrmv8/gBWh8P3gHXannvuGRERX3/9dYHln3/+eRxyyCHRqFGjqFmzZuy4447x1FNPFdhmyZIlMWjQoGjXrl3UrFkzGjduHHvssUe88MIL2W2Kmnto0aJF0b9//2jatGnUq1cvDjzwwPjuu+8KZevXr19svPHGhZYXN0/V/fffHx06dIhatWpFo0aN4ogjjogpU6as8jH49ttv47TTTovNN988atWqFY0bN45DDz20wGErl1xySbRu3ToiIs4555zIZDKx8cYbR+fOneM///lPfPvtt9nDXX6fedGiRTFw4MDYdNNNIy8vL1q1ahXnnntuobkuMplMnHHGGTF8+PDYeuutIy8vb6WHIW288cax//77x8svvxw77rhj1KpVK7bddtvsXDQjR46MbbfdNmrWrBkdOnSIDz/8sNA+Xnrppdhzzz2jTp060bBhwzjooIPis88+K7Tda6+9FjvttFPUrFkz2rZtW+whEb+fH6e4xyui6PmOIiKeffbZ6NSpU9SrVy/q168fO+20U4wYMSK7/tVXX41DDz00Ntpoo+xj2b9//1iwYEF2m379+sXNN9+cfUyXX37/OK84gurDDz+Mnj17Rv369aNu3bqx9957x1tvvVVgm+WZX3/99Tj77LOjadOmUadOnfjjH/+YLd/K2/Dhw2O99daL/fbbLw455JBi5wL65JNPomvXrlGrVq3YcMMN47LLLov8/PxC2604p9TixYvj4osvjg4dOkSDBg2iTp06seeee8aYMWMK3TY/Pz+uv/767HOqadOmse+++8Z7771XYLuSvAY7d+4c22yzTXz66afRpUuXqF27dmywwQZx1VVXZbd5+eWXY6eddoqIiOOOOy77cxw2bFj89NNPMWfOnGIPj2rWrFmB6yV9DRb3vlSWeciuueaa2G233aJx48ZRq1at6NChQzz66KOl2kd5euKJJ2LjjTeOtm3blup2M2bMiOOPPz7WX3/9qFmzZvzhD38oNApo+aGA11xzTQwZMiTatm0beXl58emnn2bnyHrooYfi73//ezRv3jzq1KkTBx54YJHvzY888kj2+dOkSZM4+uijCx1e3q9fv6hbt258/fXX0atXr6hXr1786U9/yq7fZ599Yv78+QU+h9LyxBNPxC677FJsGTZ8+PDYZ599okuXLrHlllsW+Zou7vOtqPfN/Pz8uOSSS6Jly5ZRu3bt6NKlS3z66adFzu9XEgsWLIjHH388OzfXggUL4sknnyxy25W9X6/sc3HFOaWuueaayGQy8e233xa6jwEDBkSNGjXil19+yS57++23Y999940GDRpE7dq1o1OnTkXO39WhQ4do1KhRsfmXW9Vn+I033hhbb7111K5dO9Zbb73sIfsREY8++mhkMpkYO3Zsof3efvvtkclk4uOPP17p/QNEGCkFrOOWf8Fdb731sss++eST2H333WODDTaI888/P+rUqRMPP/xw9O7dOx577LH44x//GBG/fXkePHhwnHDCCbHzzjvHnDlz4r333osPPvgg9tlnn2Lv84QTToj7778/jjrqqNhtt93ipZdeiv3222+1/h+XX355XHTRRXHYYYfFCSecED/++GPceOONsddee8WHH3640tFN7777brzxxhtxxBFHxIYbbhjffPNN3HrrrdG5c+f49NNPo3bt2nHwwQdHw4YNo3///nHkkUdGr169om7dulGnTp2YPXt2gcNelv9Ckp+fHwceeGC89tprcdJJJ8WWW24ZH330UVx33XXx5ZdfFpok/aWXXoqHH344zjjjjGjSpEmRhdzvTZw4MY466qg4+eST4+ijj45rrrkmDjjggLjtttvi73//e5x22mkRETF48OA47LDDChxe9+KLL0bPnj1jk002iUsuuSQWLFgQN954Y+y+++7xwQcfZO/7o48+iu7du0fTpk3jkksuiaVLl8bAgQNj/fXXX2m24h6v4gwbNiz+/Oc/x9Zbbx0DBgyIhg0bxocffhjPPfdcHHXUURHx2y+sv/76a5x66qnRuHHjeOedd+LGG2+M7777Lh555JGIiDj55JPjhx9+iBdeeKHQ4UlF+eSTT2LPPfeM+vXrx7nnnhvVq1eP22+/PTp37hxjx44tNJfTmWeeGeutt14MHDgwvvnmmxgyZEicccYZ8dBDD63yvkpr+PDhcfDBB0eNGjXiyCOPjFtvvTXefffdbFkTETFt2rTo0qVLLF26NPta/fe//x21atVa5f7nzJkTd955Zxx55JFx4oknxty5c+Ouu+6KHj16xDvvvFPgkLnjjz8+hg0bFj179owTTjghli5dGq+++mq89dZb2ZGXpXkN/vLLL7HvvvvGwQcfHIcddlg8+uijcd5558W2224bPXv2jC233DIuvfTSuPjii+Okk07Klue77bZbNGvWLGrVqhVPP/10nHnmmdGoUaNi/4+leQ2W5/vS9ddfHwceeGD86U9/isWLF8eDDz4Yhx56aDzzzDOr/V5XFm+88UbssMMOpbrNggULonPnzjFx4sQ444wzok2bNvHII49Ev379YtasWfGXv/ylwPZDhw6NhQsXxkknnRR5eXnRqFGjmDVrVkT89txYPo/SjBkzYsiQIdGtW7cYN25c9rk6bNiwOO6442KnnXaKwYMHx/Tp0+P666+P119/vdDzZ+nSpdGjR4/YY4894pprronatWtn12211VZRq1ateP3117OfVWlYsmRJvPvuu3HqqacWuf6HH36IMWPGZEu9I488Mq677rq46aabyjyibMCAAXHVVVfFAQccED169Ijx48dHjx49YuHChWXa31NPPRXz5s2LI444Ipo3bx6dO3eO4cOHZ9+Dl1vV+/UFF1xQ7Ofiig477LA499xz4+GHH45zzjmnwLqHH344unfvnv1+8tJLL0XPnj2jQ4cOMXDgwKhSpUoMHTo0unbtGq+++mrsvPPOBW6/ww47rHLC+ZVlveOOO+Kss86KQw45JP7yl7/EwoULY8KECfH222/HUUcdFfvtt1/UrVs3Hn744UIj3h566KHYeuutY5tttlnp/QNEREQCsA4YOnRoEhHJiy++mPz444/JlClTkkcffTRp2rRpkpeXl0yZMiW77d57751su+22ycKFC7PL8vPzk9122y1p165ddtkf/vCHZL/99lvp/Q4cODD5/VvtuHHjkohITjvttALbHXXUUUlEJAMHDswu69u3b9K6detV7vObb75Jqlatmlx++eUFtvvoo4+SatWqFVq+ol9//bXQsjfffDOJiOTee+/NLps0aVISEcnVV19dYNv99tuvyJz33XdfUqVKleTVV18tsPy2225LIiJ5/fXXs8siIqlSpUryySefrDTrcq1bt04iInnjjTeyy0aNGpVERFKrVq3k22+/zS6//fbbk4hIxowZk122/fbbJ82aNUt+/vnn7LLx48cnVapUSY499tjsst69eyc1a9YssL9PP/00qVq1arLiR2jr1q2Tvn37Zq8X93gtfy5OmjQpSZIkmTVrVlKvXr1kl112SRYsWFBg2/z8/Oy/i/o5DR48OMlkMgXynX766YWyLbfic6x3795JjRo1kq+//jq77Icffkjq1auX7LXXXoUyd+vWrUCm/v37J1WrVk1mzZpV5P2V1XvvvZdERPLCCy8kSfLb47Dhhhsmf/nLXwps99e//jWJiOTtt9/OLpsxY0bSoEGDAo9xkiRJp06dkk6dOmWvL126NFm0aFGB/f3yyy/J+uuvn/z5z3/OLnvppZeSiEjOOuusQjmXPxaleQ126tSp0Gtr0aJFSfPmzZM+ffpkl7377rtJRCRDhw4tdL8XX3xxEhFJnTp1kp49eyaXX3558v777xfarqSvwdK8L5XEis/VxYsXJ9tss03StWvXAstXfM2MGTOm0Gt1dS1ZsiTJZDLJ//3f/610uxXfx4YMGZJERHL//fdnly1evDjp2LFjUrdu3WTOnDlJkvy/13n9+vWTGTNmFNjn8v/PBhtskN0+SZLk4YcfTiIiuf7667P7bdasWbLNNtsUeA945plnkohILr744uyyvn37JhGRnH/++cX+XzbbbLOkZ8+eK/3/lreJEycmEZHceOONRa6/5pprklq1amUfhy+//DKJiOTxxx8vsN2Kn2/Lrfi+OW3atKRatWpJ7969C2x3ySWXJBFR4HlVUvvvv3+y++67Z6//+9//TqpVq1bg51rS9+viPheXP19+/7ru2LFj0qFDhwLbvfPOOwXeJ/Lz85N27dolPXr0KPS50KZNm2SfffYpdF8nnXRSUqtWrVX+v4vLetBBByVbb731Sm975JFHJs2aNUuWLl2aXTZ16tSkSpUqyaWXXrrK+wZIkiRx+B6wTunWrVs0bdo0WrVqFYccckjUqVMnnnrqqdhwww0jImLmzJnx0ksvxWGHHRZz586Nn376KX766af4+eefo0ePHvHVV19lD6do2LBhfPLJJ/HVV1+V+P7/+9//RkTEWWedVWD56kwwOnLkyMjPz4/DDjssm/enn36K5s2bR7t27Yo8HOn3fj+qZMmSJfHzzz/HpptuGg0bNowPPvigzLkeeeSR2HLLLWOLLbYokGv5/BYr5urUqVNstdVWJd7/VlttFR07dsxeXz6qp2vXrrHRRhsVWv6///0vIiKmTp0a48aNi379+hUYZbLddtvFPvvsk/0ZLVu2LEaNGhW9e/cusL8tt9wyevToUeKcq/LCCy/E3Llz4/zzzy8018jvD2P5/c9p/vz58dNPP8Vuu+0WSZIUeXjiqixbtiyef/756N27d2yyySbZ5S1atIijjjoqXnvttZgzZ06B25x00kkFMu25556xbNmyIg89WR3Dhw+P9ddfP7p06RIRvz0Ohx9+eDz44IOxbNmy7Hb//e9/Y9dddy0wQqBp06YFDmcqTtWqVbMjNPLz82PmzJmxdOnS2HHHHQs87x977LHIZDJFzme1/LEo7Wuwbt26BeYuqlGjRuy8887Z5+iqDBo0KEaMGBHt27ePUaNGxQUXXBAdOnSIHXbYocAhqCV9DZb3+9Lvn6u//PJLzJ49O/bcc8/Vej8pq5kzZ0aSJAVGw5bEf//732jevHkceeSR2WXVq1ePs846K+bNm1fokKU+ffoUOyn6scceG/Xq1cteP+SQQ6JFixbZx/29996LGTNmxGmnnVbgPWC//faLLbbYIv7zn/8U2mdxI5Iifhv5+9NPP5XsP1pOfv755+x9F2X48OGx3377ZR+Hdu3aRYcOHYo9LHdVRo8eHUuXLs2OiF3uzDPPLNP+fv755xg1alSBn3efPn0ik8nEww8/nF1W0vfr0jj88MPj/fffLzCNwEMPPRR5eXlx0EEHRUTEuHHj4quvvoqjjjoqfv755+xref78+bH33nvHK6+8Uuiw5fXWWy8WLFgQv/76a5lyNWzYML777rt49913V5p9xowZ2UPnI347rC8/Pz8OP/zwMt0vsO5RSgHrlJtvvjleeOGFePTRR6NXr17x008/FZigeuLEiZEkSVx00UXRtGnTApflv5Qun0j40ksvjVmzZsVmm20W2267bZxzzjkxYcKEld7/t99+G1WqVCk0t8nmm29e5v/TV199FUmSRLt27Qpl/uyzz1Y58fGCBQvi4osvjlatWkVeXl40adIkmjZtGrNmzYrZs2evVq5PPvmkUKbNNtssIgpPyNymTZtS7f/3RVFERIMGDSIiolWrVkUuXz4vx/ICpajHfMstt8x+0f/xxx9jwYIF0a5du0Lbrc7Pa0XLfxFZ1WEOkydPzhZpdevWjaZNm2YPmSjLz+nHH3+MX3/9tdjHIT8/v9C8Nys+5st/Af39nCcrWrBgQUybNq3AZWWWLVsWDz74YHTp0iUmTZoUEydOjIkTJ8Yuu+wS06dPj9GjR2e3/fbbb1fr53PPPffEdtttl50TrmnTpvGf//ynwOP59ddfR8uWLVd6mFxpX4MbbrhhoV9g11tvvZU+jis68sgj49VXX41ffvklnn/++TjqqKPiww8/jAMOOCB7+FJJX4Pl/b70zDPPxK677ho1a9aMRo0aRdOmTePWW29drfeT1ZUkSam2X/7c+v0ZNSN+e20sX/97K3v/WvE5mslkYtNNN80ePr6y96Qtttii0H1Vq1Yt+4eUoiRJkrOz2hX1OH/22Wfx4Ycfxu677559PU+cODE6d+4czzzzTKHyuySWPyYrnpGzUaNGpS4gI34rgZYsWRLt27fP5ps5c2bssssuBYqzkr5fl8ahhx4aVapUyR4GnSRJPPLII9m5/iIi+8evvn37Fno933nnnbFo0aJCr6/lP4uyPhfOO++8qFu3buy8887Rrl27OP300wsdDrh8fqvfH8L90EMPxfbbb599nwFYFXNKAeuUnXfeOTsHTO/evWOPPfaIo446Kr744ouoW7du9i+Nf/vb34odDbP8S/Bee+0VX3/9dTz55JPx/PPPx5133hnXXXdd3HbbbXHCCSesdtbivkj+fqRIxG+jPDKZTDz77LNFntVtVWdgOvPMM2Po0KHx17/+NTp27BgNGjSITCYTRxxxRJETRpdUfn5+bLvttnHttdcWuX7F8qgk8wD9XnFnsCtueWl/Ka1Ili1bFvvss0/MnDkzzjvvvNhiiy2iTp068f3330e/fv1W6+dUGmV5bB966KE47rjjSrz98lOyP/jgg/Hggw8WWj98+PDo3r17CRMX7/77749+/fpF796945xzzolmzZpF1apVY/DgwYVOfLAqpX0NludztH79+rHPPvvEPvvsE9WrV4977rkn3n777ejUqVOpX4Pl4dVXX40DDzww9tprr7jllluiRYsWUb169Rg6dGiBifvT0qhRo8hkMqUq/MqitO9fqyMvL69QWfZ7v/zyS5Fl7ZrUuHHj7H2v6P7774+IiP79+0f//v0LrX/sscey7xEl/dwrb8uLp+JOIPC///2vwIjS8tSyZcvYc8894+GHH46///3v8dZbb8XkyZPjyiuvzG6z/D3+6quvLjDf3e+t+D7zyy+/RO3atcv83Nxyyy3jiy++iGeeeSaee+65eOyxx+KWW26Jiy++OAYNGhQRvz0Xe/fuHY8//njccsstMX369Hj99dfjn//8Z5nuE1g3KaWAddbyX0C7dOkSN910U5x//vnZL53Vq1ePbt26rXIfjRo1iuOOOy6OO+64mDdvXuy1115xySWXFFtKtW7dOvLz8+Prr78u8FfxL774otC26623Xnai3N9b8a/mbdu2jSRJok2bNmX6y+Sjjz4affv2jX/961/ZZQsXLizyvotS3C8Rbdu2jfHjx8fee++ds7/aF2X5WfGKesw///zzaNKkSdSpUydq1qwZtWrVKvLwzKJuW1bLR6d8/PHHhf7qv9xHH30UX375Zdxzzz1x7LHHZpcXdYatkj7WTZs2jdq1axf7OFSpUqVcSosePXqU6kxgw4cPj2bNmmXPIvh7I0eOjMcffzxuu+22qFWrVrRu3brMP59HH300Ntlkkxg5cmSBx2zFw/Tatm0bo0aNipkzZxY7Wmp1X4NFKctrZscdd4x77rknpk6dms1Vktdgad6XVuWxxx6LmjVrxqhRowqMQh06dGip91UeqlWrFm3bto1JkyaV6natW7eOCRMmRH5+foEC6PPPP8+uL6kVn6NJksTEiRNju+22K7CvL774Into5XJffPFFqe5r6dKlMWXKlDjwwANLfJvysNFGG0WtWrUKPc5JksSIESOiS5cuhQ61i4j4xz/+EcOHD8+WUstHOc2aNavA5O4rfu4tf0wmTpxYYJTazz//XOoCctKkSfHGG2/EGWecUWjC7vz8/DjmmGNixIgRceGFF5bo/Tqi9K/fww8/PE477bT44osv4qGHHoratWvHAQcckF2//H7r169fou8my/9fy0f2rczKstapUycOP/zwOPzww2Px4sVx8MEHx+WXXx4DBgzIHr54+OGHxz333BOjR4+Ozz77LJIkcegeUCoO3wPWaZ07d46dd945hgwZEgsXLoxmzZpF586d4/bbb8/+Yvd7P/74Y/bfy+fQWK5u3bqx6aabFjrV+u/17NkzIiJuuOGGAsuHDBlSaNu2bdvG7NmzCxwSOHXq1Hj88ccLbHfwwQdH1apVY9CgQYVGWiRJUijniqpWrVrodjfeeGOJ/zK9/Ax8KzrssMPi+++/jzvuuKPQugULFsT8+fNLtP/y1qJFi9h+++3jnnvuKVC8ffzxx/H8889Hr169IuK3x6VHjx7xxBNPxOTJk7PbffbZZzFq1Khyy9O9e/eoV69eDB48uNBZo5b/XJaPrPn9zylJkrj++usL7a9OnToREassFatWrRrdu3ePJ598ssBp1qdPnx4jRoyIPfbYI3voyOpo0aJFdOvWrcClOAsWLIiRI0fG/vvvH4ccckihyxlnnBFz586Np556KiIievXqFW+99Va888472X38+OOPJZqnpqjH9O23344333yzwHZ9+vSJJEmyIwN+b/ltV/c1WJTifo6//vproYzLPfvssxHx/w4DK+lrsDTvS6tStWrVyGQyBd4/vvnmm0Jn20xTx44d47333ivVbXr16hXTpk0rcFjS0qVL48Ybb4y6desWKi9W5t577425c+dmrz/66KMxderU7OO+4447RrNmzeK2224r8Pnx7LPPxmeffVaqMxZ++umnsXDhwthtt91KfJvyUL169dhxxx0LPc6vv/56fPPNN3HccccV+Zo+/PDDY8yYMfHDDz9ExP8rX1555ZXsPubPn589a99ye++9d1SrVi1uvfXWAstvuummUmdf/n5x7rnnFsp32GGHRadOnbLblOT9OqL4z8Xi9OnTJ6pWrRoPPPBAPPLII7H//vtn3wMiIjp06BBt27aNa665JubNm1fo9r//brLcBx98UOh5MHny5GyxuqqsK75v1ahRI7baaqtIkiSWLFmSXd6tW7do1KhRPPTQQ/HQQw/FzjvvXOrD8YF1m5FSwDrvnHPOiUMPPTSGDRsWp5xyStx8882xxx57xLbbbhsnnnhibLLJJjF9+vR4880347vvvovx48dHxG8TbXfu3Dk6dOgQjRo1ivfeey8effTROOOMM4q9r+233z6OPPLIuOWWW2L27Nmx2267xejRo2PixImFtj3iiCPivPPOiz/+8Y9x1llnxa+//hq33nprbLbZZgUmDG7btm1cdtllMWDAgPjmm2+id+/eUa9evZg0aVI8/vjjcdJJJ8Xf/va3YjPtv//+cd9990WDBg1iq622ijfffDNefPHF7OEYq9KhQ4d46KGH4uyzz46ddtop6tatGwcccEAcc8wx8fDDD8cpp5wSY8aMid133z2WLVsWn3/+eTz88MMxatSo7KGUabv66qujZ8+e0bFjxzj++ONjwYIFceONN0aDBg3ikksuyW43aNCgeO6552LPPfeM0047LftL6dZbb73K+cNKqn79+nHdddfFCSecEDvttFMcddRRsd5668X48ePj119/jXvuuSe22GKLaNu2bfztb3+L77//PurXrx+PPfZYkSMCOnToEBG/TVrdo0ePqFq1ahxxxBFF3vdll10WL7zwQuyxxx5x2mmnRbVq1eL222+PRYsWxVVXXVUu/7/SeOqpp2Lu3LnFjvLYddddo2nTpjF8+PA4/PDD49xzz4377rsv9t133/jLX/4SderUiX//+9/ZUS4rs//++8fIkSPjj3/8Y+y3334xadKkuO2222KrrbYq8Etfly5d4phjjokbbrghvvrqq9h3330jPz8/Xn311ejSpUucccYZq/0aLErbtm2jYcOGcdttt0W9evWiTp06scsuu0S9evVit912i1133TX23XffaNWqVcyaNSueeOKJePXVV6N3797Rvn37iIgSvwZL8760Kvvtt19ce+21se+++8ZRRx0VM2bMiJtvvjk23XTTcnvNlNZBBx0U9913X3z55ZclHsl20kknxe233x79+vWL999/PzbeeON49NFH4/XXX48hQ4YUmLh8VRo1ahR77LFHHHfccTF9+vQYMmRIbLrppnHiiSdGxG+FzpVXXhnHHXdcdOrUKY488siYPn16XH/99bHxxhsXechbcV544YWoXbt27LPPPiW+TXk56KCD4oILLog5c+ZkC+3hw4dH1apViy3WDjzwwLjgggviwQcfjLPPPju6d+8eG220URx//PFxzjnnRNWqVePuu++Opk2bFvjjwPrrrx9/+ctf4l//+lcceOCBse+++8b48ePj2WefjSZNmpRqpNLw4cNj++23L3Zk6IEHHhhnnnlmfPDBB7HDDjus8v06ovjPxeI0a9YsunTpEtdee23MnTu30EijKlWqxJ133hk9e/aMrbfeOo477rjYYIMN4vvvv48xY8ZE/fr14+mnn85u//7778fMmTOzE6Uvd+yxx8bYsWMLFGjFZe3evXs0b948dt9991h//fXjs88+i5tuuqnAhPURvz1/Dz744HjwwQdj/vz5cc0115T4sQeIiGLOGQ2wlll+Oul333230Lply5Ylbdu2Tdq2bZs9rfHXX3+dHHvssUnz5s2T6tWrJxtssEGy//77J48++mj2dpdddlmy8847Jw0bNkxq1aqVbLHFFsnll1+eLF68OLtNUae3XrBgQXLWWWcljRs3TurUqZMccMAByZQpU4o89frzzz+fbLPNNkmNGjWSzTffPLn//vuLPWX2Y489luyxxx5JnTp1kjp16iRbbLFFcvrppydffPHFSh+bX375JTnuuOOSJk2aJHXr1k169OiRfP7554VO1778VNZXX311gdvPmzcvOeqoo5KGDRsmEVHg1NKLFy9OrrzyymTrrbdO8vLykvXWWy/p0KFDMmjQoGT27NnZ7SIiOf3001ea8/dat26d7LfffoWWF7Wf4nK/+OKLye67757UqlUrqV+/fnLAAQckn376aaF9jh07NunQoUNSo0aNZJNNNkluu+22In8GJX28Vjy1+XJPPfVUsttuu2Xz7LzzzskDDzyQXf/pp58m3bp1S+rWrZs0adIkOfHEE5Px48cXOr340qVLkzPPPDNp2rRpkslkCuQs6jn2wQcfJD169Ejq1q2b1K5dO+nSpUvyxhtvFJl5xdfP8lPejxkzptDjVhYHHHBAUrNmzWT+/PnFbtOvX7+kevXqyU8//ZQkSZJMmDAh6dSpU1KzZs1kgw02SP7xj38kd911V6HHuFOnTkmnTp2y1/Pz85N//vOfSevWrZO8vLykffv2yTPPPJP07du30OnRly5dmlx99dXJFltskdSoUSNp2rRp0rNnz+T9998vsF1JXoOdOnUq8jTrRd3vk08+mWy11VZJtWrVsj/nJUuWJHfccUfSu3fvbPbatWsn7du3T66++upk0aJFBfZR0tdgad6XVuWuu+5K2rVrl+Tl5SVbbLFFMnTo0BK9Zsr7+bTcokWLkiZNmiT/+Mc/it1mv/32K/T4T58+PfveWKNGjWTbbbct8FpLkuJf50ny//4/DzzwQDJgwICkWbNmSa1atZL99tsv+fbbbwtt/9BDDyXt27dP8vLykkaNGiV/+tOfku+++67ANn379k3q1KlT7P9jl112SY4++uhi169J06dPT6pVq5bcd999SZL89txr3Lhxsueee670dm3atEnat2+fvf7+++8nu+yyS1KjRo1ko402Sq699toi3zeXLl2aXHTRRUnz5s2TWrVqJV27dk0+++yzpHHjxskpp5xSoszvv/9+EhHJRRddVOw233zzTRIRSf/+/bPLVvV+Xdzn4vLny4rPoyRJkjvuuCOJiKRevXrJggULiszy4YcfJgcffHDSuHHjJC8vL2ndunVy2GGHJaNHjy6w3XnnnZdstNFGSX5+foHlnTp1KvQ6LC7r7bffnuy1117Z+2rbtm1yzjnnFHjfWO6FF15IIiLJZDLJlClTin0sAYqSSZJKPPMrAEAJ7LnnnpGXlxcvvvhirqNUKplMJgYOHFhgBGFl9I9//COGDh0aX331VbETzZe3l19+Obp06RKPPPJIHHLIIWv8/saNGxc77LBDfPDBB8VOhr2mHX/88fHll1/Gq6++mpP7nzVrVqy33npx2WWXxQUXXJCTDLm2aNGi2HjjjeP888+Pv/zlL7mOA7BK5pQCANZ6U6dOjSZNmuQ6BjnSv3//mDdvXpFndFxbXHHFFXHIIYfkrJCK+O1EAe+++268/vrra/y+FixYUGjZ8nnQOnfuvMbvv6IaOnRoVK9ePU455ZRcRwEoEXNKAQBrrTfeeCNGjhwZX3/9dZx33nm5jlNpLVu2rMjJlH+vbt26hU5LX1YLFixY5UTRixcvjho1ahS7vlatWtGgQYNsthkzZpRLtoqqqMLtxx9/XOlJK1b1GC6/7cpGl/3+577RRhsVmgB8TXnooYdi2LBh0atXr6hbt2689tpr8cADD0T37t1j9913j4iIadOmrXQfv3+OrC1OOeUUhRRQqSilAIC11h133BHPPvts/PWvf82edp7SmzJlyirPqFWeh/k99NBDq/3z6tu3bwwbNqxc8lRWO+20U3z77bfFru/UqVOMHTt2te4jV4d3brfddlGtWrW46qqrYs6cOdnJzy+77LLsNi1atFjpPjxHAHIvp3NK3XrrrXHrrbdmT0W99dZbx8UXX5w9RW7nzp0LfVCefPLJcdttt6UdFQBgnbVw4cJ47bXXVrrNJptsEptsskm53N/UqVPjk08+Wek2v/zyS6y33nrFrm/ZsmVstdVW5ZKnsnr99deLPMxtufXWW6/Is3gut3zUU82aNYvdpjx/7uVtVXPIeY4A5F5OS6mnn346qlatGu3atYskSeKee+6Jq6++Oj788MPYeuuto3PnzrHZZpvFpZdemr1N7dq1s6eZBQAAAKByyunhewcccECB65dffnnceuut8dZbb8XWW28dEb+VUM2bN89FPAAAAADWkAozp9SyZcvikUceifnz50fHjh2zy4cPHx73339/NG/ePA444IC46KKLonbt2sXuZ9GiRbFo0aLs9fz8/Jg5c2Y0btw4MpnMGv0/AAAAAKzrkiSJuXPnRsuWLaNKlSrFbpfzUuqjjz6Kjh07xsKFC6Nu3brx+OOPZ4/tPuqoo6J169bRsmXLmDBhQpx33nnxxRdfxMiRI4vd3+DBg2PQoEFpxQcAAACgCFOmTIkNN9yw2PU5nVMq4rdT0U6ePDlmz54djz76aNx5550xduzYIicdfOmll2LvvfeOiRMnRtu2bYvc34ojpWbPnh0bbbRRTJkyxVxUAAAAAGvYnDlzolWrVjFr1qxo0KBBsdvlvJRaUbdu3aJt27Zx++23F1o3f/78qFu3bjz33HPRo0ePEu1vzpw50aBBg5g9e7ZSCgAAAGANK2kXU/yBfTmSn59fYKTT740bNy4iIlq0aJFiIgAAAADKW07nlBowYED07NkzNtpoo5g7d26MGDEiXn755Rg1alR8/fXXMWLEiOjVq1c0btw4JkyYEP3794+99tortttuu1zGBgAAAGA15bSUmjFjRhx77LExderUaNCgQWy33XYxatSo2GeffWLKlCnx4osvxpAhQ2L+/PnRqlWr6NOnT1x44YW5jAwAAABAOahwc0qVN3NKAQAAAKSn0s4pBQAAAMDaTykFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkrlquAwAAQLEymdJtnyRrJgcAUO6MlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFJXLdcBAACgvFzx4U8l3vb89k3WYBIAYFWMlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKnlAIAAAAgdUopAAAAAFKX01Lq1ltvje222y7q168f9evXj44dO8azzz6bXb9w4cI4/fTTo3HjxlG3bt3o06dPTJ8+PYeJAQAAACgPOS2lNtxww7jiiivi/fffj/feey+6du0aBx10UHzyyScREdG/f/94+umn45FHHomxY8fGDz/8EAcffHAuIwMAAABQDjJJkiS5DvF7jRo1iquvvjoOOeSQaNq0aYwYMSIOOeSQiIj4/PPPY8stt4w333wzdt111xLtb86cOdGgQYOYPXt21K9ff01GBwCgvGUypdr8ig9+LPG257dvUto0AEAJlLSLqTBzSi1btiwefPDBmD9/fnTs2DHef//9WLJkSXTr1i27zRZbbBEbbbRRvPnmm8XuZ9GiRTFnzpwCFwAAAAAqlpyXUh999FHUrVs38vLy4pRTTonHH388ttpqq5g2bVrUqFEjGjZsWGD79ddfP6ZNm1bs/gYPHhwNGjTIXlq1arWG/wcAAAAAlFbOS6nNN988xo0bF2+//Xaceuqp0bdv3/j000/LvL8BAwbE7Nmzs5cpU6aUY1oAAAAAykO1XAeoUaNGbLrpphER0aFDh3j33Xfj+uuvj8MPPzwWL14cs2bNKjBaavr06dG8efNi95eXlxd5eXlrOjYAAAAAqyHnI6VWlJ+fH4sWLYoOHTpE9erVY/To0dl1X3zxRUyePDk6duyYw4QAAAAArK6cjpQaMGBA9OzZMzbaaKOYO3dujBgxIl5++eUYNWpUNGjQII4//vg4++yzo1GjRlG/fv0488wzo2PHjiU+8x4AAAAAFVNOS6kZM2bEscceG1OnTo0GDRrEdtttF6NGjYp99tknIiKuu+66qFKlSvTp0ycWLVoUPXr0iFtuuSWXkQEAAAAoB5kkSZJch1iT5syZEw0aNIjZs2dH/fr1cx0HAIDSyGRKtfkVH/xY4m3Pb9+ktGkAgBIoaRdT4eaUAgAAAGDtp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHU5LaUGDx4cO+20U9SrVy+aNWsWvXv3ji+++KLANp07d45MJlPgcsopp+QoMQAAAADlIael1NixY+P000+Pt956K1544YVYsmRJdO/ePebPn19guxNPPDGmTp2avVx11VU5SgwAAABAeaiWyzt/7rnnClwfNmxYNGvWLN5///3Ya6+9sstr164dzZs3TzseAAAAAGtIhZpTavbs2RER0ahRowLLhw8fHk2aNIltttkmBgwYEL/++msu4gEAAABQTnI6Uur38vPz469//Wvsvvvusc0222SXH3XUUdG6deto2bJlTJgwIc4777z44osvYuTIkUXuZ9GiRbFo0aLs9Tlz5qzx7AAAAACUToUppU4//fT4+OOP47XXXiuw/KSTTsr+e9ttt40WLVrE3nvvHV9//XW0bdu20H4GDx4cgwYNWuN5AQAAACi7CnH43hlnnBHPPPNMjBkzJjbccMOVbrvLLrtERMTEiROLXD9gwICYPXt29jJlypRyzwsAAADA6snpSKkkSeLMM8+Mxx9/PF5++eVo06bNKm8zbty4iIho0aJFkevz8vIiLy+vPGMCAAAAUM5yWkqdfvrpMWLEiHjyySejXr16MW3atIiIaNCgQdSqVSu+/vrrGDFiRPTq1SsaN24cEyZMiP79+8dee+0V2223XS6jAwAAALAaclpK3XrrrRER0blz5wLLhw4dGv369YsaNWrEiy++GEOGDIn58+dHq1atok+fPnHhhRfmIC0AAAAA5SXnh++tTKtWrWLs2LEppQEAAAAgLRVionMAAAAA1i1KKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHU5LaUGDx4cO+20U9SrVy+aNWsWvXv3ji+++KLANgsXLozTTz89GjduHHXr1o0+ffrE9OnTc5QYAAAAgPKQ01Jq7Nixcfrpp8dbb70VL7zwQixZsiS6d+8e8+fPz27Tv3//ePrpp+ORRx6JsWPHxg8//BAHH3xwDlMDAAAAsLoySZIkuQ6x3I8//hjNmjWLsWPHxl577RWzZ8+Opk2bxogRI+KQQw6JiIjPP/88ttxyy3jzzTdj1113XeU+58yZEw0aNIjZs2dH/fr11/R/AQCA8pTJlGrzKz74scTbnt++SWnTAAAlUNIupkLNKTV79uyIiGjUqFFERLz//vuxZMmS6NatW3abLbbYIjbaaKN48803c5IRAAAAgNVXLdcBlsvPz4+//vWvsfvuu8c222wTERHTpk2LGjVqRMOGDQtsu/7668e0adOK3M+iRYti0aJF2etz5sxZY5kBAAAAKJsKM1Lq9NNPj48//jgefPDB1drP4MGDo0GDBtlLq1atyikhAAAAAOWlQpRSZ5xxRjzzzDMxZsyY2HDDDbPLmzdvHosXL45Zs2YV2H769OnRvHnzIvc1YMCAmD17dvYyZcqUNRkdAAAAgDLIaSmVJEmcccYZ8fjjj8dLL70Ubdq0KbC+Q4cOUb169Rg9enR22RdffBGTJ0+Ojh07FrnPvLy8qF+/foELAAAAABVLTueUOv3002PEiBHx5JNPRr169bLzRDVo0CBq1aoVDRo0iOOPPz7OPvvsaNSoUdSvXz/OPPPM6NixY4nOvAcAAABAxZTTUurWW2+NiIjOnTsXWD506NDo169fRERcd911UaVKlejTp08sWrQoevToEbfcckvKSQEAAAAoTzktpZIkWeU2NWvWjJtvvjluvvnmFBIBAAAAkIYKMdE5AAAAAOsWpRQAAAAAqVNKAQAAAJA6pRQAAAAAqVNKAQAAAJA6pRQAAAAAqStzKTVr1qy48847Y8CAATFz5syIiPjggw/i+++/L7dwAAAAAKydqpXlRhMmTIhu3bpFgwYN4ptvvokTTzwxGjVqFCNHjozJkyfHvffeW945AQAAAFiLlGmk1Nlnnx39+vWLr776KmrWrJld3qtXr3jllVfKLRwAAAAAa6cylVLvvvtunHzyyYWWb7DBBjFt2rTVDgUAAADA2q1MpVReXl7MmTOn0PIvv/wymjZtutqhAAAAAFi7lamUOvDAA+PSSy+NJUuWREREJpOJyZMnx3nnnRd9+vQp14AAAAAArH3KVEr961//innz5kWzZs1iwYIF0alTp9h0002jXr16cfnll5d3RgAAAADWMmU6+16DBg3ihRdeiNdeey0mTJgQ8+bNix122CG6detW3vkAAAAAWAuVqZRabo899og99tijvLIAAAAAsI4oUyl1ww03FLk8k8lEzZo1Y9NNN4299torqlatulrhAAAAAFg7lamUuu666+LHH3+MX3/9NdZbb72IiPjll1+idu3aUbdu3ZgxY0ZssskmMWbMmGjVqlW5BgYAAACg8ivTROf//Oc/Y6eddoqvvvoqfv755/j555/jyy+/jF122SWuv/76mDx5cjRv3jz69+9f3nkBAAAAWAuUaaTUhRdeGI899li0bds2u2zTTTeNa665Jvr06RP/+9//4qqrroo+ffqUW1AAAAAA1h5lGik1derUWLp0aaHlS5cujWnTpkVERMuWLWPu3Lmrlw4AAACAtVKZSqkuXbrEySefHB9++GF22YcffhinnnpqdO3aNSIiPvroo2jTpk35pAQAAABgrVKmUuquu+6KRo0aRYcOHSIvLy/y8vJixx13jEaNGsVdd90VERF169aNf/3rX+UaFgAAAIC1Q5nmlGrevHm88MIL8fnnn8eXX34ZERGbb755bL755tltunTpUj4JAQAAAFjrlKmUWm6LLbaILbbYoryyAAAAALCOKHMp9d1338VTTz0VkydPjsWLFxdYd+211652MAAAAADWXmUqpUaPHh0HHnhgbLLJJvH555/HNttsE998800kSRI77LBDeWcEAAAAYC1TponOBwwYEH/729/io48+ipo1a8Zjjz0WU6ZMiU6dOsWhhx5a3hkBAAAAWMuUqZT67LPP4thjj42IiGrVqsWCBQuibt26cemll8aVV15ZrgEBAAAAWPuUqZSqU6dOdh6pFi1axNdff51d99NPP5VPMgAAAADWWmWaU2rXXXeN1157Lbbccsvo1atX/N///V989NFHMXLkyNh1113LOyMAAAAAa5kylVLXXnttzJs3LyIiBg0aFPPmzYuHHnoo2rVr58x7AAAAAKxSmUqpTTbZJPvvOnXqxG233VZugQAAAABY+5VpTqlNNtkkfv7550LLZ82aVaCwAgAAAICilKmU+uabb2LZsmWFli9atCi+//771Q4FAAAAwNqtVIfvPfXUU9l/jxo1Kho0aJC9vmzZshg9enRsvPHG5RYOAAAAgLVTqUqp3r17R0REJpOJvn37FlhXvXr12HjjjeNf//pXuYUDAAAAYO1UqlIqPz8/IiLatGkT7777bjRp0mSNhAIAAABg7Vams+9NmjSpvHMAAAAAsA4pUykVETF69OgYPXp0zJgxIzuCarm77757tYMBAAAAsPYqUyk1aNCguPTSS2PHHXeMFi1aRCaTKe9cAAAAAKzFylRK3XbbbTFs2LA45phjyjsPAAAAAOuAKmW50eLFi2O33XYr7ywAAAAArCPKVEqdcMIJMWLEiPLOAgAAAMA6okyH7y1cuDD+/e9/x4svvhjbbbddVK9evcD6a6+9tlzCAQAAALB2KlMpNWHChNh+++0jIuLjjz8usM6k5wAAAACsSplKqTFjxpR3DgAAAADWIWWaU2q5iRMnxqhRo2LBggUREZEkSbmEAgAAAGDtVqZS6ueff4699947Nttss+jVq1dMnTo1IiKOP/74+L//+79yDQgAAADA2qdMpVT//v2jevXqMXny5Khdu3Z2+eGHHx7PPfdcuYUDAAAAYO1Upjmlnn/++Rg1alRsuOGGBZa3a9cuvv3223IJBgAAAMDaq0wjpebPn19ghNRyM2fOjLy8vNUOBQAAAMDarUyl1J577hn33ntv9nomk4n8/Py46qqrokuXLuUWDgAAAIC1U5kO37vqqqti7733jvfeey8WL14c5557bnzyyScxc+bMeP3118s7IwAAAABrmTKNlNpmm23iyy+/jD322CMOOuigmD9/fhx88MHx4YcfRtu2bcs7IwAAAABrmTKNlIqIaNCgQVxwwQXlmQUAAACAdUSZRkoNHTo0HnnkkULLH3nkkbjnnntWOxQAAAAAa7cylVKDBw+OJk2aFFrerFmz+Oc//7naoQAAAABYu5WplJo8eXK0adOm0PLWrVvH5MmTVzsUAAAAAGu3MpVSzZo1iwkTJhRaPn78+GjcuPFqhwIAAABg7VamUurII4+Ms846K8aMGRPLli2LZcuWxUsvvRR/+ctf4ogjjijvjAAAAACsZcp09r1//OMf8c0338Tee+8d1ar9tov8/Pw49thjzSkFAAAAwCqVupRKkiSmTZsWw4YNi8suuyzGjRsXtWrVim233TZat269JjICAAAAsJYpUym16aabxieffBLt2rWLdu3arYlcAAAAAKzFSj2nVJUqVaJdu3bx888/r4k8AAAAAKwDyjTR+RVXXBHnnHNOfPzxx+WdBwAAAIB1QJkmOj/22GPj119/jT/84Q9Ro0aNqFWrVoH1M2fOLJdwAAAAAKydylRKDRkypJxjAAAAALAuKVMp1bdv3/LOAQAAAMA6pExzSkVEfP3113HhhRfGkUceGTNmzIiIiGeffTY++eSTcgsHAAAAwNqpTKXU2LFjY9ttt4233347Ro4cGfPmzYuIiPHjx8fAgQPLNSAAAAAAa58ylVLnn39+XHbZZfHCCy9EjRo1ssu7du0ab731VrmFAwAAAGDtVKZS6qOPPoo//vGPhZY3a9Ysfvrpp9UOBQAAAMDarUylVMOGDWPq1KmFln/44YexwQYbrHYoAAAAANZuZSqljjjiiDjvvPNi2rRpkclkIj8/P15//fX429/+Fscee2x5ZwQAAABgLVOmUuqf//xnbLnllrHRRhvFvHnzYquttoq99tordtttt7jwwgvLOyMAAAAAa5lqpdk4Pz8/rr766njqqadi8eLFccwxx0SfPn1i3rx50b59+2jXrt2aygkAAADAWqRUpdTll18el1xySXTr1i1q1aoVI0aMiCRJ4u67715T+QAAAABYC5Xq8L177703brnllhg1alQ88cQT8fTTT8fw4cMjPz9/TeUDAAAAYC1UqlJq8uTJ0atXr+z1bt26RSaTiR9++KHcgwEAAACw9ipVKbV06dKoWbNmgWXVq1ePJUuWlGsoAAAAANZupZpTKkmS6NevX+Tl5WWXLVy4ME455ZSoU6dOdtnIkSPLLyEAAAAAa51SlVJ9+/YttOzoo48utzAAAAAArBtKVUoNHTp0TeUAAAAAYB1SqjmlAAAAAKA8KKUAAAAASF1OS6lXXnklDjjggGjZsmVkMpl44oknCqzv169fZDKZApd99903N2EBAAAAKDc5LaXmz58ff/jDH+Lmm28udpt99903pk6dmr088MADKSYEAAAAYE0o1UTn5a1nz57Rs2fPlW6Tl5cXzZs3TykRAAAAAGmo8HNKvfzyy9GsWbPYfPPN49RTT42ff/55pdsvWrQo5syZU+ACAAAAQMVSoUupfffdN+69994YPXp0XHnllTF27Njo2bNnLFu2rNjbDB48OBo0aJC9tGrVKsXEAAAAAJRETg/fW5Ujjjgi++9tt902tttuu2jbtm28/PLLsffeexd5mwEDBsTZZ5+dvT5nzhzFFAAAAEAFU6FHSq1ok002iSZNmsTEiROL3SYvLy/q169f4AIAAABAxVKpSqnvvvsufv7552jRokWuowAAAACwGnJ6+N68efMKjHqaNGlSjBs3Lho1ahSNGjWKQYMGRZ8+faJ58+bx9ddfx7nnnhubbrpp9OjRI4epAQAAAFhdOS2l3nvvvejSpUv2+vK5oPr27Ru33nprTJgwIe65556YNWtWtGzZMrp37x7/+Mc/Ii8vL1eRAQAAACgHOS2lOnfuHEmSFLt+1KhRKaYBAAAAIC2Vak4pAAAAANYOSikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUlct1wFgXXfFhz+Vavvz2zdZQ0kAAAAgPUZKAQAAAJA6pRQAAAAAqVNKAQAAAJA6pRQAAAAAqVNKAQAAAJA6pRQAAAAAqVNKAQAAAJA6pRQAAAAAqctpKfXKK6/EAQccEC1btoxMJhNPPPFEgfVJksTFF18cLVq0iFq1akW3bt3iq6++yk1YAAAAAMpNTkup+fPnxx/+8Ie4+eabi1x/1VVXxQ033BC33XZbvP3221GnTp3o0aNHLFy4MOWkAAAAAJSnarm88549e0bPnj2LXJckSQwZMiQuvPDCOOiggyIi4t577431118/nnjiiTjiiCPSjAoAAABAOaqwc0pNmjQppk2bFt26dcsua9CgQeyyyy7x5ptvFnu7RYsWxZw5cwpcAAAAAKhYKmwpNW3atIiIWH/99QssX3/99bPrijJ48OBo0KBB9tKqVas1mhMAAACA0quwpVRZDRgwIGbPnp29TJkyJdeRAAAAAFhBhS2lmjdvHhER06dPL7B8+vTp2XVFycvLi/r16xe4AAAAAFCxVNhSqk2bNtG8efMYPXp0dtmcOXPi7bffjo4dO+YwGQAAAACrK6dn35s3b15MnDgxe33SpEkxbty4aNSoUWy00Ubx17/+NS677LJo165dtGnTJi666KJo2bJl9O7dO3ehAQAAAFhtOS2l3nvvvejSpUv2+tlnnx0REX379o1hw4bFueeeG/Pnz4+TTjopZs2aFXvssUc899xzUbNmzVxFBgAAAKAc5LSU6ty5cyRJUuz6TCYTl156aVx66aUppgIAAABgTauwc0oBAAAAsPZSSgEAAACQOqUUAAAAAKlTSgEAAACQupxOdA4AVG6ZTKZU26/sBCcAAKxbjJQCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSVy3XAQAAAIqTyWRKtX2SJGsoCQDlzUgpAAAAAFKnlII1IZMp+QUAAADWQQ7fAwAA1h6l/aOfw/0AcsZIKQAAAABSp5QCAAAAIHVKKQAAAABSZ04poMyu+PCnUm1/fvsmaygJAAAAlY2RUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkrlquAwAAAOTKFR/+VOJtz2/fZA0mAVj3GCkFFJTJlPwCAAAAZaSUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUufse7CWy5RyQvJkDeUAAACA3zNSCgAAAIDUKaUAAAAASJ1SCgAAAIDUmVMKAKiwrvjwp1Jtf377JmsoCQAA5c1IKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAgPZlM6S4AAKy1lFIAAAAApK5argNQOplS/tU4SZI1lAQAAACg7IyUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUlct1wEAAFh3ZDKZUm2frKEcAEDuGSkFAAAAQOqUUgAAAACkTikFAAAAQOrMKQWwBpRmzpQkMWMKAACw7jFSCgAAAIDUKaUAAAAASJ1SCgAAAIDUmVMKAHKsNHOQRZiHDACAtYORUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOpMdA4AAGs5J1SgtDxngDQYKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKSuQpdSl1xySWQymQKXLbbYItexWAet+Dxc1QVKJZMp3QUAqBR8hwRYuWq5DrAqW2+9dbz44ovZ69WqVfjIAAAAAKxChW94qlWrFs2bN891DAAAAADKUYU+fC8i4quvvoqWLVvGJptsEn/6059i8uTJuY4EAAAAwGqq0COldtlllxg2bFhsvvnmMXXq1Bg0aFDsueee8fHHH0e9evWKvM2iRYti0aJF2etz5sxJKy4AAAAAJVShS6mePXtm/73ddtvFLrvsEq1bt46HH344jj/++CJvM3jw4Bg0aFBaEQEAAAAogwp/+N7vNWzYMDbbbLOYOHFisdsMGDAgZs+enb1MmTIlxYQAAAAAlESlKqXmzZsXX3/9dbRo0aLYbfLy8qJ+/foFLgAAsLoymUypLgDAylXoUupvf/tbjB07Nr755pt444034o9//GNUrVo1jjzyyFxHAwAAAGA1VOg5pb777rs48sgj4+eff46mTZvGHnvsEW+99VY0bdo019EAAAAAWA0VupR68MEHcx0BAAAAgDWgQpdSABR2xYc/lWr789s3WUNJAAAAyq5CzykFAAAAwNpJKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6qrlOgAAlJdMJlPibZMkWYNJAACAVTFSCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASF21XAdgDSvF6dHD6dEBAIgo3XfICN8jASgTpRRZV3z4U6m2P799kzWUBAAAAFjbOXwPAAAAgNQppQAAAABInVIKAAAAgNSZUwoAoJLJlHIS6sQk1ABABaSUAgCACqA0J51xwhkA1gZKKQAAAFZPaUZwGr0J/P/MKQUAAABA6pRSAAAAAKROKQUAAABA6swpBVRYpTm7lDNLAQAAVC5KKQAAAFJTmjNNRjjbJKzNHL4HAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAsCZkMqW7AMA6xkTnAADAainNxNUmrQZgOSOlAAAAAEidUgoAAACA1CmlAAAAAEidUgoAAACA1CmlAAAAAEidUgoAAACA1CmlAAAAAEhdtVwHAAAAoPSu+PCnEm97fvsmazAJQNkYKQUAAFARZDKluwBUckZKAVBAphRfcpMkWYNJKpbS/DU6wl+kqdw83wFyozTfwyLWre9irJ2MlAIAAAAgdUopANZNDpEAAICcUkoBAAAAkDqlFAAAAACpU0oBAAAAkDpn3wMA1knOcAQAkFtKKWCd5HTnAAAAueXwPQAAAABSp5QCAAAAIHVKKWDtkMmU7gKwLvEeCQBUQEopAAAAAFKnlAIAAAAgdUopAAAAAFJXLdcBWHdkSjlHRZIkaygJQCVX2jl/vJ8CAFABKaUAYC13xYc/lWr789s3WUNJAADg/3H4HgAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpc/Y9AIA1wFkPAQBWzkgpAICSyGRKdwEAYKWUUgAAAACkTikFAAAAQOqUUgAAAACkzkTnAAAAUBmVZg7DJFlzOaCMlFIAlF1pJ3P2ZQgAWMMypfx+kvh+Ajnj8D0AAAAAUmekFACpueLDn0q1/fntm6yhJAAAQK4ZKQUAAABA6oyUAgAAgLWcEetUREZKAQAAAJA6pRQAAAAAqVNKAQAAAJA6pRQAAAAAqVNKAQAAAJA6Z9+j4spkSr5tkqy5HAAAwNqrFL93XPHBj6XatTPYwcoppQAAAADWgCs+/KlU269rRaZSirWCFzoAAABrXGmO6ImIKOXounWNOaUAAAAASJ1SCgAAAIDUVYpS6uabb46NN944atasGbvssku88847uY4EAAAAwGqo8KXUQw89FGeffXYMHDgwPvjgg/jDH/4QPXr0iBkzZuQ6GgAAAABlVOFLqWuvvTZOPPHEOO6442KrrbaK2267LWrXrh133313rqMBAAAAUEYV+ux7ixcvjvfffz8GDBiQXValSpXo1q1bvPnmm0XeZtGiRbFo0aLs9dmzZ0dExJw5c9Zs2AqqNP/rhfPmlm7fc2qULkwpyV7MvmUvet+l3F728iH7SvYve9H7LuX2spcP2YvZdwX6TI2QvbzIXsy+K1D2ivQeEyF7sftew9nXFevq4768g0mSZKXbZZJVbZFDP/zwQ2ywwQbxxhtvRMeOHbPLzz333Bg7dmy8/fbbhW5zySWXxKBBg9KMCQAAAMAKpkyZEhtuuGGx6yv0SKmyGDBgQJx99tnZ6/n5+TFz5sxo3LhxZDKZHCarOObMmROtWrWKKVOmRP369XMdp8Qqa+4I2XNF9tyQPTdkzw3Zc0P23JA9N2TPjcqavbLmjpB9bZMkScydOzdatmy50u0qdCnVpEmTqFq1akyfPr3A8unTp0fz5s2LvE1eXl7k5eUVWNawYcM1FbFSq1+/fqV8wVTW3BGy54rsuSF7bsieG7Lnhuy5IXtuyJ4blTV7Zc0dIfvapEGDBqvcpkJPdF6jRo3o0KFDjB49OrssPz8/Ro8eXeBwPgAAAAAqlwo9Uioi4uyzz46+ffvGjjvuGDvvvHMMGTIk5s+fH8cdd1yuowEAAABQRhW+lDr88MPjxx9/jIsvvjimTZsW22+/fTz33HOx/vrr5zpapZWXlxcDBw4sdJhjRVdZc0fIniuy54bsuSF7bsieG7Lnhuy5IXtuVNbslTV3hOzrqgp99j0AAAAA1k4Vek4pAAAAANZOSikAAAAAUqeUAgAAACB1SikAAACoRH744YdcR4ByoZRax3388ce5jgDAGjBv3rxcRwBW8NNPP8WcOXNyHYMK6tJLL41ff/011zHKpDL/TjFp0qRcRyiTrbfeOkaMGJHrGLDalFLroLlz58a///3v2HnnneMPf/hDruMU6aWXXoqtttqqyC9us2fPjq233jpeffXVHCSjIvvf//4XlfGEogsWLIhnnnkme33AgAFx9tlnZy/nnHNOLFy4MIcJqWiuu+66la6fO3du9OjRI6U05eu7776Lk046KdcxoNzMmjUrTj/99GjSpEmsv/76sd5660Xz5s1jwIABlbaAWG7BggW5jlCkN998s8DnakTEvffeG23atIlmzZrFSSedFIsWLcpRuuINGjSo0v5BYbvttotddtkl7rjjjpg7d26u45RK27Zto02bNvHnP/857rvvvvjuu+9yHalELr/88jj55JPj0EMPjZkzZ+Y6DpRZJqmMv8FRJq+88krcdddd8dhjj0XLli3j4IMPjj59+sROO+2U62iFHHjggdGlS5fo379/ketvuOGGGDNmTDz++OMpJyu5/Pz8GDZsWIwcOTK++eabyGQy0aZNmzjkkEPimGOOiUwmk+uIRerVq1c88MAD0aBBg4iIuOKKK+KUU06Jhg0bRkTEzz//HHvuuWd8+umnOUxZtKpVq8bUqVOjWbNmERFx+OGHxw033BDrr79+jpOt3G233Rb/+c9/4umnn46IiHr16sXWW28dtWrVioiIzz//PM4999xiXw+5VtK/+tevX38NJym9gw8+uETbjRw5cg0nKZ1atWrF7bffHscee2yhdfPnz4/u3bvHzz//HJ9//nkO0q2e8ePHxw477BDLli3LdZRC/vznP5dou7vvvnsNJ1m35Ofnx9VXXx1PPfVULF68OPbee+8YOHBg9j2yIps5c2Z07Ngxvv/++/jTn/4UW265ZUREfPrppzFixIjYYost4rXXXosJEybEW2+9FWeddVaOE5fMokWL4qabboqrr746pk2blus4hfTs2TM6d+4c5513XkREfPTRR7HDDjtEv379Ysstt4yrr746Tj755LjkkktyG3QFVapUiWnTpmW/x1Qmr776agwdOjQeffTRyM/Pjz59+sQJJ5wQe+65Z66jrdLLL7+cvbz99tuxePHi2GSTTaJr167RpUuX6NKlS4X9Ljlp0qQ4/vjj49NPP4077rgjDjjggFxHWm3z5s2L/Pz8Assq4nfI5X744Yd47bXXYsaMGYVyV5b39FxTSq3lpk2bFsOGDYu77ror5syZE4cddljcdtttMX78+Nhqq61yHa9YrVu3jueeey775W1Fn3/+eXTv3j0mT56ccrKSSZIkDjjggPjvf/8bf/jDH2KLLbaIJEnis88+i48++igOPPDAeOKJJ3Ids0grFjv169ePcePGxSabbBIREdOnT4+WLVtWyF8YV/wyV69evRg/fnw2e0W15557xrnnnpv9IrFi7vvvvz9uvvnmePPNN3MZs1hVqlRZacmaJElkMpkK+Zw57rjjSrTd0KFD13CS0nn00UfjmGOOiYceeigOPPDA7PL58+dHjx49YsaMGTF27Nho0aJFDlOWTUUupapUqRKtW7eO9u3br3RUZkX8g8mqXqcREZlMJpYuXZpSopL7xz/+EZdcckl069YtatWqFaNGjYojjzyyUpR/f/3rX2P06NHx4osvFvqldtq0adG9e/fYfPPN4/nnn48bbrgh+vbtm6OkhS1atCguueSSeOGFF6JGjRpx7rnnRu/evWPo0KFxwQUXRNWqVeOMM87IFj8VSYsWLeLpp5+OHXfcMSIiLrjgghg7dmy89tprERHxyCOPxMCBAyvcH9iqVKkS06dPj6ZNm+Y6SpnNnz8/Hn744Rg2bFi8+uqrsemmm8bxxx8fffv2jebNm+c63iotXLgw3njjjWxJ9c4778SSJUtiiy22iE8++STX8Yp10003Rf/+/WPLLbeMatWqFVj3wQcf5ChVyU2aNCnOOOOMePnllwscHVCRv0NGRAwbNixOPvnkqFGjRjRu3LjA52wmk4n//e9/OUxXeVRb9SZUVgcccEC88sorsd9++8WQIUNi3333japVq8Ztt92W62irNH369KhevXqx66tVqxY//vhjiolKZ9iwYfHKK6/E6NGjo0uXLgXWvfTSS9G7d++49957ixzlkGsr/qKlt17zJk6cGNtuu232es2aNaNKlf93dPXOO+8cp59+ei6ilciYMWOy/06SJHr16hV33nlnbLDBBjlMVTIVrWwqqUMOOSRmzZoVRx55ZPznP/+Jzp07x/z582PfffeN6dOnV9pCqqI79dRT44EHHohJkybFcccdF0cffXQ0atQo17FKZGVF2Ztvvhk33HBDob/wVhT33ntv3HLLLXHyySdHRMSLL74Y++23X9x5550F3isroieeeCJuv/32IkdZNG/ePK666qro1atXDBw4sEIVUhERF198cdx+++3RrVu3eOONN+LQQw+N4447Lt5666249tpr49BDD42qVavmOmaRfvnllwKP+dixY6Nnz57Z6zvttFNMmTIlF9FWabPNNltlgVyRD9WqU6dOHHfccXHcccfFxIkTY+jQoXHzzTfHRRddFPvuu2889dRTuY64UjVr1oyuXbvGHnvsEV26dIlnn302br/99go98vjbb7+NkSNHxnrrrRcHHXRQoVKqMjj66KMjSZK4++67Y/3116+wR5Ss6KKLLoqLL744BgwYUOE/jyqyyveMpcSeffbZOOuss+LUU0+Ndu3a5TpOqWywwQbx8ccfx6abblrk+gkTJlToX7geeOCB+Pvf/16okIqI6Nq1a5x//vkxfPjwCllKVWaZTKbQh1hl+FCbNWtWgbktVixc8/PzK+TcF8t16tSpwPWqVavGrrvuWuFHqFV2J5xwQsycOTMOOuigePLJJ+Piiy+OH374IcaOHRstW7bMdby10s033xzXXnttjBw5Mu6+++4YMGBA7LfffnH88cdH9+7dK/T7zUEHHVRo2RdffBHnn39+PP300/GnP/0pLr300hwkW7XJkydHr169ste7desWmUwmfvjhh9hwww1zmGzVpk6dGltvvXWx67fZZpuoUqVKDBw4MMVUJfPII4/EvffeGwceeGB8/PHHsd1228XSpUtj/PjxFfq5HhGx/vrrx6RJk6JVq1axePHi+OCDD2LQoEHZ9XPnzl3pHz9zadCgQdkpFCq7TTfdNP7+979H69atY8CAAfGf//wn15GKtXjx4njrrbdizJgx2cP4WrVqFXvttVfcdNNNhb7rVBR33HFH/N///V9069YtPvnkk0o7ym78+PHx/vvvx+abb57rKKXy66+/xhFHHKGQWk1KqbXYa6+9FnfddVd06NAhttxyyzjmmGPiiCOOyHWsEunVq1f2Lyo1a9YssG7BggUxcODA2H///XOUbtUmTJgQV111VbHre/bsGTfccEOKiUqushY7Eb+N0unXr1/k5eVFxG9DsE855ZSoU6dOge0q2vxAG264YXz88cfFfhBPmDChwv/iRW6ce+65MXPmzNh7771j4403jpdffrnCP1dWNY/XrFmz0glSRnl5eXHkkUfGkUceGd9++20MGzYsTjvttFi6dGl88sknUbdu3VxHXKUffvghBg4cGPfcc0/06NEjxo0bF9tss02uYxVr6dKlhb4LVK9ePZYsWZKjRCXXpEmT+Oabb4p9XU6aNKnCzh/03XffRYcOHSLit/IsLy8v+vfvXym+E/Tq1SvOP//8uPLKK+OJJ56I2rVrF5jbaMKECdG2bdscJizeEUccUWGfE6XxyiuvxN133x2PPfZYVKlSJQ477LA4/vjjcx2rSF27do2333472rRpE506dYqTTz45RowYUaH/AB4Rse+++8Y777wTN910U6X/Q/fy0YuVrZQ6/vjj45FHHonzzz8/11EqNaXUWmzXXXeNXXfdNYYMGRIPPfRQ3H333XH22WdHfn5+vPDCC9GqVauoV69ermMW6cILL4yRI0fGZpttFmeccUb2Derzzz+Pm2++OZYtWxYXXHBBjlMWb+bMmSudEHH99dePX375JcVEJbeqYqcij9hZ8dCHo48+OkdJSqdXr15x8cUXx3777VdkCTto0KDYb7/9cpSOimjFYqd69erRpEmT+Mtf/lJgeUUrYCNilSMAGjRoUGm+XC+fpylJkgo738XvzZ49O/75z3/GjTfeGNtvv32MHj26UkxCvOLnUkTRf3SoiM/3Hj16xAUXXJCdl+n3Fi1alP0DXEW0bNmyApmrVatWKUrXiN/mITv44IOjU6dOUbdu3bjnnnsK/F/uvvvu6N69ew4TFq0yFH4r88MPP8SwYcNi2LBhMXHixNhtt93ihhtuiMMOO6zQHwgrkldffTVatGgRXbt2jc6dO0enTp2icePGuY61SsuWLVtr/nB55513ximnnBLff/99bLPNNoVGMm633XY5SrZygwcPjv333z+ee+652HbbbQvlvvbaa3OUrHIx0fk65osvvoi77ror7rvvvpg1a1bss88+FfbY7m+//TZOPfXUGDVqVHZeo0wmEz169Iibb7452rRpk+OExatatWpMmzat2CG0FXmy8Mo68XNlNn369Nh+++2jRo0accYZZ8Rmm20WEb+9Xm+66aZYunRpfPjhhxX2zC8rqlevXkyYMKFCv0YrO6/T3Fm0aFH28L3XXnst9t9//zjuuONi3333rdDD96+66qq48soro3nz5vHPf/6zyMP5KqrK/Hz/7rvvYscdd4y8vLw4/fTTC5z45JZbbolFixbFu+++GxtttFGuoxZSpUqV6NmzZ7YMfPrpp6Nr164VfvTx782ePTvq1q1baO6rmTNnRt26dQsVhblWmc++17Nnz3jxxRejSZMmceyxx8af//znSjPqZf78+fHqq6/Gyy+/HGPGjIlx48bFZpttFp06dcqWVJX1sLjK4q233oqjjjoqvvnmm+yy5X/0qcgTnV922WVx8cUXx+abb15oLqxMJhMvvfRSDtNVHkqpddSyZcvi6aefjrvvvrvCllLL/fLLLzFx4sRIkiTatWsX6623Xq4jrdKKX+RWtGjRonjuuecq7Bss6Zs0aVKceuqp8cILLxQoYffZZ5+45ZZbKvT8TCuO2qmMv7hASZx22mnx4IMPRqtWreLPf/5z/OlPf4omTZrkOlaJVKlSJWrVqhXdunVb6eTUXqflb9KkSXHaaafF888/X+j9/aabbip2/sxcq8xlIOk78MAD4/jjj4/999+/wk6AX1Jz586N1157LTu/1Pjx46Ndu3bx8ccf5zraWmurrbaKLbfcMs4999wiJzpv3bp1jpKt3HrrrRfXXXdd9OvXL9dRKjWlFKwBvshRVjNnzoyJEydGxG8ThFaGM3t5vrOuqFKlSmy00UbRvn37lR5mUxGLnX79+pXo0CCv0zXnl19+ia+++ioiKs/7O6yL8vPz4913340xY8bEmDFj4rXXXouFCxf6Y/IaVKdOnRg/fnyFLemL07x583j11Vcr3UnFKhqlFABACSh2ANY++fn58d5772UP33v99ddj/vz5scEGG0SXLl2yl4o6WmdtcMABB0S/fv2iT58+uY5SKoMHD46pU6dW2BNYVRZKKQAAANZJ9evXj/nz50fz5s2zBVTnzp0r7Bka10b//ve/47LLLos///nPRU4YfuCBB+Yo2cr98Y9/jJdeeikaN24cW2+9daHcFXHkdEWklAIAAGCddPvtt0eXLl2yJ5ohfSs7UUhFnuh8VVNYGDldMkopAAAAgBJaunRpjBgxIrp37x7NmzfPdZxKreKeuxgAAABYay1ZsiSqVatW6c5uWK1atTjllFNi0aJFuY5S6SmlAAAAgNRVr149Ntpoowp7iN7K7LzzzvHhhx/mOkalVy3XAQAAAIB10wUXXBB///vf47777otGjRrlOk6JnXbaafF///d/8d1330WHDh2iTp06BdZvt912OUpWuZhTCgAAAMiJ9u3bx8SJE2PJkiXRunXrQuXOBx98kKNkK1fUBO2ZTCaSJKnQE7RXNEZKAQAAADnRu3fvXEcok0mTJuU6wlrBSCkAAAAAUmeicwAAACBnZs2aFXfeeWcMGDAgZs6cGRG/Hbb3/fff5zjZyt13332x++67R8uWLePbb7+NiIghQ4bEk08+meNklYdSCgAAAMiJCRMmxGabbRZXXnllXHPNNTFr1qyIiBg5cmQMGDAgt+FW4tZbb42zzz47evXqFbNmzcrOIdWwYcMYMmRIbsNVIkopAAAAICfOPvvs6NevX3z11VdRs2bN7PJevXrFK6+8ksNkK3fjjTfGHXfcERdccEFUrVo1u3zHHXeMjz76KIfJKhelFAAAAJAT7777bpx88smFlm+wwQYxbdq0HCQqmUmTJkX79u0LLc/Ly4v58+fnIFHlpJQCAAAAciIvLy/mzJlTaPmXX34ZTZs2zUGikmnTpk2MGzeu0PLnnnsuttxyy/QDVVLVch0AAAAAWDcdeOCBcemll8bDDz8cERGZTCYmT54c5513XvTp0yfH6Yp39tlnx+mnnx4LFy6MJEninXfeiQceeCAGDx4cd955Z67jVRqZJEmSXIcAAAAA1j2zZ8+OQw45JN57772YO3dutGzZMqZNmxYdO3aM//73v1GnTp1cRyzW8OHD45JLLomvv/46IiJatmwZgwYNiuOPPz7HySoPpRQAAACQU6+//nqMHz8+5s2bFzvssEN069Yt15FK7Ndff4158+ZFs2bNch2l0lFKAQAAADlx7733xuGHHx55eXkFli9evDgefPDBOPbYY3OUbOW6du0aI0eOjIYNGxZYPmfOnOjdu3e89NJLuQlWySilAAAAgJyoWrVqTJ06tdAoo59//jmaNWsWy5Yty1GylatSpUpMmzatUO4ZM2bEBhtsEEuWLMlRssrFROcAAABATiRJEplMptDy7777Lho0aJCDRCs3YcKE7L8//fTTmDZtWvb6smXL4rnnnosNNtggF9EqJaUUAAAAkKr27dtHJpOJTCYTe++9d1Sr9v/qiWXLlsWkSZNi3333zWHCom2//fbZ3F27di20vlatWnHjjTfmIFnlpJQCAAAAUtW7d++IiBg3blz06NEj6tatm11Xo0aN2HjjjaNPnz45Sle8SZMmRZIksckmm8Q777wTTZs2za6rUaNGNGvWLKpWrZrDhJWLOaUAAACAnLjnnnvi8MMPj5o1a+Y6CjlQJdcBAAAAgHVT3759s4XUaaedFj/99FOOE5Ve/fr143//+1+uY1RKSikAAAAg5+6///6YM2dOrmOUmgPQyk4pBQAAAOSccmfdo5QCAAAAKKOjjz466tevn+sYlZJSCgAAAMi5uXPnxiabbJLrGCv10ksvxVZbbVXgMMNbb701mjRpErNnz46tt946Xn311RwmrFyUUgAAAECqqlSpElWrVl3ppVq1armOWciQIUPixBNPLHJkVIMGDeLkk0+Oa6+9NgfJKqdM4qBNAAAAIEVPPvlksevefPPNuOGGGyI/Pz8WLlyYYqpVa926dTz33HOx5ZZbFrn+888/j+7du8fkyZNTTlY5VbzaEQAAAFirHXTQQYWWffHFF3H++efH008/HX/605/i0ksvzUGylZs+fXpUr1692PXVqlWLH3/8McVElZvD9wAAAICc+eGHH+LEE0+MbbfdNpYuXRrjxo2Le+65J1q3bp3raIVssMEG8fHHHxe7fsKECdGiRYsUE1VuSikAAAAgdbNnz47zzjsvNt100/jkk09i9OjR8fTTT8c222yT62jF6tWrV1x00UVFHla4YMGCGDhwYOy///45SFY5mVMKAAAASNVVV10VV155ZTRv3jz++c9/Fnk4X0U0ffr02GGHHaJq1apxxhlnxOabbx4Rv80ldfPNN8eyZcvigw8+iPXXXz/HSSsHpRQAAACQqipVqkStWrWiW7duUbVq1WK3GzlyZIqpSubbb7+NU089NUaNGhXLK5VMJhM9evSIm2++Odq0aZPjhJWHUgoAAABIVb9+/SKTyaxyu6FDh6aQpmx++eWXmDhxYiRJEu3atYv11lsv15EqHaUUAAAAAKkz0TkAAAAAqVNKAQAAAJC6/6+9u4+p+f3jOP46lW5WDZvWCeGQCDkyY/WHe01/tFlhDGuUYfWHRhZTmKk0d8OR+1Pulo2JVTI2bJghN9ON4lixyT2bclfx++M7Zw7Z1NeO377n+dg+2znXua7rvK//zl67rusQSgEAAAAAAMDpCKUAAAAAAADgdIRSAAAAf0hdXZ0MBoNu3779yz4XLlyQwWDQ27dvnVYXAADA/yNCKQAA4DK+/f20wWBQp06dZDKZtGzZMn38+PGPzB8cHKyGhgYNGTLkj8zXUd/W+Ktn9erVf7U+AAAASfL42wUAAAA40+TJk2W1WtXc3Kzy8nIlJCTIYDBo/fr1/3pud3d3GY3GP1Dlv9PQ0GB/ffToUWVmZqqmpsbe5ufn9zfKAgAAcMBOKQAA4FK8vLxkNBoVHBysKVOmaOLEiTp79qz98y9fvig7O1smk0k+Pj4ym806duyY/fM3b95o1qxZCggIkI+Pj/r37y+r1Sqp7eN7paWlCg0NlY+Pj8aNG6e6ujqHelavXq1hw4Y5tG3ZskV9+vRxaNu7d6/CwsLk7e2tgQMHaseOHb9co9FotD+dO3eWwWCQ0WiUv7+/QkNDVVZW5tC/qKhIvr6+evfunX0NhYWFioqKkre3t4YMGaKLFy86jKmoqFBMTIz8/PwUGBioOXPm6OXLl7+sCQAA4EeEUgAAwGVVVFToypUr8vT0tLdlZ2frwIED2rlzpyorK5WamqrZs2fbQ5mMjAxVVVXp9OnTqq6uVl5enrp169bm/I8fP1ZcXJxiY2N1+/ZtJSUlKT09vd11Hj58WJmZmVq3bp2qq6uVlZWljIwMFRQUtGseX19fzZgxwx6ifWO1WjV16lT5+/vb29LS0rRkyRLdunVLkZGRio2N1atXryRJb9++1fjx4xUREaEbN26orKxMz5490/Tp09u9NgAA4Lo4vgcAAFxKcXGx/Pz81NLSok+fPsnNzU3bt2+XJH369ElZWVk6d+6cIiMjJUl9+/bVpUuXtGvXLo0ZM0aPHj1SRESERowYIUk/7Wj6Xl5envr166eNGzdKkgYMGKC7d++2+6jgqlWrtHHjRsXFxUmSTCaTqqqqtGvXLiUkJLRrrqSkJEVFRamhoUFBQUF6/vy5SktLde7cOYd+KSkpio+Pt6+jrKxM+/bt07Jly7R9+3ZFREQoKyvL3n///v0KDg5WbW2tQkND21UTAABwTYRSAADApYwbN055eXlqamrS5s2b5eHhYQ9fHjx4oPfv32vSpEkOYz5//qyIiAhJ0qJFixQfH6+bN28qOjpaU6ZMUVRUVJvfVV1drVGjRjm0fQu7fldTU5NsNpsSExM1f/58e3tLS4s6d+7crrkkaeTIkRo8eLAKCgqUnp6uQ4cOqXfv3ho9evQv6/Tw8NCIESNUXV0tSbpz547Onz/f5t1UNpuNUAoAAPwWQikAAOBSfH19FRISIumf3T1ms1n79u1TYmKiGhsbJUklJSXq0aOHwzgvLy9JUkxMjOrr61VaWqqzZ89qwoQJSk5O1oYNGzpUj5ubm75+/erQ1tzcbH/9raY9e/b8FHC5u7t36DuTkpJksViUnp4uq9WquXPnymAw/Pb4xsZGxcbGtrnjKygoqEM1AQAA18OdUgAAwGW5ublpxYoVWrlypT58+KBBgwbJy8tLjx49UkhIiMMTHBxsHxcQEKCEhAQdOnRIW7Zs0e7du9ucPywsTNeuXXNou3r1qsP7gIAAPX361CGY+v6i9MDAQHXv3l0PHz78qSaTydShdc+ePVv19fXaunWrqqqq2jwC+H2dLS0tKi8vV1hYmCRp+PDhqqysVJ8+fX6qydfXt0M1AQAA10MoBQAAXNq0adPk7u4ui8Uif39/LV26VKmpqSooKJDNZtPNmze1bds2+6XimZmZOnnypB48eKDKykoVFxfbw5ofLVy4UPfv31daWppqamp05MgR5efnO/QZO3asXrx4odzcXNlsNlksFp0+fdqhz5o1a5Sdna2tW7eqtrZWd+/eldVq1aZNmzq05q5duyouLk5paWmKjo5Wz549f+pjsVh04sQJ3bt3T8nJyXrz5o3mzZsnSUpOTtbr1681c+ZMXb9+XTabTWfOnNHcuXPV2traoZoAAIDrIZQCAAAuzcPDQykpKcrNzVVTU5PWrl2rjIwMZWdnKywsTJMnT1ZJSYl9V5Knp6eWL1+uoUOHavTo0XJ3d1dhYWGbc/fq1UvHjx9XUVGRzGazdu7c6XA5uPTPbqodO3bIYrHIbDbr2rVrWrp0qUOfpKQk7d27V1arVeHh4RozZozy8/M7vFNKkhITE/X582d70PSjnJwc5eTkyGw269KlSzp16pT9Xwa7d++uy5cvq7W1VdHR0QoPD9fixYvVpUsXubnx8xIAAPwew9cfLzEAAADAf97BgweVmpqqJ0+eyNPT095eV1cnk8mkW7duadiwYX+vQAAA8J/HRecAAAAu5P3792poaFBOTo4WLFjgEEgBAAA4E/urAQAAXEhubq4GDhwoo9Go5cuX/+1yAACAC+P4HgAAAAAAAJyOnVIAAAAAAABwOkIpAAAAAAAAOB2hFAAAAAAAAJyOUAoAAAAAAABORygFAAAAAAAApyOUAgAAAAAAgNMRSgEAAAAAAMDpCKUAAAAAAADgdIRSAAAAAAAAcLr/ARfZhFb+Vk6ZAAAAAElFTkSuQmCC\n"
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of Unique Accession Numbers for Significant Data in AdjacentSeq_all__(Torpor)___(Aug_Active).tsv: 1337\n",
            "Number of Unique Accession Numbers for Not Significant Data in AdjacentSeq_all__(Torpor)___(Aug_Active).tsv: 1890\n",
            "Total number of modifications in AdjacentSeq_all__(Torpor)___(Aug_Active).tsv: 17793\n",
            "Significant N-terminal (Met-loss) modifications: 19\n",
            "Non-significant N-terminal (Met-loss) modifications: 24\n",
            "Processing file: /content/output/modlist_files/AdjacentSeq_all__(Mar_Active)___(Torpor).tsv\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 1200x800 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAABKUAAAMWCAYAAAAgRDUeAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAB1UklEQVR4nOzdd5hU9f0+/PfQlt5BLIiAKNgRGzZAkWZDMRYSBWMXY9TYiImKUYmarxIbGAtYQI2KNRHs2DuKXTEoqIAoUqXvef7wYX4uuwu7y3JmF16v65oL9pwzM/fOzpzZvedzPieTJEkSAAAAAJCiKrkOAAAAAMCGRykFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFAAAAQOqUUgAAAACkTikFsI5deumlkclkSrRtJpOJSy+9dN0GWktffPFF9OjRIxo0aBCZTCYeeeSRXEfKiaJ+rltssUUMHDiwwLKiHq9Ro0ZFJpOJr776Kr3A/7/K8BwrD0U9xl27do2uXbvmLBOFrfqaeeGFFyKTycQLL7xQ7ve1YMGCaN68eYwePbrcb3t9sK73DePGjYu6devGrFmz1tl9lESfPn3ipJNOymmGdWnEiBGx+eabx5IlS3IdBaBElFLABmPlH6krL9WqVYtNN900Bg4cGN9++22u41UaAwYMiA8++CCuuOKKuPvuu2OXXXaJMWPGxLBhw3IdrUIq6vFa1/773/+uN8XTzTffHJlMJnbfffdcR0ndd999F5deemm89957Ra5//PHHo0uXLtG8efOoXbt2tGnTJo488sgYN25cukEriX/+859Rr169OProo7PLVpbLVapUiWnTphW6zrx586JWrVqRyWTijDPOSDNuIbvttltkMpkYPnx4mW8jl/uGXr16xZZbbhlDhw7Nyf1HRLzyyivx1FNPxQUXXBARv5Siv/69oLjLqFGjcpa5tAYOHBhLly6NW265pUTbe/8Gcq1argMApO2yyy6L1q1bx+LFi+P111+PUaNGxcsvvxwffvhh1KxZs9zv7y9/+UtceOGF5X67ubBo0aJ47bXX4qKLLirwB9qYMWPiww8/jLPOOit34SqAzz77LKpU+X+f9xT3eB177LFx9NFHR15e3jrJ8d///jduuummIv/4XLRoUVSrVnne/kePHh1bbLFFvPnmmzF58uTYcssty3xbTz31VDkmW/e+++67GDJkSGyxxRax0047FVj3j3/8I84777zo0qVLDB48OGrXrh2TJ0+OZ555Ju67777o1atXbkJXUMuWLYt//vOfcfbZZ0fVqlULrc/Ly4t77703zj///ALLx44dm1bE1friiy/irbfeii222CJGjx4dp512WpluJ9f7hlNOOSXOPffcGDJkSNSrV2+d3ldRrrnmmth///2z+5Fhw4bFggULsuv/+9//xr333hvXXXddNG3aNLt8zz33TD1rWdWsWTMGDBgQ1157bfzhD39Y40ht799ArlWe30oByknv3r2zo1VOPPHEaNq0aVx11VXx2GOPxZFHHlnu91etWrVKVQKszsrDLho2bLjO7ys/Pz+WLl26TorCdWXVkqm4x6tq1apF/mGchsr0eE6ZMiVeffXVGDt2bJxyyikxevTouOSSS8p8ezVq1CjHdLmzfPny+Nvf/hYHHHBAkUXb999/n4NUFdsTTzwRs2bNKnYf36dPnyJLqTFjxsSBBx4YDz30ULllWb58eeTn55fq+XjPPfdE8+bN4//+7//iiCOOiK+++iq22GKLcssUkc6+oV+/fvGHP/whHnjggfj973+/zu/v177//vv4z3/+EyNGjMgu69u3b4FtZsyYEffee2/07dt3rR/fJEli8eLFUatWrbW6nZJauHBh1KlTJyIijjzyyLj66qvj+eefj/322y+V+wcoK4fvARu8ffbZJyIivvzyywLLP/300zjiiCOicePGUbNmzdhll13iscceK7DNsmXLYsiQIdGuXbuoWbNmNGnSJPbee+94+umns9sUNffQkiVL4uyzz45mzZpFvXr14pBDDolvvvmmULaBAwcW+YtxcfNU3XPPPdGpU6eoVatWNG7cOI4++ugiD0lZ1ddffx2nn356bL311lGrVq1o0qRJ/OY3vykwH8+ll14arVq1ioiI8847LzKZTGyxxRbRtWvX+M9//hNff/119lCHX2desmRJXHLJJbHllltGXl5etGzZMs4///xC812sPDxm9OjRse2220ZeXt5qD0PaYost4qCDDooXXnghdtlll6hVq1Zsv/322bloxo4dG9tvv33UrFkzOnXqFBMnTix0G88991zss88+UadOnWjYsGEceuih8cknnxTa7uWXX45dd901atasGW3bti32sIhfz49T3OMVUfR8RxERTz75ZHTp0iXq1asX9evXj1133TXGjBmTXf/SSy/Fb37zm9h8882zj+XZZ58dixYtym4zcODAuOmmm7KP6crLrx/nVUdJTJw4MXr37h3169ePunXrxv777x+vv/56gW1WZn7llVfinHPOiWbNmkWdOnXisMMOW2dzxIwePToaNWoUBx54YBxxxBHFzgX00UcfxX777Re1atWKzTbbLC6//PLIz88vtN2qc0otXbo0Lr744ujUqVM0aNAg6tSpE/vss088//zzha6bn58f//znP7PPqWbNmkWvXr3i7bffLrBdSV6DXbt2je222y4+/vjj6NatW9SuXTs23XTTuPrqq7PbvPDCC7HrrrtGRMTxxx9f4DCiH374IebNmxd77bVXkY9H8+bNC3xd0tdgcfulssw19I9//CP23HPPaNKkSdSqVSs6deoUDz74YKluozw98sgjscUWW0Tbtm2LXN+/f/9477334tNPP80umzFjRjz33HPRv3//QtuX9Lnz1VdfRSaTiX/84x8xbNiwaNu2beTl5cXHH39cqvxjxoyJI444Ig466KBo0KBBgf3Cr73xxhvRp0+faNSoUdSpUyd22GGH+Oc//xkRpds3PPjgg5HJZGLChAmF7uOWW26JTCYTH374YXZZSd4vI355bu6www7x6KOPlur7Lw//+c9/Yvny5dG9e/dSXW9lCbzyZ7fFFlvEn//850Kvn5XvSePHj8++J618r/j1+9vWW2+dfV968cUXC91fafbHEyZMiNNPPz2aN28em222WXZ9p06donHjxmt8nNf0/n3DDTfEtttuG7Vr145GjRplD9ePKP1zBKA468dH9wBrYWUx0KhRo+yyjz76KPbaa6/YdNNN48ILL4w6derEv//97+jbt2889NBDcdhhh0XEL8XD0KFD48QTT4zddtst5s2bF2+//Xa8++67ccABBxR7nyeeeGLcc8890b9//9hzzz3jueeeiwMPPHCtvo8rrrgi/vrXv8aRRx4ZJ554YsyaNStuuOGG2HfffWPixImrHd301ltvxauvvhpHH310bLbZZvHVV1/F8OHDo2vXrvHxxx9H7dq14/DDD4+GDRvG2WefHcccc0z06dMn6tatG3Xq1Im5c+fGN998E9ddd11ERNStWzcifvlD/pBDDomXX345Tj755OjQoUN88MEHcd1118Xnn39eaJL05557Lv7973/HGWecEU2bNl3jJ9WTJ0+O/v37xymnnBK/+93v4h//+EccfPDBMWLEiPjzn/8cp59+ekREDB06NI488sgCh9c988wz0bt372jTpk1ceumlsWjRorjhhhtir732infffTd73x988EH06NEjmjVrFpdeemksX748Lrnkkthoo41Wm624x6s4o0aNit///vex7bbbxuDBg6Nhw4YxceLEGDduXPaP4gceeCB+/vnnOO2006JJkybx5ptvxg033BDffPNNPPDAAxHxy+Ex3333XTz99NNx9913rzZjxC/P9X322Sfq168f559/flSvXj1uueWW6Nq1a0yYMKHQXE5/+MMfolGjRnHJJZfEV199FcOGDYszzjgj7r///jXeV2mNHj06Dj/88KhRo0Ycc8wxMXz48HjrrbeyZU3EL8VBt27dYvny5dnX6r/+9a8SjU6YN29e3HbbbXHMMcfESSedFPPnz4/bb789evbsGW+++WaBQ+ZOOOGEGDVqVPTu3TtOPPHEWL58ebz00kvx+uuvZ0deluY1+NNPP0WvXr3i8MMPjyOPPDIefPDBuOCCC2L77beP3r17R4cOHeKyyy6Liy++OE4++eRseb7nnntG8+bNo1atWvH444/HH/7wh2jcuHGx32NpXoPluV/65z//GYccckj89re/jaVLl8Z9990Xv/nNb+KJJ55Y631dWbz66qux8847F7t+3333jc022yzGjBkTl112WURE3H///VG3bt0i85bmuRMRMXLkyFi8eHGcfPLJkZeXt9qf2areeOONmDx5cowcOTJq1KgRhx9+eIwePTr+/Oc/F9ju6aefjoMOOig23njj+OMf/xgtWrSITz75JJ544on44x//WKp9w4EHHhh169aNf//739GlS5cC6+6///7YdtttY7vttouIkr9frtSpU6ecnCDj1VdfjSZNmmQ/LCipE088Me6888444ogj4k9/+lO88cYbMXTo0Pjkk0/i4YcfLrDtZ599Fsccc0yccsopcdJJJ8XWW2+dXTdhwoS4//7748wzz4y8vLy4+eabo1evXvHmm28WeCxLsz8+/fTTo1mzZnHxxRfHwoULC6zbeeed45VXXlnt93bRRRcV+/596623xplnnhlHHHFE/PGPf4zFixfHpEmT4o033oj+/fuX6jkCsFoJwAZi5MiRSUQkzzzzTDJr1qxk2rRpyYMPPpg0a9YsycvLS6ZNm5bddv/990+23377ZPHixdll+fn5yZ577pm0a9cuu2zHHXdMDjzwwNXe7yWXXJL8enf73nvvJRGRnH766QW269+/fxIRySWXXJJdNmDAgKRVq1ZrvM2vvvoqqVq1anLFFVcU2O6DDz5IqlWrVmj5qn7++edCy1577bUkIpK77roru2zKlClJRCTXXHNNgW0PPPDAInPefffdSZUqVZKXXnqpwPIRI0YkEZG88sor2WURkVSpUiX56KOPVpt1pVatWiURkbz66qvZZePHj08iIqlVq1by9ddfZ5ffcsstSUQkzz//fHbZTjvtlDRv3jz58ccfs8vef//9pEqVKslxxx2XXda3b9+kZs2aBW7v448/TqpWrZqs+jbaqlWrZMCAAdmvi3u8Vj4Xp0yZkiRJksyZMyepV69esvvuuyeLFi0qsG1+fn72/0X9nIYOHZpkMpkC+QYNGlQo20qrPsf69u2b1KhRI/nyyy+zy7777rukXr16yb777lsoc/fu3QtkOvvss5OqVasmc+bMKfL+yurtt99OIiJ5+umnkyT55XHYbLPNkj/+8Y8FtjvrrLOSiEjeeOON7LLvv/8+adCgQYHHOEmSpEuXLkmXLl2yXy9fvjxZsmRJgdv76aefko022ij5/e9/n1323HPPJRGRnHnmmYVyrnwsSvMa7NKlS6HX1pIlS5IWLVok/fr1yy576623kohIRo4cWeh+L7744iQikjp16iS9e/dOrrjiiuSdd94ptF1JX4Ol2S+VxKrP1aVLlybbbbddst9++xVYvupr5vnnny/0Wl1by5YtSzKZTPKnP/2p0LqV+9JZs2Yl5557brLllltm1+26667J8ccfnyTJL6+bQYMGZdeV9Lmzch9Qv3795Pvvvy9T/jPOOCNp2bJl9rn21FNPJRGRTJw4sUCe1q1bJ61atUp++umnAtf/9eu1NPuGY445JmnevHmyfPny7LLp06cnVapUSS677LLsspK+X6505ZVXJhGRzJw5s0Tff3nZe++9k06dOq12m2uuuabAfmPl6+LEE08ssN25556bRETy3HPPZZetfE8aN25coduNiCQikrfffju77Ouvv05q1qyZHHbYYdllpd0f77333gV+Pr928sknJ7Vq1Vrt95skxb9/H3roocm222672uuW9DkCsDoO3wM2ON27d49mzZpFy5Yt44gjjog6derEY489lh36Pnv27HjuuefiyCOPjPnz58cPP/wQP/zwQ/z444/Rs2fP+OKLL7Jn62vYsGF89NFH8cUXX5T4/v/73/9GRMSZZ55ZYPnaTDI6duzYyM/PjyOPPDKb94cffogWLVpEu3btijwc6dd+Papk2bJl8eOPP8aWW24ZDRs2jHfffbfMuR544IHo0KFDtG/fvkCulXNcrJqrS5cusc0225T49rfZZpvo3Llz9uuVnyLvt99+sfnmmxda/r///S8iIqZPnx7vvfdeDBw4sMCIhR122CEOOOCA7M9oxYoVMX78+Ojbt2+B2+vQoUP07NmzxDnX5Omnn4758+fHhRdeWGhel18fXvPrn9PChQvjhx9+iD333DOSJCny8MQ1WbFiRTz11FPRt2/faNOmTXb5xhtvHP3794+XX3455s2bV+A6J598coFM++yzT6xYsSK+/vrrUt//6owePTo22mij6NatW0T88jgcddRRcd9998WKFSuy2/33v/+NPfbYI3bbbbfssmbNmsVvf/vbNd5H1apVs/P65Ofnx+zZs2P58uWxyy67FHjeP/TQQ5HJZIqcz2rlY1Ha12DdunXjd7/7XfbrGjVqxG677ZZ9jq7JkCFDYsyYMdGxY8cYP358XHTRRdGpU6fYeeedCxyCWtLXYHnvl379XP3pp59i7ty5sc8++6zV/qSsZs+eHUmSFBgNW5T+/fvH5MmT46233sr+W9ShexElf+6s1K9fv2jWrFmpsy9fvjzuv//+OOqoo7LPtf322y+aN29e4HDWiRMnxpQpU+Kss84qNCp2TRNdF+eoo46K77//PntIdMQvh2zl5+fHUUcdFRGle79caeXP4YcffihTrrL68ccf1/gcWNXK18U555xTYPmf/vSniPjlkMBfa926dbHvDZ07d45OnTplv958883j0EMPjfHjx8eKFSvKtD8+6aSTip2fsFGjRrFo0aL4+eefS/jdFtSwYcP45ptv4q233ip2m5I8RwDWRCkFbHBuuummePrpp+PBBx+MPn36xA8//FBggurJkydHkiTx17/+NZo1a1bgsvKP0pUTCV922WUxZ86c2GqrrWL77beP8847LyZNmrTa+//666+jSpUqheY2+fUw/9L64osvIkmSaNeuXaHMn3zyyRonPl60aFFcfPHF0bJly8jLy4umTZtGs2bNYs6cOTF37ty1yvXRRx8VyrTVVltFROEJmVu3bl2q2/91URQR0aBBg4iIaNmyZZHLf/rpp4iIbIFS1GPeoUOH+OGHH2LhwoUxa9asWLRoUbRr167Qdmvz81rVyvnM1nSow9SpU7NFWt26daNZs2bZwybK8nOaNWtW/Pzzz8U+Dvn5+YXmQ1r1MV/5R97Kx7YoixYtihkzZhS4rM6KFSvivvvui27dusWUKVNi8uTJMXny5Nh9991j5syZ8eyzz2a3/frrr9fq53PnnXfGDjvskJ0TrlmzZvGf//ynwOP55ZdfxiabbLLaQ65K+xrcbLPNCpUFjRo1Wu3juKpjjjkmXnrppfjpp5/iqaeeiv79+8fEiRPj4IMPjsWLF2dzleQ1WN77pSeeeCL22GOPqFmzZjRu3DiaNWsWw4cPX6v9ydpKkmS16zt27Bjt27ePMWPGxOjRo6NFixarnSS6JM+dlUq7b1vpqaeeilmzZsVuu+2WfR1MmTIlunXrFvfee2927rSS7kNKo1evXtGgQYMCh+bef//9sdNOO2WfP6V5v1xp5c+hrGXZ2ljTc2BVK18Xq571s0WLFtGwYcNCZfzqfs5F7ae22mqr+Pnnn2PWrFll2h+v7v7W9nG+4IILom7durHbbrtFu3btYtCgQYUOByzJcwRgTcwpBWxwdtttt+wcMH379o299947+vfvH5999lnUrVs3+0v+ueeeW+wnnit/Qd13333jyy+/jEcffTSeeuqpuO222+K6666LESNGxIknnrjWWYv7ZfLXI0UifvmkPpPJxJNPPlnkp6arm8so4pd5gkaOHBlnnXVWdO7cORo0aBCZTCaOPvroIieMLqn8/PzYfvvt49prry1y/arlUWnPUlTcJ8TFLS/tHyQVyYoVK+KAAw6I2bNnxwUXXBDt27ePOnXqxLfffhsDBw5cq59TaZTlsb3//vvj+OOPL/H2zz33XEyfPj3uu+++uO+++wqtHz16dPTo0aOEiYt3zz33xMCBA6Nv375x3nnnRfPmzaNq1aoxdOjQQic+WJPSvgbL8zlav379OOCAA+KAAw6I6tWrx5133hlvvPFGdOnSpdSvwfLw0ksvxSGHHBL77rtv3HzzzbHxxhtH9erVY+TIkcVO0L0uNW7cODKZTIkKv/79+8fw4cOjXr16cdRRR2XnoFtVaZ87ZT0D28rRUMWdNXDChAnZ0YTlLS8vL/r27RsPP/xw3HzzzTFz5sx45ZVX4sorr8xuU5r3y5VW/hyaNm26TnIXp0mTJqUqfX+tpMVOWmfaK8n9/fTTT1G7du0yZ+rQoUN89tln8cQTT8S4cePioYceiptvvjkuvvjiGDJkSESU7DkCsCZKKWCDtvKPiG7dusWNN94YF154YXbYfPXq1Ut0lp7GjRvH8ccfH8cff3wsWLAg9t1337j00kuLLaVatWoV+fn58eWXXxb4RPSzzz4rtG2jRo1izpw5hZav+uls27ZtI0mSaN26dZk+nXzwwQdjwIAB8X//93/ZZYsXLy7yvotS3C/sbdu2jffffz/233//nHwqXpyVE90W9Zh/+umn0bRp06hTp07UrFkzatWqVeThmUVdt6xWjk758MMPC/0Bt9IHH3wQn3/+edx5551x3HHHZZf/+kyPK5X0sW7WrFnUrl272MehSpUq5VJa9OzZs8icxRk9enQ0b948e6awXxs7dmw8/PDDMWLEiKhVq1a0atWqzD+fBx98MNq0aRNjx44t8Jitephe27ZtY/z48TF79uxiR0ut7WuwKGV5zeyyyy5x5513xvTp07O5SvIaLM1+aU0eeuihqFmzZowfP77AKNSRI0eW+rbKQ7Vq1aJt27YxZcqUNW7bv3//uPjii2P69OmrnQy8pM+dtbFw4cJ49NFH46ijjoojjjii0PozzzwzRo8eHd26dSuwD1nd+1Zpn1NHHXVU3HnnnfHss8/GJ598EkmSFDgsq7TvlxERU6ZMyY7GTVP79u3joYceKtV1Vr4uvvjii+jQoUN2+cyZM2POnDmlmjS9qP3U559/HrVr184+FuW5P54yZUqBzMVZ3XOiTp06cdRRR8VRRx0VS5cujcMPPzyuuOKKGDx4cPZQ8zU9RwDWxOF7wAava9eusdtuu8WwYcNi8eLF0bx58+jatWvccsst2T/sfm3WrFnZ///4448F1tWtWze23HLLQqeK/rXevXtHRMT1119fYPmwYcMKbdu2bduYO3dugUMCp0+fXuiMP4cffnhUrVo1hgwZUmikRZIkhXKuqmrVqoWud8MNNxQakVWclWfgW9WRRx4Z3377bdx6662F1i1atKjQ2YLSsvHGG8dOO+0Ud955Z4Hi7cMPP4ynnnoq+vTpExG/PC49e/aMRx55JKZOnZrd7pNPPonx48eXW54ePXpEvXr1YujQodnDrlZa+XNZObLm1z+nJEmyp3v/tTp16kRErLFUrFq1avTo0SMeffTR7FkoI375g2vMmDGx9957R/369cvyLRWw8cYbR/fu3QtcirNo0aIYO3ZsHHTQQXHEEUcUupxxxhkxf/787Onm+/TpE6+//nq8+eab2duYNWtWgfl2ilPUY/rGG2/Ea6+9VmC7fv36RZIk2dEBv7byumv7GixKcT/Hn3/+uVDGlZ588smI+H+H3ZX0NVia/dKaVK1aNTKZTIH9x1dffZWTM66t1Llz53j77bfXuF3btm1j2LBhMXTo0ALzlK2qpM+dtfHwww/HwoULY9CgQUW+Fg466KB46KGHYsmSJbHzzjtH69atY9iwYYWeL7/OWNJ9w0rdu3ePxo0bx/333x/3339/7LbbbgUOGSvN++VK77zzToG5ANPSuXPn+Omnn0o8b1tEZN8LVn0drBx5WJozSb722msF5hubNm1aPProo9GjR4+oWrVque+P33333dhzzz0LLJs6dWp8+umnBZYV9/696j6rRo0asc0220SSJLFs2bLs8jU9RwDWxEgpgIg477zz4je/+U2MGjUqTj311Ljpppti7733ju233z5OOumkaNOmTcycOTNee+21+Oabb+L999+PiF8m2u7atWt06tQpGjduHG+//XY8+OCDccYZZxR7XzvttFMcc8wxcfPNN8fcuXNjzz33jGeffTYmT55caNujjz46LrjggjjssMPizDPPjJ9//jmGDx8eW221VYFfbtu2bRuXX355DB48OL766qvo27dv1KtXL6ZMmRIPP/xwnHzyyXHuuecWm+mggw6Ku+++Oxo0aBDbbLNNvPbaa/HMM89EkyZNSvT4derUKe6///4455xzYtddd426devGwQcfHMcee2z8+9//jlNPPTWef/752GuvvWLFihXx6aefxr///e8YP3589lDKtF1zzTXRu3fv6Ny5c5xwwgmxaNGiuOGGG6JBgwZx6aWXZrcbMmRIjBs3LvbZZ584/fTTY/ny5XHDDTfEtttuu8b5w0qqfv36cd1118WJJ54Yu+66a/Tv3z8aNWoU77//fvz8889x5513Rvv27aNt27Zx7rnnxrfffhv169ePhx56qMjDUVZOpnvmmWdGz549o2rVqnH00UcXed+XX355PP3007H33nvH6aefHtWqVYtbbrkllixZEldffXW5fH+l8dhjj8X8+fPjkEMOKXL9HnvsEc2aNYvRo0fHUUcdFeeff37cfffd0atXr/jjH/8YderUiX/961/RqlWrNf58DjrooBg7dmwcdthhceCBB8aUKVNixIgRsc0228SCBQuy23Xr1i2OPfbYuP766+OLL76IXr16RX5+frz00kvRrVu3OOOMM9b6NViUtm3bRsOGDWPEiBFRr169qFOnTuy+++5Rr1692HPPPWOPPfaIXr16RcuWLWPOnDnxyCOPxEsvvRR9+/aNjh07RkSU+DVYmv3Smhx44IFx7bXXRq9evaJ///7x/fffx0033RRbbrllub1mSuvQQw+Nu+++Oz7//PM1jmT74x//uMbbK+lzZ22MHj06mjRpUqhYWOmQQw6JW2+9Nf7zn//E4YcfHsOHD4+DDz44dtpppzj++ONj4403jk8//TQ++uijbIlemn1DxC8joA4//PC47777YuHChfGPf/yj0DYlfb+M+GV+qUmTJsWgQYPW5qEpkwMPPDCqVasWzzzzTJx88sklus6OO+4YAwYMiH/9618xZ86c6NKlS7z55ptx5513Rt++fUt16OR2220XPXv2jDPPPDPy8vLi5ptvjogoUHaX1/74nXfeidmzZ8ehhx5aYPlxxx0XEyZMKFBUFvf+3aNHj2jRokXstddesdFGG8Unn3wSN954Yxx44IFRr1697PVL8hwBWK11fn4/gApi5SmU33rrrULrVqxYkbRt2zZp27Zt9tTGX375ZXLcccclLVq0SKpXr55suummyUEHHZQ8+OCD2etdfvnlyW677ZY0bNgwqVWrVtK+ffvkiiuuSJYuXZrdZuUpx39t0aJFyZlnnpk0adIkqVOnTnLwwQcn06ZNK/LU60899VSy3XbbJTVq1Ei23nrr5J577inyNpMkSR566KFk7733TurUqZPUqVMnad++fTJo0KDks88+W+1j89NPPyXHH3980rRp06Ru3bpJz549k08//bTQ6dpXnt78mmuuKXD9BQsWJP37908aNmyYRESB00svXbo0ueqqq5Jtt902ycvLSxo1apR06tQpGTJkSDJ37tzsdrHKKdfXpFWrVsmBBx5YaHlRt1Nc7meeeSbZa6+9klq1aiX169dPDj744OTjjz8udJsTJkxIOnXqlNSoUSNp06ZNMmLEiCJ/BiV9vFY+F1eednylxx57LNlzzz2zeXbbbbfk3nvvza7/+OOPk+7duyd169ZNmjZtmpx00knJ+++/n0REMnLkyOx2y5cvT/7whz8kzZo1SzKZTIGcRT3H3n333aRnz55J3bp1k9q1ayfdunVLXn311SIzr/r6ef7555OISJ5//vlCj1tZHHzwwUnNmjWThQsXFrvNwIEDk+rVqyc//PBDkiRJMmnSpKRLly5JzZo1k0033TT529/+ltx+++2FHuMuXbokXbp0yX6dn5+fXHnllUmrVq2SvLy8pGPHjskTTzyRDBgwoNAp0pcvX55cc801Sfv27ZMaNWokzZo1S3r37p288847BbYryWuwS5cuRZ5qvaj7ffTRR5NtttkmqVatWvbnvGzZsuTWW29N+vbtm81eu3btpGPHjsk111yTLFmypMBtlPQ1WJr90prcfvvtSbt27ZK8vLykffv2yciRI0v0minv59NKS5YsSZo2bZr87W9/K7B8ZaZZs2at9vqr7ldK+twpbh+wJjNnzkyqVauWHHvsscVu8/PPPye1a9dODjvssOyyl19+OTnggAOSevXqJXXq1El22GGH5IYbbsiuL+2+IUmS5Omnn04iIslkMsm0adOKzFKS98skSZLhw4cntWvXTubNm1fSh6JcHXLIIcn+++9f7Pprrrmm0H5j2bJlyZAhQ5LWrVsn1atXT1q2bJkMHjw4Wbx4cYHrFveelCT/7/lzzz33ZF8XHTt2LPJ5vjb745UuuOCCZPPNN0/y8/MLLO/SpUuh12Bx79+33HJLsu+++yZNmjRJ8vLykrZt2ybnnXdegX3GSiV5jgAUJ5MklXjWVwCAEtpnn30iLy8vnnnmmVxHqVQymUxccsklBUYQVkZ/+9vfYuTIkfHFF18UO9E861bHjh2ja9eucd111+Xk/l966aXo2rVrfPrpp0WeDW9dyWQyMWjQoLjxxhvX+X0tWbIktthii7jwwgtLNOoPINfMKQUAbBCmT5+e+hm/qDjOPvvsWLBgQZFndGTdGzduXHzxxRcxePDgnGXYZ599okePHjk5NDktI0eOjOrVq8epp56a6ygAJWJOKQBgvfbqq6/G2LFj48svv4wLLrgg13EqrRUrVhQ5cfWv1a1bN+rWrVsu97do0aIiJ2D+taVLl0aNGjWKXV+rVq1o0KBBNtv3339fLtnW1qxZs1Z7IokaNWoUe6bHyqpXr16F5ttasGDBaufgWvkYrW5k25qeA1WrVi1wpr+VJwNYX5166qkKKaBSUUoBAOu1W2+9NZ588sk466yz4vjjj891nEpr2rRpazyrVnke5nf//fev9c9rwIABMWrUqHLJU5523XXX+Prrr4td36VLl3jhhRfSC5Qj//jHP4o8q2VpdOnSJSZMmFDs+latWhU4mx0AFUtO55QaPnx4DB8+PPtGse2228bFF1+cPS1x165dC73JnHLKKTFixIi0owIAbNAWL14cL7/88mq3adOmTbRp06Zc7m/69Onx0UcfrXabn376KRo1alTs+k022SS22WabcslTnl555ZVYtGhRsesbNWqUPVPe+ux///tf/O9//yt2/eLFiyMiombNmsVu06hRoyLPQrpSrVq1Yq+99ip7SADWqZyWUo8//nhUrVo12rVrF0mSxJ133hnXXHNNTJw4Mbbddtvo2rVrbLXVVnHZZZdlr1O7du2oX79+riIDAAAAUA5yevjewQcfXODrK664IoYPHx6vv/56bLvtthHxSwnVokWLXMQDAAAAYB2pMHNKrVixIh544IFYuHBhdO7cObt89OjRcc8990SLFi3i4IMPjr/+9a9Ru3btYm9nyZIlsWTJkuzX+fn5MXv27GjSpElkMpl1+j0AAAAAbOiSJIn58+fHJptsElWqVCl2u5yXUh988EF07tw5Fi9eHHXr1o2HH344e+x///79o1WrVrHJJpvEpEmT4oILLojPPvssxo4dW+ztDR06dK0nTAQAAABg7UybNi0222yzYtfndE6piF9O4zp16tSYO3duPPjgg3HbbbfFhAkTipyU8rnnnov9998/Jk+eHG3bti3y9lYdKTV37tzYfPPNY9q0aeaiAgAAAFjH5s2bFy1btow5c+ZEgwYNit0u56XUqrp37x5t27aNW265pdC6hQsXRt26dWPcuHHRs2fPEt3evHnzokGDBjF37lylFAAAAMA6VtIupvgD+3IkPz+/wEinX3vvvfciImLjjTdOMREAAAAA5S2nc0oNHjw4evfuHZtvvnnMnz8/xowZEy+88EKMHz8+vvzyyxgzZkz06dMnmjRpEpMmTYqzzz479t1339hhhx1yGRsAAACAtZTTUur777+P4447LqZPnx4NGjSIHXbYIcaPHx8HHHBATJs2LZ555pkYNmxYLFy4MFq2bBn9+vWLv/zlL7mMDAAAAEA5qHBzSpU3c0oBAAAApKfSzikFAAAAwPpPKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKROKQUAAABA6pRSAAAAAKSuWq4DAABAsTKZ0m2fJOsmBwBQ7oyUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB11XIdAAAAysvfJ/5Q4m0v7Nh0HSYBANbESCkAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUpfTUmr48OGxww47RP369aN+/frRuXPnePLJJ7PrFy9eHIMGDYomTZpE3bp1o1+/fjFz5swcJgYAAACgPOS0lNpss83i73//e7zzzjvx9ttvx3777ReHHnpofPTRRxERcfbZZ8fjjz8eDzzwQEyYMCG+++67OPzww3MZGQAAAIBykEmSJMl1iF9r3LhxXHPNNXHEEUdEs2bNYsyYMXHEEUdERMSnn34aHTp0iNdeey322GOPEt3evHnzokGDBjF37tyoX7/+uowOAEB5y2RKtfnf351V4m0v7Ni0tGkAgBIoaRdTYeaUWrFiRdx3332xcOHC6Ny5c7zzzjuxbNmy6N69e3ab9u3bx+abbx6vvfZasbezZMmSmDdvXoELAAAAABVLzkupDz74IOrWrRt5eXlx6qmnxsMPPxzbbLNNzJgxI2rUqBENGzYssP1GG20UM2bMKPb2hg4dGg0aNMheWrZsuY6/AwAAAABKK+el1NZbbx3vvfdevPHGG3HaaafFgAED4uOPPy7z7Q0ePDjmzp2bvUybNq0c0wIAAABQHqrlOkCNGjViyy23jIiITp06xVtvvRX//Oc/46ijjoqlS5fGnDlzCoyWmjlzZrRo0aLY28vLy4u8vLx1HRsAAACAtZDzkVKrys/PjyVLlkSnTp2ievXq8eyzz2bXffbZZzF16tTo3LlzDhMCAAAAsLZyOlJq8ODB0bt379h8881j/vz5MWbMmHjhhRdi/Pjx0aBBgzjhhBPinHPOicaNG0f9+vXjD3/4Q3Tu3LnEZ94DAAAAoGLKaSn1/fffx3HHHRfTp0+PBg0axA477BDjx4+PAw44ICIirrvuuqhSpUr069cvlixZEj179oybb745l5EBAAAAKAeZJEmSXIdYl+bNmxcNGjSIuXPnRv369XMdBwCA0shkSrX539+dVeJtL+zYtLRpAIASKGkXU+HmlAIAAABg/aeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1SikAAAAAUqeUAgAAACB1OS2lhg4dGrvuumvUq1cvmjdvHn379o3PPvuswDZdu3aNTCZT4HLqqafmKDEAAAAA5SGnpdSECRNi0KBB8frrr8fTTz8dy5Ytix49esTChQsLbHfSSSfF9OnTs5err746R4kBAAAAKA/Vcnnn48aNK/D1qFGjonnz5vHOO+/Evvvum11eu3btaNGiRdrxAAAAAFhHKtScUnPnzo2IiMaNGxdYPnr06GjatGlst912MXjw4Pj5559zEQ8AAACAcpLTkVK/lp+fH2eddVbstddesd1222WX9+/fP1q1ahWbbLJJTJo0KS644IL47LPPYuzYsUXezpIlS2LJkiXZr+fNm7fOswMAAABQOhWmlBo0aFB8+OGH8fLLLxdYfvLJJ2f/v/3228fGG28c+++/f3z55ZfRtm3bQrczdOjQGDJkyDrPCwAAAEDZVYjD984444x44okn4vnnn4/NNttstdvuvvvuERExefLkItcPHjw45s6dm71Mmzat3PMCAAAAsHZyOlIqSZL4wx/+EA8//HC88MIL0bp16zVe57333ouIiI033rjI9Xl5eZGXl1eeMQEAAAAoZzktpQYNGhRjxoyJRx99NOrVqxczZsyIiIgGDRpErVq14ssvv4wxY8ZEnz59okmTJjFp0qQ4++yzY999940ddtghl9EBAAAAWAs5LaWGDx8eERFdu3YtsHzkyJExcODAqFGjRjzzzDMxbNiwWLhwYbRs2TL69esXf/nLX3KQFgAAAIDykvPD91anZcuWMWHChJTSAAAAAJCWCjHROQAAAAAbFqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKlTSgEAAACQupyWUkOHDo1dd9016tWrF82bN4++ffvGZ599VmCbxYsXx6BBg6JJkyZRt27d6NevX8ycOTNHiQEAAAAoDzktpSZMmBCDBg2K119/PZ5++ulYtmxZ9OjRIxYuXJjd5uyzz47HH388HnjggZgwYUJ89913cfjhh+cwNQAAAABrK5MkSZLrECvNmjUrmjdvHhMmTIh999035s6dG82aNYsxY8bEEUccERERn376aXTo0CFee+212GOPPdZ4m/PmzYsGDRrE3Llzo379+uv6WwAAoDxlMqXa/O/vzirxthd2bFraNABACZS0i6lQc0rNnTs3IiIaN24cERHvvPNOLFu2LLp3757dpn379rH55pvHa6+9lpOMAAAAAKy9arkOsFJ+fn6cddZZsddee8V2220XEREzZsyIGjVqRMOGDQtsu9FGG8WMGTOKvJ0lS5bEkiVLsl/PmzdvnWUGAAAAoGwqzEipQYMGxYcffhj33XffWt3O0KFDo0GDBtlLy5YtyykhAAAAAOWlQpRSZ5xxRjzxxBPx/PPPx2abbZZd3qJFi1i6dGnMmTOnwPYzZ86MFi1aFHlbgwcPjrlz52Yv06ZNW5fRAQAAACiDnJZSSZLEGWecEQ8//HA899xz0bp16wLrO3XqFNWrV49nn302u+yzzz6LqVOnRufOnYu8zby8vKhfv36BCwAAAAAVS07nlBo0aFCMGTMmHn300ahXr152nqgGDRpErVq1okGDBnHCCSfEOeecE40bN4769evHH/7wh+jcuXOJzrwHAAAAQMWU01Jq+PDhERHRtWvXAstHjhwZAwcOjIiI6667LqpUqRL9+vWLJUuWRM+ePePmm29OOSkAAAAA5SmnpVSSJGvcpmbNmnHTTTfFTTfdlEIiAAAAANJQISY6BwAAAGDDopQCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVlLqXmzJkTt912WwwePDhmz54dERHvvvtufPvtt+UWDgAAAID1U7WyXGnSpEnRvXv3aNCgQXz11Vdx0kknRePGjWPs2LExderUuOuuu8o7JwAAAADrkTKNlDrnnHNi4MCB8cUXX0TNmjWzy/v06RMvvvhiuYUDAAAAYP1UplLqrbfeilNOOaXQ8k033TRmzJix1qEAAAAAWL+VqZTKy8uLefPmFVr++eefR7NmzdY6FAAAAADrtzKVUoccckhcdtllsWzZsoiIyGQyMXXq1LjggguiX79+5RoQAAAAgPVPmUqp//u//4sFCxZE8+bNY9GiRdGlS5fYcssto169enHFFVeUd0YAAAAA1jNlOvtegwYN4umnn46XX345Jk2aFAsWLIidd945unfvXt75AAAAAFgPlamUWmnvvfeOvffeu7yyAAAAALCBKFMpdf311xe5PJPJRM2aNWPLLbeMfffdN6pWrbpW4QAAAABYP5WplLruuuti1qxZ8fPPP0ejRo0iIuKnn36K2rVrR926deP777+PNm3axPPPPx8tW7Ys18AAAAAAVH5lmuj8yiuvjF133TW++OKL+PHHH+PHH3+Mzz//PHbffff45z//GVOnTo0WLVrE2WefXd55AQAAAFgPlGmk1F/+8pd46KGHom3bttllW265ZfzjH/+Ifv36xf/+97+4+uqro1+/fuUWFAAAAID1R5lGSk2fPj2WL19eaPny5ctjxowZERGxySabxPz589cuHQAAAADrpTKVUt26dYtTTjklJk6cmF02ceLEOO2002K//faLiIgPPvggWrduXT4pAQAAAFivlKmUuv3226Nx48bRqVOnyMvLi7y8vNhll12icePGcfvtt0dERN26deP//u//yjUsAAAAAOuHMs0p1aJFi3j66afj008/jc8//zwiIrbeeuvYeuuts9t069atfBICAAAAsN4pUym1Uvv27aN9+/bllQUAAACADUSZS6lvvvkmHnvssZg6dWosXbq0wLprr712rYMBAAAAsP4qUyn17LPPxiGHHBJt2rSJTz/9NLbbbrv46quvIkmS2Hnnncs7IwAAAADrmTJNdD548OA499xz44MPPoiaNWvGQw89FNOmTYsuXbrEb37zm/LOCAAAAMB6pkyl1CeffBLHHXdcRERUq1YtFi1aFHXr1o3LLrssrrrqqnINCAAAAMD6p0ylVJ06dbLzSG288cbx5ZdfZtf98MMP5ZMMAAAAgPVWmeaU2mOPPeLll1+ODh06RJ8+feJPf/pTfPDBBzF27NjYY489yjsjAAAAAOuZMpVS1157bSxYsCAiIoYMGRILFiyI+++/P9q1a+fMewAAAACsUZlKqTZt2mT/X6dOnRgxYkS5BQIAAABg/VemOaXatGkTP/74Y6Hlc+bMKVBYAQAAAEBRylRKffXVV7FixYpCy5csWRLffvvtWocCAAAAYP1WqsP3Hnvssez/x48fHw0aNMh+vWLFinj22Wdjiy22KLdwAAAAAKyfSlVK9e3bNyIiMplMDBgwoMC66tWrxxZbbBH/93//V27hAAAAAFg/laqUys/Pj4iI1q1bx1tvvRVNmzZdJ6EAAAAAWL+V6ex7U6ZMKe8cAAAAAGxAylRKRUQ8++yz8eyzz8b333+fHUG10h133LHWwQAAAABYf5WplBoyZEhcdtllscsuu8TGG28cmUymvHMBAAAAsB4rUyk1YsSIGDVqVBx77LHlnQcAAACADUCVslxp6dKlseeee5Z3FgAAAAA2EGUqpU488cQYM2ZMeWcBAAAAYANRpsP3Fi9eHP/617/imWeeiR122CGqV69eYP21115bLuEAAAAAWD+VqZSaNGlS7LTTThER8eGHHxZYZ9JzAAAAANakTKXU888/X945AAAAANiAlGlOqZUmT54c48ePj0WLFkVERJIk5RIKAAAAgPVbmUqpH3/8Mfbff//Yaqutok+fPjF9+vSIiDjhhBPiT3/6U7kGBAAAAGD9U6ZS6uyzz47q1avH1KlTo3bt2tnlRx11VIwbN67cwgEAAACwfirTnFJPPfVUjB8/PjbbbLMCy9u1axdff/11uQQDAAAAYP1VppFSCxcuLDBCaqXZs2dHXl7eWocCAAAAYP1WplJqn332ibvuuiv7dSaTifz8/Lj66qujW7du5RYOAAAAgPVTmQ7fu/rqq2P//fePt99+O5YuXRrnn39+fPTRRzF79ux45ZVXyjsjAAAAAOuZMo2U2m677eLzzz+PvffeOw499NBYuHBhHH744TFx4sRo27ZteWcEAAAAYD1TppFSERENGjSIiy66qDyzAAAAALCBKNNIqZEjR8YDDzxQaPkDDzwQd95551qHAgAAAGD9VqZSaujQodG0adNCy5s3bx5XXnnlWocCAAAAYP1WplJq6tSp0bp160LLW7VqFVOnTl3rUAAAAACs38pUSjVv3jwmTZpUaPn7778fTZo0WetQAAAAAKzfylRKHXPMMXHmmWfG888/HytWrIgVK1bEc889F3/84x/j6KOPLu+MAAAAAKxnynT2vb/97W/x1Vdfxf777x/Vqv1yE/n5+XHccceZUwoAAACANSp1KZUkScyYMSNGjRoVl19+ebz33ntRq1at2H777aNVq1brIiMAAAAA65kylVJbbrllfPTRR9GuXbto167dusgFAAAAwHqs1HNKValSJdq1axc//vjjusgDAAAAwAagTBOd//3vf4/zzjsvPvzww/LOAwAAAMAGoEwTnR933HHx888/x4477hg1atSIWrVqFVg/e/bscgkHAAAAwPqpTKXUsGHDyjkGAAAAABuSMpVSAwYMKO8cAAAAAGxAyjSnVETEl19+GX/5y1/imGOOie+//z4iIp588sn46KOPyi0cAAAAAOunMpVSEyZMiO233z7eeOONGDt2bCxYsCAiIt5///245JJLyjUgAAAAAOufMpVSF154YVx++eXx9NNPR40aNbLL99tvv3j99dfLLRwAAAAA66cylVIffPBBHHbYYYWWN2/ePH744Ye1DgUAAADA+q1MpVTDhg1j+vTphZZPnDgxNt1007UOBQAAAMD6rUyl1NFHHx0XXHBBzJgxIzKZTOTn58crr7wS5557bhx33HHlnREAAACA9UyZSqkrr7wyOnToEJtvvnksWLAgttlmm9h3331jzz33jL/85S/lnREAAACA9Uy10mycn58f11xzTTz22GOxdOnSOPbYY6Nfv36xYMGC6NixY7Rr125d5QQAAABgPVKqUuqKK66ISy+9NLp37x61atWKMWPGRJIkcccdd6yrfAAAAACsh0p1+N5dd90VN998c4wfPz4eeeSRePzxx2P06NGRn5+/rvIBAAAAsB4qVSk1derU6NOnT/br7t27RyaTie+++67cgwEAAACw/ipVKbV8+fKoWbNmgWXVq1ePZcuWlWsoAAAAANZvpZpTKkmSGDhwYOTl5WWXLV68OE499dSoU6dOdtnYsWPLLyEAAAAA651SlVIDBgwotOx3v/tduYUBAAAAYMNQqlJq5MiR6yoHAAAAABuQUs0pBQAAAADlQSkFAAAAQOpyWkq9+OKLcfDBB8cmm2wSmUwmHnnkkQLrBw4cGJlMpsClV69euQkLAAAAQLnJaSm1cOHC2HHHHeOmm24qdptevXrF9OnTs5d77703xYQAAAAArAulmui8vPXu3Tt69+692m3y8vKiRYsWKSUCAAAAIA0Vfk6pF154IZo3bx5bb711nHbaafHjjz+udvslS5bEvHnzClwAAAAAqFgqdCnVq1evuOuuu+LZZ5+Nq666KiZMmBC9e/eOFStWFHudoUOHRoMGDbKXli1bppgYAAAAgJLI6eF7a3L00Udn/7/99tvHDjvsEG3bto0XXngh9t9//yKvM3jw4DjnnHOyX8+bN08xBQAAAFDBVOiRUqtq06ZNNG3aNCZPnlzsNnl5eVG/fv0CFwAAAAAqlkpVSn3zzTfx448/xsYbb5zrKAAAAACshZwevrdgwYICo56mTJkS7733XjRu3DgaN24cQ4YMiX79+kWLFi3iyy+/jPPPPz+23HLL6NmzZw5TAwAAALC2clpKvf3229GtW7fs1yvnghowYEAMHz48Jk2aFHfeeWfMmTMnNtlkk+jRo0f87W9/i7y8vFxFBgAAAKAc5LSU6tq1ayRJUuz68ePHp5gGAAAAgLRUqjmlAAAAAFg/KKUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUKaUAAAAASJ1SCgAAAIDUVct1ANjQ/X3iD6Xa/sKOTddREgAAAEiPkVIAAAAApE4pBQAAAEDqlFIAAAAApC6npdSLL74YBx98cGyyySaRyWTikUceKbA+SZK4+OKLY+ONN45atWpF9+7d44svvshNWAAAAADKTU5LqYULF8aOO+4YN910U5Hrr7766rj++utjxIgR8cYbb0SdOnWiZ8+esXjx4pSTAgAAAFCecnr2vd69e0fv3r2LXJckSQwbNiz+8pe/xKGHHhoREXfddVdstNFG8cgjj8TRRx+dZlQAAAAAylGFnVNqypQpMWPGjOjevXt2WYMGDWL33XeP1157rdjrLVmyJObNm1fgAgAAAEDFUmFLqRkzZkRExEYbbVRg+UYbbZRdV5ShQ4dGgwYNspeWLVuu05wAAAAAlF6FLaXKavDgwTF37tzsZdq0abmOBAAAAMAqKmwp1aJFi4iImDlzZoHlM2fOzK4rSl5eXtSvX7/ABQAAAICKpcKWUq1bt44WLVrEs88+m102b968eOONN6Jz5845TAYAAADA2srp2fcWLFgQkydPzn49ZcqUeO+996Jx48ax+eabx1lnnRWXX355tGvXLlq3bh1//etfY5NNNom+ffvmLjQAAAAAay2npdTbb78d3bp1y359zjnnRETEgAEDYtSoUXH++efHwoUL4+STT445c+bE3nvvHePGjYuaNWvmKjIAAAAA5SCnpVTXrl0jSZJi12cymbjsssvisssuSzEVAAAAAOtahZ1TCgAAAID1l1IKAAAAgNQppQAAAABInVIKAAAAgNTldKJzAKByy2Qypdp+dSc4AQBgw2KkFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDoTnQMAAOuPUp6AIZyAASBnjJQCAAAAIHVGSgEAABVWppQjn4x7Aqg8jJQCAAAAIHVKKQAAAABSp5SCdSGTKfkFAAAANkBKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHXVch0AqLz+PvGHUm1/Ycem6ygJAAAAlY2RUgAAAACkTikFFJTJlPwCAAAAZaSUAgAAACB1SikAAAAAUqeUAgAAACB1zr4HAABssEpzNmFnEgYoX0ZKAQAAAJA6pRQAAAAAqVNKAQAAAJA6pRQAAAAAqTPROaznMplMqbZP1lEOAAAA+DUjpQAAAABInZFSAECFVZpTtUc4XTsAQGVipBQAAAAAqVNKAQAAAJA6pRQAAAAAqVNKAQAAAJA6pRQAAAAAqVNKAQAAAJA6pRQAAAAAqVNKAQAAAJA6pRQAAAAAqVNKAQAAAJA6pRQAAAAAqVNKAQAAAJA6pRQAAAAAqVNKAQAAAJC6arkOQOlkMplSbZ8kyTpKAgAAAFB2RkoBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFACQnkymdBcAANZbSikAAAAAUqeUAgAAACB11XIdAACADUemlIdlJusoBwCQe0ZKAQAAAJA6pRQAAAAAqVNKAQAAAJA6c0oBrAOlmTMlScyYAgAAbHiMlAIAAAAgdUopAAAAAFKnlAIAAAAgdeaUAoAcK80cZBHmIQMAYP1gpBQAAAAAqVNKAQAAAJA6pRQAAAAAqVNKAQAAAJA6E50DAMB6zgkVKC3PGSANRkoBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpU0oBAAAAkDqlFAAAAACpq9Cl1KWXXhqZTKbApX379rmOxQZo1efhmi5QKplM6S4AQKXgd0iA1auW6wBrsu2228YzzzyT/bpatQofGQAAAIA1qPANT7Vq1aJFixa5jgEAAABAOarQh+9FRHzxxRexySabRJs2beK3v/1tTJ06NdeRAAAAAFhLFXqk1O677x6jRo2KrbfeOqZPnx5DhgyJffbZJz788MOoV69ekddZsmRJLFmyJPv1vHnz0ooLAAAAQAlV6FKqd+/e2f/vsMMOsfvuu0erVq3i3//+d5xwwglFXmfo0KExZMiQtCICAAAAUAYV/vC9X2vYsGFstdVWMXny5GK3GTx4cMydOzd7mTZtWooJAQAAACiJSlVKLViwIL788svYeOONi90mLy8v6tevX+ACAABrK5PJlOoCAKxehS6lzj333JgwYUJ89dVX8eqrr8Zhhx0WVatWjWOOOSbX0QAAAABYCxV6TqlvvvkmjjnmmPjxxx+jWbNmsffee8frr78ezZo1y3U0gJz5+8QfSrX9hR2brqMkAAAAZVehS6n77rsv1xEAAAAAWAcq9OF7AAAAAKyflFIAAAAApE4pBQAAAEDqlFIAAAAApE4pBQAAAEDqlFIAAAAApE4pBQAAAEDqlFIAAAAApE4pBQAAAEDqquU6AACUl0wmU+JtkyRZh0kAAIA1UUoBAAAFlaLkj4gIRT8AZeDwPQAAAABSp5QCAAAAIHVKKQAAAABSp5QCAAAAIHUmOifr7xN/KNX2F3Zsuo6SAAAAAOs7I6UAAAAASJ1SCgAAAIDUKaXWd5lMyS8AAAAAKVFKAQAAAJA6pRQAAAAAqXP2PQCASiZTysPukyRZR0kAAMpOKQUAAMDaKU1ZrigH/n8O3wMAAAAgdUopAAAAAFLn8D0AAKgA/j7xhxJve2HHpuswCQCkw0gpAAAAAFJnpBRQYZXm7FLOLAUAUDmUZlRghJGBsD4zUgoAAACA1CmlAAAAAEidUgoAANaFTKZ0FwDYwCilAAAAAEidUgoAAACA1CmlAAAAAEhdtVwHAAAAKre/T/yhxNte2LHpOkwCQGVipBQAAAAAqVNKAQAAAJA6pRQAAAAAqVNKAQAAAJA6E50DAABUQiaYByo7I6UAAAAASJ1SCgAAAIDUKaUAAAAASJ05pQAoIJPJlHjbJEnWYZKKpTTzdkSYuwMAKL3S/B4WsWH9Lsb6SSkFAEABSljIkVIWEvHurHWTAyAlDt8DYMOUyZTuAgAAlCulFAAAAACpU0oBAAAAkDqlFAAAAACpM9E5ALBBcoYjAIDcMlIKAAAAgNQZKQVskJzuHAAAILeMlALWD5lM6S4AAADklFIKAAAAgNQppQAAAABInVIKAGB95xBnAKACMtE5AKznTOwPAEBFpJQiNZlSfvKaJMk6SgIAAADkmsP3AAAAAEidUgoAKhvzAwEAsB5QSgEAAACQOqUUAAAAAKlTSgEAAACQOqUUAAAAAKmrlusAAADro79P/KFU21/Ysek6SgIAUDEZKQUAAABA6pRSAAAlkcmU7gIAwGoppQAAAABInVIKAAAAgNSZ6ByAsivtIUpJsm5yAAAAlY5SCgAAgPVGppQfmiU+NIOccfgeAAAAAKkzUgqA1Px94g+l2v7Cjk3XURIAgPVAKUaF/f3dWaW6ab+HkQYjpQAAAABInVIKAAAAgNQppQAAAABInVIKAAAAgNQppQAAAABInVIKAAAAgNRVy3UAAAAAqAz+PvGHUm1/Ycem6ygJrB+MlKLiymRKfgEAACgLf3dAzhgpBQAAALAOGF23ekop1gte6AAAAKxzpR0x9+6sdZNjPeHwPQAAAABSp5QCAAAAIHWVopS66aabYosttoiaNWvG7rvvHm+++WauIwEAAACwFip8KXX//ffHOeecE5dcckm8++67seOOO0bPnj3j+++/z3U0AAAAAMqowpdS1157bZx00klx/PHHxzbbbBMjRoyI2rVrxx133JHraAAAAACUUYU++97SpUvjnXfeicGDB2eXValSJbp37x6vvfZakddZsmRJLFmyJPv13LlzIyJi3rx56zZsBVWa73rxgvmlu+15NUoXppRkL+a2ZS/6tku5vezlQ/bV3L7sRd92KbeXvXzIXsxtV6D31AjZy4vsxdx2BcpekfYxEbIXe9vrOPuGYkN93Fd2MEmSrHa7TLKmLXLou+++i0033TReffXV6Ny5c3b5+eefHxMmTIg33nij0HUuvfTSGDJkSJoxAQAAAFjFtGnTYrPNNit2fYUeKVUWgwcPjnPOOSf7dX5+fsyePTuaNGkSmUwmh8kqjnnz5kXLli1j2rRpUb9+/VzHKbHKmjtC9lyRPTdkzw3Zc0P23JA9N2TPDdlzo7Jmr6y5I2Rf3yRJEvPnz49NNtlktdtV6FKqadOmUbVq1Zg5c2aB5TNnzowWLVoUeZ28vLzIy8srsKxhw4brKmKlVr9+/Ur5gqmsuSNkzxXZc0P23JA9N2TPDdlzQ/bckD03Kmv2ypo7Qvb1SYMGDda4TYWe6LxGjRrRqVOnePbZZ7PL8vPz49lnny1wOB8AAAAAlUuFHikVEXHOOefEgAEDYpdddonddtsthg0bFgsXLozjjz8+19EAAAAAKKMKX0odddRRMWvWrLj44otjxowZsdNOO8W4ceNio402ynW0SisvLy8uueSSQoc5VnSVNXeE7Lkie27Inhuy54bsuSF7bsieG7LnRmXNXllzR8i+oarQZ98DAAAAYP1UoeeUAgAAAGD9pJQCAAAAIHVKKQAAAABSp5QCAACASuS7777LdQQoF0qpDdyHH36Y6wgArAMLFizIdQRgFT/88EPMmzcv1zGooC677LL4+eefcx2jTCrz3xRTpkzJdYQy2XbbbWPMmDG5jgFrTSm1AZo/f37861//it122y123HHHXMcp0nPPPRfbbLNNkb+4zZ07N7bddtt46aWXcpCMiux///tfVMYTii5atCieeOKJ7NeDBw+Oc845J3s577zzYvHixTlMSEVz3XXXrXb9/Pnzo2fPnimlKV/ffPNNnHzyybmOAeVmzpw5MWjQoGjatGlstNFG0ahRo2jRokUMHjy40hYQKy1atCjXEYr02muvFXhfjYi46667onXr1tG8efM4+eSTY8mSJTlKV7whQ4ZU2g8Udthhh9h9993j1ltvjfnz5+c6Tqm0bds2WrduHb///e/j7rvvjm+++SbXkUrkiiuuiFNOOSV+85vfxOzZs3MdB8osk1TGv+AokxdffDFuv/32eOihh2KTTTaJww8/PPr16xe77rprrqMVcsghh0S3bt3i7LPPLnL99ddfH88//3w8/PDDKScrufz8/Bg1alSMHTs2vvrqq8hkMtG6des44ogj4thjj41MJpPriEXq06dP3HvvvdGgQYOIiPj73/8ep556ajRs2DAiIn788cfYZ5994uOPP85hyqJVrVo1pk+fHs2bN4+IiKOOOiquv/762GijjXKcbPVGjBgR//nPf+Lxxx+PiIh69erFtttuG7Vq1YqIiE8//TTOP//8Yl8PuVbST/3r16+/jpOU3uGHH16i7caOHbuOk5ROrVq14pZbbonjjjuu0LqFCxdGjx494scff4xPP/00B+nWzvvvvx8777xzrFixItdRCvn9739fou3uuOOOdZxkw5Kfnx/XXHNNPPbYY7F06dLYf//945JLLsnuIyuy2bNnR+fOnePbb7+N3/72t9GhQ4eIiPj4449jzJgx0b59+3j55Zdj0qRJ8frrr8eZZ56Z48Qls2TJkrjxxhvjmmuuiRkzZuQ6TiG9e/eOrl27xgUXXBARER988EHsvPPOMXDgwOjQoUNcc801ccopp8Sll16a26CrqFKlSsyYMSP7e0xl8tJLL8XIkSPjwQcfjPz8/OjXr1+ceOKJsc8+++Q62hq98MIL2csbb7wRS5cujTZt2sR+++0X3bp1i27dulXY3yWnTJkSJ5xwQnz88cdx6623xsEHH5zrSGttwYIFkZ+fX2BZRfwdcqXvvvsuXn755fj+++8L5a4s+/RcU0qt52bMmBGjRo2K22+/PebNmxdHHnlkjBgxIt5///3YZpttch2vWK1atYpx48Zlf3lb1aeffho9evSIqVOnppysZJIkiYMPPjj++9//xo477hjt27ePJEnik08+iQ8++CAOOeSQeOSRR3Ids0irFjv169eP9957L9q0aRMRETNnzoxNNtmkQv7BuOovc/Xq1Yv3338/m72i2meffeL888/P/iKxau577rknbrrppnjttddyGbNYVapUWW3JmiRJZDKZCvmcOf7440u03ciRI9dxktJ58MEH49hjj437778/DjnkkOzyhQsXRs+ePeP777+PCRMmxMYbb5zDlGVTkUupKlWqRKtWraJjx46rHZVZET8wWdPrNCIik8nE8uXLU0pUcn/729/i0ksvje7du0etWrVi/Pjxccwxx1SK8u+ss86KZ599Np555plCf9TOmDEjevToEVtvvXU89dRTcf3118eAAQNylLSwJUuWxKWXXhpPP/101KhRI84///zo27dvjBw5Mi666KKoWrVqnHHGGdnipyLZeOON4/HHH49ddtklIiIuuuiimDBhQrz88ssREfHAAw/EJZdcUuE+YKtSpUrMnDkzmjVrlusoZbZw4cL497//HaNGjYqXXnopttxyyzjhhBNiwIAB0aJFi1zHW6PFixfHq6++mi2p3nzzzVi2bFm0b98+Pvroo1zHK9aNN94YZ599dnTo0CGqVatWYN27776bo1QlN2XKlDjjjDPihRdeKHB0QEX+HTIiYtSoUXHKKadEjRo1okmTJgXeZzOZTPzvf//LYbrKo9qaN6GyOvjgg+PFF1+MAw88MIYNGxa9evWKqlWrxogRI3IdbY1mzpwZ1atXL3Z9tWrVYtasWSkmKp1Ro0bFiy++GM8++2x069atwLrnnnsu+vbtG3fddVeRoxxybdU/tPTW697kyZNj++23z35ds2bNqFLl/x1dvdtuu8WgQYNyEa1Enn/++ez/kySJPn36xG233RabbrppDlOVTEUrm0rqiCOOiDlz5sQxxxwT//nPf6Jr166xcOHC6NWrV8ycObPSFlIV3WmnnRb33ntvTJkyJY4//vj43e9+F40bN851rBJZXVH22muvxfXXX1/oE96K4q677oqbb745TjnllIiIeOaZZ+LAAw+M2267rcC+siJ65JFH4pZbbilylEWLFi3i6quvjj59+sQll1xSoQqpiIiLL744brnllujevXu8+uqr8Zvf/CaOP/74eP311+Paa6+N3/zmN1G1atVcxyzSTz/9VOAxnzBhQvTu3Tv79a677hrTpk3LRbQ12mqrrdZYIFfkQ7Xq1KkTxx9/fBx//PExefLkGDlyZNx0003x17/+NXr16hWPPfZYriOuVs2aNWO//faLvffeO7p16xZPPvlk3HLLLRV65PHXX38dY8eOjUaNGsWhhx5aqJSqDH73u99FkiRxxx13xEYbbVRhjyhZ1V//+te4+OKLY/DgwRX+/agiq3zPWErsySefjDPPPDNOO+20aNeuXa7jlMqmm24aH374YWy55ZZFrp80aVKF/oPr3nvvjT//+c+FCqmIiP322y8uvPDCGD16dIUspSqzTCZT6E2sMrypzZkzp8DcFqsWrvn5+RVy7ouVunTpUuDrqlWrxh577FHhR6hVdieeeGLMnj07Dj300Hj00Ufj4osvju+++y4mTJgQm2yySa7jrZduuummuPbaa2Ps2LFxxx13xODBg+PAAw+ME044IXr06FGh9zeHHnpooWWfffZZXHjhhfH444/Hb3/727jssstykGzNpk6dGn369Ml+3b1798hkMvHdd9/FZpttlsNkazZ9+vTYdttti12/3XbbRZUqVeKSSy5JMVXJPPDAA3HXXXfFIYccEh9++GHssMMOsXz58nj//fcr9HM9ImKjjTaKKVOmRMuWLWPp0qXx7rvvxpAhQ7Lr58+fv9oPP3NpyJAh2SkUKrstt9wy/vznP0erVq1i8ODB8Z///CfXkYq1dOnSeP311+P555/PHsbXsmXL2HfffePGG28s9LtORXHrrbfGn/70p+jevXt89NFHlXaU3fvvvx/vvPNObL311rmOUio///xzHH300QqptaSUWo+9/PLLcfvtt0enTp2iQ4cOceyxx8bRRx+d61gl0qdPn+wnKjVr1iywbtGiRXHJJZfEQQcdlKN0azZp0qS4+uqri13fu3fvuP7661NMVHKVtdiJ+GWUzsCBAyMvLy8ifhmCfeqpp0adOnUKbFfR5gfabLPN4sMPPyz2jXjSpEkV/g8vcuP888+P2bNnx/777x9bbLFFvPDCCxX+ubKmebzmzJmTTpAyysvLi2OOOSaOOeaY+Prrr2PUqFFx+umnx/Lly+Ojjz6KunXr5jriGn333XdxySWXxJ133hk9e/aM9957L7bbbrtcxyrW8uXLC/0uUL169Vi2bFmOEpVc06ZN46uvvir2dTllypQKO3/QN998E506dYqIX8qzvLy8OPvssyvF7wR9+vSJCy+8MK666qp45JFHonbt2gXmNpo0aVK0bds2hwmLd/TRR1fY50RpvPjii3HHHXfEQw89FFWqVIkjjzwyTjjhhFzHKtJ+++0Xb7zxRrRu3Tq6dOkSp5xySowZM6ZCfwAeEdGrV694880348Ybb6z0H3SvHL1Y2UqpE044IR544IG48MILcx2lUlNKrcf22GOP2GOPPWLYsGFx//33xx133BHnnHNO5Ofnx9NPPx0tW7aMevXq5Tpmkf7yl7/E2LFjY6uttoozzjgju4P69NNP46abbooVK1bERRddlOOUxZs9e/ZqJ0TcaKON4qeffkoxUcmtqdipyCN2Vj304Xe/+12OkpROnz594uKLL44DDzywyBJ2yJAhceCBB+YoHRXRqsVO9erVo2nTpvHHP/6xwPKKVsBGxBpHADRo0KDS/HK9cp6mJEkq7HwXvzZ37ty48sor44Ybboiddtopnn322UoxCfGq70sRRX/oUBGf7z179oyLLrooOy/Try1ZsiT7AVxFtGLFigKZq1WrVilK14hf5iE7/PDDo0uXLlG3bt248847C3wvd9xxR/To0SOHCYtWGQq/1fnuu+9i1KhRMWrUqJg8eXLsueeecf3118eRRx5Z6APCiuSll16KjTfeOPbbb7/o2rVrdOnSJZo0aZLrWGu0YsWK9eaDy9tuuy1OPfXU+Pbbb2O77bYrNJJxhx12yFGy1Rs6dGgcdNBBMW7cuNh+++0L5b722mtzlKxyMdH5Buazzz6L22+/Pe6+++6YM2dOHHDAARX22O6vv/46TjvttBg/fnx2XqNMJhM9e/aMm266KVq3bp3jhMWrWrVqzJgxo9ghtBV5svDKOvFzZTZz5szYaaedokaNGnHGGWfEVlttFRG/vF5vvPHGWL58eUycOLHCnvllVfXq1YtJkyZV6NdoZed1mjtLlizJHr738ssvx0EHHRTHH3989OrVq0IP37/66qvjqquuihYtWsSVV15Z5OF8FVVlfr5/8803scsuu0ReXl4MGjSowIlPbr755liyZEm89dZbsfnmm+c6aiFVqlSJ3r17Z8vAxx9/PPbbb78KP/r41+bOnRt169YtNPfV7Nmzo27duoWKwlyrzGff6927dzzzzDPRtGnTOO644+L3v/99pRn1snDhwnjppZfihRdeiOeffz7ee++92GqrraJLly7ZkqqyHhZXWbz++uvRv3//+Oqrr7LLVn7oU5EnOr/88svj4osvjq233rrQXFiZTCaee+65HKarPJRSG6gVK1bE448/HnfccUeFLaVW+umnn2Ly5MmRJEm0a9cuGjVqlOtIa7TqL3KrWrJkSYwbN67C7mBJ35QpU+K0006Lp59+ukAJe8ABB8TNN99coednWnXUTmX8wwVK4vTTT4/77rsvWrZsGb///e/jt7/9bTRt2jTXsUqkSpUqUatWrejevftqJ6f2Oi1/U6ZMidNPPz2eeuqpQvv3G2+8sdj5M3OtMpeBpO+QQw6JE044IQ466KAKOwF+Sc2fPz9efvnl7PxS77//frRr1y4+/PDDXEdbb22zzTbRoUOHOP/884uc6LxVq1Y5SrZ6jRo1iuuuuy4GDhyY6yiVmlIK1gG/yFFWs2fPjsmTJ0fELxOEVoYze3m+s6GoUqVKbL755tGxY8fVHmZTEYudgQMHlujQIK/Tdeenn36KL774IiIqz/4dNkT5+fnx1ltvxfPPPx/PP/98vPzyy7F48WIfJq9DderUiffff7/ClvTFadGiRbz00kuV7qRiFY1SCgCgBBQ7AOuf/Pz8ePvtt7OH773yyiuxcOHC2HTTTaNbt27ZS0UdrbM+OPjgg2PgwIHRr1+/XEcplaFDh8b06dMr7AmsKgulFAAAABuk+vXrx8KFC6NFixbZAqpr164V9gyN66N//etfcfnll8fvf//7IicMP+SQQ3KUbPUOO+yweO6556JJkyax7bbbFspdEUdOV0RKKQAAADZIt9xyS3Tr1i17ohnSt7oThVTkic7XNIWFkdMlo5QCAAAAKKHly5fHmDFjokePHtGiRYtcx6nUKu65iwEAAID11rJly6JatWqV7uyG1apVi1NPPTWWLFmS6yiVnlIKAAAASF316tVj8803r7CH6K3ObrvtFhMnTsx1jEqvWq4DAAAAABumiy66KP785z/H3XffHY0bN851nBI7/fTT409/+lN888030alTp6hTp06B9TvssEOOklUu5pQCAAAAcqJjx44xefLkWLZsWbRq1apQufPuu+/mKNnqFTVBeyaTiSRJKvQE7RWNkVIAAABATvTt2zfXEcpkypQpuY6wXjBSCgAAAIDUmegcAAAAyJk5c+bEbbfdFoMHD47Zs2dHxC+H7X377bc5TrZ6d999d+y1116xySabxNdffx0REcOGDYtHH300x8kqD6UUAAAAkBOTJk2KrbbaKq666qr4xz/+EXPmzImIiLFjx8bgwYNzG241hg8fHuecc0706dMn5syZk51DqmHDhjFs2LDchqtElFIAAABATpxzzjkxcODA+OKLL6JmzZrZ5X369IkXX3wxh8lW74Ybbohbb701LrrooqhatWp2+S677BIffPBBDpNVLkopAAAAICfeeuutOOWUUwot33TTTWPGjBk5SFQyU6ZMiY4dOxZanpeXFwsXLsxBospJKQUAAADkRF5eXsybN6/Q8s8//zyaNWuWg0Ql07p163jvvfcKLR83blx06NAh/UCVVLVcBwAAAAA2TIccckhcdtll8e9//zsiIjKZTEydOjUuuOCC6NevX47TFe+cc86JQYMGxeLFiyNJknjzzTfj3nvvjaFDh8Ztt92W63iVRiZJkiTXIQAAAIANz9y5c+OII46It99+O+bPnx+bbLJJzJgxIzp37hz//e9/o06dOrmOWKzRo0fHpZdeGl9++WVERGyyySYxZMiQOOGEE3KcrPJQSgEAAAA59corr8T7778fCxYsiJ133jm6d++e60gl9vPPP8eCBQuiefPmuY5S6SilAAAAgJy466674qijjoq8vLwCy5cuXRr33XdfHHfccTlKtnr77bdfjB07Nho2bFhg+bx586Jv377x3HPP5SZYJaOUAgAAAHKiatWqMX369EKjjH788cdo3rx5rFixIkfJVq9KlSoxY8aMQrm///772HTTTWPZsmU5Sla5mOgcAAAAyIkkSSKTyRRa/s0330SDBg1ykGj1Jk2alP3/xx9/HDNmzMh+vWLFihg3blxsuummuYhWKSmlAAAAgFR17NgxMplMZDKZ2H///aNatf9XT6xYsSKmTJkSvXr1ymHCou20007Z3Pvtt1+h9bVq1YobbrghB8kqJ6UUAAAAkKq+fftGRMR7770XPXv2jLp162bX1ahRI7bYYovo169fjtIVb8qUKZEkSbRp0ybefPPNaNasWXZdjRo1onnz5lG1atUcJqxczCkFAAAA5MSdd94ZRx11VNSsWTPXUciBKrkOAAAAAGyYBgwYkC2kTj/99Pjhhx9ynKj06tevH//73/9yHaNSUkoBAAAAOXfPPffEvHnzch2j1ByAVnZKKQAAACDnlDsbHqUUAAAAQBn97ne/i/r16+c6RqWklAIAAABybv78+dGmTZtcx1it5557LrbZZpsChxkOHz48mjZtGnPnzo1tt902XnrppRwmrFyUUgAAAECqqlSpElWrVl3tpVq1armOWciwYcPipJNOKnJkVIMGDeKUU06Ja6+9NgfJKqdM4qBNAAAAIEWPPvposetee+21uP766yM/Pz8WL16cYqo1a9WqVYwbNy46dOhQ5PpPP/00evToEVOnTk05WeVU8WpHAAAAYL126KGHFlr22WefxYUXXhiPP/54/Pa3v43LLrssB8lWb+bMmVG9evVi11erVi1mzZqVYqLKzeF7AAAAQM589913cdJJJ8X2228fy5cvj/feey/uvPPOaNWqVa6jFbLpppvGhx9+WOz6SZMmxcYbb5xiospNKQUAAACkbu7cuXHBBRfElltuGR999FE8++yz8fjjj8d2222X62jF6tOnT/z1r38t8rDCRYsWxSWXXBIHHXRQDpJVTuaUAgAAAFJ19dVXx1VXXRUtWrSIK6+8ssjD+SqimTNnxs477xxVq1aNM844I7beeuuI+GUuqZtuuilWrFgR7777bmy00UY5Tlo5KKUAAACAVFWpUiVq1aoV3bt3j6pVqxa73dixY1NMVTJff/11nHbaaTF+/PhYWalkMpno2bNn3HTTTdG6descJ6w8lFIAAABAqgYOHBiZTGaN240cOTKFNGXz008/xeTJkyNJkmjXrl00atQo15EqHaUUAAAAAKkz0TkAAAAAqVNKAQAAAJA6pRQAAAAAqVNKAQAAAJA6pRQAQDn56quvIpPJxHvvvVfsNi+88EJkMpmYM2dOarkAACoipRQAsMFYefrpTCYT1atXj9atW8f5558fixcvLpfbb9myZUyfPj222267crm9slr5PRZ3ufTSS3OaDwAgIqJargMAAKSpV69eMXLkyFi2bFm88847MWDAgMhkMnHVVVet9W1XrVo1WrRoUQ4p18706dOz/7///vvj4osvjs8++yy7rG7durmIBQBQgJFSAMAGJS8vL1q0aBEtW7aMvn37Rvfu3ePpp5/Ors/Pz4+hQ4dG69ato1atWrHjjjvGgw8+mF3/008/xW9/+9to1qxZ1KpVK9q1axcjR46MiKIP3/vvf/8bW221VdSqVSu6desWX331VYE8l156aey0004Flg0bNiy22GKLAstuu+226NChQ9SsWTPat28fN998c7HfY4sWLbKXBg0aRCaTiRYtWkS9evViq622inHjxhXY/pFHHok6derE/Pnzs9/DfffdF3vuuWfUrFkztttuu5gwYUKB63z44YfRu3fvqFu3bmy00UZx7LHHxg8//FBsJgCAVSmlAIAN1ocffhivvvpq1KhRI7ts6NChcdddd8WIESPio48+irPPPjt+97vfZUuZv/71r/Hxxx/Hk08+GZ988kkMHz48mjZtWuTtT5s2LQ4//PA4+OCD47333osTTzwxLrzwwlLnHD16dFx88cVxxRVXxCeffBJXXnll/PWvf40777yzVLdTp06dOProo7Ml2kojR46MI444IurVq5dddt5558Wf/vSnmDhxYnTu3DkOPvjg+PHHHyMiYs6cObHffvtFx44d4+23345x48bFzJkz48gjjyz19wYAbLgcvgcAbFCeeOKJqFu3bixfvjyWLFkSVapUiRtvvPH/a+/+Qpp64ziOf9yGXkhUF1JaViupVthaSJEXWkSjLoJQuggCsS0q5kWChkZ/iEDHyArrZFY2JYluoj+kFQndGISREaKWeSK9yCioG2dkir+LaHTSH+guFrT3Cwbbw/M853vuxofveY4k6fv376qqqlJbW5s2btwoSVq6dKna29tVX1+v/Px8DQ4OyuPxKCcnR5ImdTT9rq6uTsuWLVNNTY0kacWKFerq6prxo4InTpxQTU2NCgoKJElOp1M9PT2qr69XUVHRjPby+/3Kzc3V0NCQ0tPT9enTJ7W2tqqtrc0yr6SkRIWFhdH7ePjwoRoaGnT48GFduHBBHo9HVVVV0fnXrl1TZmam+vr6tHz58hnVBAAAEhOhFAAASCibN29WXV2dIpGIzp49K4fDEQ1f+vv7NTIyoq1bt1rWjI6OyuPxSJIOHjyowsJCdXZ2yuv1aufOncrNzZ3yWr29vdqwYYNl7FfYNV2RSESmacrn82nfvn3R8bGxMc2ePXtGe0nS+vXrtXr1ajU1NamiokLNzc1avHix8vLy/rdOh8OhnJwc9fb2SpJevXqlJ0+eTHk2lWmahFIAAGBaCKUAAEBCSU1NVVZWlqSf3T1ut1sNDQ3y+XwaHh6WJLW0tGjBggWWdSkpKZKk7du3a2BgQK2trXr8+LG2bNmiQCCg06dPx1SPzWbTxMSEZezHjx/R779qunLlyqSAy263x3RNv98vwzBUUVGhcDis4uJiJSUlTXv98PCwduzYMWXHV3p6ekw1AQCAxMOZUgAAIGHZbDYdOXJER48e1bdv37Rq1SqlpKRocHBQWVlZlk9mZmZ0XVpamoqKitTc3Kxz587p8uXLU+7vcrnU0dFhGXv27Jnld1pamj5+/GgJpn4/KH3evHnKyMjQu3fvJtXkdDpjuu89e/ZoYGBAtbW16unpmfIRwN/rHBsb04sXL+RyuSRJ69atU3d3t5YsWTKpptTU1JhqAgAAiYdQCgAAJLRdu3bJbrfLMAzNmjVLZWVlKi0tVVNTk0zTVGdnp86fPx89VPz48eO6e/eu+vv71d3drfv370fDmj8dOHBAb9++VXl5ud68eaMbN26osbHRMmfTpk36/PmzQqGQTNOUYRh68OCBZc7JkydVXV2t2tpa9fX1qaurS+FwWGfOnInpnufOnauCggKVl5fL6/Vq4cKFk+YYhqHbt2/r9evXCgQC+vr1q/bu3StJCgQC+vLli3bv3q3nz5/LNE09evRIxcXFGh8fj6kmAACQeAilAABAQnM4HCopKVEoFFIkEtGpU6d07NgxVVdXy+Vyadu2bWppaYl2JSUnJ6uyslJr1qxRXl6e7Ha7bt68OeXeixYt0q1bt3Tnzh253W5dunTJcji49LOb6uLFizIMQ263Wx0dHSorK7PM8fv9unr1qsLhsLKzs5Wfn6/GxsaYO6UkyefzaXR0NBo0/SkYDCoYDMrtdqu9vV337t2LvmUwIyNDT58+1fj4uLxer7Kzs3Xo0CHNmTNHNht/LwEAwPQkTfx5iAEAAAD+edevX1dpaak+fPig5OTk6Pj79+/ldDr18uVLrV279u8VCAAA/nkcdA4AAJBARkZGNDQ0pGAwqP3791sCKQAAgHiivxoAACCBhEIhrVy5UvPnz1dlZeXfLgcAACQwHt8DAAAAAABA3NEpBQAAAAAAgLgjlAIAAAAAAEDcEUoBAAAAAAAg7gilAAAAAAAAEHeEUgAAAAAAAIg7QikAAAAAAADEHaEUAAAAAAAA4o5QCgAAAAAAAHFHKAUAAAAAAIC4+w9P10kPpp+4ZwAAAABJRU5ErkJggg==\n"
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of Unique Accession Numbers for Significant Data in AdjacentSeq_all__(Mar_Active)___(Torpor).tsv: 1062\n",
            "Number of Unique Accession Numbers for Not Significant Data in AdjacentSeq_all__(Mar_Active)___(Torpor).tsv: 2165\n",
            "Total number of modifications in AdjacentSeq_all__(Mar_Active)___(Torpor).tsv: 17793\n",
            "Significant N-terminal (Met-loss) modifications: 13\n",
            "Non-significant N-terminal (Met-loss) modifications: 30\n",
            "Processing complete.\n"
          ]
        }
      ],
      "source": [
        "#@title Plot of amino acid type AFTER modification (significant vs not significant modifications)\n",
        "\n",
        "import os\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import numpy as np\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "def calculate_aa_percentages(fasta_dir):\n",
        "    aa_counts = {'A': 0, 'C': 0, 'D': 0, 'E': 0, 'F': 0, 'G': 0, 'H': 0, 'I': 0, 'K': 0, 'L': 0, 'M': 0, 'N': 0, 'P': 0, 'Q': 0, 'R': 0, 'S': 0, 'T': 0, 'V': 0, 'W': 0, 'Y': 0}\n",
        "    n_term_counts = {key: 0 for key in aa_counts}\n",
        "    c_term_counts = {key: 0 for key in aa_counts}\n",
        "    total_residues = 0\n",
        "\n",
        "    for filename in os.listdir(fasta_dir):\n",
        "        if filename.endswith('.fasta'):\n",
        "            with open(os.path.join(fasta_dir, filename), 'r') as f:\n",
        "                _ = f.readline()  # Skip header\n",
        "                sequence = f.readline().strip()\n",
        "                if sequence:\n",
        "                    total_residues += len(sequence)\n",
        "                    for aa in sequence:\n",
        "                        if aa in aa_counts:\n",
        "                            aa_counts[aa] += 1\n",
        "                    if sequence[0] in n_term_counts:\n",
        "                        n_term_counts[sequence[0]] += 1\n",
        "                    if sequence[-1] in c_term_counts:\n",
        "                        c_term_counts[sequence[-1]] += 1\n",
        "\n",
        "    aa_percentages = {key: (count / total_residues) * 100 for key, count in aa_counts.items()}\n",
        "    n_term_percentages = {key: (count / total_residues) * 100 for key, count in n_term_counts.items()}\n",
        "    c_term_percentages = {key: (count / total_residues) * 100 for key, count in c_term_counts.items()}\n",
        "\n",
        "    # Print statements for total residues, A residues, and percentage of A residues\n",
        "    print(f\"Total number of residues: {total_residues}\")\n",
        "    print(f\"Number of A residues: {aa_counts['A']}\")\n",
        "    print(f\"Percentage of A residues: {aa_percentages['A']}%\")\n",
        "\n",
        "    return aa_percentages, n_term_percentages, c_term_percentages\n",
        "\n",
        "def plot_last_column(file_path, log_file, aa_percentages, n_term_percentages, c_term_percentages):\n",
        "    significant_characters_count = {key: 0 for key in aa_percentages}\n",
        "    non_significant_characters_count = {key: 0 for key in aa_percentages}\n",
        "\n",
        "    # Include 'N-term' and 'C-term' in the dictionaries\n",
        "    significant_characters_count['N-term'] = 0\n",
        "    significant_characters_count['C-term'] = 0\n",
        "    non_significant_characters_count['N-term'] = 0\n",
        "    non_significant_characters_count['C-term'] = 0\n",
        "\n",
        "    unique_accession_numbers_significant = set()\n",
        "    unique_accession_numbers_non_significant = set()\n",
        "\n",
        "    total_modifications = 0\n",
        "    significant_modifications = 0\n",
        "    non_significant_modifications = 0\n",
        "\n",
        "    significant_met_loss_count = 0\n",
        "    non_significant_met_loss_count = 0\n",
        "\n",
        "    with open(file_path, 'r') as file:\n",
        "        # Skip header\n",
        "        header = next(file)\n",
        "        write_log(log_file, f\"Processing file: {file_path}\")\n",
        "\n",
        "        for line in file:\n",
        "            line = line.strip()\n",
        "            if line:\n",
        "                columns = line.split('\\t')\n",
        "                if len(columns) != 6:  # Ensure the line has exactly 6 columns\n",
        "                    write_log(log_file, f\"Skipping malformed line: {line}\")\n",
        "                    continue\n",
        "\n",
        "                accession_number = columns[0]\n",
        "                p_value = float(columns[3])\n",
        "                last_column = columns[4]  # Ensure the modifications are in the correct column\n",
        "                sequence_column = columns[5]  # Ensure the sequence is in the correct column\n",
        "\n",
        "                total_modifications += 1\n",
        "\n",
        "                is_n_term = last_column.startswith('Met-loss')\n",
        "                is_met_loss_acetyl = 'Met-loss+Acetyl' in last_column\n",
        "                character = 'C-term' if len(sequence_column) < 9 else sequence_column[8]\n",
        "\n",
        "                if character not in significant_characters_count:\n",
        "                    significant_characters_count[character] = 0\n",
        "                if character not in non_significant_characters_count:\n",
        "                    non_significant_characters_count[character] = 0\n",
        "\n",
        "                if p_value < 0.05:\n",
        "                    significant_modifications += 1\n",
        "                    unique_accession_numbers_significant.add(accession_number)\n",
        "                    significant_characters_count[character] += 1\n",
        "                    if is_n_term:\n",
        "                        significant_met_loss_count += 1\n",
        "                else:\n",
        "                    non_significant_modifications += 1\n",
        "                    unique_accession_numbers_non_significant.add(accession_number)\n",
        "                    non_significant_characters_count[character] += 1\n",
        "                    if is_n_term:\n",
        "                        non_significant_met_loss_count += 1\n",
        "\n",
        "    # Calculate percentages for significant and non-significant modifications\n",
        "    significant_characters_percent = {key: (value / significant_modifications) * 100 if significant_modifications > 0 else 0 for key, value in significant_characters_count.items()}\n",
        "    non_significant_characters_percent = {key: (value / non_significant_modifications) * 100 if non_significant_modifications > 0 else 0 for key, value in non_significant_characters_count.items()}\n",
        "\n",
        "    # Plot data\n",
        "    categories = list(significant_characters_percent.keys())\n",
        "    bar_width = 0.25\n",
        "    x = np.arange(len(categories))\n",
        "    fig, ax1 = plt.subplots(figsize=(12, 8))\n",
        "\n",
        "    ax1.bar(x - bar_width, [aa_percentages.get(cat, 0) for cat in categories], bar_width, color='black', label='All Data')\n",
        "    ax1.bar(x, [significant_characters_percent.get(cat, 0) for cat in categories], bar_width, color='red', label='Significant (p < 0.05)')\n",
        "    ax1.bar(x + bar_width, [non_significant_characters_percent.get(cat, 0) for cat in categories], bar_width, color='skyblue', label='Not Significant (p >= 0.05)')\n",
        "\n",
        "    ax1.set_ylabel('Percentage')\n",
        "    ax1.set_xticks(x)\n",
        "    ax1.set_xticklabels(categories)\n",
        "    ax1.set_xlabel('Residue Type')\n",
        "    ax1.set_title(f'Residue after modification - {os.path.basename(file_path)}')\n",
        "    ax1.set_ylim(0, 35)\n",
        "\n",
        "    plt.xticks(rotation=90)\n",
        "    plt.tight_layout()\n",
        "\n",
        "    # Save plot\n",
        "    output_dir = 'output/plots/adjacent_aa'\n",
        "    if not os.path.exists(output_dir):\n",
        "        os.makedirs(output_dir)\n",
        "    output_filename = os.path.basename(file_path).replace('.tsv', '')\n",
        "    plt.savefig(os.path.join(output_dir, f'AA_aftermod_{output_filename}.eps'))      # Saves as eps format image\n",
        "\n",
        "    plt.show()\n",
        "\n",
        "    # Log number of unique accession numbers and modifications\n",
        "    write_log(log_file, f\"Number of Unique Accession Numbers for Significant Data in {os.path.basename(file_path)}: {len(unique_accession_numbers_significant)}\")\n",
        "    write_log(log_file, f\"Number of Unique Accession Numbers for Not Significant Data in {os.path.basename(file_path)}: {len(unique_accession_numbers_non_significant)}\")\n",
        "    write_log(log_file, f\"Total number of modifications in {os.path.basename(file_path)}: {total_modifications}\")\n",
        "\n",
        "    # Print statements for Met-loss modifications\n",
        "    print(f\"Significant N-terminal (Met-loss) modifications: {significant_met_loss_count}\")\n",
        "    print(f\"Non-significant N-terminal (Met-loss) modifications: {non_significant_met_loss_count}\")\n",
        "\n",
        "# Directory containing the files\n",
        "directory = '/content/output/modlist_files'\n",
        "fasta_dir = '/content/output/libraries/all_protein_seq'\n",
        "logs_directory = os.path.join('output/logs')\n",
        "os.makedirs(logs_directory, exist_ok=True)\n",
        "log_file = os.path.join(logs_directory, \"AA_after_mod.log\")\n",
        "\n",
        "# Calculate amino acid percentages for all protein residues\n",
        "aa_percentages, n_term_percentages, c_term_percentages = calculate_aa_percentages(fasta_dir)\n",
        "\n",
        "# Process each file\n",
        "for filename in os.listdir(directory):\n",
        "    if filename.startswith('AdjacentSeq') and filename.endswith('.tsv'):\n",
        "        file_path = os.path.join(directory, filename)\n",
        "        plot_last_column(file_path, log_file, aa_percentages, n_term_percentages, c_term_percentages)\n",
        "\n",
        "write_log(log_file, \"Processing complete.\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "cellView": "form",
        "id": "BQ3St8VvzV3H",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "49fce254-11b6-4a6e-ff09-dfb8370f509a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Processing file: all_mods_ss_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "Plot saved to /content/output/plots/SStype/SStype_percent_all_mods_ss_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).eps\n",
            "Counts saved to /content/output/text_files/SStype__Abundance_Ratio_(log2)-(Aug_Active)-_(Mar_Active).txt\n",
            "File: all_mods_ss_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "Number of unique accession numbers: 15\n",
            "Number of significant modifications: 94\n",
            "Number of all modifications: 16026\n",
            "Number of not significant modifications: 15932\n",
            "Number of all residues: 2349857.0\n",
            "Processing file: all_mods_ss_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "Plot saved to /content/output/plots/SStype/SStype_percent_all_mods_ss_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).eps\n",
            "Counts saved to /content/output/text_files/SStype__Abundance_Ratio_(log2)-(Torpor)-_(Aug_Active).txt\n",
            "File: all_mods_ss_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "Number of unique accession numbers: 1286\n",
            "Number of significant modifications: 7716\n",
            "Number of all modifications: 16026\n",
            "Number of not significant modifications: 8310\n",
            "Number of all residues: 2349857.0\n",
            "Processing file: all_mods_ss_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "Plot saved to /content/output/plots/SStype/SStype_percent_all_mods_ss_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).eps\n",
            "Counts saved to /content/output/text_files/SStype__Abundance_Ratio_(log2)-(Mar_Active)-_(Torpor).txt\n",
            "File: all_mods_ss_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "Number of unique accession numbers: 1027\n",
            "Number of significant modifications: 6075\n",
            "Number of all modifications: 16026\n",
            "Number of not significant modifications: 9951\n",
            "Number of all residues: 2349857.0\n"
          ]
        }
      ],
      "source": [
        "#@title Plot of secondary structure element for modifications (all, sig, and not sig)\n",
        "import os\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import numpy as np\n",
        "from matplotlib.ticker import PercentFormatter\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "def unify_indices(*dataframes):\n",
        "    unified_index = dataframes[0].index\n",
        "    for df in dataframes[1:]:\n",
        "        unified_index = unified_index.union(df.index)\n",
        "    return [df.reindex(unified_index, fill_value=0) for df in dataframes]\n",
        "\n",
        "def calculate_all_residues_structure(directory):\n",
        "    \"\"\"Calculate secondary structure counts for all residues in all proteins in the dataset.\"\"\"\n",
        "    all_residue_counts = pd.Series(dtype=int)\n",
        "    for filename in os.listdir(directory):\n",
        "        if filename.endswith(\"_ss.tsv\"):\n",
        "            df = pd.read_csv(os.path.join(directory, filename), delimiter='\\t')\n",
        "            df.columns = df.columns.str.strip()\n",
        "            if 'Secondary Structure' not in df.columns:\n",
        "                continue\n",
        "            structure_counts = df['Secondary Structure'].value_counts()\n",
        "            all_residue_counts = all_residue_counts.add(structure_counts, fill_value=0)\n",
        "    return all_residue_counts\n",
        "\n",
        "def plot_secondary_structures(all_mods_directory, all_residues_directory, output_figures_directory, text_files_directory, log_file):\n",
        "    os.makedirs(output_figures_directory, exist_ok=True)\n",
        "    os.makedirs(text_files_directory, exist_ok=True)\n",
        "\n",
        "    # Dictionary for full names of DSSP codes\n",
        "    structure_names = {\n",
        "        'H': 'Alpha-helix',\n",
        "        'B': 'Isolated beta-bridge',\n",
        "        'E': 'Beta-strand',\n",
        "        'G': '3-10 helix',\n",
        "        'I': 'Pi helix',\n",
        "        'T': 'Turn',\n",
        "        'S': 'Bend',\n",
        "        '-': 'None'\n",
        "    }\n",
        "\n",
        "    # Order to plot secondary structures\n",
        "    plot_order = ['None', 'Bend', 'Turn', 'Alpha-helix', '3-10 helix', 'Pi helix', 'Beta-strand', 'Isolated beta-bridge']\n",
        "\n",
        "    # Calculate all residues structure counts\n",
        "    all_residue_counts = calculate_all_residues_structure(all_residues_directory)\n",
        "    all_residue_counts.index = all_residue_counts.index.map(structure_names)\n",
        "\n",
        "    for filename in os.listdir(all_mods_directory):\n",
        "        if \"failed\" not in filename and filename.startswith(\"all_mods_ss_\") and filename.endswith(\".tsv\"):\n",
        "            write_log(log_file, f\"Processing file: {filename}\")\n",
        "            all_mods_file_path = os.path.join(all_mods_directory, filename)\n",
        "\n",
        "            all_mods_df = pd.read_csv(all_mods_file_path, delimiter='\\t')\n",
        "            all_mods_df.columns = all_mods_df.columns.str.strip()\n",
        "\n",
        "            if 'Secondary Structure' not in all_mods_df.columns or 'Adj. P-Value' not in all_mods_df.columns:\n",
        "                write_log(log_file, \"Required column missing in the dataframe.\")\n",
        "                continue\n",
        "\n",
        "            all_mods_df['Secondary Structure'] = all_mods_df['Secondary Structure'].str.strip().map(structure_names)\n",
        "\n",
        "            sig_df = all_mods_df[all_mods_df['Adj. P-Value'] < 0.05]\n",
        "            notsig_df = all_mods_df[all_mods_df['Adj. P-Value'] >= 0.05]\n",
        "\n",
        "            sig_structure_counts = sig_df['Secondary Structure'].value_counts()\n",
        "            all_mods_structure_counts = all_mods_df['Secondary Structure'].value_counts()\n",
        "            notsig_mods_structure_counts = notsig_df['Secondary Structure'].value_counts()\n",
        "\n",
        "            # Calculate percentages for plotting\n",
        "            sig_total = sig_structure_counts.sum()\n",
        "            all_total = all_mods_structure_counts.sum()\n",
        "            notsig_total = notsig_mods_structure_counts.sum()\n",
        "            all_residues_total = all_residue_counts.sum()\n",
        "\n",
        "            sig_structure_percentages = (sig_structure_counts / sig_total * 100)\n",
        "            all_mods_structure_percentages = (all_mods_structure_counts / all_total * 100)\n",
        "            notsig_mods_structure_percentages = (notsig_mods_structure_counts / notsig_total * 100)\n",
        "            all_residues_percentages = (all_residue_counts / all_residues_total * 100)\n",
        "\n",
        "            sig_structure_percentages, all_mods_structure_percentages, notsig_mods_structure_percentages, all_residues_percentages = unify_indices(\n",
        "                sig_structure_percentages, all_mods_structure_percentages, notsig_mods_structure_percentages, all_residues_percentages)\n",
        "\n",
        "            # Reorder percentages according to plot_order\n",
        "            sig_structure_percentages = sig_structure_percentages.reindex(plot_order, fill_value=0)\n",
        "            all_mods_structure_percentages = all_mods_structure_percentages.reindex(plot_order, fill_value=0)\n",
        "            notsig_mods_structure_percentages = notsig_mods_structure_percentages.reindex(plot_order, fill_value=0)\n",
        "            all_residues_percentages = all_residues_percentages.reindex(plot_order, fill_value=0)\n",
        "\n",
        "            plt.rcParams['font.family'] = 'Liberation Sans'  # Set the font to Liberation Sans\n",
        "\n",
        "            fig, ax = plt.subplots(figsize=(8, 8))  # Make the plot square\n",
        "            index = np.arange(len(sig_structure_percentages))\n",
        "            bar_width = 0.2\n",
        "\n",
        "            ax.bar(index, all_residues_percentages, bar_width, label='All Residues', color='black')\n",
        "            ax.bar(index + bar_width, sig_structure_percentages, bar_width, label='Significant', color='red')\n",
        "\n",
        "            ax.set_xlabel('Secondary Structure')\n",
        "            ax.set_ylabel('Percentage')\n",
        "            ax.set_ylim(0, 100)\n",
        "            ax.yaxis.set_major_formatter(PercentFormatter())\n",
        "            ax.set_title(f'Percentage of Each Secondary Structure Type in {filename}')\n",
        "            ax.set_xticks(index)\n",
        "            ax.set_xticklabels(plot_order)\n",
        "\n",
        "            # Move legend to top right corner and remove the frame\n",
        "            legend = ax.legend(loc='upper right', frameon=False)\n",
        "\n",
        "            # Display the number of unique accession numbers\n",
        "            sig_unique_accessions = len(sig_df['Accession Number'].unique())\n",
        "            all_unique_accessions = len(all_mods_df['Accession Number'].unique())\n",
        "            notsig_unique_accessions = len(notsig_df['Accession Number'].unique())\n",
        "            ax.text(0.885, 0.70, f'Significant: {sig_unique_accessions}', transform=ax.transAxes, verticalalignment='top')\n",
        "            ax.text(0.885, 0.65, f'All Mods: {all_unique_accessions}', transform=ax.transAxes, verticalalignment='top')\n",
        "            ax.text(0.885, 0.60, f'Not Significant: {notsig_unique_accessions}', transform=ax.transAxes, verticalalignment='top')\n",
        "\n",
        "            plt.xticks(rotation=45)\n",
        "            plt.tight_layout()\n",
        "\n",
        "            plot_filename = os.path.join(output_figures_directory, f'SStype_percent_{filename[:-4]}.eps')\n",
        "            plt.savefig(plot_filename)\n",
        "            plt.close()\n",
        "            write_log(log_file, f\"Plot saved to {plot_filename}\")\n",
        "\n",
        "            # Save counts to text file\n",
        "            text_filename = os.path.join(text_files_directory, f'SStype_{filename[11:-4].replace(\"__\", \"-\")}.txt')\n",
        "            with open(text_filename, 'w') as text_file:\n",
        "                text_file.write(f\"File: {filename}\\n\")\n",
        "                text_file.write(f\"Total modifications (Significant): {sig_total}\\n\")\n",
        "                text_file.write(f\"Total modifications (All Mods): {all_total}\\n\")\n",
        "                text_file.write(f\"Total modifications (Not Significant): {notsig_total}\\n\")\n",
        "                text_file.write(f\"Total residues (All Residues): {all_residues_total}\\n\")\n",
        "                text_file.write(\"\\nSecondary Structure Counts (Significant):\\n\")\n",
        "                for structure, count in sig_structure_counts.reindex(plot_order, fill_value=0).items():\n",
        "                    text_file.write(f\"{structure}: {count}\\n\")\n",
        "                text_file.write(\"\\nSecondary Structure Counts (All Mods):\\n\")\n",
        "                for structure, count in all_mods_structure_counts.reindex(plot_order, fill_value=0).items():\n",
        "                    text_file.write(f\"{structure}: {count}\\n\")\n",
        "                text_file.write(\"\\nSecondary Structure Counts (Not Significant):\\n\")\n",
        "                for structure, count in notsig_mods_structure_counts.reindex(plot_order, fill_value=0).items():\n",
        "                    text_file.write(f\"{structure}: {count}\\n\")\n",
        "                text_file.write(\"\\nSecondary Structure Counts (All Residues):\\n\")\n",
        "                for structure, count in all_residue_counts.reindex(plot_order, fill_value=0).items():\n",
        "                    text_file.write(f\"{structure}: {count}\\n\")\n",
        "\n",
        "            write_log(log_file, f\"Counts saved to {text_filename}\")\n",
        "\n",
        "            # Print statements for the unique accession numbers and modifications\n",
        "            print(f\"File: {filename}\")\n",
        "            print(f\"Number of unique accession numbers: {len(sig_df['Accession Number'].unique())}\")\n",
        "            print(f\"Number of significant modifications: {sig_total}\")\n",
        "            print(f\"Number of all modifications: {all_total}\")\n",
        "            print(f\"Number of not significant modifications: {notsig_total}\")\n",
        "            print(f\"Number of all residues: {all_residues_total}\")\n",
        "\n",
        "def main():\n",
        "    all_mods_directory = \"/content/output/modlist_files\"  # Directory containing all mods files\n",
        "    all_residues_directory = \"/content/output/libraries/all_protein_ss\"  # Directory containing all residues files\n",
        "    output_figures_directory = \"/content/output/plots/SStype\"  # Directory to save plots\n",
        "    text_files_directory = \"/content/output/text_files\"  # Directory to save text files\n",
        "    logs_directory = \"/content/output/logs\"\n",
        "    os.makedirs(logs_directory, exist_ok=True)\n",
        "    log_file = os.path.join(logs_directory, \"ss-type_process.log\")\n",
        "\n",
        "    plot_secondary_structures(all_mods_directory, all_residues_directory, output_figures_directory, text_files_directory, log_file)\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    main()\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 57,
      "metadata": {
        "id": "a1LxNynD04G8",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "36b0db3b-2822-4a91-e0e7-d3a6249a7a26"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Processing file: domain_sig_modlist_sig_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "Processed domain_sig_modlist_sig_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv: Sig Domain: 525, Sig No Domain: 6418\n",
            "Processing file: domain_sig_modlist_sig_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "Processed domain_sig_modlist_sig_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv: Sig Domain: 19, Sig No Domain: 75\n",
            "Processing file: domain_sig_modlist_sig_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "Processed domain_sig_modlist_sig_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv: Sig Domain: 677, Sig No Domain: 8296\n",
            "                                             Dataset Domain No Domain  \\\n",
            "0                                       All residues    26%       74%   \n",
            "1  domain_sig_modlist_sig_modlist_Abundance_Ratio...     8%       92%   \n",
            "2  domain_sig_modlist_sig_modlist_Abundance_Ratio...    20%       80%   \n",
            "3  domain_sig_modlist_sig_modlist_Abundance_Ratio...     8%       92%   \n",
            "4                                    All significant     8%       92%   \n",
            "\n",
            "   # residues  # proteins  \n",
            "0     2697361        3311  \n",
            "1        6943        1076  \n",
            "2          94          15  \n",
            "3        8973        1354  \n",
            "4       16010        1430  \n",
            "Shortest protein length: 35\n",
            "Longest protein length: 7747\n",
            "Average protein length: 814.7\n",
            "Standard deviation of protein lengths: 665.4\n",
            "Results saved to /content/output/text_files/domain_presence_summary.csv\n"
          ]
        }
      ],
      "source": [
        "#@title Identifies whether residues are in a domain vs \"no domain\" for all protein residues and modified residues for significant datasets\n",
        "\n",
        "import os\n",
        "import re\n",
        "import pandas as pd\n",
        "from Bio import SeqIO\n",
        "import numpy as np\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "def count_domain_presence(df):\n",
        "    \"\"\"Count the number of modification sites in 'Domain' and 'No Domain' categories.\"\"\"\n",
        "    domain_count = df[df['Domain Type'] != 'No Domain'].shape[0]\n",
        "    no_domain_count = df[df['Domain Type'] == 'No Domain'].shape[0]\n",
        "    return domain_count, no_domain_count\n",
        "\n",
        "def calculate_percentage(domain_count, total_count):\n",
        "    \"\"\"Calculate the percentage of domain presence.\"\"\"\n",
        "    if total_count == 0:\n",
        "        return 0\n",
        "    return (domain_count / total_count) * 100\n",
        "\n",
        "def extract_filename_info(filename):\n",
        "    \"\"\"Extract the relevant part of the filename (e.g., (Torpor) vs (Aug Active))\"\"\"\n",
        "    match = re.search(r'domain_sig_modlist_all_modlist_Abundance_Ratio_\\(log2\\)__\\(([^)]+)\\)___\\(([^)]+)\\).tsv', filename)\n",
        "    if match:\n",
        "        return f\"{match.group(1)} vs {match.group(2)}\"\n",
        "    else:\n",
        "        return filename\n",
        "\n",
        "def read_fasta_sequences(fasta_directory):\n",
        "    \"\"\"Read all FASTA files in the directory and return a dictionary of sequences.\"\"\"\n",
        "    sequences = {}\n",
        "    for fasta_file in os.listdir(fasta_directory):\n",
        "        if fasta_file.endswith(\".fasta\"):\n",
        "            accession_number = os.path.splitext(fasta_file)[0]\n",
        "            fasta_path = os.path.join(fasta_directory, fasta_file)\n",
        "            for record in SeqIO.parse(fasta_path, \"fasta\"):\n",
        "                sequences[accession_number] = str(record.seq)\n",
        "    return sequences\n",
        "\n",
        "def process_files(all_directory, fasta_directory, domain_library_path, log_file):\n",
        "    results = []\n",
        "    combined_sig_df = pd.DataFrame()  # DataFrame to hold all significant data combined\n",
        "\n",
        "    # Check if the directory exists\n",
        "    if not os.path.exists(all_directory):\n",
        "        write_log(log_file, f\"Directory does not exist: {all_directory}\")\n",
        "        return\n",
        "\n",
        "    # Read domain library\n",
        "    domain_library = pd.read_csv(domain_library_path, delimiter='\\t')\n",
        "\n",
        "    # Read FASTA sequences\n",
        "    sequences = read_fasta_sequences(fasta_directory)\n",
        "\n",
        "    all_residues_domain_count = 0\n",
        "    all_residues_no_domain_count = 0\n",
        "    all_residues_total_count = 0\n",
        "\n",
        "    protein_lengths = []\n",
        "\n",
        "    for accession, sequence in sequences.items():\n",
        "        protein_length = len(sequence)\n",
        "        all_residues_total_count += protein_length\n",
        "        protein_lengths.append(protein_length)\n",
        "\n",
        "        domains = domain_library[domain_library['Accession'] == accession]\n",
        "        domain_ranges = [(row['Domain Start'], row['Domain End']) for index, row in domains.iterrows()]\n",
        "\n",
        "        for i in range(1, protein_length + 1):\n",
        "            in_domain = False\n",
        "            for start, end in domain_ranges:\n",
        "                if start <= i <= end:\n",
        "                    in_domain = True\n",
        "                    break\n",
        "            if in_domain:\n",
        "                all_residues_domain_count += 1\n",
        "            else:\n",
        "                all_residues_no_domain_count += 1\n",
        "\n",
        "    all_residues_domain_percentage = calculate_percentage(all_residues_domain_count, all_residues_total_count)\n",
        "    all_residues_no_domain_percentage = calculate_percentage(all_residues_no_domain_count, all_residues_total_count)\n",
        "\n",
        "    # Add all residues summary row\n",
        "    results.append({\n",
        "        'Dataset': 'All residues',\n",
        "        'Domain': f\"{all_residues_domain_percentage:.0f}%\",\n",
        "        'No Domain': f\"{all_residues_no_domain_percentage:.0f}%\",\n",
        "        '# residues': all_residues_total_count,\n",
        "        '# proteins': len(sequences)\n",
        "    })\n",
        "\n",
        "    for filename in os.listdir(all_directory):\n",
        "        if filename.startswith(\"domain_sig_modlist\") and filename.endswith(\".tsv\") and \"notsig\" not in filename:\n",
        "            all_file_path = os.path.join(all_directory, filename)\n",
        "\n",
        "            write_log(log_file, f\"Processing file: {filename}\")\n",
        "\n",
        "            try:\n",
        "                all_df = pd.read_csv(all_file_path, delimiter='\\t').drop_duplicates()\n",
        "\n",
        "                # Determine significant data\n",
        "                sig_df = all_df[all_df['Adj. P-Value'] < 0.05].copy()\n",
        "\n",
        "                # Ensure 'Domain Type' column is present and filled with 'No Domain' where missing\n",
        "                if 'Domain Type' not in all_df.columns:\n",
        "                    all_df['Domain Type'] = 'No Domain'\n",
        "                else:\n",
        "                    all_df['Domain Type'] = all_df['Domain Type'].fillna('No Domain')\n",
        "\n",
        "                if 'Domain Type' not in sig_df.columns:\n",
        "                    sig_df['Domain Type'] = 'No Domain'\n",
        "                else:\n",
        "                    sig_df['Domain Type'] = sig_df['Domain Type'].fillna('No Domain')\n",
        "\n",
        "                # Combine significant data for the final row\n",
        "                combined_sig_df = pd.concat([combined_sig_df, sig_df]).drop_duplicates()\n",
        "\n",
        "                # Count domain presence for significant data\n",
        "                sig_domain_count, sig_no_domain_count = count_domain_presence(sig_df)\n",
        "                total_sig_count = sig_df[sig_df['Accession Number'].isin(combined_sig_df['Accession Number'])].shape[0]\n",
        "                sig_domain_percentage = calculate_percentage(sig_domain_count, total_sig_count)\n",
        "                sig_no_domain_percentage = calculate_percentage(sig_no_domain_count, total_sig_count)\n",
        "                sig_unique_proteins = sig_df['Accession Number'].nunique()\n",
        "\n",
        "                # Extract relevant filename info\n",
        "                filename_info = extract_filename_info(filename)\n",
        "\n",
        "                # Add the results to the list for significant data\n",
        "                results.append({\n",
        "                    'Dataset': filename_info,\n",
        "                    'Domain': f\"{sig_domain_percentage:.0f}%\",\n",
        "                    'No Domain': f\"{sig_no_domain_percentage:.0f}%\",\n",
        "                    '# residues': total_sig_count,\n",
        "                    '# proteins': sig_unique_proteins\n",
        "                })\n",
        "\n",
        "                write_log(log_file, f\"Processed {filename}: Sig Domain: {sig_domain_count}, Sig No Domain: {sig_no_domain_count}\")\n",
        "\n",
        "            except Exception as e:\n",
        "                write_log(log_file, f\"Error processing {filename}: {e}\")\n",
        "\n",
        "    # Combine significant data across all files\n",
        "    combined_sig_df = combined_sig_df.drop_duplicates()\n",
        "    sig_domain_count, sig_no_domain_count = count_domain_presence(combined_sig_df)\n",
        "    total_sig_count = combined_sig_df[combined_sig_df['Accession Number'].isin(sequences.keys())].shape[0]\n",
        "    sig_domain_percentage = calculate_percentage(sig_domain_count, total_sig_count)\n",
        "    sig_no_domain_percentage = calculate_percentage(sig_no_domain_count, total_sig_count)\n",
        "    sig_unique_proteins = combined_sig_df['Accession Number'].nunique()\n",
        "\n",
        "    # Add the combined significant data row\n",
        "    results.append({\n",
        "        'Dataset': 'All significant',\n",
        "        'Domain': f\"{sig_domain_percentage:.0f}%\",\n",
        "        'No Domain': f\"{sig_no_domain_percentage:.0f}%\",\n",
        "        '# residues': total_sig_count,\n",
        "        '# proteins': sig_unique_proteins\n",
        "    })\n",
        "\n",
        "    # Create a DataFrame from the results\n",
        "    results_df = pd.DataFrame(results)\n",
        "\n",
        "    # Print the table\n",
        "    print(results_df)\n",
        "\n",
        "    # Print protein length statistics\n",
        "    shortest_protein = min(protein_lengths)\n",
        "    longest_protein = max(protein_lengths)\n",
        "    average_protein = np.mean(protein_lengths)\n",
        "    stddev_protein = np.std(protein_lengths)\n",
        "\n",
        "    print(f\"Shortest protein length: {shortest_protein}\")\n",
        "    print(f\"Longest protein length: {longest_protein}\")\n",
        "    print(f\"Average protein length: {average_protein:.1f}\")\n",
        "    print(f\"Standard deviation of protein lengths: {stddev_protein:.1f}\")\n",
        "\n",
        "    # Ensure the output directory exists\n",
        "    output_text_directory = \"/content/output/text_files\"\n",
        "    os.makedirs(output_text_directory, exist_ok=True)\n",
        "\n",
        "    # Save the table to a CSV file\n",
        "    output_file_path = os.path.join(output_text_directory, \"domain_presence_summary.csv\")\n",
        "    results_df.to_csv(output_file_path, index=False)\n",
        "    write_log(log_file, f\"Results saved to {output_file_path}\")\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    all_directory = \"/content/output/modlist_files\"  # Directory containing all files\n",
        "    fasta_directory = \"/content/output/libraries/all_protein_seq\"\n",
        "    domain_library_path = \"/content/output/libraries/domainlibrary.tsv\"\n",
        "    logs_directory = \"/content/output/logs\"\n",
        "    os.makedirs(logs_directory, exist_ok=True)\n",
        "    log_file = os.path.join(logs_directory, \"domain_presence_process.log\")\n",
        "    process_files(all_directory, fasta_directory, domain_library_path, log_file)\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def combine_domain_sig_files(all_directory, output_combined_file):\n",
        "    \"\"\"Combine all domain_sig files into one file with all phosphosite residues.\"\"\"\n",
        "    combined_df = pd.DataFrame()\n",
        "\n",
        "    # List all files in the directory for debugging\n",
        "    print(f\"Files in directory: {os.listdir(all_directory)}\")\n",
        "\n",
        "    for filename in os.listdir(all_directory):\n",
        "        if filename.startswith(\"domain_sig_modlist\") and filename.endswith(\".tsv\") and \"notsig\" not in filename:\n",
        "            file_path = os.path.join(all_directory, filename)\n",
        "            print(f\"Processing file: {filename}\")\n",
        "\n",
        "            try:\n",
        "                df = pd.read_csv(file_path, delimiter='\\t')\n",
        "                print(f\"Number of rows in {filename}: {df.shape[0]}\")\n",
        "                combined_df = pd.concat([combined_df, df], ignore_index=True)\n",
        "            except Exception as e:\n",
        "                print(f\"Error processing {filename}: {e}\")\n",
        "\n",
        "    # Check the total number of rows after combining\n",
        "    print(f\"Total rows in combined DataFrame before saving: {combined_df.shape[0]}\")\n",
        "\n",
        "    # Save the combined DataFrame to a file\n",
        "    combined_df.to_csv(output_combined_file, index=False)\n",
        "\n",
        "    # Confirm the number of rows in the saved file\n",
        "    saved_df = pd.read_csv(output_combined_file)\n",
        "    print(f\"Total rows in combined file: {saved_df.shape[0]}\")\n",
        "\n",
        "    return combined_df\n"
      ],
      "metadata": {
        "id": "yCnoh0dz_Avi"
      },
      "execution_count": 51,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "id": "x4mN5Qpxy-_7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form",
        "outputId": "7ad75c2b-6b54-40d1-929b-3f06e9b46072"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading domain library...\n",
            "Total Kinase domains found: 270\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287CSI2.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287CRX6.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287CWW4.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287CX14.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D132.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D281.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D2C7.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D345.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D3E9.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D3P7.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D3E2.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D610.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D720.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287DAB5.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287DDG1.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287DDK1.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3LWN4.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3LWM4.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3M1P9.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3M3X4.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3M557.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3M8K9.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3M9M0.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MB56.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MB56.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MDX4.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MHI3.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MJI3.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MJQ4.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MJQ4.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3ML28.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MLK8.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MNQ6.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MNZ2.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MQ63.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MPK5.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MPD8.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MQU9.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MRW0.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MS67.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MSI1.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MYX5.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MZR7.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3N0V5.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3N200.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3N1A6.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3NET9.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3NFA8.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3NFI1.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3NFH8.fasta\n",
            "Processed file: sig_mods_ss_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "Total modifications in Kinase domains: 86\n",
            "Total modifications to Kinase proteins, not in Kinase domains: 703\n",
            "Results written to: /content/output/text_files/Kinases/KinaseDomain_modifications_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).txt\n",
            "Results written to: /content/output/text_files/Kinases/Kinase_mod_notinKD_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).txt\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287CSI2.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287CRX6.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287CWW4.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287CX14.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D132.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D281.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D2C7.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D345.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D3E9.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D3P7.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D610.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287D720.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/A0A287DDG1.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3LWN4.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3LWM4.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3LWD8.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3M1P9.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3M3X4.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3M557.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3M7P3.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3M7P3.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3M9M0.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MB56.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MB56.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MDX4.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MHI3.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MJI3.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MJQ4.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MJQ4.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3ML28.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MLK8.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MNQ6.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MNZ2.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MQ63.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MPK5.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MPD8.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MQU9.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MRW0.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MS67.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MY19.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MY19.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3N0V5.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3N200.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3NDL9.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3NFA8.fasta\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3NFI1.fasta\n",
            "Processed file: sig_mods_ss_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "Total modifications in Kinase domains: 80\n",
            "Total modifications to Kinase proteins, not in Kinase domains: 708\n",
            "Results written to: /content/output/text_files/Kinases/KinaseDomain_modifications_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).txt\n",
            "Results written to: /content/output/text_files/Kinases/Kinase_mod_notinKD_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).txt\n",
            "Searching FASTA file: /content//output/libraries/all_protein_seq/I3MS67.fasta\n",
            "Processed file: sig_mods_ss_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "Total modifications in Kinase domains: 6\n",
            "Total modifications to Kinase proteins, not in Kinase domains: 12\n",
            "Results written to: /content/output/text_files/Kinases/KinaseDomain_modifications_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).txt\n",
            "Results written to: /content/output/text_files/Kinases/Kinase_mod_notinKD_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).txt\n",
            "Total T-loop sequences identified: 22\n",
            "Total T-loop sequences identified: 22\n",
            "Accession A0A287CX14 T-loop sequence: DFGVSAQLDRtVGRRNTFIGTPYWMAPE\n",
            "Accession A0A287CX14 T-loop sequence: DFGVSAQLDRtVGRRNTFIGTPYWMAPE\n",
            "Accession A0A287D2C7 T-loop sequence: DFGFSNEFTFGNKLDTFCGsPPYAAPE\n",
            "Accession A0A287D2C7 T-loop sequence: DFGFSNEFTFGNKLDTFCGsPPYAAPE\n",
            "Accession A0A287D610 T-loop sequence: DFGLARTAGTSFMMtPYVVTRYYRAPE\n",
            "Accession A0A287D610 T-loop sequence: DFGLARTAGTSFMMtPYVVTRYYRAPE\n",
            "Accession A0A287D610 T-loop sequence: DFGLARTAGTSFMMTPyVVTRYYRAPE\n",
            "Accession A0A287D610 T-loop sequence: DFGLARTAGTSFMMTPyVVTRYYRAPE\n",
            "Accession I3LWM4 T-loop sequence: DFGFCAQVSKEVPKRKsLVGTPYWMAPE\n",
            "Accession I3LWM4 T-loop sequence: DFGFCAQVSKEVPKRKsLVGTPYWMAPE\n",
            "Accession I3M557 T-loop sequence: DFGFSNEFTVGNKLDTFCGsPPYAAPE\n",
            "Accession I3M557 T-loop sequence: DFGFSNEFTVGNKLDTFCGsPPYAAPE\n",
            "Accession I3MJQ4 T-loop sequence: DFGMCKENVFPGTTTRtFCGTPDYIAPE\n",
            "Accession I3MJQ4 T-loop sequence: DFGMCKENVFPGTTTRtFCGTPDYIAPE\n",
            "Accession I3MJQ4 T-loop sequence: DFGMCKENVFPGTTTRTFCGtPDYIAPE\n",
            "Accession I3MJQ4 T-loop sequence: DFGMCKENVFPGTTTRTFCGtPDYIAPE\n",
            "Accession I3MYX5 T-loop sequence: DFGLAKSFGsPNRAYTHQVVTRWYRAPE\n",
            "Accession I3MYX5 T-loop sequence: DFGLAKSFGsPNRAYTHQVVTRWYRAPE\n",
            "Accession I3MYX5 T-loop sequence: DFGLAKSFGSPNRAYtHQVVTRWYRAPE\n",
            "Accession I3MYX5 T-loop sequence: DFGLAKSFGSPNRAYtHQVVTRWYRAPE\n",
            "Accession I3N0V5 T-loop sequence: DFGLARTACTNFMMtPYVVTRYYRAPE\n",
            "Accession I3N0V5 T-loop sequence: DFGLARTACTNFMMtPYVVTRYYRAPE\n",
            "Accession I3N0V5 T-loop sequence: DFGLARTACTNFMMTPyVVTRYYRAPE\n",
            "Accession I3N0V5 T-loop sequence: DFGLARTACTNFMMTPyVVTRYYRAPE\n",
            "Accession A0A287CX14 T-loop sequence: DFGVSAQLDRtVGRRNTFIGTPYWMAPE\n",
            "Accession A0A287CX14 T-loop sequence: DFGVSAQLDRtVGRRNTFIGTPYWMAPE\n",
            "Accession A0A287D2C7 T-loop sequence: DFGFSNEFTFGNKLDTFCGsPPYAAPE\n",
            "Accession A0A287D2C7 T-loop sequence: DFGFSNEFTFGNKLDTFCGsPPYAAPE\n",
            "Accession A0A287D610 T-loop sequence: DFGLARTAGTSFMMtPYVVTRYYRAPE\n",
            "Accession A0A287D610 T-loop sequence: DFGLARTAGTSFMMtPYVVTRYYRAPE\n",
            "Accession A0A287D610 T-loop sequence: DFGLARTAGTSFMMTPyVVTRYYRAPE\n",
            "Accession A0A287D610 T-loop sequence: DFGLARTAGTSFMMTPyVVTRYYRAPE\n",
            "Accession I3LWM4 T-loop sequence: DFGFCAQVSKEVPKRKsLVGTPYWMAPE\n",
            "Accession I3LWM4 T-loop sequence: DFGFCAQVSKEVPKRKsLVGTPYWMAPE\n",
            "Accession I3M557 T-loop sequence: DFGFSNEFTVGNKLDTFCGsPPYAAPE\n",
            "Accession I3M557 T-loop sequence: DFGFSNEFTVGNKLDTFCGsPPYAAPE\n",
            "Accession I3MJQ4 T-loop sequence: DFGMCKENVFPGTTTRtFCGTPDYIAPE\n",
            "Accession I3MJQ4 T-loop sequence: DFGMCKENVFPGTTTRtFCGTPDYIAPE\n",
            "Accession I3MJQ4 T-loop sequence: DFGMCKENVFPGTTTRTFCGtPDYIAPE\n",
            "Accession I3MJQ4 T-loop sequence: DFGMCKENVFPGTTTRTFCGtPDYIAPE\n",
            "Accession I3N0V5 T-loop sequence: DFGLARTACTNFMMtPYVVTRYYRAPE\n",
            "Accession I3N0V5 T-loop sequence: DFGLARTACTNFMMtPYVVTRYYRAPE\n",
            "Accession I3N0V5 T-loop sequence: DFGLARTACTNFMMTPyVVTRYYRAPE\n",
            "Accession I3N0V5 T-loop sequence: DFGLARTACTNFMMTPyVVTRYYRAPE\n",
            "Process complete.\n"
          ]
        }
      ],
      "source": [
        "#@title Identify Kinases and map PTMs to Kinase domains and predict T-loop\n",
        "import os\n",
        "import shutil\n",
        "import pandas as pd\n",
        "import re\n",
        "\n",
        "# Define directories\n",
        "sig_mods_dir = '/content/output/modlist_files'\n",
        "domain_library_path = '/content/output/libraries/domainlibrary.tsv'\n",
        "disorder_files_dir = '/content//output/libraries/IUPred2A_disorder'\n",
        "output_text_dir = '/content/output/text_files/Kinases'\n",
        "output_plots_dir = '/content/output/plots/Kinases'\n",
        "logs_directory = '/content/output/logs'\n",
        "fasta_seq_dir = '/content//output/libraries/all_protein_seq'\n",
        "log_file = os.path.join(logs_directory, \"kinasedomainPTM_process.log\")\n",
        "\n",
        "# Ensure output directories exist\n",
        "os.makedirs(output_text_dir, exist_ok=True)\n",
        "os.makedirs(output_plots_dir, exist_ok=True)\n",
        "os.makedirs(logs_directory, exist_ok=True)\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "def extract_iupred_score(disorder_files_dir, accession, residue_number):\n",
        "    \"\"\"Extracts the IUPred score for a given residue number from the corresponding disorder file.\"\"\"\n",
        "    disorder_file_path = os.path.join(disorder_files_dir, f'{accession}_iupred2a.txt')\n",
        "    if os.path.exists(disorder_file_path):\n",
        "        with open(disorder_file_path, 'r') as file:\n",
        "            for line in file:\n",
        "                if line.startswith('#') or not line.strip():\n",
        "                    continue\n",
        "                parts = line.split()\n",
        "                if len(parts) == 3:\n",
        "                    try:\n",
        "                        residue = int(parts[0])\n",
        "                        disorder_score = float(parts[2])\n",
        "                        if residue == residue_number:\n",
        "                            return disorder_score\n",
        "                    except ValueError:\n",
        "                        continue\n",
        "    return 'N/A'\n",
        "\n",
        "def extract_t_loop_sequence(fasta_seq_dir, accession):\n",
        "    \"\"\"Extracts the T-loop sequence from the fasta sequence file.\"\"\"\n",
        "    fasta_file_path = os.path.join(fasta_seq_dir, f'{accession}.fasta')\n",
        "    if not os.path.exists(fasta_file_path):\n",
        "        write_log(log_file, f\"FASTA file not found: {fasta_file_path}\")\n",
        "        return None, None, None\n",
        "\n",
        "    write_log(log_file, f\"Searching FASTA file: {fasta_file_path}\")\n",
        "\n",
        "    with open(fasta_file_path, 'r') as file:\n",
        "        sequence = ''.join([line.strip() for line in file if not line.startswith('>')])\n",
        "\n",
        "    dfg_match = re.search(r'DFG', sequence)\n",
        "    ape_match = re.search(r'PE', sequence)\n",
        "\n",
        "    if dfg_match and ape_match and 0 < ape_match.start() - dfg_match.end() <= 50:\n",
        "        t_loop_start = dfg_match.start()\n",
        "        t_loop_end = ape_match.end()\n",
        "        return t_loop_start + 1, t_loop_end + 1, sequence[t_loop_start:t_loop_end]\n",
        "\n",
        "    return None, None, None\n",
        "\n",
        "def extract_modifications_in_kinase_domains(sig_mods_dir, filtered_domains_df, disorder_files_dir, fasta_seq_dir, output_text_dir, output_plots_dir):\n",
        "    t_loop_identified = 0\n",
        "    t_loop_sequences = []\n",
        "\n",
        "    for file in os.listdir(sig_mods_dir):\n",
        "        if file.startswith('sig_mods_ss_') and file.endswith('.tsv'):\n",
        "            sig_mods_path = os.path.join(sig_mods_dir, file)\n",
        "            suffix = file[len('sig_mods_ss_'):-4]\n",
        "            output_kd_file_path = os.path.join(output_text_dir, f'KinaseDomain_modifications_{suffix}.txt')\n",
        "            output_not_in_kd_file_path = os.path.join(output_text_dir, f'Kinase_mod_notinKD_{suffix}.txt')\n",
        "\n",
        "            with open(output_kd_file_path, 'w') as kd_outfile, open(output_not_in_kd_file_path, 'w') as notinkd_outfile:\n",
        "                header = 'Accession Number\\tGene ID\\tAbundance Ratio\\tModification\\tDomain\\tSecondary Structure\\tSolvent Accessibility\\tIUPred Score\\tT-loop\\tT-loop Sequence\\n'\n",
        "                kd_outfile.write(header)\n",
        "                notinkd_outfile.write(header)\n",
        "\n",
        "                sig_mods_df = pd.read_csv(sig_mods_path, delimiter='\\t')\n",
        "                total_modifications_kd = 0\n",
        "                total_modifications_notinkd = 0\n",
        "\n",
        "                for index, row in filtered_domains_df.iterrows():\n",
        "                    accession = row['Accession']\n",
        "                    domain_type = row['Domain Type']\n",
        "                    domain_start = row['Domain Start']\n",
        "                    domain_end = row['Domain End']\n",
        "\n",
        "                    def extract_residue_number(modification):\n",
        "                        if 'Met-loss' in modification:\n",
        "                            return 1\n",
        "                        try:\n",
        "                            return int(''.join(filter(str.isdigit, str(modification))))\n",
        "                        except ValueError:\n",
        "                            return None\n",
        "\n",
        "                    sig_mods_df['Residue Number'] = sig_mods_df['Modification'].apply(extract_residue_number)\n",
        "\n",
        "                    domain_mods_df = sig_mods_df[(sig_mods_df['Accession Number'] == accession) &\n",
        "                                                 (sig_mods_df['Residue Number'].notna()) &\n",
        "                                                 (sig_mods_df['Residue Number'] >= domain_start) &\n",
        "                                                 (sig_mods_df['Residue Number'] <= domain_end)]\n",
        "\n",
        "                    not_in_domain_mods_df = sig_mods_df[(sig_mods_df['Accession Number'] == accession) &\n",
        "                                                        (sig_mods_df['Residue Number'].notna()) &\n",
        "                                                        ((sig_mods_df['Residue Number'] < domain_start) |\n",
        "                                                         (sig_mods_df['Residue Number'] > domain_end))]\n",
        "\n",
        "                    if not domain_mods_df.empty:\n",
        "                        t_loop_start, t_loop_end, t_loop_seq = extract_t_loop_sequence(fasta_seq_dir, accession)\n",
        "\n",
        "                    # Count modifications in Kinase domains\n",
        "                    for mod_index, mod_row in domain_mods_df.iterrows():\n",
        "                        gene_id = mod_row['Gene ID']\n",
        "                        abundance_ratio = mod_row['Abundance Ratio (log2)']\n",
        "                        modification = mod_row['Modification']\n",
        "                        sec_structure = mod_row['Secondary Structure']\n",
        "                        sol_access = mod_row['Solvent Accessibility']\n",
        "                        residue_number = mod_row['Residue Number']\n",
        "                        iupred_score = extract_iupred_score(disorder_files_dir, accession, residue_number)\n",
        "\n",
        "                        t_loop_flag = 'No'\n",
        "                        t_loop_seq_mod = 'N/A'\n",
        "\n",
        "                        if t_loop_start is not None and t_loop_end is not None and t_loop_start <= residue_number <= t_loop_end:\n",
        "                            t_loop_flag = 'Yes'\n",
        "                            mod_position = residue_number - t_loop_start\n",
        "                            t_loop_seq_mod = t_loop_seq[:mod_position] + t_loop_seq[mod_position].lower() + t_loop_seq[mod_position + 1:]\n",
        "                            t_loop_identified += 1\n",
        "                            t_loop_sequences.append((accession, t_loop_seq_mod))\n",
        "\n",
        "                        kd_outfile.write(f'{accession}\\t{gene_id}\\t{abundance_ratio}\\t{modification}\\t{domain_type}\\t{sec_structure}\\t{sol_access}\\t{iupred_score}\\t{t_loop_flag}\\t{t_loop_seq_mod}\\n')\n",
        "                        total_modifications_kd += 1\n",
        "\n",
        "                    # Count modifications not in Kinase domains\n",
        "                    for mod_index, mod_row in not_in_domain_mods_df.iterrows():\n",
        "                        gene_id = mod_row['Gene ID']\n",
        "                        abundance_ratio = mod_row['Abundance Ratio (log2)']\n",
        "                        modification = mod_row['Modification']\n",
        "                        sec_structure = mod_row['Secondary Structure']\n",
        "                        sol_access = mod_row['Solvent Accessibility']\n",
        "                        residue_number = mod_row['Residue Number']\n",
        "                        iupred_score = extract_iupred_score(disorder_files_dir, accession, residue_number)\n",
        "\n",
        "                        notinkd_outfile.write(f'{accession}\\t{gene_id}\\t{abundance_ratio}\\t{modification}\\tNone\\t{sec_structure}\\t{sol_access}\\t{iupred_score}\\tNo\\tN/A\\n')\n",
        "                        total_modifications_notinkd += 1\n",
        "\n",
        "                        # Copy relevant plots to output directory\n",
        "                        for plot_type in ['disorder', 'sa']:\n",
        "                            plot_file = f'{accession}_{plot_type}.png'\n",
        "                            plot_path = os.path.join(f'/content/output/plots/{plot_type}_plots', plot_file)\n",
        "                            if os.path.exists(plot_path):\n",
        "                                shutil.copy(plot_path, output_plots_dir)\n",
        "\n",
        "            write_log(log_file, f\"Processed file: {file}\")\n",
        "            write_log(log_file, f\"Total modifications in Kinase domains: {total_modifications_kd}\")\n",
        "            write_log(log_file, f\"Total modifications to Kinase proteins, not in Kinase domains: {total_modifications_notinkd}\")\n",
        "            write_log(log_file, f\"Results written to: {output_kd_file_path}\")\n",
        "            write_log(log_file, f\"Results written to: {output_not_in_kd_file_path}\")\n",
        "\n",
        "    print(f\"Total T-loop sequences identified: {t_loop_identified}\")\n",
        "    write_log(log_file, f\"Total T-loop sequences identified: {t_loop_identified}\")\n",
        "    for accession, t_loop_seq_mod in t_loop_sequences:\n",
        "        print(f\"Accession {accession} T-loop sequence: {t_loop_seq_mod}\")\n",
        "        write_log(log_file, f\"Accession {accession} T-loop sequence: {t_loop_seq_mod}\")\n",
        "\n",
        "# Load domain library and filter for Kinase domains\n",
        "print(\"Reading domain library...\")\n",
        "domain_library_df = pd.read_csv(domain_library_path, delimiter='\\t')\n",
        "kinase_domains_pattern = re.compile(r'.*kinase.*', re.IGNORECASE)\n",
        "filtered_domains_df = domain_library_df[domain_library_df['Domain Type'].str.contains(kinase_domains_pattern)]\n",
        "print(f\"Total Kinase domains found: {len(filtered_domains_df)}\")\n",
        "\n",
        "# Run the function to extract modifications\n",
        "extract_modifications_in_kinase_domains(sig_mods_dir, filtered_domains_df, disorder_files_dir, fasta_seq_dir, output_text_dir, output_plots_dir)\n",
        "print(\"Process complete.\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "cellView": "form",
        "id": "xvk2SRYAFpRD",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "3010a124-277d-450f-e4b6-d36ee58eccc4"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Listing actual filenames in modlist_dir:\n",
            "domain_sig_modlist_sig_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "notsig_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "all_mods_ss_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "notsig_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "sig_mods_ss_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "sig_mods_ss_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "sig_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "all_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "domain_all_modlist_all_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "AdjacentSeq_all__(Aug_Active)___(Mar_Active).tsv\n",
            "all_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "sig_mods_ss_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "notsig_mods_ss_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "domain_sig_modlist_notsig_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "notsig_mods_ss_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "notsig_mods_ss_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "domain_sig_modlist_sig_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "domain_all_modlist_all_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "AdjacentSeq_all__(Torpor)___(Aug_Active).tsv\n",
            "AdjacentSeq_all__(Mar_Active)___(Torpor).tsv\n",
            "sig_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "all_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "all_mods_ss_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "phosphoprotein_isoforms.tsv\n",
            "domain_sig_modlist_sig_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "domain_sig_modlist_notsig_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "domain_all_modlist_all_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "domain_sig_modlist_notsig_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "all_mods_ss_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "sig_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "notsig_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "Processing file: /content/output/modlist_files/all_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "Number of significant accessions extracted from /content/output/modlist_files/all_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv: 1337\n",
            "All residues: 2697361 residues, mean IUPred score = 0.41\n",
            "Significant: 8955 modifications, mean IUPred score = 0.70\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:matplotlib.backends.backend_ps:The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Welch's ANOVA test result: T-statistic = -128.273, p-value = 0.000e+00\n",
            "Plot saved to /content/output/plots/iupred_histograms/iupred_histogram_all_modlist_Abundance_Ratio_log2__Torpor___Aug_Active.tsv.eps\n",
            "Processing file: /content/output/modlist_files/all_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "Number of significant accessions extracted from /content/output/modlist_files/all_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv: 15\n",
            "All residues: 2697361 residues, mean IUPred score = 0.41\n",
            "Significant: 94 modifications, mean IUPred score = 0.57\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:matplotlib.backends.backend_ps:The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Welch's ANOVA test result: T-statistic = -6.830, p-value = 8.641e-10\n",
            "Plot saved to /content/output/plots/iupred_histograms/iupred_histogram_all_modlist_Abundance_Ratio_log2__Aug_Active___Mar_Active.tsv.eps\n",
            "Processing file: /content/output/modlist_files/all_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "Number of significant accessions extracted from /content/output/modlist_files/all_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv: 1062\n",
            "All residues: 2697361 residues, mean IUPred score = 0.41\n",
            "Significant: 6928 modifications, mean IUPred score = 0.70\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:matplotlib.backends.backend_ps:The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Welch's ANOVA test result: T-statistic = -114.948, p-value = 0.000e+00\n",
            "Plot saved to /content/output/plots/iupred_histograms/iupred_histogram_all_modlist_Abundance_Ratio_log2__Mar_Active___Torpor.tsv.eps\n",
            "Processing complete.\n"
          ]
        }
      ],
      "source": [
        "#@title Histogram of IUPred score for significant modification vs all residues\n",
        "import os\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "from statsmodels.stats.weightstats import ttest_ind\n",
        "from matplotlib.ticker import PercentFormatter\n",
        "import numpy as np\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "# Define directories\n",
        "directory_to_iupred_files = \"/content/output/libraries/IUPred2A_disorder/\"\n",
        "modlist_dir = \"/content/output/modlist_files/\"\n",
        "output_figures_path = \"/content/output/plots/iupred_histograms\"\n",
        "logs_directory = \"/content/output/logs\"\n",
        "log_file = os.path.join(logs_directory, \"iupred_histogram_process.log\")\n",
        "\n",
        "# Ensure the output directory and logs directory exist\n",
        "os.makedirs(output_figures_path, exist_ok=True)\n",
        "os.makedirs(logs_directory, exist_ok=True)\n",
        "\n",
        "def read_iupred_scores(filepath):\n",
        "    \"\"\"Reads IUPred scores from a given IUPred file.\"\"\"\n",
        "    iupred_scores = {}\n",
        "    try:\n",
        "        with open(filepath, 'r') as file:\n",
        "            for line in file:\n",
        "                if line.startswith('#') or not line.strip():\n",
        "                    continue\n",
        "                parts = line.strip().split()\n",
        "                if len(parts) == 3:\n",
        "                    try:\n",
        "                        residue_number = int(parts[0])\n",
        "                        iupred_score = float(parts[2])\n",
        "                        iupred_scores[residue_number] = iupred_score\n",
        "                    except ValueError:\n",
        "                        continue  # Ignore conversion errors and continue\n",
        "    except Exception as e:\n",
        "        write_log(log_file, f\"Error reading IUPred scores from {filepath}: {e}\")\n",
        "    return iupred_scores\n",
        "\n",
        "def read_all_iupred_scores(directory):\n",
        "    \"\"\"Reads IUPred scores from files in the specified directory.\"\"\"\n",
        "    all_scores = []\n",
        "    residue_count = 0\n",
        "    for filename in os.listdir(directory):\n",
        "        if filename.endswith('.txt'):\n",
        "            filepath = os.path.join(directory, filename)\n",
        "            iupred_scores = read_iupred_scores(filepath)\n",
        "            all_scores.extend(iupred_scores.values())\n",
        "            residue_count += len(iupred_scores)\n",
        "    return all_scores, residue_count\n",
        "\n",
        "def read_modifications(filepath):\n",
        "    \"\"\"Reads modification information from the modlist file.\"\"\"\n",
        "    modifications = {}\n",
        "    significant_data = {}\n",
        "    not_significant_data = {}\n",
        "    significant_accessions = set()\n",
        "    try:\n",
        "        data = pd.read_csv(filepath, delimiter='\\t')  # Specify TSV delimiter\n",
        "        data.columns = data.columns.str.strip()\n",
        "\n",
        "        abundance_cols = [col for col in data.columns if 'Abundance Ratio (log2)' in col]\n",
        "        p_value_cols = [col for col in data.columns if 'Adj. P-Value' in col]\n",
        "\n",
        "        for p_col in p_value_cols:\n",
        "            corresponding_abundance_col = p_col.replace('Adj. P-Value', 'Abundance Ratio (log2)').strip()\n",
        "            if corresponding_abundance_col in abundance_cols:\n",
        "                timepoint = p_col.replace('Adj. P-Value', '').strip()\n",
        "                significant_data[timepoint] = []\n",
        "                for index, row in data.iterrows():\n",
        "                    try:\n",
        "                        mod = str(row['Modifications'])\n",
        "                        mod_residue = 1 if 'Met-loss' in mod else int(''.join(filter(str.isdigit, mod)))\n",
        "                        p_value = float(row[p_col])\n",
        "                        accession = row['Accession Number']\n",
        "                        if pd.notna(mod_residue):\n",
        "                            modifications.setdefault(accession, []).append(mod_residue)\n",
        "                            if p_value < 0.05:\n",
        "                                significant_data[timepoint].append((accession, mod_residue))\n",
        "                                significant_accessions.add(accession)\n",
        "                    except ValueError:\n",
        "                        continue  # Ignore conversion errors and continue\n",
        "            else:\n",
        "                write_log(log_file, f\"No matching abundance ratio column for {p_col} in file: {filepath}\")\n",
        "\n",
        "        write_log(log_file, f\"Number of significant accessions extracted from {filepath}: {len(significant_accessions)}\")\n",
        "    except Exception as e:\n",
        "        write_log(log_file, f\"Error reading modifications from {filepath}: {e}\")\n",
        "    return modifications, significant_data, significant_accessions\n",
        "\n",
        "def perform_welchs_anova(data1, data2, log_file):\n",
        "    \"\"\"Performs Welch's ANOVA test between two datasets.\"\"\"\n",
        "    if len(data1) < 2 or len(data2) < 2:\n",
        "        write_log(log_file, \"Not enough data to perform Welch's ANOVA test.\")\n",
        "        return float('nan')\n",
        "    t_stat, p_value, _ = ttest_ind(data1, data2, usevar='unequal')\n",
        "    write_log(log_file, f\"Welch's ANOVA test result: T-statistic = {t_stat:.3f}, p-value = {p_value:.3e}\")\n",
        "    return p_value\n",
        "\n",
        "def process_file(filename):\n",
        "    filepath = os.path.join(modlist_dir, filename)\n",
        "    write_log(log_file, f\"Processing file: {filepath}\")\n",
        "    if not os.path.exists(filepath):\n",
        "        write_log(log_file, f\"File does not exist: {filepath}\")\n",
        "        return\n",
        "\n",
        "    modifications, significant_data, significant_accessions = read_modifications(filepath)\n",
        "\n",
        "    if not significant_data:\n",
        "        write_log(log_file, f\"No significant data found in file: {filepath}\")\n",
        "        return\n",
        "\n",
        "    # Read all IUPred scores for significant accessions\n",
        "    all_scores, total_residues = read_all_iupred_scores(directory_to_iupred_files)\n",
        "\n",
        "    if not all_scores:\n",
        "        write_log(log_file, f\"No matching IUPred scores found in {directory_to_iupred_files}\")\n",
        "        return\n",
        "\n",
        "    significant_unstructured_scores = []\n",
        "\n",
        "    # Extract IUPred scores for significant modifications\n",
        "    for timepoint, mods in significant_data.items():\n",
        "        for accession, residue in mods:\n",
        "            iupred_filepath = os.path.join(directory_to_iupred_files, f\"{accession}_iupred2a.txt\")\n",
        "            iupred_scores = read_iupred_scores(iupred_filepath)\n",
        "            if residue in iupred_scores:\n",
        "                significant_unstructured_scores.append(iupred_scores[residue])\n",
        "\n",
        "    # Create plot with two different y-axes\n",
        "    fig, ax1 = plt.subplots(figsize=(10, 6))\n",
        "\n",
        "    # Histogram and KDE for all residues\n",
        "    sns.histplot(all_scores, bins=50, kde=True, color='gray', label='All Residues', ax=ax1)\n",
        "    all_mean = np.mean(all_scores)\n",
        "    write_log(log_file, f\"All residues: {total_residues} residues, mean IUPred score = {all_mean:.2f}\")\n",
        "\n",
        "    ax1.set_xlabel('IUPred Score', fontname='Liberation Sans', color='black')\n",
        "    ax1.set_ylabel('Frequency for All Residues', fontname='Liberation Sans', color='black')\n",
        "    ax1.tick_params(axis='y', labelcolor='black')\n",
        "    ax1.set_ylim(0, 140000)\n",
        "\n",
        "    # Create a secondary y-axis for significant residues\n",
        "    ax2 = ax1.twinx()\n",
        "    sns.histplot(significant_unstructured_scores, bins=50, kde=True, color='red', label='Significant Modifications', ax=ax2)\n",
        "    significant_mean = np.mean(significant_unstructured_scores)\n",
        "    write_log(log_file, f\"Significant: {len(significant_unstructured_scores)} modifications, mean IUPred score = {significant_mean:.2f}\")\n",
        "\n",
        "    ax2.set_ylabel('Frequency for Significant Modifications', fontname='Liberation Sans', color='red')\n",
        "    ax2.tick_params(axis='y', labelcolor='red')\n",
        "\n",
        "    # Adding a title\n",
        "    plt.title(f'Histogram of IUPred Scores: All Residues vs Significant Modifications\\n{filename}', fontname='Liberation Sans', color='black')\n",
        "\n",
        "    # Perform Welch's ANOVA tests and log results\n",
        "    p_value_sig_vs_all = perform_welchs_anova(all_scores, significant_unstructured_scores, log_file)\n",
        "\n",
        "    # Set y-axis limits based on the maximum value in the histograms\n",
        "    max_y_value = max(ax2.get_ylim()[1], 10)\n",
        "    ylim = 500 if max_y_value > 10 else 10\n",
        "    ax2.set_ylim(0, ylim)\n",
        "\n",
        "    # Save the figure\n",
        "    plot_filename = os.path.join(output_figures_path, f'iupred_histogram_{filename.replace(\" \", \"_\").replace(\"/\", \"_\").replace(\"(\", \"\").replace(\")\", \"\")}.eps')\n",
        "    plt.savefig(plot_filename)\n",
        "    plt.close()\n",
        "\n",
        "    write_log(log_file, f\"Plot saved to {plot_filename}\")\n",
        "\n",
        "# List the actual filenames in the directory\n",
        "write_log(log_file, \"Listing actual filenames in modlist_dir:\")\n",
        "actual_filenames = os.listdir(modlist_dir)\n",
        "for filename in actual_filenames:\n",
        "    write_log(log_file, filename)\n",
        "\n",
        "# Process each file with the all_modlist_ prefix\n",
        "for filename in actual_filenames:\n",
        "    if filename.startswith('all_modlist_'):\n",
        "        process_file(filename)\n",
        "\n",
        "write_log(log_file, \"Processing complete.\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "cellView": "form",
        "id": "RaZYm8RGv-TA",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "25ce9811-ce9d-472a-f118-d0de2f86e26e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Listing actual filenames in Modifications_dir:\n",
            "domain_sig_modlist_sig_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "notsig_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "all_mods_ss_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "notsig_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "sig_mods_ss_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "sig_mods_ss_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "sig_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "all_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "domain_all_modlist_all_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "AdjacentSeq_all__(Aug_Active)___(Mar_Active).tsv\n",
            "all_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "sig_mods_ss_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "notsig_mods_ss_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "domain_sig_modlist_notsig_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "notsig_mods_ss_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "notsig_mods_ss_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "domain_sig_modlist_sig_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "domain_all_modlist_all_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "AdjacentSeq_all__(Torpor)___(Aug_Active).tsv\n",
            "AdjacentSeq_all__(Mar_Active)___(Torpor).tsv\n",
            "sig_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "all_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "all_mods_ss_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "phosphoprotein_isoforms.tsv\n",
            "domain_sig_modlist_sig_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "domain_sig_modlist_notsig_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "domain_all_modlist_all_modlist_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "domain_sig_modlist_notsig_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "all_mods_ss_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "sig_modlist_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "notsig_modlist_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "Processing file: /content/output/modlist_files/all_mods_ss_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv\n",
            "Number of significant accessions extracted from /content/output/modlist_files/all_mods_ss_Abundance_Ratio_(log2)__(Aug_Active)___(Mar_Active).tsv: 15\n",
            "Total modifications: 16026\n",
            "Significant modifications: 94\n",
            "Not significant modifications: 15932\n",
            "All residues: 2349857 residues, mean solvent accessibility = 0.54\n",
            "Significant : 94 modifications, mean solvent accessibility = 0.69\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:matplotlib.backends.backend_ps:The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Welch's ANOVA test result: T-statistic = -2.713, p-value = 7.928e-03\n",
            "Welch's ANOVA between Significant and Not Significant : p-value = 7.928e-03\n",
            "Plot saved to /content/output/plots/sa_histograms/sa_compare_histogram_all_mods_ss_Abundance_Ratio_log2__Aug_Active___Mar_Active.tsv.eps\n",
            "Processing file: /content/output/modlist_files/all_mods_ss_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv\n",
            "Number of significant accessions extracted from /content/output/modlist_files/all_mods_ss_Abundance_Ratio_(log2)__(Torpor)___(Aug_Active).tsv: 1286\n",
            "Total modifications: 16026\n",
            "Significant modifications: 7716\n",
            "Not significant modifications: 8310\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:matplotlib.backends.backend_ps:The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "All residues: 2349857 residues, mean solvent accessibility = 0.54\n",
            "Significant : 7716 modifications, mean solvent accessibility = 0.76\n",
            "Welch's ANOVA test result: T-statistic = 5.408, p-value = 6.477e-08\n",
            "Welch's ANOVA between Significant and Not Significant : p-value = 6.477e-08\n",
            "Plot saved to /content/output/plots/sa_histograms/sa_compare_histogram_all_mods_ss_Abundance_Ratio_log2__Torpor___Aug_Active.tsv.eps\n",
            "Processing file: /content/output/modlist_files/all_mods_ss_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv\n",
            "Number of significant accessions extracted from /content/output/modlist_files/all_mods_ss_Abundance_Ratio_(log2)__(Mar_Active)___(Torpor).tsv: 1027\n",
            "Total modifications: 16026\n",
            "Significant modifications: 6075\n",
            "Not significant modifications: 9951\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:matplotlib.backends.backend_ps:The PostScript backend does not support transparency; partially transparent artists will be rendered opaque.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "All residues: 2349857 residues, mean solvent accessibility = 0.54\n",
            "Significant : 6075 modifications, mean solvent accessibility = 0.76\n",
            "Welch's ANOVA test result: T-statistic = 4.164, p-value = 3.147e-05\n",
            "Welch's ANOVA between Significant and Not Significant : p-value = 3.147e-05\n",
            "Plot saved to /content/output/plots/sa_histograms/sa_compare_histogram_all_mods_ss_Abundance_Ratio_log2__Mar_Active___Torpor.tsv.eps\n",
            "Processing complete.\n"
          ]
        }
      ],
      "source": [
        "#@title Solvent accessibility histogram for significant, not significant, and all mods\n",
        "\n",
        "import os\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "from statsmodels.stats.weightstats import ttest_ind\n",
        "from matplotlib.ticker import PercentFormatter\n",
        "import numpy as np\n",
        "\n",
        "def write_log(log_file, message):\n",
        "    \"\"\"Writes a message to the log file.\"\"\"\n",
        "    with open(log_file, 'a') as f:\n",
        "        f.write(message + '\\n')\n",
        "    print(message)  # Also print to console\n",
        "\n",
        "# Define directories\n",
        "directory_to_residue_files = \"/content/output/libraries/all_protein_ss/\"\n",
        "Modifications_dir = \"/content/output/modlist_files/\"\n",
        "output_figures_path = \"/content/output/plots/sa_histograms\"\n",
        "logs_directory = \"/content/output/logs\"\n",
        "log_file = os.path.join(logs_directory, \"sa_histogram_process.log\")\n",
        "\n",
        "# Ensure the output directory and logs directory exist\n",
        "os.makedirs(output_figures_path, exist_ok=True)\n",
        "os.makedirs(logs_directory, exist_ok=True)\n",
        "\n",
        "def read_all_residue_accessibility(directory):\n",
        "    \"\"\"Reads solvent accessibility from files in the specified directory.\"\"\"\n",
        "    all_accessibility = []\n",
        "    residue_count = 0\n",
        "    for filename in os.listdir(directory):\n",
        "        if filename.endswith('_ss.tsv'):\n",
        "            filepath = os.path.join(directory, filename)\n",
        "            if os.path.exists(filepath):\n",
        "                with open(filepath, 'r') as file:\n",
        "                    for line in file:\n",
        "                        parts = line.strip().split('\\t')  # Split using tab delimiter for TSV files\n",
        "                        if len(parts) > 1:  # Ensure there's enough parts in the line\n",
        "                            try:\n",
        "                                accessibility = float(parts[-1])\n",
        "                                all_accessibility.append(accessibility)\n",
        "                                residue_count += 1\n",
        "                            except ValueError:\n",
        "                                continue  # Ignore conversion errors and continue\n",
        "            else:\n",
        "                write_log(log_file, f\"File {filepath} not found\")\n",
        "    return all_accessibility, residue_count\n",
        "\n",
        "def read_Modification_accessibility(filepath):\n",
        "    \"\"\"Reads solvent accessibility specifically for Modifications, separated by significance.\"\"\"\n",
        "    significant_data = {}\n",
        "    not_significant_data = {}\n",
        "    significant_accessions = set()\n",
        "    total_modifications = 0\n",
        "    try:\n",
        "        data = pd.read_csv(filepath, delimiter='\\t')  # Specify TSV delimiter\n",
        "        data.columns = data.columns.str.strip()\n",
        "\n",
        "        abundance_cols = [col for col in data.columns if 'Abundance Ratio (log2)' in col]\n",
        "        p_value_cols = [col for col in data.columns if 'Adj. P-Value' in col]\n",
        "\n",
        "        for p_col in p_value_cols:\n",
        "            corresponding_abundance_col = p_col.replace('Adj. P-Value', 'Abundance Ratio (log2)').strip()\n",
        "            if corresponding_abundance_col in abundance_cols:\n",
        "                timepoint = p_col.replace('Adj. P-Value', '').strip()\n",
        "                significant_data[timepoint] = []\n",
        "                not_significant_data[timepoint] = []\n",
        "                for index, row in data.iterrows():\n",
        "                    try:\n",
        "                        accessibility = float(row['Solvent Accessibility'])\n",
        "                        p_value = float(row[p_col])\n",
        "                        total_modifications += 1\n",
        "                        if p_value < 0.05:\n",
        "                            significant_data[timepoint].append(accessibility)\n",
        "                            significant_accessions.add(row['Accession Number'])\n",
        "                        else:\n",
        "                            not_significant_data[timepoint].append(accessibility)\n",
        "                    except ValueError:\n",
        "                        continue  # Ignore conversion errors and continue\n",
        "            else:\n",
        "                write_log(log_file, f\"No matching abundance ratio column for {p_col} in file: {filepath}\")\n",
        "\n",
        "        write_log(log_file, f\"Number of significant accessions extracted from {filepath}: {len(significant_accessions)}\")\n",
        "    except Exception as e:\n",
        "        write_log(log_file, f\"Error reading Modification accessibility from {filepath}: {e}\")\n",
        "    return significant_data, not_significant_data, significant_accessions, total_modifications\n",
        "\n",
        "def perform_welchs_anova(data1, data2, log_file):\n",
        "    \"\"\"Performs Welch's ANOVA test between two datasets.\"\"\"\n",
        "    if len(data1) < 2 or len(data2) < 2:\n",
        "        write_log(log_file, \"Not enough data to perform Welch's ANOVA test.\")\n",
        "        return float('nan')\n",
        "    t_stat, p_value, _ = ttest_ind(data1, data2, usevar='unequal')\n",
        "    write_log(log_file, f\"Welch's ANOVA test result: T-statistic = {t_stat:.3f}, p-value = {p_value:.3e}\")\n",
        "    return p_value\n",
        "\n",
        "def process_file(filename):\n",
        "    filepath = os.path.join(Modifications_dir, filename)\n",
        "    write_log(log_file, f\"Processing file: {filepath}\")\n",
        "    if not os.path.exists(filepath):\n",
        "        write_log(log_file, f\"File does not exist: {filepath}\")\n",
        "        return\n",
        "\n",
        "    significant_data, not_significant_data, significant_accessions, total_modifications = read_Modification_accessibility(filepath)\n",
        "\n",
        "    if not significant_data and not not_significant_data:\n",
        "        write_log(log_file, f\"No significant or not significant data found in file: {filepath}\")\n",
        "        return\n",
        "\n",
        "    # Read all residue accessibility data for significant accessions\n",
        "    all_accessibility, total_residues = read_all_residue_accessibility(directory_to_residue_files)\n",
        "\n",
        "    if not all_accessibility:\n",
        "        write_log(log_file, f\"No matching residues found in {directory_to_residue_files}\")\n",
        "        return\n",
        "\n",
        "    # Check if the sum of significant and not significant modifications equals total modifications\n",
        "    sig_mod_count = sum(len(data) for data in significant_data.values())\n",
        "    not_sig_mod_count = sum(len(data) for data in not_significant_data.values())\n",
        "    write_log(log_file, f\"Total modifications: {total_modifications}\")\n",
        "    write_log(log_file, f\"Significant modifications: {sig_mod_count}\")\n",
        "    write_log(log_file, f\"Not significant modifications: {not_sig_mod_count}\")\n",
        "\n",
        "    if sig_mod_count + not_sig_mod_count != total_modifications:\n",
        "        write_log(log_file, \"Warning: Significant and not significant modifications do not sum up to total modifications!\")\n",
        "\n",
        "    # Create plot with two different y-axes\n",
        "    fig, ax1 = plt.subplots()\n",
        "\n",
        "    # Histogram and KDE for all residues\n",
        "    sns.histplot(all_accessibility, bins=50, kde=True, color='gray', label='All Residues', ax=ax1)\n",
        "    all_mean = np.mean(all_accessibility)\n",
        "    write_log(log_file, f\"All residues: {len(all_accessibility)} residues, mean solvent accessibility = {all_mean:.2f}\")\n",
        "\n",
        "    ax1.set_xlabel('Solvent Accessibility', fontname='Liberation Sans', color='black')\n",
        "    ax1.set_ylabel('Frequency for All Residues', fontname='Liberation Sans', color='black')\n",
        "    ax1.tick_params(axis='y', labelcolor='black')\n",
        "    ax1.set_ylim(0, 250000)\n",
        "\n",
        "    # Create a secondary y-axis for significant residues\n",
        "    ax2 = ax1.twinx()\n",
        "    significant_unstructured_residues = []\n",
        "\n",
        "    for idx, (timepoint, data) in enumerate(significant_data.items()):\n",
        "        sns.histplot(data, bins=50, kde=True, color='red', label=f'Significant Modifications {timepoint}', ax=ax2)\n",
        "        mean = np.mean(data)\n",
        "        write_log(log_file, f\"Significant {timepoint}: {len(data)} modifications, mean solvent accessibility = {mean:.2f}\")\n",
        "        significant_unstructured_residues.extend(data)\n",
        "\n",
        "    ax2.set_ylabel('Frequency for Significant Modifications', fontname='Liberation Sans', color='red')\n",
        "    ax2.tick_params(axis='y', labelcolor='red')\n",
        "\n",
        "    # Set y-axis limits for significant modifications based on the maximum value\n",
        "    max_y_value = max(ax2.get_ylim()[1], 10)\n",
        "    ylim = 700 if max_y_value > 10 else 10\n",
        "    ax2.set_ylim(0, ylim)\n",
        "\n",
        "    # Adding a title\n",
        "    plt.title(f'Histogram of Solvent Accessibility: All Residues vs Modifications\\n{filename}', fontname='Liberation Sans', color='black')\n",
        "\n",
        "    # Perform Welch's ANOVA tests and log results\n",
        "    for timepoint, sig_data in significant_data.items():\n",
        "        if timepoint in not_significant_data:\n",
        "            not_sig_data = not_significant_data[timepoint]\n",
        "            p_value_sig_vs_not_sig = perform_welchs_anova(sig_data, not_sig_data, log_file)\n",
        "            write_log(log_file, f\"Welch's ANOVA between Significant and Not Significant {timepoint}: p-value = {p_value_sig_vs_not_sig:.3e}\")\n",
        "\n",
        "    # Saving the figure\n",
        "    plot_filename = os.path.join(output_figures_path, f'sa_compare_histogram_{filename.replace(\" \", \"_\").replace(\"/\", \"_\").replace(\"(\", \"\").replace(\")\", \"\")}.eps')\n",
        "    plt.savefig(plot_filename)\n",
        "    plt.close()\n",
        "\n",
        "    write_log(log_file, f\"Plot saved to {plot_filename}\")\n",
        "\n",
        "# List the actual filenames in the directory\n",
        "write_log(log_file, \"Listing actual filenames in Modifications_dir:\")\n",
        "actual_filenames = os.listdir(Modifications_dir)\n",
        "for filename in actual_filenames:\n",
        "    write_log(log_file, filename)\n",
        "\n",
        "# Process each file with the all_mods_ss_Abundance prefix\n",
        "for filename in actual_filenames:\n",
        "    if filename.startswith('all_mods_ss_Abundance'):\n",
        "        process_file(filename)\n",
        "\n",
        "write_log(log_file, \"Processing complete.\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VeXRQ7rDWgs0"
      },
      "source": [
        "# **Zip and download data**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "id": "33h9oBL5A0UF",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        },
        "outputId": "f98e3e99-fd33-46e4-da83-3f171c115623"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Downloading zip file: /content/07262024_all_proteins_output.zip\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_e0905fcb-195d-4cef-a2a6-0d11d4106ce4\", \"07262024_all_proteins_output.zip\", 516074421)"
            ]
          },
          "metadata": {}
        }
      ],
      "source": [
        "import os\n",
        "import zipfile\n",
        "from google.colab import files\n",
        "\n",
        "def find_first_csv(directory):\n",
        "    \"\"\"Finds the first CSV file in the given directory and returns its full path.\"\"\"\n",
        "    for file in os.listdir(directory):\n",
        "        if file.endswith('.csv'):\n",
        "            return os.path.join(directory, file)\n",
        "    return None\n",
        "\n",
        "def zip_directory(directory_to_zip, zip_output_path):\n",
        "    \"\"\"Create a ZIP file containing all files in the specified directory.\"\"\"\n",
        "    with zipfile.ZipFile(zip_output_path, 'w', zipfile.ZIP_DEFLATED) as zipf:\n",
        "        # Walk through the directory\n",
        "        for root, dirs, files in os.walk(directory_to_zip):\n",
        "            for file in files:\n",
        "                # Create the full path to the file\n",
        "                file_path = os.path.join(root, file)\n",
        "                # Arcname strips the leading part of the path to avoid excessive folder structure in zip\n",
        "                arcname = os.path.relpath(file_path, directory_to_zip)\n",
        "                zipf.write(file_path, arcname=arcname)\n",
        "\n",
        "def main():\n",
        "    # Define the path to the directory you want to zip and where to find the CSV file\n",
        "    directory_to_zip = \"/content/output\"  # Assuming you want to zip everything in /content/output\n",
        "    directory_with_csv = \"/content/input\"  # Assuming CSV is in /content/input\n",
        "\n",
        "    # Find the first CSV file to determine the name of the zip file\n",
        "    csv_file_path = find_first_csv(directory_with_csv)\n",
        "    if csv_file_path is None:\n",
        "        print(\"No CSV file found.\")\n",
        "        return\n",
        "\n",
        "    csv_basename = os.path.splitext(os.path.basename(csv_file_path))[0]  # Remove extension\n",
        "    zip_filename = f\"{csv_basename}_output.zip\"  # Append 'output' to the CSV base name\n",
        "    zip_output_path = os.path.join(\"/content\", zip_filename)\n",
        "\n",
        "    # Create a ZIP file\n",
        "    zip_directory(directory_to_zip, zip_output_path)\n",
        "\n",
        "    # Download the zip file, which triggers the browser's save dialog in Google Colab\n",
        "    print(f\"Downloading zip file: {zip_output_path}\")\n",
        "    files.download(zip_output_path)\n",
        "\n",
        "if __name__ == '__main__':\n",
        "    main()\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyPJnGiMFdpzrERSTfEN/WCD",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}